// @injectivelabs/core-proto-ts-v2/esm/generated/injective/evm/v1/chain_config_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class ChainConfig$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.ChainConfig", [
      { no: 1, name: "homestead_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"homestead_block"' } },
      { no: 2, name: "dao_fork_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "DAOForkBlock", "gogoproto.moretags": 'yaml:"dao_fork_block"' } },
      { no: 3, name: "dao_fork_support", kind: "scalar", T: 8, options: { "gogoproto.customname": "DAOForkSupport", "gogoproto.moretags": 'yaml:"dao_fork_support"' } },
      { no: 4, name: "eip150_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "EIP150Block", "gogoproto.moretags": 'yaml:"eip150_block"' } },
      { no: 5, name: "eip150_hash", kind: "scalar", T: 9, options: { "gogoproto.customname": "EIP150Hash", "gogoproto.moretags": 'yaml:"byzantium_block"' } },
      { no: 6, name: "eip155_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "EIP155Block", "gogoproto.moretags": 'yaml:"eip155_block"' } },
      { no: 7, name: "eip158_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "EIP158Block", "gogoproto.moretags": 'yaml:"eip158_block"' } },
      { no: 8, name: "byzantium_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"byzantium_block"' } },
      { no: 9, name: "constantinople_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"constantinople_block"' } },
      { no: 10, name: "petersburg_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"petersburg_block"' } },
      { no: 11, name: "istanbul_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"istanbul_block"' } },
      { no: 12, name: "muir_glacier_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"muir_glacier_block"' } },
      { no: 13, name: "berlin_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"berlin_block"' } },
      { no: 17, name: "london_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"london_block"' } },
      { no: 18, name: "arrow_glacier_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"arrow_glacier_block"' } },
      { no: 20, name: "gray_glacier_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"gray_glacier_block"' } },
      { no: 21, name: "merge_netsplit_block", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"merge_netsplit_block"' } },
      { no: 22, name: "shanghai_time", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"shanghai_time"' } },
      { no: 23, name: "cancun_time", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"cancun_time"' } },
      { no: 24, name: "prague_time", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.moretags": 'yaml:"prague_time"' } },
      { no: 25, name: "eip155_chain_id", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "EIP155ChainID", "gogoproto.jsontag": "eip155ChainID" } },
      { no: 26, name: "blob_schedule_config", kind: "message", T: () => BlobScheduleConfig, options: { "gogoproto.moretags": 'yaml:"blob_schedule_config"' } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.homesteadBlock = "";
    message.daoForkBlock = "";
    message.daoForkSupport = false;
    message.eip150Block = "";
    message.eip150Hash = "";
    message.eip155Block = "";
    message.eip158Block = "";
    message.byzantiumBlock = "";
    message.constantinopleBlock = "";
    message.petersburgBlock = "";
    message.istanbulBlock = "";
    message.muirGlacierBlock = "";
    message.berlinBlock = "";
    message.londonBlock = "";
    message.arrowGlacierBlock = "";
    message.grayGlacierBlock = "";
    message.mergeNetsplitBlock = "";
    message.shanghaiTime = "";
    message.cancunTime = "";
    message.pragueTime = "";
    message.eip155ChainId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string homestead_block */
        1:
          message.homesteadBlock = reader.string();
          break;
        case /* string dao_fork_block */
        2:
          message.daoForkBlock = reader.string();
          break;
        case /* bool dao_fork_support */
        3:
          message.daoForkSupport = reader.bool();
          break;
        case /* string eip150_block */
        4:
          message.eip150Block = reader.string();
          break;
        case /* string eip150_hash */
        5:
          message.eip150Hash = reader.string();
          break;
        case /* string eip155_block */
        6:
          message.eip155Block = reader.string();
          break;
        case /* string eip158_block */
        7:
          message.eip158Block = reader.string();
          break;
        case /* string byzantium_block */
        8:
          message.byzantiumBlock = reader.string();
          break;
        case /* string constantinople_block */
        9:
          message.constantinopleBlock = reader.string();
          break;
        case /* string petersburg_block */
        10:
          message.petersburgBlock = reader.string();
          break;
        case /* string istanbul_block */
        11:
          message.istanbulBlock = reader.string();
          break;
        case /* string muir_glacier_block */
        12:
          message.muirGlacierBlock = reader.string();
          break;
        case /* string berlin_block */
        13:
          message.berlinBlock = reader.string();
          break;
        case /* string london_block */
        17:
          message.londonBlock = reader.string();
          break;
        case /* string arrow_glacier_block */
        18:
          message.arrowGlacierBlock = reader.string();
          break;
        case /* string gray_glacier_block */
        20:
          message.grayGlacierBlock = reader.string();
          break;
        case /* string merge_netsplit_block */
        21:
          message.mergeNetsplitBlock = reader.string();
          break;
        case /* string shanghai_time */
        22:
          message.shanghaiTime = reader.string();
          break;
        case /* string cancun_time */
        23:
          message.cancunTime = reader.string();
          break;
        case /* string prague_time */
        24:
          message.pragueTime = reader.string();
          break;
        case /* string eip155_chain_id */
        25:
          message.eip155ChainId = reader.string();
          break;
        case /* injective.evm.v1.BlobScheduleConfig blob_schedule_config */
        26:
          message.blobScheduleConfig = BlobScheduleConfig.internalBinaryRead(reader, reader.uint32(), options, message.blobScheduleConfig);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.homesteadBlock !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.homesteadBlock);
    if (message.daoForkBlock !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.daoForkBlock);
    if (message.daoForkSupport !== false)
      writer.tag(3, WireType.Varint).bool(message.daoForkSupport);
    if (message.eip150Block !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.eip150Block);
    if (message.eip150Hash !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.eip150Hash);
    if (message.eip155Block !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.eip155Block);
    if (message.eip158Block !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.eip158Block);
    if (message.byzantiumBlock !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.byzantiumBlock);
    if (message.constantinopleBlock !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.constantinopleBlock);
    if (message.petersburgBlock !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.petersburgBlock);
    if (message.istanbulBlock !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.istanbulBlock);
    if (message.muirGlacierBlock !== "")
      writer.tag(12, WireType.LengthDelimited).string(message.muirGlacierBlock);
    if (message.berlinBlock !== "")
      writer.tag(13, WireType.LengthDelimited).string(message.berlinBlock);
    if (message.londonBlock !== "")
      writer.tag(17, WireType.LengthDelimited).string(message.londonBlock);
    if (message.arrowGlacierBlock !== "")
      writer.tag(18, WireType.LengthDelimited).string(message.arrowGlacierBlock);
    if (message.grayGlacierBlock !== "")
      writer.tag(20, WireType.LengthDelimited).string(message.grayGlacierBlock);
    if (message.mergeNetsplitBlock !== "")
      writer.tag(21, WireType.LengthDelimited).string(message.mergeNetsplitBlock);
    if (message.shanghaiTime !== "")
      writer.tag(22, WireType.LengthDelimited).string(message.shanghaiTime);
    if (message.cancunTime !== "")
      writer.tag(23, WireType.LengthDelimited).string(message.cancunTime);
    if (message.pragueTime !== "")
      writer.tag(24, WireType.LengthDelimited).string(message.pragueTime);
    if (message.eip155ChainId !== "")
      writer.tag(25, WireType.LengthDelimited).string(message.eip155ChainId);
    if (message.blobScheduleConfig)
      BlobScheduleConfig.internalBinaryWrite(message.blobScheduleConfig, writer.tag(26, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ChainConfig = new ChainConfig$Type();
class BlobScheduleConfig$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.BlobScheduleConfig", [
      { no: 1, name: "cancun", kind: "message", T: () => BlobConfig },
      { no: 2, name: "prague", kind: "message", T: () => BlobConfig },
      { no: 3, name: "osaka", kind: "message", T: () => BlobConfig },
      { no: 4, name: "verkle", kind: "message", T: () => BlobConfig }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.evm.v1.BlobConfig cancun */
        1:
          message.cancun = BlobConfig.internalBinaryRead(reader, reader.uint32(), options, message.cancun);
          break;
        case /* injective.evm.v1.BlobConfig prague */
        2:
          message.prague = BlobConfig.internalBinaryRead(reader, reader.uint32(), options, message.prague);
          break;
        case /* injective.evm.v1.BlobConfig osaka */
        3:
          message.osaka = BlobConfig.internalBinaryRead(reader, reader.uint32(), options, message.osaka);
          break;
        case /* injective.evm.v1.BlobConfig verkle */
        4:
          message.verkle = BlobConfig.internalBinaryRead(reader, reader.uint32(), options, message.verkle);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.cancun)
      BlobConfig.internalBinaryWrite(message.cancun, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.prague)
      BlobConfig.internalBinaryWrite(message.prague, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.osaka)
      BlobConfig.internalBinaryWrite(message.osaka, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.verkle)
      BlobConfig.internalBinaryWrite(message.verkle, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BlobScheduleConfig = new BlobScheduleConfig$Type();
class BlobConfig$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.BlobConfig", [
      {
        no: 1,
        name: "target",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "max",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "base_fee_update_fraction", kind: "scalar", T: 4, L: 0, options: { "gogoproto.moretags": 'yaml:"base_fee_update_fraction"' } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.target = 0n;
    message.max = 0n;
    message.baseFeeUpdateFraction = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 target */
        1:
          message.target = reader.uint64().toBigInt();
          break;
        case /* uint64 max */
        2:
          message.max = reader.uint64().toBigInt();
          break;
        case /* uint64 base_fee_update_fraction */
        3:
          message.baseFeeUpdateFraction = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.target !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.target);
    if (message.max !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.max);
    if (message.baseFeeUpdateFraction !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.baseFeeUpdateFraction);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BlobConfig = new BlobConfig$Type();
export {
  BlobConfig,
  BlobScheduleConfig,
  ChainConfig
};
