// @injectivelabs/core-proto-ts-v2/esm/generated/injective/evm/v1/events_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class EventEthereumTx$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.EventEthereumTx", [
      {
        no: 1,
        name: "amount",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "eth_hash",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "index",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "gas_used",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "hash",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "recipient",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "eth_tx_failed",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.amount = "";
    message.ethHash = "";
    message.index = "";
    message.gasUsed = "";
    message.hash = "";
    message.recipient = "";
    message.ethTxFailed = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string amount */
        1:
          message.amount = reader.string();
          break;
        case /* string eth_hash */
        2:
          message.ethHash = reader.string();
          break;
        case /* string index */
        3:
          message.index = reader.string();
          break;
        case /* string gas_used */
        4:
          message.gasUsed = reader.string();
          break;
        case /* string hash */
        5:
          message.hash = reader.string();
          break;
        case /* string recipient */
        6:
          message.recipient = reader.string();
          break;
        case /* string eth_tx_failed */
        7:
          message.ethTxFailed = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.amount !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.amount);
    if (message.ethHash !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.ethHash);
    if (message.index !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.index);
    if (message.gasUsed !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.gasUsed);
    if (message.hash !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.hash);
    if (message.recipient !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.recipient);
    if (message.ethTxFailed !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.ethTxFailed);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventEthereumTx = new EventEthereumTx$Type();
class EventTxLog$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.EventTxLog", [
      {
        no: 1,
        name: "tx_logs",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.txLogs = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string tx_logs */
        1:
          message.txLogs.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.txLogs.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.txLogs[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventTxLog = new EventTxLog$Type();
class EventMessage$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.EventMessage", [
      {
        no: 1,
        name: "module",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "tx_type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.module = "";
    message.sender = "";
    message.txType = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string module */
        1:
          message.module = reader.string();
          break;
        case /* string sender */
        2:
          message.sender = reader.string();
          break;
        case /* string tx_type */
        3:
          message.txType = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.module !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.module);
    if (message.sender !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.sender);
    if (message.txType !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.txType);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventMessage = new EventMessage$Type();
class EventBlockBloom$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.EventBlockBloom", [
      {
        no: 1,
        name: "bloom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.bloom = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string bloom */
        1:
          message.bloom = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.bloom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.bloom);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventBlockBloom = new EventBlockBloom$Type();
export {
  EventBlockBloom,
  EventEthereumTx,
  EventMessage,
  EventTxLog
};
