// @injectivelabs/core-proto-ts-v2/esm/generated/injective/evm/v1/log_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Log$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.Log", [
      {
        no: 1,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "topics",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 4, name: "block_number", kind: "scalar", T: 4, L: 0, options: { "gogoproto.jsontag": "blockNumber" } },
      { no: 5, name: "tx_hash", kind: "scalar", T: 9, options: { "gogoproto.jsontag": "transactionHash" } },
      { no: 6, name: "tx_index", kind: "scalar", T: 4, L: 0, options: { "gogoproto.jsontag": "transactionIndex" } },
      { no: 7, name: "block_hash", kind: "scalar", T: 9, options: { "gogoproto.jsontag": "blockHash" } },
      { no: 8, name: "index", kind: "scalar", T: 4, L: 0, options: { "gogoproto.jsontag": "logIndex" } },
      {
        no: 9,
        name: "removed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.topics = [];
    message.data = new Uint8Array(0);
    message.blockNumber = 0n;
    message.txHash = "";
    message.txIndex = 0n;
    message.blockHash = "";
    message.index = 0n;
    message.removed = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* repeated string topics */
        2:
          message.topics.push(reader.string());
          break;
        case /* bytes data */
        3:
          message.data = reader.bytes();
          break;
        case /* uint64 block_number */
        4:
          message.blockNumber = reader.uint64().toBigInt();
          break;
        case /* string tx_hash */
        5:
          message.txHash = reader.string();
          break;
        case /* uint64 tx_index */
        6:
          message.txIndex = reader.uint64().toBigInt();
          break;
        case /* string block_hash */
        7:
          message.blockHash = reader.string();
          break;
        case /* uint64 index */
        8:
          message.index = reader.uint64().toBigInt();
          break;
        case /* bool removed */
        9:
          message.removed = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    for (let i = 0; i < message.topics.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.topics[i]);
    if (message.data.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.data);
    if (message.blockNumber !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.blockNumber);
    if (message.txHash !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.txHash);
    if (message.txIndex !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.txIndex);
    if (message.blockHash !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.blockHash);
    if (message.index !== 0n)
      writer.tag(8, WireType.Varint).uint64(message.index);
    if (message.removed !== false)
      writer.tag(9, WireType.Varint).bool(message.removed);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Log = new Log$Type();
export {
  Log
};
