// @injectivelabs/core-proto-ts-v2/esm/generated/injective/evm/v1/params_pb.js
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { WireType } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ChainConfig } from "./chain_config_pb.js";
class Params$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.Params", [
      { no: 1, name: "evm_denom", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"evm_denom"' } },
      { no: 2, name: "enable_create", kind: "scalar", T: 8, options: { "gogoproto.moretags": 'yaml:"enable_create"' } },
      { no: 3, name: "enable_call", kind: "scalar", T: 8, options: { "gogoproto.moretags": 'yaml:"enable_call"' } },
      { no: 4, name: "extra_eips", kind: "scalar", repeat: 1, T: 3, L: 0, options: { "gogoproto.customname": "ExtraEIPs", "gogoproto.moretags": 'yaml:"extra_eips"' } },
      { no: 5, name: "chain_config", kind: "message", T: () => ChainConfig, options: { "gogoproto.nullable": false } },
      {
        no: 6,
        name: "allow_unprotected_txs",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 7,
        name: "authorized_deployers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "permissioned",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.evmDenom = "";
    message.enableCreate = false;
    message.enableCall = false;
    message.extraEips = [];
    message.allowUnprotectedTxs = false;
    message.authorizedDeployers = [];
    message.permissioned = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string evm_denom */
        1:
          message.evmDenom = reader.string();
          break;
        case /* bool enable_create */
        2:
          message.enableCreate = reader.bool();
          break;
        case /* bool enable_call */
        3:
          message.enableCall = reader.bool();
          break;
        case /* repeated int64 extra_eips */
        4:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.extraEips.push(reader.int64().toBigInt());
          else
            message.extraEips.push(reader.int64().toBigInt());
          break;
        case /* injective.evm.v1.ChainConfig chain_config */
        5:
          message.chainConfig = ChainConfig.internalBinaryRead(reader, reader.uint32(), options, message.chainConfig);
          break;
        case /* bool allow_unprotected_txs */
        6:
          message.allowUnprotectedTxs = reader.bool();
          break;
        case /* repeated string authorized_deployers */
        7:
          message.authorizedDeployers.push(reader.string());
          break;
        case /* bool permissioned */
        8:
          message.permissioned = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.evmDenom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.evmDenom);
    if (message.enableCreate !== false)
      writer.tag(2, WireType.Varint).bool(message.enableCreate);
    if (message.enableCall !== false)
      writer.tag(3, WireType.Varint).bool(message.enableCall);
    if (message.extraEips.length) {
      writer.tag(4, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.extraEips.length; i++)
        writer.int64(message.extraEips[i]);
      writer.join();
    }
    if (message.chainConfig)
      ChainConfig.internalBinaryWrite(message.chainConfig, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.allowUnprotectedTxs !== false)
      writer.tag(6, WireType.Varint).bool(message.allowUnprotectedTxs);
    for (let i = 0; i < message.authorizedDeployers.length; i++)
      writer.tag(7, WireType.LengthDelimited).string(message.authorizedDeployers[i]);
    if (message.permissioned !== false)
      writer.tag(8, WireType.Varint).bool(message.permissioned);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
export {
  Params
};
