// @injectivelabs/core-proto-ts-v2/esm/generated/injective/evm/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Account queries an Ethereum account.
   *
   * @generated from protobuf rpc: Account
   */
  account(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * CosmosAccount queries an Ethereum account's Cosmos Address.
   *
   * @generated from protobuf rpc: CosmosAccount
   */
  cosmosAccount(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ValidatorAccount queries an Ethereum account's from a validator consensus
   * Address.
   *
   * @generated from protobuf rpc: ValidatorAccount
   */
  validatorAccount(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Balance queries the balance of a the EVM denomination for a single
   * EthAccount.
   *
   * @generated from protobuf rpc: Balance
   */
  balance(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Storage queries the balance of all coins for a single account.
   *
   * @generated from protobuf rpc: Storage
   */
  storage(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Code queries the code of a single account.
   *
   * @generated from protobuf rpc: Code
   */
  code(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Params queries the parameters of x/evm module.
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * EthCall implements the `eth_call` rpc api
   *
   * @generated from protobuf rpc: EthCall
   */
  ethCall(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * EstimateGas implements the `eth_estimateGas` rpc api
   *
   * @generated from protobuf rpc: EstimateGas
   */
  estimateGas(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TraceTx implements the `debug_traceTransaction` rpc api
   *
   * @generated from protobuf rpc: TraceTx
   */
  traceTx(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TraceBlock implements the `debug_traceBlockByNumber` and
   * `debug_traceBlockByHash` rpc api
   *
   * @generated from protobuf rpc: TraceBlock
   */
  traceBlock(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TraceCall implements the `debug_traceCall` rpc api
   *
   * @generated from protobuf rpc: TraceCall
   */
  traceCall(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * BaseFee queries the base fee of the parent block of the current block,
   * it's similar to feemarket module's method, but also checks london hardfork
   * status.
   *
   * @generated from protobuf rpc: BaseFee
   */
  baseFee(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
