// @injectivelabs/core-proto-ts-v2/esm/generated/injective/evm/v1/trace_config_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ChainConfig } from "./chain_config_pb.js";
class TraceConfig$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.TraceConfig", [
      {
        no: 1,
        name: "tracer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "timeout",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "reexec",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 5, name: "disable_stack", kind: "scalar", T: 8, options: { "gogoproto.jsontag": "disableStack" } },
      { no: 6, name: "disable_storage", kind: "scalar", T: 8, options: { "gogoproto.jsontag": "disableStorage" } },
      {
        no: 8,
        name: "debug",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 9,
        name: "limit",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      { no: 10, name: "overrides", kind: "message", T: () => ChainConfig },
      { no: 11, name: "enable_memory", kind: "scalar", T: 8, options: { "gogoproto.jsontag": "enableMemory" } },
      { no: 12, name: "enable_return_data", kind: "scalar", T: 8, options: { "gogoproto.jsontag": "enableReturnData" } },
      { no: 13, name: "tracer_json_config", kind: "scalar", T: 9, options: { "gogoproto.jsontag": "tracerConfig" } },
      { no: 14, name: "state_overrides", kind: "scalar", T: 12, options: { "gogoproto.jsontag": "stateOverrides" } },
      { no: 15, name: "block_overrides", kind: "scalar", T: 12, options: { "gogoproto.jsontag": "blockOverrides" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.tracer = "";
    message.timeout = "";
    message.reexec = 0n;
    message.disableStack = false;
    message.disableStorage = false;
    message.debug = false;
    message.limit = 0;
    message.enableMemory = false;
    message.enableReturnData = false;
    message.tracerJsonConfig = "";
    message.stateOverrides = new Uint8Array(0);
    message.blockOverrides = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string tracer */
        1:
          message.tracer = reader.string();
          break;
        case /* string timeout */
        2:
          message.timeout = reader.string();
          break;
        case /* uint64 reexec */
        3:
          message.reexec = reader.uint64().toBigInt();
          break;
        case /* bool disable_stack */
        5:
          message.disableStack = reader.bool();
          break;
        case /* bool disable_storage */
        6:
          message.disableStorage = reader.bool();
          break;
        case /* bool debug */
        8:
          message.debug = reader.bool();
          break;
        case /* int32 limit */
        9:
          message.limit = reader.int32();
          break;
        case /* injective.evm.v1.ChainConfig overrides */
        10:
          message.overrides = ChainConfig.internalBinaryRead(reader, reader.uint32(), options, message.overrides);
          break;
        case /* bool enable_memory */
        11:
          message.enableMemory = reader.bool();
          break;
        case /* bool enable_return_data */
        12:
          message.enableReturnData = reader.bool();
          break;
        case /* string tracer_json_config */
        13:
          message.tracerJsonConfig = reader.string();
          break;
        case /* bytes state_overrides */
        14:
          message.stateOverrides = reader.bytes();
          break;
        case /* bytes block_overrides */
        15:
          message.blockOverrides = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.tracer !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.tracer);
    if (message.timeout !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.timeout);
    if (message.reexec !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.reexec);
    if (message.disableStack !== false)
      writer.tag(5, WireType.Varint).bool(message.disableStack);
    if (message.disableStorage !== false)
      writer.tag(6, WireType.Varint).bool(message.disableStorage);
    if (message.debug !== false)
      writer.tag(8, WireType.Varint).bool(message.debug);
    if (message.limit !== 0)
      writer.tag(9, WireType.Varint).int32(message.limit);
    if (message.overrides)
      ChainConfig.internalBinaryWrite(message.overrides, writer.tag(10, WireType.LengthDelimited).fork(), options).join();
    if (message.enableMemory !== false)
      writer.tag(11, WireType.Varint).bool(message.enableMemory);
    if (message.enableReturnData !== false)
      writer.tag(12, WireType.Varint).bool(message.enableReturnData);
    if (message.tracerJsonConfig !== "")
      writer.tag(13, WireType.LengthDelimited).string(message.tracerJsonConfig);
    if (message.stateOverrides.length)
      writer.tag(14, WireType.LengthDelimited).bytes(message.stateOverrides);
    if (message.blockOverrides.length)
      writer.tag(15, WireType.LengthDelimited).bytes(message.blockOverrides);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TraceConfig = new TraceConfig$Type();
export {
  TraceConfig
};
