// @injectivelabs/core-proto-ts-v2/esm/generated/injective/evm/v1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * EthereumTx defines a method submitting Ethereum transactions.
   *
   * @generated from protobuf rpc: EthereumTx
   */
  ethereumTx(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateParams defined a governance operation for updating the x/evm module
   * parameters. The authority is hard-coded to the Cosmos SDK x/gov module
   * account
   *
   * @generated from protobuf rpc: UpdateParams
   */
  updateParams(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
