// @injectivelabs/core-proto-ts-v2/esm/generated/injective/evm/v1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./params_pb.js";
import { Log } from "./log_pb.js";
import { AccessTuple } from "./access_tuple_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
class MsgEthereumTx$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.MsgEthereumTx", [
      { no: 1, name: "data", kind: "message", T: () => Any },
      { no: 2, name: "size", kind: "scalar", T: 1, options: { "gogoproto.jsontag": "-" } },
      { no: 3, name: "deprecated_hash", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'rlp:"-"' } },
      {
        no: 4,
        name: "deprecated_from",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "from",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 6, name: "raw", kind: "scalar", T: 12, options: { "gogoproto.nullable": false, "gogoproto.customtype": "EthereumTx" } }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["from"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.size = 0;
    message.deprecatedHash = "";
    message.deprecatedFrom = "";
    message.from = new Uint8Array(0);
    message.raw = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any data */
        1:
          message.data = Any.internalBinaryRead(reader, reader.uint32(), options, message.data);
          break;
        case /* double size */
        2:
          message.size = reader.double();
          break;
        case /* string deprecated_hash */
        3:
          message.deprecatedHash = reader.string();
          break;
        case /* string deprecated_from = 4 [deprecated = true] */
        4:
          message.deprecatedFrom = reader.string();
          break;
        case /* bytes from */
        5:
          message.from = reader.bytes();
          break;
        case /* bytes raw */
        6:
          message.raw = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.data)
      Any.internalBinaryWrite(message.data, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.size !== 0)
      writer.tag(2, WireType.Bit64).double(message.size);
    if (message.deprecatedHash !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.deprecatedHash);
    if (message.deprecatedFrom !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.deprecatedFrom);
    if (message.from.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.from);
    if (message.raw.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.raw);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgEthereumTx = new MsgEthereumTx$Type();
class LegacyTx$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.LegacyTx", [
      {
        no: 1,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "gas_price", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      { no: 3, name: "gas", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "GasLimit" } },
      {
        no: 4,
        name: "to",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 5, name: "value", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "Amount" } },
      {
        no: 6,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 7,
        name: "v",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 8,
        name: "r",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 9,
        name: "s",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos_proto.implements_interface": ["TxData"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nonce = 0n;
    message.gasPrice = "";
    message.gas = 0n;
    message.to = "";
    message.value = "";
    message.data = new Uint8Array(0);
    message.v = new Uint8Array(0);
    message.r = new Uint8Array(0);
    message.s = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 nonce */
        1:
          message.nonce = reader.uint64().toBigInt();
          break;
        case /* string gas_price */
        2:
          message.gasPrice = reader.string();
          break;
        case /* uint64 gas */
        3:
          message.gas = reader.uint64().toBigInt();
          break;
        case /* string to */
        4:
          message.to = reader.string();
          break;
        case /* string value */
        5:
          message.value = reader.string();
          break;
        case /* bytes data */
        6:
          message.data = reader.bytes();
          break;
        case /* bytes v */
        7:
          message.v = reader.bytes();
          break;
        case /* bytes r */
        8:
          message.r = reader.bytes();
          break;
        case /* bytes s */
        9:
          message.s = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.nonce);
    if (message.gasPrice !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.gasPrice);
    if (message.gas !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.gas);
    if (message.to !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.to);
    if (message.value !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.value);
    if (message.data.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.data);
    if (message.v.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.v);
    if (message.r.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.r);
    if (message.s.length)
      writer.tag(9, WireType.LengthDelimited).bytes(message.s);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const LegacyTx = new LegacyTx$Type();
class AccessListTx$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.AccessListTx", [
      { no: 1, name: "chain_id", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "ChainID", "gogoproto.jsontag": "chainID" } },
      {
        no: 2,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "gas_price", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      { no: 4, name: "gas", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "GasLimit" } },
      {
        no: 5,
        name: "to",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 6, name: "value", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "Amount" } },
      {
        no: 7,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 8, name: "accesses", kind: "message", repeat: 2, T: () => AccessTuple, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "accessList", "gogoproto.castrepeated": "AccessList" } },
      {
        no: 9,
        name: "v",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 10,
        name: "r",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 11,
        name: "s",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos_proto.implements_interface": ["TxData"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chainId = "";
    message.nonce = 0n;
    message.gasPrice = "";
    message.gas = 0n;
    message.to = "";
    message.value = "";
    message.data = new Uint8Array(0);
    message.accesses = [];
    message.v = new Uint8Array(0);
    message.r = new Uint8Array(0);
    message.s = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string chain_id */
        1:
          message.chainId = reader.string();
          break;
        case /* uint64 nonce */
        2:
          message.nonce = reader.uint64().toBigInt();
          break;
        case /* string gas_price */
        3:
          message.gasPrice = reader.string();
          break;
        case /* uint64 gas */
        4:
          message.gas = reader.uint64().toBigInt();
          break;
        case /* string to */
        5:
          message.to = reader.string();
          break;
        case /* string value */
        6:
          message.value = reader.string();
          break;
        case /* bytes data */
        7:
          message.data = reader.bytes();
          break;
        case /* repeated injective.evm.v1.AccessTuple accesses */
        8:
          message.accesses.push(AccessTuple.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* bytes v */
        9:
          message.v = reader.bytes();
          break;
        case /* bytes r */
        10:
          message.r = reader.bytes();
          break;
        case /* bytes s */
        11:
          message.s = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.chainId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.chainId);
    if (message.nonce !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.nonce);
    if (message.gasPrice !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.gasPrice);
    if (message.gas !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.gas);
    if (message.to !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.to);
    if (message.value !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.value);
    if (message.data.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.data);
    for (let i = 0; i < message.accesses.length; i++)
      AccessTuple.internalBinaryWrite(message.accesses[i], writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    if (message.v.length)
      writer.tag(9, WireType.LengthDelimited).bytes(message.v);
    if (message.r.length)
      writer.tag(10, WireType.LengthDelimited).bytes(message.r);
    if (message.s.length)
      writer.tag(11, WireType.LengthDelimited).bytes(message.s);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AccessListTx = new AccessListTx$Type();
class DynamicFeeTx$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.DynamicFeeTx", [
      { no: 1, name: "chain_id", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "ChainID", "gogoproto.jsontag": "chainID" } },
      {
        no: 2,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "gas_tip_cap", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      { no: 4, name: "gas_fee_cap", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      { no: 5, name: "gas", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "GasLimit" } },
      {
        no: 6,
        name: "to",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 7, name: "value", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.customname": "Amount" } },
      {
        no: 8,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 9, name: "accesses", kind: "message", repeat: 2, T: () => AccessTuple, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "accessList", "gogoproto.castrepeated": "AccessList" } },
      {
        no: 10,
        name: "v",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 11,
        name: "r",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 12,
        name: "s",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos_proto.implements_interface": ["TxData"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chainId = "";
    message.nonce = 0n;
    message.gasTipCap = "";
    message.gasFeeCap = "";
    message.gas = 0n;
    message.to = "";
    message.value = "";
    message.data = new Uint8Array(0);
    message.accesses = [];
    message.v = new Uint8Array(0);
    message.r = new Uint8Array(0);
    message.s = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string chain_id */
        1:
          message.chainId = reader.string();
          break;
        case /* uint64 nonce */
        2:
          message.nonce = reader.uint64().toBigInt();
          break;
        case /* string gas_tip_cap */
        3:
          message.gasTipCap = reader.string();
          break;
        case /* string gas_fee_cap */
        4:
          message.gasFeeCap = reader.string();
          break;
        case /* uint64 gas */
        5:
          message.gas = reader.uint64().toBigInt();
          break;
        case /* string to */
        6:
          message.to = reader.string();
          break;
        case /* string value */
        7:
          message.value = reader.string();
          break;
        case /* bytes data */
        8:
          message.data = reader.bytes();
          break;
        case /* repeated injective.evm.v1.AccessTuple accesses */
        9:
          message.accesses.push(AccessTuple.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* bytes v */
        10:
          message.v = reader.bytes();
          break;
        case /* bytes r */
        11:
          message.r = reader.bytes();
          break;
        case /* bytes s */
        12:
          message.s = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.chainId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.chainId);
    if (message.nonce !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.nonce);
    if (message.gasTipCap !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.gasTipCap);
    if (message.gasFeeCap !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.gasFeeCap);
    if (message.gas !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.gas);
    if (message.to !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.to);
    if (message.value !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.value);
    if (message.data.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.data);
    for (let i = 0; i < message.accesses.length; i++)
      AccessTuple.internalBinaryWrite(message.accesses[i], writer.tag(9, WireType.LengthDelimited).fork(), options).join();
    if (message.v.length)
      writer.tag(10, WireType.LengthDelimited).bytes(message.v);
    if (message.r.length)
      writer.tag(11, WireType.LengthDelimited).bytes(message.r);
    if (message.s.length)
      writer.tag(12, WireType.LengthDelimited).bytes(message.s);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DynamicFeeTx = new DynamicFeeTx$Type();
class ExtensionOptionsEthereumTx$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.ExtensionOptionsEthereumTx", [], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ExtensionOptionsEthereumTx = new ExtensionOptionsEthereumTx$Type();
class MsgEthereumTxResponse$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.MsgEthereumTxResponse", [
      {
        no: 1,
        name: "hash",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "logs", kind: "message", repeat: 2, T: () => Log },
      {
        no: 3,
        name: "ret",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "vm_error",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "gas_used",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "block_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.hash = "";
    message.logs = [];
    message.ret = new Uint8Array(0);
    message.vmError = "";
    message.gasUsed = 0n;
    message.blockHash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string hash */
        1:
          message.hash = reader.string();
          break;
        case /* repeated injective.evm.v1.Log logs */
        2:
          message.logs.push(Log.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* bytes ret */
        3:
          message.ret = reader.bytes();
          break;
        case /* string vm_error */
        4:
          message.vmError = reader.string();
          break;
        case /* uint64 gas_used */
        5:
          message.gasUsed = reader.uint64().toBigInt();
          break;
        case /* bytes block_hash */
        6:
          message.blockHash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.hash !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.hash);
    for (let i = 0; i < message.logs.length; i++)
      Log.internalBinaryWrite(message.logs[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.ret.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.ret);
    if (message.vmError !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.vmError);
    if (message.gasUsed !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.gasUsed);
    if (message.blockHash.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.blockHash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgEthereumTxResponse = new MsgEthereumTxResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.MsgUpdateParams", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ], { "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* injective.evm.v1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
const Msg = new ServiceType("injective.evm.v1.Msg", [
  { name: "EthereumTx", options: { "google.api.http": { post: "/injective/evm/v1/ethereum_tx" } }, I: MsgEthereumTx, O: MsgEthereumTxResponse },
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse }
], { "cosmos.msg.v1.service": true });
export {
  AccessListTx,
  DynamicFeeTx,
  ExtensionOptionsEthereumTx,
  LegacyTx,
  Msg,
  MsgEthereumTx,
  MsgEthereumTxResponse,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
