// @injectivelabs/core-proto-ts-v2/esm/generated/injective/evm/v1/tx_result_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { TransactionLogs } from "./transaction_logs_pb.js";
class TxResult$Type extends MessageType {
  constructor() {
    super("injective.evm.v1.TxResult", [
      { no: 1, name: "contract_address", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"contract_address"' } },
      {
        no: 2,
        name: "bloom",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "tx_logs", kind: "message", T: () => TransactionLogs, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"tx_logs"' } },
      {
        no: 4,
        name: "ret",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "reverted",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 6,
        name: "gas_used",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contractAddress = "";
    message.bloom = new Uint8Array(0);
    message.ret = new Uint8Array(0);
    message.reverted = false;
    message.gasUsed = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract_address */
        1:
          message.contractAddress = reader.string();
          break;
        case /* bytes bloom */
        2:
          message.bloom = reader.bytes();
          break;
        case /* injective.evm.v1.TransactionLogs tx_logs */
        3:
          message.txLogs = TransactionLogs.internalBinaryRead(reader, reader.uint32(), options, message.txLogs);
          break;
        case /* bytes ret */
        4:
          message.ret = reader.bytes();
          break;
        case /* bool reverted */
        5:
          message.reverted = reader.bool();
          break;
        case /* uint64 gas_used */
        6:
          message.gasUsed = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contractAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contractAddress);
    if (message.bloom.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.bloom);
    if (message.txLogs)
      TransactionLogs.internalBinaryWrite(message.txLogs, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.ret.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.ret);
    if (message.reverted !== false)
      writer.tag(5, WireType.Varint).bool(message.reverted);
    if (message.gasUsed !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.gasUsed);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxResult = new TxResult$Type();
export {
  TxResult
};
