// @injectivelabs/core-proto-ts-v2/esm/generated/injective/erc20/v1beta1/erc20_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class TokenPair$Type extends MessageType {
  constructor() {
    super("injective.erc20.v1beta1.TokenPair", [
      {
        no: 1,
        name: "bank_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "erc20_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.bankDenom = "";
    message.erc20Address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string bank_denom */
        1:
          message.bankDenom = reader.string();
          break;
        case /* string erc20_address */
        2:
          message.erc20Address = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.bankDenom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.bankDenom);
    if (message.erc20Address !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.erc20Address);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TokenPair = new TokenPair$Type();
export {
  TokenPair
};
