// @injectivelabs/core-proto-ts-v2/esm/generated/injective/erc20/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Params defines a gRPC query method that returns the erc20 module's
   * parameters.
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * AllTokenPairs defines a gRPC query method that returns the erc20
   * module's created token pairs.
   *
   * @generated from protobuf rpc: AllTokenPairs
   */
  allTokenPairs(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TokenPairByDenom defines a gRPC query method that returns the erc20
   * module's token pair associated with the provided bank denom.
   *
   * @generated from protobuf rpc: TokenPairByDenom
   */
  tokenPairByDenom(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TokenPairByERC20Address defines a gRPC query method that returns the erc20
   * module's token pair associated with the provided erc20 contract address.
   *
   * @generated from protobuf rpc: TokenPairByERC20Address
   */
  tokenPairByERC20Address(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
