// @injectivelabs/core-proto-ts-v2/esm/generated/injective/erc20/v1beta1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * @generated from protobuf rpc: UpdateParams
   */
  updateParams(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: CreateTokenPair
   */
  createTokenPair(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: DeleteTokenPair
   */
  deleteTokenPair(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
