// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/store/snapshots/v1/snapshot_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Snapshot$Type extends MessageType {
  constructor() {
    super("cosmos.store.snapshots.v1.Snapshot", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "format",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "chunks",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 4,
        name: "hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 5, name: "metadata", kind: "message", T: () => Metadata, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    message.format = 0;
    message.chunks = 0;
    message.hash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 height */
        1:
          message.height = reader.uint64().toBigInt();
          break;
        case /* uint32 format */
        2:
          message.format = reader.uint32();
          break;
        case /* uint32 chunks */
        3:
          message.chunks = reader.uint32();
          break;
        case /* bytes hash */
        4:
          message.hash = reader.bytes();
          break;
        case /* cosmos.store.snapshots.v1.Metadata metadata */
        5:
          message.metadata = Metadata.internalBinaryRead(reader, reader.uint32(), options, message.metadata);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.height);
    if (message.format !== 0)
      writer.tag(2, WireType.Varint).uint32(message.format);
    if (message.chunks !== 0)
      writer.tag(3, WireType.Varint).uint32(message.chunks);
    if (message.hash.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.hash);
    if (message.metadata)
      Metadata.internalBinaryWrite(message.metadata, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Snapshot = new Snapshot$Type();
class Metadata$Type extends MessageType {
  constructor() {
    super("cosmos.store.snapshots.v1.Metadata", [
      {
        no: 1,
        name: "chunk_hashes",
        kind: "scalar",
        repeat: 2,
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chunkHashes = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated bytes chunk_hashes */
        1:
          message.chunkHashes.push(reader.bytes());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.chunkHashes.length; i++)
      writer.tag(1, WireType.LengthDelimited).bytes(message.chunkHashes[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Metadata = new Metadata$Type();
class SnapshotItem$Type extends MessageType {
  constructor() {
    super("cosmos.store.snapshots.v1.SnapshotItem", [
      { no: 1, name: "store", kind: "message", oneof: "item", T: () => SnapshotStoreItem },
      { no: 2, name: "iavl", kind: "message", oneof: "item", T: () => SnapshotIAVLItem, options: { "gogoproto.customname": "IAVL" } },
      { no: 3, name: "extension", kind: "message", oneof: "item", T: () => SnapshotExtensionMeta },
      { no: 4, name: "extension_payload", kind: "message", oneof: "item", T: () => SnapshotExtensionPayload }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.item = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.store.snapshots.v1.SnapshotStoreItem store */
        1:
          message.item = {
            oneofKind: "store",
            store: SnapshotStoreItem.internalBinaryRead(reader, reader.uint32(), options, message.item.store)
          };
          break;
        case /* cosmos.store.snapshots.v1.SnapshotIAVLItem iavl */
        2:
          message.item = {
            oneofKind: "iavl",
            iavl: SnapshotIAVLItem.internalBinaryRead(reader, reader.uint32(), options, message.item.iavl)
          };
          break;
        case /* cosmos.store.snapshots.v1.SnapshotExtensionMeta extension */
        3:
          message.item = {
            oneofKind: "extension",
            extension: SnapshotExtensionMeta.internalBinaryRead(reader, reader.uint32(), options, message.item.extension)
          };
          break;
        case /* cosmos.store.snapshots.v1.SnapshotExtensionPayload extension_payload */
        4:
          message.item = {
            oneofKind: "extensionPayload",
            extensionPayload: SnapshotExtensionPayload.internalBinaryRead(reader, reader.uint32(), options, message.item.extensionPayload)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.item.oneofKind === "store")
      SnapshotStoreItem.internalBinaryWrite(message.item.store, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.item.oneofKind === "iavl")
      SnapshotIAVLItem.internalBinaryWrite(message.item.iavl, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.item.oneofKind === "extension")
      SnapshotExtensionMeta.internalBinaryWrite(message.item.extension, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.item.oneofKind === "extensionPayload")
      SnapshotExtensionPayload.internalBinaryWrite(message.item.extensionPayload, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SnapshotItem = new SnapshotItem$Type();
class SnapshotStoreItem$Type extends MessageType {
  constructor() {
    super("cosmos.store.snapshots.v1.SnapshotStoreItem", [
      {
        no: 1,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.name = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string name */
        1:
          message.name = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.name !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.name);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SnapshotStoreItem = new SnapshotStoreItem$Type();
class SnapshotIAVLItem$Type extends MessageType {
  constructor() {
    super("cosmos.store.snapshots.v1.SnapshotIAVLItem", [
      {
        no: 1,
        name: "key",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "value",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "version",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "height",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.key = new Uint8Array(0);
    message.value = new Uint8Array(0);
    message.version = 0n;
    message.height = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes key */
        1:
          message.key = reader.bytes();
          break;
        case /* bytes value */
        2:
          message.value = reader.bytes();
          break;
        case /* int64 version */
        3:
          message.version = reader.int64().toBigInt();
          break;
        case /* int32 height */
        4:
          message.height = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.key.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.key);
    if (message.value.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.value);
    if (message.version !== 0n)
      writer.tag(3, WireType.Varint).int64(message.version);
    if (message.height !== 0)
      writer.tag(4, WireType.Varint).int32(message.height);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SnapshotIAVLItem = new SnapshotIAVLItem$Type();
class SnapshotExtensionMeta$Type extends MessageType {
  constructor() {
    super("cosmos.store.snapshots.v1.SnapshotExtensionMeta", [
      {
        no: 1,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "format",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.name = "";
    message.format = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string name */
        1:
          message.name = reader.string();
          break;
        case /* uint32 format */
        2:
          message.format = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.name !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.name);
    if (message.format !== 0)
      writer.tag(2, WireType.Varint).uint32(message.format);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SnapshotExtensionMeta = new SnapshotExtensionMeta$Type();
class SnapshotExtensionPayload$Type extends MessageType {
  constructor() {
    super("cosmos.store.snapshots.v1.SnapshotExtensionPayload", [
      {
        no: 1,
        name: "payload",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.payload = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes payload */
        1:
          message.payload = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.payload.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.payload);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SnapshotExtensionPayload = new SnapshotExtensionPayload$Type();
export {
  Metadata,
  Snapshot,
  SnapshotExtensionMeta,
  SnapshotExtensionPayload,
  SnapshotIAVLItem,
  SnapshotItem,
  SnapshotStoreItem
};
