// @injectivelabs/core-proto-ts-v2/esm/generated/injective/downtimedetector/v1beta1/downtime_duration_pb.js
var Downtime = /* @__PURE__ */ ((Downtime2) => {
  Downtime2[Downtime2["DURATION_30S"] = 0] = "DURATION_30S";
  Downtime2[Downtime2["DURATION_1M"] = 1] = "DURATION_1M";
  Downtime2[Downtime2["DURATION_2M"] = 2] = "DURATION_2M";
  Downtime2[Downtime2["DURATION_3M"] = 3] = "DURATION_3M";
  Downtime2[Downtime2["DURATION_4M"] = 4] = "DURATION_4M";
  Downtime2[Downtime2["DURATION_5M"] = 5] = "DURATION_5M";
  Downtime2[Downtime2["DURATION_10M"] = 6] = "DURATION_10M";
  Downtime2[Downtime2["DURATION_20M"] = 7] = "DURATION_20M";
  Downtime2[Downtime2["DURATION_30M"] = 8] = "DURATION_30M";
  Downtime2[Downtime2["DURATION_40M"] = 9] = "DURATION_40M";
  Downtime2[Downtime2["DURATION_50M"] = 10] = "DURATION_50M";
  Downtime2[Downtime2["DURATION_1H"] = 11] = "DURATION_1H";
  Downtime2[Downtime2["DURATION_1_5H"] = 12] = "DURATION_1_5H";
  Downtime2[Downtime2["DURATION_2H"] = 13] = "DURATION_2H";
  Downtime2[Downtime2["DURATION_2_5H"] = 14] = "DURATION_2_5H";
  Downtime2[Downtime2["DURATION_3H"] = 15] = "DURATION_3H";
  Downtime2[Downtime2["DURATION_4H"] = 16] = "DURATION_4H";
  Downtime2[Downtime2["DURATION_5H"] = 17] = "DURATION_5H";
  Downtime2[Downtime2["DURATION_6H"] = 18] = "DURATION_6H";
  Downtime2[Downtime2["DURATION_9H"] = 19] = "DURATION_9H";
  Downtime2[Downtime2["DURATION_12H"] = 20] = "DURATION_12H";
  Downtime2[Downtime2["DURATION_18H"] = 21] = "DURATION_18H";
  Downtime2[Downtime2["DURATION_24H"] = 22] = "DURATION_24H";
  Downtime2[Downtime2["DURATION_36H"] = 23] = "DURATION_36H";
  Downtime2[Downtime2["DURATION_48H"] = 24] = "DURATION_48H";
  return Downtime2;
})(Downtime || {});
export {
  Downtime
};
