// @injectivelabs/core-proto-ts-v2/esm/generated/injective/downtimedetector/v1beta1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Downtime } from "./downtime_duration_pb.js";
class GenesisDowntimeEntry$Type extends MessageType {
  constructor() {
    super("injective.downtimedetector.v1beta1.GenesisDowntimeEntry", [
      { no: 1, name: "duration", kind: "enum", T: () => ["injective.downtimedetector.v1beta1.Downtime", Downtime], options: { "gogoproto.moretags": 'yaml:"duration"' } },
      { no: 2, name: "last_downtime", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"last_downtime"', "gogoproto.stdtime": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.duration = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.downtimedetector.v1beta1.Downtime duration */
        1:
          message.duration = reader.int32();
          break;
        case /* google.protobuf.Timestamp last_downtime */
        2:
          message.lastDowntime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.lastDowntime);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.duration !== 0)
      writer.tag(1, WireType.Varint).int32(message.duration);
    if (message.lastDowntime)
      Timestamp.internalBinaryWrite(message.lastDowntime, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisDowntimeEntry = new GenesisDowntimeEntry$Type();
class GenesisState$Type extends MessageType {
  constructor() {
    super("injective.downtimedetector.v1beta1.GenesisState", [
      { no: 1, name: "downtimes", kind: "message", repeat: 2, T: () => GenesisDowntimeEntry, options: { "gogoproto.nullable": false } },
      { no: 2, name: "last_block_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"last_block_time"', "gogoproto.stdtime": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.downtimes = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective.downtimedetector.v1beta1.GenesisDowntimeEntry downtimes */
        1:
          message.downtimes.push(GenesisDowntimeEntry.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* google.protobuf.Timestamp last_block_time */
        2:
          message.lastBlockTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.lastBlockTime);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.downtimes.length; i++)
      GenesisDowntimeEntry.internalBinaryWrite(message.downtimes[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.lastBlockTime)
      Timestamp.internalBinaryWrite(message.lastBlockTime, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
export {
  GenesisDowntimeEntry,
  GenesisState
};
