// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/app/v1alpha1/config_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../google/protobuf/any_pb.js";
class Config$Type extends MessageType {
  constructor() {
    super("cosmos.app.v1alpha1.Config", [
      { no: 1, name: "modules", kind: "message", repeat: 2, T: () => ModuleConfig },
      { no: 2, name: "golang_bindings", kind: "message", repeat: 2, T: () => GolangBinding }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.modules = [];
    message.golangBindings = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.app.v1alpha1.ModuleConfig modules */
        1:
          message.modules.push(ModuleConfig.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.app.v1alpha1.GolangBinding golang_bindings */
        2:
          message.golangBindings.push(GolangBinding.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.modules.length; i++)
      ModuleConfig.internalBinaryWrite(message.modules[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.golangBindings.length; i++)
      GolangBinding.internalBinaryWrite(message.golangBindings[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Config = new Config$Type();
class ModuleConfig$Type extends MessageType {
  constructor() {
    super("cosmos.app.v1alpha1.ModuleConfig", [
      {
        no: 1,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "config", kind: "message", T: () => Any },
      { no: 3, name: "golang_bindings", kind: "message", repeat: 2, T: () => GolangBinding }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.name = "";
    message.golangBindings = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string name */
        1:
          message.name = reader.string();
          break;
        case /* google.protobuf.Any config */
        2:
          message.config = Any.internalBinaryRead(reader, reader.uint32(), options, message.config);
          break;
        case /* repeated cosmos.app.v1alpha1.GolangBinding golang_bindings */
        3:
          message.golangBindings.push(GolangBinding.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.name !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.name);
    if (message.config)
      Any.internalBinaryWrite(message.config, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.golangBindings.length; i++)
      GolangBinding.internalBinaryWrite(message.golangBindings[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ModuleConfig = new ModuleConfig$Type();
class GolangBinding$Type extends MessageType {
  constructor() {
    super("cosmos.app.v1alpha1.GolangBinding", [
      {
        no: 1,
        name: "interface_type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "implementation",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.interfaceType = "";
    message.implementation = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string interface_type */
        1:
          message.interfaceType = reader.string();
          break;
        case /* string implementation */
        2:
          message.implementation = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.interfaceType !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.interfaceType);
    if (message.implementation !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.implementation);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GolangBinding = new GolangBinding$Type();
export {
  Config,
  GolangBinding,
  ModuleConfig
};
