// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/app/v1alpha1/module_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class ModuleDescriptor$Type extends MessageType {
  constructor() {
    super("cosmos.app.v1alpha1.ModuleDescriptor", [
      {
        no: 1,
        name: "go_import",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "use_package", kind: "message", repeat: 2, T: () => PackageReference },
      { no: 3, name: "can_migrate_from", kind: "message", repeat: 2, T: () => MigrateFromInfo }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.goImport = "";
    message.usePackage = [];
    message.canMigrateFrom = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string go_import */
        1:
          message.goImport = reader.string();
          break;
        case /* repeated cosmos.app.v1alpha1.PackageReference use_package */
        2:
          message.usePackage.push(PackageReference.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.app.v1alpha1.MigrateFromInfo can_migrate_from */
        3:
          message.canMigrateFrom.push(MigrateFromInfo.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.goImport !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.goImport);
    for (let i = 0; i < message.usePackage.length; i++)
      PackageReference.internalBinaryWrite(message.usePackage[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.canMigrateFrom.length; i++)
      MigrateFromInfo.internalBinaryWrite(message.canMigrateFrom[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ModuleDescriptor = new ModuleDescriptor$Type();
class PackageReference$Type extends MessageType {
  constructor() {
    super("cosmos.app.v1alpha1.PackageReference", [
      {
        no: 1,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "revision",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.name = "";
    message.revision = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string name */
        1:
          message.name = reader.string();
          break;
        case /* uint32 revision */
        2:
          message.revision = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.name !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.name);
    if (message.revision !== 0)
      writer.tag(2, WireType.Varint).uint32(message.revision);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PackageReference = new PackageReference$Type();
class MigrateFromInfo$Type extends MessageType {
  constructor() {
    super("cosmos.app.v1alpha1.MigrateFromInfo", [
      {
        no: 1,
        name: "module",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.module = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string module */
        1:
          message.module = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.module !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.module);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MigrateFromInfo = new MigrateFromInfo$Type();
export {
  MigrateFromInfo,
  ModuleDescriptor,
  PackageReference
};
