// @injectivelabs/core-proto-ts-v2/esm/generated/capability/v1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { CapabilityOwners } from "./capability_pb.js";
class GenesisOwners$Type extends MessageType {
  constructor() {
    super("capability.v1.GenesisOwners", [
      {
        no: 1,
        name: "index",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "index_owners", kind: "message", T: () => CapabilityOwners, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.index = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 index */
        1:
          message.index = reader.uint64().toBigInt();
          break;
        case /* capability.v1.CapabilityOwners index_owners */
        2:
          message.indexOwners = CapabilityOwners.internalBinaryRead(reader, reader.uint32(), options, message.indexOwners);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.index !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.index);
    if (message.indexOwners)
      CapabilityOwners.internalBinaryWrite(message.indexOwners, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisOwners = new GenesisOwners$Type();
class GenesisState$Type extends MessageType {
  constructor() {
    super("capability.v1.GenesisState", [
      {
        no: 1,
        name: "index",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "owners", kind: "message", repeat: 2, T: () => GenesisOwners, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.index = 0n;
    message.owners = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 index */
        1:
          message.index = reader.uint64().toBigInt();
          break;
        case /* repeated capability.v1.GenesisOwners owners */
        2:
          message.owners.push(GenesisOwners.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.index !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.index);
    for (let i = 0; i < message.owners.length; i++)
      GenesisOwners.internalBinaryWrite(message.owners[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
export {
  GenesisOwners,
  GenesisState
};
