// @injectivelabs/networks/dist/cjs/index.cjs
let __injectivelabs_ts_types = require("@injectivelabs/ts-types");

//#region src/types.ts
const Network = {
	MainnetK8s: "mainnetK8s",
	MainnetLB: "mainnetLB",
	Mainnet: "mainnet",
	MainnetSentry: "mainnetSentry",
	MainnetOld: "mainnetOld",
	Staging: "staging",
	Internal: "internal",
	TestnetK8s: "testnetK8s",
	TestnetOld: "testnetOld",
	TestnetSentry: "testnetSentry",
	Testnet: "testnet",
	Devnet1: "devnet1",
	Devnet2: "devnet2",
	Devnet3: "devnet3",
	Devnet: "devnet",
	Local: "local"
};

//#endregion
//#region src/chainInfos.ts
const INJ_DENOM = "inj";
const getMainnetChainInfo = () => ({
	feeDenom: INJ_DENOM,
	chainId: __injectivelabs_ts_types.ChainId.Mainnet,
	evmChainId: __injectivelabs_ts_types.EvmChainId.Mainnet,
	env: "mainnet"
});
const getTestnetChainInfo = () => ({
	feeDenom: INJ_DENOM,
	chainId: __injectivelabs_ts_types.ChainId.Testnet,
	evmChainId: __injectivelabs_ts_types.EvmChainId.Sepolia,
	env: "testnet"
});
const getDevnetChainInfo = () => ({
	feeDenom: INJ_DENOM,
	chainId: __injectivelabs_ts_types.ChainId.Devnet,
	evmChainId: __injectivelabs_ts_types.EvmChainId.Sepolia,
	env: "devnet"
});
const getLocalChainInfo = () => ({
	feeDenom: INJ_DENOM,
	chainId: __injectivelabs_ts_types.ChainId.Mainnet,
	evmChainId: __injectivelabs_ts_types.EvmChainId.Sepolia,
	env: "local"
});

//#endregion
//#region src/endpoints.ts
const getEndpointsMainnetSentry = () => ({
	indexer: "https://sentry.exchange.grpc-web.injective.network",
	grpc: "https://sentry.chain.grpc-web.injective.network",
	rpc: "https://sentry.tm.injective.network",
	rest: "https://sentry.lcd.injective.network",
	chronos: "https://sentry.exchange.grpc-web.injective.network",
	explorer: "https://sentry.explorer.grpc-web.injective.network",
	cacheGrpc: "https://sentry.chain.grpc-web.injective.network",
	cacheRest: "https://staging.gateway.grpc-web.injective.network",
	web3gw: "https://sentry.exchange.grpc-web.injective.network",
	chart: "https://sentry.exchange.grpc-web.injective.network"
});
const getEndpointsStaging = () => ({
	indexer: "https://k8s.mainnet.staging.exchange.grpc-web.injective.network",
	grpc: "https://k8s.global.mainnet.chain.grpc-web.injective.network",
	rpc: "https://k8s.global.mainnet.tm.injective.network",
	rest: "https://k8s.global.mainnet.lcd.injective.network",
	chronos: "https://k8s.mainnet.staging.exchange.grpc-web.injective.network",
	explorer: "https://k8s.mainnet.staging.exchange.grpc-web.injective.network",
	cacheGrpc: "https://k8s.global.mainnet.chain.grpc-web.injective.network",
	cacheRest: "https://k8s.mainnet.staging.exchange.grpc-web.injective.network",
	web3gw: "https://k8s.mainnet.staging.exchange.grpc-web.injective.network",
	chart: "https://k8s.mainnet.staging.chart.grpc-web.injective.network"
});
const getEndpointsInternal = () => ({
	indexer: "https://products.exchange.grpc-web.injective.network",
	grpc: "https://products.chain.grpc-web.injective.network",
	rpc: "https://products.tm.injective.network",
	rest: "https://products.lcd.injective.network",
	chronos: "https://products.chronos.grpc-web.injective.network",
	explorer: "https://products.explorer.grpc-web.injective.network",
	cacheGrpc: "https://products.chain.grpc-web.injective.network",
	cacheRest: "https://staging.gateway.grpc-web.injective.network",
	web3gw: "https://products.web3-gateway.injective.network",
	chart: "https://products.exchange.grpc-web.injective.network"
});
const getEndpointsTestnetSentry = () => ({
	indexer: "https://testnet.sentry.exchange.grpc-web.injective.network",
	grpc: "https://testnet.sentry.chain.grpc-web.injective.network",
	rpc: "https://testnet.sentry.tm.injective.network",
	rest: "https://testnet.sentry.lcd.injective.network",
	chronos: "https://testnet.sentry.exchange.grpc-web.injective.network",
	explorer: "https://testnet.sentry.exchange.grpc-web.injective.network",
	cacheGrpc: "https://testnet.sentry.chain.grpc-web.injective.network",
	cacheRest: "https://testnet.sentry.exchange.grpc-web.injective.network",
	web3gw: "https://testnet.sentry.exchange.grpc-web.injective.network",
	chart: "https://k8s.testnet.chart.grpc-web.injective.network"
});
const getEndpointsDevnet = () => ({
	indexer: "https://devnet.api.injective.dev",
	grpc: "https://devnet.grpc.injective.dev",
	rpc: "https://devnet.tm.injective.dev",
	rest: "https://devnet.lcd.injective.dev",
	chronos: "https://devnet.api.injective.dev",
	explorer: "https://devnet.api.injective.dev",
	cacheGrpc: "https://devnet.grpc.injective.dev",
	cacheRest: "https://devnet.api.injective.dev",
	web3gw: "https://devnet.api.injective.dev",
	chart: "https://devnet.api.injective.dev"
});
const getEndpointsDevnet1 = () => ({
	indexer: "https://devnet-1.api.injective.dev",
	grpc: "https://devnet-1.grpc.injective.dev",
	rpc: "https://devnet-1.tm.injective.dev",
	rest: "https://devnet-1.lcd.injective.dev",
	chronos: "https://devnet-1.api.injective.dev",
	explorer: "https://devnet-1.api.injective.dev",
	cacheGrpc: "https://devnet-1.grpc.injective.dev",
	cacheRest: "https://devnet-1.api.injective.dev",
	web3gw: "https://devnet-1.api.injective.dev",
	chart: "https://devnet-1.api.injective.dev"
});
const getEndpointsDevnet2 = () => ({
	indexer: "https://devnet-2.api.injective.dev",
	grpc: "https://devnet-2.grpc.injective.dev",
	rpc: "https://devnet-2.tm.injective.dev",
	rest: "https://devnet-2.lcd.injective.dev",
	chronos: "https://devnet-2.api.injective.dev",
	explorer: "https://devnet-2.api.injective.dev",
	cacheGrpc: "https://devnet-2.grpc.injective.dev",
	cacheRest: "https://devnet-2.api.injective.dev",
	web3gw: "https://devnet-2.api.injective.dev",
	chart: "https://devnet-2.api.injective.dev"
});
const getEndpointsDevnet3 = () => ({
	indexer: "https://devnet-3.api.injective.dev",
	grpc: "https://devnet-3.grpc.injective.dev",
	rpc: "https://devnet-3.tm.injective.dev",
	rest: "https://devnet-3.lcd.injective.dev",
	chronos: "https://devnet-3.api.injective.dev",
	explorer: "https://devnet-3.api.injective.dev",
	cacheGrpc: "https://devnet-3.grpc.injective.dev",
	cacheRest: "https://devnet-3.api.injective.dev",
	web3gw: "https://devnet-3.api.injective.dev",
	chart: "https://devnet-3.api.injective.dev"
});
const getEndpointsLocal = () => ({
	indexer: "https://localhost:4444",
	grpc: "http://localhost:10337",
	rpc: "http://localhost:9092",
	rest: "http://localhost:10337",
	chronos: "https://localhost:4442",
	explorer: "http://localhost:4443",
	cacheGrpc: "http://localhost:10337",
	cacheRest: "https://localhost:4444",
	web3gw: "https://localhost:4444"
});
/**
* @deprecated use TestnetSentry instead
*/
const getEndpointsTestnetOld = () => ({
	indexer: "https://testnet.exchange.grpc-web.injective.network",
	grpc: "https://testnet.chain.grpc-web.injective.network",
	rpc: "https://testnet.tm.injective.network",
	rest: "https://testnet.lcd.injective.network",
	chronos: "https://testnet.exchange.grpc-web.injective.network",
	explorer: "https://testnet.exchange.grpc-web.injective.network",
	web3gw: "https://testnet.exchange.grpc-web.injective.network",
	cacheGrpc: "https://testnet.exchange.grpc-web.injective.network/",
	cacheRest: "https://testnet.exchange.grpc-web.injective.network"
});
/**
* @deprecated use TestnetSentry instead
*/
const getEndpointsTestnetK8s = () => ({
	indexer: "https://k8s.testnet.exchange.grpc-web.injective.network",
	grpc: "https://k8s.testnet.chain.grpc-web.injective.network",
	rpc: "https://k8s.testnet.tm.injective.network",
	rest: "https://k8s.testnet.lcd.injective.network",
	chronos: "https://k8s.testnet.exchange.grpc-web.injective.network",
	explorer: "https://k8s.testnet.explorer.grpc-web.injective.network",
	cacheGrpc: "https://k8s.testnet.gateway.grpc.injective.network",
	cacheRest: "https://k8s.testnet.gateway.grpc-web.injective.network",
	web3gw: "https://k8s.testnet.exchange.grpc-web.injective.network"
});
/**
* @deprecated use MainnetSentry instead
*/
const getEndpointsMainnetLB = () => ({
	indexer: "https://k8s.global.mainnet.exchange.grpc-web.injective.network",
	grpc: "https://k8s.global.mainnet.chain.grpc-web.injective.network",
	rpc: "https://k8s.global.mainnet.tm.injective.network",
	rest: "https://k8s.global.mainnet.lcd.injective.network",
	chronos: "https://k8s.global.mainnet.exchange.grpc-web.injective.network",
	explorer: "https://k8s.global.mainnet.exchange.grpc-web.injective.network",
	cacheGrpc: "https://k8s.global.mainnet.chain.grpc-web.injective.network",
	cacheRest: "https://k8s.global.mainnet.exchange.grpc-web.injective.network",
	web3gw: "https://k8s.global.mainnet.exchange.grpc-web.injective.network",
	chart: "https://k8s.global.mainnet.chart.grpc-web.injective.network"
});
/**
* @deprecated use MainnetSentry instead
*/
const getEndpointsMainnetOld = () => ({
	indexer: "https://api.injective.network",
	grpc: "https://grpc.injective.network",
	cacheGrpc: "https://grpc.injective.network",
	rpc: "https://tm.injective.network",
	rest: "https://lcd.injective.network",
	cacheRest: "https://api.injective.network",
	chronos: "https://api.injective.network",
	explorer: "https://api.injective.network",
	web3gw: "https://api.injective.network"
});
/**
* @deprecated use MainnetSentry instead
*/
const getEndpointsMainnetK8s = () => ({
	indexer: "https://k8s.mainnet.exchange.grpc-web.injective.network",
	grpc: "https://k8s.mainnet.chain.grpc-web.injective.network",
	rpc: "https://k8s.mainnet.tm.injective.network",
	rest: "https://k8s.mainnet.lcd.injective.network",
	chronos: "https://k8s.mainnet.exchange.grpc-web.injective.network",
	explorer: "https://k8s.mainnet.exchange.grpc-web.injective.network",
	cacheGrpc: "https://k8s.mainnet.chain.grpc-web.injective.network",
	cacheRest: "https://k8s.mainnet.exchange.grpc-web.injective.network",
	web3gw: "https://k8s.mainnet.exchange.grpc-web.injective.network"
});

//#endregion
//#region src/network.ts
const getNetworkEndpoints = (network) => {
	switch (network) {
		case Network.MainnetLB: return getEndpointsMainnetLB();
		case Network.MainnetK8s: return getEndpointsMainnetK8s();
		case Network.MainnetSentry:
		case Network.Mainnet: return getEndpointsMainnetSentry();
		case Network.MainnetOld: return getEndpointsMainnetOld();
		case Network.Staging: return getEndpointsStaging();
		case Network.Internal: return getEndpointsInternal();
		case Network.Devnet: return getEndpointsDevnet();
		case Network.Devnet1: return getEndpointsDevnet1();
		case Network.Devnet2: return getEndpointsDevnet2();
		case Network.Devnet3: return getEndpointsDevnet3();
		case Network.TestnetK8s: return getEndpointsTestnetK8s();
		case Network.TestnetOld: return getEndpointsTestnetOld();
		case Network.TestnetSentry:
		case Network.Testnet: return getEndpointsTestnetSentry();
		case Network.Local: return getEndpointsLocal();
		default: throw new Error(`Unknown network: ${network}`);
	}
};
/**
* @deprecated - use getNetworkChainInfo instead
* @param network deprecated
* @returns
*/
const getChainInfoForNetwork = (network) => getNetworkChainInfo(network);
const getNetworkChainInfo = (network) => {
	if (network === Network.Local) return getLocalChainInfo();
	if (isTestnet(network)) return getTestnetChainInfo();
	if (isDevnet(network)) return getDevnetChainInfo();
	return getMainnetChainInfo();
};
const getNetworkInfo = (network) => ({
	...getNetworkChainInfo(network),
	...getNetworkEndpoints(network)
});
const isMainnet = (network) => [
	Network.Staging,
	Network.Mainnet,
	Network.MainnetOld,
	Network.MainnetK8s,
	Network.MainnetSentry,
	Network.Internal,
	Network.MainnetLB
].includes(network);
const isDevnet = (network) => [
	Network.Devnet,
	Network.Devnet1,
	Network.Devnet2,
	Network.Devnet3,
	Network.Local
].includes(network);
const isTestnet = (network) => [
	Network.Testnet,
	Network.TestnetOld,
	Network.TestnetK8s,
	Network.TestnetSentry
].includes(network);
const isTestnetOrDevnet = (network) => isDevnet(network) || isTestnet(network);

//#endregion
//#region src/utils.ts
const CW20_CODE_IDS_BY_NETWORK = (network = Network.Mainnet) => {
	if (isTestnet(network)) return ["25"];
	return [
		"28",
		"5",
		"42"
	];
};
const getCw20AdapterContractForNetwork = (network = Network.Mainnet) => {
	if (isDevnet(network)) return "inj1uukt3kqela4vsllvrqnrgllkna5wn3cm588w6k";
	if (isTestnet(network)) return "inj1hdvy6tl89llqy3ze8lv6mz5qh66sx9enn0jxg6";
	return "inj14ejqjyq8um4p3xfqj74yld5waqljf88f9eneuk";
};
const getCw20SwapContractForNetwork = (network = Network.Mainnet) => {
	if (isDevnet(network)) return "inj177yh38g3ctu7cemxpa3c2kvwh2yslfxfmfa66h";
	if (isTestnet(network)) return "inj14d7h5j6ddq6pqppl65z24w7xrtmpcrqjxj8d43";
	if (network === Network.Staging) return "inj12yj3mtjarujkhcp6lg3klxjjfrx2v7v8yswgp9";
	return "inj1psk3468yr9teahgz73amwvpfjehnhczvkrhhqx";
};
const getIncentivesContractForNetwork = (network = Network.Mainnet) => {
	if (isDevnet(network)) return "";
	if (isTestnet(network)) return "inj16twru668nsl7tqzahxd9q033swhr6a5xuslpkt";
	return "";
};
const getInjNameRegistryContractForNetwork = (network = Network.Mainnet) => {
	if (isDevnet(network)) return "inj1aw59rkpd9afp2ws6rx23nz5mrvq8dlckeslwfa";
	if (isTestnet(network)) return "inj1aw59rkpd9afp2ws6rx23nz5mrvq8dlckeslwfa";
	return "inj1hm8vs8sr2h9nk0x66vctfs528wrp6k3gtgg275";
};
const getInjNameReverseResolverContractForNetwork = (network = Network.Mainnet) => {
	if (isDevnet(network)) return "inj1knf6puyscuuqqhgqglskfc0k99d4885qw5uv7v";
	if (isTestnet(network)) return "inj1knf6puyscuuqqhgqglskfc0k99d4885qw5uv7v";
	return "inj1x9m0hceug9qylcyrrtwqtytslv2jrph433thgu";
};
const getPeggyGraphQlEndpointForNetwork = (network) => {
	if (isDevnet(network)) return "https://api.thegraph.com/subgraphs/name/injectivelabsdev/injective-peggo-devnet";
	if (isTestnet(network)) return "https://api.thegraph.com/subgraphs/name/injectivelabs/injective-peggo-goerli";
	return "https://api.thegraph.com/subgraphs/name/injectivelabs/injective-peggo-mainnet";
};
const getAssetPriceServiceForNetwork = (network) => {
	if (isDevnet(network)) return "https://devnet.asset.injective.dev/asset-price/v1";
	if (isTestnet(network)) return "https://k8s.testnet.asset.injective.network/asset-price/v1";
	return "https://k8s.mainnet.asset.injective.network/asset-price/v1";
};

//#endregion
exports.CW20_CODE_IDS_BY_NETWORK = CW20_CODE_IDS_BY_NETWORK;
exports.Network = Network;
exports.getAssetPriceServiceForNetwork = getAssetPriceServiceForNetwork;
exports.getChainInfoForNetwork = getChainInfoForNetwork;
exports.getCw20AdapterContractForNetwork = getCw20AdapterContractForNetwork;
exports.getCw20SwapContractForNetwork = getCw20SwapContractForNetwork;
exports.getIncentivesContractForNetwork = getIncentivesContractForNetwork;
exports.getInjNameRegistryContractForNetwork = getInjNameRegistryContractForNetwork;
exports.getInjNameReverseResolverContractForNetwork = getInjNameReverseResolverContractForNetwork;
exports.getNetworkChainInfo = getNetworkChainInfo;
exports.getNetworkEndpoints = getNetworkEndpoints;
exports.getNetworkInfo = getNetworkInfo;
exports.getPeggyGraphQlEndpointForNetwork = getPeggyGraphQlEndpointForNetwork;
exports.isDevnet = isDevnet;
exports.isMainnet = isMainnet;
exports.isTestnet = isTestnet;
exports.isTestnetOrDevnet = isTestnetOrDevnet;
