// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/bank/module/v1/module_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Module$Type extends MessageType {
  constructor() {
    super("cosmos.bank.module.v1.Module", [
      {
        no: 1,
        name: "blocked_module_accounts_override",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "authority",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "restrictions_order",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "cosmos.app.v1alpha1.module": { goImport: "github.com/cosmos/cosmos-sdk/x/bank" } });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.blockedModuleAccountsOverride = [];
    message.authority = "";
    message.restrictionsOrder = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string blocked_module_accounts_override */
        1:
          message.blockedModuleAccountsOverride.push(reader.string());
          break;
        case /* string authority */
        2:
          message.authority = reader.string();
          break;
        case /* repeated string restrictions_order */
        3:
          message.restrictionsOrder.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.blockedModuleAccountsOverride.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.blockedModuleAccountsOverride[i]);
    if (message.authority !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.authority);
    for (let i = 0; i < message.restrictionsOrder.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.restrictionsOrder[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Module = new Module$Type();
export {
  Module
};
