// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/interchain_accounts/controller/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * InterchainAccount returns the interchain account address for a given owner address on a given connection
   *
   * @generated from protobuf rpc: InterchainAccount
   */
  interchainAccount(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Params queries all parameters of the ICA controller submodule.
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
