// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/interchain_accounts/controller/v1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./controller_pb.js";
import { InterchainAccountPacketData } from "../../v1/packet_pb.js";
import { Order } from "../../../../core/channel/v1/channel_pb.js";
class MsgRegisterInterchainAccount$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.controller.v1.MsgRegisterInterchainAccount", [
      {
        no: 1,
        name: "owner",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "ordering", kind: "enum", T: () => ["ibc.core.channel.v1.Order", Order, "ORDER_"] }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["owner"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.owner = "";
    message.connectionId = "";
    message.version = "";
    message.ordering = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string owner */
        1:
          message.owner = reader.string();
          break;
        case /* string connection_id */
        2:
          message.connectionId = reader.string();
          break;
        case /* string version */
        3:
          message.version = reader.string();
          break;
        case /* ibc.core.channel.v1.Order ordering */
        4:
          message.ordering = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.owner !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.owner);
    if (message.connectionId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.connectionId);
    if (message.version !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.version);
    if (message.ordering !== 0)
      writer.tag(4, WireType.Varint).int32(message.ordering);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRegisterInterchainAccount = new MsgRegisterInterchainAccount$Type();
class MsgRegisterInterchainAccountResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.controller.v1.MsgRegisterInterchainAccountResponse", [
      {
        no: 1,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channelId = "";
    message.portId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string channel_id */
        1:
          message.channelId = reader.string();
          break;
        case /* string port_id */
        2:
          message.portId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channelId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.channelId);
    if (message.portId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.portId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRegisterInterchainAccountResponse = new MsgRegisterInterchainAccountResponse$Type();
class MsgSendTx$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.controller.v1.MsgSendTx", [
      {
        no: 1,
        name: "owner",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "packet_data", kind: "message", T: () => InterchainAccountPacketData, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "relative_timeout",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["owner"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.owner = "";
    message.connectionId = "";
    message.relativeTimeout = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string owner */
        1:
          message.owner = reader.string();
          break;
        case /* string connection_id */
        2:
          message.connectionId = reader.string();
          break;
        case /* ibc.applications.interchain_accounts.v1.InterchainAccountPacketData packet_data */
        3:
          message.packetData = InterchainAccountPacketData.internalBinaryRead(reader, reader.uint32(), options, message.packetData);
          break;
        case /* uint64 relative_timeout */
        4:
          message.relativeTimeout = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.owner !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.owner);
    if (message.connectionId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.connectionId);
    if (message.packetData)
      InterchainAccountPacketData.internalBinaryWrite(message.packetData, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.relativeTimeout !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.relativeTimeout);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSendTx = new MsgSendTx$Type();
class MsgSendTxResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.controller.v1.MsgSendTxResponse", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSendTxResponse = new MsgSendTxResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.controller.v1.MsgUpdateParams", [
      {
        no: 1,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string signer */
        1:
          message.signer = reader.string();
          break;
        case /* ibc.applications.interchain_accounts.controller.v1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signer !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.signer);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.controller.v1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
const Msg = new ServiceType("ibc.applications.interchain_accounts.controller.v1.Msg", [
  { name: "RegisterInterchainAccount", options: {}, I: MsgRegisterInterchainAccount, O: MsgRegisterInterchainAccountResponse },
  { name: "SendTx", options: {}, I: MsgSendTx, O: MsgSendTxResponse },
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgRegisterInterchainAccount,
  MsgRegisterInterchainAccountResponse,
  MsgSendTx,
  MsgSendTxResponse,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
