// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/interchain_accounts/controller/v1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * RegisterInterchainAccount defines a rpc handler for MsgRegisterInterchainAccount.
   *
   * @generated from protobuf rpc: RegisterInterchainAccount
   */
  registerInterchainAccount(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * SendTx defines a rpc handler for MsgSendTx.
   *
   * @generated from protobuf rpc: SendTx
   */
  sendTx(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateParams defines a rpc handler for MsgUpdateParams.
   *
   * @generated from protobuf rpc: UpdateParams
   */
  updateParams(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
