// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/gov/v1/gov_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Duration } from "../../../google/protobuf/duration_pb.js";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
import { Coin } from "../../base/v1beta1/coin_pb.js";
var VoteOption = /* @__PURE__ */ ((VoteOption2) => {
  VoteOption2[VoteOption2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  VoteOption2[VoteOption2["YES"] = 1] = "YES";
  VoteOption2[VoteOption2["ABSTAIN"] = 2] = "ABSTAIN";
  VoteOption2[VoteOption2["NO"] = 3] = "NO";
  VoteOption2[VoteOption2["NO_WITH_VETO"] = 4] = "NO_WITH_VETO";
  return VoteOption2;
})(VoteOption || {});
var ProposalStatus = /* @__PURE__ */ ((ProposalStatus2) => {
  ProposalStatus2[ProposalStatus2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  ProposalStatus2[ProposalStatus2["DEPOSIT_PERIOD"] = 1] = "DEPOSIT_PERIOD";
  ProposalStatus2[ProposalStatus2["VOTING_PERIOD"] = 2] = "VOTING_PERIOD";
  ProposalStatus2[ProposalStatus2["PASSED"] = 3] = "PASSED";
  ProposalStatus2[ProposalStatus2["REJECTED"] = 4] = "REJECTED";
  ProposalStatus2[ProposalStatus2["FAILED"] = 5] = "FAILED";
  return ProposalStatus2;
})(ProposalStatus || {});
class WeightedVoteOption$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1.WeightedVoteOption", [
      { no: 1, name: "option", kind: "enum", T: () => ["cosmos.gov.v1.VoteOption", VoteOption, "VOTE_OPTION_"] },
      { no: 2, name: "weight", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.option = 0;
    message.weight = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.gov.v1.VoteOption option */
        1:
          message.option = reader.int32();
          break;
        case /* string weight */
        2:
          message.weight = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.option !== 0)
      writer.tag(1, WireType.Varint).int32(message.option);
    if (message.weight !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.weight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const WeightedVoteOption = new WeightedVoteOption$Type();
class Deposit$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1.Deposit", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "depositor", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "amount", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.depositor = "";
    message.amount = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string depositor */
        2:
          message.depositor = reader.string();
          break;
        case /* repeated cosmos.base.v1beta1.Coin amount */
        3:
          message.amount.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.depositor !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.depositor);
    for (let i = 0; i < message.amount.length; i++)
      Coin.internalBinaryWrite(message.amount[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Deposit = new Deposit$Type();
class Proposal$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1.Proposal", [
      {
        no: 1,
        name: "id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "messages", kind: "message", repeat: 2, T: () => Any },
      { no: 3, name: "status", kind: "enum", T: () => ["cosmos.gov.v1.ProposalStatus", ProposalStatus, "PROPOSAL_STATUS_"] },
      { no: 4, name: "final_tally_result", kind: "message", T: () => TallyResult },
      { no: 5, name: "submit_time", kind: "message", T: () => Timestamp, options: { "gogoproto.stdtime": true } },
      { no: 6, name: "deposit_end_time", kind: "message", T: () => Timestamp, options: { "gogoproto.stdtime": true } },
      { no: 7, name: "total_deposit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 8, name: "voting_start_time", kind: "message", T: () => Timestamp, options: { "gogoproto.stdtime": true } },
      { no: 9, name: "voting_end_time", kind: "message", T: () => Timestamp, options: { "gogoproto.stdtime": true } },
      {
        no: 10,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 12,
        name: "summary",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 13, name: "proposer", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 14,
        name: "expedited",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 15,
        name: "failed_reason",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.id = 0n;
    message.messages = [];
    message.status = 0;
    message.totalDeposit = [];
    message.metadata = "";
    message.title = "";
    message.summary = "";
    message.proposer = "";
    message.expedited = false;
    message.failedReason = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 id */
        1:
          message.id = reader.uint64().toBigInt();
          break;
        case /* repeated google.protobuf.Any messages */
        2:
          message.messages.push(Any.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.gov.v1.ProposalStatus status */
        3:
          message.status = reader.int32();
          break;
        case /* cosmos.gov.v1.TallyResult final_tally_result */
        4:
          message.finalTallyResult = TallyResult.internalBinaryRead(reader, reader.uint32(), options, message.finalTallyResult);
          break;
        case /* google.protobuf.Timestamp submit_time */
        5:
          message.submitTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.submitTime);
          break;
        case /* google.protobuf.Timestamp deposit_end_time */
        6:
          message.depositEndTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.depositEndTime);
          break;
        case /* repeated cosmos.base.v1beta1.Coin total_deposit */
        7:
          message.totalDeposit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* google.protobuf.Timestamp voting_start_time */
        8:
          message.votingStartTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.votingStartTime);
          break;
        case /* google.protobuf.Timestamp voting_end_time */
        9:
          message.votingEndTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.votingEndTime);
          break;
        case /* string metadata */
        10:
          message.metadata = reader.string();
          break;
        case /* string title */
        11:
          message.title = reader.string();
          break;
        case /* string summary */
        12:
          message.summary = reader.string();
          break;
        case /* string proposer */
        13:
          message.proposer = reader.string();
          break;
        case /* bool expedited */
        14:
          message.expedited = reader.bool();
          break;
        case /* string failed_reason */
        15:
          message.failedReason = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.id !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.id);
    for (let i = 0; i < message.messages.length; i++)
      Any.internalBinaryWrite(message.messages[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.status !== 0)
      writer.tag(3, WireType.Varint).int32(message.status);
    if (message.finalTallyResult)
      TallyResult.internalBinaryWrite(message.finalTallyResult, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.submitTime)
      Timestamp.internalBinaryWrite(message.submitTime, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.depositEndTime)
      Timestamp.internalBinaryWrite(message.depositEndTime, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.totalDeposit.length; i++)
      Coin.internalBinaryWrite(message.totalDeposit[i], writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.votingStartTime)
      Timestamp.internalBinaryWrite(message.votingStartTime, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    if (message.votingEndTime)
      Timestamp.internalBinaryWrite(message.votingEndTime, writer.tag(9, WireType.LengthDelimited).fork(), options).join();
    if (message.metadata !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.metadata);
    if (message.title !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.title);
    if (message.summary !== "")
      writer.tag(12, WireType.LengthDelimited).string(message.summary);
    if (message.proposer !== "")
      writer.tag(13, WireType.LengthDelimited).string(message.proposer);
    if (message.expedited !== false)
      writer.tag(14, WireType.Varint).bool(message.expedited);
    if (message.failedReason !== "")
      writer.tag(15, WireType.LengthDelimited).string(message.failedReason);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Proposal = new Proposal$Type();
class TallyResult$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1.TallyResult", [
      { no: 1, name: "yes_count", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Int" } },
      { no: 2, name: "abstain_count", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Int" } },
      { no: 3, name: "no_count", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Int" } },
      { no: 4, name: "no_with_veto_count", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Int" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.yesCount = "";
    message.abstainCount = "";
    message.noCount = "";
    message.noWithVetoCount = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string yes_count */
        1:
          message.yesCount = reader.string();
          break;
        case /* string abstain_count */
        2:
          message.abstainCount = reader.string();
          break;
        case /* string no_count */
        3:
          message.noCount = reader.string();
          break;
        case /* string no_with_veto_count */
        4:
          message.noWithVetoCount = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.yesCount !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.yesCount);
    if (message.abstainCount !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.abstainCount);
    if (message.noCount !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.noCount);
    if (message.noWithVetoCount !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.noWithVetoCount);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TallyResult = new TallyResult$Type();
class Vote$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1.Vote", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "voter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 4, name: "options", kind: "message", repeat: 2, T: () => WeightedVoteOption },
      {
        no: 5,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.voter = "";
    message.options = [];
    message.metadata = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string voter */
        2:
          message.voter = reader.string();
          break;
        case /* repeated cosmos.gov.v1.WeightedVoteOption options */
        4:
          message.options.push(WeightedVoteOption.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string metadata */
        5:
          message.metadata = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.voter !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.voter);
    for (let i = 0; i < message.options.length; i++)
      WeightedVoteOption.internalBinaryWrite(message.options[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.metadata !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.metadata);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Vote = new Vote$Type();
class DepositParams$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1.DepositParams", [
      { no: 1, name: "min_deposit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "min_deposit,omitempty" } },
      { no: 2, name: "max_deposit_period", kind: "message", T: () => Duration, options: { "gogoproto.jsontag": "max_deposit_period,omitempty", "gogoproto.stdduration": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.minDeposit = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin min_deposit */
        1:
          message.minDeposit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* google.protobuf.Duration max_deposit_period */
        2:
          message.maxDepositPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.maxDepositPeriod);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.minDeposit.length; i++)
      Coin.internalBinaryWrite(message.minDeposit[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.maxDepositPeriod)
      Duration.internalBinaryWrite(message.maxDepositPeriod, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DepositParams = new DepositParams$Type();
class VotingParams$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1.VotingParams", [
      { no: 1, name: "voting_period", kind: "message", T: () => Duration, options: { "gogoproto.stdduration": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Duration voting_period */
        1:
          message.votingPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.votingPeriod);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.votingPeriod)
      Duration.internalBinaryWrite(message.votingPeriod, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const VotingParams = new VotingParams$Type();
class TallyParams$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1.TallyParams", [
      { no: 1, name: "quorum", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 2, name: "threshold", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 3, name: "veto_threshold", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.quorum = "";
    message.threshold = "";
    message.vetoThreshold = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string quorum */
        1:
          message.quorum = reader.string();
          break;
        case /* string threshold */
        2:
          message.threshold = reader.string();
          break;
        case /* string veto_threshold */
        3:
          message.vetoThreshold = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.quorum !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.quorum);
    if (message.threshold !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.threshold);
    if (message.vetoThreshold !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.vetoThreshold);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TallyParams = new TallyParams$Type();
class Params$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1.Params", [
      { no: 1, name: "min_deposit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "max_deposit_period", kind: "message", T: () => Duration, options: { "gogoproto.stdduration": true } },
      { no: 3, name: "voting_period", kind: "message", T: () => Duration, options: { "gogoproto.stdduration": true } },
      { no: 4, name: "quorum", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 5, name: "threshold", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 6, name: "veto_threshold", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 7, name: "min_initial_deposit_ratio", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 8, name: "proposal_cancel_ratio", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 9, name: "proposal_cancel_dest", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 10, name: "expedited_voting_period", kind: "message", T: () => Duration, options: { "gogoproto.stdduration": true } },
      { no: 11, name: "expedited_threshold", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 12, name: "expedited_min_deposit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 13,
        name: "burn_vote_quorum",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 14,
        name: "burn_proposal_deposit_prevote",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 15,
        name: "burn_vote_veto",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 16, name: "min_deposit_ratio", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.Dec" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.minDeposit = [];
    message.quorum = "";
    message.threshold = "";
    message.vetoThreshold = "";
    message.minInitialDepositRatio = "";
    message.proposalCancelRatio = "";
    message.proposalCancelDest = "";
    message.expeditedThreshold = "";
    message.expeditedMinDeposit = [];
    message.burnVoteQuorum = false;
    message.burnProposalDepositPrevote = false;
    message.burnVoteVeto = false;
    message.minDepositRatio = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin min_deposit */
        1:
          message.minDeposit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* google.protobuf.Duration max_deposit_period */
        2:
          message.maxDepositPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.maxDepositPeriod);
          break;
        case /* google.protobuf.Duration voting_period */
        3:
          message.votingPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.votingPeriod);
          break;
        case /* string quorum */
        4:
          message.quorum = reader.string();
          break;
        case /* string threshold */
        5:
          message.threshold = reader.string();
          break;
        case /* string veto_threshold */
        6:
          message.vetoThreshold = reader.string();
          break;
        case /* string min_initial_deposit_ratio */
        7:
          message.minInitialDepositRatio = reader.string();
          break;
        case /* string proposal_cancel_ratio */
        8:
          message.proposalCancelRatio = reader.string();
          break;
        case /* string proposal_cancel_dest */
        9:
          message.proposalCancelDest = reader.string();
          break;
        case /* google.protobuf.Duration expedited_voting_period */
        10:
          message.expeditedVotingPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.expeditedVotingPeriod);
          break;
        case /* string expedited_threshold */
        11:
          message.expeditedThreshold = reader.string();
          break;
        case /* repeated cosmos.base.v1beta1.Coin expedited_min_deposit */
        12:
          message.expeditedMinDeposit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* bool burn_vote_quorum */
        13:
          message.burnVoteQuorum = reader.bool();
          break;
        case /* bool burn_proposal_deposit_prevote */
        14:
          message.burnProposalDepositPrevote = reader.bool();
          break;
        case /* bool burn_vote_veto */
        15:
          message.burnVoteVeto = reader.bool();
          break;
        case /* string min_deposit_ratio */
        16:
          message.minDepositRatio = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.minDeposit.length; i++)
      Coin.internalBinaryWrite(message.minDeposit[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.maxDepositPeriod)
      Duration.internalBinaryWrite(message.maxDepositPeriod, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.votingPeriod)
      Duration.internalBinaryWrite(message.votingPeriod, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.quorum !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.quorum);
    if (message.threshold !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.threshold);
    if (message.vetoThreshold !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.vetoThreshold);
    if (message.minInitialDepositRatio !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.minInitialDepositRatio);
    if (message.proposalCancelRatio !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.proposalCancelRatio);
    if (message.proposalCancelDest !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.proposalCancelDest);
    if (message.expeditedVotingPeriod)
      Duration.internalBinaryWrite(message.expeditedVotingPeriod, writer.tag(10, WireType.LengthDelimited).fork(), options).join();
    if (message.expeditedThreshold !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.expeditedThreshold);
    for (let i = 0; i < message.expeditedMinDeposit.length; i++)
      Coin.internalBinaryWrite(message.expeditedMinDeposit[i], writer.tag(12, WireType.LengthDelimited).fork(), options).join();
    if (message.burnVoteQuorum !== false)
      writer.tag(13, WireType.Varint).bool(message.burnVoteQuorum);
    if (message.burnProposalDepositPrevote !== false)
      writer.tag(14, WireType.Varint).bool(message.burnProposalDepositPrevote);
    if (message.burnVoteVeto !== false)
      writer.tag(15, WireType.Varint).bool(message.burnVoteVeto);
    if (message.minDepositRatio !== "")
      writer.tag(16, WireType.LengthDelimited).string(message.minDepositRatio);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
export {
  Deposit,
  DepositParams,
  Params,
  Proposal,
  ProposalStatus,
  TallyParams,
  TallyResult,
  Vote,
  VoteOption,
  VotingParams,
  WeightedVoteOption
};
