// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/gov/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Constitution queries the chain's constitution.
   *
   * @generated from protobuf rpc: Constitution
   */
  constitution(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Proposal queries proposal details based on ProposalID.
   *
   * @generated from protobuf rpc: Proposal
   */
  proposal(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Proposals queries all proposals based on given status.
   *
   * @generated from protobuf rpc: Proposals
   */
  proposals(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Vote queries voted information based on proposalID, voterAddr.
   *
   * @generated from protobuf rpc: Vote
   */
  vote(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Votes queries votes of a given proposal.
   *
   * @generated from protobuf rpc: Votes
   */
  votes(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Params queries all parameters of the gov module.
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Deposit queries single deposit information based on proposalID, depositAddr.
   *
   * @generated from protobuf rpc: Deposit
   */
  deposit(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Deposits queries all deposits of a single proposal.
   *
   * @generated from protobuf rpc: Deposits
   */
  deposits(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TallyResult queries the tally of a proposal vote.
   *
   * @generated from protobuf rpc: TallyResult
   */
  tallyResult(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
