// @injectivelabs/mito-proto-ts-v2/esm/generated/goadesign_goagen_mito_api_pb.client.js
import { MitoAPI } from "./goadesign_goagen_mito_api_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MitoAPIClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = MitoAPI.typeName;
    this.methods = MitoAPI.methods;
    this.options = MitoAPI.options;
  }
  /**
   * List all vaults
   *
   * @generated from protobuf rpc: GetVaults
   */
  getVaults(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List a vault by contract address
   *
   * @generated from protobuf rpc: GetVault
   */
  getVault(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * LPTokenPriceChart implements LPTokenPriceChart.
   *
   * @generated from protobuf rpc: LPTokenPriceChart
   */
  lPTokenPriceChart(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TVLChart implements TVLChart.
   *
   * @generated from protobuf rpc: TVLChart
   */
  tVLChart(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get list of vaults that user subscribes to
   *
   * @generated from protobuf rpc: VaultsByHolderAddress
   */
  vaultsByHolderAddress(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get list of address that holds this LP token
   *
   * @generated from protobuf rpc: LPHolders
   */
  lPHolders(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Portfolio summary of an account
   *
   * @generated from protobuf rpc: Portfolio
   */
  portfolio(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Leaderboard ranking
   *
   * @generated from protobuf rpc: Leaderboard
   */
  leaderboard(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get leaderboard epochs
   *
   * @generated from protobuf rpc: LeaderboardEpochs
   */
  leaderboardEpochs(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Return historical transfers (i.e Subscribe - Deposit / Redeem - withdrawal)
   *
   * @generated from protobuf rpc: TransfersHistory
   */
  transfersHistory(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * [Staking] Return staking pools
   *
   * @generated from protobuf rpc: GetStakingPools
   */
  getStakingPools(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * [Staking] Return staking rewards
   *
   * @generated from protobuf rpc: StakingRewardByAccount
   */
  stakingRewardByAccount(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * [Staking] Return staking historical activities
   *
   * @generated from protobuf rpc: StakingHistory
   */
  stakingHistory(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * [Staking] Return the amount of a denom staked at a given height
   *
   * @generated from protobuf rpc: StakingAmountAtHeight
   */
  stakingAmountAtHeight(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream up-to-date vault transfers
   *
   * @generated from protobuf rpc: StreamTransfers
   */
  streamTransfers(input, options) {
    const method = this.methods[14], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Stream up-to-date vault data
   *
   * @generated from protobuf rpc: StreamVault
   */
  streamVault(input, options) {
    const method = this.methods[15], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Stream up-to-date vault subscription
   *
   * @generated from protobuf rpc: StreamHolderSubscription
   */
  streamHolderSubscription(input, options) {
    const method = this.methods[16], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * [Staking] Stream staking update of an account (staker)
   *
   * @generated from protobuf rpc: StreamStakingRewardByAccount
   */
  streamStakingRewardByAccount(input, options) {
    const method = this.methods[17], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * [Staking] Stream historical staking activities
   *
   * @generated from protobuf rpc: StreamHistoricalStaking
   */
  streamHistoricalStaking(input, options) {
    const method = this.methods[18], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Return binary version and health check
   *
   * @generated from protobuf rpc: Health
   */
  health(input, options) {
    const method = this.methods[19], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Return contract execution log (mainly for debug purpose)
   *
   * @generated from protobuf rpc: Execution
   */
  execution(input, options) {
    const method = this.methods[20], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Return mission status of an account
   *
   * @generated from protobuf rpc: Missions
   */
  missions(input, options) {
    const method = this.methods[21], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Return ALL-TIME mission leaderboard across mito users
   *
   * @generated from protobuf rpc: MissionLeaderboard
   */
  missionLeaderboard(input, options) {
    const method = this.methods[22], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Return all IDOs that are taking place
   *
   * @generated from protobuf rpc: ListIDOs
   */
  listIDOs(input, options) {
    const method = this.methods[23], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Return a single IDO
   *
   * @generated from protobuf rpc: GetIDO
   */
  getIDO(input, options) {
    const method = this.methods[24], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Return list of subscribers in the IDOs (not whitelisted members)
   *
   * @generated from protobuf rpc: GetIDOSubscribers
   */
  getIDOSubscribers(input, options) {
    const method = this.methods[25], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Subscription detail in the IDO
   *
   * @generated from protobuf rpc: GetIDOSubscription
   */
  getIDOSubscription(input, options) {
    const method = this.methods[26], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Subscription activities in the IDO
   *
   * @generated from protobuf rpc: GetIDOActivities
   */
  getIDOActivities(input, options) {
    const method = this.methods[27], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get white list accounts of a given IDO
   *
   * @generated from protobuf rpc: GetWhitelist
   */
  getWhitelist(input, options) {
    const method = this.methods[28], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get token metadata that current backend is processing, mainly for debug
   * purpose
   *
   * @generated from protobuf rpc: TokenMetadata
   */
  tokenMetadata(input, options) {
    const method = this.methods[29], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get claim references of a given IDO and account
   *
   * @generated from protobuf rpc: GetClaimReferences
   */
  getClaimReferences(input, options) {
    const method = this.methods[30], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MitoAPIClient
};
