// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/reflection/v1/reflection_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { FileDescriptorProto } from "../../../google/protobuf/descriptor_pb.js";
class FileDescriptorsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.reflection.v1.FileDescriptorsRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FileDescriptorsRequest = new FileDescriptorsRequest$Type();
class FileDescriptorsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.reflection.v1.FileDescriptorsResponse", [
      { no: 1, name: "files", kind: "message", repeat: 2, T: () => FileDescriptorProto }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.files = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated google.protobuf.FileDescriptorProto files */
        1:
          message.files.push(FileDescriptorProto.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.files.length; i++)
      FileDescriptorProto.internalBinaryWrite(message.files[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FileDescriptorsResponse = new FileDescriptorsResponse$Type();
const ReflectionService = new ServiceType("cosmos.reflection.v1.ReflectionService", [
  { name: "FileDescriptors", options: { "cosmos.query.v1.module_query_safe": false }, I: FileDescriptorsRequest, O: FileDescriptorsResponse }
]);
export {
  FileDescriptorsRequest,
  FileDescriptorsResponse,
  ReflectionService
};
