// @injectivelabs/indexer-proto-ts-v2/esm/index.js
import { EventProviderAPIClient } from "./generated/event_provider_api_pb.client.js";
import { HealthClient } from "./generated/health_pb.client.js";
import { InjectiveAccountsRPCClient } from "./generated/injective_accounts_rpc_pb.client.js";
import { InjectiveArchiverRPCClient } from "./generated/injective_archiver_rpc_pb.client.js";
import { InjectiveAuctionRPCClient } from "./generated/injective_auction_rpc_pb.client.js";
import { InjectiveCampaignRPCClient } from "./generated/injective_campaign_rpc_pb.client.js";
import { InjectiveChartRPCClient } from "./generated/injective_chart_rpc_pb.client.js";
import { InjectiveDerivativeExchangeRPCClient } from "./generated/injective_derivative_exchange_rpc_pb.client.js";
import { InjectiveExchangeRPCClient } from "./generated/injective_exchange_rpc_pb.client.js";
import { InjectiveExplorerRPCClient } from "./generated/injective_explorer_rpc_pb.client.js";
import { InjectiveInsuranceRPCClient } from "./generated/injective_insurance_rpc_pb.client.js";
import { InjectiveMegavaultRPCClient } from "./generated/injective_megavault_rpc_pb.client.js";
import { InjectiveMetaRPCClient } from "./generated/injective_meta_rpc_pb.client.js";
import { InjectiveOracleRPCClient } from "./generated/injective_oracle_rpc_pb.client.js";
import { InjectivePortfolioRPCClient } from "./generated/injective_portfolio_rpc_pb.client.js";
import { InjectiveReferralRPCClient } from "./generated/injective_referral_rpc_pb.client.js";
import { InjectiveSpotExchangeRPCClient } from "./generated/injective_spot_exchange_rpc_pb.client.js";
import { InjectiveTradingRPCClient } from "./generated/injective_trading_rpc_pb.client.js";
import * as EventProviderApiPb from "./generated/event_provider_api_pb.js";
import * as HealthPb from "./generated/health_pb.js";
import * as InjectiveAccountsRpcPb from "./generated/injective_accounts_rpc_pb.js";
import * as InjectiveArchiverRpcPb from "./generated/injective_archiver_rpc_pb.js";
import * as InjectiveAuctionRpcPb from "./generated/injective_auction_rpc_pb.js";
import * as InjectiveCampaignRpcPb from "./generated/injective_campaign_rpc_pb.js";
import * as InjectiveChartRpcPb from "./generated/injective_chart_rpc_pb.js";
import * as InjectiveDerivativeExchangeRpcPb from "./generated/injective_derivative_exchange_rpc_pb.js";
import * as InjectiveExchangeRpcPb from "./generated/injective_exchange_rpc_pb.js";
import * as InjectiveExplorerRpcPb from "./generated/injective_explorer_rpc_pb.js";
import * as InjectiveInsuranceRpcPb from "./generated/injective_insurance_rpc_pb.js";
import * as InjectiveMegavaultRpcPb from "./generated/injective_megavault_rpc_pb.js";
import * as InjectiveMetaRpcPb from "./generated/injective_meta_rpc_pb.js";
import * as InjectiveOracleRpcPb from "./generated/injective_oracle_rpc_pb.js";
import * as InjectivePortfolioRpcPb from "./generated/injective_portfolio_rpc_pb.js";
import * as InjectiveReferralRpcPb from "./generated/injective_referral_rpc_pb.js";
import * as InjectiveSpotExchangeRpcPb from "./generated/injective_spot_exchange_rpc_pb.js";
import * as InjectiveTradingRpcPb from "./generated/injective_trading_rpc_pb.js";
export {
  EventProviderAPIClient,
  EventProviderApiPb,
  HealthClient,
  HealthPb,
  InjectiveAccountsRPCClient,
  InjectiveAccountsRpcPb,
  InjectiveArchiverRPCClient,
  InjectiveArchiverRpcPb,
  InjectiveAuctionRPCClient,
  InjectiveAuctionRpcPb,
  InjectiveCampaignRPCClient,
  InjectiveCampaignRpcPb,
  InjectiveChartRPCClient,
  InjectiveChartRpcPb,
  InjectiveDerivativeExchangeRPCClient,
  InjectiveDerivativeExchangeRpcPb,
  InjectiveExchangeRPCClient,
  InjectiveExchangeRpcPb,
  InjectiveExplorerRPCClient,
  InjectiveExplorerRpcPb,
  InjectiveInsuranceRPCClient,
  InjectiveInsuranceRpcPb,
  InjectiveMegavaultRPCClient,
  InjectiveMegavaultRpcPb,
  InjectiveMetaRPCClient,
  InjectiveMetaRpcPb,
  InjectiveOracleRPCClient,
  InjectiveOracleRpcPb,
  InjectivePortfolioRPCClient,
  InjectivePortfolioRpcPb,
  InjectiveReferralRPCClient,
  InjectiveReferralRpcPb,
  InjectiveSpotExchangeRPCClient,
  InjectiveSpotExchangeRpcPb,
  InjectiveTradingRPCClient,
  InjectiveTradingRpcPb
};
