// @injectivelabs/core-proto-ts-v2/esm/generated/packetforward/v1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class GenesisState$Type extends MessageType {
  constructor() {
    super("packetforward.v1.GenesisState", [
      { no: 2, name: "in_flight_packets", kind: "map", K: 9, V: { kind: "message", T: () => InFlightPacket }, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"in_flight_packets"' } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.inFlightPackets = {};
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* map<string, packetforward.v1.InFlightPacket> in_flight_packets */
        2:
          this.binaryReadMap2(message.inFlightPackets, reader, options);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  binaryReadMap2(map, reader, options) {
    let len = reader.uint32(), end = reader.pos + len, key, val;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case 1:
          key = reader.string();
          break;
        case 2:
          val = InFlightPacket.internalBinaryRead(reader, reader.uint32(), options);
          break;
        default:
          throw new globalThis.Error("unknown map entry field for packetforward.v1.GenesisState.in_flight_packets");
      }
    }
    map[key ?? ""] = val ?? InFlightPacket.create();
  }
  internalBinaryWrite(message, writer, options) {
    for (let k of globalThis.Object.keys(message.inFlightPackets)) {
      writer.tag(2, WireType.LengthDelimited).fork().tag(1, WireType.LengthDelimited).string(k);
      writer.tag(2, WireType.LengthDelimited).fork();
      InFlightPacket.internalBinaryWrite(message.inFlightPackets[k], writer, options);
      writer.join().join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class InFlightPacket$Type extends MessageType {
  constructor() {
    super("packetforward.v1.InFlightPacket", [
      {
        no: 1,
        name: "original_sender_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "refund_channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "refund_port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "packet_src_channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "packet_src_port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "packet_timeout_timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "packet_timeout_height",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "packet_data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 9,
        name: "refund_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 10,
        name: "retries_remaining",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      {
        no: 11,
        name: "timeout",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 12,
        name: "nonrefundable",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.originalSenderAddress = "";
    message.refundChannelId = "";
    message.refundPortId = "";
    message.packetSrcChannelId = "";
    message.packetSrcPortId = "";
    message.packetTimeoutTimestamp = 0n;
    message.packetTimeoutHeight = "";
    message.packetData = new Uint8Array(0);
    message.refundSequence = 0n;
    message.retriesRemaining = 0;
    message.timeout = 0n;
    message.nonrefundable = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string original_sender_address */
        1:
          message.originalSenderAddress = reader.string();
          break;
        case /* string refund_channel_id */
        2:
          message.refundChannelId = reader.string();
          break;
        case /* string refund_port_id */
        3:
          message.refundPortId = reader.string();
          break;
        case /* string packet_src_channel_id */
        4:
          message.packetSrcChannelId = reader.string();
          break;
        case /* string packet_src_port_id */
        5:
          message.packetSrcPortId = reader.string();
          break;
        case /* uint64 packet_timeout_timestamp */
        6:
          message.packetTimeoutTimestamp = reader.uint64().toBigInt();
          break;
        case /* string packet_timeout_height */
        7:
          message.packetTimeoutHeight = reader.string();
          break;
        case /* bytes packet_data */
        8:
          message.packetData = reader.bytes();
          break;
        case /* uint64 refund_sequence */
        9:
          message.refundSequence = reader.uint64().toBigInt();
          break;
        case /* int32 retries_remaining */
        10:
          message.retriesRemaining = reader.int32();
          break;
        case /* uint64 timeout */
        11:
          message.timeout = reader.uint64().toBigInt();
          break;
        case /* bool nonrefundable */
        12:
          message.nonrefundable = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.originalSenderAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.originalSenderAddress);
    if (message.refundChannelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.refundChannelId);
    if (message.refundPortId !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.refundPortId);
    if (message.packetSrcChannelId !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.packetSrcChannelId);
    if (message.packetSrcPortId !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.packetSrcPortId);
    if (message.packetTimeoutTimestamp !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.packetTimeoutTimestamp);
    if (message.packetTimeoutHeight !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.packetTimeoutHeight);
    if (message.packetData.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.packetData);
    if (message.refundSequence !== 0n)
      writer.tag(9, WireType.Varint).uint64(message.refundSequence);
    if (message.retriesRemaining !== 0)
      writer.tag(10, WireType.Varint).int32(message.retriesRemaining);
    if (message.timeout !== 0n)
      writer.tag(11, WireType.Varint).uint64(message.timeout);
    if (message.nonrefundable !== false)
      writer.tag(12, WireType.Varint).bool(message.nonrefundable);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const InFlightPacket = new InFlightPacket$Type();
export {
  GenesisState,
  InFlightPacket
};
