// @injectivelabs/core-proto-ts-v2/esm/generated/tendermint/blocksync/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ExtendedCommit } from "../types/types_pb.js";
import { Block } from "../types/block_pb.js";
class BlockRequest$Type extends MessageType {
  constructor() {
    super("tendermint.blocksync.BlockRequest", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BlockRequest = new BlockRequest$Type();
class NoBlockResponse$Type extends MessageType {
  constructor() {
    super("tendermint.blocksync.NoBlockResponse", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const NoBlockResponse = new NoBlockResponse$Type();
class BlockResponse$Type extends MessageType {
  constructor() {
    super("tendermint.blocksync.BlockResponse", [
      { no: 1, name: "block", kind: "message", T: () => Block },
      { no: 2, name: "ext_commit", kind: "message", T: () => ExtendedCommit }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.Block block */
        1:
          message.block = Block.internalBinaryRead(reader, reader.uint32(), options, message.block);
          break;
        case /* tendermint.types.ExtendedCommit ext_commit */
        2:
          message.extCommit = ExtendedCommit.internalBinaryRead(reader, reader.uint32(), options, message.extCommit);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.block)
      Block.internalBinaryWrite(message.block, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.extCommit)
      ExtendedCommit.internalBinaryWrite(message.extCommit, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BlockResponse = new BlockResponse$Type();
class StatusRequest$Type extends MessageType {
  constructor() {
    super("tendermint.blocksync.StatusRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StatusRequest = new StatusRequest$Type();
class StatusResponse$Type extends MessageType {
  constructor() {
    super("tendermint.blocksync.StatusResponse", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "base",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    message.base = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        case /* int64 base */
        2:
          message.base = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    if (message.base !== 0n)
      writer.tag(2, WireType.Varint).int64(message.base);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StatusResponse = new StatusResponse$Type();
class Message$Type extends MessageType {
  constructor() {
    super("tendermint.blocksync.Message", [
      { no: 1, name: "block_request", kind: "message", oneof: "sum", T: () => BlockRequest },
      { no: 2, name: "no_block_response", kind: "message", oneof: "sum", T: () => NoBlockResponse },
      { no: 3, name: "block_response", kind: "message", oneof: "sum", T: () => BlockResponse },
      { no: 4, name: "status_request", kind: "message", oneof: "sum", T: () => StatusRequest },
      { no: 5, name: "status_response", kind: "message", oneof: "sum", T: () => StatusResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sum = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.blocksync.BlockRequest block_request */
        1:
          message.sum = {
            oneofKind: "blockRequest",
            blockRequest: BlockRequest.internalBinaryRead(reader, reader.uint32(), options, message.sum.blockRequest)
          };
          break;
        case /* tendermint.blocksync.NoBlockResponse no_block_response */
        2:
          message.sum = {
            oneofKind: "noBlockResponse",
            noBlockResponse: NoBlockResponse.internalBinaryRead(reader, reader.uint32(), options, message.sum.noBlockResponse)
          };
          break;
        case /* tendermint.blocksync.BlockResponse block_response */
        3:
          message.sum = {
            oneofKind: "blockResponse",
            blockResponse: BlockResponse.internalBinaryRead(reader, reader.uint32(), options, message.sum.blockResponse)
          };
          break;
        case /* tendermint.blocksync.StatusRequest status_request */
        4:
          message.sum = {
            oneofKind: "statusRequest",
            statusRequest: StatusRequest.internalBinaryRead(reader, reader.uint32(), options, message.sum.statusRequest)
          };
          break;
        case /* tendermint.blocksync.StatusResponse status_response */
        5:
          message.sum = {
            oneofKind: "statusResponse",
            statusResponse: StatusResponse.internalBinaryRead(reader, reader.uint32(), options, message.sum.statusResponse)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sum.oneofKind === "blockRequest")
      BlockRequest.internalBinaryWrite(message.sum.blockRequest, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "noBlockResponse")
      NoBlockResponse.internalBinaryWrite(message.sum.noBlockResponse, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "blockResponse")
      BlockResponse.internalBinaryWrite(message.sum.blockResponse, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "statusRequest")
      StatusRequest.internalBinaryWrite(message.sum.statusRequest, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "statusResponse")
      StatusResponse.internalBinaryWrite(message.sum.statusResponse, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Message = new Message$Type();
export {
  BlockRequest,
  BlockResponse,
  Message,
  NoBlockResponse,
  StatusRequest,
  StatusResponse
};
