// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/slashing/v1beta1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ValidatorSigningInfo } from "./slashing_pb.js";
import { Params } from "./slashing_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("cosmos.slashing.v1beta1.GenesisState", [
      { no: 1, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "signing_infos", kind: "message", repeat: 2, T: () => SigningInfo, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 3, name: "missed_blocks", kind: "message", repeat: 2, T: () => ValidatorMissedBlocks, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signingInfos = [];
    message.missedBlocks = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.slashing.v1beta1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        case /* repeated cosmos.slashing.v1beta1.SigningInfo signing_infos */
        2:
          message.signingInfos.push(SigningInfo.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.slashing.v1beta1.ValidatorMissedBlocks missed_blocks */
        3:
          message.missedBlocks.push(ValidatorMissedBlocks.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.signingInfos.length; i++)
      SigningInfo.internalBinaryWrite(message.signingInfos[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.missedBlocks.length; i++)
      ValidatorMissedBlocks.internalBinaryWrite(message.missedBlocks[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class SigningInfo$Type extends MessageType {
  constructor() {
    super("cosmos.slashing.v1beta1.SigningInfo", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ConsensusAddressString" } },
      { no: 2, name: "validator_signing_info", kind: "message", T: () => ValidatorSigningInfo, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* cosmos.slashing.v1beta1.ValidatorSigningInfo validator_signing_info */
        2:
          message.validatorSigningInfo = ValidatorSigningInfo.internalBinaryRead(reader, reader.uint32(), options, message.validatorSigningInfo);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.validatorSigningInfo)
      ValidatorSigningInfo.internalBinaryWrite(message.validatorSigningInfo, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SigningInfo = new SigningInfo$Type();
class ValidatorMissedBlocks$Type extends MessageType {
  constructor() {
    super("cosmos.slashing.v1beta1.ValidatorMissedBlocks", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ConsensusAddressString" } },
      { no: 2, name: "missed_blocks", kind: "message", repeat: 2, T: () => MissedBlock, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.missedBlocks = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* repeated cosmos.slashing.v1beta1.MissedBlock missed_blocks */
        2:
          message.missedBlocks.push(MissedBlock.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    for (let i = 0; i < message.missedBlocks.length; i++)
      MissedBlock.internalBinaryWrite(message.missedBlocks[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ValidatorMissedBlocks = new ValidatorMissedBlocks$Type();
class MissedBlock$Type extends MessageType {
  constructor() {
    super("cosmos.slashing.v1beta1.MissedBlock", [
      {
        no: 1,
        name: "index",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "missed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.index = 0n;
    message.missed = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 index */
        1:
          message.index = reader.int64().toBigInt();
          break;
        case /* bool missed */
        2:
          message.missed = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.index !== 0n)
      writer.tag(1, WireType.Varint).int64(message.index);
    if (message.missed !== false)
      writer.tag(2, WireType.Varint).bool(message.missed);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MissedBlock = new MissedBlock$Type();
export {
  GenesisState,
  MissedBlock,
  SigningInfo,
  ValidatorMissedBlocks
};
