// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/slashing/v1beta1/slashing_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Duration } from "../../../google/protobuf/duration_pb.js";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
class ValidatorSigningInfo$Type extends MessageType {
  constructor() {
    super("cosmos.slashing.v1beta1.ValidatorSigningInfo", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ConsensusAddressString" } },
      {
        no: 2,
        name: "start_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "index_offset",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "jailed_until", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      {
        no: 5,
        name: "tombstoned",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 6,
        name: "missed_blocks_counter",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.startHeight = 0n;
    message.indexOffset = 0n;
    message.tombstoned = false;
    message.missedBlocksCounter = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* int64 start_height */
        2:
          message.startHeight = reader.int64().toBigInt();
          break;
        case /* int64 index_offset */
        3:
          message.indexOffset = reader.int64().toBigInt();
          break;
        case /* google.protobuf.Timestamp jailed_until */
        4:
          message.jailedUntil = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.jailedUntil);
          break;
        case /* bool tombstoned */
        5:
          message.tombstoned = reader.bool();
          break;
        case /* int64 missed_blocks_counter */
        6:
          message.missedBlocksCounter = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.startHeight !== 0n)
      writer.tag(2, WireType.Varint).int64(message.startHeight);
    if (message.indexOffset !== 0n)
      writer.tag(3, WireType.Varint).int64(message.indexOffset);
    if (message.jailedUntil)
      Timestamp.internalBinaryWrite(message.jailedUntil, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.tombstoned !== false)
      writer.tag(5, WireType.Varint).bool(message.tombstoned);
    if (message.missedBlocksCounter !== 0n)
      writer.tag(6, WireType.Varint).int64(message.missedBlocksCounter);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ValidatorSigningInfo = new ValidatorSigningInfo$Type();
class Params$Type extends MessageType {
  constructor() {
    super("cosmos.slashing.v1beta1.Params", [
      {
        no: 1,
        name: "signed_blocks_window",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "min_signed_per_window", kind: "scalar", T: 12, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } },
      { no: 3, name: "downtime_jail_duration", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true, "amino.dont_omitempty": true } },
      { no: 4, name: "slash_fraction_double_sign", kind: "scalar", T: 12, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } },
      { no: 5, name: "slash_fraction_downtime", kind: "scalar", T: 12, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } }
    ], { "amino.name": "cosmos-sdk/x/slashing/Params" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signedBlocksWindow = 0n;
    message.minSignedPerWindow = new Uint8Array(0);
    message.slashFractionDoubleSign = new Uint8Array(0);
    message.slashFractionDowntime = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 signed_blocks_window */
        1:
          message.signedBlocksWindow = reader.int64().toBigInt();
          break;
        case /* bytes min_signed_per_window */
        2:
          message.minSignedPerWindow = reader.bytes();
          break;
        case /* google.protobuf.Duration downtime_jail_duration */
        3:
          message.downtimeJailDuration = Duration.internalBinaryRead(reader, reader.uint32(), options, message.downtimeJailDuration);
          break;
        case /* bytes slash_fraction_double_sign */
        4:
          message.slashFractionDoubleSign = reader.bytes();
          break;
        case /* bytes slash_fraction_downtime */
        5:
          message.slashFractionDowntime = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signedBlocksWindow !== 0n)
      writer.tag(1, WireType.Varint).int64(message.signedBlocksWindow);
    if (message.minSignedPerWindow.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.minSignedPerWindow);
    if (message.downtimeJailDuration)
      Duration.internalBinaryWrite(message.downtimeJailDuration, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.slashFractionDoubleSign.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.slashFractionDoubleSign);
    if (message.slashFractionDowntime.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.slashFractionDowntime);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
export {
  Params,
  ValidatorSigningInfo
};
