// @injectivelabs/core-proto-ts-v2/esm/generated/cometbft/p2p/v1/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class NetAddress$Type extends MessageType {
  constructor() {
    super("cometbft.p2p.v1.NetAddress", [
      { no: 1, name: "id", kind: "scalar", T: 9, options: { "gogoproto.customname": "ID" } },
      { no: 2, name: "ip", kind: "scalar", T: 9, options: { "gogoproto.customname": "IP" } },
      {
        no: 3,
        name: "port",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.id = "";
    message.ip = "";
    message.port = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string id */
        1:
          message.id = reader.string();
          break;
        case /* string ip */
        2:
          message.ip = reader.string();
          break;
        case /* uint32 port */
        3:
          message.port = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.id !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.id);
    if (message.ip !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.ip);
    if (message.port !== 0)
      writer.tag(3, WireType.Varint).uint32(message.port);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const NetAddress = new NetAddress$Type();
class ProtocolVersion$Type extends MessageType {
  constructor() {
    super("cometbft.p2p.v1.ProtocolVersion", [
      { no: 1, name: "p2p", kind: "scalar", jsonName: "p2p", T: 4, L: 0, options: { "gogoproto.customname": "P2P" } },
      {
        no: 2,
        name: "block",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "app",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.p2P = 0n;
    message.block = 0n;
    message.app = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 p2p */
        1:
          message.p2P = reader.uint64().toBigInt();
          break;
        case /* uint64 block */
        2:
          message.block = reader.uint64().toBigInt();
          break;
        case /* uint64 app */
        3:
          message.app = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.p2P !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.p2P);
    if (message.block !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.block);
    if (message.app !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.app);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ProtocolVersion = new ProtocolVersion$Type();
class DefaultNodeInfo$Type extends MessageType {
  constructor() {
    super("cometbft.p2p.v1.DefaultNodeInfo", [
      { no: 1, name: "protocol_version", kind: "message", T: () => ProtocolVersion, options: { "gogoproto.nullable": false } },
      { no: 2, name: "default_node_id", kind: "scalar", T: 9, options: { "gogoproto.customname": "DefaultNodeID" } },
      {
        no: 3,
        name: "listen_addr",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "network",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "channels",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 7,
        name: "moniker",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 8, name: "other", kind: "message", T: () => DefaultNodeInfoOther, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.defaultNodeId = "";
    message.listenAddr = "";
    message.network = "";
    message.version = "";
    message.channels = new Uint8Array(0);
    message.moniker = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.p2p.v1.ProtocolVersion protocol_version */
        1:
          message.protocolVersion = ProtocolVersion.internalBinaryRead(reader, reader.uint32(), options, message.protocolVersion);
          break;
        case /* string default_node_id */
        2:
          message.defaultNodeId = reader.string();
          break;
        case /* string listen_addr */
        3:
          message.listenAddr = reader.string();
          break;
        case /* string network */
        4:
          message.network = reader.string();
          break;
        case /* string version */
        5:
          message.version = reader.string();
          break;
        case /* bytes channels */
        6:
          message.channels = reader.bytes();
          break;
        case /* string moniker */
        7:
          message.moniker = reader.string();
          break;
        case /* cometbft.p2p.v1.DefaultNodeInfoOther other */
        8:
          message.other = DefaultNodeInfoOther.internalBinaryRead(reader, reader.uint32(), options, message.other);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.protocolVersion)
      ProtocolVersion.internalBinaryWrite(message.protocolVersion, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.defaultNodeId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.defaultNodeId);
    if (message.listenAddr !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.listenAddr);
    if (message.network !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.network);
    if (message.version !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.version);
    if (message.channels.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.channels);
    if (message.moniker !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.moniker);
    if (message.other)
      DefaultNodeInfoOther.internalBinaryWrite(message.other, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DefaultNodeInfo = new DefaultNodeInfo$Type();
class DefaultNodeInfoOther$Type extends MessageType {
  constructor() {
    super("cometbft.p2p.v1.DefaultNodeInfoOther", [
      {
        no: 1,
        name: "tx_index",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "rpc_address", kind: "scalar", T: 9, options: { "gogoproto.customname": "RPCAddress" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.txIndex = "";
    message.rpcAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string tx_index */
        1:
          message.txIndex = reader.string();
          break;
        case /* string rpc_address */
        2:
          message.rpcAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.txIndex !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.txIndex);
    if (message.rpcAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.rpcAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DefaultNodeInfoOther = new DefaultNodeInfoOther$Type();
export {
  DefaultNodeInfo,
  DefaultNodeInfoOther,
  NetAddress,
  ProtocolVersion
};
