// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/nft/v1beta1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { NFT } from "./nft_pb.js";
import { Class } from "./nft_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("cosmos.nft.v1beta1.GenesisState", [
      { no: 1, name: "classes", kind: "message", repeat: 2, T: () => Class },
      { no: 2, name: "entries", kind: "message", repeat: 2, T: () => Entry }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.classes = [];
    message.entries = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.nft.v1beta1.Class classes */
        1:
          message.classes.push(Class.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.nft.v1beta1.Entry entries */
        2:
          message.entries.push(Entry.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.classes.length; i++)
      Class.internalBinaryWrite(message.classes[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.entries.length; i++)
      Entry.internalBinaryWrite(message.entries[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class Entry$Type extends MessageType {
  constructor() {
    super("cosmos.nft.v1beta1.Entry", [
      {
        no: 1,
        name: "owner",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "nfts", kind: "message", repeat: 2, T: () => NFT }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.owner = "";
    message.nfts = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string owner */
        1:
          message.owner = reader.string();
          break;
        case /* repeated cosmos.nft.v1beta1.NFT nfts */
        2:
          message.nfts.push(NFT.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.owner !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.owner);
    for (let i = 0; i < message.nfts.length; i++)
      NFT.internalBinaryWrite(message.nfts[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Entry = new Entry$Type();
export {
  Entry,
  GenesisState
};
