// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/nft/v1beta1/nft_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../google/protobuf/any_pb.js";
class Class$Type extends MessageType {
  constructor() {
    super("cosmos.nft.v1beta1.Class", [
      {
        no: 1,
        name: "id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "symbol",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "uri",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "uri_hash",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 7, name: "data", kind: "message", T: () => Any }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.id = "";
    message.name = "";
    message.symbol = "";
    message.description = "";
    message.uri = "";
    message.uriHash = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string id */
        1:
          message.id = reader.string();
          break;
        case /* string name */
        2:
          message.name = reader.string();
          break;
        case /* string symbol */
        3:
          message.symbol = reader.string();
          break;
        case /* string description */
        4:
          message.description = reader.string();
          break;
        case /* string uri */
        5:
          message.uri = reader.string();
          break;
        case /* string uri_hash */
        6:
          message.uriHash = reader.string();
          break;
        case /* google.protobuf.Any data */
        7:
          message.data = Any.internalBinaryRead(reader, reader.uint32(), options, message.data);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.id !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.id);
    if (message.name !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.name);
    if (message.symbol !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.symbol);
    if (message.description !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.description);
    if (message.uri !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.uri);
    if (message.uriHash !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.uriHash);
    if (message.data)
      Any.internalBinaryWrite(message.data, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Class = new Class$Type();
class NFT$Type extends MessageType {
  constructor() {
    super("cosmos.nft.v1beta1.NFT", [
      {
        no: 1,
        name: "class_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "uri",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "uri_hash",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 10, name: "data", kind: "message", T: () => Any }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.classId = "";
    message.id = "";
    message.uri = "";
    message.uriHash = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string class_id */
        1:
          message.classId = reader.string();
          break;
        case /* string id */
        2:
          message.id = reader.string();
          break;
        case /* string uri */
        3:
          message.uri = reader.string();
          break;
        case /* string uri_hash */
        4:
          message.uriHash = reader.string();
          break;
        case /* google.protobuf.Any data */
        10:
          message.data = Any.internalBinaryRead(reader, reader.uint32(), options, message.data);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.classId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.classId);
    if (message.id !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.id);
    if (message.uri !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.uri);
    if (message.uriHash !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.uriHash);
    if (message.data)
      Any.internalBinaryWrite(message.data, writer.tag(10, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const NFT = new NFT$Type();
export {
  Class,
  NFT
};
