// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/nft/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Balance queries the number of NFTs of a given class owned by the owner, same as balanceOf in ERC721
   *
   * @generated from protobuf rpc: Balance
   */
  balance(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Owner queries the owner of the NFT based on its class and id, same as ownerOf in ERC721
   *
   * @generated from protobuf rpc: Owner
   */
  owner(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Supply queries the number of NFTs from the given class, same as totalSupply of ERC721.
   *
   * @generated from protobuf rpc: Supply
   */
  supply(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * NFTs queries all NFTs of a given class or owner,choose at least one of the two, similar to tokenByIndex in
   * ERC721Enumerable
   *
   * @generated from protobuf rpc: NFTs
   */
  nFTs(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * NFT queries an NFT based on its class and id.
   *
   * @generated from protobuf rpc: NFT
   */
  nFT(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Class queries an NFT class based on its id
   *
   * @generated from protobuf rpc: Class
   */
  class(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Classes queries all NFT classes
   *
   * @generated from protobuf rpc: Classes
   */
  classes(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
