// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/app/runtime/v1alpha1/module_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Module$Type extends MessageType {
  constructor() {
    super("cosmos.app.runtime.v1alpha1.Module", [
      {
        no: 1,
        name: "app_name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "begin_blockers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "end_blockers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "init_genesis",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "export_genesis",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 6, name: "override_store_keys", kind: "message", repeat: 2, T: () => StoreKeyConfig },
      {
        no: 7,
        name: "order_migrations",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "precommiters",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "prepare_check_staters",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "pre_blockers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "cosmos.app.v1alpha1.module": { goImport: "github.com/cosmos/cosmos-sdk/runtime", usePackage: [{ name: "cosmos.app.v1alpha1" }] } });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.appName = "";
    message.beginBlockers = [];
    message.endBlockers = [];
    message.initGenesis = [];
    message.exportGenesis = [];
    message.overrideStoreKeys = [];
    message.orderMigrations = [];
    message.precommiters = [];
    message.prepareCheckStaters = [];
    message.preBlockers = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string app_name */
        1:
          message.appName = reader.string();
          break;
        case /* repeated string begin_blockers */
        2:
          message.beginBlockers.push(reader.string());
          break;
        case /* repeated string end_blockers */
        3:
          message.endBlockers.push(reader.string());
          break;
        case /* repeated string init_genesis */
        4:
          message.initGenesis.push(reader.string());
          break;
        case /* repeated string export_genesis */
        5:
          message.exportGenesis.push(reader.string());
          break;
        case /* repeated cosmos.app.runtime.v1alpha1.StoreKeyConfig override_store_keys */
        6:
          message.overrideStoreKeys.push(StoreKeyConfig.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated string order_migrations */
        7:
          message.orderMigrations.push(reader.string());
          break;
        case /* repeated string precommiters */
        8:
          message.precommiters.push(reader.string());
          break;
        case /* repeated string prepare_check_staters */
        9:
          message.prepareCheckStaters.push(reader.string());
          break;
        case /* repeated string pre_blockers */
        10:
          message.preBlockers.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.appName !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.appName);
    for (let i = 0; i < message.beginBlockers.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.beginBlockers[i]);
    for (let i = 0; i < message.endBlockers.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.endBlockers[i]);
    for (let i = 0; i < message.initGenesis.length; i++)
      writer.tag(4, WireType.LengthDelimited).string(message.initGenesis[i]);
    for (let i = 0; i < message.exportGenesis.length; i++)
      writer.tag(5, WireType.LengthDelimited).string(message.exportGenesis[i]);
    for (let i = 0; i < message.overrideStoreKeys.length; i++)
      StoreKeyConfig.internalBinaryWrite(message.overrideStoreKeys[i], writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.orderMigrations.length; i++)
      writer.tag(7, WireType.LengthDelimited).string(message.orderMigrations[i]);
    for (let i = 0; i < message.precommiters.length; i++)
      writer.tag(8, WireType.LengthDelimited).string(message.precommiters[i]);
    for (let i = 0; i < message.prepareCheckStaters.length; i++)
      writer.tag(9, WireType.LengthDelimited).string(message.prepareCheckStaters[i]);
    for (let i = 0; i < message.preBlockers.length; i++)
      writer.tag(10, WireType.LengthDelimited).string(message.preBlockers[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Module = new Module$Type();
class StoreKeyConfig$Type extends MessageType {
  constructor() {
    super("cosmos.app.runtime.v1alpha1.StoreKeyConfig", [
      {
        no: 1,
        name: "module_name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "kv_store_key",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.moduleName = "";
    message.kvStoreKey = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string module_name */
        1:
          message.moduleName = reader.string();
          break;
        case /* string kv_store_key */
        2:
          message.kvStoreKey = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.moduleName !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.moduleName);
    if (message.kvStoreKey !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.kvStoreKey);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StoreKeyConfig = new StoreKeyConfig$Type();
export {
  Module,
  StoreKeyConfig
};
