// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/interchain_accounts/v1/account_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { BaseAccount } from "../../../../cosmos/auth/v1beta1/auth_pb.js";
class InterchainAccount$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.v1.InterchainAccount", [
      { no: 1, name: "base_account", kind: "message", T: () => BaseAccount, options: { "gogoproto.embed": true } },
      {
        no: 2,
        name: "account_owner",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.goproto_stringer": false, "cosmos_proto.implements_interface": ["ibc.applications.interchain_accounts.v1.InterchainAccountI"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.accountOwner = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.auth.v1beta1.BaseAccount base_account */
        1:
          message.baseAccount = BaseAccount.internalBinaryRead(reader, reader.uint32(), options, message.baseAccount);
          break;
        case /* string account_owner */
        2:
          message.accountOwner = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.baseAccount)
      BaseAccount.internalBinaryWrite(message.baseAccount, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.accountOwner !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.accountOwner);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const InterchainAccount = new InterchainAccount$Type();
export {
  InterchainAccount
};
