// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/interchain_accounts/v1/metadata_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Metadata$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.v1.Metadata", [
      {
        no: 1,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "controller_connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "host_connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "encoding",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "tx_type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.version = "";
    message.controllerConnectionId = "";
    message.hostConnectionId = "";
    message.address = "";
    message.encoding = "";
    message.txType = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string version */
        1:
          message.version = reader.string();
          break;
        case /* string controller_connection_id */
        2:
          message.controllerConnectionId = reader.string();
          break;
        case /* string host_connection_id */
        3:
          message.hostConnectionId = reader.string();
          break;
        case /* string address */
        4:
          message.address = reader.string();
          break;
        case /* string encoding */
        5:
          message.encoding = reader.string();
          break;
        case /* string tx_type */
        6:
          message.txType = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.version !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.version);
    if (message.controllerConnectionId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.controllerConnectionId);
    if (message.hostConnectionId !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.hostConnectionId);
    if (message.address !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.address);
    if (message.encoding !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.encoding);
    if (message.txType !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.txType);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Metadata = new Metadata$Type();
export {
  Metadata
};
