// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/interchain_accounts/v1/packet_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../../google/protobuf/any_pb.js";
var Type = /* @__PURE__ */ ((Type2) => {
  Type2[Type2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  Type2[Type2["EXECUTE_TX"] = 1] = "EXECUTE_TX";
  return Type2;
})(Type || {});
class InterchainAccountPacketData$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.v1.InterchainAccountPacketData", [
      { no: 1, name: "type", kind: "enum", T: () => ["ibc.applications.interchain_accounts.v1.Type", Type, "TYPE_"] },
      {
        no: 2,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "memo",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.type = 0;
    message.data = new Uint8Array(0);
    message.memo = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.applications.interchain_accounts.v1.Type type */
        1:
          message.type = reader.int32();
          break;
        case /* bytes data */
        2:
          message.data = reader.bytes();
          break;
        case /* string memo */
        3:
          message.memo = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.type !== 0)
      writer.tag(1, WireType.Varint).int32(message.type);
    if (message.data.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.data);
    if (message.memo !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.memo);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const InterchainAccountPacketData = new InterchainAccountPacketData$Type();
class CosmosTx$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.v1.CosmosTx", [
      { no: 1, name: "messages", kind: "message", repeat: 2, T: () => Any }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.messages = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated google.protobuf.Any messages */
        1:
          message.messages.push(Any.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.messages.length; i++)
      Any.internalBinaryWrite(message.messages[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CosmosTx = new CosmosTx$Type();
export {
  CosmosTx,
  InterchainAccountPacketData,
  Type
};
