// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/base/reflection/v1beta1/reflection_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class ListAllInterfacesRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.reflection.v1beta1.ListAllInterfacesRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListAllInterfacesRequest = new ListAllInterfacesRequest$Type();
class ListAllInterfacesResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.reflection.v1beta1.ListAllInterfacesResponse", [
      {
        no: 1,
        name: "interface_names",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.interfaceNames = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string interface_names */
        1:
          message.interfaceNames.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.interfaceNames.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.interfaceNames[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListAllInterfacesResponse = new ListAllInterfacesResponse$Type();
class ListImplementationsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.reflection.v1beta1.ListImplementationsRequest", [
      {
        no: 1,
        name: "interface_name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.interfaceName = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string interface_name */
        1:
          message.interfaceName = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.interfaceName !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.interfaceName);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListImplementationsRequest = new ListImplementationsRequest$Type();
class ListImplementationsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.reflection.v1beta1.ListImplementationsResponse", [
      {
        no: 1,
        name: "implementation_message_names",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.implementationMessageNames = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string implementation_message_names */
        1:
          message.implementationMessageNames.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.implementationMessageNames.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.implementationMessageNames[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListImplementationsResponse = new ListImplementationsResponse$Type();
const ReflectionService = new ServiceType("cosmos.base.reflection.v1beta1.ReflectionService", [
  { name: "ListAllInterfaces", options: { "google.api.http": { get: "/cosmos/base/reflection/v1beta1/interfaces" } }, I: ListAllInterfacesRequest, O: ListAllInterfacesResponse },
  { name: "ListImplementations", options: { "google.api.http": { get: "/cosmos/base/reflection/v1beta1/interfaces/{interface_name}/implementations" } }, I: ListImplementationsRequest, O: ListImplementationsResponse }
]);
export {
  ListAllInterfacesRequest,
  ListAllInterfacesResponse,
  ListImplementationsRequest,
  ListImplementationsResponse,
  ReflectionService
};
