// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/base/reflection/v1beta1/reflection_pb.client.js
import { ReflectionService } from "./reflection_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class ReflectionServiceClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = ReflectionService.typeName;
    this.methods = ReflectionService.methods;
    this.options = ReflectionService.options;
  }
  /**
   * ListAllInterfaces lists all the interfaces registered in the interface
   * registry.
   *
   * @generated from protobuf rpc: ListAllInterfaces
   */
  listAllInterfaces(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ListImplementations list all the concrete types that implement a given
   * interface.
   *
   * @generated from protobuf rpc: ListImplementations
   */
  listImplementations(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  ReflectionServiceClient
};
