// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/base/query/v1beta1/pagination_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class PageRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.query.v1beta1.PageRequest", [
      {
        no: 1,
        name: "key",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "offset",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "limit",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "count_total",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 5,
        name: "reverse",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.key = new Uint8Array(0);
    message.offset = 0n;
    message.limit = 0n;
    message.countTotal = false;
    message.reverse = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes key */
        1:
          message.key = reader.bytes();
          break;
        case /* uint64 offset */
        2:
          message.offset = reader.uint64().toBigInt();
          break;
        case /* uint64 limit */
        3:
          message.limit = reader.uint64().toBigInt();
          break;
        case /* bool count_total */
        4:
          message.countTotal = reader.bool();
          break;
        case /* bool reverse */
        5:
          message.reverse = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.key.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.key);
    if (message.offset !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.offset);
    if (message.limit !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.limit);
    if (message.countTotal !== false)
      writer.tag(4, WireType.Varint).bool(message.countTotal);
    if (message.reverse !== false)
      writer.tag(5, WireType.Varint).bool(message.reverse);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PageRequest = new PageRequest$Type();
class PageResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.query.v1beta1.PageResponse", [
      {
        no: 1,
        name: "next_key",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "total",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nextKey = new Uint8Array(0);
    message.total = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes next_key */
        1:
          message.nextKey = reader.bytes();
          break;
        case /* uint64 total */
        2:
          message.total = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nextKey.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.nextKey);
    if (message.total !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.total);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PageResponse = new PageResponse$Type();
export {
  PageRequest,
  PageResponse
};
