// @injectivelabs/core-proto-ts-v2/esm/generated/cometbft/types/v1/evidence_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Validator } from "./validator_pb.js";
import { LightBlock } from "./types_pb.js";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Vote } from "./types_pb.js";
class Evidence$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.Evidence", [
      { no: 1, name: "duplicate_vote_evidence", kind: "message", oneof: "sum", T: () => DuplicateVoteEvidence },
      { no: 2, name: "light_client_attack_evidence", kind: "message", oneof: "sum", T: () => LightClientAttackEvidence }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sum = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.DuplicateVoteEvidence duplicate_vote_evidence */
        1:
          message.sum = {
            oneofKind: "duplicateVoteEvidence",
            duplicateVoteEvidence: DuplicateVoteEvidence.internalBinaryRead(reader, reader.uint32(), options, message.sum.duplicateVoteEvidence)
          };
          break;
        case /* cometbft.types.v1.LightClientAttackEvidence light_client_attack_evidence */
        2:
          message.sum = {
            oneofKind: "lightClientAttackEvidence",
            lightClientAttackEvidence: LightClientAttackEvidence.internalBinaryRead(reader, reader.uint32(), options, message.sum.lightClientAttackEvidence)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sum.oneofKind === "duplicateVoteEvidence")
      DuplicateVoteEvidence.internalBinaryWrite(message.sum.duplicateVoteEvidence, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "lightClientAttackEvidence")
      LightClientAttackEvidence.internalBinaryWrite(message.sum.lightClientAttackEvidence, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Evidence = new Evidence$Type();
class DuplicateVoteEvidence$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.DuplicateVoteEvidence", [
      { no: 1, name: "vote_a", kind: "message", T: () => Vote },
      { no: 2, name: "vote_b", kind: "message", T: () => Vote },
      {
        no: 3,
        name: "total_voting_power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "validator_power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 5, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.totalVotingPower = 0n;
    message.validatorPower = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.Vote vote_a */
        1:
          message.voteA = Vote.internalBinaryRead(reader, reader.uint32(), options, message.voteA);
          break;
        case /* cometbft.types.v1.Vote vote_b */
        2:
          message.voteB = Vote.internalBinaryRead(reader, reader.uint32(), options, message.voteB);
          break;
        case /* int64 total_voting_power */
        3:
          message.totalVotingPower = reader.int64().toBigInt();
          break;
        case /* int64 validator_power */
        4:
          message.validatorPower = reader.int64().toBigInt();
          break;
        case /* google.protobuf.Timestamp timestamp */
        5:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.voteA)
      Vote.internalBinaryWrite(message.voteA, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.voteB)
      Vote.internalBinaryWrite(message.voteB, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.totalVotingPower !== 0n)
      writer.tag(3, WireType.Varint).int64(message.totalVotingPower);
    if (message.validatorPower !== 0n)
      writer.tag(4, WireType.Varint).int64(message.validatorPower);
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DuplicateVoteEvidence = new DuplicateVoteEvidence$Type();
class LightClientAttackEvidence$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.LightClientAttackEvidence", [
      { no: 1, name: "conflicting_block", kind: "message", T: () => LightBlock },
      {
        no: 2,
        name: "common_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "byzantine_validators", kind: "message", repeat: 2, T: () => Validator },
      {
        no: 4,
        name: "total_voting_power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 5, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.commonHeight = 0n;
    message.byzantineValidators = [];
    message.totalVotingPower = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.LightBlock conflicting_block */
        1:
          message.conflictingBlock = LightBlock.internalBinaryRead(reader, reader.uint32(), options, message.conflictingBlock);
          break;
        case /* int64 common_height */
        2:
          message.commonHeight = reader.int64().toBigInt();
          break;
        case /* repeated cometbft.types.v1.Validator byzantine_validators */
        3:
          message.byzantineValidators.push(Validator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* int64 total_voting_power */
        4:
          message.totalVotingPower = reader.int64().toBigInt();
          break;
        case /* google.protobuf.Timestamp timestamp */
        5:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.conflictingBlock)
      LightBlock.internalBinaryWrite(message.conflictingBlock, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.commonHeight !== 0n)
      writer.tag(2, WireType.Varint).int64(message.commonHeight);
    for (let i = 0; i < message.byzantineValidators.length; i++)
      Validator.internalBinaryWrite(message.byzantineValidators[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.totalVotingPower !== 0n)
      writer.tag(4, WireType.Varint).int64(message.totalVotingPower);
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const LightClientAttackEvidence = new LightClientAttackEvidence$Type();
class EvidenceList$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.EvidenceList", [
      { no: 1, name: "evidence", kind: "message", repeat: 2, T: () => Evidence, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.evidence = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cometbft.types.v1.Evidence evidence */
        1:
          message.evidence.push(Evidence.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.evidence.length; i++)
      Evidence.internalBinaryWrite(message.evidence[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EvidenceList = new EvidenceList$Type();
export {
  DuplicateVoteEvidence,
  Evidence,
  EvidenceList,
  LightClientAttackEvidence
};
