// @injectivelabs/core-proto-ts-v2/esm/generated/cometbft/types/v1/params_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Int64Value } from "../../../google/protobuf/wrappers_pb.js";
import { Duration } from "../../../google/protobuf/duration_pb.js";
class ConsensusParams$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.ConsensusParams", [
      { no: 1, name: "block", kind: "message", T: () => BlockParams },
      { no: 2, name: "evidence", kind: "message", T: () => EvidenceParams },
      { no: 3, name: "validator", kind: "message", T: () => ValidatorParams },
      { no: 4, name: "version", kind: "message", T: () => VersionParams },
      { no: 5, name: "abci", kind: "message", T: () => ABCIParams },
      { no: 6, name: "synchrony", kind: "message", T: () => SynchronyParams },
      { no: 7, name: "feature", kind: "message", T: () => FeatureParams }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.BlockParams block */
        1:
          message.block = BlockParams.internalBinaryRead(reader, reader.uint32(), options, message.block);
          break;
        case /* cometbft.types.v1.EvidenceParams evidence */
        2:
          message.evidence = EvidenceParams.internalBinaryRead(reader, reader.uint32(), options, message.evidence);
          break;
        case /* cometbft.types.v1.ValidatorParams validator */
        3:
          message.validator = ValidatorParams.internalBinaryRead(reader, reader.uint32(), options, message.validator);
          break;
        case /* cometbft.types.v1.VersionParams version */
        4:
          message.version = VersionParams.internalBinaryRead(reader, reader.uint32(), options, message.version);
          break;
        case /* cometbft.types.v1.ABCIParams abci = 5 [deprecated = true] */
        5:
          message.abci = ABCIParams.internalBinaryRead(reader, reader.uint32(), options, message.abci);
          break;
        case /* cometbft.types.v1.SynchronyParams synchrony */
        6:
          message.synchrony = SynchronyParams.internalBinaryRead(reader, reader.uint32(), options, message.synchrony);
          break;
        case /* cometbft.types.v1.FeatureParams feature */
        7:
          message.feature = FeatureParams.internalBinaryRead(reader, reader.uint32(), options, message.feature);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.block)
      BlockParams.internalBinaryWrite(message.block, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.evidence)
      EvidenceParams.internalBinaryWrite(message.evidence, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.validator)
      ValidatorParams.internalBinaryWrite(message.validator, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.version)
      VersionParams.internalBinaryWrite(message.version, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.abci)
      ABCIParams.internalBinaryWrite(message.abci, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.synchrony)
      SynchronyParams.internalBinaryWrite(message.synchrony, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.feature)
      FeatureParams.internalBinaryWrite(message.feature, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConsensusParams = new ConsensusParams$Type();
class BlockParams$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.BlockParams", [
      {
        no: 1,
        name: "max_bytes",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "max_gas",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.maxBytes = 0n;
    message.maxGas = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 max_bytes */
        1:
          message.maxBytes = reader.int64().toBigInt();
          break;
        case /* int64 max_gas */
        2:
          message.maxGas = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.maxBytes !== 0n)
      writer.tag(1, WireType.Varint).int64(message.maxBytes);
    if (message.maxGas !== 0n)
      writer.tag(2, WireType.Varint).int64(message.maxGas);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BlockParams = new BlockParams$Type();
class EvidenceParams$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.EvidenceParams", [
      {
        no: 1,
        name: "max_age_num_blocks",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "max_age_duration", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true } },
      {
        no: 3,
        name: "max_bytes",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.maxAgeNumBlocks = 0n;
    message.maxBytes = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 max_age_num_blocks */
        1:
          message.maxAgeNumBlocks = reader.int64().toBigInt();
          break;
        case /* google.protobuf.Duration max_age_duration */
        2:
          message.maxAgeDuration = Duration.internalBinaryRead(reader, reader.uint32(), options, message.maxAgeDuration);
          break;
        case /* int64 max_bytes */
        3:
          message.maxBytes = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.maxAgeNumBlocks !== 0n)
      writer.tag(1, WireType.Varint).int64(message.maxAgeNumBlocks);
    if (message.maxAgeDuration)
      Duration.internalBinaryWrite(message.maxAgeDuration, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.maxBytes !== 0n)
      writer.tag(3, WireType.Varint).int64(message.maxBytes);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EvidenceParams = new EvidenceParams$Type();
class ValidatorParams$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.ValidatorParams", [
      {
        no: 1,
        name: "pub_key_types",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.populate": true, "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.pubKeyTypes = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string pub_key_types */
        1:
          message.pubKeyTypes.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.pubKeyTypes.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.pubKeyTypes[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ValidatorParams = new ValidatorParams$Type();
class VersionParams$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.VersionParams", [
      {
        no: 1,
        name: "app",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.populate": true, "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.app = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 app */
        1:
          message.app = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.app !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.app);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const VersionParams = new VersionParams$Type();
class HashedParams$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.HashedParams", [
      {
        no: 1,
        name: "block_max_bytes",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "block_max_gas",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.blockMaxBytes = 0n;
    message.blockMaxGas = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 block_max_bytes */
        1:
          message.blockMaxBytes = reader.int64().toBigInt();
          break;
        case /* int64 block_max_gas */
        2:
          message.blockMaxGas = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockMaxBytes !== 0n)
      writer.tag(1, WireType.Varint).int64(message.blockMaxBytes);
    if (message.blockMaxGas !== 0n)
      writer.tag(2, WireType.Varint).int64(message.blockMaxGas);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const HashedParams = new HashedParams$Type();
class SynchronyParams$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.SynchronyParams", [
      { no: 1, name: "precision", kind: "message", T: () => Duration, options: { "gogoproto.stdduration": true } },
      { no: 2, name: "message_delay", kind: "message", T: () => Duration, options: { "gogoproto.stdduration": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Duration precision */
        1:
          message.precision = Duration.internalBinaryRead(reader, reader.uint32(), options, message.precision);
          break;
        case /* google.protobuf.Duration message_delay */
        2:
          message.messageDelay = Duration.internalBinaryRead(reader, reader.uint32(), options, message.messageDelay);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.precision)
      Duration.internalBinaryWrite(message.precision, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.messageDelay)
      Duration.internalBinaryWrite(message.messageDelay, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SynchronyParams = new SynchronyParams$Type();
class FeatureParams$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.FeatureParams", [
      { no: 1, name: "vote_extensions_enable_height", kind: "message", T: () => Int64Value, options: { "gogoproto.nullable": true } },
      { no: 2, name: "pbts_enable_height", kind: "message", T: () => Int64Value, options: { "gogoproto.nullable": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Int64Value vote_extensions_enable_height */
        1:
          message.voteExtensionsEnableHeight = Int64Value.internalBinaryRead(reader, reader.uint32(), options, message.voteExtensionsEnableHeight);
          break;
        case /* google.protobuf.Int64Value pbts_enable_height */
        2:
          message.pbtsEnableHeight = Int64Value.internalBinaryRead(reader, reader.uint32(), options, message.pbtsEnableHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.voteExtensionsEnableHeight)
      Int64Value.internalBinaryWrite(message.voteExtensionsEnableHeight, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pbtsEnableHeight)
      Int64Value.internalBinaryWrite(message.pbtsEnableHeight, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FeatureParams = new FeatureParams$Type();
class ABCIParams$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.ABCIParams", [
      {
        no: 1,
        name: "vote_extensions_enable_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.voteExtensionsEnableHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 vote_extensions_enable_height */
        1:
          message.voteExtensionsEnableHeight = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.voteExtensionsEnableHeight !== 0n)
      writer.tag(1, WireType.Varint).int64(message.voteExtensionsEnableHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ABCIParams = new ABCIParams$Type();
export {
  ABCIParams,
  BlockParams,
  ConsensusParams,
  EvidenceParams,
  FeatureParams,
  HashedParams,
  SynchronyParams,
  ValidatorParams,
  VersionParams
};
