// @injectivelabs/core-proto-ts-v2/esm/generated/cometbft/types/v1/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ValidatorSet } from "./validator_pb.js";
import { BlockIDFlag } from "./validator_pb.js";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Consensus } from "../../version/v1/types_pb.js";
import { Proof } from "../../crypto/v1/proof_pb.js";
var SignedMsgType = /* @__PURE__ */ ((SignedMsgType2) => {
  SignedMsgType2[SignedMsgType2["UNKNOWN"] = 0] = "UNKNOWN";
  SignedMsgType2[SignedMsgType2["PREVOTE"] = 1] = "PREVOTE";
  SignedMsgType2[SignedMsgType2["PRECOMMIT"] = 2] = "PRECOMMIT";
  SignedMsgType2[SignedMsgType2["PROPOSAL"] = 32] = "PROPOSAL";
  return SignedMsgType2;
})(SignedMsgType || {});
class PartSetHeader$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.PartSetHeader", [
      {
        no: 1,
        name: "total",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 2,
        name: "hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.total = 0;
    message.hash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 total */
        1:
          message.total = reader.uint32();
          break;
        case /* bytes hash */
        2:
          message.hash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.total !== 0)
      writer.tag(1, WireType.Varint).uint32(message.total);
    if (message.hash.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.hash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PartSetHeader = new PartSetHeader$Type();
class Part$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.Part", [
      {
        no: 1,
        name: "index",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 2,
        name: "bytes",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof", kind: "message", T: () => Proof, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.index = 0;
    message.bytes = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 index */
        1:
          message.index = reader.uint32();
          break;
        case /* bytes bytes */
        2:
          message.bytes = reader.bytes();
          break;
        case /* cometbft.crypto.v1.Proof proof */
        3:
          message.proof = Proof.internalBinaryRead(reader, reader.uint32(), options, message.proof);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.index !== 0)
      writer.tag(1, WireType.Varint).uint32(message.index);
    if (message.bytes.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.bytes);
    if (message.proof)
      Proof.internalBinaryWrite(message.proof, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Part = new Part$Type();
class BlockID$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.BlockID", [
      {
        no: 1,
        name: "hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "part_set_header", kind: "message", T: () => PartSetHeader, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.hash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes hash */
        1:
          message.hash = reader.bytes();
          break;
        case /* cometbft.types.v1.PartSetHeader part_set_header */
        2:
          message.partSetHeader = PartSetHeader.internalBinaryRead(reader, reader.uint32(), options, message.partSetHeader);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.hash.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.hash);
    if (message.partSetHeader)
      PartSetHeader.internalBinaryWrite(message.partSetHeader, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BlockID = new BlockID$Type();
class Header$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.Header", [
      { no: 1, name: "version", kind: "message", T: () => Consensus, options: { "gogoproto.nullable": false } },
      { no: 2, name: "chain_id", kind: "scalar", T: 9, options: { "gogoproto.customname": "ChainID" } },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      { no: 5, name: "last_block_id", kind: "message", T: () => BlockID, options: { "gogoproto.nullable": false } },
      {
        no: 6,
        name: "last_commit_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 7,
        name: "data_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 8,
        name: "validators_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 9,
        name: "next_validators_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 10,
        name: "consensus_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 11,
        name: "app_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 12,
        name: "last_results_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 13,
        name: "evidence_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 14,
        name: "proposer_address",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chainId = "";
    message.height = 0n;
    message.lastCommitHash = new Uint8Array(0);
    message.dataHash = new Uint8Array(0);
    message.validatorsHash = new Uint8Array(0);
    message.nextValidatorsHash = new Uint8Array(0);
    message.consensusHash = new Uint8Array(0);
    message.appHash = new Uint8Array(0);
    message.lastResultsHash = new Uint8Array(0);
    message.evidenceHash = new Uint8Array(0);
    message.proposerAddress = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.version.v1.Consensus version */
        1:
          message.version = Consensus.internalBinaryRead(reader, reader.uint32(), options, message.version);
          break;
        case /* string chain_id */
        2:
          message.chainId = reader.string();
          break;
        case /* int64 height */
        3:
          message.height = reader.int64().toBigInt();
          break;
        case /* google.protobuf.Timestamp time */
        4:
          message.time = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.time);
          break;
        case /* cometbft.types.v1.BlockID last_block_id */
        5:
          message.lastBlockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.lastBlockId);
          break;
        case /* bytes last_commit_hash */
        6:
          message.lastCommitHash = reader.bytes();
          break;
        case /* bytes data_hash */
        7:
          message.dataHash = reader.bytes();
          break;
        case /* bytes validators_hash */
        8:
          message.validatorsHash = reader.bytes();
          break;
        case /* bytes next_validators_hash */
        9:
          message.nextValidatorsHash = reader.bytes();
          break;
        case /* bytes consensus_hash */
        10:
          message.consensusHash = reader.bytes();
          break;
        case /* bytes app_hash */
        11:
          message.appHash = reader.bytes();
          break;
        case /* bytes last_results_hash */
        12:
          message.lastResultsHash = reader.bytes();
          break;
        case /* bytes evidence_hash */
        13:
          message.evidenceHash = reader.bytes();
          break;
        case /* bytes proposer_address */
        14:
          message.proposerAddress = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.version)
      Consensus.internalBinaryWrite(message.version, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.chainId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.chainId);
    if (message.height !== 0n)
      writer.tag(3, WireType.Varint).int64(message.height);
    if (message.time)
      Timestamp.internalBinaryWrite(message.time, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.lastBlockId)
      BlockID.internalBinaryWrite(message.lastBlockId, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.lastCommitHash.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.lastCommitHash);
    if (message.dataHash.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.dataHash);
    if (message.validatorsHash.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.validatorsHash);
    if (message.nextValidatorsHash.length)
      writer.tag(9, WireType.LengthDelimited).bytes(message.nextValidatorsHash);
    if (message.consensusHash.length)
      writer.tag(10, WireType.LengthDelimited).bytes(message.consensusHash);
    if (message.appHash.length)
      writer.tag(11, WireType.LengthDelimited).bytes(message.appHash);
    if (message.lastResultsHash.length)
      writer.tag(12, WireType.LengthDelimited).bytes(message.lastResultsHash);
    if (message.evidenceHash.length)
      writer.tag(13, WireType.LengthDelimited).bytes(message.evidenceHash);
    if (message.proposerAddress.length)
      writer.tag(14, WireType.LengthDelimited).bytes(message.proposerAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Header = new Header$Type();
class Data$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.Data", [
      {
        no: 1,
        name: "txs",
        kind: "scalar",
        repeat: 2,
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.txs = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated bytes txs */
        1:
          message.txs.push(reader.bytes());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.txs.length; i++)
      writer.tag(1, WireType.LengthDelimited).bytes(message.txs[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Data = new Data$Type();
class Vote$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.Vote", [
      { no: 1, name: "type", kind: "enum", T: () => ["cometbft.types.v1.SignedMsgType", SignedMsgType, "SIGNED_MSG_TYPE_"] },
      {
        no: 2,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "round",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      { no: 4, name: "block_id", kind: "message", T: () => BlockID, options: { "gogoproto.nullable": false, "gogoproto.customname": "BlockID" } },
      { no: 5, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      {
        no: 6,
        name: "validator_address",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 7,
        name: "validator_index",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      {
        no: 8,
        name: "signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 9,
        name: "extension",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 10,
        name: "extension_signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.type = 0;
    message.height = 0n;
    message.round = 0;
    message.validatorAddress = new Uint8Array(0);
    message.validatorIndex = 0;
    message.signature = new Uint8Array(0);
    message.extension = new Uint8Array(0);
    message.extensionSignature = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.SignedMsgType type */
        1:
          message.type = reader.int32();
          break;
        case /* int64 height */
        2:
          message.height = reader.int64().toBigInt();
          break;
        case /* int32 round */
        3:
          message.round = reader.int32();
          break;
        case /* cometbft.types.v1.BlockID block_id */
        4:
          message.blockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* google.protobuf.Timestamp timestamp */
        5:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        case /* bytes validator_address */
        6:
          message.validatorAddress = reader.bytes();
          break;
        case /* int32 validator_index */
        7:
          message.validatorIndex = reader.int32();
          break;
        case /* bytes signature */
        8:
          message.signature = reader.bytes();
          break;
        case /* bytes extension */
        9:
          message.extension = reader.bytes();
          break;
        case /* bytes extension_signature */
        10:
          message.extensionSignature = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.type !== 0)
      writer.tag(1, WireType.Varint).int32(message.type);
    if (message.height !== 0n)
      writer.tag(2, WireType.Varint).int64(message.height);
    if (message.round !== 0)
      writer.tag(3, WireType.Varint).int32(message.round);
    if (message.blockId)
      BlockID.internalBinaryWrite(message.blockId, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.validatorAddress.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.validatorAddress);
    if (message.validatorIndex !== 0)
      writer.tag(7, WireType.Varint).int32(message.validatorIndex);
    if (message.signature.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.signature);
    if (message.extension.length)
      writer.tag(9, WireType.LengthDelimited).bytes(message.extension);
    if (message.extensionSignature.length)
      writer.tag(10, WireType.LengthDelimited).bytes(message.extensionSignature);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Vote = new Vote$Type();
class Commit$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.Commit", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "round",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      { no: 3, name: "block_id", kind: "message", T: () => BlockID, options: { "gogoproto.nullable": false, "gogoproto.customname": "BlockID" } },
      { no: 4, name: "signatures", kind: "message", repeat: 2, T: () => CommitSig, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    message.round = 0;
    message.signatures = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        case /* int32 round */
        2:
          message.round = reader.int32();
          break;
        case /* cometbft.types.v1.BlockID block_id */
        3:
          message.blockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* repeated cometbft.types.v1.CommitSig signatures */
        4:
          message.signatures.push(CommitSig.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    if (message.round !== 0)
      writer.tag(2, WireType.Varint).int32(message.round);
    if (message.blockId)
      BlockID.internalBinaryWrite(message.blockId, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.signatures.length; i++)
      CommitSig.internalBinaryWrite(message.signatures[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Commit = new Commit$Type();
class CommitSig$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.CommitSig", [
      { no: 1, name: "block_id_flag", kind: "enum", T: () => ["cometbft.types.v1.BlockIDFlag", BlockIDFlag] },
      {
        no: 2,
        name: "validator_address",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      {
        no: 4,
        name: "signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.blockIdFlag = 0;
    message.validatorAddress = new Uint8Array(0);
    message.signature = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.BlockIDFlag block_id_flag */
        1:
          message.blockIdFlag = reader.int32();
          break;
        case /* bytes validator_address */
        2:
          message.validatorAddress = reader.bytes();
          break;
        case /* google.protobuf.Timestamp timestamp */
        3:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        case /* bytes signature */
        4:
          message.signature = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockIdFlag !== 0)
      writer.tag(1, WireType.Varint).int32(message.blockIdFlag);
    if (message.validatorAddress.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.validatorAddress);
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.signature.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.signature);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CommitSig = new CommitSig$Type();
class ExtendedCommit$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.ExtendedCommit", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "round",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      { no: 3, name: "block_id", kind: "message", T: () => BlockID, options: { "gogoproto.nullable": false, "gogoproto.customname": "BlockID" } },
      { no: 4, name: "extended_signatures", kind: "message", repeat: 2, T: () => ExtendedCommitSig, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    message.round = 0;
    message.extendedSignatures = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        case /* int32 round */
        2:
          message.round = reader.int32();
          break;
        case /* cometbft.types.v1.BlockID block_id */
        3:
          message.blockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* repeated cometbft.types.v1.ExtendedCommitSig extended_signatures */
        4:
          message.extendedSignatures.push(ExtendedCommitSig.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    if (message.round !== 0)
      writer.tag(2, WireType.Varint).int32(message.round);
    if (message.blockId)
      BlockID.internalBinaryWrite(message.blockId, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.extendedSignatures.length; i++)
      ExtendedCommitSig.internalBinaryWrite(message.extendedSignatures[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ExtendedCommit = new ExtendedCommit$Type();
class ExtendedCommitSig$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.ExtendedCommitSig", [
      { no: 1, name: "block_id_flag", kind: "enum", T: () => ["cometbft.types.v1.BlockIDFlag", BlockIDFlag] },
      {
        no: 2,
        name: "validator_address",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      {
        no: 4,
        name: "signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "extension",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 6,
        name: "extension_signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.blockIdFlag = 0;
    message.validatorAddress = new Uint8Array(0);
    message.signature = new Uint8Array(0);
    message.extension = new Uint8Array(0);
    message.extensionSignature = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.BlockIDFlag block_id_flag */
        1:
          message.blockIdFlag = reader.int32();
          break;
        case /* bytes validator_address */
        2:
          message.validatorAddress = reader.bytes();
          break;
        case /* google.protobuf.Timestamp timestamp */
        3:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        case /* bytes signature */
        4:
          message.signature = reader.bytes();
          break;
        case /* bytes extension */
        5:
          message.extension = reader.bytes();
          break;
        case /* bytes extension_signature */
        6:
          message.extensionSignature = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockIdFlag !== 0)
      writer.tag(1, WireType.Varint).int32(message.blockIdFlag);
    if (message.validatorAddress.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.validatorAddress);
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.signature.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.signature);
    if (message.extension.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.extension);
    if (message.extensionSignature.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.extensionSignature);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ExtendedCommitSig = new ExtendedCommitSig$Type();
class Proposal$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.Proposal", [
      { no: 1, name: "type", kind: "enum", T: () => ["cometbft.types.v1.SignedMsgType", SignedMsgType, "SIGNED_MSG_TYPE_"] },
      {
        no: 2,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "round",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      {
        no: 4,
        name: "pol_round",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      { no: 5, name: "block_id", kind: "message", T: () => BlockID, options: { "gogoproto.nullable": false, "gogoproto.customname": "BlockID" } },
      { no: 6, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      {
        no: 7,
        name: "signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.type = 0;
    message.height = 0n;
    message.round = 0;
    message.polRound = 0;
    message.signature = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.SignedMsgType type */
        1:
          message.type = reader.int32();
          break;
        case /* int64 height */
        2:
          message.height = reader.int64().toBigInt();
          break;
        case /* int32 round */
        3:
          message.round = reader.int32();
          break;
        case /* int32 pol_round */
        4:
          message.polRound = reader.int32();
          break;
        case /* cometbft.types.v1.BlockID block_id */
        5:
          message.blockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* google.protobuf.Timestamp timestamp */
        6:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        case /* bytes signature */
        7:
          message.signature = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.type !== 0)
      writer.tag(1, WireType.Varint).int32(message.type);
    if (message.height !== 0n)
      writer.tag(2, WireType.Varint).int64(message.height);
    if (message.round !== 0)
      writer.tag(3, WireType.Varint).int32(message.round);
    if (message.polRound !== 0)
      writer.tag(4, WireType.Varint).int32(message.polRound);
    if (message.blockId)
      BlockID.internalBinaryWrite(message.blockId, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.signature.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.signature);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Proposal = new Proposal$Type();
class SignedHeader$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.SignedHeader", [
      { no: 1, name: "header", kind: "message", T: () => Header },
      { no: 2, name: "commit", kind: "message", T: () => Commit }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.Header header */
        1:
          message.header = Header.internalBinaryRead(reader, reader.uint32(), options, message.header);
          break;
        case /* cometbft.types.v1.Commit commit */
        2:
          message.commit = Commit.internalBinaryRead(reader, reader.uint32(), options, message.commit);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.header)
      Header.internalBinaryWrite(message.header, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.commit)
      Commit.internalBinaryWrite(message.commit, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignedHeader = new SignedHeader$Type();
class LightBlock$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.LightBlock", [
      { no: 1, name: "signed_header", kind: "message", T: () => SignedHeader },
      { no: 2, name: "validator_set", kind: "message", T: () => ValidatorSet }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.SignedHeader signed_header */
        1:
          message.signedHeader = SignedHeader.internalBinaryRead(reader, reader.uint32(), options, message.signedHeader);
          break;
        case /* cometbft.types.v1.ValidatorSet validator_set */
        2:
          message.validatorSet = ValidatorSet.internalBinaryRead(reader, reader.uint32(), options, message.validatorSet);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signedHeader)
      SignedHeader.internalBinaryWrite(message.signedHeader, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.validatorSet)
      ValidatorSet.internalBinaryWrite(message.validatorSet, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const LightBlock = new LightBlock$Type();
class BlockMeta$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.BlockMeta", [
      { no: 1, name: "block_id", kind: "message", T: () => BlockID, options: { "gogoproto.nullable": false, "gogoproto.customname": "BlockID" } },
      {
        no: 2,
        name: "block_size",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "header", kind: "message", T: () => Header, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "num_txs",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.blockSize = 0n;
    message.numTxs = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.BlockID block_id */
        1:
          message.blockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* int64 block_size */
        2:
          message.blockSize = reader.int64().toBigInt();
          break;
        case /* cometbft.types.v1.Header header */
        3:
          message.header = Header.internalBinaryRead(reader, reader.uint32(), options, message.header);
          break;
        case /* int64 num_txs */
        4:
          message.numTxs = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockId)
      BlockID.internalBinaryWrite(message.blockId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.blockSize !== 0n)
      writer.tag(2, WireType.Varint).int64(message.blockSize);
    if (message.header)
      Header.internalBinaryWrite(message.header, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.numTxs !== 0n)
      writer.tag(4, WireType.Varint).int64(message.numTxs);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BlockMeta = new BlockMeta$Type();
class TxProof$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.TxProof", [
      {
        no: 1,
        name: "root_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof", kind: "message", T: () => Proof }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.rootHash = new Uint8Array(0);
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes root_hash */
        1:
          message.rootHash = reader.bytes();
          break;
        case /* bytes data */
        2:
          message.data = reader.bytes();
          break;
        case /* cometbft.crypto.v1.Proof proof */
        3:
          message.proof = Proof.internalBinaryRead(reader, reader.uint32(), options, message.proof);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.rootHash.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.rootHash);
    if (message.data.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.data);
    if (message.proof)
      Proof.internalBinaryWrite(message.proof, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxProof = new TxProof$Type();
export {
  BlockID,
  BlockMeta,
  Commit,
  CommitSig,
  Data,
  ExtendedCommit,
  ExtendedCommitSig,
  Header,
  LightBlock,
  Part,
  PartSetHeader,
  Proposal,
  SignedHeader,
  SignedMsgType,
  TxProof,
  Vote
};
