// @injectivelabs/core-proto-ts-v2/esm/generated/cometbft/types/v1/validator_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { PublicKey } from "../../crypto/v1/keys_pb.js";
var BlockIDFlag = /* @__PURE__ */ ((BlockIDFlag2) => {
  BlockIDFlag2[BlockIDFlag2["BLOCK_ID_FLAG_UNKNOWN"] = 0] = "BLOCK_ID_FLAG_UNKNOWN";
  BlockIDFlag2[BlockIDFlag2["BLOCK_ID_FLAG_ABSENT"] = 1] = "BLOCK_ID_FLAG_ABSENT";
  BlockIDFlag2[BlockIDFlag2["BLOCK_ID_FLAG_COMMIT"] = 2] = "BLOCK_ID_FLAG_COMMIT";
  BlockIDFlag2[BlockIDFlag2["BLOCK_ID_FLAG_NIL"] = 3] = "BLOCK_ID_FLAG_NIL";
  return BlockIDFlag2;
})(BlockIDFlag || {});
class ValidatorSet$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.ValidatorSet", [
      { no: 1, name: "validators", kind: "message", repeat: 2, T: () => Validator },
      { no: 2, name: "proposer", kind: "message", T: () => Validator },
      {
        no: 3,
        name: "total_voting_power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validators = [];
    message.totalVotingPower = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cometbft.types.v1.Validator validators */
        1:
          message.validators.push(Validator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cometbft.types.v1.Validator proposer */
        2:
          message.proposer = Validator.internalBinaryRead(reader, reader.uint32(), options, message.proposer);
          break;
        case /* int64 total_voting_power */
        3:
          message.totalVotingPower = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.validators.length; i++)
      Validator.internalBinaryWrite(message.validators[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.proposer)
      Validator.internalBinaryWrite(message.proposer, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.totalVotingPower !== 0n)
      writer.tag(3, WireType.Varint).int64(message.totalVotingPower);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ValidatorSet = new ValidatorSet$Type();
class Validator$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.Validator", [
      {
        no: 1,
        name: "address",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "pub_key", kind: "message", T: () => PublicKey },
      {
        no: 3,
        name: "voting_power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "proposer_priority",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "pub_key_bytes",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 6,
        name: "pub_key_type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = new Uint8Array(0);
    message.votingPower = 0n;
    message.proposerPriority = 0n;
    message.pubKeyBytes = new Uint8Array(0);
    message.pubKeyType = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes address */
        1:
          message.address = reader.bytes();
          break;
        case /* cometbft.crypto.v1.PublicKey pub_key = 2 [deprecated = true] */
        2:
          message.pubKey = PublicKey.internalBinaryRead(reader, reader.uint32(), options, message.pubKey);
          break;
        case /* int64 voting_power */
        3:
          message.votingPower = reader.int64().toBigInt();
          break;
        case /* int64 proposer_priority */
        4:
          message.proposerPriority = reader.int64().toBigInt();
          break;
        case /* bytes pub_key_bytes */
        5:
          message.pubKeyBytes = reader.bytes();
          break;
        case /* string pub_key_type */
        6:
          message.pubKeyType = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.address);
    if (message.pubKey)
      PublicKey.internalBinaryWrite(message.pubKey, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.votingPower !== 0n)
      writer.tag(3, WireType.Varint).int64(message.votingPower);
    if (message.proposerPriority !== 0n)
      writer.tag(4, WireType.Varint).int64(message.proposerPriority);
    if (message.pubKeyBytes.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.pubKeyBytes);
    if (message.pubKeyType !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.pubKeyType);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Validator = new Validator$Type();
class SimpleValidator$Type extends MessageType {
  constructor() {
    super("cometbft.types.v1.SimpleValidator", [
      { no: 1, name: "pub_key", kind: "message", T: () => PublicKey },
      {
        no: 2,
        name: "voting_power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.votingPower = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.crypto.v1.PublicKey pub_key */
        1:
          message.pubKey = PublicKey.internalBinaryRead(reader, reader.uint32(), options, message.pubKey);
          break;
        case /* int64 voting_power */
        2:
          message.votingPower = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pubKey)
      PublicKey.internalBinaryWrite(message.pubKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.votingPower !== 0n)
      writer.tag(2, WireType.Varint).int64(message.votingPower);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SimpleValidator = new SimpleValidator$Type();
export {
  BlockIDFlag,
  SimpleValidator,
  Validator,
  ValidatorSet
};
