// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/feegrant/v1beta1/feegrant_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../google/protobuf/any_pb.js";
import { Duration } from "../../../google/protobuf/duration_pb.js";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Coin } from "../../base/v1beta1/coin_pb.js";
class BasicAllowance$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.BasicAllowance", [
      { no: 1, name: "spend_limit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } },
      { no: 2, name: "expiration", kind: "message", T: () => Timestamp, options: { "gogoproto.stdtime": true } }
    ], { "cosmos_proto.implements_interface": ["cosmos.feegrant.v1beta1.FeeAllowanceI"], "amino.name": "cosmos-sdk/BasicAllowance" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.spendLimit = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin spend_limit */
        1:
          message.spendLimit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* google.protobuf.Timestamp expiration */
        2:
          message.expiration = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.expiration);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.spendLimit.length; i++)
      Coin.internalBinaryWrite(message.spendLimit[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.expiration)
      Timestamp.internalBinaryWrite(message.expiration, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BasicAllowance = new BasicAllowance$Type();
class PeriodicAllowance$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.PeriodicAllowance", [
      { no: 1, name: "basic", kind: "message", T: () => BasicAllowance, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "period", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true, "amino.dont_omitempty": true } },
      { no: 3, name: "period_spend_limit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } },
      { no: 4, name: "period_can_spend", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } },
      { no: 5, name: "period_reset", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } }
    ], { "cosmos_proto.implements_interface": ["cosmos.feegrant.v1beta1.FeeAllowanceI"], "amino.name": "cosmos-sdk/PeriodicAllowance" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.periodSpendLimit = [];
    message.periodCanSpend = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.feegrant.v1beta1.BasicAllowance basic */
        1:
          message.basic = BasicAllowance.internalBinaryRead(reader, reader.uint32(), options, message.basic);
          break;
        case /* google.protobuf.Duration period */
        2:
          message.period = Duration.internalBinaryRead(reader, reader.uint32(), options, message.period);
          break;
        case /* repeated cosmos.base.v1beta1.Coin period_spend_limit */
        3:
          message.periodSpendLimit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.base.v1beta1.Coin period_can_spend */
        4:
          message.periodCanSpend.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* google.protobuf.Timestamp period_reset */
        5:
          message.periodReset = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.periodReset);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.basic)
      BasicAllowance.internalBinaryWrite(message.basic, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.period)
      Duration.internalBinaryWrite(message.period, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.periodSpendLimit.length; i++)
      Coin.internalBinaryWrite(message.periodSpendLimit[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.periodCanSpend.length; i++)
      Coin.internalBinaryWrite(message.periodCanSpend[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.periodReset)
      Timestamp.internalBinaryWrite(message.periodReset, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PeriodicAllowance = new PeriodicAllowance$Type();
class AllowedMsgAllowance$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.AllowedMsgAllowance", [
      { no: 1, name: "allowance", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.feegrant.v1beta1.FeeAllowanceI" } },
      {
        no: 2,
        name: "allowed_messages",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos_proto.implements_interface": ["cosmos.feegrant.v1beta1.FeeAllowanceI"], "amino.name": "cosmos-sdk/AllowedMsgAllowance" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.allowedMessages = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any allowance */
        1:
          message.allowance = Any.internalBinaryRead(reader, reader.uint32(), options, message.allowance);
          break;
        case /* repeated string allowed_messages */
        2:
          message.allowedMessages.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.allowance)
      Any.internalBinaryWrite(message.allowance, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.allowedMessages.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.allowedMessages[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AllowedMsgAllowance = new AllowedMsgAllowance$Type();
class Grant$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.Grant", [
      { no: 1, name: "granter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "grantee", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "allowance", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.feegrant.v1beta1.FeeAllowanceI" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.granter = "";
    message.grantee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string granter */
        1:
          message.granter = reader.string();
          break;
        case /* string grantee */
        2:
          message.grantee = reader.string();
          break;
        case /* google.protobuf.Any allowance */
        3:
          message.allowance = Any.internalBinaryRead(reader, reader.uint32(), options, message.allowance);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.granter !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.granter);
    if (message.grantee !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.grantee);
    if (message.allowance)
      Any.internalBinaryWrite(message.allowance, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Grant = new Grant$Type();
export {
  AllowedMsgAllowance,
  BasicAllowance,
  Grant,
  PeriodicAllowance
};
