// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/feegrant/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Allowance returns granted allwance to the grantee by the granter.
   *
   * @generated from protobuf rpc: Allowance
   */
  allowance(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Allowances returns all the grants for the given grantee address.
   *
   * @generated from protobuf rpc: Allowances
   */
  allowances(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * AllowancesByGranter returns all the grants given by an address
   *
   * Since: cosmos-sdk 0.46
   *
   * @generated from protobuf rpc: AllowancesByGranter
   */
  allowancesByGranter(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
