// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/feegrant/v1beta1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../google/protobuf/any_pb.js";
class MsgGrantAllowance$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.MsgGrantAllowance", [
      { no: 1, name: "granter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "grantee", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "allowance", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.feegrant.v1beta1.FeeAllowanceI" } }
    ], { "amino.name": "cosmos-sdk/MsgGrantAllowance", "cosmos.msg.v1.signer": ["granter"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.granter = "";
    message.grantee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string granter */
        1:
          message.granter = reader.string();
          break;
        case /* string grantee */
        2:
          message.grantee = reader.string();
          break;
        case /* google.protobuf.Any allowance */
        3:
          message.allowance = Any.internalBinaryRead(reader, reader.uint32(), options, message.allowance);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.granter !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.granter);
    if (message.grantee !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.grantee);
    if (message.allowance)
      Any.internalBinaryWrite(message.allowance, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgGrantAllowance = new MsgGrantAllowance$Type();
class MsgGrantAllowanceResponse$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.MsgGrantAllowanceResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgGrantAllowanceResponse = new MsgGrantAllowanceResponse$Type();
class MsgRevokeAllowance$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.MsgRevokeAllowance", [
      { no: 1, name: "granter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "grantee", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "amino.name": "cosmos-sdk/MsgRevokeAllowance", "cosmos.msg.v1.signer": ["granter"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.granter = "";
    message.grantee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string granter */
        1:
          message.granter = reader.string();
          break;
        case /* string grantee */
        2:
          message.grantee = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.granter !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.granter);
    if (message.grantee !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.grantee);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRevokeAllowance = new MsgRevokeAllowance$Type();
class MsgRevokeAllowanceResponse$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.MsgRevokeAllowanceResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRevokeAllowanceResponse = new MsgRevokeAllowanceResponse$Type();
class MsgPruneAllowances$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.MsgPruneAllowances", [
      { no: 1, name: "pruner", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "cosmos.msg.v1.signer": ["pruner"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.pruner = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string pruner */
        1:
          message.pruner = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pruner !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.pruner);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPruneAllowances = new MsgPruneAllowances$Type();
class MsgPruneAllowancesResponse$Type extends MessageType {
  constructor() {
    super("cosmos.feegrant.v1beta1.MsgPruneAllowancesResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPruneAllowancesResponse = new MsgPruneAllowancesResponse$Type();
const Msg = new ServiceType("cosmos.feegrant.v1beta1.Msg", [
  { name: "GrantAllowance", options: {}, I: MsgGrantAllowance, O: MsgGrantAllowanceResponse },
  { name: "RevokeAllowance", options: {}, I: MsgRevokeAllowance, O: MsgRevokeAllowanceResponse },
  { name: "PruneAllowances", options: {}, I: MsgPruneAllowances, O: MsgPruneAllowancesResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgGrantAllowance,
  MsgGrantAllowanceResponse,
  MsgPruneAllowances,
  MsgPruneAllowancesResponse,
  MsgRevokeAllowance,
  MsgRevokeAllowanceResponse
};
