// @injectivelabs/core-proto-ts-v2/esm/generated/tendermint/types/canonical_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Timestamp } from "../../google/protobuf/timestamp_pb.js";
import { SignedMsgType } from "./types_pb.js";
class CanonicalBlockID$Type extends MessageType {
  constructor() {
    super("tendermint.types.CanonicalBlockID", [
      {
        no: 1,
        name: "hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "part_set_header", kind: "message", T: () => CanonicalPartSetHeader, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.hash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes hash */
        1:
          message.hash = reader.bytes();
          break;
        case /* tendermint.types.CanonicalPartSetHeader part_set_header */
        2:
          message.partSetHeader = CanonicalPartSetHeader.internalBinaryRead(reader, reader.uint32(), options, message.partSetHeader);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.hash.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.hash);
    if (message.partSetHeader)
      CanonicalPartSetHeader.internalBinaryWrite(message.partSetHeader, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CanonicalBlockID = new CanonicalBlockID$Type();
class CanonicalPartSetHeader$Type extends MessageType {
  constructor() {
    super("tendermint.types.CanonicalPartSetHeader", [
      {
        no: 1,
        name: "total",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 2,
        name: "hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.total = 0;
    message.hash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 total */
        1:
          message.total = reader.uint32();
          break;
        case /* bytes hash */
        2:
          message.hash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.total !== 0)
      writer.tag(1, WireType.Varint).uint32(message.total);
    if (message.hash.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.hash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CanonicalPartSetHeader = new CanonicalPartSetHeader$Type();
class CanonicalProposal$Type extends MessageType {
  constructor() {
    super("tendermint.types.CanonicalProposal", [
      { no: 1, name: "type", kind: "enum", T: () => ["tendermint.types.SignedMsgType", SignedMsgType, "SIGNED_MSG_TYPE_"] },
      {
        no: 2,
        name: "height",
        kind: "scalar",
        T: 16,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "round",
        kind: "scalar",
        T: 16,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "pol_round", kind: "scalar", T: 3, L: 0, options: { "gogoproto.customname": "POLRound" } },
      { no: 5, name: "block_id", kind: "message", T: () => CanonicalBlockID, options: { "gogoproto.customname": "BlockID" } },
      { no: 6, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      { no: 7, name: "chain_id", kind: "scalar", T: 9, options: { "gogoproto.customname": "ChainID" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.type = 0;
    message.height = 0n;
    message.round = 0n;
    message.polRound = 0n;
    message.chainId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.SignedMsgType type */
        1:
          message.type = reader.int32();
          break;
        case /* sfixed64 height */
        2:
          message.height = reader.sfixed64().toBigInt();
          break;
        case /* sfixed64 round */
        3:
          message.round = reader.sfixed64().toBigInt();
          break;
        case /* int64 pol_round */
        4:
          message.polRound = reader.int64().toBigInt();
          break;
        case /* tendermint.types.CanonicalBlockID block_id */
        5:
          message.blockId = CanonicalBlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* google.protobuf.Timestamp timestamp */
        6:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        case /* string chain_id */
        7:
          message.chainId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.type !== 0)
      writer.tag(1, WireType.Varint).int32(message.type);
    if (message.height !== 0n)
      writer.tag(2, WireType.Bit64).sfixed64(message.height);
    if (message.round !== 0n)
      writer.tag(3, WireType.Bit64).sfixed64(message.round);
    if (message.polRound !== 0n)
      writer.tag(4, WireType.Varint).int64(message.polRound);
    if (message.blockId)
      CanonicalBlockID.internalBinaryWrite(message.blockId, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.chainId !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.chainId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CanonicalProposal = new CanonicalProposal$Type();
class CanonicalVote$Type extends MessageType {
  constructor() {
    super("tendermint.types.CanonicalVote", [
      { no: 1, name: "type", kind: "enum", T: () => ["tendermint.types.SignedMsgType", SignedMsgType, "SIGNED_MSG_TYPE_"] },
      {
        no: 2,
        name: "height",
        kind: "scalar",
        T: 16,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "round",
        kind: "scalar",
        T: 16,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "block_id", kind: "message", T: () => CanonicalBlockID, options: { "gogoproto.customname": "BlockID" } },
      { no: 5, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      { no: 6, name: "chain_id", kind: "scalar", T: 9, options: { "gogoproto.customname": "ChainID" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.type = 0;
    message.height = 0n;
    message.round = 0n;
    message.chainId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.SignedMsgType type */
        1:
          message.type = reader.int32();
          break;
        case /* sfixed64 height */
        2:
          message.height = reader.sfixed64().toBigInt();
          break;
        case /* sfixed64 round */
        3:
          message.round = reader.sfixed64().toBigInt();
          break;
        case /* tendermint.types.CanonicalBlockID block_id */
        4:
          message.blockId = CanonicalBlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* google.protobuf.Timestamp timestamp */
        5:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        case /* string chain_id */
        6:
          message.chainId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.type !== 0)
      writer.tag(1, WireType.Varint).int32(message.type);
    if (message.height !== 0n)
      writer.tag(2, WireType.Bit64).sfixed64(message.height);
    if (message.round !== 0n)
      writer.tag(3, WireType.Bit64).sfixed64(message.round);
    if (message.blockId)
      CanonicalBlockID.internalBinaryWrite(message.blockId, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.chainId !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.chainId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CanonicalVote = new CanonicalVote$Type();
class CanonicalVoteExtension$Type extends MessageType {
  constructor() {
    super("tendermint.types.CanonicalVoteExtension", [
      {
        no: 1,
        name: "extension",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "height",
        kind: "scalar",
        T: 16,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "round",
        kind: "scalar",
        T: 16,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "chain_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.extension = new Uint8Array(0);
    message.height = 0n;
    message.round = 0n;
    message.chainId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes extension */
        1:
          message.extension = reader.bytes();
          break;
        case /* sfixed64 height */
        2:
          message.height = reader.sfixed64().toBigInt();
          break;
        case /* sfixed64 round */
        3:
          message.round = reader.sfixed64().toBigInt();
          break;
        case /* string chain_id */
        4:
          message.chainId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.extension.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.extension);
    if (message.height !== 0n)
      writer.tag(2, WireType.Bit64).sfixed64(message.height);
    if (message.round !== 0n)
      writer.tag(3, WireType.Bit64).sfixed64(message.round);
    if (message.chainId !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.chainId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CanonicalVoteExtension = new CanonicalVoteExtension$Type();
export {
  CanonicalBlockID,
  CanonicalPartSetHeader,
  CanonicalProposal,
  CanonicalVote,
  CanonicalVoteExtension
};
