// @injectivelabs/core-proto-ts-v2/esm/generated/google/protobuf/wrappers_pb.js
import { ScalarType } from "@protobuf-ts/runtime";
import { LongType } from "@protobuf-ts/runtime";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class DoubleValue$Type extends MessageType {
  constructor() {
    super("google.protobuf.DoubleValue", [
      {
        no: 1,
        name: "value",
        kind: "scalar",
        T: 1
        /*ScalarType.DOUBLE*/
      }
    ]);
  }
  /**
   * Encode `DoubleValue` to JSON number.
   */
  internalJsonWrite(message, options) {
    return this.refJsonWriter.scalar(2, message.value, "value", false, true);
  }
  /**
   * Decode `DoubleValue` from JSON number.
   */
  internalJsonRead(json, options, target) {
    if (!target)
      target = this.create();
    target.value = this.refJsonReader.scalar(json, 1, void 0, "value");
    return target;
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* double value */
        1:
          message.value = reader.double();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value !== 0)
      writer.tag(1, WireType.Bit64).double(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DoubleValue = new DoubleValue$Type();
class FloatValue$Type extends MessageType {
  constructor() {
    super("google.protobuf.FloatValue", [
      {
        no: 1,
        name: "value",
        kind: "scalar",
        T: 2
        /*ScalarType.FLOAT*/
      }
    ]);
  }
  /**
   * Encode `FloatValue` to JSON number.
   */
  internalJsonWrite(message, options) {
    return this.refJsonWriter.scalar(1, message.value, "value", false, true);
  }
  /**
   * Decode `FloatValue` from JSON number.
   */
  internalJsonRead(json, options, target) {
    if (!target)
      target = this.create();
    target.value = this.refJsonReader.scalar(json, 1, void 0, "value");
    return target;
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* float value */
        1:
          message.value = reader.float();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value !== 0)
      writer.tag(1, WireType.Bit32).float(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FloatValue = new FloatValue$Type();
class Int64Value$Type extends MessageType {
  constructor() {
    super("google.protobuf.Int64Value", [
      {
        no: 1,
        name: "value",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  /**
   * Encode `Int64Value` to JSON string.
   */
  internalJsonWrite(message, options) {
    return this.refJsonWriter.scalar(ScalarType.INT64, message.value, "value", false, true);
  }
  /**
   * Decode `Int64Value` from JSON string.
   */
  internalJsonRead(json, options, target) {
    if (!target)
      target = this.create();
    target.value = this.refJsonReader.scalar(json, ScalarType.INT64, LongType.BIGINT, "value");
    return target;
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 value */
        1:
          message.value = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value !== 0n)
      writer.tag(1, WireType.Varint).int64(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Int64Value = new Int64Value$Type();
class UInt64Value$Type extends MessageType {
  constructor() {
    super("google.protobuf.UInt64Value", [
      {
        no: 1,
        name: "value",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  /**
   * Encode `UInt64Value` to JSON string.
   */
  internalJsonWrite(message, options) {
    return this.refJsonWriter.scalar(ScalarType.UINT64, message.value, "value", false, true);
  }
  /**
   * Decode `UInt64Value` from JSON string.
   */
  internalJsonRead(json, options, target) {
    if (!target)
      target = this.create();
    target.value = this.refJsonReader.scalar(json, ScalarType.UINT64, LongType.BIGINT, "value");
    return target;
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 value */
        1:
          message.value = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const UInt64Value = new UInt64Value$Type();
class Int32Value$Type extends MessageType {
  constructor() {
    super("google.protobuf.Int32Value", [
      {
        no: 1,
        name: "value",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      }
    ]);
  }
  /**
   * Encode `Int32Value` to JSON string.
   */
  internalJsonWrite(message, options) {
    return this.refJsonWriter.scalar(5, message.value, "value", false, true);
  }
  /**
   * Decode `Int32Value` from JSON string.
   */
  internalJsonRead(json, options, target) {
    if (!target)
      target = this.create();
    target.value = this.refJsonReader.scalar(json, 5, void 0, "value");
    return target;
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int32 value */
        1:
          message.value = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value !== 0)
      writer.tag(1, WireType.Varint).int32(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Int32Value = new Int32Value$Type();
class UInt32Value$Type extends MessageType {
  constructor() {
    super("google.protobuf.UInt32Value", [
      {
        no: 1,
        name: "value",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  /**
   * Encode `UInt32Value` to JSON string.
   */
  internalJsonWrite(message, options) {
    return this.refJsonWriter.scalar(13, message.value, "value", false, true);
  }
  /**
   * Decode `UInt32Value` from JSON string.
   */
  internalJsonRead(json, options, target) {
    if (!target)
      target = this.create();
    target.value = this.refJsonReader.scalar(json, 13, void 0, "value");
    return target;
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 value */
        1:
          message.value = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value !== 0)
      writer.tag(1, WireType.Varint).uint32(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const UInt32Value = new UInt32Value$Type();
class BoolValue$Type extends MessageType {
  constructor() {
    super("google.protobuf.BoolValue", [
      {
        no: 1,
        name: "value",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  /**
   * Encode `BoolValue` to JSON bool.
   */
  internalJsonWrite(message, options) {
    return message.value;
  }
  /**
   * Decode `BoolValue` from JSON bool.
   */
  internalJsonRead(json, options, target) {
    if (!target)
      target = this.create();
    target.value = this.refJsonReader.scalar(json, 8, void 0, "value");
    return target;
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bool value */
        1:
          message.value = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value !== false)
      writer.tag(1, WireType.Varint).bool(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BoolValue = new BoolValue$Type();
class StringValue$Type extends MessageType {
  constructor() {
    super("google.protobuf.StringValue", [
      {
        no: 1,
        name: "value",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  /**
   * Encode `StringValue` to JSON string.
   */
  internalJsonWrite(message, options) {
    return message.value;
  }
  /**
   * Decode `StringValue` from JSON string.
   */
  internalJsonRead(json, options, target) {
    if (!target)
      target = this.create();
    target.value = this.refJsonReader.scalar(json, 9, void 0, "value");
    return target;
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string value */
        1:
          message.value = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StringValue = new StringValue$Type();
class BytesValue$Type extends MessageType {
  constructor() {
    super("google.protobuf.BytesValue", [
      {
        no: 1,
        name: "value",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  /**
   * Encode `BytesValue` to JSON string.
   */
  internalJsonWrite(message, options) {
    return this.refJsonWriter.scalar(12, message.value, "value", false, true);
  }
  /**
   * Decode `BytesValue` from JSON string.
   */
  internalJsonRead(json, options, target) {
    if (!target)
      target = this.create();
    target.value = this.refJsonReader.scalar(json, 12, void 0, "value");
    return target;
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes value */
        1:
          message.value = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BytesValue = new BytesValue$Type();
export {
  BoolValue,
  BytesValue,
  DoubleValue,
  FloatValue,
  Int32Value,
  Int64Value,
  StringValue,
  UInt32Value,
  UInt64Value
};
