// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/tx/v1beta1/service_pb.client.js
import { Service } from "./service_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class ServiceClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Service.typeName;
    this.methods = Service.methods;
    this.options = Service.options;
  }
  /**
   * Simulate simulates executing a transaction for estimating gas usage.
   *
   * @generated from protobuf rpc: Simulate
   */
  simulate(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetTx fetches a tx by hash.
   *
   * @generated from protobuf rpc: GetTx
   */
  getTx(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * BroadcastTx broadcast transaction.
   *
   * @generated from protobuf rpc: BroadcastTx
   */
  broadcastTx(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetTxsEvent fetches txs by event.
   *
   * @generated from protobuf rpc: GetTxsEvent
   */
  getTxsEvent(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetBlockWithTxs fetches a block with decoded txs.
   *
   * Since: cosmos-sdk 0.45.2
   *
   * @generated from protobuf rpc: GetBlockWithTxs
   */
  getBlockWithTxs(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TxDecode decodes the transaction.
   *
   * Since: cosmos-sdk 0.47
   *
   * @generated from protobuf rpc: TxDecode
   */
  txDecode(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TxEncode encodes the transaction.
   *
   * Since: cosmos-sdk 0.47
   *
   * @generated from protobuf rpc: TxEncode
   */
  txEncode(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TxEncodeAmino encodes an Amino transaction from JSON to encoded bytes.
   *
   * Since: cosmos-sdk 0.47
   *
   * @generated from protobuf rpc: TxEncodeAmino
   */
  txEncodeAmino(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TxDecodeAmino decodes an Amino transaction from encoded bytes to JSON.
   *
   * Since: cosmos-sdk 0.47
   *
   * @generated from protobuf rpc: TxDecodeAmino
   */
  txDecodeAmino(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  ServiceClient
};
