// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/tx/v1beta1/service_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Block } from "../../../cometbft/types/v1/block_pb.js";
import { BlockID } from "../../../cometbft/types/v1/types_pb.js";
import { Result } from "../../base/abci/v1beta1/abci_pb.js";
import { GasInfo } from "../../base/abci/v1beta1/abci_pb.js";
import { PageResponse } from "../../base/query/v1beta1/pagination_pb.js";
import { TxResponse } from "../../base/abci/v1beta1/abci_pb.js";
import { Tx } from "./tx_pb.js";
import { PageRequest } from "../../base/query/v1beta1/pagination_pb.js";
var OrderBy = /* @__PURE__ */ ((OrderBy2) => {
  OrderBy2[OrderBy2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  OrderBy2[OrderBy2["ASC"] = 1] = "ASC";
  OrderBy2[OrderBy2["DESC"] = 2] = "DESC";
  return OrderBy2;
})(OrderBy || {});
var BroadcastMode = /* @__PURE__ */ ((BroadcastMode2) => {
  BroadcastMode2[BroadcastMode2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  BroadcastMode2[BroadcastMode2["BLOCK"] = 1] = "BLOCK";
  BroadcastMode2[BroadcastMode2["SYNC"] = 2] = "SYNC";
  BroadcastMode2[BroadcastMode2["ASYNC"] = 3] = "ASYNC";
  return BroadcastMode2;
})(BroadcastMode || {});
class GetTxsEventRequest$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.GetTxsEventRequest", [
      {
        no: 1,
        name: "events",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest },
      { no: 3, name: "order_by", kind: "enum", T: () => ["cosmos.tx.v1beta1.OrderBy", OrderBy, "ORDER_BY_"] },
      {
        no: 4,
        name: "page",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "limit",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "query",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.events = [];
    message.orderBy = 0;
    message.page = 0n;
    message.limit = 0n;
    message.query = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string events = 1 [deprecated = true] */
        1:
          message.events.push(reader.string());
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination = 2 [deprecated = true] */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* cosmos.tx.v1beta1.OrderBy order_by */
        3:
          message.orderBy = reader.int32();
          break;
        case /* uint64 page */
        4:
          message.page = reader.uint64().toBigInt();
          break;
        case /* uint64 limit */
        5:
          message.limit = reader.uint64().toBigInt();
          break;
        case /* string query */
        6:
          message.query = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.events.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.events[i]);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.orderBy !== 0)
      writer.tag(3, WireType.Varint).int32(message.orderBy);
    if (message.page !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.page);
    if (message.limit !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.limit);
    if (message.query !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.query);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetTxsEventRequest = new GetTxsEventRequest$Type();
class GetTxsEventResponse$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.GetTxsEventResponse", [
      { no: 1, name: "txs", kind: "message", repeat: 2, T: () => Tx },
      { no: 2, name: "tx_responses", kind: "message", repeat: 2, T: () => TxResponse },
      { no: 3, name: "pagination", kind: "message", T: () => PageResponse },
      {
        no: 4,
        name: "total",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.txs = [];
    message.txResponses = [];
    message.total = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.tx.v1beta1.Tx txs */
        1:
          message.txs.push(Tx.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.base.abci.v1beta1.TxResponse tx_responses */
        2:
          message.txResponses.push(TxResponse.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination = 3 [deprecated = true] */
        3:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* uint64 total */
        4:
          message.total = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.txs.length; i++)
      Tx.internalBinaryWrite(message.txs[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.txResponses.length; i++)
      TxResponse.internalBinaryWrite(message.txResponses[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.total !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.total);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetTxsEventResponse = new GetTxsEventResponse$Type();
class BroadcastTxRequest$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.BroadcastTxRequest", [
      {
        no: 1,
        name: "tx_bytes",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "mode", kind: "enum", T: () => ["cosmos.tx.v1beta1.BroadcastMode", BroadcastMode, "BROADCAST_MODE_"] }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.txBytes = new Uint8Array(0);
    message.mode = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes tx_bytes */
        1:
          message.txBytes = reader.bytes();
          break;
        case /* cosmos.tx.v1beta1.BroadcastMode mode */
        2:
          message.mode = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.txBytes.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.txBytes);
    if (message.mode !== 0)
      writer.tag(2, WireType.Varint).int32(message.mode);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BroadcastTxRequest = new BroadcastTxRequest$Type();
class BroadcastTxResponse$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.BroadcastTxResponse", [
      { no: 1, name: "tx_response", kind: "message", T: () => TxResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.abci.v1beta1.TxResponse tx_response */
        1:
          message.txResponse = TxResponse.internalBinaryRead(reader, reader.uint32(), options, message.txResponse);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.txResponse)
      TxResponse.internalBinaryWrite(message.txResponse, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BroadcastTxResponse = new BroadcastTxResponse$Type();
class SimulateRequest$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.SimulateRequest", [
      { no: 1, name: "tx", kind: "message", T: () => Tx },
      {
        no: 2,
        name: "tx_bytes",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.txBytes = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.tx.v1beta1.Tx tx = 1 [deprecated = true] */
        1:
          message.tx = Tx.internalBinaryRead(reader, reader.uint32(), options, message.tx);
          break;
        case /* bytes tx_bytes */
        2:
          message.txBytes = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.tx)
      Tx.internalBinaryWrite(message.tx, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.txBytes.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.txBytes);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SimulateRequest = new SimulateRequest$Type();
class SimulateResponse$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.SimulateResponse", [
      { no: 1, name: "gas_info", kind: "message", T: () => GasInfo },
      { no: 2, name: "result", kind: "message", T: () => Result }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.abci.v1beta1.GasInfo gas_info */
        1:
          message.gasInfo = GasInfo.internalBinaryRead(reader, reader.uint32(), options, message.gasInfo);
          break;
        case /* cosmos.base.abci.v1beta1.Result result */
        2:
          message.result = Result.internalBinaryRead(reader, reader.uint32(), options, message.result);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.gasInfo)
      GasInfo.internalBinaryWrite(message.gasInfo, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.result)
      Result.internalBinaryWrite(message.result, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SimulateResponse = new SimulateResponse$Type();
class GetTxRequest$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.GetTxRequest", [
      {
        no: 1,
        name: "hash",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.hash = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string hash */
        1:
          message.hash = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.hash !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.hash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetTxRequest = new GetTxRequest$Type();
class GetTxResponse$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.GetTxResponse", [
      { no: 1, name: "tx", kind: "message", T: () => Tx },
      { no: 2, name: "tx_response", kind: "message", T: () => TxResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.tx.v1beta1.Tx tx */
        1:
          message.tx = Tx.internalBinaryRead(reader, reader.uint32(), options, message.tx);
          break;
        case /* cosmos.base.abci.v1beta1.TxResponse tx_response */
        2:
          message.txResponse = TxResponse.internalBinaryRead(reader, reader.uint32(), options, message.txResponse);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.tx)
      Tx.internalBinaryWrite(message.tx, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.txResponse)
      TxResponse.internalBinaryWrite(message.txResponse, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetTxResponse = new GetTxResponse$Type();
class GetBlockWithTxsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.GetBlockWithTxsRequest", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetBlockWithTxsRequest = new GetBlockWithTxsRequest$Type();
class GetBlockWithTxsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.GetBlockWithTxsResponse", [
      { no: 1, name: "txs", kind: "message", repeat: 2, T: () => Tx },
      { no: 2, name: "block_id", kind: "message", T: () => BlockID },
      { no: 3, name: "block", kind: "message", T: () => Block },
      { no: 4, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.txs = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.tx.v1beta1.Tx txs */
        1:
          message.txs.push(Tx.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cometbft.types.v1.BlockID block_id */
        2:
          message.blockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* cometbft.types.v1.Block block */
        3:
          message.block = Block.internalBinaryRead(reader, reader.uint32(), options, message.block);
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        4:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.txs.length; i++)
      Tx.internalBinaryWrite(message.txs[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.blockId)
      BlockID.internalBinaryWrite(message.blockId, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.block)
      Block.internalBinaryWrite(message.block, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetBlockWithTxsResponse = new GetBlockWithTxsResponse$Type();
class TxDecodeRequest$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.TxDecodeRequest", [
      {
        no: 1,
        name: "tx_bytes",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.txBytes = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes tx_bytes */
        1:
          message.txBytes = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.txBytes.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.txBytes);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxDecodeRequest = new TxDecodeRequest$Type();
class TxDecodeResponse$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.TxDecodeResponse", [
      { no: 1, name: "tx", kind: "message", T: () => Tx }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.tx.v1beta1.Tx tx */
        1:
          message.tx = Tx.internalBinaryRead(reader, reader.uint32(), options, message.tx);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.tx)
      Tx.internalBinaryWrite(message.tx, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxDecodeResponse = new TxDecodeResponse$Type();
class TxEncodeRequest$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.TxEncodeRequest", [
      { no: 1, name: "tx", kind: "message", T: () => Tx }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.tx.v1beta1.Tx tx */
        1:
          message.tx = Tx.internalBinaryRead(reader, reader.uint32(), options, message.tx);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.tx)
      Tx.internalBinaryWrite(message.tx, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxEncodeRequest = new TxEncodeRequest$Type();
class TxEncodeResponse$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.TxEncodeResponse", [
      {
        no: 1,
        name: "tx_bytes",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.txBytes = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes tx_bytes */
        1:
          message.txBytes = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.txBytes.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.txBytes);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxEncodeResponse = new TxEncodeResponse$Type();
class TxEncodeAminoRequest$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.TxEncodeAminoRequest", [
      {
        no: 1,
        name: "amino_json",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.aminoJson = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string amino_json */
        1:
          message.aminoJson = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.aminoJson !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.aminoJson);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxEncodeAminoRequest = new TxEncodeAminoRequest$Type();
class TxEncodeAminoResponse$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.TxEncodeAminoResponse", [
      {
        no: 1,
        name: "amino_binary",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.aminoBinary = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes amino_binary */
        1:
          message.aminoBinary = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.aminoBinary.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.aminoBinary);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxEncodeAminoResponse = new TxEncodeAminoResponse$Type();
class TxDecodeAminoRequest$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.TxDecodeAminoRequest", [
      {
        no: 1,
        name: "amino_binary",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.aminoBinary = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes amino_binary */
        1:
          message.aminoBinary = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.aminoBinary.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.aminoBinary);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxDecodeAminoRequest = new TxDecodeAminoRequest$Type();
class TxDecodeAminoResponse$Type extends MessageType {
  constructor() {
    super("cosmos.tx.v1beta1.TxDecodeAminoResponse", [
      {
        no: 1,
        name: "amino_json",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.aminoJson = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string amino_json */
        1:
          message.aminoJson = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.aminoJson !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.aminoJson);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxDecodeAminoResponse = new TxDecodeAminoResponse$Type();
const Service = new ServiceType("cosmos.tx.v1beta1.Service", [
  { name: "Simulate", options: { "google.api.http": { post: "/cosmos/tx/v1beta1/simulate", body: "*" } }, I: SimulateRequest, O: SimulateResponse },
  { name: "GetTx", options: { "google.api.http": { get: "/cosmos/tx/v1beta1/txs/{hash}" } }, I: GetTxRequest, O: GetTxResponse },
  { name: "BroadcastTx", options: { "google.api.http": { post: "/cosmos/tx/v1beta1/txs", body: "*" } }, I: BroadcastTxRequest, O: BroadcastTxResponse },
  { name: "GetTxsEvent", options: { "google.api.http": { get: "/cosmos/tx/v1beta1/txs" } }, I: GetTxsEventRequest, O: GetTxsEventResponse },
  { name: "GetBlockWithTxs", options: { "google.api.http": { get: "/cosmos/tx/v1beta1/txs/block/{height}" } }, I: GetBlockWithTxsRequest, O: GetBlockWithTxsResponse },
  { name: "TxDecode", options: { "google.api.http": { post: "/cosmos/tx/v1beta1/decode", body: "*" } }, I: TxDecodeRequest, O: TxDecodeResponse },
  { name: "TxEncode", options: { "google.api.http": { post: "/cosmos/tx/v1beta1/encode", body: "*" } }, I: TxEncodeRequest, O: TxEncodeResponse },
  { name: "TxEncodeAmino", options: { "google.api.http": { post: "/cosmos/tx/v1beta1/encode/amino", body: "*" } }, I: TxEncodeAminoRequest, O: TxEncodeAminoResponse },
  { name: "TxDecodeAmino", options: { "google.api.http": { post: "/cosmos/tx/v1beta1/decode/amino", body: "*" } }, I: TxDecodeAminoRequest, O: TxDecodeAminoResponse }
]);
export {
  BroadcastMode,
  BroadcastTxRequest,
  BroadcastTxResponse,
  GetBlockWithTxsRequest,
  GetBlockWithTxsResponse,
  GetTxRequest,
  GetTxResponse,
  GetTxsEventRequest,
  GetTxsEventResponse,
  OrderBy,
  Service,
  SimulateRequest,
  SimulateResponse,
  TxDecodeAminoRequest,
  TxDecodeAminoResponse,
  TxDecodeRequest,
  TxDecodeResponse,
  TxEncodeAminoRequest,
  TxEncodeAminoResponse,
  TxEncodeRequest,
  TxEncodeResponse
};
