// @injectivelabs/olp-proto-ts-v2/generated/goagen_olp_injective_dmm_v2_pb.client.js
import { InjectiveDmmV2 } from "./goagen_olp_injective_dmm_v2_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveDmmV2Client {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveDmmV2.typeName;
    this.methods = InjectiveDmmV2.methods;
    this.options = InjectiveDmmV2.options;
  }
  /**
   * Get epochs, optionally filtered by status
   *
   * @generated from protobuf rpc: GetEpochs
   */
  getEpochs(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get total scores for a market in an epoch
   *
   * @generated from protobuf rpc: GetTotalScores
   */
  getTotalScores(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get historical total scores for an account in a market
   *
   * @generated from protobuf rpc: GetTotalScoresHistory
   */
  getTotalScoresHistory(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get epoch scores
   *
   * @generated from protobuf rpc: GetEpochScores
   */
  getEpochScores(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get historical epoch scores for an account
   *
   * @generated from protobuf rpc: GetEpochScoresHistory
   */
  getEpochScoresHistory(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get market rewards for an epoch
   *
   * @generated from protobuf rpc: GetMarketRewards
   */
  getMarketRewards(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get eligible addresses for rewards in an epoch
   *
   * @generated from protobuf rpc: GetEligibleAddresses
   */
  getEligibleAddresses(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get reward distribution for an epoch
   *
   * @generated from protobuf rpc: GetRewardsDistribution
   */
  getRewardsDistribution(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get volume data for an account in an epoch
   *
   * @generated from protobuf rpc: GetAccountVolumes
   */
  getAccountVolumes(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get rewards eligibility information for an account
   *
   * @generated from protobuf rpc: GetRewardsEligibility
   */
  getRewardsEligibility(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get health status of the DMM service
   *
   * @generated from protobuf rpc: GetHealthStatus
   */
  getHealthStatus(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get market reward ranges for an epoch
   *
   * @generated from protobuf rpc: GetMarketRewardsRange
   */
  getMarketRewardsRange(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveDmmV2Client
};
