// @injectivelabs/olp-proto-ts-v2/generated/goagen_olp_injective_dmm_v2_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class GetEpochsRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetEpochsRequest", [
      {
        no: 1,
        name: "status",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional string status */
        1:
          message.status = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.status !== void 0)
      writer.tag(1, WireType.LengthDelimited).string(message.status);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetEpochsRequest = new GetEpochsRequest$Type();
class GetEpochsResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetEpochsResponse", [
      { no: 1, name: "epochs", kind: "message", repeat: 2, T: () => EpochV2 }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochs = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.EpochV2 epochs */
        1:
          message.epochs.push(EpochV2.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.epochs.length; i++)
      EpochV2.internalBinaryWrite(message.epochs[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetEpochsResponse = new GetEpochsResponse$Type();
class EpochV2$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.EpochV2", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "status",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "start_height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "end_height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "snapshot_count",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 6,
        name: "result_count",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      { no: 7, name: "config", kind: "message", T: () => EpochConfigV2 },
      {
        no: 8,
        name: "created_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "updated_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.status = "";
    message.startHeight = 0n;
    message.endHeight = 0n;
    message.snapshotCount = 0;
    message.resultCount = 0;
    message.createdAt = "";
    message.updatedAt = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string status */
        2:
          message.status = reader.string();
          break;
        case /* sint64 start_height */
        3:
          message.startHeight = reader.sint64().toBigInt();
          break;
        case /* sint64 end_height */
        4:
          message.endHeight = reader.sint64().toBigInt();
          break;
        case /* sint32 snapshot_count */
        5:
          message.snapshotCount = reader.sint32();
          break;
        case /* sint32 result_count */
        6:
          message.resultCount = reader.sint32();
          break;
        case /* injective_dmm_v2.EpochConfigV2 config */
        7:
          message.config = EpochConfigV2.internalBinaryRead(reader, reader.uint32(), options, message.config);
          break;
        case /* string created_at */
        8:
          message.createdAt = reader.string();
          break;
        case /* string updated_at */
        9:
          message.updatedAt = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.status !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.status);
    if (message.startHeight !== 0n)
      writer.tag(3, WireType.Varint).sint64(message.startHeight);
    if (message.endHeight !== 0n)
      writer.tag(4, WireType.Varint).sint64(message.endHeight);
    if (message.snapshotCount !== 0)
      writer.tag(5, WireType.Varint).sint32(message.snapshotCount);
    if (message.resultCount !== 0)
      writer.tag(6, WireType.Varint).sint32(message.resultCount);
    if (message.config)
      EpochConfigV2.internalBinaryWrite(message.config, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.createdAt !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.createdAt);
    if (message.updatedAt !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.updatedAt);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EpochV2 = new EpochV2$Type();
class EpochConfigV2$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.EpochConfigV2", [
      {
        no: 1,
        name: "start_date",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "end_date",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "reward_inj",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "markets", kind: "message", repeat: 2, T: () => Market },
      {
        no: 5,
        name: "liquidity_score_exponent",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "uptime_exponent",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "volume_exponent",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "permanence_volume_threshold",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "qualifying_volume_threshold",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "number",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 11,
        name: "is_mini_epoch",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.startDate = "";
    message.endDate = "";
    message.rewardInj = "";
    message.markets = [];
    message.liquidityScoreExponent = "";
    message.uptimeExponent = "";
    message.volumeExponent = "";
    message.number = 0;
    message.isMiniEpoch = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string start_date */
        1:
          message.startDate = reader.string();
          break;
        case /* string end_date */
        2:
          message.endDate = reader.string();
          break;
        case /* string reward_inj */
        3:
          message.rewardInj = reader.string();
          break;
        case /* repeated injective_dmm_v2.Market markets */
        4:
          message.markets.push(Market.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string liquidity_score_exponent */
        5:
          message.liquidityScoreExponent = reader.string();
          break;
        case /* string uptime_exponent */
        6:
          message.uptimeExponent = reader.string();
          break;
        case /* string volume_exponent */
        7:
          message.volumeExponent = reader.string();
          break;
        case /* optional string permanence_volume_threshold */
        8:
          message.permanenceVolumeThreshold = reader.string();
          break;
        case /* optional string qualifying_volume_threshold */
        9:
          message.qualifyingVolumeThreshold = reader.string();
          break;
        case /* sint32 number */
        10:
          message.number = reader.sint32();
          break;
        case /* bool is_mini_epoch */
        11:
          message.isMiniEpoch = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.startDate !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.startDate);
    if (message.endDate !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.endDate);
    if (message.rewardInj !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.rewardInj);
    for (let i = 0; i < message.markets.length; i++)
      Market.internalBinaryWrite(message.markets[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.liquidityScoreExponent !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.liquidityScoreExponent);
    if (message.uptimeExponent !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.uptimeExponent);
    if (message.volumeExponent !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.volumeExponent);
    if (message.permanenceVolumeThreshold !== void 0)
      writer.tag(8, WireType.LengthDelimited).string(message.permanenceVolumeThreshold);
    if (message.qualifyingVolumeThreshold !== void 0)
      writer.tag(9, WireType.LengthDelimited).string(message.qualifyingVolumeThreshold);
    if (message.number !== 0)
      writer.tag(10, WireType.Varint).sint32(message.number);
    if (message.isMiniEpoch !== false)
      writer.tag(11, WireType.Varint).bool(message.isMiniEpoch);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EpochConfigV2 = new EpochConfigV2$Type();
class Market$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.Market", [
      {
        no: 1,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "ticker",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "start_date",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "pre_allocated_reward",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.marketId = "";
    message.ticker = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string market_id */
        1:
          message.marketId = reader.string();
          break;
        case /* string ticker */
        2:
          message.ticker = reader.string();
          break;
        case /* optional string start_date */
        3:
          message.startDate = reader.string();
          break;
        case /* optional string pre_allocated_reward */
        4:
          message.preAllocatedReward = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.marketId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.marketId);
    if (message.ticker !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.ticker);
    if (message.startDate !== void 0)
      writer.tag(3, WireType.LengthDelimited).string(message.startDate);
    if (message.preAllocatedReward !== void 0)
      writer.tag(4, WireType.LengthDelimited).string(message.preAllocatedReward);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Market = new Market$Type();
class GetTotalScoresRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetTotalScoresRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "page", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.marketId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string market_id */
        2:
          message.marketId = reader.string();
          break;
        case /* injective_dmm_v2.Pagination page */
        3:
          message.page = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.page);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.marketId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    if (message.page)
      Pagination.internalBinaryWrite(message.page, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetTotalScoresRequest = new GetTotalScoresRequest$Type();
class Pagination$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.Pagination", [
      {
        no: 1,
        name: "from",
        kind: "scalar",
        opt: true,
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 2,
        name: "height",
        kind: "scalar",
        opt: true,
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 3,
        name: "per_page",
        kind: "scalar",
        opt: true,
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 4,
        name: "total",
        kind: "scalar",
        opt: true,
        T: 17
        /*ScalarType.SINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional sint32 from */
        1:
          message.from = reader.sint32();
          break;
        case /* optional sint32 height */
        2:
          message.height = reader.sint32();
          break;
        case /* optional sint32 per_page */
        3:
          message.perPage = reader.sint32();
          break;
        case /* optional sint32 total */
        4:
          message.total = reader.sint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.from !== void 0)
      writer.tag(1, WireType.Varint).sint32(message.from);
    if (message.height !== void 0)
      writer.tag(2, WireType.Varint).sint32(message.height);
    if (message.perPage !== void 0)
      writer.tag(3, WireType.Varint).sint32(message.perPage);
    if (message.total !== void 0)
      writer.tag(4, WireType.Varint).sint32(message.total);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Pagination = new Pagination$Type();
class GetTotalScoresResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetTotalScoresResponse", [
      { no: 1, name: "scores", kind: "message", repeat: 2, T: () => TotalScore },
      { no: 2, name: "next", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.scores = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.TotalScore scores */
        1:
          message.scores.push(TotalScore.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* injective_dmm_v2.Pagination next */
        2:
          message.next = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.next);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.scores.length; i++)
      TotalScore.internalBinaryWrite(message.scores[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.next)
      Pagination.internalBinaryWrite(message.next, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetTotalScoresResponse = new GetTotalScoresResponse$Type();
class TotalScore$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.TotalScore", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "start_height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "block_time",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "bid",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "ask",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "depth",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "liquidity_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "liquidity_score_ponderated",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 12,
        name: "uptime_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 13,
        name: "uptime_score_ponderated",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 14,
        name: "uptime_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 15,
        name: "start_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 16,
        name: "current_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 17,
        name: "volume_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 18,
        name: "volume_score_ponderated",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 22,
        name: "total_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 23,
        name: "reward",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 24,
        name: "reward_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 26,
        name: "created_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 27,
        name: "updated_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 28,
        name: "snapshot_count",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 29,
        name: "volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 30,
        name: "taker_start_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 31,
        name: "taker_current_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 32,
        name: "taker_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 33,
        name: "maker_start_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 34,
        name: "maker_current_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 35,
        name: "maker_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 36,
        name: "bid_snapshot",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 37,
        name: "ask_snapshot",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 38,
        name: "depth_snapshot",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 39,
        name: "liquidity_score_snapshot",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 40,
        name: "uptime_score_snapshot",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 41,
        name: "skip",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.marketId = "";
    message.accountAddress = "";
    message.height = 0n;
    message.startHeight = 0n;
    message.blockTime = "";
    message.bid = "";
    message.ask = "";
    message.depth = "";
    message.liquidityScore = "";
    message.liquidityScorePonderated = "";
    message.uptimeScore = "";
    message.uptimeScorePonderated = "";
    message.uptimePercentage = "";
    message.startVolume = "";
    message.currentVolume = "";
    message.volumeScore = "";
    message.volumeScorePonderated = "";
    message.totalScore = "";
    message.reward = "";
    message.rewardPercentage = "";
    message.createdAt = "";
    message.updatedAt = "";
    message.snapshotCount = 0;
    message.volume = "";
    message.takerStartVolume = "";
    message.takerCurrentVolume = "";
    message.takerVolume = "";
    message.makerStartVolume = "";
    message.makerCurrentVolume = "";
    message.makerVolume = "";
    message.bidSnapshot = "";
    message.askSnapshot = "";
    message.depthSnapshot = "";
    message.liquidityScoreSnapshot = "";
    message.uptimeScoreSnapshot = "";
    message.skip = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string market_id */
        2:
          message.marketId = reader.string();
          break;
        case /* string account_address */
        3:
          message.accountAddress = reader.string();
          break;
        case /* sint64 height */
        4:
          message.height = reader.sint64().toBigInt();
          break;
        case /* sint64 start_height */
        5:
          message.startHeight = reader.sint64().toBigInt();
          break;
        case /* string block_time */
        6:
          message.blockTime = reader.string();
          break;
        case /* string bid */
        7:
          message.bid = reader.string();
          break;
        case /* string ask */
        8:
          message.ask = reader.string();
          break;
        case /* string depth */
        9:
          message.depth = reader.string();
          break;
        case /* string liquidity_score */
        10:
          message.liquidityScore = reader.string();
          break;
        case /* string liquidity_score_ponderated */
        11:
          message.liquidityScorePonderated = reader.string();
          break;
        case /* string uptime_score */
        12:
          message.uptimeScore = reader.string();
          break;
        case /* string uptime_score_ponderated */
        13:
          message.uptimeScorePonderated = reader.string();
          break;
        case /* string uptime_percentage */
        14:
          message.uptimePercentage = reader.string();
          break;
        case /* string start_volume */
        15:
          message.startVolume = reader.string();
          break;
        case /* string current_volume */
        16:
          message.currentVolume = reader.string();
          break;
        case /* string volume_score */
        17:
          message.volumeScore = reader.string();
          break;
        case /* string volume_score_ponderated */
        18:
          message.volumeScorePonderated = reader.string();
          break;
        case /* string total_score */
        22:
          message.totalScore = reader.string();
          break;
        case /* string reward */
        23:
          message.reward = reader.string();
          break;
        case /* string reward_percentage */
        24:
          message.rewardPercentage = reader.string();
          break;
        case /* string created_at */
        26:
          message.createdAt = reader.string();
          break;
        case /* string updated_at */
        27:
          message.updatedAt = reader.string();
          break;
        case /* sint32 snapshot_count */
        28:
          message.snapshotCount = reader.sint32();
          break;
        case /* string volume */
        29:
          message.volume = reader.string();
          break;
        case /* string taker_start_volume */
        30:
          message.takerStartVolume = reader.string();
          break;
        case /* string taker_current_volume */
        31:
          message.takerCurrentVolume = reader.string();
          break;
        case /* string taker_volume */
        32:
          message.takerVolume = reader.string();
          break;
        case /* string maker_start_volume */
        33:
          message.makerStartVolume = reader.string();
          break;
        case /* string maker_current_volume */
        34:
          message.makerCurrentVolume = reader.string();
          break;
        case /* string maker_volume */
        35:
          message.makerVolume = reader.string();
          break;
        case /* string bid_snapshot */
        36:
          message.bidSnapshot = reader.string();
          break;
        case /* string ask_snapshot */
        37:
          message.askSnapshot = reader.string();
          break;
        case /* string depth_snapshot */
        38:
          message.depthSnapshot = reader.string();
          break;
        case /* string liquidity_score_snapshot */
        39:
          message.liquidityScoreSnapshot = reader.string();
          break;
        case /* string uptime_score_snapshot */
        40:
          message.uptimeScoreSnapshot = reader.string();
          break;
        case /* bool skip */
        41:
          message.skip = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.marketId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    if (message.accountAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.accountAddress);
    if (message.height !== 0n)
      writer.tag(4, WireType.Varint).sint64(message.height);
    if (message.startHeight !== 0n)
      writer.tag(5, WireType.Varint).sint64(message.startHeight);
    if (message.blockTime !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.blockTime);
    if (message.bid !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.bid);
    if (message.ask !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.ask);
    if (message.depth !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.depth);
    if (message.liquidityScore !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.liquidityScore);
    if (message.liquidityScorePonderated !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.liquidityScorePonderated);
    if (message.uptimeScore !== "")
      writer.tag(12, WireType.LengthDelimited).string(message.uptimeScore);
    if (message.uptimeScorePonderated !== "")
      writer.tag(13, WireType.LengthDelimited).string(message.uptimeScorePonderated);
    if (message.uptimePercentage !== "")
      writer.tag(14, WireType.LengthDelimited).string(message.uptimePercentage);
    if (message.startVolume !== "")
      writer.tag(15, WireType.LengthDelimited).string(message.startVolume);
    if (message.currentVolume !== "")
      writer.tag(16, WireType.LengthDelimited).string(message.currentVolume);
    if (message.volumeScore !== "")
      writer.tag(17, WireType.LengthDelimited).string(message.volumeScore);
    if (message.volumeScorePonderated !== "")
      writer.tag(18, WireType.LengthDelimited).string(message.volumeScorePonderated);
    if (message.totalScore !== "")
      writer.tag(22, WireType.LengthDelimited).string(message.totalScore);
    if (message.reward !== "")
      writer.tag(23, WireType.LengthDelimited).string(message.reward);
    if (message.rewardPercentage !== "")
      writer.tag(24, WireType.LengthDelimited).string(message.rewardPercentage);
    if (message.createdAt !== "")
      writer.tag(26, WireType.LengthDelimited).string(message.createdAt);
    if (message.updatedAt !== "")
      writer.tag(27, WireType.LengthDelimited).string(message.updatedAt);
    if (message.snapshotCount !== 0)
      writer.tag(28, WireType.Varint).sint32(message.snapshotCount);
    if (message.volume !== "")
      writer.tag(29, WireType.LengthDelimited).string(message.volume);
    if (message.takerStartVolume !== "")
      writer.tag(30, WireType.LengthDelimited).string(message.takerStartVolume);
    if (message.takerCurrentVolume !== "")
      writer.tag(31, WireType.LengthDelimited).string(message.takerCurrentVolume);
    if (message.takerVolume !== "")
      writer.tag(32, WireType.LengthDelimited).string(message.takerVolume);
    if (message.makerStartVolume !== "")
      writer.tag(33, WireType.LengthDelimited).string(message.makerStartVolume);
    if (message.makerCurrentVolume !== "")
      writer.tag(34, WireType.LengthDelimited).string(message.makerCurrentVolume);
    if (message.makerVolume !== "")
      writer.tag(35, WireType.LengthDelimited).string(message.makerVolume);
    if (message.bidSnapshot !== "")
      writer.tag(36, WireType.LengthDelimited).string(message.bidSnapshot);
    if (message.askSnapshot !== "")
      writer.tag(37, WireType.LengthDelimited).string(message.askSnapshot);
    if (message.depthSnapshot !== "")
      writer.tag(38, WireType.LengthDelimited).string(message.depthSnapshot);
    if (message.liquidityScoreSnapshot !== "")
      writer.tag(39, WireType.LengthDelimited).string(message.liquidityScoreSnapshot);
    if (message.uptimeScoreSnapshot !== "")
      writer.tag(40, WireType.LengthDelimited).string(message.uptimeScoreSnapshot);
    if (message.skip !== false)
      writer.tag(41, WireType.Varint).bool(message.skip);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TotalScore = new TotalScore$Type();
class GetTotalScoresHistoryRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetTotalScoresHistoryRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "page", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.marketId = "";
    message.accountAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string market_id */
        2:
          message.marketId = reader.string();
          break;
        case /* string account_address */
        3:
          message.accountAddress = reader.string();
          break;
        case /* injective_dmm_v2.Pagination page */
        4:
          message.page = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.page);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.marketId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    if (message.accountAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.accountAddress);
    if (message.page)
      Pagination.internalBinaryWrite(message.page, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetTotalScoresHistoryRequest = new GetTotalScoresHistoryRequest$Type();
class GetTotalScoresHistoryResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetTotalScoresHistoryResponse", [
      { no: 1, name: "scores", kind: "message", repeat: 2, T: () => TotalScore },
      { no: 2, name: "next", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.scores = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.TotalScore scores */
        1:
          message.scores.push(TotalScore.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* injective_dmm_v2.Pagination next */
        2:
          message.next = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.next);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.scores.length; i++)
      TotalScore.internalBinaryWrite(message.scores[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.next)
      Pagination.internalBinaryWrite(message.next, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetTotalScoresHistoryResponse = new GetTotalScoresHistoryResponse$Type();
class GetEpochScoresRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetEpochScoresRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "page", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* injective_dmm_v2.Pagination page */
        2:
          message.page = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.page);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.page)
      Pagination.internalBinaryWrite(message.page, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetEpochScoresRequest = new GetEpochScoresRequest$Type();
class GetEpochScoresResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetEpochScoresResponse", [
      { no: 1, name: "scores", kind: "message", repeat: 2, T: () => EpochScore },
      { no: 2, name: "next", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.scores = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.EpochScore scores */
        1:
          message.scores.push(EpochScore.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* injective_dmm_v2.Pagination next */
        2:
          message.next = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.next);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.scores.length; i++)
      EpochScore.internalBinaryWrite(message.scores[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.next)
      Pagination.internalBinaryWrite(message.next, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetEpochScoresResponse = new GetEpochScoresResponse$Type();
class EpochScore$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.EpochScore", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "block_time",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "start_height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "liquidity_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "liquidity_score_ponderated",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "uptime_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "uptime_score_ponderated",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "volume_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "volume_score_ponderated",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 12,
        name: "total_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 13,
        name: "volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 14,
        name: "maker_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 15,
        name: "taker_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 16,
        name: "reward",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 17,
        name: "reward_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 18,
        name: "qualifies",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 19,
        name: "volume_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 20,
        name: "created_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 21,
        name: "updated_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 22,
        name: "uptime_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 23,
        name: "depth",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.accountAddress = "";
    message.height = 0n;
    message.blockTime = "";
    message.startHeight = 0n;
    message.liquidityScore = "";
    message.liquidityScorePonderated = "";
    message.uptimeScore = "";
    message.uptimeScorePonderated = "";
    message.volumeScore = "";
    message.volumeScorePonderated = "";
    message.totalScore = "";
    message.volume = "";
    message.makerVolume = "";
    message.takerVolume = "";
    message.reward = "";
    message.rewardPercentage = "";
    message.qualifies = false;
    message.volumePercentage = "";
    message.createdAt = "";
    message.updatedAt = "";
    message.uptimePercentage = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string account_address */
        2:
          message.accountAddress = reader.string();
          break;
        case /* sint64 height */
        3:
          message.height = reader.sint64().toBigInt();
          break;
        case /* string block_time */
        4:
          message.blockTime = reader.string();
          break;
        case /* sint64 start_height */
        5:
          message.startHeight = reader.sint64().toBigInt();
          break;
        case /* string liquidity_score */
        6:
          message.liquidityScore = reader.string();
          break;
        case /* string liquidity_score_ponderated */
        7:
          message.liquidityScorePonderated = reader.string();
          break;
        case /* string uptime_score */
        8:
          message.uptimeScore = reader.string();
          break;
        case /* string uptime_score_ponderated */
        9:
          message.uptimeScorePonderated = reader.string();
          break;
        case /* string volume_score */
        10:
          message.volumeScore = reader.string();
          break;
        case /* string volume_score_ponderated */
        11:
          message.volumeScorePonderated = reader.string();
          break;
        case /* string total_score */
        12:
          message.totalScore = reader.string();
          break;
        case /* string volume */
        13:
          message.volume = reader.string();
          break;
        case /* string maker_volume */
        14:
          message.makerVolume = reader.string();
          break;
        case /* string taker_volume */
        15:
          message.takerVolume = reader.string();
          break;
        case /* string reward */
        16:
          message.reward = reader.string();
          break;
        case /* string reward_percentage */
        17:
          message.rewardPercentage = reader.string();
          break;
        case /* bool qualifies */
        18:
          message.qualifies = reader.bool();
          break;
        case /* string volume_percentage */
        19:
          message.volumePercentage = reader.string();
          break;
        case /* string created_at */
        20:
          message.createdAt = reader.string();
          break;
        case /* string updated_at */
        21:
          message.updatedAt = reader.string();
          break;
        case /* string uptime_percentage */
        22:
          message.uptimePercentage = reader.string();
          break;
        case /* optional string depth */
        23:
          message.depth = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.accountAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.accountAddress);
    if (message.height !== 0n)
      writer.tag(3, WireType.Varint).sint64(message.height);
    if (message.blockTime !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.blockTime);
    if (message.startHeight !== 0n)
      writer.tag(5, WireType.Varint).sint64(message.startHeight);
    if (message.liquidityScore !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.liquidityScore);
    if (message.liquidityScorePonderated !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.liquidityScorePonderated);
    if (message.uptimeScore !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.uptimeScore);
    if (message.uptimeScorePonderated !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.uptimeScorePonderated);
    if (message.volumeScore !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.volumeScore);
    if (message.volumeScorePonderated !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.volumeScorePonderated);
    if (message.totalScore !== "")
      writer.tag(12, WireType.LengthDelimited).string(message.totalScore);
    if (message.volume !== "")
      writer.tag(13, WireType.LengthDelimited).string(message.volume);
    if (message.makerVolume !== "")
      writer.tag(14, WireType.LengthDelimited).string(message.makerVolume);
    if (message.takerVolume !== "")
      writer.tag(15, WireType.LengthDelimited).string(message.takerVolume);
    if (message.reward !== "")
      writer.tag(16, WireType.LengthDelimited).string(message.reward);
    if (message.rewardPercentage !== "")
      writer.tag(17, WireType.LengthDelimited).string(message.rewardPercentage);
    if (message.qualifies !== false)
      writer.tag(18, WireType.Varint).bool(message.qualifies);
    if (message.volumePercentage !== "")
      writer.tag(19, WireType.LengthDelimited).string(message.volumePercentage);
    if (message.createdAt !== "")
      writer.tag(20, WireType.LengthDelimited).string(message.createdAt);
    if (message.updatedAt !== "")
      writer.tag(21, WireType.LengthDelimited).string(message.updatedAt);
    if (message.uptimePercentage !== "")
      writer.tag(22, WireType.LengthDelimited).string(message.uptimePercentage);
    if (message.depth !== void 0)
      writer.tag(23, WireType.LengthDelimited).string(message.depth);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EpochScore = new EpochScore$Type();
class GetEpochScoresHistoryRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetEpochScoresHistoryRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "page", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.accountAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string account_address */
        2:
          message.accountAddress = reader.string();
          break;
        case /* injective_dmm_v2.Pagination page */
        3:
          message.page = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.page);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.accountAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.accountAddress);
    if (message.page)
      Pagination.internalBinaryWrite(message.page, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetEpochScoresHistoryRequest = new GetEpochScoresHistoryRequest$Type();
class GetEpochScoresHistoryResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetEpochScoresHistoryResponse", [
      { no: 1, name: "scores", kind: "message", repeat: 2, T: () => EpochScore },
      { no: 2, name: "next", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.scores = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.EpochScore scores */
        1:
          message.scores.push(EpochScore.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* injective_dmm_v2.Pagination next */
        2:
          message.next = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.next);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.scores.length; i++)
      EpochScore.internalBinaryWrite(message.scores[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.next)
      Pagination.internalBinaryWrite(message.next, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetEpochScoresHistoryResponse = new GetEpochScoresHistoryResponse$Type();
class GetMarketRewardsRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetMarketRewardsRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetMarketRewardsRequest = new GetMarketRewardsRequest$Type();
class GetMarketRewardsResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetMarketRewardsResponse", [
      { no: 1, name: "rewards", kind: "message", repeat: 2, T: () => MarketReward }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.rewards = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.MarketReward rewards */
        1:
          message.rewards.push(MarketReward.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.rewards.length; i++)
      MarketReward.internalBinaryWrite(message.rewards[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetMarketRewardsResponse = new GetMarketRewardsResponse$Type();
class MarketReward$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.MarketReward", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "reward",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "reward_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "start_date",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "end_date",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "total_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "created_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "updated_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 12,
        name: "height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 13,
        name: "liquidity",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 14,
        name: "mini_epochs_reward",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.marketId = "";
    message.reward = "";
    message.rewardPercentage = "";
    message.startDate = "";
    message.endDate = "";
    message.totalScore = "";
    message.createdAt = "";
    message.updatedAt = "";
    message.height = 0n;
    message.liquidity = "";
    message.miniEpochsReward = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string market_id */
        2:
          message.marketId = reader.string();
          break;
        case /* string reward */
        3:
          message.reward = reader.string();
          break;
        case /* string reward_percentage */
        4:
          message.rewardPercentage = reader.string();
          break;
        case /* string start_date */
        5:
          message.startDate = reader.string();
          break;
        case /* string end_date */
        6:
          message.endDate = reader.string();
          break;
        case /* string total_score */
        9:
          message.totalScore = reader.string();
          break;
        case /* string created_at */
        10:
          message.createdAt = reader.string();
          break;
        case /* string updated_at */
        11:
          message.updatedAt = reader.string();
          break;
        case /* sint64 height */
        12:
          message.height = reader.sint64().toBigInt();
          break;
        case /* string liquidity */
        13:
          message.liquidity = reader.string();
          break;
        case /* string mini_epochs_reward */
        14:
          message.miniEpochsReward = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.marketId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    if (message.reward !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.reward);
    if (message.rewardPercentage !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.rewardPercentage);
    if (message.startDate !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.startDate);
    if (message.endDate !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.endDate);
    if (message.totalScore !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.totalScore);
    if (message.createdAt !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.createdAt);
    if (message.updatedAt !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.updatedAt);
    if (message.height !== 0n)
      writer.tag(12, WireType.Varint).sint64(message.height);
    if (message.liquidity !== "")
      writer.tag(13, WireType.LengthDelimited).string(message.liquidity);
    if (message.miniEpochsReward !== "")
      writer.tag(14, WireType.LengthDelimited).string(message.miniEpochsReward);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MarketReward = new MarketReward$Type();
class GetEligibleAddressesRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetEligibleAddressesRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "page", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* injective_dmm_v2.Pagination page */
        3:
          message.page = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.page);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.page)
      Pagination.internalBinaryWrite(message.page, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetEligibleAddressesRequest = new GetEligibleAddressesRequest$Type();
class GetEligibleAddressesResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetEligibleAddressesResponse", [
      { no: 1, name: "addresses", kind: "message", repeat: 2, T: () => EligibleAddress },
      { no: 2, name: "next", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.addresses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.EligibleAddress addresses */
        1:
          message.addresses.push(EligibleAddress.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* injective_dmm_v2.Pagination next */
        2:
          message.next = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.next);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.addresses.length; i++)
      EligibleAddress.internalBinaryWrite(message.addresses[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.next)
      Pagination.internalBinaryWrite(message.next, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetEligibleAddressesResponse = new GetEligibleAddressesResponse$Type();
class EligibleAddress$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.EligibleAddress", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "source",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "created_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "updated_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.accountAddress = "";
    message.height = 0n;
    message.source = "";
    message.createdAt = "";
    message.updatedAt = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string account_address */
        2:
          message.accountAddress = reader.string();
          break;
        case /* sint64 height */
        3:
          message.height = reader.sint64().toBigInt();
          break;
        case /* string source */
        4:
          message.source = reader.string();
          break;
        case /* string created_at */
        5:
          message.createdAt = reader.string();
          break;
        case /* string updated_at */
        6:
          message.updatedAt = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.accountAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.accountAddress);
    if (message.height !== 0n)
      writer.tag(3, WireType.Varint).sint64(message.height);
    if (message.source !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.source);
    if (message.createdAt !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.createdAt);
    if (message.updatedAt !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.updatedAt);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EligibleAddress = new EligibleAddress$Type();
class GetRewardsDistributionRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetRewardsDistributionRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "page", kind: "message", T: () => Pagination },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        opt: true,
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* injective_dmm_v2.Pagination page */
        2:
          message.page = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.page);
          break;
        case /* optional sint64 height */
        3:
          message.height = reader.sint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.page)
      Pagination.internalBinaryWrite(message.page, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.height !== void 0)
      writer.tag(3, WireType.Varint).sint64(message.height);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetRewardsDistributionRequest = new GetRewardsDistributionRequest$Type();
class GetRewardsDistributionResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetRewardsDistributionResponse", [
      { no: 1, name: "rewards", kind: "message", repeat: 2, T: () => RewardDistribution },
      { no: 2, name: "next", kind: "message", T: () => Pagination }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.rewards = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.RewardDistribution rewards */
        1:
          message.rewards.push(RewardDistribution.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* injective_dmm_v2.Pagination next */
        2:
          message.next = Pagination.internalBinaryRead(reader, reader.uint32(), options, message.next);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.rewards.length; i++)
      RewardDistribution.internalBinaryWrite(message.rewards[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.next)
      Pagination.internalBinaryWrite(message.next, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetRewardsDistributionResponse = new GetRewardsDistributionResponse$Type();
class RewardDistribution$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.RewardDistribution", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "start_height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "block_time",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "depth",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "reward",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "mini_epochs_reward",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "created_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "updated_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.accountAddress = "";
    message.height = 0n;
    message.startHeight = 0n;
    message.blockTime = "";
    message.depth = "";
    message.reward = "";
    message.createdAt = "";
    message.updatedAt = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string account_address */
        2:
          message.accountAddress = reader.string();
          break;
        case /* sint64 height */
        3:
          message.height = reader.sint64().toBigInt();
          break;
        case /* sint64 start_height */
        4:
          message.startHeight = reader.sint64().toBigInt();
          break;
        case /* string block_time */
        5:
          message.blockTime = reader.string();
          break;
        case /* string depth */
        6:
          message.depth = reader.string();
          break;
        case /* string reward */
        7:
          message.reward = reader.string();
          break;
        case /* optional string mini_epochs_reward */
        8:
          message.miniEpochsReward = reader.string();
          break;
        case /* string created_at */
        9:
          message.createdAt = reader.string();
          break;
        case /* string updated_at */
        10:
          message.updatedAt = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.accountAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.accountAddress);
    if (message.height !== 0n)
      writer.tag(3, WireType.Varint).sint64(message.height);
    if (message.startHeight !== 0n)
      writer.tag(4, WireType.Varint).sint64(message.startHeight);
    if (message.blockTime !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.blockTime);
    if (message.depth !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.depth);
    if (message.reward !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.reward);
    if (message.miniEpochsReward !== void 0)
      writer.tag(8, WireType.LengthDelimited).string(message.miniEpochsReward);
    if (message.createdAt !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.createdAt);
    if (message.updatedAt !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.updatedAt);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RewardDistribution = new RewardDistribution$Type();
class GetAccountVolumesRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetAccountVolumesRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.accountAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string account_address */
        2:
          message.accountAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.accountAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.accountAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetAccountVolumesRequest = new GetAccountVolumesRequest$Type();
class GetAccountVolumesResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetAccountVolumesResponse", [
      { no: 1, name: "volumes", kind: "message", repeat: 2, T: () => AccountVolume }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.volumes = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.AccountVolume volumes */
        1:
          message.volumes.push(AccountVolume.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.volumes.length; i++)
      AccountVolume.internalBinaryWrite(message.volumes[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetAccountVolumesResponse = new GetAccountVolumesResponse$Type();
class AccountVolume$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.AccountVolume", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "block_time",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "date",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "taker_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "maker_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "volume_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "maker_volume_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 12,
        name: "taker_volume_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 13,
        name: "created_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 14,
        name: "updated_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 15,
        name: "date_timestamp",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 16,
        name: "daily_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 17,
        name: "daily_maker_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 18,
        name: "daily_taker_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 19,
        name: "daily_volume_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 20,
        name: "daily_maker_volume_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 21,
        name: "daily_taker_volume_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 22,
        name: "daily_qualified",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.accountAddress = "";
    message.height = 0n;
    message.blockTime = "";
    message.date = "";
    message.volume = "";
    message.takerVolume = "";
    message.makerVolume = "";
    message.volumePercentage = "";
    message.makerVolumePercentage = "";
    message.takerVolumePercentage = "";
    message.createdAt = "";
    message.updatedAt = "";
    message.dateTimestamp = "";
    message.dailyVolume = "";
    message.dailyMakerVolume = "";
    message.dailyTakerVolume = "";
    message.dailyVolumePercentage = "";
    message.dailyMakerVolumePercentage = "";
    message.dailyTakerVolumePercentage = "";
    message.dailyQualified = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string account_address */
        2:
          message.accountAddress = reader.string();
          break;
        case /* sint64 height */
        3:
          message.height = reader.sint64().toBigInt();
          break;
        case /* string block_time */
        4:
          message.blockTime = reader.string();
          break;
        case /* string date */
        6:
          message.date = reader.string();
          break;
        case /* string volume */
        7:
          message.volume = reader.string();
          break;
        case /* string taker_volume */
        8:
          message.takerVolume = reader.string();
          break;
        case /* string maker_volume */
        9:
          message.makerVolume = reader.string();
          break;
        case /* string volume_percentage */
        10:
          message.volumePercentage = reader.string();
          break;
        case /* string maker_volume_percentage */
        11:
          message.makerVolumePercentage = reader.string();
          break;
        case /* string taker_volume_percentage */
        12:
          message.takerVolumePercentage = reader.string();
          break;
        case /* string created_at */
        13:
          message.createdAt = reader.string();
          break;
        case /* string updated_at */
        14:
          message.updatedAt = reader.string();
          break;
        case /* string date_timestamp */
        15:
          message.dateTimestamp = reader.string();
          break;
        case /* string daily_volume */
        16:
          message.dailyVolume = reader.string();
          break;
        case /* string daily_maker_volume */
        17:
          message.dailyMakerVolume = reader.string();
          break;
        case /* string daily_taker_volume */
        18:
          message.dailyTakerVolume = reader.string();
          break;
        case /* string daily_volume_percentage */
        19:
          message.dailyVolumePercentage = reader.string();
          break;
        case /* string daily_maker_volume_percentage */
        20:
          message.dailyMakerVolumePercentage = reader.string();
          break;
        case /* string daily_taker_volume_percentage */
        21:
          message.dailyTakerVolumePercentage = reader.string();
          break;
        case /* bool daily_qualified */
        22:
          message.dailyQualified = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.accountAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.accountAddress);
    if (message.height !== 0n)
      writer.tag(3, WireType.Varint).sint64(message.height);
    if (message.blockTime !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.blockTime);
    if (message.date !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.date);
    if (message.volume !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.volume);
    if (message.takerVolume !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.takerVolume);
    if (message.makerVolume !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.makerVolume);
    if (message.volumePercentage !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.volumePercentage);
    if (message.makerVolumePercentage !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.makerVolumePercentage);
    if (message.takerVolumePercentage !== "")
      writer.tag(12, WireType.LengthDelimited).string(message.takerVolumePercentage);
    if (message.createdAt !== "")
      writer.tag(13, WireType.LengthDelimited).string(message.createdAt);
    if (message.updatedAt !== "")
      writer.tag(14, WireType.LengthDelimited).string(message.updatedAt);
    if (message.dateTimestamp !== "")
      writer.tag(15, WireType.LengthDelimited).string(message.dateTimestamp);
    if (message.dailyVolume !== "")
      writer.tag(16, WireType.LengthDelimited).string(message.dailyVolume);
    if (message.dailyMakerVolume !== "")
      writer.tag(17, WireType.LengthDelimited).string(message.dailyMakerVolume);
    if (message.dailyTakerVolume !== "")
      writer.tag(18, WireType.LengthDelimited).string(message.dailyTakerVolume);
    if (message.dailyVolumePercentage !== "")
      writer.tag(19, WireType.LengthDelimited).string(message.dailyVolumePercentage);
    if (message.dailyMakerVolumePercentage !== "")
      writer.tag(20, WireType.LengthDelimited).string(message.dailyMakerVolumePercentage);
    if (message.dailyTakerVolumePercentage !== "")
      writer.tag(21, WireType.LengthDelimited).string(message.dailyTakerVolumePercentage);
    if (message.dailyQualified !== false)
      writer.tag(22, WireType.Varint).bool(message.dailyQualified);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AccountVolume = new AccountVolume$Type();
class GetRewardsEligibilityRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetRewardsEligibilityRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.accountAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string account_address */
        2:
          message.accountAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.accountAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.accountAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetRewardsEligibilityRequest = new GetRewardsEligibilityRequest$Type();
class GetRewardsEligibilityResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetRewardsEligibilityResponse", [
      { no: 1, name: "volumes", kind: "message", repeat: 2, T: () => AccountVolume },
      {
        no: 2,
        name: "current_maker_volume_percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "average_daily_maker_volume_percentage",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "eligible_for_next_epoch",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 5,
        name: "estimated_reward",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "eligible_for_current_epoch",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 7,
        name: "updated_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.volumes = [];
    message.currentMakerVolumePercentage = "";
    message.eligibleForNextEpoch = false;
    message.estimatedReward = "";
    message.eligibleForCurrentEpoch = false;
    message.updatedAt = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.AccountVolume volumes */
        1:
          message.volumes.push(AccountVolume.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string current_maker_volume_percentage */
        2:
          message.currentMakerVolumePercentage = reader.string();
          break;
        case /* optional string average_daily_maker_volume_percentage */
        3:
          message.averageDailyMakerVolumePercentage = reader.string();
          break;
        case /* bool eligible_for_next_epoch */
        4:
          message.eligibleForNextEpoch = reader.bool();
          break;
        case /* string estimated_reward */
        5:
          message.estimatedReward = reader.string();
          break;
        case /* bool eligible_for_current_epoch */
        6:
          message.eligibleForCurrentEpoch = reader.bool();
          break;
        case /* string updated_at */
        7:
          message.updatedAt = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.volumes.length; i++)
      AccountVolume.internalBinaryWrite(message.volumes[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.currentMakerVolumePercentage !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.currentMakerVolumePercentage);
    if (message.averageDailyMakerVolumePercentage !== void 0)
      writer.tag(3, WireType.LengthDelimited).string(message.averageDailyMakerVolumePercentage);
    if (message.eligibleForNextEpoch !== false)
      writer.tag(4, WireType.Varint).bool(message.eligibleForNextEpoch);
    if (message.estimatedReward !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.estimatedReward);
    if (message.eligibleForCurrentEpoch !== false)
      writer.tag(6, WireType.Varint).bool(message.eligibleForCurrentEpoch);
    if (message.updatedAt !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.updatedAt);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetRewardsEligibilityResponse = new GetRewardsEligibilityResponse$Type();
class GetHealthStatusRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetHealthStatusRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetHealthStatusRequest = new GetHealthStatusRequest$Type();
class GetHealthStatusResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetHealthStatusResponse", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "last_snapshot_block_time",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "last_snapshot_block_height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "updated_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "status",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    message.lastSnapshotBlockTime = "";
    message.lastSnapshotBlockHeight = 0n;
    message.updatedAt = "";
    message.status = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* string last_snapshot_block_time */
        2:
          message.lastSnapshotBlockTime = reader.string();
          break;
        case /* sint64 last_snapshot_block_height */
        3:
          message.lastSnapshotBlockHeight = reader.sint64().toBigInt();
          break;
        case /* string updated_at */
        4:
          message.updatedAt = reader.string();
          break;
        case /* string status */
        5:
          message.status = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.lastSnapshotBlockTime !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.lastSnapshotBlockTime);
    if (message.lastSnapshotBlockHeight !== 0n)
      writer.tag(3, WireType.Varint).sint64(message.lastSnapshotBlockHeight);
    if (message.updatedAt !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.updatedAt);
    if (message.status !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.status);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetHealthStatusResponse = new GetHealthStatusResponse$Type();
class GetMarketRewardsRangeRequest$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetMarketRewardsRangeRequest", [
      {
        no: 1,
        name: "epoch_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epochId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string epoch_id */
        1:
          message.epochId = reader.string();
          break;
        case /* optional string market_id */
        2:
          message.marketId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epochId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.epochId);
    if (message.marketId !== void 0)
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetMarketRewardsRangeRequest = new GetMarketRewardsRangeRequest$Type();
class GetMarketRewardsRangeResponse$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.GetMarketRewardsRangeResponse", [
      { no: 1, name: "ranges", kind: "message", repeat: 2, T: () => MarketRewardRange }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.ranges = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_dmm_v2.MarketRewardRange ranges */
        1:
          message.ranges.push(MarketRewardRange.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.ranges.length; i++)
      MarketRewardRange.internalBinaryWrite(message.ranges[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetMarketRewardsRangeResponse = new GetMarketRewardsRangeResponse$Type();
class MarketRewardRange$Type extends MessageType {
  constructor() {
    super("injective_dmm_v2.MarketRewardRange", [
      {
        no: 1,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "max",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "min",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.marketId = "";
    message.max = "";
    message.min = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string market_id */
        1:
          message.marketId = reader.string();
          break;
        case /* string max */
        2:
          message.max = reader.string();
          break;
        case /* string min */
        3:
          message.min = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.marketId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.marketId);
    if (message.max !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.max);
    if (message.min !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.min);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MarketRewardRange = new MarketRewardRange$Type();
const InjectiveDmmV2 = new ServiceType("injective_dmm_v2.InjectiveDmmV2", [
  { name: "GetEpochs", options: {}, I: GetEpochsRequest, O: GetEpochsResponse },
  { name: "GetTotalScores", options: {}, I: GetTotalScoresRequest, O: GetTotalScoresResponse },
  { name: "GetTotalScoresHistory", options: {}, I: GetTotalScoresHistoryRequest, O: GetTotalScoresHistoryResponse },
  { name: "GetEpochScores", options: {}, I: GetEpochScoresRequest, O: GetEpochScoresResponse },
  { name: "GetEpochScoresHistory", options: {}, I: GetEpochScoresHistoryRequest, O: GetEpochScoresHistoryResponse },
  { name: "GetMarketRewards", options: {}, I: GetMarketRewardsRequest, O: GetMarketRewardsResponse },
  { name: "GetEligibleAddresses", options: {}, I: GetEligibleAddressesRequest, O: GetEligibleAddressesResponse },
  { name: "GetRewardsDistribution", options: {}, I: GetRewardsDistributionRequest, O: GetRewardsDistributionResponse },
  { name: "GetAccountVolumes", options: {}, I: GetAccountVolumesRequest, O: GetAccountVolumesResponse },
  { name: "GetRewardsEligibility", options: {}, I: GetRewardsEligibilityRequest, O: GetRewardsEligibilityResponse },
  { name: "GetHealthStatus", options: {}, I: GetHealthStatusRequest, O: GetHealthStatusResponse },
  { name: "GetMarketRewardsRange", options: {}, I: GetMarketRewardsRangeRequest, O: GetMarketRewardsRangeResponse }
]);
export {
  AccountVolume,
  EligibleAddress,
  EpochConfigV2,
  EpochScore,
  EpochV2,
  GetAccountVolumesRequest,
  GetAccountVolumesResponse,
  GetEligibleAddressesRequest,
  GetEligibleAddressesResponse,
  GetEpochScoresHistoryRequest,
  GetEpochScoresHistoryResponse,
  GetEpochScoresRequest,
  GetEpochScoresResponse,
  GetEpochsRequest,
  GetEpochsResponse,
  GetHealthStatusRequest,
  GetHealthStatusResponse,
  GetMarketRewardsRangeRequest,
  GetMarketRewardsRangeResponse,
  GetMarketRewardsRequest,
  GetMarketRewardsResponse,
  GetRewardsDistributionRequest,
  GetRewardsDistributionResponse,
  GetRewardsEligibilityRequest,
  GetRewardsEligibilityResponse,
  GetTotalScoresHistoryRequest,
  GetTotalScoresHistoryResponse,
  GetTotalScoresRequest,
  GetTotalScoresResponse,
  InjectiveDmmV2,
  Market,
  MarketReward,
  MarketRewardRange,
  Pagination,
  RewardDistribution,
  TotalScore
};
