// @injectivelabs/olp-proto-ts-v2/generated/goagen_volatility_injective_volatility_rpc_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class GetVolatilityIndexRequest$Type extends MessageType {
  constructor() {
    super("injective_volatility_rpc.GetVolatilityIndexRequest", [
      {
        no: 1,
        name: "symbol",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "time_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.symbol = "";
    message.timeAt = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string symbol */
        1:
          message.symbol = reader.string();
          break;
        case /* string time_at */
        2:
          message.timeAt = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.symbol !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.symbol);
    if (message.timeAt !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.timeAt);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetVolatilityIndexRequest = new GetVolatilityIndexRequest$Type();
class GetVolatilityIndexResponse$Type extends MessageType {
  constructor() {
    super("injective_volatility_rpc.GetVolatilityIndexResponse", [
      {
        no: 1,
        name: "symbol",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "time_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "volatility_index",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "parameters", kind: "message", T: () => VolatilityIndexParameters }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.symbol = "";
    message.timeAt = "";
    message.volatilityIndex = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string symbol */
        1:
          message.symbol = reader.string();
          break;
        case /* string time_at */
        2:
          message.timeAt = reader.string();
          break;
        case /* string volatility_index */
        3:
          message.volatilityIndex = reader.string();
          break;
        case /* injective_volatility_rpc.VolatilityIndexParameters parameters */
        4:
          message.parameters = VolatilityIndexParameters.internalBinaryRead(reader, reader.uint32(), options, message.parameters);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.symbol !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.symbol);
    if (message.timeAt !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.timeAt);
    if (message.volatilityIndex !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.volatilityIndex);
    if (message.parameters)
      VolatilityIndexParameters.internalBinaryWrite(message.parameters, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetVolatilityIndexResponse = new GetVolatilityIndexResponse$Type();
class VolatilityIndexParameters$Type extends MessageType {
  constructor() {
    super("injective_volatility_rpc.VolatilityIndexParameters", [
      {
        no: 1,
        name: "alpha",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "theta_max",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "realized_volatility",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "price",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "average_price",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.alpha = "";
    message.thetaMax = "";
    message.realizedVolatility = "";
    message.price = "";
    message.averagePrice = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string alpha */
        1:
          message.alpha = reader.string();
          break;
        case /* string theta_max */
        2:
          message.thetaMax = reader.string();
          break;
        case /* string realized_volatility */
        3:
          message.realizedVolatility = reader.string();
          break;
        case /* string price */
        4:
          message.price = reader.string();
          break;
        case /* string average_price */
        5:
          message.averagePrice = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.alpha !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.alpha);
    if (message.thetaMax !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.thetaMax);
    if (message.realizedVolatility !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.realizedVolatility);
    if (message.price !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.price);
    if (message.averagePrice !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.averagePrice);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const VolatilityIndexParameters = new VolatilityIndexParameters$Type();
const InjectiveVolatilityRPC = new ServiceType("injective_volatility_rpc.InjectiveVolatilityRPC", [
  { name: "GetVolatilityIndex", options: {}, I: GetVolatilityIndexRequest, O: GetVolatilityIndexResponse }
]);
export {
  GetVolatilityIndexRequest,
  GetVolatilityIndexResponse,
  InjectiveVolatilityRPC,
  VolatilityIndexParameters
};
