// @injectivelabs/abacus-proto-ts-v2/esm/generated/points_svc_pb.client.js
import { PointsSvc } from "./points_svc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class PointsSvcClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = PointsSvc.typeName;
    this.methods = PointsSvc.methods;
    this.options = PointsSvc.options;
  }
  /**
   * Gets user's total points and rank by account address.
   *
   * @generated from protobuf rpc: PointsLatestForAccount
   */
  pointsLatestForAccount(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets user's daily incurred points and volume by account address.
   *
   * @generated from protobuf rpc: PointsStatsDailyForAccount
   */
  pointsStatsDailyForAccount(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets user's weekly incurred points and volume by account address.
   *
   * @generated from protobuf rpc: PointsStatsWeeklyForAccount
   */
  pointsStatsWeeklyForAccount(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to get the true leaderboard with all participants and
   * ranks.
   *
   * @generated from protobuf rpc: PointsLeaderboard
   */
  pointsLeaderboard(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to simulate points allocation for a trade without
   * persisting to database.
   *
   * @generated from protobuf rpc: PointsSimulateAllocation
   */
  pointsSimulateAllocation(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to get the current league configuration.
   *
   * @generated from protobuf rpc: PointsGetLeagueConfig
   */
  pointsGetLeagueConfig(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to update the league configuration.
   *
   * @generated from protobuf rpc: PointsSetLeagueConfig
   */
  pointsSetLeagueConfig(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to get the current points emission configuration.
   *
   * @generated from protobuf rpc: PointsGetEmissionConfig
   */
  pointsGetEmissionConfig(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to update the points emission configuration.
   *
   * @generated from protobuf rpc: PointsSetEmissionConfig
   */
  pointsSetEmissionConfig(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to list all banned account configurations.
   *
   * @generated from protobuf rpc: PointsListBanConfigs
   */
  pointsListBanConfigs(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to create a new ban configuration for an account.
   *
   * @generated from protobuf rpc: PointsCreateBanConfig
   */
  pointsCreateBanConfig(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to delete a ban configuration.
   *
   * @generated from protobuf rpc: PointsDeleteBanConfig
   */
  pointsDeleteBanConfig(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to list all points multiplier configurations.
   *
   * @generated from protobuf rpc: PointsListMultiplierConfigs
   */
  pointsListMultiplierConfigs(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to create a new points multiplier configuration.
   *
   * @generated from protobuf rpc: PointsCreateMultiplierConfig
   */
  pointsCreateMultiplierConfig(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to delete a points multiplier configuration.
   *
   * @generated from protobuf rpc: PointsDeleteMultiplierConfig
   */
  pointsDeleteMultiplierConfig(input, options) {
    const method = this.methods[14], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to list all points corrections.
   *
   * @generated from protobuf rpc: PointsListAdminCorrections
   */
  pointsListAdminCorrections(input, options) {
    const method = this.methods[15], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to create a new points correction.
   *
   * @generated from protobuf rpc: PointsCreateAdminCorrection
   */
  pointsCreateAdminCorrection(input, options) {
    const method = this.methods[16], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to create a new points corrections (sharing same type,
   * label, effective date).
   *
   * @generated from protobuf rpc: PointsCreateAdminCorrectionsBatch
   */
  pointsCreateAdminCorrectionsBatch(input, options) {
    const method = this.methods[17], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to delete a points correction.
   *
   * @generated from protobuf rpc: PointsDeleteAdminCorrection
   */
  pointsDeleteAdminCorrection(input, options) {
    const method = this.methods[18], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to reprocess trades after a given date.
   *
   * @generated from protobuf rpc: PointsReprocessTradesAfterDate
   */
  pointsReprocessTradesAfterDate(input, options) {
    const method = this.methods[19], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to capture current points totals as snapshot for S1 and
   * calculate bonus.
   *
   * @generated from protobuf rpc: PointsCaptureSeasonOneSnapshotAndBonus
   */
  pointsCaptureSeasonOneSnapshotAndBonus(input, options) {
    const method = this.methods[20], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Admin-only endpoint to recompute daily points snapshots within (last 6
   * months).
   *
   * @generated from protobuf rpc: PointsRecomputeDailyPointsSnapshots
   */
  pointsRecomputeDailyPointsSnapshots(input, options) {
    const method = this.methods[21], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  PointsSvcClient
};
