// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/lightclients/wasm/v1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * StoreCode defines a rpc handler method for MsgStoreCode.
   *
   * @generated from protobuf rpc: StoreCode
   */
  storeCode(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * RemoveChecksum defines a rpc handler method for MsgRemoveChecksum.
   *
   * @generated from protobuf rpc: RemoveChecksum
   */
  removeChecksum(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * MigrateContract defines a rpc handler method for MsgMigrateContract.
   *
   * @generated from protobuf rpc: MigrateContract
   */
  migrateContract(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
