// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/connection/v1/connection_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { MerklePrefix } from "../../commitment/v1/commitment_pb.js";
var State = /* @__PURE__ */ ((State2) => {
  State2[State2["UNINITIALIZED_UNSPECIFIED"] = 0] = "UNINITIALIZED_UNSPECIFIED";
  State2[State2["INIT"] = 1] = "INIT";
  State2[State2["TRYOPEN"] = 2] = "TRYOPEN";
  State2[State2["OPEN"] = 3] = "OPEN";
  return State2;
})(State || {});
class ConnectionEnd$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.ConnectionEnd", [
      {
        no: 1,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "versions", kind: "message", repeat: 2, T: () => Version },
      { no: 3, name: "state", kind: "enum", T: () => ["ibc.core.connection.v1.State", State, "STATE_"] },
      { no: 4, name: "counterparty", kind: "message", T: () => Counterparty, options: { "gogoproto.nullable": false } },
      {
        no: 5,
        name: "delay_period",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clientId = "";
    message.versions = [];
    message.state = 0;
    message.delayPeriod = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string client_id */
        1:
          message.clientId = reader.string();
          break;
        case /* repeated ibc.core.connection.v1.Version versions */
        2:
          message.versions.push(Version.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* ibc.core.connection.v1.State state */
        3:
          message.state = reader.int32();
          break;
        case /* ibc.core.connection.v1.Counterparty counterparty */
        4:
          message.counterparty = Counterparty.internalBinaryRead(reader, reader.uint32(), options, message.counterparty);
          break;
        case /* uint64 delay_period */
        5:
          message.delayPeriod = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.clientId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.clientId);
    for (let i = 0; i < message.versions.length; i++)
      Version.internalBinaryWrite(message.versions[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.state !== 0)
      writer.tag(3, WireType.Varint).int32(message.state);
    if (message.counterparty)
      Counterparty.internalBinaryWrite(message.counterparty, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.delayPeriod !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.delayPeriod);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConnectionEnd = new ConnectionEnd$Type();
class IdentifiedConnection$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.IdentifiedConnection", [
      {
        no: 1,
        name: "id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "versions", kind: "message", repeat: 2, T: () => Version },
      { no: 4, name: "state", kind: "enum", T: () => ["ibc.core.connection.v1.State", State, "STATE_"] },
      { no: 5, name: "counterparty", kind: "message", T: () => Counterparty, options: { "gogoproto.nullable": false } },
      {
        no: 6,
        name: "delay_period",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.id = "";
    message.clientId = "";
    message.versions = [];
    message.state = 0;
    message.delayPeriod = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string id */
        1:
          message.id = reader.string();
          break;
        case /* string client_id */
        2:
          message.clientId = reader.string();
          break;
        case /* repeated ibc.core.connection.v1.Version versions */
        3:
          message.versions.push(Version.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* ibc.core.connection.v1.State state */
        4:
          message.state = reader.int32();
          break;
        case /* ibc.core.connection.v1.Counterparty counterparty */
        5:
          message.counterparty = Counterparty.internalBinaryRead(reader, reader.uint32(), options, message.counterparty);
          break;
        case /* uint64 delay_period */
        6:
          message.delayPeriod = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.id !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.id);
    if (message.clientId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.clientId);
    for (let i = 0; i < message.versions.length; i++)
      Version.internalBinaryWrite(message.versions[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.state !== 0)
      writer.tag(4, WireType.Varint).int32(message.state);
    if (message.counterparty)
      Counterparty.internalBinaryWrite(message.counterparty, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.delayPeriod !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.delayPeriod);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const IdentifiedConnection = new IdentifiedConnection$Type();
class Counterparty$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.Counterparty", [
      {
        no: 1,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "prefix", kind: "message", T: () => MerklePrefix, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clientId = "";
    message.connectionId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string client_id */
        1:
          message.clientId = reader.string();
          break;
        case /* string connection_id */
        2:
          message.connectionId = reader.string();
          break;
        case /* ibc.core.commitment.v1.MerklePrefix prefix */
        3:
          message.prefix = MerklePrefix.internalBinaryRead(reader, reader.uint32(), options, message.prefix);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.clientId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.clientId);
    if (message.connectionId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.connectionId);
    if (message.prefix)
      MerklePrefix.internalBinaryWrite(message.prefix, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Counterparty = new Counterparty$Type();
class ClientPaths$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.ClientPaths", [
      {
        no: 1,
        name: "paths",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.paths = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string paths */
        1:
          message.paths.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.paths.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.paths[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ClientPaths = new ClientPaths$Type();
class ConnectionPaths$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.ConnectionPaths", [
      {
        no: 1,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "paths",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clientId = "";
    message.paths = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string client_id */
        1:
          message.clientId = reader.string();
          break;
        case /* repeated string paths */
        2:
          message.paths.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.clientId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.clientId);
    for (let i = 0; i < message.paths.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.paths[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConnectionPaths = new ConnectionPaths$Type();
class Version$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.Version", [
      {
        no: 1,
        name: "identifier",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "features",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.identifier = "";
    message.features = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string identifier */
        1:
          message.identifier = reader.string();
          break;
        case /* repeated string features */
        2:
          message.features.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.identifier !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.identifier);
    for (let i = 0; i < message.features.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.features[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Version = new Version$Type();
class Params$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.Params", [
      {
        no: 1,
        name: "max_expected_time_per_block",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.maxExpectedTimePerBlock = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 max_expected_time_per_block */
        1:
          message.maxExpectedTimePerBlock = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.maxExpectedTimePerBlock !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.maxExpectedTimePerBlock);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
export {
  ClientPaths,
  ConnectionEnd,
  ConnectionPaths,
  Counterparty,
  IdentifiedConnection,
  Params,
  State,
  Version
};
