// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/connection/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Connection queries an IBC connection end.
   *
   * @generated from protobuf rpc: Connection
   */
  connection(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Connections queries all the IBC connections of a chain.
   *
   * @generated from protobuf rpc: Connections
   */
  connections(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ClientConnections queries the connection paths associated with a client
   * state.
   *
   * @generated from protobuf rpc: ClientConnections
   */
  clientConnections(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ConnectionClientState queries the client state associated with the
   * connection.
   *
   * @generated from protobuf rpc: ConnectionClientState
   */
  connectionClientState(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ConnectionConsensusState queries the consensus state associated with the
   * connection.
   *
   * @generated from protobuf rpc: ConnectionConsensusState
   */
  connectionConsensusState(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ConnectionParams queries all parameters of the ibc connection submodule.
   *
   * @generated from protobuf rpc: ConnectionParams
   */
  connectionParams(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
