// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/connection/v1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * ConnectionOpenInit defines a rpc handler method for MsgConnectionOpenInit.
   *
   * @generated from protobuf rpc: ConnectionOpenInit
   */
  connectionOpenInit(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ConnectionOpenTry defines a rpc handler method for MsgConnectionOpenTry.
   *
   * @generated from protobuf rpc: ConnectionOpenTry
   */
  connectionOpenTry(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ConnectionOpenAck defines a rpc handler method for MsgConnectionOpenAck.
   *
   * @generated from protobuf rpc: ConnectionOpenAck
   */
  connectionOpenAck(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ConnectionOpenConfirm defines a rpc handler method for
   * MsgConnectionOpenConfirm.
   *
   * @generated from protobuf rpc: ConnectionOpenConfirm
   */
  connectionOpenConfirm(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateConnectionParams defines a rpc handler method for
   * MsgUpdateParams.
   *
   * @generated from protobuf rpc: UpdateConnectionParams
   */
  updateConnectionParams(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
