// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/connection/v1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./connection_pb.js";
import { Height } from "../../client/v1/client_pb.js";
import { Any } from "../../../../google/protobuf/any_pb.js";
import { Version } from "./connection_pb.js";
import { Counterparty } from "./connection_pb.js";
class MsgConnectionOpenInit$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgConnectionOpenInit", [
      {
        no: 1,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "counterparty", kind: "message", T: () => Counterparty, options: { "gogoproto.nullable": false } },
      { no: 3, name: "version", kind: "message", T: () => Version },
      {
        no: 4,
        name: "delay_period",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clientId = "";
    message.delayPeriod = 0n;
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string client_id */
        1:
          message.clientId = reader.string();
          break;
        case /* ibc.core.connection.v1.Counterparty counterparty */
        2:
          message.counterparty = Counterparty.internalBinaryRead(reader, reader.uint32(), options, message.counterparty);
          break;
        case /* ibc.core.connection.v1.Version version */
        3:
          message.version = Version.internalBinaryRead(reader, reader.uint32(), options, message.version);
          break;
        case /* uint64 delay_period */
        4:
          message.delayPeriod = reader.uint64().toBigInt();
          break;
        case /* string signer */
        5:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.clientId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.clientId);
    if (message.counterparty)
      Counterparty.internalBinaryWrite(message.counterparty, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.version)
      Version.internalBinaryWrite(message.version, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.delayPeriod !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.delayPeriod);
    if (message.signer !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgConnectionOpenInit = new MsgConnectionOpenInit$Type();
class MsgConnectionOpenInitResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgConnectionOpenInitResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgConnectionOpenInitResponse = new MsgConnectionOpenInitResponse$Type();
class MsgConnectionOpenTry$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgConnectionOpenTry", [
      {
        no: 1,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "previous_connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "client_state", kind: "message", T: () => Any },
      { no: 4, name: "counterparty", kind: "message", T: () => Counterparty, options: { "gogoproto.nullable": false } },
      {
        no: 5,
        name: "delay_period",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 6, name: "counterparty_versions", kind: "message", repeat: 2, T: () => Version },
      { no: 7, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 8,
        name: "proof_init",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 9,
        name: "proof_client",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 10,
        name: "proof_consensus",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 11, name: "consensus_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 12,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 13,
        name: "host_consensus_state_proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clientId = "";
    message.previousConnectionId = "";
    message.delayPeriod = 0n;
    message.counterpartyVersions = [];
    message.proofInit = new Uint8Array(0);
    message.proofClient = new Uint8Array(0);
    message.proofConsensus = new Uint8Array(0);
    message.signer = "";
    message.hostConsensusStateProof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string client_id */
        1:
          message.clientId = reader.string();
          break;
        case /* string previous_connection_id = 2 [deprecated = true] */
        2:
          message.previousConnectionId = reader.string();
          break;
        case /* google.protobuf.Any client_state = 3 [deprecated = true] */
        3:
          message.clientState = Any.internalBinaryRead(reader, reader.uint32(), options, message.clientState);
          break;
        case /* ibc.core.connection.v1.Counterparty counterparty */
        4:
          message.counterparty = Counterparty.internalBinaryRead(reader, reader.uint32(), options, message.counterparty);
          break;
        case /* uint64 delay_period */
        5:
          message.delayPeriod = reader.uint64().toBigInt();
          break;
        case /* repeated ibc.core.connection.v1.Version counterparty_versions */
        6:
          message.counterpartyVersions.push(Version.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* ibc.core.client.v1.Height proof_height */
        7:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* bytes proof_init */
        8:
          message.proofInit = reader.bytes();
          break;
        case /* bytes proof_client = 9 [deprecated = true] */
        9:
          message.proofClient = reader.bytes();
          break;
        case /* bytes proof_consensus = 10 [deprecated = true] */
        10:
          message.proofConsensus = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height consensus_height = 11 [deprecated = true] */
        11:
          message.consensusHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.consensusHeight);
          break;
        case /* string signer */
        12:
          message.signer = reader.string();
          break;
        case /* bytes host_consensus_state_proof = 13 [deprecated = true] */
        13:
          message.hostConsensusStateProof = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.clientId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.clientId);
    if (message.previousConnectionId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.previousConnectionId);
    if (message.clientState)
      Any.internalBinaryWrite(message.clientState, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.counterparty)
      Counterparty.internalBinaryWrite(message.counterparty, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.delayPeriod !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.delayPeriod);
    for (let i = 0; i < message.counterpartyVersions.length; i++)
      Version.internalBinaryWrite(message.counterpartyVersions[i], writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.proofInit.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.proofInit);
    if (message.proofClient.length)
      writer.tag(9, WireType.LengthDelimited).bytes(message.proofClient);
    if (message.proofConsensus.length)
      writer.tag(10, WireType.LengthDelimited).bytes(message.proofConsensus);
    if (message.consensusHeight)
      Height.internalBinaryWrite(message.consensusHeight, writer.tag(11, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(12, WireType.LengthDelimited).string(message.signer);
    if (message.hostConsensusStateProof.length)
      writer.tag(13, WireType.LengthDelimited).bytes(message.hostConsensusStateProof);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgConnectionOpenTry = new MsgConnectionOpenTry$Type();
class MsgConnectionOpenTryResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgConnectionOpenTryResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgConnectionOpenTryResponse = new MsgConnectionOpenTryResponse$Type();
class MsgConnectionOpenAck$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgConnectionOpenAck", [
      {
        no: 1,
        name: "connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "counterparty_connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "version", kind: "message", T: () => Version },
      { no: 4, name: "client_state", kind: "message", T: () => Any },
      { no: 5, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 6,
        name: "proof_try",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 7,
        name: "proof_client",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 8,
        name: "proof_consensus",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 9, name: "consensus_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 10,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "host_consensus_state_proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.connectionId = "";
    message.counterpartyConnectionId = "";
    message.proofTry = new Uint8Array(0);
    message.proofClient = new Uint8Array(0);
    message.proofConsensus = new Uint8Array(0);
    message.signer = "";
    message.hostConsensusStateProof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string connection_id */
        1:
          message.connectionId = reader.string();
          break;
        case /* string counterparty_connection_id */
        2:
          message.counterpartyConnectionId = reader.string();
          break;
        case /* ibc.core.connection.v1.Version version */
        3:
          message.version = Version.internalBinaryRead(reader, reader.uint32(), options, message.version);
          break;
        case /* google.protobuf.Any client_state = 4 [deprecated = true] */
        4:
          message.clientState = Any.internalBinaryRead(reader, reader.uint32(), options, message.clientState);
          break;
        case /* ibc.core.client.v1.Height proof_height */
        5:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* bytes proof_try */
        6:
          message.proofTry = reader.bytes();
          break;
        case /* bytes proof_client = 7 [deprecated = true] */
        7:
          message.proofClient = reader.bytes();
          break;
        case /* bytes proof_consensus = 8 [deprecated = true] */
        8:
          message.proofConsensus = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height consensus_height = 9 [deprecated = true] */
        9:
          message.consensusHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.consensusHeight);
          break;
        case /* string signer */
        10:
          message.signer = reader.string();
          break;
        case /* bytes host_consensus_state_proof = 11 [deprecated = true] */
        11:
          message.hostConsensusStateProof = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.connectionId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.connectionId);
    if (message.counterpartyConnectionId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.counterpartyConnectionId);
    if (message.version)
      Version.internalBinaryWrite(message.version, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.clientState)
      Any.internalBinaryWrite(message.clientState, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.proofTry.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.proofTry);
    if (message.proofClient.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.proofClient);
    if (message.proofConsensus.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.proofConsensus);
    if (message.consensusHeight)
      Height.internalBinaryWrite(message.consensusHeight, writer.tag(9, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.signer);
    if (message.hostConsensusStateProof.length)
      writer.tag(11, WireType.LengthDelimited).bytes(message.hostConsensusStateProof);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgConnectionOpenAck = new MsgConnectionOpenAck$Type();
class MsgConnectionOpenAckResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgConnectionOpenAckResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgConnectionOpenAckResponse = new MsgConnectionOpenAckResponse$Type();
class MsgConnectionOpenConfirm$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgConnectionOpenConfirm", [
      {
        no: 1,
        name: "connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "proof_ack",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.connectionId = "";
    message.proofAck = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string connection_id */
        1:
          message.connectionId = reader.string();
          break;
        case /* bytes proof_ack */
        2:
          message.proofAck = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        4:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.connectionId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.connectionId);
    if (message.proofAck.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proofAck);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgConnectionOpenConfirm = new MsgConnectionOpenConfirm$Type();
class MsgConnectionOpenConfirmResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgConnectionOpenConfirmResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgConnectionOpenConfirmResponse = new MsgConnectionOpenConfirmResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgUpdateParams", [
      {
        no: 1,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string signer */
        1:
          message.signer = reader.string();
          break;
        case /* ibc.core.connection.v1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signer !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.signer);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.connection.v1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
const Msg = new ServiceType("ibc.core.connection.v1.Msg", [
  { name: "ConnectionOpenInit", options: {}, I: MsgConnectionOpenInit, O: MsgConnectionOpenInitResponse },
  { name: "ConnectionOpenTry", options: {}, I: MsgConnectionOpenTry, O: MsgConnectionOpenTryResponse },
  { name: "ConnectionOpenAck", options: {}, I: MsgConnectionOpenAck, O: MsgConnectionOpenAckResponse },
  { name: "ConnectionOpenConfirm", options: {}, I: MsgConnectionOpenConfirm, O: MsgConnectionOpenConfirmResponse },
  { name: "UpdateConnectionParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgConnectionOpenAck,
  MsgConnectionOpenAckResponse,
  MsgConnectionOpenConfirm,
  MsgConnectionOpenConfirmResponse,
  MsgConnectionOpenInit,
  MsgConnectionOpenInitResponse,
  MsgConnectionOpenTry,
  MsgConnectionOpenTryResponse,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
