// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/base/tendermint/v1beta1/query_pb.client.js
import { Service } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class ServiceClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Service.typeName;
    this.methods = Service.methods;
    this.options = Service.options;
  }
  /**
   * GetNodeInfo queries the current node info.
   *
   * @generated from protobuf rpc: GetNodeInfo
   */
  getNodeInfo(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetSyncing queries node syncing.
   *
   * @generated from protobuf rpc: GetSyncing
   */
  getSyncing(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetLatestBlock returns the latest block.
   *
   * @generated from protobuf rpc: GetLatestBlock
   */
  getLatestBlock(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetBlockByHeight queries block for given height.
   *
   * @generated from protobuf rpc: GetBlockByHeight
   */
  getBlockByHeight(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetLatestValidatorSet queries latest validator-set.
   *
   * @generated from protobuf rpc: GetLatestValidatorSet
   */
  getLatestValidatorSet(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetValidatorSetByHeight queries validator-set at a given height.
   *
   * @generated from protobuf rpc: GetValidatorSetByHeight
   */
  getValidatorSetByHeight(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ABCIQuery defines a query handler that supports ABCI queries directly to the
   * application, bypassing Tendermint completely. The ABCI query must contain
   * a valid and supported path, including app, custom, p2p, and store.
   *
   * Since: cosmos-sdk 0.46
   *
   * @generated from protobuf rpc: ABCIQuery
   */
  aBCIQuery(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  ServiceClient
};
