// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/base/tendermint/v1beta1/query_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { DefaultNodeInfo } from "../../../../cometbft/p2p/v1/types_pb.js";
import { Block as Block$ } from "./types_pb.js";
import { Block } from "../../../../cometbft/types/v1/block_pb.js";
import { BlockID } from "../../../../cometbft/types/v1/types_pb.js";
import { Any } from "../../../../google/protobuf/any_pb.js";
import { PageResponse } from "../../query/v1beta1/pagination_pb.js";
import { PageRequest } from "../../query/v1beta1/pagination_pb.js";
class GetValidatorSetByHeightRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetValidatorSetByHeightRequest", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetValidatorSetByHeightRequest = new GetValidatorSetByHeightRequest$Type();
class GetValidatorSetByHeightResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetValidatorSetByHeightResponse", [
      {
        no: 1,
        name: "block_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "validators", kind: "message", repeat: 2, T: () => Validator },
      { no: 3, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.blockHeight = 0n;
    message.validators = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 block_height */
        1:
          message.blockHeight = reader.int64().toBigInt();
          break;
        case /* repeated cosmos.base.tendermint.v1beta1.Validator validators */
        2:
          message.validators.push(Validator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        3:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockHeight !== 0n)
      writer.tag(1, WireType.Varint).int64(message.blockHeight);
    for (let i = 0; i < message.validators.length; i++)
      Validator.internalBinaryWrite(message.validators[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetValidatorSetByHeightResponse = new GetValidatorSetByHeightResponse$Type();
class GetLatestValidatorSetRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetLatestValidatorSetRequest", [
      { no: 1, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        1:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetLatestValidatorSetRequest = new GetLatestValidatorSetRequest$Type();
class GetLatestValidatorSetResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetLatestValidatorSetResponse", [
      {
        no: 1,
        name: "block_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "validators", kind: "message", repeat: 2, T: () => Validator },
      { no: 3, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.blockHeight = 0n;
    message.validators = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 block_height */
        1:
          message.blockHeight = reader.int64().toBigInt();
          break;
        case /* repeated cosmos.base.tendermint.v1beta1.Validator validators */
        2:
          message.validators.push(Validator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        3:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockHeight !== 0n)
      writer.tag(1, WireType.Varint).int64(message.blockHeight);
    for (let i = 0; i < message.validators.length; i++)
      Validator.internalBinaryWrite(message.validators[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetLatestValidatorSetResponse = new GetLatestValidatorSetResponse$Type();
class Validator$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.Validator", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "pub_key", kind: "message", T: () => Any },
      {
        no: 3,
        name: "voting_power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "proposer_priority",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.votingPower = 0n;
    message.proposerPriority = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* google.protobuf.Any pub_key */
        2:
          message.pubKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.pubKey);
          break;
        case /* int64 voting_power */
        3:
          message.votingPower = reader.int64().toBigInt();
          break;
        case /* int64 proposer_priority */
        4:
          message.proposerPriority = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.pubKey)
      Any.internalBinaryWrite(message.pubKey, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.votingPower !== 0n)
      writer.tag(3, WireType.Varint).int64(message.votingPower);
    if (message.proposerPriority !== 0n)
      writer.tag(4, WireType.Varint).int64(message.proposerPriority);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Validator = new Validator$Type();
class GetBlockByHeightRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetBlockByHeightRequest", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetBlockByHeightRequest = new GetBlockByHeightRequest$Type();
class GetBlockByHeightResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetBlockByHeightResponse", [
      { no: 1, name: "block_id", kind: "message", T: () => BlockID },
      { no: 2, name: "block", kind: "message", T: () => Block },
      { no: 3, name: "sdk_block", kind: "message", T: () => Block$ }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.BlockID block_id */
        1:
          message.blockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* cometbft.types.v1.Block block */
        2:
          message.block = Block.internalBinaryRead(reader, reader.uint32(), options, message.block);
          break;
        case /* cosmos.base.tendermint.v1beta1.Block sdk_block */
        3:
          message.sdkBlock = Block$.internalBinaryRead(reader, reader.uint32(), options, message.sdkBlock);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockId)
      BlockID.internalBinaryWrite(message.blockId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.block)
      Block.internalBinaryWrite(message.block, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.sdkBlock)
      Block$.internalBinaryWrite(message.sdkBlock, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetBlockByHeightResponse = new GetBlockByHeightResponse$Type();
class GetLatestBlockRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetLatestBlockRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetLatestBlockRequest = new GetLatestBlockRequest$Type();
class GetLatestBlockResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetLatestBlockResponse", [
      { no: 1, name: "block_id", kind: "message", T: () => BlockID },
      { no: 2, name: "block", kind: "message", T: () => Block },
      { no: 3, name: "sdk_block", kind: "message", T: () => Block$ }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.BlockID block_id */
        1:
          message.blockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.blockId);
          break;
        case /* cometbft.types.v1.Block block */
        2:
          message.block = Block.internalBinaryRead(reader, reader.uint32(), options, message.block);
          break;
        case /* cosmos.base.tendermint.v1beta1.Block sdk_block */
        3:
          message.sdkBlock = Block$.internalBinaryRead(reader, reader.uint32(), options, message.sdkBlock);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockId)
      BlockID.internalBinaryWrite(message.blockId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.block)
      Block.internalBinaryWrite(message.block, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.sdkBlock)
      Block$.internalBinaryWrite(message.sdkBlock, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetLatestBlockResponse = new GetLatestBlockResponse$Type();
class GetSyncingRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetSyncingRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetSyncingRequest = new GetSyncingRequest$Type();
class GetSyncingResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetSyncingResponse", [
      {
        no: 1,
        name: "syncing",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.syncing = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bool syncing */
        1:
          message.syncing = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.syncing !== false)
      writer.tag(1, WireType.Varint).bool(message.syncing);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetSyncingResponse = new GetSyncingResponse$Type();
class GetNodeInfoRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetNodeInfoRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetNodeInfoRequest = new GetNodeInfoRequest$Type();
class GetNodeInfoResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.GetNodeInfoResponse", [
      { no: 1, name: "default_node_info", kind: "message", T: () => DefaultNodeInfo },
      { no: 2, name: "application_version", kind: "message", T: () => VersionInfo }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.p2p.v1.DefaultNodeInfo default_node_info */
        1:
          message.defaultNodeInfo = DefaultNodeInfo.internalBinaryRead(reader, reader.uint32(), options, message.defaultNodeInfo);
          break;
        case /* cosmos.base.tendermint.v1beta1.VersionInfo application_version */
        2:
          message.applicationVersion = VersionInfo.internalBinaryRead(reader, reader.uint32(), options, message.applicationVersion);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.defaultNodeInfo)
      DefaultNodeInfo.internalBinaryWrite(message.defaultNodeInfo, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.applicationVersion)
      VersionInfo.internalBinaryWrite(message.applicationVersion, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetNodeInfoResponse = new GetNodeInfoResponse$Type();
class VersionInfo$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.VersionInfo", [
      {
        no: 1,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "app_name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "git_commit",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "build_tags",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "go_version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 7, name: "build_deps", kind: "message", repeat: 2, T: () => Module },
      {
        no: 8,
        name: "cosmos_sdk_version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.name = "";
    message.appName = "";
    message.version = "";
    message.gitCommit = "";
    message.buildTags = "";
    message.goVersion = "";
    message.buildDeps = [];
    message.cosmosSdkVersion = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string name */
        1:
          message.name = reader.string();
          break;
        case /* string app_name */
        2:
          message.appName = reader.string();
          break;
        case /* string version */
        3:
          message.version = reader.string();
          break;
        case /* string git_commit */
        4:
          message.gitCommit = reader.string();
          break;
        case /* string build_tags */
        5:
          message.buildTags = reader.string();
          break;
        case /* string go_version */
        6:
          message.goVersion = reader.string();
          break;
        case /* repeated cosmos.base.tendermint.v1beta1.Module build_deps */
        7:
          message.buildDeps.push(Module.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string cosmos_sdk_version */
        8:
          message.cosmosSdkVersion = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.name !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.name);
    if (message.appName !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.appName);
    if (message.version !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.version);
    if (message.gitCommit !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.gitCommit);
    if (message.buildTags !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.buildTags);
    if (message.goVersion !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.goVersion);
    for (let i = 0; i < message.buildDeps.length; i++)
      Module.internalBinaryWrite(message.buildDeps[i], writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.cosmosSdkVersion !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.cosmosSdkVersion);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const VersionInfo = new VersionInfo$Type();
class Module$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.Module", [
      {
        no: 1,
        name: "path",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "sum",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.path = "";
    message.version = "";
    message.sum = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string path */
        1:
          message.path = reader.string();
          break;
        case /* string version */
        2:
          message.version = reader.string();
          break;
        case /* string sum */
        3:
          message.sum = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.path);
    if (message.version !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.version);
    if (message.sum !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.sum);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Module = new Module$Type();
class ABCIQueryRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.ABCIQueryRequest", [
      {
        no: 1,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "path",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "prove",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.data = new Uint8Array(0);
    message.path = "";
    message.height = 0n;
    message.prove = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes data */
        1:
          message.data = reader.bytes();
          break;
        case /* string path */
        2:
          message.path = reader.string();
          break;
        case /* int64 height */
        3:
          message.height = reader.int64().toBigInt();
          break;
        case /* bool prove */
        4:
          message.prove = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.data.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.data);
    if (message.path !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.path);
    if (message.height !== 0n)
      writer.tag(3, WireType.Varint).int64(message.height);
    if (message.prove !== false)
      writer.tag(4, WireType.Varint).bool(message.prove);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ABCIQueryRequest = new ABCIQueryRequest$Type();
class ABCIQueryResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.ABCIQueryResponse", [
      {
        no: 1,
        name: "code",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "log",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "info",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "index",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "key",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 7,
        name: "value",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 8, name: "proof_ops", kind: "message", T: () => ProofOps },
      {
        no: 9,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 10,
        name: "codespace",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.code = 0;
    message.log = "";
    message.info = "";
    message.index = 0n;
    message.key = new Uint8Array(0);
    message.value = new Uint8Array(0);
    message.height = 0n;
    message.codespace = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 code */
        1:
          message.code = reader.uint32();
          break;
        case /* string log */
        3:
          message.log = reader.string();
          break;
        case /* string info */
        4:
          message.info = reader.string();
          break;
        case /* int64 index */
        5:
          message.index = reader.int64().toBigInt();
          break;
        case /* bytes key */
        6:
          message.key = reader.bytes();
          break;
        case /* bytes value */
        7:
          message.value = reader.bytes();
          break;
        case /* cosmos.base.tendermint.v1beta1.ProofOps proof_ops */
        8:
          message.proofOps = ProofOps.internalBinaryRead(reader, reader.uint32(), options, message.proofOps);
          break;
        case /* int64 height */
        9:
          message.height = reader.int64().toBigInt();
          break;
        case /* string codespace */
        10:
          message.codespace = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.code !== 0)
      writer.tag(1, WireType.Varint).uint32(message.code);
    if (message.log !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.log);
    if (message.info !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.info);
    if (message.index !== 0n)
      writer.tag(5, WireType.Varint).int64(message.index);
    if (message.key.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.key);
    if (message.value.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.value);
    if (message.proofOps)
      ProofOps.internalBinaryWrite(message.proofOps, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    if (message.height !== 0n)
      writer.tag(9, WireType.Varint).int64(message.height);
    if (message.codespace !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.codespace);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ABCIQueryResponse = new ABCIQueryResponse$Type();
class ProofOp$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.ProofOp", [
      {
        no: 1,
        name: "type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "key",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.type = "";
    message.key = new Uint8Array(0);
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string type */
        1:
          message.type = reader.string();
          break;
        case /* bytes key */
        2:
          message.key = reader.bytes();
          break;
        case /* bytes data */
        3:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.type !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.type);
    if (message.key.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.key);
    if (message.data.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ProofOp = new ProofOp$Type();
class ProofOps$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.ProofOps", [
      { no: 1, name: "ops", kind: "message", repeat: 2, T: () => ProofOp, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.ops = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.tendermint.v1beta1.ProofOp ops */
        1:
          message.ops.push(ProofOp.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.ops.length; i++)
      ProofOp.internalBinaryWrite(message.ops[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ProofOps = new ProofOps$Type();
const Service = new ServiceType("cosmos.base.tendermint.v1beta1.Service", [
  { name: "GetNodeInfo", options: { "google.api.http": { get: "/cosmos/base/tendermint/v1beta1/node_info" } }, I: GetNodeInfoRequest, O: GetNodeInfoResponse },
  { name: "GetSyncing", options: { "google.api.http": { get: "/cosmos/base/tendermint/v1beta1/syncing" } }, I: GetSyncingRequest, O: GetSyncingResponse },
  { name: "GetLatestBlock", options: { "google.api.http": { get: "/cosmos/base/tendermint/v1beta1/blocks/latest" } }, I: GetLatestBlockRequest, O: GetLatestBlockResponse },
  { name: "GetBlockByHeight", options: { "google.api.http": { get: "/cosmos/base/tendermint/v1beta1/blocks/{height}" } }, I: GetBlockByHeightRequest, O: GetBlockByHeightResponse },
  { name: "GetLatestValidatorSet", options: { "google.api.http": { get: "/cosmos/base/tendermint/v1beta1/validatorsets/latest" } }, I: GetLatestValidatorSetRequest, O: GetLatestValidatorSetResponse },
  { name: "GetValidatorSetByHeight", options: { "google.api.http": { get: "/cosmos/base/tendermint/v1beta1/validatorsets/{height}" } }, I: GetValidatorSetByHeightRequest, O: GetValidatorSetByHeightResponse },
  { name: "ABCIQuery", options: { "google.api.http": { get: "/cosmos/base/tendermint/v1beta1/abci_query" } }, I: ABCIQueryRequest, O: ABCIQueryResponse }
]);
export {
  ABCIQueryRequest,
  ABCIQueryResponse,
  GetBlockByHeightRequest,
  GetBlockByHeightResponse,
  GetLatestBlockRequest,
  GetLatestBlockResponse,
  GetLatestValidatorSetRequest,
  GetLatestValidatorSetResponse,
  GetNodeInfoRequest,
  GetNodeInfoResponse,
  GetSyncingRequest,
  GetSyncingResponse,
  GetValidatorSetByHeightRequest,
  GetValidatorSetByHeightResponse,
  Module,
  ProofOp,
  ProofOps,
  Service,
  Validator,
  VersionInfo
};
