// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/base/tendermint/v1beta1/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { BlockID } from "../../../../cometbft/types/v1/types_pb.js";
import { Timestamp } from "../../../../google/protobuf/timestamp_pb.js";
import { Consensus } from "../../../../cometbft/version/v1/types_pb.js";
import { Commit } from "../../../../cometbft/types/v1/types_pb.js";
import { EvidenceList } from "../../../../cometbft/types/v1/evidence_pb.js";
import { Data } from "../../../../cometbft/types/v1/types_pb.js";
class Block$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.Block", [
      { no: 1, name: "header", kind: "message", T: () => Header, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "data", kind: "message", T: () => Data, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 3, name: "evidence", kind: "message", T: () => EvidenceList, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 4, name: "last_commit", kind: "message", T: () => Commit }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.tendermint.v1beta1.Header header */
        1:
          message.header = Header.internalBinaryRead(reader, reader.uint32(), options, message.header);
          break;
        case /* cometbft.types.v1.Data data */
        2:
          message.data = Data.internalBinaryRead(reader, reader.uint32(), options, message.data);
          break;
        case /* cometbft.types.v1.EvidenceList evidence */
        3:
          message.evidence = EvidenceList.internalBinaryRead(reader, reader.uint32(), options, message.evidence);
          break;
        case /* cometbft.types.v1.Commit last_commit */
        4:
          message.lastCommit = Commit.internalBinaryRead(reader, reader.uint32(), options, message.lastCommit);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.header)
      Header.internalBinaryWrite(message.header, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.data)
      Data.internalBinaryWrite(message.data, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.evidence)
      EvidenceList.internalBinaryWrite(message.evidence, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.lastCommit)
      Commit.internalBinaryWrite(message.lastCommit, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Block = new Block$Type();
class Header$Type extends MessageType {
  constructor() {
    super("cosmos.base.tendermint.v1beta1.Header", [
      { no: 1, name: "version", kind: "message", T: () => Consensus, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "chain_id", kind: "scalar", T: 9, options: { "gogoproto.customname": "ChainID" } },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      { no: 5, name: "last_block_id", kind: "message", T: () => BlockID, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 6,
        name: "last_commit_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 7,
        name: "data_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 8,
        name: "validators_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 9,
        name: "next_validators_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 10,
        name: "consensus_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 11,
        name: "app_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 12,
        name: "last_results_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 13,
        name: "evidence_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 14,
        name: "proposer_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chainId = "";
    message.height = 0n;
    message.lastCommitHash = new Uint8Array(0);
    message.dataHash = new Uint8Array(0);
    message.validatorsHash = new Uint8Array(0);
    message.nextValidatorsHash = new Uint8Array(0);
    message.consensusHash = new Uint8Array(0);
    message.appHash = new Uint8Array(0);
    message.lastResultsHash = new Uint8Array(0);
    message.evidenceHash = new Uint8Array(0);
    message.proposerAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.version.v1.Consensus version */
        1:
          message.version = Consensus.internalBinaryRead(reader, reader.uint32(), options, message.version);
          break;
        case /* string chain_id */
        2:
          message.chainId = reader.string();
          break;
        case /* int64 height */
        3:
          message.height = reader.int64().toBigInt();
          break;
        case /* google.protobuf.Timestamp time */
        4:
          message.time = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.time);
          break;
        case /* cometbft.types.v1.BlockID last_block_id */
        5:
          message.lastBlockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.lastBlockId);
          break;
        case /* bytes last_commit_hash */
        6:
          message.lastCommitHash = reader.bytes();
          break;
        case /* bytes data_hash */
        7:
          message.dataHash = reader.bytes();
          break;
        case /* bytes validators_hash */
        8:
          message.validatorsHash = reader.bytes();
          break;
        case /* bytes next_validators_hash */
        9:
          message.nextValidatorsHash = reader.bytes();
          break;
        case /* bytes consensus_hash */
        10:
          message.consensusHash = reader.bytes();
          break;
        case /* bytes app_hash */
        11:
          message.appHash = reader.bytes();
          break;
        case /* bytes last_results_hash */
        12:
          message.lastResultsHash = reader.bytes();
          break;
        case /* bytes evidence_hash */
        13:
          message.evidenceHash = reader.bytes();
          break;
        case /* string proposer_address */
        14:
          message.proposerAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.version)
      Consensus.internalBinaryWrite(message.version, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.chainId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.chainId);
    if (message.height !== 0n)
      writer.tag(3, WireType.Varint).int64(message.height);
    if (message.time)
      Timestamp.internalBinaryWrite(message.time, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.lastBlockId)
      BlockID.internalBinaryWrite(message.lastBlockId, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.lastCommitHash.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.lastCommitHash);
    if (message.dataHash.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.dataHash);
    if (message.validatorsHash.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.validatorsHash);
    if (message.nextValidatorsHash.length)
      writer.tag(9, WireType.LengthDelimited).bytes(message.nextValidatorsHash);
    if (message.consensusHash.length)
      writer.tag(10, WireType.LengthDelimited).bytes(message.consensusHash);
    if (message.appHash.length)
      writer.tag(11, WireType.LengthDelimited).bytes(message.appHash);
    if (message.lastResultsHash.length)
      writer.tag(12, WireType.LengthDelimited).bytes(message.lastResultsHash);
    if (message.evidenceHash.length)
      writer.tag(13, WireType.LengthDelimited).bytes(message.evidenceHash);
    if (message.proposerAddress !== "")
      writer.tag(14, WireType.LengthDelimited).string(message.proposerAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Header = new Header$Type();
export {
  Block,
  Header
};
