// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/store/v1beta1/listening_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { FinalizeBlockResponse } from "../../../cometbft/abci/v1/types_pb.js";
import { FinalizeBlockRequest } from "../../../cometbft/abci/v1/types_pb.js";
import { CommitResponse } from "../../../cometbft/abci/v1/types_pb.js";
class StoreKVPair$Type extends MessageType {
  constructor() {
    super("cosmos.store.v1beta1.StoreKVPair", [
      {
        no: 1,
        name: "store_key",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "delete",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 3,
        name: "key",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "value",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.storeKey = "";
    message.delete = false;
    message.key = new Uint8Array(0);
    message.value = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string store_key */
        1:
          message.storeKey = reader.string();
          break;
        case /* bool delete */
        2:
          message.delete = reader.bool();
          break;
        case /* bytes key */
        3:
          message.key = reader.bytes();
          break;
        case /* bytes value */
        4:
          message.value = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.storeKey !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.storeKey);
    if (message.delete !== false)
      writer.tag(2, WireType.Varint).bool(message.delete);
    if (message.key.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.key);
    if (message.value.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StoreKVPair = new StoreKVPair$Type();
class BlockMetadata$Type extends MessageType {
  constructor() {
    super("cosmos.store.v1beta1.BlockMetadata", [
      { no: 6, name: "response_commit", kind: "message", T: () => CommitResponse },
      { no: 7, name: "request_finalize_block", kind: "message", T: () => FinalizeBlockRequest },
      { no: 8, name: "response_finalize_block", kind: "message", T: () => FinalizeBlockResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.abci.v1.CommitResponse response_commit */
        6:
          message.responseCommit = CommitResponse.internalBinaryRead(reader, reader.uint32(), options, message.responseCommit);
          break;
        case /* cometbft.abci.v1.FinalizeBlockRequest request_finalize_block */
        7:
          message.requestFinalizeBlock = FinalizeBlockRequest.internalBinaryRead(reader, reader.uint32(), options, message.requestFinalizeBlock);
          break;
        case /* cometbft.abci.v1.FinalizeBlockResponse response_finalize_block */
        8:
          message.responseFinalizeBlock = FinalizeBlockResponse.internalBinaryRead(reader, reader.uint32(), options, message.responseFinalizeBlock);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.responseCommit)
      CommitResponse.internalBinaryWrite(message.responseCommit, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.requestFinalizeBlock)
      FinalizeBlockRequest.internalBinaryWrite(message.requestFinalizeBlock, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.responseFinalizeBlock)
      FinalizeBlockResponse.internalBinaryWrite(message.responseFinalizeBlock, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BlockMetadata = new BlockMetadata$Type();
export {
  BlockMetadata,
  StoreKVPair
};
