// @injectivelabs/sdk-ts/dist/esm/AbacusGrpcApi-H-P0UiPh.js
import { t as _defineProperty } from "./defineProperty-Bq6QU9Te.js";
import { t as BaseGrpcConsumer } from "./BaseGrpcConsumer-a7fB9A_e.js";
import { IndexerErrorModule } from "@injectivelabs/exceptions";
import * as PointsSvcPb from "@injectivelabs/abacus-proto-ts-v2/generated/points_svc_pb";
import { PointsSvcClient } from "@injectivelabs/abacus-proto-ts-v2/generated/points_svc_pb.client";

//#region src/client/abacus/grpc/transformers/index.ts
var AbacusGrpcTransformer = class {
	static grpcPointsStatsDailyToPointsStatsDaily(response) {
		return response.rows.map((row) => ({
			day: row.day,
			week: row.week,
			volume: row.volume,
			points: row.points.toString(),
			pointsPrecise: row.pointsPrecise
		}));
	}
	static grpcPointsStatsWeeklyToPointsStatsWeekly(response) {
		return response.rows.map((row) => ({
			day: row.day,
			week: row.week,
			volume: row.volume,
			points: row.points.toString(),
			pointsPrecise: row.pointsPrecise
		}));
	}
	static grpcPointsLatestToPointsLatest(response) {
		return {
			league: response.league,
			updatedAt: response.updatedAt,
			rank: response.rank.toString(),
			pointsBonus: response.pointsBonus,
			pointsSeason1: response.pointsSeason1,
			totalPoints: response.totalPoints.toString(),
			totalPointsPrecise: response.totalPointsPrecise
		};
	}
};

//#endregion
//#region src/client/abacus/grpc/AbacusGrpcApi.ts
var AbacusGrpcApi = class extends BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "module", IndexerErrorModule.Abacus);
	}
	get client() {
		return this.initClient(PointsSvcClient);
	}
	async fetchAccountLatestPoints(address) {
		const request = PointsSvcPb.PointsLatestForAccountRequest.create({ accountAddress: address });
		const response = await this.executeGrpcCall(request, this.client.pointsLatestForAccount.bind(this.client));
		return AbacusGrpcTransformer.grpcPointsLatestToPointsLatest(response);
	}
	async fetchAccountDailyPoints(address, daysLimit) {
		const request = PointsSvcPb.PointsStatsDailyForAccountRequest.create({
			accountAddress: address,
			daysLimit: daysLimit ? BigInt(daysLimit) : void 0
		});
		const response = await this.executeGrpcCall(request, this.client.pointsStatsDailyForAccount.bind(this.client));
		return AbacusGrpcTransformer.grpcPointsStatsDailyToPointsStatsDaily(response);
	}
	async fetchAccountWeeklyPoints(address, weeksLimit) {
		const request = PointsSvcPb.PointsStatsWeeklyForAccountRequest.create({
			accountAddress: address,
			weeksLimit: weeksLimit ? BigInt(weeksLimit) : void 0
		});
		const response = await this.executeGrpcCall(request, this.client.pointsStatsWeeklyForAccount.bind(this.client));
		return AbacusGrpcTransformer.grpcPointsStatsWeeklyToPointsStatsWeekly(response);
	}
};

//#endregion
export { AbacusGrpcApi as t };
