// @injectivelabs/sdk-ts/dist/esm/ExecArgNeptuneWithdraw-BRJZq8Vb.js
import { Wt as fromUtf8, Yt as toBase64, pn as safeBigIntStringify } from "./utils-CewUFGKA.js";
import { t as _defineProperty } from "./defineProperty-Bq6QU9Te.js";
import { t as MsgBase } from "./MsgBase-D4pO6org.js";
import { GeneralException } from "@injectivelabs/exceptions";
import * as InjectiveWasmxV1TxPb from "@injectivelabs/core-proto-ts-v2/generated/injective/wasmx/v1/tx_pb";

//#region src/core/modules/wasm/ExecArgBase.ts
const dataToExecData = (action, data) => {
	return { [action]: data };
};
var ExecArgBase = class {
	constructor(params) {
		_defineProperty(this, "params", void 0);
		this.params = params;
	}
	toJSON() {
		return fromUtf8(safeBigIntStringify(this.params));
	}
	toExecJSON() {
		return fromUtf8(safeBigIntStringify(this.toExecData()));
	}
};

//#endregion
//#region src/core/modules/wasm/exec-args/ExecArgNeptuneDeposit.ts
/**
* @category Contract Exec Arguments
*/
var ExecArgNeptuneDeposit = class ExecArgNeptuneDeposit extends ExecArgBase {
	static fromJSON(params) {
		return new ExecArgNeptuneDeposit(params);
	}
	toData() {
		return {};
	}
	toExecData() {
		return dataToExecData("lend", this.toData());
	}
};

//#endregion
//#region src/core/modules/wasm/msgs/MsgExecuteContractCompat.ts
/**
* @category Messages
*/
var MsgExecuteContractCompat = class MsgExecuteContractCompat extends MsgBase {
	static fromJSON(params) {
		return new MsgExecuteContractCompat(params);
	}
	toProto() {
		const { params } = this;
		const msg = this.getMsgObject();
		const funds = params.funds ? (Array.isArray(params.funds) ? params.funds : [params.funds]).map((coin) => `${coin.amount}${coin.denom}`).join(",") : "0";
		return InjectiveWasmxV1TxPb.MsgExecuteContractCompat.create({
			sender: params.sender,
			contract: params.contractAddress,
			msg: safeBigIntStringify(msg),
			funds
		});
	}
	toData() {
		const proto = this.toProto();
		return {
			"@type": "/injective.wasmx.v1.MsgExecuteContractCompat",
			...proto
		};
	}
	toAmino() {
		const proto = this.toProto();
		return {
			type: "wasmx/MsgExecuteContractCompat",
			value: {
				sender: proto.sender,
				contract: proto.contract,
				msg: safeBigIntStringify(this.getMsgObject()),
				funds: proto.funds
			}
		};
	}
	toWeb3Gw() {
		const { value } = this.toAmino();
		return {
			"@type": "/injective.wasmx.v1.MsgExecuteContractCompat",
			...value
		};
	}
	toDirectSign() {
		return {
			type: "/injective.wasmx.v1.MsgExecuteContractCompat",
			message: this.toProto()
		};
	}
	toBinary() {
		return InjectiveWasmxV1TxPb.MsgExecuteContractCompat.toBinary(this.toProto());
	}
	getMsgObject() {
		const { params } = this;
		if ((params.exec || params.msg) && params.execArgs) throw new GeneralException(/* @__PURE__ */ new Error("Please provide only one exec|msg argument"));
		if (params.execArgs) return params.execArgs.toExecData();
		if (params.exec) return { [params.exec.action]: params.exec.msg };
		if (params.msg) return params.msg;
		throw new GeneralException(/* @__PURE__ */ new Error("Please provide at least one exec argument"));
	}
};

//#endregion
//#region src/core/modules/wasm/exec-args/ExecArgNeptuneWithdraw.ts
/**
* @category Contract Exec Arguments
*/
var ExecArgNeptuneWithdraw = class ExecArgNeptuneWithdraw extends ExecArgBase {
	static fromJSON(params) {
		return new ExecArgNeptuneWithdraw(params);
	}
	toData() {
		const { params } = this;
		const encodedMsg = toBase64({ redeem: {} });
		return {
			amount: params.amount,
			contract: params.contract,
			msg: encodedMsg
		};
	}
	toExecData() {
		return dataToExecData("send", this.toData());
	}
};

//#endregion
export { dataToExecData as a, ExecArgBase as i, MsgExecuteContractCompat as n, ExecArgNeptuneDeposit as r, ExecArgNeptuneWithdraw as t };
