// @injectivelabs/sdk-ts/dist/esm/OLPGrpcApi-CX-m2wvC.js
import { t as _defineProperty } from "./defineProperty-Bq6QU9Te.js";
import { t as BaseIndexerGrpcConsumer } from "./BaseIndexerGrpcConsumer-CrwqSxDq.js";
import { IndexerErrorModule } from "@injectivelabs/exceptions";
import * as DmmPb from "@injectivelabs/olp-proto-ts-v2/generated/goagen_olp_injective_dmm_v2_pb";
import { InjectiveDmmV2Client } from "@injectivelabs/olp-proto-ts-v2/generated/goagen_olp_injective_dmm_v2_pb.client";

//#region src/client/olp/grpc/transformers/index.ts
var DmmGrpcTransformer = class DmmGrpcTransformer {
	static epochsResponseToEpochs(response) {
		return response.epochs.map(DmmGrpcTransformer.grpcEpochToEpoch);
	}
	static grpcEpochToEpoch(epoch) {
		var _epoch$endHeight, _epoch$startHeight;
		return {
			status: epoch.status,
			epochId: epoch.epochId,
			resultCount: epoch.resultCount,
			snapshotCount: epoch.snapshotCount,
			endHeight: (_epoch$endHeight = epoch.endHeight) === null || _epoch$endHeight === void 0 ? void 0 : _epoch$endHeight.toString(),
			startHeight: (_epoch$startHeight = epoch.startHeight) === null || _epoch$startHeight === void 0 ? void 0 : _epoch$startHeight.toString(),
			config: epoch.config ? {
				number: epoch.config.number,
				rewardInj: epoch.config.rewardInj,
				isMiniEpoch: epoch.config.isMiniEpoch,
				uptimeExponent: epoch.config.uptimeExponent,
				volumeExponent: epoch.config.volumeExponent,
				liquidityScoreExponent: epoch.config.liquidityScoreExponent,
				permanenceVolumeThreshold: epoch.config.permanenceVolumeThreshold,
				qualifyingVolumeThreshold: epoch.config.qualifyingVolumeThreshold,
				startDate: epoch.config.startDate ? new Date(epoch.config.startDate) : void 0,
				endDate: epoch.config.endDate ? new Date(epoch.config.endDate) : void 0,
				markets: epoch.config.markets.map((market) => ({
					ticker: market.ticker,
					marketId: market.marketId,
					startDate: market.startDate ? new Date(market.startDate) : void 0,
					preAllocatedReward: market.preAllocatedReward
				}))
			} : void 0,
			createdAt: epoch.createdAt ? new Date(epoch.createdAt) : void 0,
			updatedAt: epoch.updatedAt ? new Date(epoch.updatedAt) : void 0
		};
	}
	static marketRewardsResponseToMarketRewards(response) {
		return response.rewards.map(DmmGrpcTransformer.grpcMarketRewardToMarketReward);
	}
	static grpcMarketRewardToMarketReward(marketReward) {
		var _marketReward$height;
		return {
			reward: marketReward.reward,
			epochId: marketReward.epochId,
			marketId: marketReward.marketId,
			liquidity: marketReward.liquidity,
			totalScore: marketReward.totalScore,
			height: (_marketReward$height = marketReward.height) === null || _marketReward$height === void 0 ? void 0 : _marketReward$height.toString(),
			rewardPercentage: marketReward.rewardPercentage,
			miniEpochsReward: marketReward.miniEpochsReward,
			startDate: marketReward.startDate ? new Date(marketReward.startDate) : void 0,
			endDate: marketReward.endDate ? new Date(marketReward.endDate) : void 0,
			createdAt: marketReward.createdAt ? new Date(marketReward.createdAt) : void 0,
			updatedAt: marketReward.updatedAt ? new Date(marketReward.updatedAt) : void 0
		};
	}
	static eligibleAddressesResponseToEligibleAddresses(response) {
		const addresses = response.addresses;
		return {
			next: response.next,
			addresses: addresses.map(DmmGrpcTransformer.grpcEligibleAddresssesToEligibileAddresses)
		};
	}
	static grpcEligibleAddresssesToEligibileAddresses(eligibleAddress) {
		var _eligibleAddress$heig;
		return {
			source: eligibleAddress.source,
			epochId: eligibleAddress.epochId,
			height: (_eligibleAddress$heig = eligibleAddress.height) === null || _eligibleAddress$heig === void 0 ? void 0 : _eligibleAddress$heig.toString(),
			accountAddress: eligibleAddress.accountAddress,
			createdAt: eligibleAddress.createdAt ? new Date(eligibleAddress.createdAt) : void 0,
			updatedAt: eligibleAddress.updatedAt ? new Date(eligibleAddress.updatedAt) : void 0
		};
	}
	static epochScoresResponseToEpochScores(response) {
		const scores = response.scores;
		return {
			next: response.next,
			scores: scores.map(DmmGrpcTransformer.grpcEpochScoresToEpochScores)
		};
	}
	static grpcEpochScoresToEpochScores(score) {
		var _score$height, _score$startHeight;
		return {
			volume: score.volume,
			reward: score.reward,
			epochId: score.epochId,
			qualifies: score.qualifies,
			totalScore: score.totalScore,
			uptimeScore: score.uptimeScore,
			volumeScore: score.volumeScore,
			makerVolume: score.makerVolume,
			takerVolume: score.takerVolume,
			height: (_score$height = score.height) === null || _score$height === void 0 ? void 0 : _score$height.toString(),
			accountAddress: score.accountAddress,
			liquidityScore: score.liquidityScore,
			uptimePercentage: score.uptimePercentage,
			rewardPercentage: score.rewardPercentage,
			volumePercentage: score.volumePercentage,
			startHeight: (_score$startHeight = score.startHeight) === null || _score$startHeight === void 0 ? void 0 : _score$startHeight.toString(),
			uptimeScorePonderated: score.uptimeScorePonderated,
			volumeScorePonderated: score.volumeScorePonderated,
			liquidityScorePonderated: score.liquidityScorePonderated,
			blockTime: score.blockTime ? new Date(score.blockTime) : void 0,
			createdAt: score.createdAt ? new Date(score.createdAt) : void 0,
			updatedAt: score.updatedAt ? new Date(score.updatedAt) : void 0
		};
	}
	static epochScoresHistoryResponseToEpochScoresHistory(response) {
		const scores = response.scores;
		return {
			next: response.next,
			scores: scores.map(DmmGrpcTransformer.grpcEpochScoresToEpochScores)
		};
	}
	static totalScoresResponseToTotalScores(response) {
		const scores = response.scores;
		return {
			next: response.next,
			scores: scores.map(DmmGrpcTransformer.grpcTotalScoresToTotalScores)
		};
	}
	static grpcTotalScoresToTotalScores(score) {
		var _score$height2, _score$startHeight2;
		return {
			bid: score.bid,
			ask: score.ask,
			depth: score.depth,
			volume: score.volume,
			reward: score.reward,
			epochId: score.epochId,
			marketId: score.marketId,
			totalScore: score.totalScore,
			uptimeScore: score.uptimeScore,
			startVolume: score.startVolume,
			volumeScore: score.volumeScore,
			takerVolume: score.takerVolume,
			makerVolume: score.makerVolume,
			height: (_score$height2 = score.height) === null || _score$height2 === void 0 ? void 0 : _score$height2.toString(),
			snapshotCount: score.snapshotCount,
			currentVolume: score.currentVolume,
			accountAddress: score.accountAddress,
			liquidityScore: score.liquidityScore,
			uptimePercentage: score.uptimePercentage,
			takerStartVolume: score.takerStartVolume,
			makerStartVolume: score.makerStartVolume,
			rewardPercentage: score.rewardPercentage,
			startHeight: (_score$startHeight2 = score.startHeight) === null || _score$startHeight2 === void 0 ? void 0 : _score$startHeight2.toString(),
			takerCurrentVolume: score.takerCurrentVolume,
			makerCurrentVolume: score.makerCurrentVolume,
			uptimeScorePonderated: score.uptimeScorePonderated,
			volumeScorePonderated: score.volumeScorePonderated,
			liquidityScorePonderated: score.liquidityScorePonderated,
			blockTime: score.blockTime ? new Date(score.blockTime) : void 0,
			createdAt: score.createdAt ? new Date(score.createdAt) : void 0,
			updatedAt: score.updatedAt ? new Date(score.updatedAt) : void 0
		};
	}
	static totalScoresHistoryResponseToTotalScoresHistory(response) {
		const scores = response.scores;
		return {
			next: response.next,
			scores: scores.map(DmmGrpcTransformer.grpcTotalScoresToTotalScores)
		};
	}
	static rewardsDistributionResponseToRewardsDistribution(response) {
		const rewards = response.rewards;
		return {
			next: response.next,
			rewards: rewards.map(DmmGrpcTransformer.grpcRewardsDistributionToRewardsDistribution)
		};
	}
	static grpcRewardsDistributionToRewardsDistribution(reward) {
		var _reward$height, _reward$startHeight;
		return {
			depth: reward.depth,
			reward: reward.reward,
			epochId: reward.epochId,
			height: (_reward$height = reward.height) === null || _reward$height === void 0 ? void 0 : _reward$height.toString(),
			accountAddress: reward.accountAddress,
			miniEpochsReward: reward.miniEpochsReward,
			startHeight: (_reward$startHeight = reward.startHeight) === null || _reward$startHeight === void 0 ? void 0 : _reward$startHeight.toString(),
			blockTime: reward.blockTime ? new Date(reward.blockTime) : void 0,
			createdAt: reward.createdAt ? new Date(reward.createdAt) : void 0,
			updatedAt: reward.updatedAt ? new Date(reward.updatedAt) : void 0
		};
	}
	static accountVolumesResponseToAccountVolumes(response) {
		return response.volumes.map(DmmGrpcTransformer.grpcAccountVolumesToAccountVolumes);
	}
	static grpcAccountVolumesToAccountVolumes(reward) {
		var _reward$height2;
		return {
			date: reward.date,
			volume: reward.volume,
			epochId: reward.epochId,
			takerVolume: reward.takerVolume,
			makerVolume: reward.makerVolume,
			dailyVolume: reward.dailyVolume,
			height: (_reward$height2 = reward.height) === null || _reward$height2 === void 0 ? void 0 : _reward$height2.toString(),
			accountAddress: reward.accountAddress,
			dailyQualified: reward.dailyQualified,
			volumePercentage: reward.volumePercentage,
			dailyMakerVolume: reward.dailyMakerVolume,
			dailyTakerVolume: reward.dailyTakerVolume,
			makerVolumePercentage: reward.makerVolumePercentage,
			takerVolumePercentage: reward.takerVolumePercentage,
			dailyVolumePercentage: reward.dailyVolumePercentage,
			dailyMakerVolumePercentage: reward.dailyMakerVolumePercentage,
			dailyTakerVolumePercentage: reward.dailyTakerVolumePercentage,
			blockTime: reward.blockTime ? new Date(reward.blockTime) : void 0,
			createdAt: reward.createdAt ? new Date(reward.createdAt) : void 0,
			updatedAt: reward.updatedAt ? new Date(reward.updatedAt) : void 0,
			dateTimestamp: reward.dateTimestamp ? new Date(reward.dateTimestamp) : void 0
		};
	}
	static rewardsEligibilityResponseToRewardsEligibility(response) {
		const volumes = response.volumes;
		return {
			estimatedReward: response.estimatedReward,
			eligibleForNextEpoch: response.eligibleForNextEpoch,
			eligibleForCurrentEpoch: response.eligibleForCurrentEpoch,
			currentMakerVolumePercentage: response.currentMakerVolumePercentage,
			updatedAt: response.updatedAt ? new Date(response.updatedAt) : void 0,
			averageDailyMakerVolumePercentage: response.averageDailyMakerVolumePercentage,
			volumes: volumes.map(DmmGrpcTransformer.grpcAccountVolumesToAccountVolumes)
		};
	}
	static marketRewardsRangeResponseToMarketRewardsRange(response) {
		const formattedMinCurrentEpochRewards = {};
		const formattedMaxCurrentEpochRewards = {};
		response.ranges.forEach((item) => {
			formattedMinCurrentEpochRewards[item.marketId] = item.min;
			formattedMaxCurrentEpochRewards[item.marketId] = item.max;
		});
		return {
			minRewards: formattedMinCurrentEpochRewards,
			maxRewards: formattedMaxCurrentEpochRewards
		};
	}
};

//#endregion
//#region src/client/olp/grpc/OLPGrpcApi.ts
var OLPGrpcApi = class extends BaseIndexerGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		_defineProperty(this, "module", IndexerErrorModule.OLP);
	}
	get client() {
		return this.initClient(InjectiveDmmV2Client);
	}
	async fetchEpochs(status) {
		const request = DmmPb.GetEpochsRequest.create();
		if (status) request.status = status;
		const response = await this.executeGrpcCall(request, this.client.getEpochs.bind(this.client));
		return DmmGrpcTransformer.epochsResponseToEpochs(response);
	}
	async fetchMarketRewards(epochId) {
		const request = DmmPb.GetMarketRewardsRequest.create();
		request.epochId = epochId.toString();
		const response = await this.executeGrpcCall(request, this.client.getMarketRewards.bind(this.client));
		return DmmGrpcTransformer.marketRewardsResponseToMarketRewards(response);
	}
	async fetchEligibleAddresses({ epochId, page }) {
		const request = DmmPb.GetEligibleAddressesRequest.create();
		request.epochId = epochId;
		if (page) request.page = page;
		const response = await this.executeGrpcCall(request, this.client.getEligibleAddresses.bind(this.client));
		return DmmGrpcTransformer.eligibleAddressesResponseToEligibleAddresses(response);
	}
	async fetchEpochScores({ epochId, page }) {
		const request = DmmPb.GetEpochScoresRequest.create();
		request.epochId = epochId;
		if (page) request.page = page;
		const response = await this.executeGrpcCall(request, this.client.getEpochScores.bind(this.client));
		return DmmGrpcTransformer.epochScoresResponseToEpochScores(response);
	}
	async fetchEpochScoresHistory({ epochId, accountAddress, page }) {
		const request = DmmPb.GetEpochScoresHistoryRequest.create();
		request.epochId = epochId;
		request.accountAddress = accountAddress;
		if (page) request.page = page;
		const response = await this.executeGrpcCall(request, this.client.getEpochScoresHistory.bind(this.client));
		return DmmGrpcTransformer.epochScoresHistoryResponseToEpochScoresHistory(response);
	}
	async fetchTotalScores({ epochId, marketId, page }) {
		const request = DmmPb.GetTotalScoresRequest.create();
		request.epochId = epochId;
		request.marketId = marketId;
		if (page) request.page = page;
		const response = await this.executeGrpcCall(request, this.client.getTotalScores.bind(this.client));
		return DmmGrpcTransformer.totalScoresResponseToTotalScores(response);
	}
	async fetchTotalScoresHistory({ epochId, marketId, accountAddress, page }) {
		const request = DmmPb.GetTotalScoresHistoryRequest.create();
		request.epochId = epochId;
		request.marketId = marketId;
		request.accountAddress = accountAddress;
		if (page) request.page = page;
		const response = await this.executeGrpcCall(request, this.client.getTotalScoresHistory.bind(this.client));
		return DmmGrpcTransformer.totalScoresHistoryResponseToTotalScoresHistory(response);
	}
	async fetchRewardsDistribution({ epochId, height, page }) {
		const request = DmmPb.GetRewardsDistributionRequest.create();
		request.epochId = epochId;
		if (height) request.height = BigInt(height);
		if (page) request.page = page;
		const response = await this.executeGrpcCall(request, this.client.getRewardsDistribution.bind(this.client));
		return DmmGrpcTransformer.rewardsDistributionResponseToRewardsDistribution(response);
	}
	async fetchAccountVolumes({ epochId, accountAddress }) {
		const request = DmmPb.GetAccountVolumesRequest.create();
		request.epochId = epochId;
		request.accountAddress = accountAddress;
		const response = await this.executeGrpcCall(request, this.client.getAccountVolumes.bind(this.client));
		return DmmGrpcTransformer.accountVolumesResponseToAccountVolumes(response);
	}
	async fetchRewardsEligibility({ epochId, accountAddress }) {
		const request = DmmPb.GetRewardsEligibilityRequest.create();
		if (epochId) request.epochId = epochId;
		if (accountAddress) request.accountAddress = accountAddress;
		const response = await this.executeGrpcCall(request, this.client.getRewardsEligibility.bind(this.client));
		return DmmGrpcTransformer.rewardsEligibilityResponseToRewardsEligibility(response);
	}
	async fetchMarketRewardsRange({ epochId, marketId }) {
		const request = DmmPb.GetMarketRewardsRangeRequest.create();
		request.epochId = epochId;
		request.marketId = marketId;
		const response = await this.executeGrpcCall(request, this.client.getMarketRewardsRange.bind(this.client));
		return DmmGrpcTransformer.marketRewardsRangeResponseToMarketRewardsRange(response);
	}
};

//#endregion
export { OLPGrpcApi as t };
