// @injectivelabs/sdk-ts/dist/esm/types-Cf4oVv93.js
import * as InjectiveExchangeV1Beta1ExchangePb from "@injectivelabs/core-proto-ts-v2/generated/injective/exchange/v1beta1/exchange_pb";

//#region src/types/token.ts
const TokenType = {
	Ibc: "ibc",
	Cw20: "cw20",
	Spl: "spl",
	Erc20: "erc20",
	Lp: "lp",
	Evm: "evm",
	Native: "native",
	Symbol: "symbol",
	TokenFactory: "tokenFactory",
	InsuranceFund: "insuranceFund",
	Unknown: "unknown"
};
const TokenVerification = {
	Verified: "verified",
	Submitted: "submitted",
	Internal: "internal",
	External: "external",
	Unverified: "unverified"
};
const TokenSource = {
	Aptos: "aptos",
	Solana: "solana",
	Cosmos: "cosmos",
	Ethereum: "ethereum",
	EthereumWh: "ethereum-wormhole",
	Polygon: "polygon",
	Klaytn: "klaytn",
	Arbitrum: "arbitrum",
	Sui: "sui",
	Ibc: "ibc",
	BinanceSmartChain: "binance-smart-chain",
	Axelar: "axelar"
};

//#endregion
//#region src/types/stream.ts
/**
* Subscription interface for stream management with event-based lifecycle
*/
/**
* gRPC Status Codes
* @see https://grpc.io/docs/guides/status-codes/
*/
const GrpcStatusCode = {
	OK: 0,
	CANCELLED: 1,
	UNKNOWN: 2,
	INVALID_ARGUMENT: 3,
	DEADLINE_EXCEEDED: 4,
	NOT_FOUND: 5,
	ALREADY_EXISTS: 6,
	PERMISSION_DENIED: 7,
	RESOURCE_EXHAUSTED: 8,
	FAILED_PRECONDITION: 9,
	ABORTED: 10,
	OUT_OF_RANGE: 11,
	UNIMPLEMENTED: 12,
	INTERNAL: 13,
	UNAVAILABLE: 14,
	DATA_LOSS: 15,
	UNAUTHENTICATED: 16
};
/**
* Enhanced disconnect reasons with gRPC error code mapping
*/
const StreamDisconnectReason = {
	UserStopped: "user-stopped",
	StreamError: "stream-error",
	NetworkError: "network-error",
	Timeout: "timeout",
	AuthenticationError: "authentication-error",
	InvalidRequest: "invalid-request",
	RateLimited: "rate-limited",
	MaxRetries: "max-retries",
	StreamEnded: "stream-ended"
};
const StreamEvent = {
	Data: "data",
	Connect: "connect",
	Disconnect: "disconnect",
	Retry: "retry",
	StateChange: "state:change",
	Error: "error",
	Warn: "warn"
};
/**
* StreamManager
*/
const StreamState = {
	Idle: "idle",
	Connecting: "connecting",
	Connected: "connected",
	Reconnecting: "reconnecting",
	Stopped: "stopped"
};

//#endregion
//#region src/types/exchange.ts
/**
* Trade and Order types
*/
const TradeExecutionType = {
	Market: "market",
	LimitFill: "limitFill",
	LimitMatchRestingOrder: "limitMatchRestingOrder",
	LimitMatchNewOrder: "limitMatchNewOrder"
};
const TradeExecutionSide = {
	Maker: "maker",
	Taker: "taker"
};
const TradeDirection = {
	Buy: "buy",
	Sell: "sell",
	Long: "long",
	Short: "short"
};
const OrderState = {
	Unfilled: "unfilled",
	Booked: "booked",
	PartialFilled: "partial_filled",
	PartiallyFilled: "partially_filled",
	Filled: "filled",
	Canceled: "canceled",
	Triggered: "triggered"
};
const OrderSide = {
	Unspecified: "unspecified",
	Buy: "buy",
	Sell: "sell",
	StopBuy: "stop_buy",
	StopSell: "stop_sell",
	TakeBuy: "take_buy",
	TakeSell: "take_sell",
	BuyPO: "buy_po",
	SellPO: "sell_po",
	BuyAtomic: "buy_atomic",
	SellAtomic: "sell_atomic",
	Unrecognized: "unrecognized"
};
const OrderMaskMap = InjectiveExchangeV1Beta1ExchangePb.OrderMask;

//#endregion
//#region src/types/stream-legacy.ts
/**
* Legacy stream operation types
* Used by indexer stream transformers
*/
const StreamOperation = {
	Insert: "insert",
	Delete: "delete",
	Replace: "replace",
	Update: "update",
	Invalidate: "invalidate"
};

//#endregion
export { TradeDirection as a, GrpcStatusCode as c, StreamState as d, TokenSource as f, OrderState as i, StreamDisconnectReason as l, TokenVerification as m, OrderMaskMap as n, TradeExecutionSide as o, TokenType as p, OrderSide as r, TradeExecutionType as s, StreamOperation as t, StreamEvent as u };
