// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/orm/v1alpha1/schema_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
var StorageType = /* @__PURE__ */ ((StorageType2) => {
  StorageType2[StorageType2["DEFAULT_UNSPECIFIED"] = 0] = "DEFAULT_UNSPECIFIED";
  StorageType2[StorageType2["MEMORY"] = 1] = "MEMORY";
  StorageType2[StorageType2["TRANSIENT"] = 2] = "TRANSIENT";
  return StorageType2;
})(StorageType || {});
class ModuleSchemaDescriptor$Type extends MessageType {
  constructor() {
    super("cosmos.orm.v1alpha1.ModuleSchemaDescriptor", [
      { no: 1, name: "schema_file", kind: "message", repeat: 2, T: () => ModuleSchemaDescriptor_FileEntry },
      {
        no: 2,
        name: "prefix",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.schemaFile = [];
    message.prefix = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.orm.v1alpha1.ModuleSchemaDescriptor.FileEntry schema_file */
        1:
          message.schemaFile.push(ModuleSchemaDescriptor_FileEntry.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* bytes prefix */
        2:
          message.prefix = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.schemaFile.length; i++)
      ModuleSchemaDescriptor_FileEntry.internalBinaryWrite(message.schemaFile[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.prefix.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.prefix);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ModuleSchemaDescriptor = new ModuleSchemaDescriptor$Type();
class ModuleSchemaDescriptor_FileEntry$Type extends MessageType {
  constructor() {
    super("cosmos.orm.v1alpha1.ModuleSchemaDescriptor.FileEntry", [
      {
        no: 1,
        name: "id",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 2,
        name: "proto_file_name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "storage_type", kind: "enum", T: () => ["cosmos.orm.v1alpha1.StorageType", StorageType, "STORAGE_TYPE_"] }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.id = 0;
    message.protoFileName = "";
    message.storageType = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 id */
        1:
          message.id = reader.uint32();
          break;
        case /* string proto_file_name */
        2:
          message.protoFileName = reader.string();
          break;
        case /* cosmos.orm.v1alpha1.StorageType storage_type */
        3:
          message.storageType = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.id !== 0)
      writer.tag(1, WireType.Varint).uint32(message.id);
    if (message.protoFileName !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.protoFileName);
    if (message.storageType !== 0)
      writer.tag(3, WireType.Varint).int32(message.storageType);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ModuleSchemaDescriptor_FileEntry = new ModuleSchemaDescriptor_FileEntry$Type();
export {
  ModuleSchemaDescriptor,
  ModuleSchemaDescriptor_FileEntry,
  StorageType
};
