// @injectivelabs/abacus-proto-ts-v2/esm/generated/danielvladco/protobuf/graphql_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
var Type = /* @__PURE__ */ ((Type2) => {
  Type2[Type2["DEFAULT"] = 0] = "DEFAULT";
  Type2[Type2["MUTATION"] = 1] = "MUTATION";
  Type2[Type2["QUERY"] = 2] = "QUERY";
  return Type2;
})(Type || {});
var Upstream = /* @__PURE__ */ ((Upstream2) => {
  Upstream2[Upstream2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  Upstream2[Upstream2["SERVER"] = 1] = "SERVER";
  Upstream2[Upstream2["CLIENT"] = 2] = "CLIENT";
  return Upstream2;
})(Upstream || {});
class Oneof$Type extends MessageType {
  constructor() {
    super("danielvladco.protobuf.graphql.Oneof", [
      {
        no: 4,
        name: "ignore",
        kind: "scalar",
        opt: true,
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 5,
        name: "name",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional bool ignore */
        4:
          message.ignore = reader.bool();
          break;
        case /* optional string name */
        5:
          message.name = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.ignore !== void 0)
      writer.tag(4, WireType.Varint).bool(message.ignore);
    if (message.name !== void 0)
      writer.tag(5, WireType.LengthDelimited).string(message.name);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Oneof = new Oneof$Type();
class Field$Type extends MessageType {
  constructor() {
    super("danielvladco.protobuf.graphql.Field", [
      {
        no: 1,
        name: "required",
        kind: "scalar",
        opt: true,
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 2,
        name: "params",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "dirs",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "ignore",
        kind: "scalar",
        opt: true,
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 5,
        name: "name",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional bool required */
        1:
          message.required = reader.bool();
          break;
        case /* optional string params */
        2:
          message.params = reader.string();
          break;
        case /* optional string dirs */
        3:
          message.dirs = reader.string();
          break;
        case /* optional bool ignore */
        4:
          message.ignore = reader.bool();
          break;
        case /* optional string name */
        5:
          message.name = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.required !== void 0)
      writer.tag(1, WireType.Varint).bool(message.required);
    if (message.params !== void 0)
      writer.tag(2, WireType.LengthDelimited).string(message.params);
    if (message.dirs !== void 0)
      writer.tag(3, WireType.LengthDelimited).string(message.dirs);
    if (message.ignore !== void 0)
      writer.tag(4, WireType.Varint).bool(message.ignore);
    if (message.name !== void 0)
      writer.tag(5, WireType.LengthDelimited).string(message.name);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Field = new Field$Type();
class Rpc$Type extends MessageType {
  constructor() {
    super("danielvladco.protobuf.graphql.Rpc", [
      { no: 1, name: "type", kind: "enum", opt: true, T: () => ["danielvladco.protobuf.graphql.Type", Type] },
      {
        no: 2,
        name: "ignore",
        kind: "scalar",
        opt: true,
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 3,
        name: "name",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional danielvladco.protobuf.graphql.Type type */
        1:
          message.type = reader.int32();
          break;
        case /* optional bool ignore */
        2:
          message.ignore = reader.bool();
          break;
        case /* optional string name */
        3:
          message.name = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.type !== void 0)
      writer.tag(1, WireType.Varint).int32(message.type);
    if (message.ignore !== void 0)
      writer.tag(2, WireType.Varint).bool(message.ignore);
    if (message.name !== void 0)
      writer.tag(3, WireType.LengthDelimited).string(message.name);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Rpc = new Rpc$Type();
class Svc$Type extends MessageType {
  constructor() {
    super("danielvladco.protobuf.graphql.Svc", [
      { no: 1, name: "type", kind: "enum", opt: true, T: () => ["danielvladco.protobuf.graphql.Type", Type] },
      {
        no: 2,
        name: "ignore",
        kind: "scalar",
        opt: true,
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 3,
        name: "name",
        kind: "scalar",
        opt: true,
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "upstream", kind: "enum", opt: true, T: () => ["danielvladco.protobuf.graphql.Upstream", Upstream, "UPSTREAM_"] }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* optional danielvladco.protobuf.graphql.Type type */
        1:
          message.type = reader.int32();
          break;
        case /* optional bool ignore */
        2:
          message.ignore = reader.bool();
          break;
        case /* optional string name */
        3:
          message.name = reader.string();
          break;
        case /* optional danielvladco.protobuf.graphql.Upstream upstream */
        4:
          message.upstream = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.type !== void 0)
      writer.tag(1, WireType.Varint).int32(message.type);
    if (message.ignore !== void 0)
      writer.tag(2, WireType.Varint).bool(message.ignore);
    if (message.name !== void 0)
      writer.tag(3, WireType.LengthDelimited).string(message.name);
    if (message.upstream !== void 0)
      writer.tag(4, WireType.Varint).int32(message.upstream);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Svc = new Svc$Type();
export {
  Field,
  Oneof,
  Rpc,
  Svc,
  Type,
  Upstream
};
