// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/mint/v1beta1/mint_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Minter$Type extends MessageType {
  constructor() {
    super("cosmos.mint.v1beta1.Minter", [
      { no: 1, name: "inflation", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 2, name: "annual_provisions", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.inflation = "";
    message.annualProvisions = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string inflation */
        1:
          message.inflation = reader.string();
          break;
        case /* string annual_provisions */
        2:
          message.annualProvisions = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.inflation !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.inflation);
    if (message.annualProvisions !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.annualProvisions);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Minter = new Minter$Type();
class Params$Type extends MessageType {
  constructor() {
    super("cosmos.mint.v1beta1.Params", [
      {
        no: 1,
        name: "mint_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "inflation_rate_change", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } },
      { no: 3, name: "inflation_max", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } },
      { no: 4, name: "inflation_min", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } },
      { no: 5, name: "goal_bonded", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } },
      {
        no: 6,
        name: "blocks_per_year",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "amino.name": "cosmos-sdk/x/mint/Params" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.mintDenom = "";
    message.inflationRateChange = "";
    message.inflationMax = "";
    message.inflationMin = "";
    message.goalBonded = "";
    message.blocksPerYear = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string mint_denom */
        1:
          message.mintDenom = reader.string();
          break;
        case /* string inflation_rate_change */
        2:
          message.inflationRateChange = reader.string();
          break;
        case /* string inflation_max */
        3:
          message.inflationMax = reader.string();
          break;
        case /* string inflation_min */
        4:
          message.inflationMin = reader.string();
          break;
        case /* string goal_bonded */
        5:
          message.goalBonded = reader.string();
          break;
        case /* uint64 blocks_per_year */
        6:
          message.blocksPerYear = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.mintDenom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.mintDenom);
    if (message.inflationRateChange !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.inflationRateChange);
    if (message.inflationMax !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.inflationMax);
    if (message.inflationMin !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.inflationMin);
    if (message.goalBonded !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.goalBonded);
    if (message.blocksPerYear !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.blocksPerYear);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
export {
  Minter,
  Params
};
