// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/mint/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Params returns the total set of minting parameters.
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Inflation returns the current minting inflation value.
   *
   * @generated from protobuf rpc: Inflation
   */
  inflation(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * AnnualProvisions current minting annual provisions value.
   *
   * @generated from protobuf rpc: AnnualProvisions
   */
  annualProvisions(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
