// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/crypto/hd/v1/hd_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class BIP44Params$Type extends MessageType {
  constructor() {
    super("cosmos.crypto.hd.v1.BIP44Params", [
      {
        no: 1,
        name: "purpose",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 2,
        name: "coin_type",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "account",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 4,
        name: "change",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 5,
        name: "address_index",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ], { "gogoproto.goproto_stringer": false, "amino.name": "crypto/keys/hd/BIP44Params" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.purpose = 0;
    message.coinType = 0;
    message.account = 0;
    message.change = false;
    message.addressIndex = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 purpose */
        1:
          message.purpose = reader.uint32();
          break;
        case /* uint32 coin_type */
        2:
          message.coinType = reader.uint32();
          break;
        case /* uint32 account */
        3:
          message.account = reader.uint32();
          break;
        case /* bool change */
        4:
          message.change = reader.bool();
          break;
        case /* uint32 address_index */
        5:
          message.addressIndex = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.purpose !== 0)
      writer.tag(1, WireType.Varint).uint32(message.purpose);
    if (message.coinType !== 0)
      writer.tag(2, WireType.Varint).uint32(message.coinType);
    if (message.account !== 0)
      writer.tag(3, WireType.Varint).uint32(message.account);
    if (message.change !== false)
      writer.tag(4, WireType.Varint).bool(message.change);
    if (message.addressIndex !== 0)
      writer.tag(5, WireType.Varint).uint32(message.addressIndex);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BIP44Params = new BIP44Params$Type();
export {
  BIP44Params
};
