// @injectivelabs/core-proto-ts-v2/esm/generated/injective/auction/v1beta1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { LastAuctionResult } from "./auction_pb.js";
import { Bid } from "./auction_pb.js";
import { Params } from "./auction_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("injective.auction.v1beta1.GenesisState", [
      { no: 1, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "auction_round",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "highest_bid", kind: "message", T: () => Bid },
      {
        no: 4,
        name: "auction_ending_timestamp",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 5, name: "last_auction_result", kind: "message", T: () => LastAuctionResult }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.auctionRound = 0n;
    message.auctionEndingTimestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.auction.v1beta1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        case /* uint64 auction_round */
        2:
          message.auctionRound = reader.uint64().toBigInt();
          break;
        case /* injective.auction.v1beta1.Bid highest_bid */
        3:
          message.highestBid = Bid.internalBinaryRead(reader, reader.uint32(), options, message.highestBid);
          break;
        case /* int64 auction_ending_timestamp */
        4:
          message.auctionEndingTimestamp = reader.int64().toBigInt();
          break;
        case /* injective.auction.v1beta1.LastAuctionResult last_auction_result */
        5:
          message.lastAuctionResult = LastAuctionResult.internalBinaryRead(reader, reader.uint32(), options, message.lastAuctionResult);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.auctionRound !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.auctionRound);
    if (message.highestBid)
      Bid.internalBinaryWrite(message.highestBid, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.auctionEndingTimestamp !== 0n)
      writer.tag(4, WireType.Varint).int64(message.auctionEndingTimestamp);
    if (message.lastAuctionResult)
      LastAuctionResult.internalBinaryWrite(message.lastAuctionResult, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
export {
  GenesisState
};
