// @injectivelabs/core-proto-ts-v2/esm/generated/injective/auction/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Retrieves auction params
   *
   * @generated from protobuf rpc: AuctionParams
   */
  auctionParams(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves current auction basket with current highest bid and bidder
   *
   * @generated from protobuf rpc: CurrentAuctionBasket
   */
  currentAuctionBasket(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the entire auction module's state
   *
   * @generated from protobuf rpc: AuctionModuleState
   */
  auctionModuleState(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: LastAuctionResult
   */
  lastAuctionResult(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
