// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/lightclients/tendermint/v1/tendermint_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ValidatorSet } from "../../../../cometbft/types/v1/validator_pb.js";
import { SignedHeader } from "../../../../cometbft/types/v1/types_pb.js";
import { MerkleRoot } from "../../../core/commitment/v1/commitment_pb.js";
import { Timestamp } from "../../../../google/protobuf/timestamp_pb.js";
import { ProofSpec } from "../../../../cosmos/ics23/v1/proofs_pb.js";
import { Height } from "../../../core/client/v1/client_pb.js";
import { Duration } from "../../../../google/protobuf/duration_pb.js";
class ClientState$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.tendermint.v1.ClientState", [
      {
        no: 1,
        name: "chain_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "trust_level", kind: "message", T: () => Fraction, options: { "gogoproto.nullable": false } },
      { no: 3, name: "trusting_period", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true } },
      { no: 4, name: "unbonding_period", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true } },
      { no: 5, name: "max_clock_drift", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true } },
      { no: 6, name: "frozen_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      { no: 7, name: "latest_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      { no: 8, name: "proof_specs", kind: "message", repeat: 2, T: () => ProofSpec },
      {
        no: 9,
        name: "upgrade_path",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "allow_update_after_expiry",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 11,
        name: "allow_update_after_misbehaviour",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chainId = "";
    message.proofSpecs = [];
    message.upgradePath = [];
    message.allowUpdateAfterExpiry = false;
    message.allowUpdateAfterMisbehaviour = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string chain_id */
        1:
          message.chainId = reader.string();
          break;
        case /* ibc.lightclients.tendermint.v1.Fraction trust_level */
        2:
          message.trustLevel = Fraction.internalBinaryRead(reader, reader.uint32(), options, message.trustLevel);
          break;
        case /* google.protobuf.Duration trusting_period */
        3:
          message.trustingPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.trustingPeriod);
          break;
        case /* google.protobuf.Duration unbonding_period */
        4:
          message.unbondingPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.unbondingPeriod);
          break;
        case /* google.protobuf.Duration max_clock_drift */
        5:
          message.maxClockDrift = Duration.internalBinaryRead(reader, reader.uint32(), options, message.maxClockDrift);
          break;
        case /* ibc.core.client.v1.Height frozen_height */
        6:
          message.frozenHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.frozenHeight);
          break;
        case /* ibc.core.client.v1.Height latest_height */
        7:
          message.latestHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.latestHeight);
          break;
        case /* repeated cosmos.ics23.v1.ProofSpec proof_specs */
        8:
          message.proofSpecs.push(ProofSpec.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated string upgrade_path */
        9:
          message.upgradePath.push(reader.string());
          break;
        case /* bool allow_update_after_expiry = 10 [deprecated = true] */
        10:
          message.allowUpdateAfterExpiry = reader.bool();
          break;
        case /* bool allow_update_after_misbehaviour = 11 [deprecated = true] */
        11:
          message.allowUpdateAfterMisbehaviour = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.chainId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.chainId);
    if (message.trustLevel)
      Fraction.internalBinaryWrite(message.trustLevel, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.trustingPeriod)
      Duration.internalBinaryWrite(message.trustingPeriod, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.unbondingPeriod)
      Duration.internalBinaryWrite(message.unbondingPeriod, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.maxClockDrift)
      Duration.internalBinaryWrite(message.maxClockDrift, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.frozenHeight)
      Height.internalBinaryWrite(message.frozenHeight, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.latestHeight)
      Height.internalBinaryWrite(message.latestHeight, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.proofSpecs.length; i++)
      ProofSpec.internalBinaryWrite(message.proofSpecs[i], writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.upgradePath.length; i++)
      writer.tag(9, WireType.LengthDelimited).string(message.upgradePath[i]);
    if (message.allowUpdateAfterExpiry !== false)
      writer.tag(10, WireType.Varint).bool(message.allowUpdateAfterExpiry);
    if (message.allowUpdateAfterMisbehaviour !== false)
      writer.tag(11, WireType.Varint).bool(message.allowUpdateAfterMisbehaviour);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ClientState = new ClientState$Type();
class ConsensusState$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.tendermint.v1.ConsensusState", [
      { no: 1, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      { no: 2, name: "root", kind: "message", T: () => MerkleRoot, options: { "gogoproto.nullable": false } },
      { no: 3, name: "next_validators_hash", kind: "scalar", T: 12, options: { "gogoproto.casttype": "github.com/cometbft/cometbft/libs/bytes.HexBytes" } }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nextValidatorsHash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Timestamp timestamp */
        1:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        case /* ibc.core.commitment.v1.MerkleRoot root */
        2:
          message.root = MerkleRoot.internalBinaryRead(reader, reader.uint32(), options, message.root);
          break;
        case /* bytes next_validators_hash */
        3:
          message.nextValidatorsHash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.root)
      MerkleRoot.internalBinaryWrite(message.root, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.nextValidatorsHash.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.nextValidatorsHash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConsensusState = new ConsensusState$Type();
class Misbehaviour$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.tendermint.v1.Misbehaviour", [
      {
        no: 1,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "header_1", kind: "message", T: () => Header, options: { "gogoproto.customname": "Header1" } },
      { no: 3, name: "header_2", kind: "message", T: () => Header, options: { "gogoproto.customname": "Header2" } }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clientId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string client_id = 1 [deprecated = true] */
        1:
          message.clientId = reader.string();
          break;
        case /* ibc.lightclients.tendermint.v1.Header header_1 */
        2:
          message.header1 = Header.internalBinaryRead(reader, reader.uint32(), options, message.header1);
          break;
        case /* ibc.lightclients.tendermint.v1.Header header_2 */
        3:
          message.header2 = Header.internalBinaryRead(reader, reader.uint32(), options, message.header2);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.clientId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.clientId);
    if (message.header1)
      Header.internalBinaryWrite(message.header1, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.header2)
      Header.internalBinaryWrite(message.header2, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Misbehaviour = new Misbehaviour$Type();
class Header$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.tendermint.v1.Header", [
      { no: 1, name: "signed_header", kind: "message", T: () => SignedHeader, options: { "gogoproto.embed": true } },
      { no: 2, name: "validator_set", kind: "message", T: () => ValidatorSet },
      { no: 3, name: "trusted_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      { no: 4, name: "trusted_validators", kind: "message", T: () => ValidatorSet }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.SignedHeader signed_header */
        1:
          message.signedHeader = SignedHeader.internalBinaryRead(reader, reader.uint32(), options, message.signedHeader);
          break;
        case /* cometbft.types.v1.ValidatorSet validator_set */
        2:
          message.validatorSet = ValidatorSet.internalBinaryRead(reader, reader.uint32(), options, message.validatorSet);
          break;
        case /* ibc.core.client.v1.Height trusted_height */
        3:
          message.trustedHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.trustedHeight);
          break;
        case /* cometbft.types.v1.ValidatorSet trusted_validators */
        4:
          message.trustedValidators = ValidatorSet.internalBinaryRead(reader, reader.uint32(), options, message.trustedValidators);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signedHeader)
      SignedHeader.internalBinaryWrite(message.signedHeader, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.validatorSet)
      ValidatorSet.internalBinaryWrite(message.validatorSet, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.trustedHeight)
      Height.internalBinaryWrite(message.trustedHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.trustedValidators)
      ValidatorSet.internalBinaryWrite(message.trustedValidators, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Header = new Header$Type();
class Fraction$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.tendermint.v1.Fraction", [
      {
        no: 1,
        name: "numerator",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "denominator",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.numerator = 0n;
    message.denominator = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 numerator */
        1:
          message.numerator = reader.uint64().toBigInt();
          break;
        case /* uint64 denominator */
        2:
          message.denominator = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.numerator !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.numerator);
    if (message.denominator !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.denominator);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Fraction = new Fraction$Type();
export {
  ClientState,
  ConsensusState,
  Fraction,
  Header,
  Misbehaviour
};
