// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/fee/v1/ack_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class IncentivizedAcknowledgement$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.IncentivizedAcknowledgement", [
      {
        no: 1,
        name: "app_acknowledgement",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "forward_relayer_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "underlying_app_success",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.appAcknowledgement = new Uint8Array(0);
    message.forwardRelayerAddress = "";
    message.underlyingAppSuccess = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes app_acknowledgement */
        1:
          message.appAcknowledgement = reader.bytes();
          break;
        case /* string forward_relayer_address */
        2:
          message.forwardRelayerAddress = reader.string();
          break;
        case /* bool underlying_app_success */
        3:
          message.underlyingAppSuccess = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.appAcknowledgement.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.appAcknowledgement);
    if (message.forwardRelayerAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.forwardRelayerAddress);
    if (message.underlyingAppSuccess !== false)
      writer.tag(3, WireType.Varint).bool(message.underlyingAppSuccess);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const IncentivizedAcknowledgement = new IncentivizedAcknowledgement$Type();
export {
  IncentivizedAcknowledgement
};
