// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/fee/v1/fee_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { PacketId } from "../../../core/channel/v1/channel_pb.js";
import { Coin } from "../../../../cosmos/base/v1beta1/coin_pb.js";
class Fee$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.Fee", [
      { no: 1, name: "recv_fee", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins" } },
      { no: 2, name: "ack_fee", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins" } },
      { no: 3, name: "timeout_fee", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.recvFee = [];
    message.ackFee = [];
    message.timeoutFee = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin recv_fee */
        1:
          message.recvFee.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.base.v1beta1.Coin ack_fee */
        2:
          message.ackFee.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.base.v1beta1.Coin timeout_fee */
        3:
          message.timeoutFee.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.recvFee.length; i++)
      Coin.internalBinaryWrite(message.recvFee[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.ackFee.length; i++)
      Coin.internalBinaryWrite(message.ackFee[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.timeoutFee.length; i++)
      Coin.internalBinaryWrite(message.timeoutFee[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Fee = new Fee$Type();
class PacketFee$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.PacketFee", [
      { no: 1, name: "fee", kind: "message", T: () => Fee, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "refund_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "relayers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "cosmos.msg.v1.signer": ["refund_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.refundAddress = "";
    message.relayers = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.applications.fee.v1.Fee fee */
        1:
          message.fee = Fee.internalBinaryRead(reader, reader.uint32(), options, message.fee);
          break;
        case /* string refund_address */
        2:
          message.refundAddress = reader.string();
          break;
        case /* repeated string relayers */
        3:
          message.relayers.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fee)
      Fee.internalBinaryWrite(message.fee, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.refundAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.refundAddress);
    for (let i = 0; i < message.relayers.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.relayers[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketFee = new PacketFee$Type();
class PacketFees$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.PacketFees", [
      { no: 1, name: "packet_fees", kind: "message", repeat: 2, T: () => PacketFee, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.packetFees = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.applications.fee.v1.PacketFee packet_fees */
        1:
          message.packetFees.push(PacketFee.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.packetFees.length; i++)
      PacketFee.internalBinaryWrite(message.packetFees[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketFees = new PacketFees$Type();
class IdentifiedPacketFees$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.IdentifiedPacketFees", [
      { no: 1, name: "packet_id", kind: "message", T: () => PacketId, options: { "gogoproto.nullable": false } },
      { no: 2, name: "packet_fees", kind: "message", repeat: 2, T: () => PacketFee, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.packetFees = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.PacketId packet_id */
        1:
          message.packetId = PacketId.internalBinaryRead(reader, reader.uint32(), options, message.packetId);
          break;
        case /* repeated ibc.applications.fee.v1.PacketFee packet_fees */
        2:
          message.packetFees.push(PacketFee.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packetId)
      PacketId.internalBinaryWrite(message.packetId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.packetFees.length; i++)
      PacketFee.internalBinaryWrite(message.packetFees[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const IdentifiedPacketFees = new IdentifiedPacketFees$Type();
export {
  Fee,
  IdentifiedPacketFees,
  PacketFee,
  PacketFees
};
