// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/fee/v1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { PacketId } from "../../../core/channel/v1/channel_pb.js";
import { IdentifiedPacketFees } from "./fee_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.GenesisState", [
      { no: 1, name: "identified_fees", kind: "message", repeat: 2, T: () => IdentifiedPacketFees, options: { "gogoproto.nullable": false } },
      { no: 2, name: "fee_enabled_channels", kind: "message", repeat: 2, T: () => FeeEnabledChannel, options: { "gogoproto.nullable": false } },
      { no: 3, name: "registered_payees", kind: "message", repeat: 2, T: () => RegisteredPayee, options: { "gogoproto.nullable": false } },
      { no: 4, name: "registered_counterparty_payees", kind: "message", repeat: 2, T: () => RegisteredCounterpartyPayee, options: { "gogoproto.nullable": false } },
      { no: 5, name: "forward_relayers", kind: "message", repeat: 2, T: () => ForwardRelayerAddress, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.identifiedFees = [];
    message.feeEnabledChannels = [];
    message.registeredPayees = [];
    message.registeredCounterpartyPayees = [];
    message.forwardRelayers = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.applications.fee.v1.IdentifiedPacketFees identified_fees */
        1:
          message.identifiedFees.push(IdentifiedPacketFees.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.applications.fee.v1.FeeEnabledChannel fee_enabled_channels */
        2:
          message.feeEnabledChannels.push(FeeEnabledChannel.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.applications.fee.v1.RegisteredPayee registered_payees */
        3:
          message.registeredPayees.push(RegisteredPayee.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.applications.fee.v1.RegisteredCounterpartyPayee registered_counterparty_payees */
        4:
          message.registeredCounterpartyPayees.push(RegisteredCounterpartyPayee.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.applications.fee.v1.ForwardRelayerAddress forward_relayers */
        5:
          message.forwardRelayers.push(ForwardRelayerAddress.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.identifiedFees.length; i++)
      IdentifiedPacketFees.internalBinaryWrite(message.identifiedFees[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.feeEnabledChannels.length; i++)
      FeeEnabledChannel.internalBinaryWrite(message.feeEnabledChannels[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.registeredPayees.length; i++)
      RegisteredPayee.internalBinaryWrite(message.registeredPayees[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.registeredCounterpartyPayees.length; i++)
      RegisteredCounterpartyPayee.internalBinaryWrite(message.registeredCounterpartyPayees[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.forwardRelayers.length; i++)
      ForwardRelayerAddress.internalBinaryWrite(message.forwardRelayers[i], writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class FeeEnabledChannel$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.FeeEnabledChannel", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FeeEnabledChannel = new FeeEnabledChannel$Type();
class RegisteredPayee$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.RegisteredPayee", [
      {
        no: 1,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "relayer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "payee",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channelId = "";
    message.relayer = "";
    message.payee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string channel_id */
        1:
          message.channelId = reader.string();
          break;
        case /* string relayer */
        2:
          message.relayer = reader.string();
          break;
        case /* string payee */
        3:
          message.payee = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channelId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.channelId);
    if (message.relayer !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.relayer);
    if (message.payee !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.payee);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RegisteredPayee = new RegisteredPayee$Type();
class RegisteredCounterpartyPayee$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.RegisteredCounterpartyPayee", [
      {
        no: 1,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "relayer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "counterparty_payee",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channelId = "";
    message.relayer = "";
    message.counterpartyPayee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string channel_id */
        1:
          message.channelId = reader.string();
          break;
        case /* string relayer */
        2:
          message.relayer = reader.string();
          break;
        case /* string counterparty_payee */
        3:
          message.counterpartyPayee = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channelId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.channelId);
    if (message.relayer !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.relayer);
    if (message.counterpartyPayee !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.counterpartyPayee);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RegisteredCounterpartyPayee = new RegisteredCounterpartyPayee$Type();
class ForwardRelayerAddress$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.ForwardRelayerAddress", [
      {
        no: 1,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "packet_id", kind: "message", T: () => PacketId, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* ibc.core.channel.v1.PacketId packet_id */
        2:
          message.packetId = PacketId.internalBinaryRead(reader, reader.uint32(), options, message.packetId);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.packetId)
      PacketId.internalBinaryWrite(message.packetId, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ForwardRelayerAddress = new ForwardRelayerAddress$Type();
export {
  FeeEnabledChannel,
  ForwardRelayerAddress,
  GenesisState,
  RegisteredCounterpartyPayee,
  RegisteredPayee
};
