// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/fee/v1/query_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { FeeEnabledChannel } from "./genesis_pb.js";
import { Coin } from "../../../../cosmos/base/v1beta1/coin_pb.js";
import { PacketId } from "../../../core/channel/v1/channel_pb.js";
import { PageResponse } from "../../../../cosmos/base/query/v1beta1/pagination_pb.js";
import { IdentifiedPacketFees } from "./fee_pb.js";
import { PageRequest } from "../../../../cosmos/base/query/v1beta1/pagination_pb.js";
class QueryIncentivizedPacketsRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryIncentivizedPacketsRequest", [
      { no: 1, name: "pagination", kind: "message", T: () => PageRequest },
      {
        no: 2,
        name: "query_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.queryHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        1:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* uint64 query_height */
        2:
          message.queryHeight = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.queryHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.queryHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryIncentivizedPacketsRequest = new QueryIncentivizedPacketsRequest$Type();
class QueryIncentivizedPacketsResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryIncentivizedPacketsResponse", [
      { no: 1, name: "incentivized_packets", kind: "message", repeat: 2, T: () => IdentifiedPacketFees, options: { "gogoproto.nullable": false } },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.incentivizedPackets = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.applications.fee.v1.IdentifiedPacketFees incentivized_packets */
        1:
          message.incentivizedPackets.push(IdentifiedPacketFees.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.incentivizedPackets.length; i++)
      IdentifiedPacketFees.internalBinaryWrite(message.incentivizedPackets[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryIncentivizedPacketsResponse = new QueryIncentivizedPacketsResponse$Type();
class QueryIncentivizedPacketRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryIncentivizedPacketRequest", [
      { no: 1, name: "packet_id", kind: "message", T: () => PacketId, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "query_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.queryHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.PacketId packet_id */
        1:
          message.packetId = PacketId.internalBinaryRead(reader, reader.uint32(), options, message.packetId);
          break;
        case /* uint64 query_height */
        2:
          message.queryHeight = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packetId)
      PacketId.internalBinaryWrite(message.packetId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.queryHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.queryHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryIncentivizedPacketRequest = new QueryIncentivizedPacketRequest$Type();
class QueryIncentivizedPacketResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryIncentivizedPacketResponse", [
      { no: 1, name: "incentivized_packet", kind: "message", T: () => IdentifiedPacketFees, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.applications.fee.v1.IdentifiedPacketFees incentivized_packet */
        1:
          message.incentivizedPacket = IdentifiedPacketFees.internalBinaryRead(reader, reader.uint32(), options, message.incentivizedPacket);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.incentivizedPacket)
      IdentifiedPacketFees.internalBinaryWrite(message.incentivizedPacket, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryIncentivizedPacketResponse = new QueryIncentivizedPacketResponse$Type();
class QueryIncentivizedPacketsForChannelRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryIncentivizedPacketsForChannelRequest", [
      { no: 1, name: "pagination", kind: "message", T: () => PageRequest },
      {
        no: 2,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "query_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.queryHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        1:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* string port_id */
        2:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        3:
          message.channelId = reader.string();
          break;
        case /* uint64 query_height */
        4:
          message.queryHeight = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.portId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.channelId);
    if (message.queryHeight !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.queryHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryIncentivizedPacketsForChannelRequest = new QueryIncentivizedPacketsForChannelRequest$Type();
class QueryIncentivizedPacketsForChannelResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryIncentivizedPacketsForChannelResponse", [
      { no: 1, name: "incentivized_packets", kind: "message", repeat: 2, T: () => IdentifiedPacketFees },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.incentivizedPackets = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.applications.fee.v1.IdentifiedPacketFees incentivized_packets */
        1:
          message.incentivizedPackets.push(IdentifiedPacketFees.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.incentivizedPackets.length; i++)
      IdentifiedPacketFees.internalBinaryWrite(message.incentivizedPackets[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryIncentivizedPacketsForChannelResponse = new QueryIncentivizedPacketsForChannelResponse$Type();
class QueryTotalRecvFeesRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryTotalRecvFeesRequest", [
      { no: 1, name: "packet_id", kind: "message", T: () => PacketId, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.PacketId packet_id */
        1:
          message.packetId = PacketId.internalBinaryRead(reader, reader.uint32(), options, message.packetId);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packetId)
      PacketId.internalBinaryWrite(message.packetId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryTotalRecvFeesRequest = new QueryTotalRecvFeesRequest$Type();
class QueryTotalRecvFeesResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryTotalRecvFeesResponse", [
      { no: 1, name: "recv_fees", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.recvFees = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin recv_fees */
        1:
          message.recvFees.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.recvFees.length; i++)
      Coin.internalBinaryWrite(message.recvFees[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryTotalRecvFeesResponse = new QueryTotalRecvFeesResponse$Type();
class QueryTotalAckFeesRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryTotalAckFeesRequest", [
      { no: 1, name: "packet_id", kind: "message", T: () => PacketId, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.PacketId packet_id */
        1:
          message.packetId = PacketId.internalBinaryRead(reader, reader.uint32(), options, message.packetId);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packetId)
      PacketId.internalBinaryWrite(message.packetId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryTotalAckFeesRequest = new QueryTotalAckFeesRequest$Type();
class QueryTotalAckFeesResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryTotalAckFeesResponse", [
      { no: 1, name: "ack_fees", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.ackFees = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin ack_fees */
        1:
          message.ackFees.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.ackFees.length; i++)
      Coin.internalBinaryWrite(message.ackFees[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryTotalAckFeesResponse = new QueryTotalAckFeesResponse$Type();
class QueryTotalTimeoutFeesRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryTotalTimeoutFeesRequest", [
      { no: 1, name: "packet_id", kind: "message", T: () => PacketId, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.PacketId packet_id */
        1:
          message.packetId = PacketId.internalBinaryRead(reader, reader.uint32(), options, message.packetId);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packetId)
      PacketId.internalBinaryWrite(message.packetId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryTotalTimeoutFeesRequest = new QueryTotalTimeoutFeesRequest$Type();
class QueryTotalTimeoutFeesResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryTotalTimeoutFeesResponse", [
      { no: 1, name: "timeout_fees", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.timeoutFees = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin timeout_fees */
        1:
          message.timeoutFees.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.timeoutFees.length; i++)
      Coin.internalBinaryWrite(message.timeoutFees[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryTotalTimeoutFeesResponse = new QueryTotalTimeoutFeesResponse$Type();
class QueryPayeeRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryPayeeRequest", [
      {
        no: 1,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "relayer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channelId = "";
    message.relayer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string channel_id */
        1:
          message.channelId = reader.string();
          break;
        case /* string relayer */
        2:
          message.relayer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channelId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.channelId);
    if (message.relayer !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.relayer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPayeeRequest = new QueryPayeeRequest$Type();
class QueryPayeeResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryPayeeResponse", [
      {
        no: 1,
        name: "payee_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.payeeAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string payee_address */
        1:
          message.payeeAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.payeeAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.payeeAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPayeeResponse = new QueryPayeeResponse$Type();
class QueryCounterpartyPayeeRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryCounterpartyPayeeRequest", [
      {
        no: 1,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "relayer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channelId = "";
    message.relayer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string channel_id */
        1:
          message.channelId = reader.string();
          break;
        case /* string relayer */
        2:
          message.relayer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channelId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.channelId);
    if (message.relayer !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.relayer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryCounterpartyPayeeRequest = new QueryCounterpartyPayeeRequest$Type();
class QueryCounterpartyPayeeResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryCounterpartyPayeeResponse", [
      {
        no: 1,
        name: "counterparty_payee",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.counterpartyPayee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string counterparty_payee */
        1:
          message.counterpartyPayee = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.counterpartyPayee !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.counterpartyPayee);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryCounterpartyPayeeResponse = new QueryCounterpartyPayeeResponse$Type();
class QueryFeeEnabledChannelsRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryFeeEnabledChannelsRequest", [
      { no: 1, name: "pagination", kind: "message", T: () => PageRequest },
      {
        no: 2,
        name: "query_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.queryHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        1:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* uint64 query_height */
        2:
          message.queryHeight = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.queryHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.queryHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryFeeEnabledChannelsRequest = new QueryFeeEnabledChannelsRequest$Type();
class QueryFeeEnabledChannelsResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryFeeEnabledChannelsResponse", [
      { no: 1, name: "fee_enabled_channels", kind: "message", repeat: 2, T: () => FeeEnabledChannel, options: { "gogoproto.nullable": false } },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.feeEnabledChannels = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.applications.fee.v1.FeeEnabledChannel fee_enabled_channels */
        1:
          message.feeEnabledChannels.push(FeeEnabledChannel.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.feeEnabledChannels.length; i++)
      FeeEnabledChannel.internalBinaryWrite(message.feeEnabledChannels[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryFeeEnabledChannelsResponse = new QueryFeeEnabledChannelsResponse$Type();
class QueryFeeEnabledChannelRequest$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryFeeEnabledChannelRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryFeeEnabledChannelRequest = new QueryFeeEnabledChannelRequest$Type();
class QueryFeeEnabledChannelResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.QueryFeeEnabledChannelResponse", [
      {
        no: 1,
        name: "fee_enabled",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.feeEnabled = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bool fee_enabled */
        1:
          message.feeEnabled = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feeEnabled !== false)
      writer.tag(1, WireType.Varint).bool(message.feeEnabled);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryFeeEnabledChannelResponse = new QueryFeeEnabledChannelResponse$Type();
const Query = new ServiceType("ibc.applications.fee.v1.Query", [
  { name: "IncentivizedPackets", options: { "google.api.http": { get: "/ibc/apps/fee/v1/incentivized_packets" } }, I: QueryIncentivizedPacketsRequest, O: QueryIncentivizedPacketsResponse },
  { name: "IncentivizedPacket", options: { "google.api.http": { get: "/ibc/apps/fee/v1/channels/{packet_id.channel_id}/ports/{packet_id.port_id}/sequences/{packet_id.sequence}/incentivized_packet" } }, I: QueryIncentivizedPacketRequest, O: QueryIncentivizedPacketResponse },
  { name: "IncentivizedPacketsForChannel", options: { "google.api.http": { get: "/ibc/apps/fee/v1/channels/{channel_id}/ports/{port_id}/incentivized_packets" } }, I: QueryIncentivizedPacketsForChannelRequest, O: QueryIncentivizedPacketsForChannelResponse },
  { name: "TotalRecvFees", options: { "google.api.http": { get: "/ibc/apps/fee/v1/channels/{packet_id.channel_id}/ports/{packet_id.port_id}/sequences/{packet_id.sequence}/total_recv_fees" } }, I: QueryTotalRecvFeesRequest, O: QueryTotalRecvFeesResponse },
  { name: "TotalAckFees", options: { "google.api.http": { get: "/ibc/apps/fee/v1/channels/{packet_id.channel_id}/ports/{packet_id.port_id}/sequences/{packet_id.sequence}/total_ack_fees" } }, I: QueryTotalAckFeesRequest, O: QueryTotalAckFeesResponse },
  { name: "TotalTimeoutFees", options: { "google.api.http": { get: "/ibc/apps/fee/v1/channels/{packet_id.channel_id}/ports/{packet_id.port_id}/sequences/{packet_id.sequence}/total_timeout_fees" } }, I: QueryTotalTimeoutFeesRequest, O: QueryTotalTimeoutFeesResponse },
  { name: "Payee", options: { "google.api.http": { get: "/ibc/apps/fee/v1/channels/{channel_id}/relayers/{relayer}/payee" } }, I: QueryPayeeRequest, O: QueryPayeeResponse },
  { name: "CounterpartyPayee", options: { "google.api.http": { get: "/ibc/apps/fee/v1/channels/{channel_id}/relayers/{relayer}/counterparty_payee" } }, I: QueryCounterpartyPayeeRequest, O: QueryCounterpartyPayeeResponse },
  { name: "FeeEnabledChannels", options: { "google.api.http": { get: "/ibc/apps/fee/v1/fee_enabled" } }, I: QueryFeeEnabledChannelsRequest, O: QueryFeeEnabledChannelsResponse },
  { name: "FeeEnabledChannel", options: { "google.api.http": { get: "/ibc/apps/fee/v1/channels/{channel_id}/ports/{port_id}/fee_enabled" } }, I: QueryFeeEnabledChannelRequest, O: QueryFeeEnabledChannelResponse }
]);
export {
  Query,
  QueryCounterpartyPayeeRequest,
  QueryCounterpartyPayeeResponse,
  QueryFeeEnabledChannelRequest,
  QueryFeeEnabledChannelResponse,
  QueryFeeEnabledChannelsRequest,
  QueryFeeEnabledChannelsResponse,
  QueryIncentivizedPacketRequest,
  QueryIncentivizedPacketResponse,
  QueryIncentivizedPacketsForChannelRequest,
  QueryIncentivizedPacketsForChannelResponse,
  QueryIncentivizedPacketsRequest,
  QueryIncentivizedPacketsResponse,
  QueryPayeeRequest,
  QueryPayeeResponse,
  QueryTotalAckFeesRequest,
  QueryTotalAckFeesResponse,
  QueryTotalRecvFeesRequest,
  QueryTotalRecvFeesResponse,
  QueryTotalTimeoutFeesRequest,
  QueryTotalTimeoutFeesResponse
};
