// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/fee/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * IncentivizedPackets returns all incentivized packets and their associated fees
   *
   * @generated from protobuf rpc: IncentivizedPackets
   */
  incentivizedPackets(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * IncentivizedPacket returns all packet fees for a packet given its identifier
   *
   * @generated from protobuf rpc: IncentivizedPacket
   */
  incentivizedPacket(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets all incentivized packets for a specific channel
   *
   * @generated from protobuf rpc: IncentivizedPacketsForChannel
   */
  incentivizedPacketsForChannel(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TotalRecvFees returns the total receive fees for a packet given its identifier
   *
   * @generated from protobuf rpc: TotalRecvFees
   */
  totalRecvFees(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TotalAckFees returns the total acknowledgement fees for a packet given its identifier
   *
   * @generated from protobuf rpc: TotalAckFees
   */
  totalAckFees(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TotalTimeoutFees returns the total timeout fees for a packet given its identifier
   *
   * @generated from protobuf rpc: TotalTimeoutFees
   */
  totalTimeoutFees(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Payee returns the registered payee address for a specific channel given the relayer address
   *
   * @generated from protobuf rpc: Payee
   */
  payee(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * CounterpartyPayee returns the registered counterparty payee for forward relaying
   *
   * @generated from protobuf rpc: CounterpartyPayee
   */
  counterpartyPayee(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * FeeEnabledChannels returns a list of all fee enabled channels
   *
   * @generated from protobuf rpc: FeeEnabledChannels
   */
  feeEnabledChannels(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * FeeEnabledChannel returns true if the provided port and channel identifiers belong to a fee enabled channel
   *
   * @generated from protobuf rpc: FeeEnabledChannel
   */
  feeEnabledChannel(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
