// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/fee/v1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { PacketFee } from "./fee_pb.js";
import { PacketId } from "../../../core/channel/v1/channel_pb.js";
import { Fee } from "./fee_pb.js";
class MsgRegisterPayee$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.MsgRegisterPayee", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "relayer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "payee",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "amino.name": "cosmos-sdk/MsgRegisterPayee", "cosmos.msg.v1.signer": ["relayer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.relayer = "";
    message.payee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* string relayer */
        3:
          message.relayer = reader.string();
          break;
        case /* string payee */
        4:
          message.payee = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.relayer !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.relayer);
    if (message.payee !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.payee);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRegisterPayee = new MsgRegisterPayee$Type();
class MsgRegisterPayeeResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.MsgRegisterPayeeResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRegisterPayeeResponse = new MsgRegisterPayeeResponse$Type();
class MsgRegisterCounterpartyPayee$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.MsgRegisterCounterpartyPayee", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "relayer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "counterparty_payee",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "amino.name": "cosmos-sdk/MsgRegisterCounterpartyPayee", "cosmos.msg.v1.signer": ["relayer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.relayer = "";
    message.counterpartyPayee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* string relayer */
        3:
          message.relayer = reader.string();
          break;
        case /* string counterparty_payee */
        4:
          message.counterpartyPayee = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.relayer !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.relayer);
    if (message.counterpartyPayee !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.counterpartyPayee);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRegisterCounterpartyPayee = new MsgRegisterCounterpartyPayee$Type();
class MsgRegisterCounterpartyPayeeResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.MsgRegisterCounterpartyPayeeResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRegisterCounterpartyPayeeResponse = new MsgRegisterCounterpartyPayeeResponse$Type();
class MsgPayPacketFee$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.MsgPayPacketFee", [
      { no: 1, name: "fee", kind: "message", T: () => Fee, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 2,
        name: "source_port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "source_channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "relayers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "amino.name": "cosmos-sdk/MsgPayPacketFee", "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sourcePortId = "";
    message.sourceChannelId = "";
    message.signer = "";
    message.relayers = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.applications.fee.v1.Fee fee */
        1:
          message.fee = Fee.internalBinaryRead(reader, reader.uint32(), options, message.fee);
          break;
        case /* string source_port_id */
        2:
          message.sourcePortId = reader.string();
          break;
        case /* string source_channel_id */
        3:
          message.sourceChannelId = reader.string();
          break;
        case /* string signer */
        4:
          message.signer = reader.string();
          break;
        case /* repeated string relayers */
        5:
          message.relayers.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fee)
      Fee.internalBinaryWrite(message.fee, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.sourcePortId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.sourcePortId);
    if (message.sourceChannelId !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.sourceChannelId);
    if (message.signer !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.signer);
    for (let i = 0; i < message.relayers.length; i++)
      writer.tag(5, WireType.LengthDelimited).string(message.relayers[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPayPacketFee = new MsgPayPacketFee$Type();
class MsgPayPacketFeeResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.MsgPayPacketFeeResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPayPacketFeeResponse = new MsgPayPacketFeeResponse$Type();
class MsgPayPacketFeeAsync$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.MsgPayPacketFeeAsync", [
      { no: 1, name: "packet_id", kind: "message", T: () => PacketId, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "packet_fee", kind: "message", T: () => PacketFee, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "amino.name": "cosmos-sdk/MsgPayPacketFeeAsync", "cosmos.msg.v1.signer": ["packet_fee"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.PacketId packet_id */
        1:
          message.packetId = PacketId.internalBinaryRead(reader, reader.uint32(), options, message.packetId);
          break;
        case /* ibc.applications.fee.v1.PacketFee packet_fee */
        2:
          message.packetFee = PacketFee.internalBinaryRead(reader, reader.uint32(), options, message.packetFee);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packetId)
      PacketId.internalBinaryWrite(message.packetId, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.packetFee)
      PacketFee.internalBinaryWrite(message.packetFee, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPayPacketFeeAsync = new MsgPayPacketFeeAsync$Type();
class MsgPayPacketFeeAsyncResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.fee.v1.MsgPayPacketFeeAsyncResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPayPacketFeeAsyncResponse = new MsgPayPacketFeeAsyncResponse$Type();
const Msg = new ServiceType("ibc.applications.fee.v1.Msg", [
  { name: "RegisterPayee", options: {}, I: MsgRegisterPayee, O: MsgRegisterPayeeResponse },
  { name: "RegisterCounterpartyPayee", options: {}, I: MsgRegisterCounterpartyPayee, O: MsgRegisterCounterpartyPayeeResponse },
  { name: "PayPacketFee", options: {}, I: MsgPayPacketFee, O: MsgPayPacketFeeResponse },
  { name: "PayPacketFeeAsync", options: {}, I: MsgPayPacketFeeAsync, O: MsgPayPacketFeeAsyncResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgPayPacketFee,
  MsgPayPacketFeeAsync,
  MsgPayPacketFeeAsyncResponse,
  MsgPayPacketFeeResponse,
  MsgRegisterCounterpartyPayee,
  MsgRegisterCounterpartyPayeeResponse,
  MsgRegisterPayee,
  MsgRegisterPayeeResponse
};
