// @injectivelabs/indexer-proto-ts-v2/esm/generated/event_provider_api_pb.client.js
import { EventProviderAPI } from "./event_provider_api_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class EventProviderAPIClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = EventProviderAPI.typeName;
    this.methods = EventProviderAPI.methods;
    this.options = EventProviderAPI.options;
  }
  /**
   * Get latest block from event provider
   *
   * @generated from protobuf rpc: GetLatestHeight
   */
  getLatestHeight(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream latest block information
   *
   * @generated from protobuf rpc: StreamLatestHeight
   */
  streamLatestHeight(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Stream processed block events for selected backend
   *
   * @generated from protobuf rpc: StreamBlockEvents
   */
  streamBlockEvents(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Get processed block events for selected backend
   *
   * @generated from protobuf rpc: GetBlockEventsRPC
   */
  getBlockEventsRPC(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get custom processed block events for selected backend
   *
   * @generated from protobuf rpc: GetCustomEventsRPC
   */
  getCustomEventsRPC(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get raw block events for selected height
   *
   * @generated from protobuf rpc: GetABCIBlockEvents
   */
  getABCIBlockEvents(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get all raw block events for selected height
   *
   * @generated from protobuf rpc: GetABCIBlockEventsAtHeight
   */
  getABCIBlockEventsAtHeight(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  EventProviderAPIClient
};
