// @injectivelabs/indexer-proto-ts-v2/esm/generated/health_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class GetStatusRequest$Type extends MessageType {
  constructor() {
    super("api.v1.GetStatusRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetStatusRequest = new GetStatusRequest$Type();
class GetStatusResponse$Type extends MessageType {
  constructor() {
    super("api.v1.GetStatusResponse", [
      {
        no: 1,
        name: "s",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "errmsg",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "data", kind: "message", T: () => HealthStatus },
      {
        no: 4,
        name: "status",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.s = "";
    message.errmsg = "";
    message.status = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string s */
        1:
          message.s = reader.string();
          break;
        case /* string errmsg */
        2:
          message.errmsg = reader.string();
          break;
        case /* api.v1.HealthStatus data */
        3:
          message.data = HealthStatus.internalBinaryRead(reader, reader.uint32(), options, message.data);
          break;
        case /* string status */
        4:
          message.status = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.s !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.s);
    if (message.errmsg !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.errmsg);
    if (message.data)
      HealthStatus.internalBinaryWrite(message.data, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.status !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.status);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetStatusResponse = new GetStatusResponse$Type();
class HealthStatus$Type extends MessageType {
  constructor() {
    super("api.v1.HealthStatus", [
      {
        no: 1,
        name: "local_height",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 2,
        name: "local_timestamp",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 3,
        name: "horacle_height",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 4,
        name: "horacle_timestamp",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 5,
        name: "migration_last_version",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 6,
        name: "ep_height",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 7,
        name: "ep_timestamp",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.localHeight = 0;
    message.localTimestamp = 0;
    message.horacleHeight = 0;
    message.horacleTimestamp = 0;
    message.migrationLastVersion = 0;
    message.epHeight = 0;
    message.epTimestamp = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* sint32 local_height */
        1:
          message.localHeight = reader.sint32();
          break;
        case /* sint32 local_timestamp */
        2:
          message.localTimestamp = reader.sint32();
          break;
        case /* sint32 horacle_height */
        3:
          message.horacleHeight = reader.sint32();
          break;
        case /* sint32 horacle_timestamp */
        4:
          message.horacleTimestamp = reader.sint32();
          break;
        case /* sint32 migration_last_version */
        5:
          message.migrationLastVersion = reader.sint32();
          break;
        case /* sint32 ep_height */
        6:
          message.epHeight = reader.sint32();
          break;
        case /* sint32 ep_timestamp */
        7:
          message.epTimestamp = reader.sint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.localHeight !== 0)
      writer.tag(1, WireType.Varint).sint32(message.localHeight);
    if (message.localTimestamp !== 0)
      writer.tag(2, WireType.Varint).sint32(message.localTimestamp);
    if (message.horacleHeight !== 0)
      writer.tag(3, WireType.Varint).sint32(message.horacleHeight);
    if (message.horacleTimestamp !== 0)
      writer.tag(4, WireType.Varint).sint32(message.horacleTimestamp);
    if (message.migrationLastVersion !== 0)
      writer.tag(5, WireType.Varint).sint32(message.migrationLastVersion);
    if (message.epHeight !== 0)
      writer.tag(6, WireType.Varint).sint32(message.epHeight);
    if (message.epTimestamp !== 0)
      writer.tag(7, WireType.Varint).sint32(message.epTimestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const HealthStatus = new HealthStatus$Type();
const Health = new ServiceType("api.v1.Health", [
  { name: "GetStatus", options: {}, I: GetStatusRequest, O: GetStatusResponse }
]);
export {
  GetStatusRequest,
  GetStatusResponse,
  Health,
  HealthStatus
};
