// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_referral_rpc_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class GetReferrerDetailsRequest$Type extends MessageType {
  constructor() {
    super("injective_referral_rpc.GetReferrerDetailsRequest", [
      {
        no: 1,
        name: "referrer_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.referrerAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string referrer_address */
        1:
          message.referrerAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.referrerAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.referrerAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetReferrerDetailsRequest = new GetReferrerDetailsRequest$Type();
class GetReferrerDetailsResponse$Type extends MessageType {
  constructor() {
    super("injective_referral_rpc.GetReferrerDetailsResponse", [
      { no: 1, name: "invitees", kind: "message", repeat: 2, T: () => ReferralInvitee },
      {
        no: 2,
        name: "total_commission",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "total_trading_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "referrer_code",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.invitees = [];
    message.totalCommission = "";
    message.totalTradingVolume = "";
    message.referrerCode = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_referral_rpc.ReferralInvitee invitees */
        1:
          message.invitees.push(ReferralInvitee.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string total_commission */
        2:
          message.totalCommission = reader.string();
          break;
        case /* string total_trading_volume */
        3:
          message.totalTradingVolume = reader.string();
          break;
        case /* string referrer_code */
        4:
          message.referrerCode = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.invitees.length; i++)
      ReferralInvitee.internalBinaryWrite(message.invitees[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.totalCommission !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.totalCommission);
    if (message.totalTradingVolume !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.totalTradingVolume);
    if (message.referrerCode !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.referrerCode);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetReferrerDetailsResponse = new GetReferrerDetailsResponse$Type();
class ReferralInvitee$Type extends MessageType {
  constructor() {
    super("injective_referral_rpc.ReferralInvitee", [
      {
        no: 1,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "commission",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "trading_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "join_date",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.commission = "";
    message.tradingVolume = "";
    message.joinDate = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* string commission */
        2:
          message.commission = reader.string();
          break;
        case /* string trading_volume */
        3:
          message.tradingVolume = reader.string();
          break;
        case /* string join_date */
        4:
          message.joinDate = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.commission !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.commission);
    if (message.tradingVolume !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.tradingVolume);
    if (message.joinDate !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.joinDate);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ReferralInvitee = new ReferralInvitee$Type();
class GetInviteeDetailsRequest$Type extends MessageType {
  constructor() {
    super("injective_referral_rpc.GetInviteeDetailsRequest", [
      {
        no: 1,
        name: "invitee_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.inviteeAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string invitee_address */
        1:
          message.inviteeAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.inviteeAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.inviteeAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetInviteeDetailsRequest = new GetInviteeDetailsRequest$Type();
class GetInviteeDetailsResponse$Type extends MessageType {
  constructor() {
    super("injective_referral_rpc.GetInviteeDetailsResponse", [
      {
        no: 1,
        name: "referrer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "used_code",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "trading_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "joined_at",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "active",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.referrer = "";
    message.usedCode = "";
    message.tradingVolume = "";
    message.joinedAt = "";
    message.active = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string referrer */
        1:
          message.referrer = reader.string();
          break;
        case /* string used_code */
        2:
          message.usedCode = reader.string();
          break;
        case /* string trading_volume */
        3:
          message.tradingVolume = reader.string();
          break;
        case /* string joined_at */
        4:
          message.joinedAt = reader.string();
          break;
        case /* bool active */
        5:
          message.active = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.referrer !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.referrer);
    if (message.usedCode !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.usedCode);
    if (message.tradingVolume !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.tradingVolume);
    if (message.joinedAt !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.joinedAt);
    if (message.active !== false)
      writer.tag(5, WireType.Varint).bool(message.active);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetInviteeDetailsResponse = new GetInviteeDetailsResponse$Type();
class GetReferrerByCodeRequest$Type extends MessageType {
  constructor() {
    super("injective_referral_rpc.GetReferrerByCodeRequest", [
      {
        no: 1,
        name: "referral_code",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.referralCode = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string referral_code */
        1:
          message.referralCode = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.referralCode !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.referralCode);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetReferrerByCodeRequest = new GetReferrerByCodeRequest$Type();
class GetReferrerByCodeResponse$Type extends MessageType {
  constructor() {
    super("injective_referral_rpc.GetReferrerByCodeResponse", [
      {
        no: 1,
        name: "referrer_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.referrerAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string referrer_address */
        1:
          message.referrerAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.referrerAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.referrerAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetReferrerByCodeResponse = new GetReferrerByCodeResponse$Type();
const InjectiveReferralRPC = new ServiceType("injective_referral_rpc.InjectiveReferralRPC", [
  { name: "GetReferrerDetails", options: {}, I: GetReferrerDetailsRequest, O: GetReferrerDetailsResponse },
  { name: "GetInviteeDetails", options: {}, I: GetInviteeDetailsRequest, O: GetInviteeDetailsResponse },
  { name: "GetReferrerByCode", options: {}, I: GetReferrerByCodeRequest, O: GetReferrerByCodeResponse }
]);
export {
  GetInviteeDetailsRequest,
  GetInviteeDetailsResponse,
  GetReferrerByCodeRequest,
  GetReferrerByCodeResponse,
  GetReferrerDetailsRequest,
  GetReferrerDetailsResponse,
  InjectiveReferralRPC,
  ReferralInvitee
};
