// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_chart_rpc_pb.client.js
import { InjectiveChartRPC } from "./injective_chart_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveChartRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveChartRPC.typeName;
    this.methods = InjectiveChartRPC.methods;
    this.options = InjectiveChartRPC.options;
  }
  /**
   * Request for history bars of spotMarket for TradingView.
   *
   * @generated from protobuf rpc: SpotMarketHistory
   */
  spotMarketHistory(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Request for history bars of derivativeMarket for TradingView.
   *
   * @generated from protobuf rpc: DerivativeMarketHistory
   */
  derivativeMarketHistory(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets spot market summary for the latest interval (hour, day, month)
   *
   * @generated from protobuf rpc: SpotMarketSummary
   */
  spotMarketSummary(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets batch summary for all active markets, for the latest interval (hour,
   * day, month)
   *
   * @generated from protobuf rpc: AllSpotMarketSummary
   */
  allSpotMarketSummary(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets derivative market summary for the latest interval (hour, day, month)
   *
   * @generated from protobuf rpc: DerivativeMarketSummary
   */
  derivativeMarketSummary(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets batch summary for all active markets, for the latest interval (hour,
   * day, month)
   *
   * @generated from protobuf rpc: AllDerivativeMarketSummary
   */
  allDerivativeMarketSummary(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Request for cached markets history bars. Max age is 1h.
   *
   * @generated from protobuf rpc: MarketSnapshot
   */
  marketSnapshot(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveChartRPCClient
};
