// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_spot_exchange_rpc_pb.client.js
import { InjectiveSpotExchangeRPC } from "./injective_spot_exchange_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveSpotExchangeRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveSpotExchangeRPC.typeName;
    this.methods = InjectiveSpotExchangeRPC.methods;
    this.options = InjectiveSpotExchangeRPC.options;
  }
  /**
   * Get a list of Spot Markets
   *
   * @generated from protobuf rpc: Markets
   */
  markets(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get details of a single spot market
   *
   * @generated from protobuf rpc: Market
   */
  market(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream live updates of selected spot markets
   *
   * @generated from protobuf rpc: StreamMarkets
   */
  streamMarkets(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Orderbook of a Spot Market
   *
   * @generated from protobuf rpc: OrderbookV2
   */
  orderbookV2(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Orderbook of Spot Markets
   *
   * @generated from protobuf rpc: OrderbooksV2
   */
  orderbooksV2(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream live snapshot updates of selected spot market orderbook
   *
   * @generated from protobuf rpc: StreamOrderbookV2
   */
  streamOrderbookV2(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Stream live level updates of selected spot market orderbook
   *
   * @generated from protobuf rpc: StreamOrderbookUpdate
   */
  streamOrderbookUpdate(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Orders of a Spot Market
   *
   * @generated from protobuf rpc: Orders
   */
  orders(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream updates to individual orders of a Spot Market
   *
   * @generated from protobuf rpc: StreamOrders
   */
  streamOrders(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Trades of a Spot Market
   *
   * @generated from protobuf rpc: Trades
   */
  trades(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream newly executed trades from Spot Market
   *
   * @generated from protobuf rpc: StreamTrades
   */
  streamTrades(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Trades of a Spot Market
   *
   * @generated from protobuf rpc: TradesV2
   */
  tradesV2(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream newly executed trades from Spot Market
   *
   * @generated from protobuf rpc: StreamTradesV2
   */
  streamTradesV2(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * List orders posted from this subaccount
   *
   * @generated from protobuf rpc: SubaccountOrdersList
   */
  subaccountOrdersList(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List trades executed by this subaccount
   *
   * @generated from protobuf rpc: SubaccountTradesList
   */
  subaccountTradesList(input, options) {
    const method = this.methods[14], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Lists history orders posted from this subaccount
   *
   * @generated from protobuf rpc: OrdersHistory
   */
  ordersHistory(input, options) {
    const method = this.methods[15], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream updates to historical orders of a spot Market
   *
   * @generated from protobuf rpc: StreamOrdersHistory
   */
  streamOrdersHistory(input, options) {
    const method = this.methods[16], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Get historical atomic swaps
   *
   * @generated from protobuf rpc: AtomicSwapHistory
   */
  atomicSwapHistory(input, options) {
    const method = this.methods[17], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveSpotExchangeRPCClient
};
