// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_meta_rpc_pb.client.js
import { InjectiveMetaRPC } from "./injective_meta_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveMetaRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveMetaRPC.typeName;
    this.methods = InjectiveMetaRPC.methods;
    this.options = InjectiveMetaRPC.options;
  }
  /**
   * Endpoint for checking server health.
   *
   * @generated from protobuf rpc: Ping
   */
  ping(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Returns injective-exchange version.
   *
   * @generated from protobuf rpc: Version
   */
  version(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets connection info
   *
   * @generated from protobuf rpc: Info
   */
  info(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream keepalive, if server exits, a shutdown event will be sent over this
   * channel.
   *
   * @generated from protobuf rpc: StreamKeepalive
   */
  streamKeepalive(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Get tokens metadata. Can be filtered by denom
   *
   * @generated from protobuf rpc: TokenMetadata
   */
  tokenMetadata(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveMetaRPCClient
};
