// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_insurance_rpc_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class FundsRequest$Type extends MessageType {
  constructor() {
    super("injective_insurance_rpc.FundsRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FundsRequest = new FundsRequest$Type();
class FundsResponse$Type extends MessageType {
  constructor() {
    super("injective_insurance_rpc.FundsResponse", [
      { no: 1, name: "funds", kind: "message", repeat: 2, T: () => InsuranceFund }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.funds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_insurance_rpc.InsuranceFund funds */
        1:
          message.funds.push(InsuranceFund.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.funds.length; i++)
      InsuranceFund.internalBinaryWrite(message.funds[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FundsResponse = new FundsResponse$Type();
class InsuranceFund$Type extends MessageType {
  constructor() {
    super("injective_insurance_rpc.InsuranceFund", [
      {
        no: 1,
        name: "market_ticker",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "deposit_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "pool_token_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "redemption_notice_period_duration",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "balance",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "total_share",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "oracle_base",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "oracle_quote",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "oracle_type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "expiry",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 12, name: "deposit_token_meta", kind: "message", T: () => TokenMeta }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.marketTicker = "";
    message.marketId = "";
    message.depositDenom = "";
    message.poolTokenDenom = "";
    message.redemptionNoticePeriodDuration = 0n;
    message.balance = "";
    message.totalShare = "";
    message.oracleBase = "";
    message.oracleQuote = "";
    message.oracleType = "";
    message.expiry = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string market_ticker */
        1:
          message.marketTicker = reader.string();
          break;
        case /* string market_id */
        2:
          message.marketId = reader.string();
          break;
        case /* string deposit_denom */
        3:
          message.depositDenom = reader.string();
          break;
        case /* string pool_token_denom */
        4:
          message.poolTokenDenom = reader.string();
          break;
        case /* sint64 redemption_notice_period_duration */
        5:
          message.redemptionNoticePeriodDuration = reader.sint64().toBigInt();
          break;
        case /* string balance */
        6:
          message.balance = reader.string();
          break;
        case /* string total_share */
        7:
          message.totalShare = reader.string();
          break;
        case /* string oracle_base */
        8:
          message.oracleBase = reader.string();
          break;
        case /* string oracle_quote */
        9:
          message.oracleQuote = reader.string();
          break;
        case /* string oracle_type */
        10:
          message.oracleType = reader.string();
          break;
        case /* sint64 expiry */
        11:
          message.expiry = reader.sint64().toBigInt();
          break;
        case /* injective_insurance_rpc.TokenMeta deposit_token_meta */
        12:
          message.depositTokenMeta = TokenMeta.internalBinaryRead(reader, reader.uint32(), options, message.depositTokenMeta);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.marketTicker !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.marketTicker);
    if (message.marketId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    if (message.depositDenom !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.depositDenom);
    if (message.poolTokenDenom !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.poolTokenDenom);
    if (message.redemptionNoticePeriodDuration !== 0n)
      writer.tag(5, WireType.Varint).sint64(message.redemptionNoticePeriodDuration);
    if (message.balance !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.balance);
    if (message.totalShare !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.totalShare);
    if (message.oracleBase !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.oracleBase);
    if (message.oracleQuote !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.oracleQuote);
    if (message.oracleType !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.oracleType);
    if (message.expiry !== 0n)
      writer.tag(11, WireType.Varint).sint64(message.expiry);
    if (message.depositTokenMeta)
      TokenMeta.internalBinaryWrite(message.depositTokenMeta, writer.tag(12, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const InsuranceFund = new InsuranceFund$Type();
class TokenMeta$Type extends MessageType {
  constructor() {
    super("injective_insurance_rpc.TokenMeta", [
      {
        no: 1,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "symbol",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "logo",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "decimals",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 6,
        name: "updated_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.name = "";
    message.address = "";
    message.symbol = "";
    message.logo = "";
    message.decimals = 0;
    message.updatedAt = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string name */
        1:
          message.name = reader.string();
          break;
        case /* string address */
        2:
          message.address = reader.string();
          break;
        case /* string symbol */
        3:
          message.symbol = reader.string();
          break;
        case /* string logo */
        4:
          message.logo = reader.string();
          break;
        case /* sint32 decimals */
        5:
          message.decimals = reader.sint32();
          break;
        case /* sint64 updated_at */
        6:
          message.updatedAt = reader.sint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.name !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.name);
    if (message.address !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.address);
    if (message.symbol !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.symbol);
    if (message.logo !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.logo);
    if (message.decimals !== 0)
      writer.tag(5, WireType.Varint).sint32(message.decimals);
    if (message.updatedAt !== 0n)
      writer.tag(6, WireType.Varint).sint64(message.updatedAt);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TokenMeta = new TokenMeta$Type();
class FundRequest$Type extends MessageType {
  constructor() {
    super("injective_insurance_rpc.FundRequest", [
      {
        no: 1,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.denom = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string denom */
        1:
          message.denom = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.denom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.denom);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FundRequest = new FundRequest$Type();
class FundResponse$Type extends MessageType {
  constructor() {
    super("injective_insurance_rpc.FundResponse", [
      { no: 1, name: "fund", kind: "message", T: () => InsuranceFund }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective_insurance_rpc.InsuranceFund fund */
        1:
          message.fund = InsuranceFund.internalBinaryRead(reader, reader.uint32(), options, message.fund);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fund)
      InsuranceFund.internalBinaryWrite(message.fund, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FundResponse = new FundResponse$Type();
class RedemptionsRequest$Type extends MessageType {
  constructor() {
    super("injective_insurance_rpc.RedemptionsRequest", [
      {
        no: 1,
        name: "redeemer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "redemption_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "status",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.redeemer = "";
    message.redemptionDenom = "";
    message.status = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string redeemer */
        1:
          message.redeemer = reader.string();
          break;
        case /* string redemption_denom */
        2:
          message.redemptionDenom = reader.string();
          break;
        case /* string status */
        3:
          message.status = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.redeemer !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.redeemer);
    if (message.redemptionDenom !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.redemptionDenom);
    if (message.status !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.status);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RedemptionsRequest = new RedemptionsRequest$Type();
class RedemptionsResponse$Type extends MessageType {
  constructor() {
    super("injective_insurance_rpc.RedemptionsResponse", [
      { no: 1, name: "redemption_schedules", kind: "message", repeat: 2, T: () => RedemptionSchedule }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.redemptionSchedules = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_insurance_rpc.RedemptionSchedule redemption_schedules */
        1:
          message.redemptionSchedules.push(RedemptionSchedule.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.redemptionSchedules.length; i++)
      RedemptionSchedule.internalBinaryWrite(message.redemptionSchedules[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RedemptionsResponse = new RedemptionsResponse$Type();
class RedemptionSchedule$Type extends MessageType {
  constructor() {
    super("injective_insurance_rpc.RedemptionSchedule", [
      {
        no: 1,
        name: "redemption_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "status",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "redeemer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "claimable_redemption_time",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "redemption_amount",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "redemption_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "requested_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 8,
        name: "disbursed_amount",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "disbursed_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "disbursed_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.redemptionId = 0n;
    message.status = "";
    message.redeemer = "";
    message.claimableRedemptionTime = 0n;
    message.redemptionAmount = "";
    message.redemptionDenom = "";
    message.requestedAt = 0n;
    message.disbursedAmount = "";
    message.disbursedDenom = "";
    message.disbursedAt = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 redemption_id */
        1:
          message.redemptionId = reader.uint64().toBigInt();
          break;
        case /* string status */
        2:
          message.status = reader.string();
          break;
        case /* string redeemer */
        3:
          message.redeemer = reader.string();
          break;
        case /* sint64 claimable_redemption_time */
        4:
          message.claimableRedemptionTime = reader.sint64().toBigInt();
          break;
        case /* string redemption_amount */
        5:
          message.redemptionAmount = reader.string();
          break;
        case /* string redemption_denom */
        6:
          message.redemptionDenom = reader.string();
          break;
        case /* sint64 requested_at */
        7:
          message.requestedAt = reader.sint64().toBigInt();
          break;
        case /* string disbursed_amount */
        8:
          message.disbursedAmount = reader.string();
          break;
        case /* string disbursed_denom */
        9:
          message.disbursedDenom = reader.string();
          break;
        case /* sint64 disbursed_at */
        10:
          message.disbursedAt = reader.sint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.redemptionId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.redemptionId);
    if (message.status !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.status);
    if (message.redeemer !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.redeemer);
    if (message.claimableRedemptionTime !== 0n)
      writer.tag(4, WireType.Varint).sint64(message.claimableRedemptionTime);
    if (message.redemptionAmount !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.redemptionAmount);
    if (message.redemptionDenom !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.redemptionDenom);
    if (message.requestedAt !== 0n)
      writer.tag(7, WireType.Varint).sint64(message.requestedAt);
    if (message.disbursedAmount !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.disbursedAmount);
    if (message.disbursedDenom !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.disbursedDenom);
    if (message.disbursedAt !== 0n)
      writer.tag(10, WireType.Varint).sint64(message.disbursedAt);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RedemptionSchedule = new RedemptionSchedule$Type();
const InjectiveInsuranceRPC = new ServiceType("injective_insurance_rpc.InjectiveInsuranceRPC", [
  { name: "Funds", options: {}, I: FundsRequest, O: FundsResponse },
  { name: "Fund", options: {}, I: FundRequest, O: FundResponse },
  { name: "Redemptions", options: {}, I: RedemptionsRequest, O: RedemptionsResponse }
]);
export {
  FundRequest,
  FundResponse,
  FundsRequest,
  FundsResponse,
  InjectiveInsuranceRPC,
  InsuranceFund,
  RedemptionSchedule,
  RedemptionsRequest,
  RedemptionsResponse,
  TokenMeta
};
