// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_archiver_rpc_pb.client.js
import { InjectiveArchiverRPC } from "./injective_archiver_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveArchiverRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveArchiverRPC.typeName;
    this.methods = InjectiveArchiverRPC.methods;
    this.options = InjectiveArchiverRPC.options;
  }
  /**
   * Provide historical balance data for a given account address.
   *
   * @generated from protobuf rpc: Balance
   */
  balance(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide all-time stats for a given account address.
   *
   * @generated from protobuf rpc: AccountStats
   */
  accountStats(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide historical realized profit and loss data for a given account address.
   *
   * @generated from protobuf rpc: Rpnl
   */
  rpnl(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide historical volumes for a given account address.
   *
   * @generated from protobuf rpc: Volumes
   */
  volumes(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide pnl leaderboard data.
   *
   * @generated from protobuf rpc: PnlLeaderboard
   */
  pnlLeaderboard(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide volume leaderboard data.
   *
   * @generated from protobuf rpc: VolLeaderboard
   */
  volLeaderboard(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide pnl leaderboard data.
   *
   * @generated from protobuf rpc: PnlLeaderboardFixedResolution
   */
  pnlLeaderboardFixedResolution(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide volume leaderboard data.
   *
   * @generated from protobuf rpc: VolLeaderboardFixedResolution
   */
  volLeaderboardFixedResolution(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide a list of addresses holding a specific denom
   *
   * @generated from protobuf rpc: DenomHolders
   */
  denomHolders(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide historical trades data.
   *
   * @generated from protobuf rpc: HistoricalTrades
   */
  historicalTrades(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream spot markets average entries for a given account address.
   *
   * @generated from protobuf rpc: StreamSpotAverageEntries
   */
  streamSpotAverageEntries(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
}
export {
  InjectiveArchiverRPCClient
};
