// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_explorer_rpc_pb.client.js
import { InjectiveExplorerRPC } from "./injective_explorer_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveExplorerRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveExplorerRPC.typeName;
    this.methods = InjectiveExplorerRPC.methods;
    this.options = InjectiveExplorerRPC.options;
  }
  /**
   * GetAccountTxs returns transactions involving in an account based upon
   * params. Deprecated: use GetAccountTxsV2 instead.
   *
   * @generated from protobuf rpc: GetAccountTxs
   */
  getAccountTxs(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetAccountTxs returns tranctions involving in an account based upon params.
   *
   * @generated from protobuf rpc: GetAccountTxsV2
   */
  getAccountTxsV2(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetContractTxs returns contract-related transactions. Deprecated: use
   * GetContractTxsV2 instead.
   *
   * @generated from protobuf rpc: GetContractTxs
   */
  getContractTxs(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetContractTxs returns contract-related transactions
   *
   * @generated from protobuf rpc: GetContractTxsV2
   */
  getContractTxsV2(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetBlocks returns blocks based upon the request params. Deprecated: use
   * GetBlocksV2 instead.
   *
   * @generated from protobuf rpc: GetBlocks
   */
  getBlocks(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetBlocks returns blocks based upon the request params
   *
   * @generated from protobuf rpc: GetBlocksV2
   */
  getBlocksV2(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetBlock returns block based upon the height or hash
   *
   * @generated from protobuf rpc: GetBlock
   */
  getBlock(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetValidators returns validators on the active chain
   *
   * @generated from protobuf rpc: GetValidators
   */
  getValidators(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetValidator returns validator information on the active chain
   *
   * @generated from protobuf rpc: GetValidator
   */
  getValidator(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetValidatorUptime returns validator uptime information on the active chain
   *
   * @generated from protobuf rpc: GetValidatorUptime
   */
  getValidatorUptime(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetTxs returns transactions based upon the request params. Deprecated: use
   * GetTxsV2 instead.
   *
   * @generated from protobuf rpc: GetTxs
   */
  getTxs(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetTxs returns transactions based upon the request params
   *
   * @generated from protobuf rpc: GetTxsV2
   */
  getTxsV2(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetTxByTxHash returns certain transaction information by its tx hash.
   *
   * @generated from protobuf rpc: GetTxByTxHash
   */
  getTxByTxHash(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetPeggyDepositTxs returns the peggy deposit transactions based upon the
   * request params
   *
   * @generated from protobuf rpc: GetPeggyDepositTxs
   */
  getPeggyDepositTxs(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetPeggyWithdrawalTxs returns the peggy withdrawal transactions based upon
   * the request params
   *
   * @generated from protobuf rpc: GetPeggyWithdrawalTxs
   */
  getPeggyWithdrawalTxs(input, options) {
    const method = this.methods[14], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetIBCTransferTxs returns the ibc transfer transactions based upon the
   * request params
   *
   * @generated from protobuf rpc: GetIBCTransferTxs
   */
  getIBCTransferTxs(input, options) {
    const method = this.methods[15], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetWasmCodes lists all stored code
   *
   * @generated from protobuf rpc: GetWasmCodes
   */
  getWasmCodes(input, options) {
    const method = this.methods[16], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetWasmCodeById list cosmwasm code infor by ID
   *
   * @generated from protobuf rpc: GetWasmCodeByID
   */
  getWasmCodeByID(input, options) {
    const method = this.methods[17], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetWasmContracts lists all contracts
   *
   * @generated from protobuf rpc: GetWasmContracts
   */
  getWasmContracts(input, options) {
    const method = this.methods[18], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetWasmContractByAddress list cosmwasm contract info by its address
   *
   * @generated from protobuf rpc: GetWasmContractByAddress
   */
  getWasmContractByAddress(input, options) {
    const method = this.methods[19], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetCw20Balance lists all cw20 balances of an injective account
   *
   * @generated from protobuf rpc: GetCw20Balance
   */
  getCw20Balance(input, options) {
    const method = this.methods[20], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Request relayers infos by marketIDs. If no ids are provided, all market with
   * associated relayers are returned
   *
   * @generated from protobuf rpc: Relayers
   */
  relayers(input, options) {
    const method = this.methods[21], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetBankTransfers returns bank transfers.
   *
   * @generated from protobuf rpc: GetBankTransfers
   */
  getBankTransfers(input, options) {
    const method = this.methods[22], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StreamTxs returns transactions based upon the request params
   *
   * @generated from protobuf rpc: StreamTxs
   */
  streamTxs(input, options) {
    const method = this.methods[23], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * StreamBlocks returns the latest blocks
   *
   * @generated from protobuf rpc: StreamBlocks
   */
  streamBlocks(input, options) {
    const method = this.methods[24], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * GetStats returns global exchange statistics in the last 24hs
   *
   * @generated from protobuf rpc: GetStats
   */
  getStats(input, options) {
    const method = this.methods[25], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveExplorerRPCClient
};
