// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_auction_rpc_pb.client.js
import { InjectiveAuctionRPC } from "./injective_auction_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveAuctionRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveAuctionRPC.typeName;
    this.methods = InjectiveAuctionRPC.methods;
    this.options = InjectiveAuctionRPC.options;
  }
  /**
   * Provide historical auction info for a given auction
   *
   * @generated from protobuf rpc: AuctionEndpoint
   */
  auctionEndpoint(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide the historical auctions info
   *
   * @generated from protobuf rpc: Auctions
   */
  auctions(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide the historical auctions info
   *
   * @generated from protobuf rpc: AuctionsHistoryV2
   */
  auctionsHistoryV2(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide historical auction info for a given auction
   *
   * @generated from protobuf rpc: AuctionV2
   */
  auctionV2(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide the historical auctions info for a given account
   *
   * @generated from protobuf rpc: AccountAuctionsV2
   */
  accountAuctionsV2(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get the allowlist status for a specific account in an auction round
   *
   * @generated from protobuf rpc: AuctionAccountStatus
   */
  auctionAccountStatus(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StreamBids streams new bids of an auction.
   *
   * @generated from protobuf rpc: StreamBids
   */
  streamBids(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * InjBurntEndpoint returns the total amount of INJ burnt in auctions.
   *
   * @generated from protobuf rpc: InjBurntEndpoint
   */
  injBurntEndpoint(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Returns total INJ burnt and its total USD value in auctions.
   *
   * @generated from protobuf rpc: AuctionsStats
   */
  auctionsStats(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveAuctionRPCClient
};
