// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_campaign_rpc_pb.client.js
import { InjectiveCampaignRPC } from "./injective_campaign_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveCampaignRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveCampaignRPC.typeName;
    this.methods = InjectiveCampaignRPC.methods;
    this.options = InjectiveCampaignRPC.options;
  }
  /**
   * Lists all participants in campaign
   *
   * @generated from protobuf rpc: Ranking
   */
  ranking(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List current round active campaigns
   *
   * @generated from protobuf rpc: Campaigns
   */
  campaigns(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List campaigns v2
   *
   * @generated from protobuf rpc: CampaignsV2
   */
  campaignsV2(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List guilds by campaign
   *
   * @generated from protobuf rpc: ListGuilds
   */
  listGuilds(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List guild members of given campaign and guildId
   *
   * @generated from protobuf rpc: ListGuildMembers
   */
  listGuildMembers(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get single member guild info
   *
   * @generated from protobuf rpc: GetGuildMember
   */
  getGuildMember(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveCampaignRPCClient
};
