// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_portfolio_rpc_pb.client.js
import { InjectivePortfolioRPC } from "./injective_portfolio_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectivePortfolioRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectivePortfolioRPC.typeName;
    this.methods = InjectivePortfolioRPC.methods;
    this.options = InjectivePortfolioRPC.options;
  }
  /**
   * Provide a list of addresses holding a specific token
   *
   * @generated from protobuf rpc: TokenHolders
   */
  tokenHolders(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide the account's portfolio
   *
   * @generated from protobuf rpc: AccountPortfolio
   */
  accountPortfolio(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide the account's portfolio balances
   *
   * @generated from protobuf rpc: AccountPortfolioBalances
   */
  accountPortfolioBalances(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream the account's portfolio
   *
   * @generated from protobuf rpc: StreamAccountPortfolio
   */
  streamAccountPortfolio(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
}
export {
  InjectivePortfolioRPCClient
};
