// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_derivative_exchange_rpc_pb.client.js
import { InjectiveDerivativeExchangeRPC } from "./injective_derivative_exchange_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveDerivativeExchangeRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveDerivativeExchangeRPC.typeName;
    this.methods = InjectiveDerivativeExchangeRPC.methods;
    this.options = InjectiveDerivativeExchangeRPC.options;
  }
  /**
   * Markets gets a list of Derivative Markets
   *
   * @generated from protobuf rpc: Markets
   */
  markets(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Market gets details of a single derivative market
   *
   * @generated from protobuf rpc: Market
   */
  market(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StreamMarket streams live updates of selected derivative markets
   *
   * @generated from protobuf rpc: StreamMarket
   */
  streamMarket(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * BinaryOptionsMarkets gets a list of Binary Options Markets
   *
   * @generated from protobuf rpc: BinaryOptionsMarkets
   */
  binaryOptionsMarkets(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * BinaryOptionMarket gets details of a single binary options market
   *
   * @generated from protobuf rpc: BinaryOptionsMarket
   */
  binaryOptionsMarket(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Orderbook gets the Orderbook of a Derivative Market
   *
   * @generated from protobuf rpc: OrderbookV2
   */
  orderbookV2(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Orderbooks gets the Orderbooks of requested derivative markets
   *
   * @generated from protobuf rpc: OrderbooksV2
   */
  orderbooksV2(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream live snapshot updates of selected derivative market orderbook
   *
   * @generated from protobuf rpc: StreamOrderbookV2
   */
  streamOrderbookV2(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Stream live level updates of selected derivative market orderbook
   *
   * @generated from protobuf rpc: StreamOrderbookUpdate
   */
  streamOrderbookUpdate(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * DerivativeLimitOrders gets the limit orders of a derivative Market.
   *
   * @generated from protobuf rpc: Orders
   */
  orders(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Positions gets the positions for a trader.
   *
   * @generated from protobuf rpc: Positions
   */
  positions(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Positions gets the positions for a trader. V2 removed some redundant fields
   * and had performance improvements
   *
   * @generated from protobuf rpc: PositionsV2
   */
  positionsV2(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * LiquidablePositions gets all the liquidable positions.
   *
   * @generated from protobuf rpc: LiquidablePositions
   */
  liquidablePositions(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * FundingPayments gets the funding payments for a trader.
   *
   * @generated from protobuf rpc: FundingPayments
   */
  fundingPayments(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * FundingRates gets the historical funding rates for a market.
   *
   * @generated from protobuf rpc: FundingRates
   */
  fundingRates(input, options) {
    const method = this.methods[14], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StreamPositions streams derivatives position updates. This is the legacy
   * version of the streamPositionsV2 endpoint. Use streamPositionsV2 instead.
   *
   * @generated from protobuf rpc: StreamPositions
   */
  streamPositions(input, options) {
    const method = this.methods[15], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * StreamPositionsV2 streams derivatives position updates.
   *
   * @generated from protobuf rpc: StreamPositionsV2
   */
  streamPositionsV2(input, options) {
    const method = this.methods[16], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * StreamOrders streams updates to individual orders of a Derivative Market.
   *
   * @generated from protobuf rpc: StreamOrders
   */
  streamOrders(input, options) {
    const method = this.methods[17], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Trades gets the trades of a Derivative Market.
   *
   * @generated from protobuf rpc: Trades
   */
  trades(input, options) {
    const method = this.methods[18], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Trades gets the trades of a Derivative Market.
   *
   * @generated from protobuf rpc: TradesV2
   */
  tradesV2(input, options) {
    const method = this.methods[19], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StreamTrades streams newly executed trades from Derivative Market.
   *
   * @generated from protobuf rpc: StreamTrades
   */
  streamTrades(input, options) {
    const method = this.methods[20], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * StreamTrades streams newly executed trades from Derivative Market.
   *
   * @generated from protobuf rpc: StreamTradesV2
   */
  streamTradesV2(input, options) {
    const method = this.methods[21], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * SubaccountOrdersList lists orders posted from this subaccount.
   *
   * @generated from protobuf rpc: SubaccountOrdersList
   */
  subaccountOrdersList(input, options) {
    const method = this.methods[22], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * SubaccountTradesList gets a list of derivatives trades executed by this
   * subaccount.
   *
   * @generated from protobuf rpc: SubaccountTradesList
   */
  subaccountTradesList(input, options) {
    const method = this.methods[23], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Lists history orders posted from a subaccount
   *
   * @generated from protobuf rpc: OrdersHistory
   */
  ordersHistory(input, options) {
    const method = this.methods[24], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream updates to historical orders of a derivative Market
   *
   * @generated from protobuf rpc: StreamOrdersHistory
   */
  streamOrdersHistory(input, options) {
    const method = this.methods[25], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * OpenInterest gets the open interest for a derivative market.
   *
   * @generated from protobuf rpc: OpenInterest
   */
  openInterest(input, options) {
    const method = this.methods[26], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveDerivativeExchangeRPCClient
};
