// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_campaign_rpc_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class RankingRequest$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.RankingRequest", [
      {
        no: 1,
        name: "campaign_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "limit",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 5,
        name: "skip",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignId = "";
    message.marketId = "";
    message.accountAddress = "";
    message.limit = 0;
    message.skip = 0n;
    message.contractAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_id */
        1:
          message.campaignId = reader.string();
          break;
        case /* string market_id */
        2:
          message.marketId = reader.string();
          break;
        case /* string account_address */
        3:
          message.accountAddress = reader.string();
          break;
        case /* sint32 limit */
        4:
          message.limit = reader.sint32();
          break;
        case /* uint64 skip */
        5:
          message.skip = reader.uint64().toBigInt();
          break;
        case /* string contract_address */
        6:
          message.contractAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignId);
    if (message.marketId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    if (message.accountAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.accountAddress);
    if (message.limit !== 0)
      writer.tag(4, WireType.Varint).sint32(message.limit);
    if (message.skip !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.skip);
    if (message.contractAddress !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.contractAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RankingRequest = new RankingRequest$Type();
class RankingResponse$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.RankingResponse", [
      { no: 1, name: "campaign", kind: "message", T: () => Campaign },
      { no: 2, name: "users", kind: "message", repeat: 2, T: () => CampaignUser },
      { no: 3, name: "paging", kind: "message", T: () => Paging }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.users = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective_campaign_rpc.Campaign campaign */
        1:
          message.campaign = Campaign.internalBinaryRead(reader, reader.uint32(), options, message.campaign);
          break;
        case /* repeated injective_campaign_rpc.CampaignUser users */
        2:
          message.users.push(CampaignUser.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* injective_campaign_rpc.Paging paging */
        3:
          message.paging = Paging.internalBinaryRead(reader, reader.uint32(), options, message.paging);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaign)
      Campaign.internalBinaryWrite(message.campaign, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.users.length; i++)
      CampaignUser.internalBinaryWrite(message.users[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.paging)
      Paging.internalBinaryWrite(message.paging, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RankingResponse = new RankingResponse$Type();
class Campaign$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.Campaign", [
      {
        no: 1,
        name: "campaign_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "total_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "last_updated",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "start_date",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "end_date",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 8,
        name: "is_claimable",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 9,
        name: "round_id",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 10,
        name: "manager_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 11, name: "rewards", kind: "message", repeat: 2, T: () => Coin },
      {
        no: 12,
        name: "user_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 13,
        name: "user_claimed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 14,
        name: "subaccount_id_suffix",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 15,
        name: "reward_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 16,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 17,
        name: "type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignId = "";
    message.marketId = "";
    message.totalScore = "";
    message.lastUpdated = 0n;
    message.startDate = 0n;
    message.endDate = 0n;
    message.isClaimable = false;
    message.roundId = 0;
    message.managerContract = "";
    message.rewards = [];
    message.userScore = "";
    message.userClaimed = false;
    message.subaccountIdSuffix = "";
    message.rewardContract = "";
    message.version = "";
    message.type = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_id */
        1:
          message.campaignId = reader.string();
          break;
        case /* string market_id */
        2:
          message.marketId = reader.string();
          break;
        case /* string total_score */
        4:
          message.totalScore = reader.string();
          break;
        case /* sint64 last_updated */
        5:
          message.lastUpdated = reader.sint64().toBigInt();
          break;
        case /* sint64 start_date */
        6:
          message.startDate = reader.sint64().toBigInt();
          break;
        case /* sint64 end_date */
        7:
          message.endDate = reader.sint64().toBigInt();
          break;
        case /* bool is_claimable */
        8:
          message.isClaimable = reader.bool();
          break;
        case /* sint32 round_id */
        9:
          message.roundId = reader.sint32();
          break;
        case /* string manager_contract */
        10:
          message.managerContract = reader.string();
          break;
        case /* repeated injective_campaign_rpc.Coin rewards */
        11:
          message.rewards.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string user_score */
        12:
          message.userScore = reader.string();
          break;
        case /* bool user_claimed */
        13:
          message.userClaimed = reader.bool();
          break;
        case /* string subaccount_id_suffix */
        14:
          message.subaccountIdSuffix = reader.string();
          break;
        case /* string reward_contract */
        15:
          message.rewardContract = reader.string();
          break;
        case /* string version */
        16:
          message.version = reader.string();
          break;
        case /* string type */
        17:
          message.type = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignId);
    if (message.marketId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    if (message.totalScore !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.totalScore);
    if (message.lastUpdated !== 0n)
      writer.tag(5, WireType.Varint).sint64(message.lastUpdated);
    if (message.startDate !== 0n)
      writer.tag(6, WireType.Varint).sint64(message.startDate);
    if (message.endDate !== 0n)
      writer.tag(7, WireType.Varint).sint64(message.endDate);
    if (message.isClaimable !== false)
      writer.tag(8, WireType.Varint).bool(message.isClaimable);
    if (message.roundId !== 0)
      writer.tag(9, WireType.Varint).sint32(message.roundId);
    if (message.managerContract !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.managerContract);
    for (let i = 0; i < message.rewards.length; i++)
      Coin.internalBinaryWrite(message.rewards[i], writer.tag(11, WireType.LengthDelimited).fork(), options).join();
    if (message.userScore !== "")
      writer.tag(12, WireType.LengthDelimited).string(message.userScore);
    if (message.userClaimed !== false)
      writer.tag(13, WireType.Varint).bool(message.userClaimed);
    if (message.subaccountIdSuffix !== "")
      writer.tag(14, WireType.LengthDelimited).string(message.subaccountIdSuffix);
    if (message.rewardContract !== "")
      writer.tag(15, WireType.LengthDelimited).string(message.rewardContract);
    if (message.version !== "")
      writer.tag(16, WireType.LengthDelimited).string(message.version);
    if (message.type !== "")
      writer.tag(17, WireType.LengthDelimited).string(message.type);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Campaign = new Campaign$Type();
class Coin$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.Coin", [
      {
        no: 1,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "amount",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "usd_value",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.denom = "";
    message.amount = "";
    message.usdValue = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string denom */
        1:
          message.denom = reader.string();
          break;
        case /* string amount */
        2:
          message.amount = reader.string();
          break;
        case /* string usd_value */
        3:
          message.usdValue = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.denom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.denom);
    if (message.amount !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.amount);
    if (message.usdValue !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.usdValue);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Coin = new Coin$Type();
class CampaignUser$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.CampaignUser", [
      {
        no: 1,
        name: "campaign_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "contract_updated",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 6,
        name: "block_height",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "block_time",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 8,
        name: "purchased_amount",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "galxe_updated",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 10,
        name: "reward_claimed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignId = "";
    message.marketId = "";
    message.accountAddress = "";
    message.score = "";
    message.contractUpdated = false;
    message.blockHeight = 0n;
    message.blockTime = 0n;
    message.purchasedAmount = "";
    message.galxeUpdated = false;
    message.rewardClaimed = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_id */
        1:
          message.campaignId = reader.string();
          break;
        case /* string market_id */
        2:
          message.marketId = reader.string();
          break;
        case /* string account_address */
        3:
          message.accountAddress = reader.string();
          break;
        case /* string score */
        4:
          message.score = reader.string();
          break;
        case /* bool contract_updated */
        5:
          message.contractUpdated = reader.bool();
          break;
        case /* sint64 block_height */
        6:
          message.blockHeight = reader.sint64().toBigInt();
          break;
        case /* sint64 block_time */
        7:
          message.blockTime = reader.sint64().toBigInt();
          break;
        case /* string purchased_amount */
        8:
          message.purchasedAmount = reader.string();
          break;
        case /* bool galxe_updated */
        9:
          message.galxeUpdated = reader.bool();
          break;
        case /* bool reward_claimed */
        10:
          message.rewardClaimed = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignId);
    if (message.marketId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    if (message.accountAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.accountAddress);
    if (message.score !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.score);
    if (message.contractUpdated !== false)
      writer.tag(5, WireType.Varint).bool(message.contractUpdated);
    if (message.blockHeight !== 0n)
      writer.tag(6, WireType.Varint).sint64(message.blockHeight);
    if (message.blockTime !== 0n)
      writer.tag(7, WireType.Varint).sint64(message.blockTime);
    if (message.purchasedAmount !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.purchasedAmount);
    if (message.galxeUpdated !== false)
      writer.tag(9, WireType.Varint).bool(message.galxeUpdated);
    if (message.rewardClaimed !== false)
      writer.tag(10, WireType.Varint).bool(message.rewardClaimed);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CampaignUser = new CampaignUser$Type();
class Paging$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.Paging", [
      {
        no: 1,
        name: "total",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "from",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 3,
        name: "to",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 4,
        name: "count_by_subaccount",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "next",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.total = 0n;
    message.from = 0;
    message.to = 0;
    message.countBySubaccount = 0n;
    message.next = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* sint64 total */
        1:
          message.total = reader.sint64().toBigInt();
          break;
        case /* sint32 from */
        2:
          message.from = reader.sint32();
          break;
        case /* sint32 to */
        3:
          message.to = reader.sint32();
          break;
        case /* sint64 count_by_subaccount */
        4:
          message.countBySubaccount = reader.sint64().toBigInt();
          break;
        case /* repeated string next */
        5:
          message.next.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.total !== 0n)
      writer.tag(1, WireType.Varint).sint64(message.total);
    if (message.from !== 0)
      writer.tag(2, WireType.Varint).sint32(message.from);
    if (message.to !== 0)
      writer.tag(3, WireType.Varint).sint32(message.to);
    if (message.countBySubaccount !== 0n)
      writer.tag(4, WireType.Varint).sint64(message.countBySubaccount);
    for (let i = 0; i < message.next.length; i++)
      writer.tag(5, WireType.LengthDelimited).string(message.next[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Paging = new Paging$Type();
class CampaignsRequest$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.CampaignsRequest", [
      {
        no: 1,
        name: "round_id",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "to_round_id",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 4,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.roundId = 0n;
    message.accountAddress = "";
    message.toRoundId = 0;
    message.contractAddress = "";
    message.type = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* sint64 round_id */
        1:
          message.roundId = reader.sint64().toBigInt();
          break;
        case /* string account_address */
        2:
          message.accountAddress = reader.string();
          break;
        case /* sint32 to_round_id */
        3:
          message.toRoundId = reader.sint32();
          break;
        case /* string contract_address */
        4:
          message.contractAddress = reader.string();
          break;
        case /* string type */
        5:
          message.type = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.roundId !== 0n)
      writer.tag(1, WireType.Varint).sint64(message.roundId);
    if (message.accountAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.accountAddress);
    if (message.toRoundId !== 0)
      writer.tag(3, WireType.Varint).sint32(message.toRoundId);
    if (message.contractAddress !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.contractAddress);
    if (message.type !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.type);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CampaignsRequest = new CampaignsRequest$Type();
class CampaignsResponse$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.CampaignsResponse", [
      { no: 1, name: "campaigns", kind: "message", repeat: 2, T: () => Campaign },
      { no: 2, name: "accumulated_rewards", kind: "message", repeat: 2, T: () => Coin },
      {
        no: 3,
        name: "reward_count",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaigns = [];
    message.accumulatedRewards = [];
    message.rewardCount = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_campaign_rpc.Campaign campaigns */
        1:
          message.campaigns.push(Campaign.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective_campaign_rpc.Coin accumulated_rewards */
        2:
          message.accumulatedRewards.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* sint32 reward_count */
        3:
          message.rewardCount = reader.sint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.campaigns.length; i++)
      Campaign.internalBinaryWrite(message.campaigns[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.accumulatedRewards.length; i++)
      Coin.internalBinaryWrite(message.accumulatedRewards[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.rewardCount !== 0)
      writer.tag(3, WireType.Varint).sint32(message.rewardCount);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CampaignsResponse = new CampaignsResponse$Type();
class CampaignsV2Request$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.CampaignsV2Request", [
      {
        no: 1,
        name: "type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "active",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 3,
        name: "limit",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 4,
        name: "cursor",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "from_start_date",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "to_start_date",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "from_end_date",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 8,
        name: "to_end_date",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 9,
        name: "status",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.type = "";
    message.active = false;
    message.limit = 0;
    message.cursor = "";
    message.fromStartDate = 0n;
    message.toStartDate = 0n;
    message.fromEndDate = 0n;
    message.toEndDate = 0n;
    message.status = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string type */
        1:
          message.type = reader.string();
          break;
        case /* bool active */
        2:
          message.active = reader.bool();
          break;
        case /* sint32 limit */
        3:
          message.limit = reader.sint32();
          break;
        case /* string cursor */
        4:
          message.cursor = reader.string();
          break;
        case /* sint64 from_start_date */
        5:
          message.fromStartDate = reader.sint64().toBigInt();
          break;
        case /* sint64 to_start_date */
        6:
          message.toStartDate = reader.sint64().toBigInt();
          break;
        case /* sint64 from_end_date */
        7:
          message.fromEndDate = reader.sint64().toBigInt();
          break;
        case /* sint64 to_end_date */
        8:
          message.toEndDate = reader.sint64().toBigInt();
          break;
        case /* string status */
        9:
          message.status = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.type !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.type);
    if (message.active !== false)
      writer.tag(2, WireType.Varint).bool(message.active);
    if (message.limit !== 0)
      writer.tag(3, WireType.Varint).sint32(message.limit);
    if (message.cursor !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.cursor);
    if (message.fromStartDate !== 0n)
      writer.tag(5, WireType.Varint).sint64(message.fromStartDate);
    if (message.toStartDate !== 0n)
      writer.tag(6, WireType.Varint).sint64(message.toStartDate);
    if (message.fromEndDate !== 0n)
      writer.tag(7, WireType.Varint).sint64(message.fromEndDate);
    if (message.toEndDate !== 0n)
      writer.tag(8, WireType.Varint).sint64(message.toEndDate);
    if (message.status !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.status);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CampaignsV2Request = new CampaignsV2Request$Type();
class CampaignsV2Response$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.CampaignsV2Response", [
      { no: 1, name: "campaigns", kind: "message", repeat: 2, T: () => CampaignV2 },
      {
        no: 2,
        name: "cursor",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaigns = [];
    message.cursor = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_campaign_rpc.CampaignV2 campaigns */
        1:
          message.campaigns.push(CampaignV2.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string cursor */
        2:
          message.cursor = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.campaigns.length; i++)
      CampaignV2.internalBinaryWrite(message.campaigns[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.cursor !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.cursor);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CampaignsV2Response = new CampaignsV2Response$Type();
class CampaignV2$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.CampaignV2", [
      {
        no: 1,
        name: "campaign_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "total_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "created_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "updated_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "start_date",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 8,
        name: "end_date",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 9,
        name: "is_claimable",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 10,
        name: "round_id",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 11,
        name: "manager_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 12, name: "rewards", kind: "message", repeat: 2, T: () => Coin },
      {
        no: 13,
        name: "subaccount_id_suffix",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 14,
        name: "reward_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 15,
        name: "type",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 16,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 17,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 18,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignId = "";
    message.marketId = "";
    message.totalScore = "";
    message.createdAt = 0n;
    message.updatedAt = 0n;
    message.startDate = 0n;
    message.endDate = 0n;
    message.isClaimable = false;
    message.roundId = 0;
    message.managerContract = "";
    message.rewards = [];
    message.subaccountIdSuffix = "";
    message.rewardContract = "";
    message.type = "";
    message.version = "";
    message.name = "";
    message.description = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_id */
        1:
          message.campaignId = reader.string();
          break;
        case /* string market_id */
        2:
          message.marketId = reader.string();
          break;
        case /* string total_score */
        4:
          message.totalScore = reader.string();
          break;
        case /* sint64 created_at */
        5:
          message.createdAt = reader.sint64().toBigInt();
          break;
        case /* sint64 updated_at */
        6:
          message.updatedAt = reader.sint64().toBigInt();
          break;
        case /* sint64 start_date */
        7:
          message.startDate = reader.sint64().toBigInt();
          break;
        case /* sint64 end_date */
        8:
          message.endDate = reader.sint64().toBigInt();
          break;
        case /* bool is_claimable */
        9:
          message.isClaimable = reader.bool();
          break;
        case /* sint32 round_id */
        10:
          message.roundId = reader.sint32();
          break;
        case /* string manager_contract */
        11:
          message.managerContract = reader.string();
          break;
        case /* repeated injective_campaign_rpc.Coin rewards */
        12:
          message.rewards.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string subaccount_id_suffix */
        13:
          message.subaccountIdSuffix = reader.string();
          break;
        case /* string reward_contract */
        14:
          message.rewardContract = reader.string();
          break;
        case /* string type */
        15:
          message.type = reader.string();
          break;
        case /* string version */
        16:
          message.version = reader.string();
          break;
        case /* string name */
        17:
          message.name = reader.string();
          break;
        case /* string description */
        18:
          message.description = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignId);
    if (message.marketId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.marketId);
    if (message.totalScore !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.totalScore);
    if (message.createdAt !== 0n)
      writer.tag(5, WireType.Varint).sint64(message.createdAt);
    if (message.updatedAt !== 0n)
      writer.tag(6, WireType.Varint).sint64(message.updatedAt);
    if (message.startDate !== 0n)
      writer.tag(7, WireType.Varint).sint64(message.startDate);
    if (message.endDate !== 0n)
      writer.tag(8, WireType.Varint).sint64(message.endDate);
    if (message.isClaimable !== false)
      writer.tag(9, WireType.Varint).bool(message.isClaimable);
    if (message.roundId !== 0)
      writer.tag(10, WireType.Varint).sint32(message.roundId);
    if (message.managerContract !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.managerContract);
    for (let i = 0; i < message.rewards.length; i++)
      Coin.internalBinaryWrite(message.rewards[i], writer.tag(12, WireType.LengthDelimited).fork(), options).join();
    if (message.subaccountIdSuffix !== "")
      writer.tag(13, WireType.LengthDelimited).string(message.subaccountIdSuffix);
    if (message.rewardContract !== "")
      writer.tag(14, WireType.LengthDelimited).string(message.rewardContract);
    if (message.type !== "")
      writer.tag(15, WireType.LengthDelimited).string(message.type);
    if (message.version !== "")
      writer.tag(16, WireType.LengthDelimited).string(message.version);
    if (message.name !== "")
      writer.tag(17, WireType.LengthDelimited).string(message.name);
    if (message.description !== "")
      writer.tag(18, WireType.LengthDelimited).string(message.description);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CampaignV2 = new CampaignV2$Type();
class ListGuildsRequest$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.ListGuildsRequest", [
      {
        no: 1,
        name: "campaign_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "limit",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 3,
        name: "skip",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 4,
        name: "sort_by",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignContract = "";
    message.limit = 0;
    message.skip = 0;
    message.sortBy = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_contract */
        1:
          message.campaignContract = reader.string();
          break;
        case /* sint32 limit */
        2:
          message.limit = reader.sint32();
          break;
        case /* sint32 skip */
        3:
          message.skip = reader.sint32();
          break;
        case /* string sort_by */
        4:
          message.sortBy = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignContract !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignContract);
    if (message.limit !== 0)
      writer.tag(2, WireType.Varint).sint32(message.limit);
    if (message.skip !== 0)
      writer.tag(3, WireType.Varint).sint32(message.skip);
    if (message.sortBy !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.sortBy);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListGuildsRequest = new ListGuildsRequest$Type();
class ListGuildsResponse$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.ListGuildsResponse", [
      { no: 1, name: "guilds", kind: "message", repeat: 2, T: () => Guild },
      { no: 2, name: "paging", kind: "message", T: () => Paging },
      {
        no: 3,
        name: "updated_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "campaign_summary", kind: "message", T: () => CampaignSummary }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.guilds = [];
    message.updatedAt = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_campaign_rpc.Guild guilds */
        1:
          message.guilds.push(Guild.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* injective_campaign_rpc.Paging paging */
        2:
          message.paging = Paging.internalBinaryRead(reader, reader.uint32(), options, message.paging);
          break;
        case /* sint64 updated_at */
        3:
          message.updatedAt = reader.sint64().toBigInt();
          break;
        case /* injective_campaign_rpc.CampaignSummary campaign_summary */
        4:
          message.campaignSummary = CampaignSummary.internalBinaryRead(reader, reader.uint32(), options, message.campaignSummary);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.guilds.length; i++)
      Guild.internalBinaryWrite(message.guilds[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.paging)
      Paging.internalBinaryWrite(message.paging, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.updatedAt !== 0n)
      writer.tag(3, WireType.Varint).sint64(message.updatedAt);
    if (message.campaignSummary)
      CampaignSummary.internalBinaryWrite(message.campaignSummary, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListGuildsResponse = new ListGuildsResponse$Type();
class Guild$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.Guild", [
      {
        no: 1,
        name: "campaign_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "guild_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "master_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "created_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "tvl_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "volume_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "rank_by_volume",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 8,
        name: "rank_by_tvl",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 9,
        name: "logo",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "total_tvl",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "updated_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 14,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 13,
        name: "is_active",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 15,
        name: "master_balance",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 16,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignContract = "";
    message.guildId = "";
    message.masterAddress = "";
    message.createdAt = 0n;
    message.tvlScore = "";
    message.volumeScore = "";
    message.rankByVolume = 0;
    message.rankByTvl = 0;
    message.logo = "";
    message.totalTvl = "";
    message.updatedAt = 0n;
    message.name = "";
    message.isActive = false;
    message.masterBalance = "";
    message.description = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_contract */
        1:
          message.campaignContract = reader.string();
          break;
        case /* string guild_id */
        2:
          message.guildId = reader.string();
          break;
        case /* string master_address */
        3:
          message.masterAddress = reader.string();
          break;
        case /* sint64 created_at */
        4:
          message.createdAt = reader.sint64().toBigInt();
          break;
        case /* string tvl_score */
        5:
          message.tvlScore = reader.string();
          break;
        case /* string volume_score */
        6:
          message.volumeScore = reader.string();
          break;
        case /* sint32 rank_by_volume */
        7:
          message.rankByVolume = reader.sint32();
          break;
        case /* sint32 rank_by_tvl */
        8:
          message.rankByTvl = reader.sint32();
          break;
        case /* string logo */
        9:
          message.logo = reader.string();
          break;
        case /* string total_tvl */
        10:
          message.totalTvl = reader.string();
          break;
        case /* sint64 updated_at */
        11:
          message.updatedAt = reader.sint64().toBigInt();
          break;
        case /* string name */
        14:
          message.name = reader.string();
          break;
        case /* bool is_active */
        13:
          message.isActive = reader.bool();
          break;
        case /* string master_balance */
        15:
          message.masterBalance = reader.string();
          break;
        case /* string description */
        16:
          message.description = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignContract !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignContract);
    if (message.guildId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.guildId);
    if (message.masterAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.masterAddress);
    if (message.createdAt !== 0n)
      writer.tag(4, WireType.Varint).sint64(message.createdAt);
    if (message.tvlScore !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.tvlScore);
    if (message.volumeScore !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.volumeScore);
    if (message.rankByVolume !== 0)
      writer.tag(7, WireType.Varint).sint32(message.rankByVolume);
    if (message.rankByTvl !== 0)
      writer.tag(8, WireType.Varint).sint32(message.rankByTvl);
    if (message.logo !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.logo);
    if (message.totalTvl !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.totalTvl);
    if (message.updatedAt !== 0n)
      writer.tag(11, WireType.Varint).sint64(message.updatedAt);
    if (message.isActive !== false)
      writer.tag(13, WireType.Varint).bool(message.isActive);
    if (message.name !== "")
      writer.tag(14, WireType.LengthDelimited).string(message.name);
    if (message.masterBalance !== "")
      writer.tag(15, WireType.LengthDelimited).string(message.masterBalance);
    if (message.description !== "")
      writer.tag(16, WireType.LengthDelimited).string(message.description);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Guild = new Guild$Type();
class CampaignSummary$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.CampaignSummary", [
      {
        no: 1,
        name: "campaign_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "campaign_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "total_guilds_count",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 4,
        name: "total_tvl",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "total_average_tvl",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "total_volume",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "updated_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 8,
        name: "total_members_count",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 9,
        name: "start_time",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 10,
        name: "end_time",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignId = "";
    message.campaignContract = "";
    message.totalGuildsCount = 0;
    message.totalTvl = "";
    message.totalAverageTvl = "";
    message.totalVolume = "";
    message.updatedAt = 0n;
    message.totalMembersCount = 0;
    message.startTime = 0n;
    message.endTime = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_id */
        1:
          message.campaignId = reader.string();
          break;
        case /* string campaign_contract */
        2:
          message.campaignContract = reader.string();
          break;
        case /* sint32 total_guilds_count */
        3:
          message.totalGuildsCount = reader.sint32();
          break;
        case /* string total_tvl */
        4:
          message.totalTvl = reader.string();
          break;
        case /* string total_average_tvl */
        5:
          message.totalAverageTvl = reader.string();
          break;
        case /* string total_volume */
        6:
          message.totalVolume = reader.string();
          break;
        case /* sint64 updated_at */
        7:
          message.updatedAt = reader.sint64().toBigInt();
          break;
        case /* sint32 total_members_count */
        8:
          message.totalMembersCount = reader.sint32();
          break;
        case /* sint64 start_time */
        9:
          message.startTime = reader.sint64().toBigInt();
          break;
        case /* sint64 end_time */
        10:
          message.endTime = reader.sint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignId);
    if (message.campaignContract !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.campaignContract);
    if (message.totalGuildsCount !== 0)
      writer.tag(3, WireType.Varint).sint32(message.totalGuildsCount);
    if (message.totalTvl !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.totalTvl);
    if (message.totalAverageTvl !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.totalAverageTvl);
    if (message.totalVolume !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.totalVolume);
    if (message.updatedAt !== 0n)
      writer.tag(7, WireType.Varint).sint64(message.updatedAt);
    if (message.totalMembersCount !== 0)
      writer.tag(8, WireType.Varint).sint32(message.totalMembersCount);
    if (message.startTime !== 0n)
      writer.tag(9, WireType.Varint).sint64(message.startTime);
    if (message.endTime !== 0n)
      writer.tag(10, WireType.Varint).sint64(message.endTime);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CampaignSummary = new CampaignSummary$Type();
class ListGuildMembersRequest$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.ListGuildMembersRequest", [
      {
        no: 1,
        name: "campaign_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "guild_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "limit",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 4,
        name: "skip",
        kind: "scalar",
        T: 17
        /*ScalarType.SINT32*/
      },
      {
        no: 5,
        name: "include_guild_info",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 6,
        name: "sort_by",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignContract = "";
    message.guildId = "";
    message.limit = 0;
    message.skip = 0;
    message.includeGuildInfo = false;
    message.sortBy = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_contract */
        1:
          message.campaignContract = reader.string();
          break;
        case /* string guild_id */
        2:
          message.guildId = reader.string();
          break;
        case /* sint32 limit */
        3:
          message.limit = reader.sint32();
          break;
        case /* sint32 skip */
        4:
          message.skip = reader.sint32();
          break;
        case /* bool include_guild_info */
        5:
          message.includeGuildInfo = reader.bool();
          break;
        case /* string sort_by */
        6:
          message.sortBy = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignContract !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignContract);
    if (message.guildId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.guildId);
    if (message.limit !== 0)
      writer.tag(3, WireType.Varint).sint32(message.limit);
    if (message.skip !== 0)
      writer.tag(4, WireType.Varint).sint32(message.skip);
    if (message.includeGuildInfo !== false)
      writer.tag(5, WireType.Varint).bool(message.includeGuildInfo);
    if (message.sortBy !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.sortBy);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListGuildMembersRequest = new ListGuildMembersRequest$Type();
class ListGuildMembersResponse$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.ListGuildMembersResponse", [
      { no: 1, name: "members", kind: "message", repeat: 2, T: () => GuildMember },
      { no: 2, name: "paging", kind: "message", T: () => Paging },
      { no: 3, name: "guild_info", kind: "message", T: () => Guild }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.members = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective_campaign_rpc.GuildMember members */
        1:
          message.members.push(GuildMember.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* injective_campaign_rpc.Paging paging */
        2:
          message.paging = Paging.internalBinaryRead(reader, reader.uint32(), options, message.paging);
          break;
        case /* injective_campaign_rpc.Guild guild_info */
        3:
          message.guildInfo = Guild.internalBinaryRead(reader, reader.uint32(), options, message.guildInfo);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.members.length; i++)
      GuildMember.internalBinaryWrite(message.members[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.paging)
      Paging.internalBinaryWrite(message.paging, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.guildInfo)
      Guild.internalBinaryWrite(message.guildInfo, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListGuildMembersResponse = new ListGuildMembersResponse$Type();
class GuildMember$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.GuildMember", [
      {
        no: 1,
        name: "campaign_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "guild_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "joined_at",
        kind: "scalar",
        T: 18,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "tvl_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "volume_score",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "total_tvl",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "volume_score_percentage",
        kind: "scalar",
        T: 1
        /*ScalarType.DOUBLE*/
      },
      {
        no: 9,
        name: "tvl_score_percentage",
        kind: "scalar",
        T: 1
        /*ScalarType.DOUBLE*/
      },
      { no: 10, name: "tvl_reward", kind: "message", repeat: 2, T: () => Coin },
      { no: 11, name: "volume_reward", kind: "message", repeat: 2, T: () => Coin }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignContract = "";
    message.guildId = "";
    message.address = "";
    message.joinedAt = 0n;
    message.tvlScore = "";
    message.volumeScore = "";
    message.totalTvl = "";
    message.volumeScorePercentage = 0;
    message.tvlScorePercentage = 0;
    message.tvlReward = [];
    message.volumeReward = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_contract */
        1:
          message.campaignContract = reader.string();
          break;
        case /* string guild_id */
        2:
          message.guildId = reader.string();
          break;
        case /* string address */
        3:
          message.address = reader.string();
          break;
        case /* sint64 joined_at */
        4:
          message.joinedAt = reader.sint64().toBigInt();
          break;
        case /* string tvl_score */
        5:
          message.tvlScore = reader.string();
          break;
        case /* string volume_score */
        6:
          message.volumeScore = reader.string();
          break;
        case /* string total_tvl */
        7:
          message.totalTvl = reader.string();
          break;
        case /* double volume_score_percentage */
        8:
          message.volumeScorePercentage = reader.double();
          break;
        case /* double tvl_score_percentage */
        9:
          message.tvlScorePercentage = reader.double();
          break;
        case /* repeated injective_campaign_rpc.Coin tvl_reward */
        10:
          message.tvlReward.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective_campaign_rpc.Coin volume_reward */
        11:
          message.volumeReward.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignContract !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignContract);
    if (message.guildId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.guildId);
    if (message.address !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.address);
    if (message.joinedAt !== 0n)
      writer.tag(4, WireType.Varint).sint64(message.joinedAt);
    if (message.tvlScore !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.tvlScore);
    if (message.volumeScore !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.volumeScore);
    if (message.totalTvl !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.totalTvl);
    if (message.volumeScorePercentage !== 0)
      writer.tag(8, WireType.Bit64).double(message.volumeScorePercentage);
    if (message.tvlScorePercentage !== 0)
      writer.tag(9, WireType.Bit64).double(message.tvlScorePercentage);
    for (let i = 0; i < message.tvlReward.length; i++)
      Coin.internalBinaryWrite(message.tvlReward[i], writer.tag(10, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.volumeReward.length; i++)
      Coin.internalBinaryWrite(message.volumeReward[i], writer.tag(11, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GuildMember = new GuildMember$Type();
class GetGuildMemberRequest$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.GetGuildMemberRequest", [
      {
        no: 1,
        name: "campaign_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.campaignContract = "";
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string campaign_contract */
        1:
          message.campaignContract = reader.string();
          break;
        case /* string address */
        2:
          message.address = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.campaignContract !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.campaignContract);
    if (message.address !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.address);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetGuildMemberRequest = new GetGuildMemberRequest$Type();
class GetGuildMemberResponse$Type extends MessageType {
  constructor() {
    super("injective_campaign_rpc.GetGuildMemberResponse", [
      { no: 1, name: "info", kind: "message", T: () => GuildMember }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective_campaign_rpc.GuildMember info */
        1:
          message.info = GuildMember.internalBinaryRead(reader, reader.uint32(), options, message.info);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.info)
      GuildMember.internalBinaryWrite(message.info, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetGuildMemberResponse = new GetGuildMemberResponse$Type();
const InjectiveCampaignRPC = new ServiceType("injective_campaign_rpc.InjectiveCampaignRPC", [
  { name: "Ranking", options: {}, I: RankingRequest, O: RankingResponse },
  { name: "Campaigns", options: {}, I: CampaignsRequest, O: CampaignsResponse },
  { name: "CampaignsV2", options: {}, I: CampaignsV2Request, O: CampaignsV2Response },
  { name: "ListGuilds", options: {}, I: ListGuildsRequest, O: ListGuildsResponse },
  { name: "ListGuildMembers", options: {}, I: ListGuildMembersRequest, O: ListGuildMembersResponse },
  { name: "GetGuildMember", options: {}, I: GetGuildMemberRequest, O: GetGuildMemberResponse }
]);
export {
  Campaign,
  CampaignSummary,
  CampaignUser,
  CampaignV2,
  CampaignsRequest,
  CampaignsResponse,
  CampaignsV2Request,
  CampaignsV2Response,
  Coin,
  GetGuildMemberRequest,
  GetGuildMemberResponse,
  Guild,
  GuildMember,
  InjectiveCampaignRPC,
  ListGuildMembersRequest,
  ListGuildMembersResponse,
  ListGuildsRequest,
  ListGuildsResponse,
  Paging,
  RankingRequest,
  RankingResponse
};
