// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_insurance_rpc_pb.client.js
import { InjectiveInsuranceRPC } from "./injective_insurance_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveInsuranceRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveInsuranceRPC.typeName;
    this.methods = InjectiveInsuranceRPC.methods;
    this.options = InjectiveInsuranceRPC.options;
  }
  /**
   * Funds lists all insurance funds.
   *
   * @generated from protobuf rpc: Funds
   */
  funds(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Funds returns an insurance fund for a given insurance fund token denom.
   *
   * @generated from protobuf rpc: Fund
   */
  fund(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * PendingRedemptions lists all pending redemptions according to a filter
   *
   * @generated from protobuf rpc: Redemptions
   */
  redemptions(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveInsuranceRPCClient
};
