// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_exchange_rpc_pb.client.js
import { InjectiveExchangeRPC } from "./injective_exchange_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveExchangeRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveExchangeRPC.typeName;
    this.methods = InjectiveExchangeRPC.methods;
    this.options = InjectiveExchangeRPC.options;
  }
  /**
   * GetTx gets transaction details by hash.
   *
   * @generated from protobuf rpc: GetTx
   */
  getTx(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * PrepareTx generates a Web3-signable body for a Cosmos transaction
   *
   * @generated from protobuf rpc: PrepareTx
   */
  prepareTx(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * prepareEip712 generates EIP712 for an Injective Message
   *
   * @generated from protobuf rpc: PrepareEip712
   */
  prepareEip712(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * BroadcastTx broadcasts a signed Web3 transaction
   *
   * @generated from protobuf rpc: BroadcastTx
   */
  broadcastTx(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * PrepareCosmosTx generates a Web3-signable body for a Cosmos transaction
   *
   * @generated from protobuf rpc: PrepareCosmosTx
   */
  prepareCosmosTx(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * BroadcastCosmosTx broadcasts a signed Web3 transaction
   *
   * @generated from protobuf rpc: BroadcastCosmosTx
   */
  broadcastCosmosTx(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Return fee payer information's
   *
   * @generated from protobuf rpc: GetFeePayer
   */
  getFeePayer(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveExchangeRPCClient
};
