// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_referral_rpc_pb.client.js
import { InjectiveReferralRPC } from "./injective_referral_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveReferralRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveReferralRPC.typeName;
    this.methods = InjectiveReferralRPC.methods;
    this.options = InjectiveReferralRPC.options;
  }
  /**
   * Get referrer details including their invitees, commissions and trading
   * volumes
   *
   * @generated from protobuf rpc: GetReferrerDetails
   */
  getReferrerDetails(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get invitee details including their referrer, trading volume and join date
   *
   * @generated from protobuf rpc: GetInviteeDetails
   */
  getInviteeDetails(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get referrer details by their referral code
   *
   * @generated from protobuf rpc: GetReferrerByCode
   */
  getReferrerByCode(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveReferralRPCClient
};
