// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_accounts_rpc_pb.client.js
import { InjectiveAccountsRPC } from "./injective_accounts_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveAccountsRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveAccountsRPC.typeName;
    this.methods = InjectiveAccountsRPC.methods;
    this.options = InjectiveAccountsRPC.options;
  }
  /**
   * Provide the account's portfolio value in USD.
   *
   * @generated from protobuf rpc: Portfolio
   */
  portfolio(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List order states by order hashes
   *
   * @generated from protobuf rpc: OrderStates
   */
  orderStates(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List all subaccounts IDs of an account address
   *
   * @generated from protobuf rpc: SubaccountsList
   */
  subaccountsList(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List subaccount balances for the provided denoms.
   *
   * @generated from protobuf rpc: SubaccountBalancesList
   */
  subaccountBalancesList(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets a balance for specific coin denom
   *
   * @generated from protobuf rpc: SubaccountBalanceEndpoint
   */
  subaccountBalanceEndpoint(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StreamSubaccountBalance streams new balance changes for a specified
   * subaccount and denoms. If no denoms are provided, all denom changes are
   * streamed.
   *
   * @generated from protobuf rpc: StreamSubaccountBalance
   */
  streamSubaccountBalance(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * Get subaccount's deposits and withdrawals history
   *
   * @generated from protobuf rpc: SubaccountHistory
   */
  subaccountHistory(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get subaccount's orders summary
   *
   * @generated from protobuf rpc: SubaccountOrderSummary
   */
  subaccountOrderSummary(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide historical trading rewards
   *
   * @generated from protobuf rpc: Rewards
   */
  rewards(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Stream live data for an account and respective data
   *
   * @generated from protobuf rpc: StreamAccountData
   */
  streamAccountData(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
}
export {
  InjectiveAccountsRPCClient
};
