// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_oracle_rpc_pb.client.js
import { InjectiveOracleRPC } from "./injective_oracle_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveOracleRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveOracleRPC.typeName;
    this.methods = InjectiveOracleRPC.methods;
    this.options = InjectiveOracleRPC.options;
  }
  /**
   * List all oracles
   *
   * @generated from protobuf rpc: OracleList
   */
  oracleList(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets the price of the oracle
   *
   * @generated from protobuf rpc: Price
   */
  price(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Gets prices of the oracle
   *
   * @generated from protobuf rpc: PriceV2
   */
  priceV2(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StreamPrices streams new price changes for a specified oracle. If no oracles
   * are provided, all price changes are streamed.
   *
   * @generated from protobuf rpc: StreamPrices
   */
  streamPrices(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
  /**
   * StreamPrices streams new price changes markets
   *
   * @generated from protobuf rpc: StreamPricesByMarkets
   */
  streamPricesByMarkets(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
}
export {
  InjectiveOracleRPCClient
};
