// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_trading_rpc_pb.client.js
import { InjectiveTradingRPC } from "./injective_trading_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveTradingRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveTradingRPC.typeName;
    this.methods = InjectiveTradingRPC.methods;
    this.options = InjectiveTradingRPC.options;
  }
  /**
   * Lists all trading strategies
   *
   * @generated from protobuf rpc: ListTradingStrategies
   */
  listTradingStrategies(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetStats returns global statistics in the last 24hs
   *
   * @generated from protobuf rpc: GetTradingStats
   */
  getTradingStats(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StreamStrategy streams the trading strategies on state change
   *
   * @generated from protobuf rpc: StreamStrategy
   */
  streamStrategy(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("serverStreaming", this._transport, method, opt, input);
  }
}
export {
  InjectiveTradingRPCClient
};
