// @injectivelabs/indexer-proto-ts-v2/esm/generated/injective_megavault_rpc_pb.client.js
import { InjectiveMegavaultRPC } from "./injective_megavault_rpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class InjectiveMegavaultRPCClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = InjectiveMegavaultRPC.typeName;
    this.methods = InjectiveMegavaultRPC.methods;
    this.options = InjectiveMegavaultRPC.options;
  }
  /**
   * Get the vault information
   *
   * @generated from protobuf rpc: GetVault
   */
  getVault(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get the user information for a vault
   *
   * @generated from protobuf rpc: GetUser
   */
  getUser(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List the subscriptions for a vault
   *
   * @generated from protobuf rpc: ListSubscriptions
   */
  listSubscriptions(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * List the redemptions for a vault
   *
   * @generated from protobuf rpc: ListRedemptions
   */
  listRedemptions(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Get the redemptions buckets for an operator
   *
   * @generated from protobuf rpc: GetOperatorRedemptionBuckets
   */
  getOperatorRedemptionBuckets(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide historical TVL data.
   *
   * @generated from protobuf rpc: TvlHistory
   */
  tvlHistory(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Provide historical PnL data.
   *
   * @generated from protobuf rpc: PnlHistory
   */
  pnlHistory(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  InjectiveMegavaultRPCClient
};
