// @injectivelabs/core-proto-ts-v2/esm/generated/injective/peggy/v1/attestation_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../google/protobuf/any_pb.js";
var ClaimType = /* @__PURE__ */ ((ClaimType2) => {
  ClaimType2[ClaimType2["UNKNOWN"] = 0] = "UNKNOWN";
  ClaimType2[ClaimType2["DEPOSIT"] = 1] = "DEPOSIT";
  ClaimType2[ClaimType2["WITHDRAW"] = 2] = "WITHDRAW";
  ClaimType2[ClaimType2["ERC20_DEPLOYED"] = 3] = "ERC20_DEPLOYED";
  ClaimType2[ClaimType2["VALSET_UPDATED"] = 4] = "VALSET_UPDATED";
  return ClaimType2;
})(ClaimType || {});
class Attestation$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.Attestation", [
      {
        no: 1,
        name: "observed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 2,
        name: "votes",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "claim", kind: "message", T: () => Any }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.observed = false;
    message.votes = [];
    message.height = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bool observed */
        1:
          message.observed = reader.bool();
          break;
        case /* repeated string votes */
        2:
          message.votes.push(reader.string());
          break;
        case /* uint64 height */
        3:
          message.height = reader.uint64().toBigInt();
          break;
        case /* google.protobuf.Any claim */
        4:
          message.claim = Any.internalBinaryRead(reader, reader.uint32(), options, message.claim);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.observed !== false)
      writer.tag(1, WireType.Varint).bool(message.observed);
    for (let i = 0; i < message.votes.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.votes[i]);
    if (message.height !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.height);
    if (message.claim)
      Any.internalBinaryWrite(message.claim, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Attestation = new Attestation$Type();
class ERC20Token$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.ERC20Token", [
      {
        no: 1,
        name: "contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "amount", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contract = "";
    message.amount = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract */
        1:
          message.contract = reader.string();
          break;
        case /* string amount */
        2:
          message.amount = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contract !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contract);
    if (message.amount !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.amount);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ERC20Token = new ERC20Token$Type();
export {
  Attestation,
  ClaimType,
  ERC20Token
};
