// @injectivelabs/core-proto-ts-v2/esm/generated/injective/peggy/v1/batch_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ERC20Token } from "./attestation_pb.js";
class OutgoingTxBatch$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.OutgoingTxBatch", [
      {
        no: 1,
        name: "batch_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "batch_timeout",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "transactions", kind: "message", repeat: 2, T: () => OutgoingTransferTx },
      {
        no: 4,
        name: "token_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "block",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.batchNonce = 0n;
    message.batchTimeout = 0n;
    message.transactions = [];
    message.tokenContract = "";
    message.block = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 batch_nonce */
        1:
          message.batchNonce = reader.uint64().toBigInt();
          break;
        case /* uint64 batch_timeout */
        2:
          message.batchTimeout = reader.uint64().toBigInt();
          break;
        case /* repeated injective.peggy.v1.OutgoingTransferTx transactions */
        3:
          message.transactions.push(OutgoingTransferTx.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string token_contract */
        4:
          message.tokenContract = reader.string();
          break;
        case /* uint64 block */
        5:
          message.block = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.batchNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.batchNonce);
    if (message.batchTimeout !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.batchTimeout);
    for (let i = 0; i < message.transactions.length; i++)
      OutgoingTransferTx.internalBinaryWrite(message.transactions[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.tokenContract !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.tokenContract);
    if (message.block !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.block);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const OutgoingTxBatch = new OutgoingTxBatch$Type();
class OutgoingTransferTx$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.OutgoingTransferTx", [
      {
        no: 1,
        name: "id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "dest_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "erc20_token", kind: "message", T: () => ERC20Token },
      { no: 5, name: "erc20_fee", kind: "message", T: () => ERC20Token }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.id = 0n;
    message.sender = "";
    message.destAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 id */
        1:
          message.id = reader.uint64().toBigInt();
          break;
        case /* string sender */
        2:
          message.sender = reader.string();
          break;
        case /* string dest_address */
        3:
          message.destAddress = reader.string();
          break;
        case /* injective.peggy.v1.ERC20Token erc20_token */
        4:
          message.erc20Token = ERC20Token.internalBinaryRead(reader, reader.uint32(), options, message.erc20Token);
          break;
        case /* injective.peggy.v1.ERC20Token erc20_fee */
        5:
          message.erc20Fee = ERC20Token.internalBinaryRead(reader, reader.uint32(), options, message.erc20Fee);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.id !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.id);
    if (message.sender !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.sender);
    if (message.destAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.destAddress);
    if (message.erc20Token)
      ERC20Token.internalBinaryWrite(message.erc20Token, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.erc20Fee)
      ERC20Token.internalBinaryWrite(message.erc20Fee, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const OutgoingTransferTx = new OutgoingTransferTx$Type();
export {
  OutgoingTransferTx,
  OutgoingTxBatch
};
