// @injectivelabs/core-proto-ts-v2/esm/generated/injective/peggy/v1/events_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { BridgeValidator } from "./types_pb.js";
import { ClaimType } from "./attestation_pb.js";
var JailReason = /* @__PURE__ */ ((JailReason2) => {
  JailReason2[JailReason2["MissingValsetConfirm"] = 0] = "MissingValsetConfirm";
  JailReason2[JailReason2["MissingBatchConfirm"] = 1] = "MissingBatchConfirm";
  return JailReason2;
})(JailReason || {});
class EventAttestationObserved$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventAttestationObserved", [
      { no: 1, name: "attestation_type", kind: "enum", T: () => ["injective.peggy.v1.ClaimType", ClaimType, "CLAIM_TYPE_"] },
      {
        no: 2,
        name: "bridge_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "bridge_chain_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "attestation_id",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.attestationType = 0;
    message.bridgeContract = "";
    message.bridgeChainId = 0n;
    message.attestationId = new Uint8Array(0);
    message.nonce = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.ClaimType attestation_type */
        1:
          message.attestationType = reader.int32();
          break;
        case /* string bridge_contract */
        2:
          message.bridgeContract = reader.string();
          break;
        case /* uint64 bridge_chain_id */
        3:
          message.bridgeChainId = reader.uint64().toBigInt();
          break;
        case /* bytes attestation_id */
        4:
          message.attestationId = reader.bytes();
          break;
        case /* uint64 nonce */
        5:
          message.nonce = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.attestationType !== 0)
      writer.tag(1, WireType.Varint).int32(message.attestationType);
    if (message.bridgeContract !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.bridgeContract);
    if (message.bridgeChainId !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.bridgeChainId);
    if (message.attestationId.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.attestationId);
    if (message.nonce !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.nonce);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventAttestationObserved = new EventAttestationObserved$Type();
class EventBridgeWithdrawCanceled$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventBridgeWithdrawCanceled", [
      {
        no: 1,
        name: "bridge_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "bridge_chain_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.bridgeContract = "";
    message.bridgeChainId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string bridge_contract */
        1:
          message.bridgeContract = reader.string();
          break;
        case /* uint64 bridge_chain_id */
        2:
          message.bridgeChainId = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.bridgeContract !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.bridgeContract);
    if (message.bridgeChainId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.bridgeChainId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventBridgeWithdrawCanceled = new EventBridgeWithdrawCanceled$Type();
class EventOutgoingBatch$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventOutgoingBatch", [
      {
        no: 1,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "batch_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "batch_timeout",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "batch_tx_ids",
        kind: "scalar",
        repeat: 1,
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.denom = "";
    message.orchestratorAddress = "";
    message.batchNonce = 0n;
    message.batchTimeout = 0n;
    message.batchTxIds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string denom */
        1:
          message.denom = reader.string();
          break;
        case /* string orchestrator_address */
        2:
          message.orchestratorAddress = reader.string();
          break;
        case /* uint64 batch_nonce */
        3:
          message.batchNonce = reader.uint64().toBigInt();
          break;
        case /* uint64 batch_timeout */
        4:
          message.batchTimeout = reader.uint64().toBigInt();
          break;
        case /* repeated uint64 batch_tx_ids */
        5:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.batchTxIds.push(reader.uint64().toBigInt());
          else
            message.batchTxIds.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.denom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.denom);
    if (message.orchestratorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.orchestratorAddress);
    if (message.batchNonce !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.batchNonce);
    if (message.batchTimeout !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.batchTimeout);
    if (message.batchTxIds.length) {
      writer.tag(5, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.batchTxIds.length; i++)
        writer.uint64(message.batchTxIds[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventOutgoingBatch = new EventOutgoingBatch$Type();
class EventOutgoingBatchCanceled$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventOutgoingBatchCanceled", [
      {
        no: 1,
        name: "bridge_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "bridge_chain_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "batch_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.bridgeContract = "";
    message.bridgeChainId = 0n;
    message.batchId = 0n;
    message.nonce = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string bridge_contract */
        1:
          message.bridgeContract = reader.string();
          break;
        case /* uint64 bridge_chain_id */
        2:
          message.bridgeChainId = reader.uint64().toBigInt();
          break;
        case /* uint64 batch_id */
        3:
          message.batchId = reader.uint64().toBigInt();
          break;
        case /* uint64 nonce */
        4:
          message.nonce = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.bridgeContract !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.bridgeContract);
    if (message.bridgeChainId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.bridgeChainId);
    if (message.batchId !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.batchId);
    if (message.nonce !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.nonce);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventOutgoingBatchCanceled = new EventOutgoingBatchCanceled$Type();
class EventValsetUpdateRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventValsetUpdateRequest", [
      {
        no: 1,
        name: "valset_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "valset_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "valset_members", kind: "message", repeat: 2, T: () => BridgeValidator },
      { no: 4, name: "reward_amount", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      {
        no: 5,
        name: "reward_token",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.valsetNonce = 0n;
    message.valsetHeight = 0n;
    message.valsetMembers = [];
    message.rewardAmount = "";
    message.rewardToken = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 valset_nonce */
        1:
          message.valsetNonce = reader.uint64().toBigInt();
          break;
        case /* uint64 valset_height */
        2:
          message.valsetHeight = reader.uint64().toBigInt();
          break;
        case /* repeated injective.peggy.v1.BridgeValidator valset_members */
        3:
          message.valsetMembers.push(BridgeValidator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string reward_amount */
        4:
          message.rewardAmount = reader.string();
          break;
        case /* string reward_token */
        5:
          message.rewardToken = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.valsetNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.valsetNonce);
    if (message.valsetHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.valsetHeight);
    for (let i = 0; i < message.valsetMembers.length; i++)
      BridgeValidator.internalBinaryWrite(message.valsetMembers[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.rewardAmount !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.rewardAmount);
    if (message.rewardToken !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.rewardToken);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventValsetUpdateRequest = new EventValsetUpdateRequest$Type();
class EventSetOrchestratorAddresses$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventSetOrchestratorAddresses", [
      {
        no: 1,
        name: "validator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "operator_eth_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validatorAddress = "";
    message.orchestratorAddress = "";
    message.operatorEthAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string validator_address */
        1:
          message.validatorAddress = reader.string();
          break;
        case /* string orchestrator_address */
        2:
          message.orchestratorAddress = reader.string();
          break;
        case /* string operator_eth_address */
        3:
          message.operatorEthAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.orchestratorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.orchestratorAddress);
    if (message.operatorEthAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.operatorEthAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventSetOrchestratorAddresses = new EventSetOrchestratorAddresses$Type();
class EventValsetConfirm$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventValsetConfirm", [
      {
        no: 1,
        name: "valset_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.valsetNonce = 0n;
    message.orchestratorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 valset_nonce */
        1:
          message.valsetNonce = reader.uint64().toBigInt();
          break;
        case /* string orchestrator_address */
        2:
          message.orchestratorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.valsetNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.valsetNonce);
    if (message.orchestratorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.orchestratorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventValsetConfirm = new EventValsetConfirm$Type();
class EventSendToEth$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventSendToEth", [
      {
        no: 1,
        name: "outgoing_tx_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "receiver",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "amount", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "github.com/cosmos/cosmos-sdk/types.Coin" } },
      { no: 5, name: "bridge_fee", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "github.com/cosmos/cosmos-sdk/types.Coin" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.outgoingTxId = 0n;
    message.sender = "";
    message.receiver = "";
    message.amount = "";
    message.bridgeFee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 outgoing_tx_id */
        1:
          message.outgoingTxId = reader.uint64().toBigInt();
          break;
        case /* string sender */
        2:
          message.sender = reader.string();
          break;
        case /* string receiver */
        3:
          message.receiver = reader.string();
          break;
        case /* string amount */
        4:
          message.amount = reader.string();
          break;
        case /* string bridge_fee */
        5:
          message.bridgeFee = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.outgoingTxId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.outgoingTxId);
    if (message.sender !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.sender);
    if (message.receiver !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.receiver);
    if (message.amount !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.amount);
    if (message.bridgeFee !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.bridgeFee);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventSendToEth = new EventSendToEth$Type();
class EventConfirmBatch$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventConfirmBatch", [
      {
        no: 1,
        name: "batch_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.batchNonce = 0n;
    message.orchestratorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 batch_nonce */
        1:
          message.batchNonce = reader.uint64().toBigInt();
          break;
        case /* string orchestrator_address */
        2:
          message.orchestratorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.batchNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.batchNonce);
    if (message.orchestratorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.orchestratorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventConfirmBatch = new EventConfirmBatch$Type();
class EventAttestationVote$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventAttestationVote", [
      {
        no: 1,
        name: "event_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "attestation_id",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "voter",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.eventNonce = 0n;
    message.attestationId = new Uint8Array(0);
    message.voter = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 event_nonce */
        1:
          message.eventNonce = reader.uint64().toBigInt();
          break;
        case /* bytes attestation_id */
        2:
          message.attestationId = reader.bytes();
          break;
        case /* string voter */
        3:
          message.voter = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.eventNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.eventNonce);
    if (message.attestationId.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.attestationId);
    if (message.voter !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.voter);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventAttestationVote = new EventAttestationVote$Type();
class EventDepositClaim$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventDepositClaim", [
      {
        no: 1,
        name: "event_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "event_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "attestation_id",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "ethereum_sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "cosmos_receiver",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "token_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 7, name: "amount", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      {
        no: 8,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "data",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.eventNonce = 0n;
    message.eventHeight = 0n;
    message.attestationId = new Uint8Array(0);
    message.ethereumSender = "";
    message.cosmosReceiver = "";
    message.tokenContract = "";
    message.amount = "";
    message.orchestratorAddress = "";
    message.data = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 event_nonce */
        1:
          message.eventNonce = reader.uint64().toBigInt();
          break;
        case /* uint64 event_height */
        2:
          message.eventHeight = reader.uint64().toBigInt();
          break;
        case /* bytes attestation_id */
        3:
          message.attestationId = reader.bytes();
          break;
        case /* string ethereum_sender */
        4:
          message.ethereumSender = reader.string();
          break;
        case /* string cosmos_receiver */
        5:
          message.cosmosReceiver = reader.string();
          break;
        case /* string token_contract */
        6:
          message.tokenContract = reader.string();
          break;
        case /* string amount */
        7:
          message.amount = reader.string();
          break;
        case /* string orchestrator_address */
        8:
          message.orchestratorAddress = reader.string();
          break;
        case /* string data */
        9:
          message.data = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.eventNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.eventNonce);
    if (message.eventHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.eventHeight);
    if (message.attestationId.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.attestationId);
    if (message.ethereumSender !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.ethereumSender);
    if (message.cosmosReceiver !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.cosmosReceiver);
    if (message.tokenContract !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.tokenContract);
    if (message.amount !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.amount);
    if (message.orchestratorAddress !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.orchestratorAddress);
    if (message.data !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventDepositClaim = new EventDepositClaim$Type();
class EventWithdrawClaim$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventWithdrawClaim", [
      {
        no: 1,
        name: "event_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "event_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "attestation_id",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "batch_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "token_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.eventNonce = 0n;
    message.eventHeight = 0n;
    message.attestationId = new Uint8Array(0);
    message.batchNonce = 0n;
    message.tokenContract = "";
    message.orchestratorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 event_nonce */
        1:
          message.eventNonce = reader.uint64().toBigInt();
          break;
        case /* uint64 event_height */
        2:
          message.eventHeight = reader.uint64().toBigInt();
          break;
        case /* bytes attestation_id */
        3:
          message.attestationId = reader.bytes();
          break;
        case /* uint64 batch_nonce */
        4:
          message.batchNonce = reader.uint64().toBigInt();
          break;
        case /* string token_contract */
        5:
          message.tokenContract = reader.string();
          break;
        case /* string orchestrator_address */
        6:
          message.orchestratorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.eventNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.eventNonce);
    if (message.eventHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.eventHeight);
    if (message.attestationId.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.attestationId);
    if (message.batchNonce !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.batchNonce);
    if (message.tokenContract !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.tokenContract);
    if (message.orchestratorAddress !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.orchestratorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventWithdrawClaim = new EventWithdrawClaim$Type();
class EventERC20DeployedClaim$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventERC20DeployedClaim", [
      {
        no: 1,
        name: "event_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "event_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "attestation_id",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "cosmos_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "token_contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "symbol",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "decimals",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 9,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.eventNonce = 0n;
    message.eventHeight = 0n;
    message.attestationId = new Uint8Array(0);
    message.cosmosDenom = "";
    message.tokenContract = "";
    message.name = "";
    message.symbol = "";
    message.decimals = 0n;
    message.orchestratorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 event_nonce */
        1:
          message.eventNonce = reader.uint64().toBigInt();
          break;
        case /* uint64 event_height */
        2:
          message.eventHeight = reader.uint64().toBigInt();
          break;
        case /* bytes attestation_id */
        3:
          message.attestationId = reader.bytes();
          break;
        case /* string cosmos_denom */
        4:
          message.cosmosDenom = reader.string();
          break;
        case /* string token_contract */
        5:
          message.tokenContract = reader.string();
          break;
        case /* string name */
        6:
          message.name = reader.string();
          break;
        case /* string symbol */
        7:
          message.symbol = reader.string();
          break;
        case /* uint64 decimals */
        8:
          message.decimals = reader.uint64().toBigInt();
          break;
        case /* string orchestrator_address */
        9:
          message.orchestratorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.eventNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.eventNonce);
    if (message.eventHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.eventHeight);
    if (message.attestationId.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.attestationId);
    if (message.cosmosDenom !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.cosmosDenom);
    if (message.tokenContract !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.tokenContract);
    if (message.name !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.name);
    if (message.symbol !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.symbol);
    if (message.decimals !== 0n)
      writer.tag(8, WireType.Varint).uint64(message.decimals);
    if (message.orchestratorAddress !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.orchestratorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventERC20DeployedClaim = new EventERC20DeployedClaim$Type();
class EventValsetUpdateClaim$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventValsetUpdateClaim", [
      {
        no: 1,
        name: "event_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "event_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "attestation_id",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "valset_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 5, name: "valset_members", kind: "message", repeat: 2, T: () => BridgeValidator },
      { no: 6, name: "reward_amount", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      {
        no: 7,
        name: "reward_token",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.eventNonce = 0n;
    message.eventHeight = 0n;
    message.attestationId = new Uint8Array(0);
    message.valsetNonce = 0n;
    message.valsetMembers = [];
    message.rewardAmount = "";
    message.rewardToken = "";
    message.orchestratorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 event_nonce */
        1:
          message.eventNonce = reader.uint64().toBigInt();
          break;
        case /* uint64 event_height */
        2:
          message.eventHeight = reader.uint64().toBigInt();
          break;
        case /* bytes attestation_id */
        3:
          message.attestationId = reader.bytes();
          break;
        case /* uint64 valset_nonce */
        4:
          message.valsetNonce = reader.uint64().toBigInt();
          break;
        case /* repeated injective.peggy.v1.BridgeValidator valset_members */
        5:
          message.valsetMembers.push(BridgeValidator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string reward_amount */
        6:
          message.rewardAmount = reader.string();
          break;
        case /* string reward_token */
        7:
          message.rewardToken = reader.string();
          break;
        case /* string orchestrator_address */
        8:
          message.orchestratorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.eventNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.eventNonce);
    if (message.eventHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.eventHeight);
    if (message.attestationId.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.attestationId);
    if (message.valsetNonce !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.valsetNonce);
    for (let i = 0; i < message.valsetMembers.length; i++)
      BridgeValidator.internalBinaryWrite(message.valsetMembers[i], writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.rewardAmount !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.rewardAmount);
    if (message.rewardToken !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.rewardToken);
    if (message.orchestratorAddress !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.orchestratorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventValsetUpdateClaim = new EventValsetUpdateClaim$Type();
class EventCancelSendToEth$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventCancelSendToEth", [
      {
        no: 1,
        name: "outgoing_tx_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.outgoingTxId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 outgoing_tx_id */
        1:
          message.outgoingTxId = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.outgoingTxId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.outgoingTxId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventCancelSendToEth = new EventCancelSendToEth$Type();
class EventSubmitBadSignatureEvidence$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventSubmitBadSignatureEvidence", [
      {
        no: 1,
        name: "bad_eth_signature",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "bad_eth_signature_subject",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.badEthSignature = "";
    message.badEthSignatureSubject = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string bad_eth_signature */
        1:
          message.badEthSignature = reader.string();
          break;
        case /* string bad_eth_signature_subject */
        2:
          message.badEthSignatureSubject = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.badEthSignature !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.badEthSignature);
    if (message.badEthSignatureSubject !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.badEthSignatureSubject);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventSubmitBadSignatureEvidence = new EventSubmitBadSignatureEvidence$Type();
class EventValidatorSlash$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventValidatorSlash", [
      {
        no: 1,
        name: "power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "reason",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "consensus_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "operator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "moniker",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.power = 0n;
    message.reason = "";
    message.consensusAddress = "";
    message.operatorAddress = "";
    message.moniker = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 power */
        1:
          message.power = reader.int64().toBigInt();
          break;
        case /* string reason */
        2:
          message.reason = reader.string();
          break;
        case /* string consensus_address */
        3:
          message.consensusAddress = reader.string();
          break;
        case /* string operator_address */
        4:
          message.operatorAddress = reader.string();
          break;
        case /* string moniker */
        5:
          message.moniker = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.power !== 0n)
      writer.tag(1, WireType.Varint).int64(message.power);
    if (message.reason !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.reason);
    if (message.consensusAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.consensusAddress);
    if (message.operatorAddress !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.operatorAddress);
    if (message.moniker !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.moniker);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventValidatorSlash = new EventValidatorSlash$Type();
class EventDepositReceived$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventDepositReceived", [
      {
        no: 1,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "receiver",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "amount", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "github.com/cosmos/cosmos-sdk/types.Coin" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.receiver = "";
    message.amount = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string receiver */
        2:
          message.receiver = reader.string();
          break;
        case /* string amount */
        3:
          message.amount = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.receiver !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.receiver);
    if (message.amount !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.amount);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventDepositReceived = new EventDepositReceived$Type();
class EventWithdrawalsCompleted$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventWithdrawalsCompleted", [
      {
        no: 1,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "withdrawals", kind: "message", repeat: 2, T: () => Withdrawal }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.denom = "";
    message.withdrawals = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string denom */
        1:
          message.denom = reader.string();
          break;
        case /* repeated injective.peggy.v1.Withdrawal withdrawals */
        2:
          message.withdrawals.push(Withdrawal.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.denom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.denom);
    for (let i = 0; i < message.withdrawals.length; i++)
      Withdrawal.internalBinaryWrite(message.withdrawals[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventWithdrawalsCompleted = new EventWithdrawalsCompleted$Type();
class Withdrawal$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.Withdrawal", [
      {
        no: 1,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "receiver",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "amount", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.receiver = "";
    message.amount = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string receiver */
        2:
          message.receiver = reader.string();
          break;
        case /* string amount */
        3:
          message.amount = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.receiver !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.receiver);
    if (message.amount !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.amount);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Withdrawal = new Withdrawal$Type();
class EventValidatorJailed$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.EventValidatorJailed", [
      { no: 1, name: "reason", kind: "enum", T: () => ["injective.peggy.v1.JailReason", JailReason] },
      {
        no: 2,
        name: "power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "consensus_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "operator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "moniker",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.reason = 0;
    message.power = 0n;
    message.consensusAddress = "";
    message.operatorAddress = "";
    message.moniker = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.JailReason reason */
        1:
          message.reason = reader.int32();
          break;
        case /* int64 power */
        2:
          message.power = reader.int64().toBigInt();
          break;
        case /* string consensus_address */
        3:
          message.consensusAddress = reader.string();
          break;
        case /* string operator_address */
        4:
          message.operatorAddress = reader.string();
          break;
        case /* string moniker */
        5:
          message.moniker = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.reason !== 0)
      writer.tag(1, WireType.Varint).int32(message.reason);
    if (message.power !== 0n)
      writer.tag(2, WireType.Varint).int64(message.power);
    if (message.consensusAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.consensusAddress);
    if (message.operatorAddress !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.operatorAddress);
    if (message.moniker !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.moniker);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventValidatorJailed = new EventValidatorJailed$Type();
export {
  EventAttestationObserved,
  EventAttestationVote,
  EventBridgeWithdrawCanceled,
  EventCancelSendToEth,
  EventConfirmBatch,
  EventDepositClaim,
  EventDepositReceived,
  EventERC20DeployedClaim,
  EventOutgoingBatch,
  EventOutgoingBatchCanceled,
  EventSendToEth,
  EventSetOrchestratorAddresses,
  EventSubmitBadSignatureEvidence,
  EventValidatorJailed,
  EventValidatorSlash,
  EventValsetConfirm,
  EventValsetUpdateClaim,
  EventValsetUpdateRequest,
  EventWithdrawClaim,
  EventWithdrawalsCompleted,
  JailReason,
  Withdrawal
};
