// @injectivelabs/core-proto-ts-v2/esm/generated/injective/peggy/v1/msgs_pb.client.js
import { Msg } from "./msgs_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * @generated from protobuf rpc: ValsetConfirm
   */
  valsetConfirm(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: SendToEth
   */
  sendToEth(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: RequestBatch
   */
  requestBatch(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: ConfirmBatch
   */
  confirmBatch(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: DepositClaim
   */
  depositClaim(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: WithdrawClaim
   */
  withdrawClaim(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: ValsetUpdateClaim
   */
  valsetUpdateClaim(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: ERC20DeployedClaim
   */
  eRC20DeployedClaim(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: SetOrchestratorAddresses
   */
  setOrchestratorAddresses(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: CancelSendToEth
   */
  cancelSendToEth(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: SubmitBadSignatureEvidence
   */
  submitBadSignatureEvidence(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: UpdateParams
   */
  updateParams(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * BlacklistEthereumAddresses adds Ethereum addresses to the peggy blacklist.
   *
   * @generated from protobuf rpc: BlacklistEthereumAddresses
   */
  blacklistEthereumAddresses(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * RevokeEthereumBlacklist removes Ethereum addresses from the peggy
   * blacklist.
   *
   * @generated from protobuf rpc: RevokeEthereumBlacklist
   */
  revokeEthereumBlacklist(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   *  CreateRateLimit imposes a (notional) limit on withdrawals for a particular
   *  Peggy asset
   *
   * @generated from protobuf rpc: CreateRateLimit
   */
  createRateLimit(input, options) {
    const method = this.methods[14], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   *  UpdateRateLimit updates the rate limit's metadata for a particular Peggy
   *  asset
   *
   * @generated from protobuf rpc: UpdateRateLimit
   */
  updateRateLimit(input, options) {
    const method = this.methods[15], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   *  RemoveRateLimit lifts the rate limit for a particular Peggy asset
   *
   * @generated from protobuf rpc: RemoveRateLimit
   */
  removeRateLimit(input, options) {
    const method = this.methods[16], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
