// @injectivelabs/core-proto-ts-v2/esm/generated/injective/peggy/v1/query_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { GenesisState } from "./genesis_pb.js";
import { OutgoingTransferTx } from "./batch_pb.js";
import { LastClaimEvent } from "./types_pb.js";
import { MsgConfirmBatch } from "./msgs_pb.js";
import { OutgoingTxBatch } from "./batch_pb.js";
import { BatchFees } from "./pool_pb.js";
import { MsgValsetConfirm } from "./msgs_pb.js";
import { Valset } from "./types_pb.js";
import { Params } from "./params_pb.js";
class QueryParamsRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryParamsRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryParamsRequest = new QueryParamsRequest$Type();
class QueryParamsResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryParamsResponse", [
      { no: 1, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryParamsResponse = new QueryParamsResponse$Type();
class QueryCurrentValsetRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryCurrentValsetRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryCurrentValsetRequest = new QueryCurrentValsetRequest$Type();
class QueryCurrentValsetResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryCurrentValsetResponse", [
      { no: 1, name: "valset", kind: "message", T: () => Valset }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.Valset valset */
        1:
          message.valset = Valset.internalBinaryRead(reader, reader.uint32(), options, message.valset);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.valset)
      Valset.internalBinaryWrite(message.valset, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryCurrentValsetResponse = new QueryCurrentValsetResponse$Type();
class QueryValsetRequestRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryValsetRequestRequest", [
      {
        no: 1,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nonce = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 nonce */
        1:
          message.nonce = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.nonce);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValsetRequestRequest = new QueryValsetRequestRequest$Type();
class QueryValsetRequestResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryValsetRequestResponse", [
      { no: 1, name: "valset", kind: "message", T: () => Valset }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.Valset valset */
        1:
          message.valset = Valset.internalBinaryRead(reader, reader.uint32(), options, message.valset);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.valset)
      Valset.internalBinaryWrite(message.valset, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValsetRequestResponse = new QueryValsetRequestResponse$Type();
class QueryValsetConfirmRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryValsetConfirmRequest", [
      {
        no: 1,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nonce = 0n;
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 nonce */
        1:
          message.nonce = reader.uint64().toBigInt();
          break;
        case /* string address */
        2:
          message.address = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.nonce);
    if (message.address !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.address);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValsetConfirmRequest = new QueryValsetConfirmRequest$Type();
class QueryValsetConfirmResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryValsetConfirmResponse", [
      { no: 1, name: "confirm", kind: "message", T: () => MsgValsetConfirm }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.MsgValsetConfirm confirm */
        1:
          message.confirm = MsgValsetConfirm.internalBinaryRead(reader, reader.uint32(), options, message.confirm);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.confirm)
      MsgValsetConfirm.internalBinaryWrite(message.confirm, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValsetConfirmResponse = new QueryValsetConfirmResponse$Type();
class QueryValsetConfirmsByNonceRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryValsetConfirmsByNonceRequest", [
      {
        no: 1,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nonce = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 nonce */
        1:
          message.nonce = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.nonce);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValsetConfirmsByNonceRequest = new QueryValsetConfirmsByNonceRequest$Type();
class QueryValsetConfirmsByNonceResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryValsetConfirmsByNonceResponse", [
      { no: 1, name: "confirms", kind: "message", repeat: 2, T: () => MsgValsetConfirm }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.confirms = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective.peggy.v1.MsgValsetConfirm confirms */
        1:
          message.confirms.push(MsgValsetConfirm.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.confirms.length; i++)
      MsgValsetConfirm.internalBinaryWrite(message.confirms[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValsetConfirmsByNonceResponse = new QueryValsetConfirmsByNonceResponse$Type();
class QueryLastValsetRequestsRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryLastValsetRequestsRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLastValsetRequestsRequest = new QueryLastValsetRequestsRequest$Type();
class QueryLastValsetRequestsResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryLastValsetRequestsResponse", [
      { no: 1, name: "valsets", kind: "message", repeat: 2, T: () => Valset }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.valsets = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective.peggy.v1.Valset valsets */
        1:
          message.valsets.push(Valset.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.valsets.length; i++)
      Valset.internalBinaryWrite(message.valsets[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLastValsetRequestsResponse = new QueryLastValsetRequestsResponse$Type();
class QueryLastPendingValsetRequestByAddrRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryLastPendingValsetRequestByAddrRequest", [
      {
        no: 1,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLastPendingValsetRequestByAddrRequest = new QueryLastPendingValsetRequestByAddrRequest$Type();
class QueryLastPendingValsetRequestByAddrResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryLastPendingValsetRequestByAddrResponse", [
      { no: 1, name: "valsets", kind: "message", repeat: 2, T: () => Valset }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.valsets = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective.peggy.v1.Valset valsets */
        1:
          message.valsets.push(Valset.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.valsets.length; i++)
      Valset.internalBinaryWrite(message.valsets[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLastPendingValsetRequestByAddrResponse = new QueryLastPendingValsetRequestByAddrResponse$Type();
class QueryBatchFeeRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryBatchFeeRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryBatchFeeRequest = new QueryBatchFeeRequest$Type();
class QueryBatchFeeResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryBatchFeeResponse", [
      { no: 1, name: "batchFees", kind: "message", repeat: 2, T: () => BatchFees }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.batchFees = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective.peggy.v1.BatchFees batchFees */
        1:
          message.batchFees.push(BatchFees.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.batchFees.length; i++)
      BatchFees.internalBinaryWrite(message.batchFees[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryBatchFeeResponse = new QueryBatchFeeResponse$Type();
class QueryLastPendingBatchRequestByAddrRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryLastPendingBatchRequestByAddrRequest", [
      {
        no: 1,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLastPendingBatchRequestByAddrRequest = new QueryLastPendingBatchRequestByAddrRequest$Type();
class QueryLastPendingBatchRequestByAddrResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryLastPendingBatchRequestByAddrResponse", [
      { no: 1, name: "batch", kind: "message", T: () => OutgoingTxBatch }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.OutgoingTxBatch batch */
        1:
          message.batch = OutgoingTxBatch.internalBinaryRead(reader, reader.uint32(), options, message.batch);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.batch)
      OutgoingTxBatch.internalBinaryWrite(message.batch, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLastPendingBatchRequestByAddrResponse = new QueryLastPendingBatchRequestByAddrResponse$Type();
class QueryOutgoingTxBatchesRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryOutgoingTxBatchesRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryOutgoingTxBatchesRequest = new QueryOutgoingTxBatchesRequest$Type();
class QueryOutgoingTxBatchesResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryOutgoingTxBatchesResponse", [
      { no: 1, name: "batches", kind: "message", repeat: 2, T: () => OutgoingTxBatch }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.batches = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective.peggy.v1.OutgoingTxBatch batches */
        1:
          message.batches.push(OutgoingTxBatch.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.batches.length; i++)
      OutgoingTxBatch.internalBinaryWrite(message.batches[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryOutgoingTxBatchesResponse = new QueryOutgoingTxBatchesResponse$Type();
class QueryBatchRequestByNonceRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryBatchRequestByNonceRequest", [
      {
        no: 1,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nonce = 0n;
    message.contractAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 nonce */
        1:
          message.nonce = reader.uint64().toBigInt();
          break;
        case /* string contract_address */
        2:
          message.contractAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.nonce);
    if (message.contractAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.contractAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryBatchRequestByNonceRequest = new QueryBatchRequestByNonceRequest$Type();
class QueryBatchRequestByNonceResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryBatchRequestByNonceResponse", [
      { no: 1, name: "batch", kind: "message", T: () => OutgoingTxBatch }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.OutgoingTxBatch batch */
        1:
          message.batch = OutgoingTxBatch.internalBinaryRead(reader, reader.uint32(), options, message.batch);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.batch)
      OutgoingTxBatch.internalBinaryWrite(message.batch, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryBatchRequestByNonceResponse = new QueryBatchRequestByNonceResponse$Type();
class QueryBatchConfirmsRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryBatchConfirmsRequest", [
      {
        no: 1,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nonce = 0n;
    message.contractAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 nonce */
        1:
          message.nonce = reader.uint64().toBigInt();
          break;
        case /* string contract_address */
        2:
          message.contractAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.nonce);
    if (message.contractAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.contractAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryBatchConfirmsRequest = new QueryBatchConfirmsRequest$Type();
class QueryBatchConfirmsResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryBatchConfirmsResponse", [
      { no: 1, name: "confirms", kind: "message", repeat: 2, T: () => MsgConfirmBatch }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.confirms = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective.peggy.v1.MsgConfirmBatch confirms */
        1:
          message.confirms.push(MsgConfirmBatch.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.confirms.length; i++)
      MsgConfirmBatch.internalBinaryWrite(message.confirms[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryBatchConfirmsResponse = new QueryBatchConfirmsResponse$Type();
class QueryLastEventByAddrRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryLastEventByAddrRequest", [
      {
        no: 1,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLastEventByAddrRequest = new QueryLastEventByAddrRequest$Type();
class QueryLastEventByAddrResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryLastEventByAddrResponse", [
      { no: 1, name: "last_claim_event", kind: "message", T: () => LastClaimEvent }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.LastClaimEvent last_claim_event */
        1:
          message.lastClaimEvent = LastClaimEvent.internalBinaryRead(reader, reader.uint32(), options, message.lastClaimEvent);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.lastClaimEvent)
      LastClaimEvent.internalBinaryWrite(message.lastClaimEvent, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLastEventByAddrResponse = new QueryLastEventByAddrResponse$Type();
class QueryERC20ToDenomRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryERC20ToDenomRequest", [
      {
        no: 1,
        name: "erc20",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.erc20 = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string erc20 */
        1:
          message.erc20 = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.erc20 !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.erc20);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryERC20ToDenomRequest = new QueryERC20ToDenomRequest$Type();
class QueryERC20ToDenomResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryERC20ToDenomResponse", [
      {
        no: 1,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "cosmos_originated",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.denom = "";
    message.cosmosOriginated = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string denom */
        1:
          message.denom = reader.string();
          break;
        case /* bool cosmos_originated */
        2:
          message.cosmosOriginated = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.denom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.denom);
    if (message.cosmosOriginated !== false)
      writer.tag(2, WireType.Varint).bool(message.cosmosOriginated);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryERC20ToDenomResponse = new QueryERC20ToDenomResponse$Type();
class QueryDenomToERC20Request$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryDenomToERC20Request", [
      {
        no: 1,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.denom = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string denom */
        1:
          message.denom = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.denom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.denom);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDenomToERC20Request = new QueryDenomToERC20Request$Type();
class QueryDenomToERC20Response$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryDenomToERC20Response", [
      {
        no: 1,
        name: "erc20",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "cosmos_originated",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.erc20 = "";
    message.cosmosOriginated = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string erc20 */
        1:
          message.erc20 = reader.string();
          break;
        case /* bool cosmos_originated */
        2:
          message.cosmosOriginated = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.erc20 !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.erc20);
    if (message.cosmosOriginated !== false)
      writer.tag(2, WireType.Varint).bool(message.cosmosOriginated);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDenomToERC20Response = new QueryDenomToERC20Response$Type();
class QueryDelegateKeysByValidatorAddress$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryDelegateKeysByValidatorAddress", [
      {
        no: 1,
        name: "validator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validatorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string validator_address */
        1:
          message.validatorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.validatorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegateKeysByValidatorAddress = new QueryDelegateKeysByValidatorAddress$Type();
class QueryDelegateKeysByValidatorAddressResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryDelegateKeysByValidatorAddressResponse", [
      {
        no: 1,
        name: "eth_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.ethAddress = "";
    message.orchestratorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string eth_address */
        1:
          message.ethAddress = reader.string();
          break;
        case /* string orchestrator_address */
        2:
          message.orchestratorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.ethAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.ethAddress);
    if (message.orchestratorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.orchestratorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegateKeysByValidatorAddressResponse = new QueryDelegateKeysByValidatorAddressResponse$Type();
class QueryDelegateKeysByEthAddress$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryDelegateKeysByEthAddress", [
      {
        no: 1,
        name: "eth_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.ethAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string eth_address */
        1:
          message.ethAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.ethAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.ethAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegateKeysByEthAddress = new QueryDelegateKeysByEthAddress$Type();
class QueryDelegateKeysByEthAddressResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryDelegateKeysByEthAddressResponse", [
      {
        no: 1,
        name: "validator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validatorAddress = "";
    message.orchestratorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string validator_address */
        1:
          message.validatorAddress = reader.string();
          break;
        case /* string orchestrator_address */
        2:
          message.orchestratorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.orchestratorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.orchestratorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegateKeysByEthAddressResponse = new QueryDelegateKeysByEthAddressResponse$Type();
class QueryDelegateKeysByOrchestratorAddress$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryDelegateKeysByOrchestratorAddress", [
      {
        no: 1,
        name: "orchestrator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.orchestratorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string orchestrator_address */
        1:
          message.orchestratorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.orchestratorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.orchestratorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegateKeysByOrchestratorAddress = new QueryDelegateKeysByOrchestratorAddress$Type();
class QueryDelegateKeysByOrchestratorAddressResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryDelegateKeysByOrchestratorAddressResponse", [
      {
        no: 1,
        name: "validator_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "eth_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validatorAddress = "";
    message.ethAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string validator_address */
        1:
          message.validatorAddress = reader.string();
          break;
        case /* string eth_address */
        2:
          message.ethAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.ethAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.ethAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegateKeysByOrchestratorAddressResponse = new QueryDelegateKeysByOrchestratorAddressResponse$Type();
class QueryPendingSendToEth$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryPendingSendToEth", [
      {
        no: 1,
        name: "sender_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.senderAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender_address */
        1:
          message.senderAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.senderAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.senderAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPendingSendToEth = new QueryPendingSendToEth$Type();
class QueryPendingSendToEthResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryPendingSendToEthResponse", [
      { no: 1, name: "transfers_in_batches", kind: "message", repeat: 2, T: () => OutgoingTransferTx },
      { no: 2, name: "unbatched_transfers", kind: "message", repeat: 2, T: () => OutgoingTransferTx }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.transfersInBatches = [];
    message.unbatchedTransfers = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated injective.peggy.v1.OutgoingTransferTx transfers_in_batches */
        1:
          message.transfersInBatches.push(OutgoingTransferTx.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective.peggy.v1.OutgoingTransferTx unbatched_transfers */
        2:
          message.unbatchedTransfers.push(OutgoingTransferTx.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.transfersInBatches.length; i++)
      OutgoingTransferTx.internalBinaryWrite(message.transfersInBatches[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.unbatchedTransfers.length; i++)
      OutgoingTransferTx.internalBinaryWrite(message.unbatchedTransfers[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPendingSendToEthResponse = new QueryPendingSendToEthResponse$Type();
class QueryModuleStateRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryModuleStateRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryModuleStateRequest = new QueryModuleStateRequest$Type();
class QueryModuleStateResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.QueryModuleStateResponse", [
      { no: 1, name: "state", kind: "message", T: () => GenesisState }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.peggy.v1.GenesisState state */
        1:
          message.state = GenesisState.internalBinaryRead(reader, reader.uint32(), options, message.state);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.state)
      GenesisState.internalBinaryWrite(message.state, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryModuleStateResponse = new QueryModuleStateResponse$Type();
class MissingNoncesRequest$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.MissingNoncesRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MissingNoncesRequest = new MissingNoncesRequest$Type();
class MissingNoncesResponse$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.MissingNoncesResponse", [
      {
        no: 1,
        name: "operator_addresses",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.operatorAddresses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string operator_addresses */
        1:
          message.operatorAddresses.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.operatorAddresses.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.operatorAddresses[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MissingNoncesResponse = new MissingNoncesResponse$Type();
const Query = new ServiceType("injective.peggy.v1.Query", [
  { name: "Params", options: { "google.api.http": { get: "/peggy/v1/params" } }, I: QueryParamsRequest, O: QueryParamsResponse },
  { name: "CurrentValset", options: { "google.api.http": { get: "/peggy/v1/valset/current" } }, I: QueryCurrentValsetRequest, O: QueryCurrentValsetResponse },
  { name: "ValsetRequest", options: { "google.api.http": { get: "/peggy/v1/valset" } }, I: QueryValsetRequestRequest, O: QueryValsetRequestResponse },
  { name: "ValsetConfirm", options: { "google.api.http": { get: "/peggy/v1/valset/confirm" } }, I: QueryValsetConfirmRequest, O: QueryValsetConfirmResponse },
  { name: "ValsetConfirmsByNonce", options: { "google.api.http": { get: "/peggy/v1/confirms/{nonce}" } }, I: QueryValsetConfirmsByNonceRequest, O: QueryValsetConfirmsByNonceResponse },
  { name: "LastValsetRequests", options: { "google.api.http": { get: "/peggy/v1/valset/requests" } }, I: QueryLastValsetRequestsRequest, O: QueryLastValsetRequestsResponse },
  { name: "LastPendingValsetRequestByAddr", options: { "google.api.http": { get: "/peggy/v1/valset/last" } }, I: QueryLastPendingValsetRequestByAddrRequest, O: QueryLastPendingValsetRequestByAddrResponse },
  { name: "LastEventByAddr", options: { "google.api.http": { get: "/peggy/v1/oracle/event/{address}" } }, I: QueryLastEventByAddrRequest, O: QueryLastEventByAddrResponse },
  { name: "GetPendingSendToEth", options: { "google.api.http": { get: "/peggy/v1/pending_send_to_eth" } }, I: QueryPendingSendToEth, O: QueryPendingSendToEthResponse },
  { name: "BatchFees", options: { "google.api.http": { get: "/peggy/v1/batchfees" } }, I: QueryBatchFeeRequest, O: QueryBatchFeeResponse },
  { name: "OutgoingTxBatches", options: { "google.api.http": { get: "/peggy/v1/batch/outgoingtx" } }, I: QueryOutgoingTxBatchesRequest, O: QueryOutgoingTxBatchesResponse },
  { name: "LastPendingBatchRequestByAddr", options: { "google.api.http": { get: "/peggy/v1/batch/last" } }, I: QueryLastPendingBatchRequestByAddrRequest, O: QueryLastPendingBatchRequestByAddrResponse },
  { name: "BatchRequestByNonce", options: { "google.api.http": { get: "/peggy/v1/batch" } }, I: QueryBatchRequestByNonceRequest, O: QueryBatchRequestByNonceResponse },
  { name: "BatchConfirms", options: { "google.api.http": { get: "/peggy/v1/batch/confirms" } }, I: QueryBatchConfirmsRequest, O: QueryBatchConfirmsResponse },
  { name: "ERC20ToDenom", options: { "google.api.http": { get: "/peggy/v1/cosmos_originated/erc20_to_denom" } }, I: QueryERC20ToDenomRequest, O: QueryERC20ToDenomResponse },
  { name: "DenomToERC20", options: { "google.api.http": { get: "/peggy/v1/cosmos_originated/denom_to_erc20" } }, I: QueryDenomToERC20Request, O: QueryDenomToERC20Response },
  { name: "GetDelegateKeyByValidator", options: { "google.api.http": { get: "/peggy/v1/query_delegate_keys_by_validator" } }, I: QueryDelegateKeysByValidatorAddress, O: QueryDelegateKeysByValidatorAddressResponse },
  { name: "GetDelegateKeyByEth", options: { "google.api.http": { get: "/peggy/v1/query_delegate_keys_by_eth" } }, I: QueryDelegateKeysByEthAddress, O: QueryDelegateKeysByEthAddressResponse },
  { name: "GetDelegateKeyByOrchestrator", options: { "google.api.http": { get: "/peggy/v1/query_delegate_keys_by_orchestrator" } }, I: QueryDelegateKeysByOrchestratorAddress, O: QueryDelegateKeysByOrchestratorAddressResponse },
  { name: "PeggyModuleState", options: { "google.api.http": { get: "/peggy/v1/module_state" } }, I: QueryModuleStateRequest, O: QueryModuleStateResponse },
  { name: "MissingPeggoNonces", options: { "google.api.http": { get: "/peggy/v1/missing_nonces" } }, I: MissingNoncesRequest, O: MissingNoncesResponse }
]);
export {
  MissingNoncesRequest,
  MissingNoncesResponse,
  Query,
  QueryBatchConfirmsRequest,
  QueryBatchConfirmsResponse,
  QueryBatchFeeRequest,
  QueryBatchFeeResponse,
  QueryBatchRequestByNonceRequest,
  QueryBatchRequestByNonceResponse,
  QueryCurrentValsetRequest,
  QueryCurrentValsetResponse,
  QueryDelegateKeysByEthAddress,
  QueryDelegateKeysByEthAddressResponse,
  QueryDelegateKeysByOrchestratorAddress,
  QueryDelegateKeysByOrchestratorAddressResponse,
  QueryDelegateKeysByValidatorAddress,
  QueryDelegateKeysByValidatorAddressResponse,
  QueryDenomToERC20Request,
  QueryDenomToERC20Response,
  QueryERC20ToDenomRequest,
  QueryERC20ToDenomResponse,
  QueryLastEventByAddrRequest,
  QueryLastEventByAddrResponse,
  QueryLastPendingBatchRequestByAddrRequest,
  QueryLastPendingBatchRequestByAddrResponse,
  QueryLastPendingValsetRequestByAddrRequest,
  QueryLastPendingValsetRequestByAddrResponse,
  QueryLastValsetRequestsRequest,
  QueryLastValsetRequestsResponse,
  QueryModuleStateRequest,
  QueryModuleStateResponse,
  QueryOutgoingTxBatchesRequest,
  QueryOutgoingTxBatchesResponse,
  QueryParamsRequest,
  QueryParamsResponse,
  QueryPendingSendToEth,
  QueryPendingSendToEthResponse,
  QueryValsetConfirmRequest,
  QueryValsetConfirmResponse,
  QueryValsetConfirmsByNonceRequest,
  QueryValsetConfirmsByNonceResponse,
  QueryValsetRequestRequest,
  QueryValsetRequestResponse
};
