// @injectivelabs/core-proto-ts-v2/esm/generated/injective/peggy/v1/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class BridgeValidator$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.BridgeValidator", [
      {
        no: 1,
        name: "power",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "ethereum_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.power = 0n;
    message.ethereumAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 power */
        1:
          message.power = reader.uint64().toBigInt();
          break;
        case /* string ethereum_address */
        2:
          message.ethereumAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.power !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.power);
    if (message.ethereumAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.ethereumAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BridgeValidator = new BridgeValidator$Type();
class Valset$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.Valset", [
      {
        no: 1,
        name: "nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "members", kind: "message", repeat: 2, T: () => BridgeValidator },
      {
        no: 3,
        name: "height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "reward_amount", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      {
        no: 5,
        name: "reward_token",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nonce = 0n;
    message.members = [];
    message.height = 0n;
    message.rewardAmount = "";
    message.rewardToken = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 nonce */
        1:
          message.nonce = reader.uint64().toBigInt();
          break;
        case /* repeated injective.peggy.v1.BridgeValidator members */
        2:
          message.members.push(BridgeValidator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* uint64 height */
        3:
          message.height = reader.uint64().toBigInt();
          break;
        case /* string reward_amount */
        4:
          message.rewardAmount = reader.string();
          break;
        case /* string reward_token */
        5:
          message.rewardToken = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.nonce);
    for (let i = 0; i < message.members.length; i++)
      BridgeValidator.internalBinaryWrite(message.members[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.height !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.height);
    if (message.rewardAmount !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.rewardAmount);
    if (message.rewardToken !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.rewardToken);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Valset = new Valset$Type();
class LastObservedEthereumBlockHeight$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.LastObservedEthereumBlockHeight", [
      {
        no: 1,
        name: "cosmos_block_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "ethereum_block_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.cosmosBlockHeight = 0n;
    message.ethereumBlockHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 cosmos_block_height */
        1:
          message.cosmosBlockHeight = reader.uint64().toBigInt();
          break;
        case /* uint64 ethereum_block_height */
        2:
          message.ethereumBlockHeight = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.cosmosBlockHeight !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.cosmosBlockHeight);
    if (message.ethereumBlockHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.ethereumBlockHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const LastObservedEthereumBlockHeight = new LastObservedEthereumBlockHeight$Type();
class LastClaimEvent$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.LastClaimEvent", [
      {
        no: 1,
        name: "ethereum_event_nonce",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "ethereum_event_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.ethereumEventNonce = 0n;
    message.ethereumEventHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 ethereum_event_nonce */
        1:
          message.ethereumEventNonce = reader.uint64().toBigInt();
          break;
        case /* uint64 ethereum_event_height */
        2:
          message.ethereumEventHeight = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.ethereumEventNonce !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.ethereumEventNonce);
    if (message.ethereumEventHeight !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.ethereumEventHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const LastClaimEvent = new LastClaimEvent$Type();
class ERC20ToDenom$Type extends MessageType {
  constructor() {
    super("injective.peggy.v1.ERC20ToDenom", [
      {
        no: 1,
        name: "erc20",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.erc20 = "";
    message.denom = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string erc20 */
        1:
          message.erc20 = reader.string();
          break;
        case /* string denom */
        2:
          message.denom = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.erc20 !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.erc20);
    if (message.denom !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.denom);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ERC20ToDenom = new ERC20ToDenom$Type();
export {
  BridgeValidator,
  ERC20ToDenom,
  LastClaimEvent,
  LastObservedEthereumBlockHeight,
  Valset
};
