// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/base/reflection/v2alpha1/reflection_pb.client.js
import { ReflectionService } from "./reflection_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class ReflectionServiceClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = ReflectionService.typeName;
    this.methods = ReflectionService.methods;
    this.options = ReflectionService.options;
  }
  /**
   * GetAuthnDescriptor returns information on how to authenticate transactions in the application
   * NOTE: this RPC is still experimental and might be subject to breaking changes or removal in
   * future releases of the cosmos-sdk.
   *
   * @generated from protobuf rpc: GetAuthnDescriptor
   */
  getAuthnDescriptor(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetChainDescriptor returns the description of the chain
   *
   * @generated from protobuf rpc: GetChainDescriptor
   */
  getChainDescriptor(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetCodecDescriptor returns the descriptor of the codec of the application
   *
   * @generated from protobuf rpc: GetCodecDescriptor
   */
  getCodecDescriptor(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetConfigurationDescriptor returns the descriptor for the sdk.Config of the application
   *
   * @generated from protobuf rpc: GetConfigurationDescriptor
   */
  getConfigurationDescriptor(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetQueryServicesDescriptor returns the available gRPC queryable services of the application
   *
   * @generated from protobuf rpc: GetQueryServicesDescriptor
   */
  getQueryServicesDescriptor(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GetTxDescriptor returns information on the used transaction object and available msgs that can be used
   *
   * @generated from protobuf rpc: GetTxDescriptor
   */
  getTxDescriptor(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  ReflectionServiceClient
};
