// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/evidence/v1beta1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../google/protobuf/any_pb.js";
class MsgSubmitEvidence$Type extends MessageType {
  constructor() {
    super("cosmos.evidence.v1beta1.MsgSubmitEvidence", [
      { no: 1, name: "submitter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "evidence", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.evidence.v1beta1.Evidence" } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgSubmitEvidence", "cosmos.msg.v1.signer": ["submitter"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.submitter = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string submitter */
        1:
          message.submitter = reader.string();
          break;
        case /* google.protobuf.Any evidence */
        2:
          message.evidence = Any.internalBinaryRead(reader, reader.uint32(), options, message.evidence);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.submitter !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.submitter);
    if (message.evidence)
      Any.internalBinaryWrite(message.evidence, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSubmitEvidence = new MsgSubmitEvidence$Type();
class MsgSubmitEvidenceResponse$Type extends MessageType {
  constructor() {
    super("cosmos.evidence.v1beta1.MsgSubmitEvidenceResponse", [
      {
        no: 4,
        name: "hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.hash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes hash */
        4:
          message.hash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.hash.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.hash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSubmitEvidenceResponse = new MsgSubmitEvidenceResponse$Type();
const Msg = new ServiceType("cosmos.evidence.v1beta1.Msg", [
  { name: "SubmitEvidence", options: {}, I: MsgSubmitEvidence, O: MsgSubmitEvidenceResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgSubmitEvidence,
  MsgSubmitEvidenceResponse
};
