// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/staking/v1beta1/authz_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Coin } from "../../base/v1beta1/coin_pb.js";
var AuthorizationType = /* @__PURE__ */ ((AuthorizationType2) => {
  AuthorizationType2[AuthorizationType2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  AuthorizationType2[AuthorizationType2["DELEGATE"] = 1] = "DELEGATE";
  AuthorizationType2[AuthorizationType2["UNDELEGATE"] = 2] = "UNDELEGATE";
  AuthorizationType2[AuthorizationType2["REDELEGATE"] = 3] = "REDELEGATE";
  AuthorizationType2[AuthorizationType2["CANCEL_UNBONDING_DELEGATION"] = 4] = "CANCEL_UNBONDING_DELEGATION";
  return AuthorizationType2;
})(AuthorizationType || {});
class StakeAuthorization$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.StakeAuthorization", [
      { no: 1, name: "max_tokens", kind: "message", T: () => Coin, options: { "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coin" } },
      { no: 2, name: "allow_list", kind: "message", oneof: "validators", T: () => StakeAuthorization_Validators, options: { "amino.oneof_name": "cosmos-sdk/StakeAuthorization/AllowList" } },
      { no: 3, name: "deny_list", kind: "message", oneof: "validators", T: () => StakeAuthorization_Validators, options: { "amino.oneof_name": "cosmos-sdk/StakeAuthorization/DenyList" } },
      { no: 4, name: "authorization_type", kind: "enum", T: () => ["cosmos.staking.v1beta1.AuthorizationType", AuthorizationType, "AUTHORIZATION_TYPE_"] }
    ], { "cosmos_proto.implements_interface": ["cosmos.authz.v1beta1.Authorization"], "amino.name": "cosmos-sdk/StakeAuthorization" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validators = { oneofKind: void 0 };
    message.authorizationType = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.v1beta1.Coin max_tokens */
        1:
          message.maxTokens = Coin.internalBinaryRead(reader, reader.uint32(), options, message.maxTokens);
          break;
        case /* cosmos.staking.v1beta1.StakeAuthorization.Validators allow_list */
        2:
          message.validators = {
            oneofKind: "allowList",
            allowList: StakeAuthorization_Validators.internalBinaryRead(reader, reader.uint32(), options, message.validators.allowList)
          };
          break;
        case /* cosmos.staking.v1beta1.StakeAuthorization.Validators deny_list */
        3:
          message.validators = {
            oneofKind: "denyList",
            denyList: StakeAuthorization_Validators.internalBinaryRead(reader, reader.uint32(), options, message.validators.denyList)
          };
          break;
        case /* cosmos.staking.v1beta1.AuthorizationType authorization_type */
        4:
          message.authorizationType = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.maxTokens)
      Coin.internalBinaryWrite(message.maxTokens, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.validators.oneofKind === "allowList")
      StakeAuthorization_Validators.internalBinaryWrite(message.validators.allowList, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.validators.oneofKind === "denyList")
      StakeAuthorization_Validators.internalBinaryWrite(message.validators.denyList, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.authorizationType !== 0)
      writer.tag(4, WireType.Varint).int32(message.authorizationType);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StakeAuthorization = new StakeAuthorization$Type();
class StakeAuthorization_Validators$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.StakeAuthorization.Validators", [
      { no: 1, name: "address", kind: "scalar", repeat: 2, T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string address */
        1:
          message.address.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.address.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.address[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StakeAuthorization_Validators = new StakeAuthorization_Validators$Type();
export {
  AuthorizationType,
  StakeAuthorization,
  StakeAuthorization_Validators
};
