// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/staking/v1beta1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Redelegation } from "./staking_pb.js";
import { UnbondingDelegation } from "./staking_pb.js";
import { Delegation } from "./staking_pb.js";
import { Validator } from "./staking_pb.js";
import { Params } from "./staking_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.GenesisState", [
      { no: 1, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "last_total_power", kind: "scalar", T: 12, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int", "amino.dont_omitempty": true } },
      { no: 3, name: "last_validator_powers", kind: "message", repeat: 2, T: () => LastValidatorPower, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 4, name: "validators", kind: "message", repeat: 2, T: () => Validator, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 5, name: "delegations", kind: "message", repeat: 2, T: () => Delegation, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 6, name: "unbonding_delegations", kind: "message", repeat: 2, T: () => UnbondingDelegation, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 7, name: "redelegations", kind: "message", repeat: 2, T: () => Redelegation, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 8,
        name: "exported",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 9, name: "allowed_delegation_transfer_receivers", kind: "scalar", repeat: 2, T: 9, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.lastTotalPower = new Uint8Array(0);
    message.lastValidatorPowers = [];
    message.validators = [];
    message.delegations = [];
    message.unbondingDelegations = [];
    message.redelegations = [];
    message.exported = false;
    message.allowedDelegationTransferReceivers = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        case /* bytes last_total_power */
        2:
          message.lastTotalPower = reader.bytes();
          break;
        case /* repeated cosmos.staking.v1beta1.LastValidatorPower last_validator_powers */
        3:
          message.lastValidatorPowers.push(LastValidatorPower.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.staking.v1beta1.Validator validators */
        4:
          message.validators.push(Validator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.staking.v1beta1.Delegation delegations */
        5:
          message.delegations.push(Delegation.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.staking.v1beta1.UnbondingDelegation unbonding_delegations */
        6:
          message.unbondingDelegations.push(UnbondingDelegation.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.staking.v1beta1.Redelegation redelegations */
        7:
          message.redelegations.push(Redelegation.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* bool exported */
        8:
          message.exported = reader.bool();
          break;
        case /* repeated string allowed_delegation_transfer_receivers */
        9:
          message.allowedDelegationTransferReceivers.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.lastTotalPower.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.lastTotalPower);
    for (let i = 0; i < message.lastValidatorPowers.length; i++)
      LastValidatorPower.internalBinaryWrite(message.lastValidatorPowers[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.validators.length; i++)
      Validator.internalBinaryWrite(message.validators[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.delegations.length; i++)
      Delegation.internalBinaryWrite(message.delegations[i], writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.unbondingDelegations.length; i++)
      UnbondingDelegation.internalBinaryWrite(message.unbondingDelegations[i], writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.redelegations.length; i++)
      Redelegation.internalBinaryWrite(message.redelegations[i], writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.exported !== false)
      writer.tag(8, WireType.Varint).bool(message.exported);
    for (let i = 0; i < message.allowedDelegationTransferReceivers.length; i++)
      writer.tag(9, WireType.LengthDelimited).string(message.allowedDelegationTransferReceivers[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class LastValidatorPower$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.LastValidatorPower", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "power",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.power = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* int64 power */
        2:
          message.power = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.power !== 0n)
      writer.tag(2, WireType.Varint).int64(message.power);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const LastValidatorPower = new LastValidatorPower$Type();
export {
  GenesisState,
  LastValidatorPower
};
