// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/staking/v1beta1/query_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./staking_pb.js";
import { Pool } from "./staking_pb.js";
import { HistoricalInfo } from "./staking_pb.js";
import { RedelegationResponse } from "./staking_pb.js";
import { UnbondingDelegation } from "./staking_pb.js";
import { DelegationResponse } from "./staking_pb.js";
import { PageResponse } from "../../base/query/v1beta1/pagination_pb.js";
import { Validator } from "./staking_pb.js";
import { PageRequest } from "../../base/query/v1beta1/pagination_pb.js";
class QueryValidatorsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryValidatorsRequest", [
      {
        no: 1,
        name: "status",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.status = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string status */
        1:
          message.status = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.status !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.status);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValidatorsRequest = new QueryValidatorsRequest$Type();
class QueryValidatorsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryValidatorsResponse", [
      { no: 1, name: "validators", kind: "message", repeat: 2, T: () => Validator, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validators = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.staking.v1beta1.Validator validators */
        1:
          message.validators.push(Validator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.validators.length; i++)
      Validator.internalBinaryWrite(message.validators[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValidatorsResponse = new QueryValidatorsResponse$Type();
class QueryValidatorRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryValidatorRequest", [
      { no: 1, name: "validator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string validator_addr */
        1:
          message.validatorAddr = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.validatorAddr);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValidatorRequest = new QueryValidatorRequest$Type();
class QueryValidatorResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryValidatorResponse", [
      { no: 1, name: "validator", kind: "message", T: () => Validator, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.Validator validator */
        1:
          message.validator = Validator.internalBinaryRead(reader, reader.uint32(), options, message.validator);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validator)
      Validator.internalBinaryWrite(message.validator, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValidatorResponse = new QueryValidatorResponse$Type();
class QueryValidatorDelegationsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryValidatorDelegationsRequest", [
      { no: 1, name: "validator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string validator_addr */
        1:
          message.validatorAddr = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.validatorAddr);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValidatorDelegationsRequest = new QueryValidatorDelegationsRequest$Type();
class QueryValidatorDelegationsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryValidatorDelegationsResponse", [
      { no: 1, name: "delegation_responses", kind: "message", repeat: 2, T: () => DelegationResponse, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "DelegationResponses", "amino.dont_omitempty": true } },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegationResponses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.staking.v1beta1.DelegationResponse delegation_responses */
        1:
          message.delegationResponses.push(DelegationResponse.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.delegationResponses.length; i++)
      DelegationResponse.internalBinaryWrite(message.delegationResponses[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValidatorDelegationsResponse = new QueryValidatorDelegationsResponse$Type();
class QueryValidatorUnbondingDelegationsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryValidatorUnbondingDelegationsRequest", [
      { no: 1, name: "validator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string validator_addr */
        1:
          message.validatorAddr = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.validatorAddr);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValidatorUnbondingDelegationsRequest = new QueryValidatorUnbondingDelegationsRequest$Type();
class QueryValidatorUnbondingDelegationsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryValidatorUnbondingDelegationsResponse", [
      { no: 1, name: "unbonding_responses", kind: "message", repeat: 2, T: () => UnbondingDelegation, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.unbondingResponses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.staking.v1beta1.UnbondingDelegation unbonding_responses */
        1:
          message.unbondingResponses.push(UnbondingDelegation.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.unbondingResponses.length; i++)
      UnbondingDelegation.internalBinaryWrite(message.unbondingResponses[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryValidatorUnbondingDelegationsResponse = new QueryValidatorUnbondingDelegationsResponse$Type();
class QueryDelegationRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegationRequest", [
      { no: 1, name: "delegator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddr = "";
    message.validatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_addr */
        1:
          message.delegatorAddr = reader.string();
          break;
        case /* string validator_addr */
        2:
          message.validatorAddr = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddr);
    if (message.validatorAddr !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddr);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegationRequest = new QueryDelegationRequest$Type();
class QueryDelegationResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegationResponse", [
      { no: 1, name: "delegation_response", kind: "message", T: () => DelegationResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.DelegationResponse delegation_response */
        1:
          message.delegationResponse = DelegationResponse.internalBinaryRead(reader, reader.uint32(), options, message.delegationResponse);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegationResponse)
      DelegationResponse.internalBinaryWrite(message.delegationResponse, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegationResponse = new QueryDelegationResponse$Type();
class QueryUnbondingDelegationRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryUnbondingDelegationRequest", [
      { no: 1, name: "delegator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddr = "";
    message.validatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_addr */
        1:
          message.delegatorAddr = reader.string();
          break;
        case /* string validator_addr */
        2:
          message.validatorAddr = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddr);
    if (message.validatorAddr !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddr);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUnbondingDelegationRequest = new QueryUnbondingDelegationRequest$Type();
class QueryUnbondingDelegationResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryUnbondingDelegationResponse", [
      { no: 1, name: "unbond", kind: "message", T: () => UnbondingDelegation, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.UnbondingDelegation unbond */
        1:
          message.unbond = UnbondingDelegation.internalBinaryRead(reader, reader.uint32(), options, message.unbond);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.unbond)
      UnbondingDelegation.internalBinaryWrite(message.unbond, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUnbondingDelegationResponse = new QueryUnbondingDelegationResponse$Type();
class QueryDelegatorDelegationsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegatorDelegationsRequest", [
      { no: 1, name: "delegator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_addr */
        1:
          message.delegatorAddr = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddr);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegatorDelegationsRequest = new QueryDelegatorDelegationsRequest$Type();
class QueryDelegatorDelegationsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegatorDelegationsResponse", [
      { no: 1, name: "delegation_responses", kind: "message", repeat: 2, T: () => DelegationResponse, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegationResponses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.staking.v1beta1.DelegationResponse delegation_responses */
        1:
          message.delegationResponses.push(DelegationResponse.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.delegationResponses.length; i++)
      DelegationResponse.internalBinaryWrite(message.delegationResponses[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegatorDelegationsResponse = new QueryDelegatorDelegationsResponse$Type();
class QueryDelegatorUnbondingDelegationsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegatorUnbondingDelegationsRequest", [
      { no: 1, name: "delegator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_addr */
        1:
          message.delegatorAddr = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddr);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegatorUnbondingDelegationsRequest = new QueryDelegatorUnbondingDelegationsRequest$Type();
class QueryDelegatorUnbondingDelegationsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegatorUnbondingDelegationsResponse", [
      { no: 1, name: "unbonding_responses", kind: "message", repeat: 2, T: () => UnbondingDelegation, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.unbondingResponses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.staking.v1beta1.UnbondingDelegation unbonding_responses */
        1:
          message.unbondingResponses.push(UnbondingDelegation.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.unbondingResponses.length; i++)
      UnbondingDelegation.internalBinaryWrite(message.unbondingResponses[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegatorUnbondingDelegationsResponse = new QueryDelegatorUnbondingDelegationsResponse$Type();
class QueryRedelegationsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryRedelegationsRequest", [
      { no: 1, name: "delegator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "src_validator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "dst_validator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 4, name: "pagination", kind: "message", T: () => PageRequest }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddr = "";
    message.srcValidatorAddr = "";
    message.dstValidatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_addr */
        1:
          message.delegatorAddr = reader.string();
          break;
        case /* string src_validator_addr */
        2:
          message.srcValidatorAddr = reader.string();
          break;
        case /* string dst_validator_addr */
        3:
          message.dstValidatorAddr = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        4:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddr);
    if (message.srcValidatorAddr !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.srcValidatorAddr);
    if (message.dstValidatorAddr !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.dstValidatorAddr);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryRedelegationsRequest = new QueryRedelegationsRequest$Type();
class QueryRedelegationsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryRedelegationsResponse", [
      { no: 1, name: "redelegation_responses", kind: "message", repeat: 2, T: () => RedelegationResponse, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.redelegationResponses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.staking.v1beta1.RedelegationResponse redelegation_responses */
        1:
          message.redelegationResponses.push(RedelegationResponse.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.redelegationResponses.length; i++)
      RedelegationResponse.internalBinaryWrite(message.redelegationResponses[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryRedelegationsResponse = new QueryRedelegationsResponse$Type();
class QueryDelegatorValidatorsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegatorValidatorsRequest", [
      { no: 1, name: "delegator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_addr */
        1:
          message.delegatorAddr = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddr);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegatorValidatorsRequest = new QueryDelegatorValidatorsRequest$Type();
class QueryDelegatorValidatorsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegatorValidatorsResponse", [
      { no: 1, name: "validators", kind: "message", repeat: 2, T: () => Validator, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validators = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.staking.v1beta1.Validator validators */
        1:
          message.validators.push(Validator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.validators.length; i++)
      Validator.internalBinaryWrite(message.validators[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegatorValidatorsResponse = new QueryDelegatorValidatorsResponse$Type();
class QueryDelegatorValidatorRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegatorValidatorRequest", [
      { no: 1, name: "delegator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_addr", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddr = "";
    message.validatorAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_addr */
        1:
          message.delegatorAddr = reader.string();
          break;
        case /* string validator_addr */
        2:
          message.validatorAddr = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddr);
    if (message.validatorAddr !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddr);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegatorValidatorRequest = new QueryDelegatorValidatorRequest$Type();
class QueryDelegatorValidatorResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryDelegatorValidatorResponse", [
      { no: 1, name: "validator", kind: "message", T: () => Validator, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.Validator validator */
        1:
          message.validator = Validator.internalBinaryRead(reader, reader.uint32(), options, message.validator);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validator)
      Validator.internalBinaryWrite(message.validator, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryDelegatorValidatorResponse = new QueryDelegatorValidatorResponse$Type();
class QueryAllowedDelegationTransferReceiversRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryAllowedDelegationTransferReceiversRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryAllowedDelegationTransferReceiversRequest = new QueryAllowedDelegationTransferReceiversRequest$Type();
class QueryAllowedDelegationTransferReceiversResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryAllowedDelegationTransferReceiversResponse", [
      { no: 1, name: "addresses", kind: "scalar", repeat: 2, T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.addresses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string addresses */
        1:
          message.addresses.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.addresses.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.addresses[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryAllowedDelegationTransferReceiversResponse = new QueryAllowedDelegationTransferReceiversResponse$Type();
class QueryHistoricalInfoRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryHistoricalInfoRequest", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryHistoricalInfoRequest = new QueryHistoricalInfoRequest$Type();
class QueryHistoricalInfoResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryHistoricalInfoResponse", [
      { no: 1, name: "hist", kind: "message", T: () => HistoricalInfo }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.HistoricalInfo hist */
        1:
          message.hist = HistoricalInfo.internalBinaryRead(reader, reader.uint32(), options, message.hist);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.hist)
      HistoricalInfo.internalBinaryWrite(message.hist, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryHistoricalInfoResponse = new QueryHistoricalInfoResponse$Type();
class QueryPoolRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryPoolRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPoolRequest = new QueryPoolRequest$Type();
class QueryPoolResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryPoolResponse", [
      { no: 1, name: "pool", kind: "message", T: () => Pool, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.Pool pool */
        1:
          message.pool = Pool.internalBinaryRead(reader, reader.uint32(), options, message.pool);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pool)
      Pool.internalBinaryWrite(message.pool, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPoolResponse = new QueryPoolResponse$Type();
class QueryParamsRequest$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryParamsRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryParamsRequest = new QueryParamsRequest$Type();
class QueryParamsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.QueryParamsResponse", [
      { no: 1, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryParamsResponse = new QueryParamsResponse$Type();
const Query = new ServiceType("cosmos.staking.v1beta1.Query", [
  { name: "Validators", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/validators" }, "cosmos.query.v1.module_query_safe": true }, I: QueryValidatorsRequest, O: QueryValidatorsResponse },
  { name: "Validator", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/validators/{validator_addr}" }, "cosmos.query.v1.module_query_safe": true }, I: QueryValidatorRequest, O: QueryValidatorResponse },
  { name: "ValidatorDelegations", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/validators/{validator_addr}/delegations" }, "cosmos.query.v1.module_query_safe": true }, I: QueryValidatorDelegationsRequest, O: QueryValidatorDelegationsResponse },
  { name: "ValidatorUnbondingDelegations", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/validators/{validator_addr}/unbonding_delegations" }, "cosmos.query.v1.module_query_safe": true }, I: QueryValidatorUnbondingDelegationsRequest, O: QueryValidatorUnbondingDelegationsResponse },
  { name: "Delegation", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/validators/{validator_addr}/delegations/{delegator_addr}" }, "cosmos.query.v1.module_query_safe": true }, I: QueryDelegationRequest, O: QueryDelegationResponse },
  { name: "UnbondingDelegation", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/validators/{validator_addr}/delegations/{delegator_addr}/unbonding_delegation" }, "cosmos.query.v1.module_query_safe": true }, I: QueryUnbondingDelegationRequest, O: QueryUnbondingDelegationResponse },
  { name: "DelegatorDelegations", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/delegations/{delegator_addr}" }, "cosmos.query.v1.module_query_safe": true }, I: QueryDelegatorDelegationsRequest, O: QueryDelegatorDelegationsResponse },
  { name: "DelegatorUnbondingDelegations", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/delegators/{delegator_addr}/unbonding_delegations" }, "cosmos.query.v1.module_query_safe": true }, I: QueryDelegatorUnbondingDelegationsRequest, O: QueryDelegatorUnbondingDelegationsResponse },
  { name: "Redelegations", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/delegators/{delegator_addr}/redelegations" }, "cosmos.query.v1.module_query_safe": true }, I: QueryRedelegationsRequest, O: QueryRedelegationsResponse },
  { name: "DelegatorValidators", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/delegators/{delegator_addr}/validators" }, "cosmos.query.v1.module_query_safe": true }, I: QueryDelegatorValidatorsRequest, O: QueryDelegatorValidatorsResponse },
  { name: "DelegatorValidator", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/delegators/{delegator_addr}/validators/{validator_addr}" }, "cosmos.query.v1.module_query_safe": true }, I: QueryDelegatorValidatorRequest, O: QueryDelegatorValidatorResponse },
  { name: "AllowedDelegationTransferReceivers", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/allowed_delegation_transfer_receivers" }, "cosmos.query.v1.module_query_safe": true }, I: QueryAllowedDelegationTransferReceiversRequest, O: QueryAllowedDelegationTransferReceiversResponse },
  { name: "HistoricalInfo", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/historical_info/{height}" }, "cosmos.query.v1.module_query_safe": true }, I: QueryHistoricalInfoRequest, O: QueryHistoricalInfoResponse },
  { name: "Pool", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/pool" }, "cosmos.query.v1.module_query_safe": true }, I: QueryPoolRequest, O: QueryPoolResponse },
  { name: "Params", options: { "google.api.http": { get: "/cosmos/staking/v1beta1/params" }, "cosmos.query.v1.module_query_safe": true }, I: QueryParamsRequest, O: QueryParamsResponse }
]);
export {
  Query,
  QueryAllowedDelegationTransferReceiversRequest,
  QueryAllowedDelegationTransferReceiversResponse,
  QueryDelegationRequest,
  QueryDelegationResponse,
  QueryDelegatorDelegationsRequest,
  QueryDelegatorDelegationsResponse,
  QueryDelegatorUnbondingDelegationsRequest,
  QueryDelegatorUnbondingDelegationsResponse,
  QueryDelegatorValidatorRequest,
  QueryDelegatorValidatorResponse,
  QueryDelegatorValidatorsRequest,
  QueryDelegatorValidatorsResponse,
  QueryHistoricalInfoRequest,
  QueryHistoricalInfoResponse,
  QueryParamsRequest,
  QueryParamsResponse,
  QueryPoolRequest,
  QueryPoolResponse,
  QueryRedelegationsRequest,
  QueryRedelegationsResponse,
  QueryUnbondingDelegationRequest,
  QueryUnbondingDelegationResponse,
  QueryValidatorDelegationsRequest,
  QueryValidatorDelegationsResponse,
  QueryValidatorRequest,
  QueryValidatorResponse,
  QueryValidatorUnbondingDelegationsRequest,
  QueryValidatorUnbondingDelegationsResponse,
  QueryValidatorsRequest,
  QueryValidatorsResponse
};
