// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/staking/v1beta1/staking_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ValidatorUpdate } from "../../../cometbft/abci/v1/types_pb.js";
import { Coin } from "../../base/v1beta1/coin_pb.js";
import { Duration } from "../../../google/protobuf/duration_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Header } from "../../../cometbft/types/v1/types_pb.js";
var BondStatus = /* @__PURE__ */ ((BondStatus2) => {
  BondStatus2[BondStatus2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  BondStatus2[BondStatus2["UNBONDED"] = 1] = "UNBONDED";
  BondStatus2[BondStatus2["UNBONDING"] = 2] = "UNBONDING";
  BondStatus2[BondStatus2["BONDED"] = 3] = "BONDED";
  return BondStatus2;
})(BondStatus || {});
var Infraction = /* @__PURE__ */ ((Infraction2) => {
  Infraction2[Infraction2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  Infraction2[Infraction2["DOUBLE_SIGN"] = 1] = "DOUBLE_SIGN";
  Infraction2[Infraction2["DOWNTIME"] = 2] = "DOWNTIME";
  return Infraction2;
})(Infraction || {});
class HistoricalInfo$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.HistoricalInfo", [
      { no: 1, name: "header", kind: "message", T: () => Header, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "valset", kind: "message", repeat: 2, T: () => Validator, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.valset = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.types.v1.Header header */
        1:
          message.header = Header.internalBinaryRead(reader, reader.uint32(), options, message.header);
          break;
        case /* repeated cosmos.staking.v1beta1.Validator valset */
        2:
          message.valset.push(Validator.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.header)
      Header.internalBinaryWrite(message.header, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.valset.length; i++)
      Validator.internalBinaryWrite(message.valset[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const HistoricalInfo = new HistoricalInfo$Type();
class CommissionRates$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.CommissionRates", [
      { no: 1, name: "rate", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } },
      { no: 2, name: "max_rate", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } },
      { no: 3, name: "max_change_rate", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.rate = "";
    message.maxRate = "";
    message.maxChangeRate = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string rate */
        1:
          message.rate = reader.string();
          break;
        case /* string max_rate */
        2:
          message.maxRate = reader.string();
          break;
        case /* string max_change_rate */
        3:
          message.maxChangeRate = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.rate !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.rate);
    if (message.maxRate !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.maxRate);
    if (message.maxChangeRate !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.maxChangeRate);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CommissionRates = new CommissionRates$Type();
class Commission$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.Commission", [
      { no: 1, name: "commission_rates", kind: "message", T: () => CommissionRates, options: { "gogoproto.nullable": false, "gogoproto.embed": true, "amino.dont_omitempty": true } },
      { no: 2, name: "update_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.CommissionRates commission_rates */
        1:
          message.commissionRates = CommissionRates.internalBinaryRead(reader, reader.uint32(), options, message.commissionRates);
          break;
        case /* google.protobuf.Timestamp update_time */
        2:
          message.updateTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.updateTime);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.commissionRates)
      CommissionRates.internalBinaryWrite(message.commissionRates, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.updateTime)
      Timestamp.internalBinaryWrite(message.updateTime, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Commission = new Commission$Type();
class Description$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.Description", [
      {
        no: 1,
        name: "moniker",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "identity",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "website",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "security_contact",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "details",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.moniker = "";
    message.identity = "";
    message.website = "";
    message.securityContact = "";
    message.details = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string moniker */
        1:
          message.moniker = reader.string();
          break;
        case /* string identity */
        2:
          message.identity = reader.string();
          break;
        case /* string website */
        3:
          message.website = reader.string();
          break;
        case /* string security_contact */
        4:
          message.securityContact = reader.string();
          break;
        case /* string details */
        5:
          message.details = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.moniker !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.moniker);
    if (message.identity !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.identity);
    if (message.website !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.website);
    if (message.securityContact !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.securityContact);
    if (message.details !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.details);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Description = new Description$Type();
class Validator$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.Validator", [
      { no: 1, name: "operator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "consensus_pubkey", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.crypto.PubKey" } },
      {
        no: 3,
        name: "jailed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 4, name: "status", kind: "enum", T: () => ["cosmos.staking.v1beta1.BondStatus", BondStatus, "BOND_STATUS_"] },
      { no: 5, name: "tokens", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } },
      { no: 6, name: "delegator_shares", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 7, name: "description", kind: "message", T: () => Description, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 8,
        name: "unbonding_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 9, name: "unbonding_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      { no: 10, name: "commission", kind: "message", T: () => Commission, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 11, name: "min_self_delegation", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } },
      {
        no: 12,
        name: "unbonding_on_hold_ref_count",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 13,
        name: "unbonding_ids",
        kind: "scalar",
        repeat: 1,
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.operatorAddress = "";
    message.jailed = false;
    message.status = 0;
    message.tokens = "";
    message.delegatorShares = "";
    message.unbondingHeight = 0n;
    message.minSelfDelegation = "";
    message.unbondingOnHoldRefCount = 0n;
    message.unbondingIds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string operator_address */
        1:
          message.operatorAddress = reader.string();
          break;
        case /* google.protobuf.Any consensus_pubkey */
        2:
          message.consensusPubkey = Any.internalBinaryRead(reader, reader.uint32(), options, message.consensusPubkey);
          break;
        case /* bool jailed */
        3:
          message.jailed = reader.bool();
          break;
        case /* cosmos.staking.v1beta1.BondStatus status */
        4:
          message.status = reader.int32();
          break;
        case /* string tokens */
        5:
          message.tokens = reader.string();
          break;
        case /* string delegator_shares */
        6:
          message.delegatorShares = reader.string();
          break;
        case /* cosmos.staking.v1beta1.Description description */
        7:
          message.description = Description.internalBinaryRead(reader, reader.uint32(), options, message.description);
          break;
        case /* int64 unbonding_height */
        8:
          message.unbondingHeight = reader.int64().toBigInt();
          break;
        case /* google.protobuf.Timestamp unbonding_time */
        9:
          message.unbondingTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.unbondingTime);
          break;
        case /* cosmos.staking.v1beta1.Commission commission */
        10:
          message.commission = Commission.internalBinaryRead(reader, reader.uint32(), options, message.commission);
          break;
        case /* string min_self_delegation */
        11:
          message.minSelfDelegation = reader.string();
          break;
        case /* int64 unbonding_on_hold_ref_count */
        12:
          message.unbondingOnHoldRefCount = reader.int64().toBigInt();
          break;
        case /* repeated uint64 unbonding_ids */
        13:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.unbondingIds.push(reader.uint64().toBigInt());
          else
            message.unbondingIds.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.operatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.operatorAddress);
    if (message.consensusPubkey)
      Any.internalBinaryWrite(message.consensusPubkey, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.jailed !== false)
      writer.tag(3, WireType.Varint).bool(message.jailed);
    if (message.status !== 0)
      writer.tag(4, WireType.Varint).int32(message.status);
    if (message.tokens !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.tokens);
    if (message.delegatorShares !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.delegatorShares);
    if (message.description)
      Description.internalBinaryWrite(message.description, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.unbondingHeight !== 0n)
      writer.tag(8, WireType.Varint).int64(message.unbondingHeight);
    if (message.unbondingTime)
      Timestamp.internalBinaryWrite(message.unbondingTime, writer.tag(9, WireType.LengthDelimited).fork(), options).join();
    if (message.commission)
      Commission.internalBinaryWrite(message.commission, writer.tag(10, WireType.LengthDelimited).fork(), options).join();
    if (message.minSelfDelegation !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.minSelfDelegation);
    if (message.unbondingOnHoldRefCount !== 0n)
      writer.tag(12, WireType.Varint).int64(message.unbondingOnHoldRefCount);
    if (message.unbondingIds.length) {
      writer.tag(13, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.unbondingIds.length; i++)
        writer.uint64(message.unbondingIds[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Validator = new Validator$Type();
class ValAddresses$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.ValAddresses", [
      { no: 1, name: "addresses", kind: "scalar", repeat: 2, T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.addresses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string addresses */
        1:
          message.addresses.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.addresses.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.addresses[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ValAddresses = new ValAddresses$Type();
class DVPair$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.DVPair", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_address */
        2:
          message.validatorAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DVPair = new DVPair$Type();
class DVPairs$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.DVPairs", [
      { no: 1, name: "pairs", kind: "message", repeat: 2, T: () => DVPair, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.pairs = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.staking.v1beta1.DVPair pairs */
        1:
          message.pairs.push(DVPair.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.pairs.length; i++)
      DVPair.internalBinaryWrite(message.pairs[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DVPairs = new DVPairs$Type();
class DVVTriplet$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.DVVTriplet", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_src_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "validator_dst_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorSrcAddress = "";
    message.validatorDstAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_src_address */
        2:
          message.validatorSrcAddress = reader.string();
          break;
        case /* string validator_dst_address */
        3:
          message.validatorDstAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorSrcAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorSrcAddress);
    if (message.validatorDstAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.validatorDstAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DVVTriplet = new DVVTriplet$Type();
class DVVTriplets$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.DVVTriplets", [
      { no: 1, name: "triplets", kind: "message", repeat: 2, T: () => DVVTriplet, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.triplets = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.staking.v1beta1.DVVTriplet triplets */
        1:
          message.triplets.push(DVVTriplet.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.triplets.length; i++)
      DVVTriplet.internalBinaryWrite(message.triplets[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DVVTriplets = new DVVTriplets$Type();
class Delegation$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.Delegation", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "shares", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec" } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorAddress = "";
    message.shares = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_address */
        2:
          message.validatorAddress = reader.string();
          break;
        case /* string shares */
        3:
          message.shares = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.shares !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.shares);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Delegation = new Delegation$Type();
class UnbondingDelegation$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.UnbondingDelegation", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "entries", kind: "message", repeat: 2, T: () => UnbondingDelegationEntry, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorAddress = "";
    message.entries = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_address */
        2:
          message.validatorAddress = reader.string();
          break;
        case /* repeated cosmos.staking.v1beta1.UnbondingDelegationEntry entries */
        3:
          message.entries.push(UnbondingDelegationEntry.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddress);
    for (let i = 0; i < message.entries.length; i++)
      UnbondingDelegationEntry.internalBinaryWrite(message.entries[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const UnbondingDelegation = new UnbondingDelegation$Type();
class UnbondingDelegationEntry$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.UnbondingDelegationEntry", [
      {
        no: 1,
        name: "creation_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "completion_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      { no: 3, name: "initial_balance", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } },
      { no: 4, name: "balance", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } },
      {
        no: 5,
        name: "unbonding_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "unbonding_on_hold_ref_count",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.creationHeight = 0n;
    message.initialBalance = "";
    message.balance = "";
    message.unbondingId = 0n;
    message.unbondingOnHoldRefCount = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 creation_height */
        1:
          message.creationHeight = reader.int64().toBigInt();
          break;
        case /* google.protobuf.Timestamp completion_time */
        2:
          message.completionTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.completionTime);
          break;
        case /* string initial_balance */
        3:
          message.initialBalance = reader.string();
          break;
        case /* string balance */
        4:
          message.balance = reader.string();
          break;
        case /* uint64 unbonding_id */
        5:
          message.unbondingId = reader.uint64().toBigInt();
          break;
        case /* int64 unbonding_on_hold_ref_count */
        6:
          message.unbondingOnHoldRefCount = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.creationHeight !== 0n)
      writer.tag(1, WireType.Varint).int64(message.creationHeight);
    if (message.completionTime)
      Timestamp.internalBinaryWrite(message.completionTime, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.initialBalance !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.initialBalance);
    if (message.balance !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.balance);
    if (message.unbondingId !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.unbondingId);
    if (message.unbondingOnHoldRefCount !== 0n)
      writer.tag(6, WireType.Varint).int64(message.unbondingOnHoldRefCount);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const UnbondingDelegationEntry = new UnbondingDelegationEntry$Type();
class RedelegationEntry$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.RedelegationEntry", [
      {
        no: 1,
        name: "creation_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "completion_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      { no: 3, name: "initial_balance", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } },
      { no: 4, name: "shares_dst", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec" } },
      {
        no: 5,
        name: "unbonding_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "unbonding_on_hold_ref_count",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.creationHeight = 0n;
    message.initialBalance = "";
    message.sharesDst = "";
    message.unbondingId = 0n;
    message.unbondingOnHoldRefCount = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 creation_height */
        1:
          message.creationHeight = reader.int64().toBigInt();
          break;
        case /* google.protobuf.Timestamp completion_time */
        2:
          message.completionTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.completionTime);
          break;
        case /* string initial_balance */
        3:
          message.initialBalance = reader.string();
          break;
        case /* string shares_dst */
        4:
          message.sharesDst = reader.string();
          break;
        case /* uint64 unbonding_id */
        5:
          message.unbondingId = reader.uint64().toBigInt();
          break;
        case /* int64 unbonding_on_hold_ref_count */
        6:
          message.unbondingOnHoldRefCount = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.creationHeight !== 0n)
      writer.tag(1, WireType.Varint).int64(message.creationHeight);
    if (message.completionTime)
      Timestamp.internalBinaryWrite(message.completionTime, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.initialBalance !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.initialBalance);
    if (message.sharesDst !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.sharesDst);
    if (message.unbondingId !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.unbondingId);
    if (message.unbondingOnHoldRefCount !== 0n)
      writer.tag(6, WireType.Varint).int64(message.unbondingOnHoldRefCount);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RedelegationEntry = new RedelegationEntry$Type();
class Redelegation$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.Redelegation", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_src_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "validator_dst_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 4, name: "entries", kind: "message", repeat: 2, T: () => RedelegationEntry, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorSrcAddress = "";
    message.validatorDstAddress = "";
    message.entries = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_src_address */
        2:
          message.validatorSrcAddress = reader.string();
          break;
        case /* string validator_dst_address */
        3:
          message.validatorDstAddress = reader.string();
          break;
        case /* repeated cosmos.staking.v1beta1.RedelegationEntry entries */
        4:
          message.entries.push(RedelegationEntry.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorSrcAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorSrcAddress);
    if (message.validatorDstAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.validatorDstAddress);
    for (let i = 0; i < message.entries.length; i++)
      RedelegationEntry.internalBinaryWrite(message.entries[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Redelegation = new Redelegation$Type();
class Params$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.Params", [
      { no: 1, name: "unbonding_time", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true, "amino.dont_omitempty": true } },
      {
        no: 2,
        name: "max_validators",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "max_entries",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 4,
        name: "historical_entries",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 5,
        name: "bond_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 6, name: "min_commission_rate", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "gogoproto.moretags": 'yaml:"min_commission_rate"', "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": true, "amino.name": "cosmos-sdk/x/staking/Params" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.maxValidators = 0;
    message.maxEntries = 0;
    message.historicalEntries = 0;
    message.bondDenom = "";
    message.minCommissionRate = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Duration unbonding_time */
        1:
          message.unbondingTime = Duration.internalBinaryRead(reader, reader.uint32(), options, message.unbondingTime);
          break;
        case /* uint32 max_validators */
        2:
          message.maxValidators = reader.uint32();
          break;
        case /* uint32 max_entries */
        3:
          message.maxEntries = reader.uint32();
          break;
        case /* uint32 historical_entries */
        4:
          message.historicalEntries = reader.uint32();
          break;
        case /* string bond_denom */
        5:
          message.bondDenom = reader.string();
          break;
        case /* string min_commission_rate */
        6:
          message.minCommissionRate = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.unbondingTime)
      Duration.internalBinaryWrite(message.unbondingTime, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.maxValidators !== 0)
      writer.tag(2, WireType.Varint).uint32(message.maxValidators);
    if (message.maxEntries !== 0)
      writer.tag(3, WireType.Varint).uint32(message.maxEntries);
    if (message.historicalEntries !== 0)
      writer.tag(4, WireType.Varint).uint32(message.historicalEntries);
    if (message.bondDenom !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.bondDenom);
    if (message.minCommissionRate !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.minCommissionRate);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
class DelegationResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.DelegationResponse", [
      { no: 1, name: "delegation", kind: "message", T: () => Delegation, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "balance", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.Delegation delegation */
        1:
          message.delegation = Delegation.internalBinaryRead(reader, reader.uint32(), options, message.delegation);
          break;
        case /* cosmos.base.v1beta1.Coin balance */
        2:
          message.balance = Coin.internalBinaryRead(reader, reader.uint32(), options, message.balance);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegation)
      Delegation.internalBinaryWrite(message.delegation, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.balance)
      Coin.internalBinaryWrite(message.balance, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DelegationResponse = new DelegationResponse$Type();
class RedelegationEntryResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.RedelegationEntryResponse", [
      { no: 1, name: "redelegation_entry", kind: "message", T: () => RedelegationEntry, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 4, name: "balance", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.balance = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.RedelegationEntry redelegation_entry */
        1:
          message.redelegationEntry = RedelegationEntry.internalBinaryRead(reader, reader.uint32(), options, message.redelegationEntry);
          break;
        case /* string balance */
        4:
          message.balance = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.redelegationEntry)
      RedelegationEntry.internalBinaryWrite(message.redelegationEntry, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.balance !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.balance);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RedelegationEntryResponse = new RedelegationEntryResponse$Type();
class RedelegationResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.RedelegationResponse", [
      { no: 1, name: "redelegation", kind: "message", T: () => Redelegation, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "entries", kind: "message", repeat: 2, T: () => RedelegationEntryResponse, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.entries = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.Redelegation redelegation */
        1:
          message.redelegation = Redelegation.internalBinaryRead(reader, reader.uint32(), options, message.redelegation);
          break;
        case /* repeated cosmos.staking.v1beta1.RedelegationEntryResponse entries */
        2:
          message.entries.push(RedelegationEntryResponse.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.redelegation)
      Redelegation.internalBinaryWrite(message.redelegation, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.entries.length; i++)
      RedelegationEntryResponse.internalBinaryWrite(message.entries[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RedelegationResponse = new RedelegationResponse$Type();
class Pool$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.Pool", [
      { no: 1, name: "not_bonded_tokens", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.jsontag": "not_bonded_tokens", "cosmos_proto.scalar": "cosmos.Int", "amino.dont_omitempty": true } },
      { no: 2, name: "bonded_tokens", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "gogoproto.jsontag": "bonded_tokens", "cosmos_proto.scalar": "cosmos.Int", "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": true, "gogoproto.description": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.notBondedTokens = "";
    message.bondedTokens = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string not_bonded_tokens */
        1:
          message.notBondedTokens = reader.string();
          break;
        case /* string bonded_tokens */
        2:
          message.bondedTokens = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.notBondedTokens !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.notBondedTokens);
    if (message.bondedTokens !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.bondedTokens);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Pool = new Pool$Type();
class ValidatorUpdates$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.ValidatorUpdates", [
      { no: 1, name: "updates", kind: "message", repeat: 2, T: () => ValidatorUpdate, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.updates = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cometbft.abci.v1.ValidatorUpdate updates */
        1:
          message.updates.push(ValidatorUpdate.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.updates.length; i++)
      ValidatorUpdate.internalBinaryWrite(message.updates[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ValidatorUpdates = new ValidatorUpdates$Type();
export {
  BondStatus,
  Commission,
  CommissionRates,
  DVPair,
  DVPairs,
  DVVTriplet,
  DVVTriplets,
  Delegation,
  DelegationResponse,
  Description,
  HistoricalInfo,
  Infraction,
  Params,
  Pool,
  Redelegation,
  RedelegationEntry,
  RedelegationEntryResponse,
  RedelegationResponse,
  UnbondingDelegation,
  UnbondingDelegationEntry,
  ValAddresses,
  Validator,
  ValidatorUpdates
};
