// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/staking/v1beta1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./staking_pb.js";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Coin } from "../../base/v1beta1/coin_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
import { CommissionRates } from "./staking_pb.js";
import { Description } from "./staking_pb.js";
class MsgCreateValidator$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgCreateValidator", [
      { no: 1, name: "description", kind: "message", T: () => Description, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "commission", kind: "message", T: () => CommissionRates, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 3, name: "min_self_delegation", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int", "amino.dont_omitempty": true } },
      { no: 4, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 5, name: "validator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 6, name: "pubkey", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.crypto.PubKey" } },
      { no: 7, name: "value", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgCreateValidator", "cosmos.msg.v1.signer": ["validator_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.minSelfDelegation = "";
    message.delegatorAddress = "";
    message.validatorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.Description description */
        1:
          message.description = Description.internalBinaryRead(reader, reader.uint32(), options, message.description);
          break;
        case /* cosmos.staking.v1beta1.CommissionRates commission */
        2:
          message.commission = CommissionRates.internalBinaryRead(reader, reader.uint32(), options, message.commission);
          break;
        case /* string min_self_delegation */
        3:
          message.minSelfDelegation = reader.string();
          break;
        case /* string delegator_address = 4 [deprecated = true] */
        4:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_address */
        5:
          message.validatorAddress = reader.string();
          break;
        case /* google.protobuf.Any pubkey */
        6:
          message.pubkey = Any.internalBinaryRead(reader, reader.uint32(), options, message.pubkey);
          break;
        case /* cosmos.base.v1beta1.Coin value */
        7:
          message.value = Coin.internalBinaryRead(reader, reader.uint32(), options, message.value);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.description)
      Description.internalBinaryWrite(message.description, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.commission)
      CommissionRates.internalBinaryWrite(message.commission, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.minSelfDelegation !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.minSelfDelegation);
    if (message.delegatorAddress !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorAddress !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.pubkey)
      Any.internalBinaryWrite(message.pubkey, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.value)
      Coin.internalBinaryWrite(message.value, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateValidator = new MsgCreateValidator$Type();
class MsgCreateValidatorResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgCreateValidatorResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateValidatorResponse = new MsgCreateValidatorResponse$Type();
class MsgEditValidator$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgEditValidator", [
      { no: 1, name: "description", kind: "message", T: () => Description, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "validator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "commission_rate", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 4, name: "min_self_delegation", kind: "scalar", T: 9, options: { "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgEditValidator", "cosmos.msg.v1.signer": ["validator_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validatorAddress = "";
    message.commissionRate = "";
    message.minSelfDelegation = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.staking.v1beta1.Description description */
        1:
          message.description = Description.internalBinaryRead(reader, reader.uint32(), options, message.description);
          break;
        case /* string validator_address */
        2:
          message.validatorAddress = reader.string();
          break;
        case /* string commission_rate */
        3:
          message.commissionRate = reader.string();
          break;
        case /* string min_self_delegation */
        4:
          message.minSelfDelegation = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.description)
      Description.internalBinaryWrite(message.description, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.validatorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.commissionRate !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.commissionRate);
    if (message.minSelfDelegation !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.minSelfDelegation);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgEditValidator = new MsgEditValidator$Type();
class MsgEditValidatorResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgEditValidatorResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgEditValidatorResponse = new MsgEditValidatorResponse$Type();
class MsgDelegate$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgDelegate", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgDelegate", "cosmos.msg.v1.signer": ["delegator_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_address */
        2:
          message.validatorAddress = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        3:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgDelegate = new MsgDelegate$Type();
class MsgDelegateResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgDelegateResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgDelegateResponse = new MsgDelegateResponse$Type();
class MsgTransferDelegation$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgTransferDelegation", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "receiver_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 4, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgTransferDelegation", "cosmos.msg.v1.signer": ["delegator_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorAddress = "";
    message.receiverAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_address */
        2:
          message.validatorAddress = reader.string();
          break;
        case /* string receiver_address */
        3:
          message.receiverAddress = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        4:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.receiverAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.receiverAddress);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTransferDelegation = new MsgTransferDelegation$Type();
class MsgTransferDelegationResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgTransferDelegationResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTransferDelegationResponse = new MsgTransferDelegationResponse$Type();
class MsgBeginRedelegate$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgBeginRedelegate", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_src_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "validator_dst_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 4, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgBeginRedelegate", "cosmos.msg.v1.signer": ["delegator_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorSrcAddress = "";
    message.validatorDstAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_src_address */
        2:
          message.validatorSrcAddress = reader.string();
          break;
        case /* string validator_dst_address */
        3:
          message.validatorDstAddress = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        4:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorSrcAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorSrcAddress);
    if (message.validatorDstAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.validatorDstAddress);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgBeginRedelegate = new MsgBeginRedelegate$Type();
class MsgBeginRedelegateResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgBeginRedelegateResponse", [
      { no: 1, name: "completion_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Timestamp completion_time */
        1:
          message.completionTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.completionTime);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.completionTime)
      Timestamp.internalBinaryWrite(message.completionTime, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgBeginRedelegateResponse = new MsgBeginRedelegateResponse$Type();
class MsgUndelegate$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgUndelegate", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgUndelegate", "cosmos.msg.v1.signer": ["delegator_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_address */
        2:
          message.validatorAddress = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        3:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUndelegate = new MsgUndelegate$Type();
class MsgUndelegateResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgUndelegateResponse", [
      { no: 1, name: "completion_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      { no: 2, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Timestamp completion_time */
        1:
          message.completionTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.completionTime);
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        2:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.completionTime)
      Timestamp.internalBinaryWrite(message.completionTime, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUndelegateResponse = new MsgUndelegateResponse$Type();
class MsgCancelUnbondingDelegation$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgCancelUnbondingDelegation", [
      { no: 1, name: "delegator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "validator_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.ValidatorAddressString" } },
      { no: 3, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 4,
        name: "creation_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgCancelUnbondingDelegation", "cosmos.msg.v1.signer": ["delegator_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.delegatorAddress = "";
    message.validatorAddress = "";
    message.creationHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string delegator_address */
        1:
          message.delegatorAddress = reader.string();
          break;
        case /* string validator_address */
        2:
          message.validatorAddress = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        3:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        case /* int64 creation_height */
        4:
          message.creationHeight = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.delegatorAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.delegatorAddress);
    if (message.validatorAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.validatorAddress);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.creationHeight !== 0n)
      writer.tag(4, WireType.Varint).int64(message.creationHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCancelUnbondingDelegation = new MsgCancelUnbondingDelegation$Type();
class MsgCancelUnbondingDelegationResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgCancelUnbondingDelegationResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCancelUnbondingDelegationResponse = new MsgCancelUnbondingDelegationResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgUpdateParams", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "amino.name": "cosmos-sdk/x/staking/MsgUpdateParams", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* cosmos.staking.v1beta1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.staking.v1beta1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
const Msg = new ServiceType("cosmos.staking.v1beta1.Msg", [
  { name: "CreateValidator", options: {}, I: MsgCreateValidator, O: MsgCreateValidatorResponse },
  { name: "EditValidator", options: {}, I: MsgEditValidator, O: MsgEditValidatorResponse },
  { name: "Delegate", options: {}, I: MsgDelegate, O: MsgDelegateResponse },
  { name: "TransferDelegation", options: {}, I: MsgTransferDelegation, O: MsgTransferDelegationResponse },
  { name: "BeginRedelegate", options: {}, I: MsgBeginRedelegate, O: MsgBeginRedelegateResponse },
  { name: "Undelegate", options: {}, I: MsgUndelegate, O: MsgUndelegateResponse },
  { name: "CancelUnbondingDelegation", options: {}, I: MsgCancelUnbondingDelegation, O: MsgCancelUnbondingDelegationResponse },
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgBeginRedelegate,
  MsgBeginRedelegateResponse,
  MsgCancelUnbondingDelegation,
  MsgCancelUnbondingDelegationResponse,
  MsgCreateValidator,
  MsgCreateValidatorResponse,
  MsgDelegate,
  MsgDelegateResponse,
  MsgEditValidator,
  MsgEditValidatorResponse,
  MsgTransferDelegation,
  MsgTransferDelegationResponse,
  MsgUndelegate,
  MsgUndelegateResponse,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
