// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/staking/v1beta1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * CreateValidator defines a method for creating a new validator.
   *
   * @generated from protobuf rpc: CreateValidator
   */
  createValidator(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * EditValidator defines a method for editing an existing validator.
   *
   * @generated from protobuf rpc: EditValidator
   */
  editValidator(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Delegate defines a method for performing a delegation of coins
   * from a delegator to a validator.
   *
   * @generated from protobuf rpc: Delegate
   */
  delegate(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TransferDelegation defines a method for transferring a delegation of coins
   * from a delegator to another address.
   *
   * @generated from protobuf rpc: TransferDelegation
   */
  transferDelegation(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * BeginRedelegate defines a method for performing a redelegation
   * of coins from a delegator and source validator to a destination validator.
   *
   * @generated from protobuf rpc: BeginRedelegate
   */
  beginRedelegate(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Undelegate defines a method for performing an undelegation from a
   * delegate and a validator.
   *
   * @generated from protobuf rpc: Undelegate
   */
  undelegate(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * CancelUnbondingDelegation defines a method for performing canceling the unbonding delegation
   * and delegate back to previous validator.
   *
   * Since: cosmos-sdk 0.46
   *
   * @generated from protobuf rpc: CancelUnbondingDelegation
   */
  cancelUnbondingDelegation(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateParams defines an operation for updating the x/staking module
   * parameters.
   * Since: cosmos-sdk 0.47
   *
   * @generated from protobuf rpc: UpdateParams
   */
  updateParams(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
