// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/lightclients/solomachine/v2/solomachine_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Channel } from "../../../core/channel/v1/channel_pb.js";
import { ConnectionEnd } from "../../../core/connection/v1/connection_pb.js";
import { Any } from "../../../../google/protobuf/any_pb.js";
var DataType = /* @__PURE__ */ ((DataType2) => {
  DataType2[DataType2["UNINITIALIZED_UNSPECIFIED"] = 0] = "UNINITIALIZED_UNSPECIFIED";
  DataType2[DataType2["CLIENT_STATE"] = 1] = "CLIENT_STATE";
  DataType2[DataType2["CONSENSUS_STATE"] = 2] = "CONSENSUS_STATE";
  DataType2[DataType2["CONNECTION_STATE"] = 3] = "CONNECTION_STATE";
  DataType2[DataType2["CHANNEL_STATE"] = 4] = "CHANNEL_STATE";
  DataType2[DataType2["PACKET_COMMITMENT"] = 5] = "PACKET_COMMITMENT";
  DataType2[DataType2["PACKET_ACKNOWLEDGEMENT"] = 6] = "PACKET_ACKNOWLEDGEMENT";
  DataType2[DataType2["PACKET_RECEIPT_ABSENCE"] = 7] = "PACKET_RECEIPT_ABSENCE";
  DataType2[DataType2["NEXT_SEQUENCE_RECV"] = 8] = "NEXT_SEQUENCE_RECV";
  DataType2[DataType2["HEADER"] = 9] = "HEADER";
  return DataType2;
})(DataType || {});
class ClientState$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.ClientState", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "is_frozen",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 3, name: "consensus_state", kind: "message", T: () => ConsensusState },
      {
        no: 4,
        name: "allow_update_after_proposal",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    message.isFrozen = false;
    message.allowUpdateAfterProposal = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* bool is_frozen */
        2:
          message.isFrozen = reader.bool();
          break;
        case /* ibc.lightclients.solomachine.v2.ConsensusState consensus_state */
        3:
          message.consensusState = ConsensusState.internalBinaryRead(reader, reader.uint32(), options, message.consensusState);
          break;
        case /* bool allow_update_after_proposal */
        4:
          message.allowUpdateAfterProposal = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    if (message.isFrozen !== false)
      writer.tag(2, WireType.Varint).bool(message.isFrozen);
    if (message.consensusState)
      ConsensusState.internalBinaryWrite(message.consensusState, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.allowUpdateAfterProposal !== false)
      writer.tag(4, WireType.Varint).bool(message.allowUpdateAfterProposal);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ClientState = new ClientState$Type();
class ConsensusState$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.ConsensusState", [
      { no: 1, name: "public_key", kind: "message", T: () => Any },
      {
        no: 2,
        name: "diversifier",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.diversifier = "";
    message.timestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any public_key */
        1:
          message.publicKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.publicKey);
          break;
        case /* string diversifier */
        2:
          message.diversifier = reader.string();
          break;
        case /* uint64 timestamp */
        3:
          message.timestamp = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.publicKey)
      Any.internalBinaryWrite(message.publicKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.diversifier !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.diversifier);
    if (message.timestamp !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.timestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConsensusState = new ConsensusState$Type();
class Header$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.Header", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 4, name: "new_public_key", kind: "message", T: () => Any },
      {
        no: 5,
        name: "new_diversifier",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    message.timestamp = 0n;
    message.signature = new Uint8Array(0);
    message.newDiversifier = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* uint64 timestamp */
        2:
          message.timestamp = reader.uint64().toBigInt();
          break;
        case /* bytes signature */
        3:
          message.signature = reader.bytes();
          break;
        case /* google.protobuf.Any new_public_key */
        4:
          message.newPublicKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.newPublicKey);
          break;
        case /* string new_diversifier */
        5:
          message.newDiversifier = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    if (message.timestamp !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.timestamp);
    if (message.signature.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.signature);
    if (message.newPublicKey)
      Any.internalBinaryWrite(message.newPublicKey, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.newDiversifier !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.newDiversifier);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Header = new Header$Type();
class Misbehaviour$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.Misbehaviour", [
      {
        no: 1,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "signature_one", kind: "message", T: () => SignatureAndData },
      { no: 4, name: "signature_two", kind: "message", T: () => SignatureAndData }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clientId = "";
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string client_id */
        1:
          message.clientId = reader.string();
          break;
        case /* uint64 sequence */
        2:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* ibc.lightclients.solomachine.v2.SignatureAndData signature_one */
        3:
          message.signatureOne = SignatureAndData.internalBinaryRead(reader, reader.uint32(), options, message.signatureOne);
          break;
        case /* ibc.lightclients.solomachine.v2.SignatureAndData signature_two */
        4:
          message.signatureTwo = SignatureAndData.internalBinaryRead(reader, reader.uint32(), options, message.signatureTwo);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.clientId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.clientId);
    if (message.sequence !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.sequence);
    if (message.signatureOne)
      SignatureAndData.internalBinaryWrite(message.signatureOne, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.signatureTwo)
      SignatureAndData.internalBinaryWrite(message.signatureTwo, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Misbehaviour = new Misbehaviour$Type();
class SignatureAndData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.SignatureAndData", [
      {
        no: 1,
        name: "signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "data_type", kind: "enum", T: () => ["ibc.lightclients.solomachine.v2.DataType", DataType, "DATA_TYPE_"] },
      {
        no: 3,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signature = new Uint8Array(0);
    message.dataType = 0;
    message.data = new Uint8Array(0);
    message.timestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes signature */
        1:
          message.signature = reader.bytes();
          break;
        case /* ibc.lightclients.solomachine.v2.DataType data_type */
        2:
          message.dataType = reader.int32();
          break;
        case /* bytes data */
        3:
          message.data = reader.bytes();
          break;
        case /* uint64 timestamp */
        4:
          message.timestamp = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signature.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.signature);
    if (message.dataType !== 0)
      writer.tag(2, WireType.Varint).int32(message.dataType);
    if (message.data.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.data);
    if (message.timestamp !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.timestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignatureAndData = new SignatureAndData$Type();
class TimestampedSignatureData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.TimestampedSignatureData", [
      {
        no: 1,
        name: "signature_data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signatureData = new Uint8Array(0);
    message.timestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes signature_data */
        1:
          message.signatureData = reader.bytes();
          break;
        case /* uint64 timestamp */
        2:
          message.timestamp = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signatureData.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.signatureData);
    if (message.timestamp !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.timestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TimestampedSignatureData = new TimestampedSignatureData$Type();
class SignBytes$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.SignBytes", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "diversifier",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "data_type", kind: "enum", T: () => ["ibc.lightclients.solomachine.v2.DataType", DataType, "DATA_TYPE_"] },
      {
        no: 5,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    message.timestamp = 0n;
    message.diversifier = "";
    message.dataType = 0;
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* uint64 timestamp */
        2:
          message.timestamp = reader.uint64().toBigInt();
          break;
        case /* string diversifier */
        3:
          message.diversifier = reader.string();
          break;
        case /* ibc.lightclients.solomachine.v2.DataType data_type */
        4:
          message.dataType = reader.int32();
          break;
        case /* bytes data */
        5:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    if (message.timestamp !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.timestamp);
    if (message.diversifier !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.diversifier);
    if (message.dataType !== 0)
      writer.tag(4, WireType.Varint).int32(message.dataType);
    if (message.data.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignBytes = new SignBytes$Type();
class HeaderData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.HeaderData", [
      { no: 1, name: "new_pub_key", kind: "message", T: () => Any },
      {
        no: 2,
        name: "new_diversifier",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.newDiversifier = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any new_pub_key */
        1:
          message.newPubKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.newPubKey);
          break;
        case /* string new_diversifier */
        2:
          message.newDiversifier = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.newPubKey)
      Any.internalBinaryWrite(message.newPubKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.newDiversifier !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.newDiversifier);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const HeaderData = new HeaderData$Type();
class ClientStateData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.ClientStateData", [
      {
        no: 1,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "client_state", kind: "message", T: () => Any }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.path = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes path */
        1:
          message.path = reader.bytes();
          break;
        case /* google.protobuf.Any client_state */
        2:
          message.clientState = Any.internalBinaryRead(reader, reader.uint32(), options, message.clientState);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.path);
    if (message.clientState)
      Any.internalBinaryWrite(message.clientState, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ClientStateData = new ClientStateData$Type();
class ConsensusStateData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.ConsensusStateData", [
      {
        no: 1,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "consensus_state", kind: "message", T: () => Any }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.path = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes path */
        1:
          message.path = reader.bytes();
          break;
        case /* google.protobuf.Any consensus_state */
        2:
          message.consensusState = Any.internalBinaryRead(reader, reader.uint32(), options, message.consensusState);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.path);
    if (message.consensusState)
      Any.internalBinaryWrite(message.consensusState, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConsensusStateData = new ConsensusStateData$Type();
class ConnectionStateData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.ConnectionStateData", [
      {
        no: 1,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "connection", kind: "message", T: () => ConnectionEnd }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.path = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes path */
        1:
          message.path = reader.bytes();
          break;
        case /* ibc.core.connection.v1.ConnectionEnd connection */
        2:
          message.connection = ConnectionEnd.internalBinaryRead(reader, reader.uint32(), options, message.connection);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.path);
    if (message.connection)
      ConnectionEnd.internalBinaryWrite(message.connection, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConnectionStateData = new ConnectionStateData$Type();
class ChannelStateData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.ChannelStateData", [
      {
        no: 1,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "channel", kind: "message", T: () => Channel }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.path = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes path */
        1:
          message.path = reader.bytes();
          break;
        case /* ibc.core.channel.v1.Channel channel */
        2:
          message.channel = Channel.internalBinaryRead(reader, reader.uint32(), options, message.channel);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.path);
    if (message.channel)
      Channel.internalBinaryWrite(message.channel, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ChannelStateData = new ChannelStateData$Type();
class PacketCommitmentData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.PacketCommitmentData", [
      {
        no: 1,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "commitment",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.path = new Uint8Array(0);
    message.commitment = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes path */
        1:
          message.path = reader.bytes();
          break;
        case /* bytes commitment */
        2:
          message.commitment = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.path);
    if (message.commitment.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.commitment);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketCommitmentData = new PacketCommitmentData$Type();
class PacketAcknowledgementData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.PacketAcknowledgementData", [
      {
        no: 1,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "acknowledgement",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.path = new Uint8Array(0);
    message.acknowledgement = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes path */
        1:
          message.path = reader.bytes();
          break;
        case /* bytes acknowledgement */
        2:
          message.acknowledgement = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.path);
    if (message.acknowledgement.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.acknowledgement);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketAcknowledgementData = new PacketAcknowledgementData$Type();
class PacketReceiptAbsenceData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.PacketReceiptAbsenceData", [
      {
        no: 1,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.path = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes path */
        1:
          message.path = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.path);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketReceiptAbsenceData = new PacketReceiptAbsenceData$Type();
class NextSequenceRecvData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v2.NextSequenceRecvData", [
      {
        no: 1,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "next_seq_recv",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.path = new Uint8Array(0);
    message.nextSeqRecv = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes path */
        1:
          message.path = reader.bytes();
          break;
        case /* uint64 next_seq_recv */
        2:
          message.nextSeqRecv = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.path);
    if (message.nextSeqRecv !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.nextSeqRecv);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const NextSequenceRecvData = new NextSequenceRecvData$Type();
export {
  ChannelStateData,
  ClientState,
  ClientStateData,
  ConnectionStateData,
  ConsensusState,
  ConsensusStateData,
  DataType,
  Header,
  HeaderData,
  Misbehaviour,
  NextSequenceRecvData,
  PacketAcknowledgementData,
  PacketCommitmentData,
  PacketReceiptAbsenceData,
  SignBytes,
  SignatureAndData,
  TimestampedSignatureData
};
