// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/crypto/ed25519/keys_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class PubKey$Type extends MessageType {
  constructor() {
    super("cosmos.crypto.ed25519.PubKey", [
      { no: 1, name: "key", kind: "scalar", T: 12, options: { "gogoproto.casttype": "crypto/ed25519.PublicKey" } }
    ], { "gogoproto.goproto_stringer": false, "amino.name": "tendermint/PubKeyEd25519", "amino.message_encoding": "key_field" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.key = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes key */
        1:
          message.key = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.key.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.key);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PubKey = new PubKey$Type();
class PrivKey$Type extends MessageType {
  constructor() {
    super("cosmos.crypto.ed25519.PrivKey", [
      { no: 1, name: "key", kind: "scalar", T: 12, options: { "gogoproto.casttype": "crypto/ed25519.PrivateKey" } }
    ], { "amino.name": "tendermint/PrivKeyEd25519", "amino.message_encoding": "key_field" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.key = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes key */
        1:
          message.key = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.key.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.key);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PrivKey = new PrivKey$Type();
export {
  PrivKey,
  PubKey
};
