// @injectivelabs/core-proto-ts-v2/esm/generated/injective/exchange/v2/orderbook_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { DerivativeMarketOrder } from "./order_pb.js";
import { DerivativeLimitOrder } from "./order_pb.js";
import { SpotLimitOrder } from "./order_pb.js";
class SpotOrderBook$Type extends MessageType {
  constructor() {
    super("injective.exchange.v2.SpotOrderBook", [
      {
        no: 1,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "isBuySide",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 3, name: "orders", kind: "message", repeat: 2, T: () => SpotLimitOrder }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.marketId = "";
    message.isBuySide = false;
    message.orders = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string market_id */
        1:
          message.marketId = reader.string();
          break;
        case /* bool isBuySide */
        2:
          message.isBuySide = reader.bool();
          break;
        case /* repeated injective.exchange.v2.SpotLimitOrder orders */
        3:
          message.orders.push(SpotLimitOrder.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.marketId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.marketId);
    if (message.isBuySide !== false)
      writer.tag(2, WireType.Varint).bool(message.isBuySide);
    for (let i = 0; i < message.orders.length; i++)
      SpotLimitOrder.internalBinaryWrite(message.orders[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SpotOrderBook = new SpotOrderBook$Type();
class DerivativeOrderBook$Type extends MessageType {
  constructor() {
    super("injective.exchange.v2.DerivativeOrderBook", [
      {
        no: 1,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "isBuySide",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 3, name: "orders", kind: "message", repeat: 2, T: () => DerivativeLimitOrder }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.marketId = "";
    message.isBuySide = false;
    message.orders = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string market_id */
        1:
          message.marketId = reader.string();
          break;
        case /* bool isBuySide */
        2:
          message.isBuySide = reader.bool();
          break;
        case /* repeated injective.exchange.v2.DerivativeLimitOrder orders */
        3:
          message.orders.push(DerivativeLimitOrder.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.marketId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.marketId);
    if (message.isBuySide !== false)
      writer.tag(2, WireType.Varint).bool(message.isBuySide);
    for (let i = 0; i < message.orders.length; i++)
      DerivativeLimitOrder.internalBinaryWrite(message.orders[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DerivativeOrderBook = new DerivativeOrderBook$Type();
class ConditionalDerivativeOrderBook$Type extends MessageType {
  constructor() {
    super("injective.exchange.v2.ConditionalDerivativeOrderBook", [
      {
        no: 1,
        name: "market_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "limit_buy_orders", kind: "message", repeat: 2, T: () => DerivativeLimitOrder },
      { no: 3, name: "market_buy_orders", kind: "message", repeat: 2, T: () => DerivativeMarketOrder },
      { no: 4, name: "limit_sell_orders", kind: "message", repeat: 2, T: () => DerivativeLimitOrder },
      { no: 5, name: "market_sell_orders", kind: "message", repeat: 2, T: () => DerivativeMarketOrder }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.marketId = "";
    message.limitBuyOrders = [];
    message.marketBuyOrders = [];
    message.limitSellOrders = [];
    message.marketSellOrders = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string market_id */
        1:
          message.marketId = reader.string();
          break;
        case /* repeated injective.exchange.v2.DerivativeLimitOrder limit_buy_orders */
        2:
          message.limitBuyOrders.push(DerivativeLimitOrder.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective.exchange.v2.DerivativeMarketOrder market_buy_orders */
        3:
          message.marketBuyOrders.push(DerivativeMarketOrder.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective.exchange.v2.DerivativeLimitOrder limit_sell_orders */
        4:
          message.limitSellOrders.push(DerivativeLimitOrder.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective.exchange.v2.DerivativeMarketOrder market_sell_orders */
        5:
          message.marketSellOrders.push(DerivativeMarketOrder.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.marketId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.marketId);
    for (let i = 0; i < message.limitBuyOrders.length; i++)
      DerivativeLimitOrder.internalBinaryWrite(message.limitBuyOrders[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.marketBuyOrders.length; i++)
      DerivativeMarketOrder.internalBinaryWrite(message.marketBuyOrders[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.limitSellOrders.length; i++)
      DerivativeLimitOrder.internalBinaryWrite(message.limitSellOrders[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.marketSellOrders.length; i++)
      DerivativeMarketOrder.internalBinaryWrite(message.marketSellOrders[i], writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConditionalDerivativeOrderBook = new ConditionalDerivativeOrderBook$Type();
class SubaccountOrderbookMetadata$Type extends MessageType {
  constructor() {
    super("injective.exchange.v2.SubaccountOrderbookMetadata", [
      {
        no: 1,
        name: "vanilla_limit_order_count",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 2,
        name: "reduce_only_limit_order_count",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      { no: 3, name: "aggregate_reduce_only_quantity", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } },
      { no: 4, name: "aggregate_vanilla_quantity", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } },
      {
        no: 5,
        name: "vanilla_conditional_order_count",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 6,
        name: "reduce_only_conditional_order_count",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.vanillaLimitOrderCount = 0;
    message.reduceOnlyLimitOrderCount = 0;
    message.aggregateReduceOnlyQuantity = "";
    message.aggregateVanillaQuantity = "";
    message.vanillaConditionalOrderCount = 0;
    message.reduceOnlyConditionalOrderCount = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 vanilla_limit_order_count */
        1:
          message.vanillaLimitOrderCount = reader.uint32();
          break;
        case /* uint32 reduce_only_limit_order_count */
        2:
          message.reduceOnlyLimitOrderCount = reader.uint32();
          break;
        case /* string aggregate_reduce_only_quantity */
        3:
          message.aggregateReduceOnlyQuantity = reader.string();
          break;
        case /* string aggregate_vanilla_quantity */
        4:
          message.aggregateVanillaQuantity = reader.string();
          break;
        case /* uint32 vanilla_conditional_order_count */
        5:
          message.vanillaConditionalOrderCount = reader.uint32();
          break;
        case /* uint32 reduce_only_conditional_order_count */
        6:
          message.reduceOnlyConditionalOrderCount = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.vanillaLimitOrderCount !== 0)
      writer.tag(1, WireType.Varint).uint32(message.vanillaLimitOrderCount);
    if (message.reduceOnlyLimitOrderCount !== 0)
      writer.tag(2, WireType.Varint).uint32(message.reduceOnlyLimitOrderCount);
    if (message.aggregateReduceOnlyQuantity !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.aggregateReduceOnlyQuantity);
    if (message.aggregateVanillaQuantity !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.aggregateVanillaQuantity);
    if (message.vanillaConditionalOrderCount !== 0)
      writer.tag(5, WireType.Varint).uint32(message.vanillaConditionalOrderCount);
    if (message.reduceOnlyConditionalOrderCount !== 0)
      writer.tag(6, WireType.Varint).uint32(message.reduceOnlyConditionalOrderCount);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SubaccountOrderbookMetadata = new SubaccountOrderbookMetadata$Type();
export {
  ConditionalDerivativeOrderBook,
  DerivativeOrderBook,
  SpotOrderBook,
  SubaccountOrderbookMetadata
};
