// @injectivelabs/core-proto-ts-v2/esm/generated/tendermint/privval/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Proposal } from "../types/types_pb.js";
import { Vote } from "../types/types_pb.js";
import { PublicKey } from "../crypto/keys_pb.js";
var Errors = /* @__PURE__ */ ((Errors2) => {
  Errors2[Errors2["UNKNOWN"] = 0] = "UNKNOWN";
  Errors2[Errors2["UNEXPECTED_RESPONSE"] = 1] = "UNEXPECTED_RESPONSE";
  Errors2[Errors2["NO_CONNECTION"] = 2] = "NO_CONNECTION";
  Errors2[Errors2["CONNECTION_TIMEOUT"] = 3] = "CONNECTION_TIMEOUT";
  Errors2[Errors2["READ_TIMEOUT"] = 4] = "READ_TIMEOUT";
  Errors2[Errors2["WRITE_TIMEOUT"] = 5] = "WRITE_TIMEOUT";
  return Errors2;
})(Errors || {});
class RemoteSignerError$Type extends MessageType {
  constructor() {
    super("tendermint.privval.RemoteSignerError", [
      {
        no: 1,
        name: "code",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.code = 0;
    message.description = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int32 code */
        1:
          message.code = reader.int32();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.code !== 0)
      writer.tag(1, WireType.Varint).int32(message.code);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RemoteSignerError = new RemoteSignerError$Type();
class PubKeyRequest$Type extends MessageType {
  constructor() {
    super("tendermint.privval.PubKeyRequest", [
      {
        no: 1,
        name: "chain_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chainId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string chain_id */
        1:
          message.chainId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.chainId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.chainId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PubKeyRequest = new PubKeyRequest$Type();
class PubKeyResponse$Type extends MessageType {
  constructor() {
    super("tendermint.privval.PubKeyResponse", [
      { no: 1, name: "pub_key", kind: "message", T: () => PublicKey, options: { "gogoproto.nullable": false } },
      { no: 2, name: "error", kind: "message", T: () => RemoteSignerError }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.crypto.PublicKey pub_key */
        1:
          message.pubKey = PublicKey.internalBinaryRead(reader, reader.uint32(), options, message.pubKey);
          break;
        case /* tendermint.privval.RemoteSignerError error */
        2:
          message.error = RemoteSignerError.internalBinaryRead(reader, reader.uint32(), options, message.error);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pubKey)
      PublicKey.internalBinaryWrite(message.pubKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.error)
      RemoteSignerError.internalBinaryWrite(message.error, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PubKeyResponse = new PubKeyResponse$Type();
class SignVoteRequest$Type extends MessageType {
  constructor() {
    super("tendermint.privval.SignVoteRequest", [
      { no: 1, name: "vote", kind: "message", T: () => Vote },
      {
        no: 2,
        name: "chain_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chainId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.Vote vote */
        1:
          message.vote = Vote.internalBinaryRead(reader, reader.uint32(), options, message.vote);
          break;
        case /* string chain_id */
        2:
          message.chainId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.vote)
      Vote.internalBinaryWrite(message.vote, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.chainId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.chainId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignVoteRequest = new SignVoteRequest$Type();
class SignedVoteResponse$Type extends MessageType {
  constructor() {
    super("tendermint.privval.SignedVoteResponse", [
      { no: 1, name: "vote", kind: "message", T: () => Vote, options: { "gogoproto.nullable": false } },
      { no: 2, name: "error", kind: "message", T: () => RemoteSignerError }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.Vote vote */
        1:
          message.vote = Vote.internalBinaryRead(reader, reader.uint32(), options, message.vote);
          break;
        case /* tendermint.privval.RemoteSignerError error */
        2:
          message.error = RemoteSignerError.internalBinaryRead(reader, reader.uint32(), options, message.error);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.vote)
      Vote.internalBinaryWrite(message.vote, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.error)
      RemoteSignerError.internalBinaryWrite(message.error, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignedVoteResponse = new SignedVoteResponse$Type();
class SignProposalRequest$Type extends MessageType {
  constructor() {
    super("tendermint.privval.SignProposalRequest", [
      { no: 1, name: "proposal", kind: "message", T: () => Proposal },
      {
        no: 2,
        name: "chain_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chainId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.Proposal proposal */
        1:
          message.proposal = Proposal.internalBinaryRead(reader, reader.uint32(), options, message.proposal);
          break;
        case /* string chain_id */
        2:
          message.chainId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposal)
      Proposal.internalBinaryWrite(message.proposal, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.chainId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.chainId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignProposalRequest = new SignProposalRequest$Type();
class SignedProposalResponse$Type extends MessageType {
  constructor() {
    super("tendermint.privval.SignedProposalResponse", [
      { no: 1, name: "proposal", kind: "message", T: () => Proposal, options: { "gogoproto.nullable": false } },
      { no: 2, name: "error", kind: "message", T: () => RemoteSignerError }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.Proposal proposal */
        1:
          message.proposal = Proposal.internalBinaryRead(reader, reader.uint32(), options, message.proposal);
          break;
        case /* tendermint.privval.RemoteSignerError error */
        2:
          message.error = RemoteSignerError.internalBinaryRead(reader, reader.uint32(), options, message.error);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposal)
      Proposal.internalBinaryWrite(message.proposal, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.error)
      RemoteSignerError.internalBinaryWrite(message.error, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignedProposalResponse = new SignedProposalResponse$Type();
class PingRequest$Type extends MessageType {
  constructor() {
    super("tendermint.privval.PingRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PingRequest = new PingRequest$Type();
class PingResponse$Type extends MessageType {
  constructor() {
    super("tendermint.privval.PingResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PingResponse = new PingResponse$Type();
class Message$Type extends MessageType {
  constructor() {
    super("tendermint.privval.Message", [
      { no: 1, name: "pub_key_request", kind: "message", oneof: "sum", T: () => PubKeyRequest },
      { no: 2, name: "pub_key_response", kind: "message", oneof: "sum", T: () => PubKeyResponse },
      { no: 3, name: "sign_vote_request", kind: "message", oneof: "sum", T: () => SignVoteRequest },
      { no: 4, name: "signed_vote_response", kind: "message", oneof: "sum", T: () => SignedVoteResponse },
      { no: 5, name: "sign_proposal_request", kind: "message", oneof: "sum", T: () => SignProposalRequest },
      { no: 6, name: "signed_proposal_response", kind: "message", oneof: "sum", T: () => SignedProposalResponse },
      { no: 7, name: "ping_request", kind: "message", oneof: "sum", T: () => PingRequest },
      { no: 8, name: "ping_response", kind: "message", oneof: "sum", T: () => PingResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sum = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.privval.PubKeyRequest pub_key_request */
        1:
          message.sum = {
            oneofKind: "pubKeyRequest",
            pubKeyRequest: PubKeyRequest.internalBinaryRead(reader, reader.uint32(), options, message.sum.pubKeyRequest)
          };
          break;
        case /* tendermint.privval.PubKeyResponse pub_key_response */
        2:
          message.sum = {
            oneofKind: "pubKeyResponse",
            pubKeyResponse: PubKeyResponse.internalBinaryRead(reader, reader.uint32(), options, message.sum.pubKeyResponse)
          };
          break;
        case /* tendermint.privval.SignVoteRequest sign_vote_request */
        3:
          message.sum = {
            oneofKind: "signVoteRequest",
            signVoteRequest: SignVoteRequest.internalBinaryRead(reader, reader.uint32(), options, message.sum.signVoteRequest)
          };
          break;
        case /* tendermint.privval.SignedVoteResponse signed_vote_response */
        4:
          message.sum = {
            oneofKind: "signedVoteResponse",
            signedVoteResponse: SignedVoteResponse.internalBinaryRead(reader, reader.uint32(), options, message.sum.signedVoteResponse)
          };
          break;
        case /* tendermint.privval.SignProposalRequest sign_proposal_request */
        5:
          message.sum = {
            oneofKind: "signProposalRequest",
            signProposalRequest: SignProposalRequest.internalBinaryRead(reader, reader.uint32(), options, message.sum.signProposalRequest)
          };
          break;
        case /* tendermint.privval.SignedProposalResponse signed_proposal_response */
        6:
          message.sum = {
            oneofKind: "signedProposalResponse",
            signedProposalResponse: SignedProposalResponse.internalBinaryRead(reader, reader.uint32(), options, message.sum.signedProposalResponse)
          };
          break;
        case /* tendermint.privval.PingRequest ping_request */
        7:
          message.sum = {
            oneofKind: "pingRequest",
            pingRequest: PingRequest.internalBinaryRead(reader, reader.uint32(), options, message.sum.pingRequest)
          };
          break;
        case /* tendermint.privval.PingResponse ping_response */
        8:
          message.sum = {
            oneofKind: "pingResponse",
            pingResponse: PingResponse.internalBinaryRead(reader, reader.uint32(), options, message.sum.pingResponse)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sum.oneofKind === "pubKeyRequest")
      PubKeyRequest.internalBinaryWrite(message.sum.pubKeyRequest, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "pubKeyResponse")
      PubKeyResponse.internalBinaryWrite(message.sum.pubKeyResponse, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "signVoteRequest")
      SignVoteRequest.internalBinaryWrite(message.sum.signVoteRequest, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "signedVoteResponse")
      SignedVoteResponse.internalBinaryWrite(message.sum.signedVoteResponse, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "signProposalRequest")
      SignProposalRequest.internalBinaryWrite(message.sum.signProposalRequest, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "signedProposalResponse")
      SignedProposalResponse.internalBinaryWrite(message.sum.signedProposalResponse, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "pingRequest")
      PingRequest.internalBinaryWrite(message.sum.pingRequest, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "pingResponse")
      PingResponse.internalBinaryWrite(message.sum.pingResponse, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Message = new Message$Type();
export {
  Errors,
  Message,
  PingRequest,
  PingResponse,
  PubKeyRequest,
  PubKeyResponse,
  RemoteSignerError,
  SignProposalRequest,
  SignVoteRequest,
  SignedProposalResponse,
  SignedVoteResponse
};
