// @injectivelabs/core-proto-ts-v2/esm/generated/injective/types/v1beta1/indexer_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class TxResult$Type extends MessageType {
  constructor() {
    super("injective.types.v1beta1.TxResult", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "tx_index",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "msg_index",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 4,
        name: "eth_tx_index",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      {
        no: 5,
        name: "failed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 6,
        name: "gas_used",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "cumulative_gas_used",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    message.txIndex = 0;
    message.msgIndex = 0;
    message.ethTxIndex = 0;
    message.failed = false;
    message.gasUsed = 0n;
    message.cumulativeGasUsed = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        case /* uint32 tx_index */
        2:
          message.txIndex = reader.uint32();
          break;
        case /* uint32 msg_index */
        3:
          message.msgIndex = reader.uint32();
          break;
        case /* int32 eth_tx_index */
        4:
          message.ethTxIndex = reader.int32();
          break;
        case /* bool failed */
        5:
          message.failed = reader.bool();
          break;
        case /* uint64 gas_used */
        6:
          message.gasUsed = reader.uint64().toBigInt();
          break;
        case /* uint64 cumulative_gas_used */
        7:
          message.cumulativeGasUsed = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    if (message.txIndex !== 0)
      writer.tag(2, WireType.Varint).uint32(message.txIndex);
    if (message.msgIndex !== 0)
      writer.tag(3, WireType.Varint).uint32(message.msgIndex);
    if (message.ethTxIndex !== 0)
      writer.tag(4, WireType.Varint).int32(message.ethTxIndex);
    if (message.failed !== false)
      writer.tag(5, WireType.Varint).bool(message.failed);
    if (message.gasUsed !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.gasUsed);
    if (message.cumulativeGasUsed !== 0n)
      writer.tag(7, WireType.Varint).uint64(message.cumulativeGasUsed);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TxResult = new TxResult$Type();
export {
  TxResult
};
