// @injectivelabs/core-proto-ts-v2/esm/generated/injective/types/v1beta1/tx_ext_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class ExtensionOptionsWeb3Tx$Type extends MessageType {
  constructor() {
    super("injective.types.v1beta1.ExtensionOptionsWeb3Tx", [
      {
        no: 1,
        name: "typedDataChainID",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "feePayer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "feePayerSig",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.typedDataChainID = 0n;
    message.feePayer = "";
    message.feePayerSig = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 typedDataChainID */
        1:
          message.typedDataChainID = reader.uint64().toBigInt();
          break;
        case /* string feePayer */
        2:
          message.feePayer = reader.string();
          break;
        case /* bytes feePayerSig */
        3:
          message.feePayerSig = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.typedDataChainID !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.typedDataChainID);
    if (message.feePayer !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.feePayer);
    if (message.feePayerSig.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.feePayerSig);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ExtensionOptionsWeb3Tx = new ExtensionOptionsWeb3Tx$Type();
export {
  ExtensionOptionsWeb3Tx
};
