// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/tx/signing/v1beta1/signing_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { CompactBitArray } from "../../../crypto/multisig/v1beta1/multisig_pb.js";
import { Any } from "../../../../google/protobuf/any_pb.js";
var SignMode = /* @__PURE__ */ ((SignMode2) => {
  SignMode2[SignMode2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  SignMode2[SignMode2["DIRECT"] = 1] = "DIRECT";
  SignMode2[SignMode2["TEXTUAL"] = 2] = "TEXTUAL";
  SignMode2[SignMode2["DIRECT_AUX"] = 3] = "DIRECT_AUX";
  SignMode2[SignMode2["LEGACY_AMINO_JSON"] = 127] = "LEGACY_AMINO_JSON";
  SignMode2[SignMode2["EIP712_V2"] = 128] = "EIP712_V2";
  SignMode2[SignMode2["EIP_191"] = 191] = "EIP_191";
  return SignMode2;
})(SignMode || {});
class SignatureDescriptors$Type extends MessageType {
  constructor() {
    super("cosmos.tx.signing.v1beta1.SignatureDescriptors", [
      { no: 1, name: "signatures", kind: "message", repeat: 2, T: () => SignatureDescriptor }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signatures = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.tx.signing.v1beta1.SignatureDescriptor signatures */
        1:
          message.signatures.push(SignatureDescriptor.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.signatures.length; i++)
      SignatureDescriptor.internalBinaryWrite(message.signatures[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignatureDescriptors = new SignatureDescriptors$Type();
class SignatureDescriptor$Type extends MessageType {
  constructor() {
    super("cosmos.tx.signing.v1beta1.SignatureDescriptor", [
      { no: 1, name: "public_key", kind: "message", T: () => Any },
      { no: 2, name: "data", kind: "message", T: () => SignatureDescriptor_Data },
      {
        no: 3,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any public_key */
        1:
          message.publicKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.publicKey);
          break;
        case /* cosmos.tx.signing.v1beta1.SignatureDescriptor.Data data */
        2:
          message.data = SignatureDescriptor_Data.internalBinaryRead(reader, reader.uint32(), options, message.data);
          break;
        case /* uint64 sequence */
        3:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.publicKey)
      Any.internalBinaryWrite(message.publicKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.data)
      SignatureDescriptor_Data.internalBinaryWrite(message.data, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.sequence !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignatureDescriptor = new SignatureDescriptor$Type();
class SignatureDescriptor_Data$Type extends MessageType {
  constructor() {
    super("cosmos.tx.signing.v1beta1.SignatureDescriptor.Data", [
      { no: 1, name: "single", kind: "message", oneof: "sum", T: () => SignatureDescriptor_Data_Single },
      { no: 2, name: "multi", kind: "message", oneof: "sum", T: () => SignatureDescriptor_Data_Multi }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sum = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.tx.signing.v1beta1.SignatureDescriptor.Data.Single single */
        1:
          message.sum = {
            oneofKind: "single",
            single: SignatureDescriptor_Data_Single.internalBinaryRead(reader, reader.uint32(), options, message.sum.single)
          };
          break;
        case /* cosmos.tx.signing.v1beta1.SignatureDescriptor.Data.Multi multi */
        2:
          message.sum = {
            oneofKind: "multi",
            multi: SignatureDescriptor_Data_Multi.internalBinaryRead(reader, reader.uint32(), options, message.sum.multi)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sum.oneofKind === "single")
      SignatureDescriptor_Data_Single.internalBinaryWrite(message.sum.single, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "multi")
      SignatureDescriptor_Data_Multi.internalBinaryWrite(message.sum.multi, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignatureDescriptor_Data = new SignatureDescriptor_Data$Type();
class SignatureDescriptor_Data_Single$Type extends MessageType {
  constructor() {
    super("cosmos.tx.signing.v1beta1.SignatureDescriptor.Data.Single", [
      { no: 1, name: "mode", kind: "enum", T: () => ["cosmos.tx.signing.v1beta1.SignMode", SignMode, "SIGN_MODE_"] },
      {
        no: 2,
        name: "signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.mode = 0;
    message.signature = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.tx.signing.v1beta1.SignMode mode */
        1:
          message.mode = reader.int32();
          break;
        case /* bytes signature */
        2:
          message.signature = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.mode !== 0)
      writer.tag(1, WireType.Varint).int32(message.mode);
    if (message.signature.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.signature);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignatureDescriptor_Data_Single = new SignatureDescriptor_Data_Single$Type();
class SignatureDescriptor_Data_Multi$Type extends MessageType {
  constructor() {
    super("cosmos.tx.signing.v1beta1.SignatureDescriptor.Data.Multi", [
      { no: 1, name: "bitarray", kind: "message", T: () => CompactBitArray },
      { no: 2, name: "signatures", kind: "message", repeat: 2, T: () => SignatureDescriptor_Data }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signatures = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.crypto.multisig.v1beta1.CompactBitArray bitarray */
        1:
          message.bitarray = CompactBitArray.internalBinaryRead(reader, reader.uint32(), options, message.bitarray);
          break;
        case /* repeated cosmos.tx.signing.v1beta1.SignatureDescriptor.Data signatures */
        2:
          message.signatures.push(SignatureDescriptor_Data.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.bitarray)
      CompactBitArray.internalBinaryWrite(message.bitarray, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.signatures.length; i++)
      SignatureDescriptor_Data.internalBinaryWrite(message.signatures[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignatureDescriptor_Data_Multi = new SignatureDescriptor_Data_Multi$Type();
export {
  SignMode,
  SignatureDescriptor,
  SignatureDescriptor_Data,
  SignatureDescriptor_Data_Multi,
  SignatureDescriptor_Data_Single,
  SignatureDescriptors
};
