// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/crypto/multisig/keys_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../google/protobuf/any_pb.js";
class LegacyAminoPubKey$Type extends MessageType {
  constructor() {
    super("cosmos.crypto.multisig.LegacyAminoPubKey", [
      {
        no: 1,
        name: "threshold",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      { no: 2, name: "public_keys", kind: "message", repeat: 2, T: () => Any, options: { "gogoproto.customname": "PubKeys", "amino.field_name": "pubkeys" } }
    ], { "gogoproto.goproto_getters": false, "amino.name": "tendermint/PubKeyMultisigThreshold", "amino.message_encoding": "threshold_string" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.threshold = 0;
    message.publicKeys = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 threshold */
        1:
          message.threshold = reader.uint32();
          break;
        case /* repeated google.protobuf.Any public_keys */
        2:
          message.publicKeys.push(Any.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.threshold !== 0)
      writer.tag(1, WireType.Varint).uint32(message.threshold);
    for (let i = 0; i < message.publicKeys.length; i++)
      Any.internalBinaryWrite(message.publicKeys[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const LegacyAminoPubKey = new LegacyAminoPubKey$Type();
export {
  LegacyAminoPubKey
};
