// @injectivelabs/core-proto-ts-v2/esm/generated/injective/oracle/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Retrieves oracle params
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the band relayers
   *
   * @generated from protobuf rpc: BandRelayers
   */
  bandRelayers(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the state for all band price feeds
   *
   * @generated from protobuf rpc: BandPriceStates
   */
  bandPriceStates(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the state for all band ibc price feeds
   *
   * @generated from protobuf rpc: BandIBCPriceStates
   */
  bandIBCPriceStates(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the state for all price feeds
   *
   * @generated from protobuf rpc: PriceFeedPriceStates
   */
  priceFeedPriceStates(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the state for all coinbase price feeds
   *
   * @generated from protobuf rpc: CoinbasePriceStates
   */
  coinbasePriceStates(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the state for all pyth price feeds
   *
   * @generated from protobuf rpc: PythPriceStates
   */
  pythPriceStates(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the state for all stork price feeds
   *
   * @generated from protobuf rpc: StorkPriceStates
   */
  storkPriceStates(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves all stork publishers
   *
   * @generated from protobuf rpc: StorkPublishers
   */
  storkPublishers(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the state for all provider price feeds
   *
   * @generated from protobuf rpc: ProviderPriceState
   */
  providerPriceState(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the entire oracle module's state
   *
   * @generated from protobuf rpc: OracleModuleState
   */
  oracleModuleState(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves historical price records for a given OracleType and Symbol
   *
   * @generated from protobuf rpc: HistoricalPriceRecords
   */
  historicalPriceRecords(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves mixed volatility value for the specified pair of base/quote
   *
   * @generated from protobuf rpc: OracleVolatility
   */
  oracleVolatility(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: OracleProvidersInfo
   */
  oracleProvidersInfo(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: OracleProviderPrices
   */
  oracleProviderPrices(input, options) {
    const method = this.methods[14], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: OraclePrice
   */
  oraclePrice(input, options) {
    const method = this.methods[15], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: PythPrice
   */
  pythPrice(input, options) {
    const method = this.methods[16], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
