// @injectivelabs/core-proto-ts-v2/esm/generated/tendermint/libs/bits/types_pb.js
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { WireType } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class BitArray$Type extends MessageType {
  constructor() {
    super("tendermint.libs.bits.BitArray", [
      {
        no: 1,
        name: "bits",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "elems",
        kind: "scalar",
        repeat: 1,
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.bits = 0n;
    message.elems = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 bits */
        1:
          message.bits = reader.int64().toBigInt();
          break;
        case /* repeated uint64 elems */
        2:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.elems.push(reader.uint64().toBigInt());
          else
            message.elems.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.bits !== 0n)
      writer.tag(1, WireType.Varint).int64(message.bits);
    if (message.elems.length) {
      writer.tag(2, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.elems.length; i++)
        writer.uint64(message.elems[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BitArray = new BitArray$Type();
export {
  BitArray
};
