// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/base/node/v1beta1/query_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Timestamp } from "../../../../google/protobuf/timestamp_pb.js";
class ConfigRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.node.v1beta1.ConfigRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConfigRequest = new ConfigRequest$Type();
class ConfigResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.node.v1beta1.ConfigResponse", [
      {
        no: 1,
        name: "minimum_gas_price",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "pruning_keep_recent",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "pruning_interval",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "halt_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.minimumGasPrice = "";
    message.pruningKeepRecent = "";
    message.pruningInterval = "";
    message.haltHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string minimum_gas_price */
        1:
          message.minimumGasPrice = reader.string();
          break;
        case /* string pruning_keep_recent */
        2:
          message.pruningKeepRecent = reader.string();
          break;
        case /* string pruning_interval */
        3:
          message.pruningInterval = reader.string();
          break;
        case /* uint64 halt_height */
        4:
          message.haltHeight = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.minimumGasPrice !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.minimumGasPrice);
    if (message.pruningKeepRecent !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.pruningKeepRecent);
    if (message.pruningInterval !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.pruningInterval);
    if (message.haltHeight !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.haltHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConfigResponse = new ConfigResponse$Type();
class StatusRequest$Type extends MessageType {
  constructor() {
    super("cosmos.base.node.v1beta1.StatusRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StatusRequest = new StatusRequest$Type();
class StatusResponse$Type extends MessageType {
  constructor() {
    super("cosmos.base.node.v1beta1.StatusResponse", [
      {
        no: 1,
        name: "earliest_store_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "timestamp", kind: "message", T: () => Timestamp, options: { "gogoproto.stdtime": true } },
      {
        no: 4,
        name: "app_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "validator_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.earliestStoreHeight = 0n;
    message.height = 0n;
    message.appHash = new Uint8Array(0);
    message.validatorHash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 earliest_store_height */
        1:
          message.earliestStoreHeight = reader.uint64().toBigInt();
          break;
        case /* uint64 height */
        2:
          message.height = reader.uint64().toBigInt();
          break;
        case /* google.protobuf.Timestamp timestamp */
        3:
          message.timestamp = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.timestamp);
          break;
        case /* bytes app_hash */
        4:
          message.appHash = reader.bytes();
          break;
        case /* bytes validator_hash */
        5:
          message.validatorHash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.earliestStoreHeight !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.earliestStoreHeight);
    if (message.height !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.height);
    if (message.timestamp)
      Timestamp.internalBinaryWrite(message.timestamp, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.appHash.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.appHash);
    if (message.validatorHash.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.validatorHash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StatusResponse = new StatusResponse$Type();
const Service = new ServiceType("cosmos.base.node.v1beta1.Service", [
  { name: "Config", options: { "google.api.http": { get: "/cosmos/base/node/v1beta1/config" } }, I: ConfigRequest, O: ConfigResponse },
  { name: "Status", options: { "google.api.http": { get: "/cosmos/base/node/v1beta1/status" } }, I: StatusRequest, O: StatusResponse }
]);
export {
  ConfigRequest,
  ConfigResponse,
  Service,
  StatusRequest,
  StatusResponse
};
