// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/orm/v1/orm_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class TableDescriptor$Type extends MessageType {
  constructor() {
    super("cosmos.orm.v1.TableDescriptor", [
      { no: 1, name: "primary_key", kind: "message", T: () => PrimaryKeyDescriptor },
      { no: 2, name: "index", kind: "message", repeat: 2, T: () => SecondaryIndexDescriptor },
      {
        no: 3,
        name: "id",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.index = [];
    message.id = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.orm.v1.PrimaryKeyDescriptor primary_key */
        1:
          message.primaryKey = PrimaryKeyDescriptor.internalBinaryRead(reader, reader.uint32(), options, message.primaryKey);
          break;
        case /* repeated cosmos.orm.v1.SecondaryIndexDescriptor index */
        2:
          message.index.push(SecondaryIndexDescriptor.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* uint32 id */
        3:
          message.id = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.primaryKey)
      PrimaryKeyDescriptor.internalBinaryWrite(message.primaryKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.index.length; i++)
      SecondaryIndexDescriptor.internalBinaryWrite(message.index[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.id !== 0)
      writer.tag(3, WireType.Varint).uint32(message.id);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TableDescriptor = new TableDescriptor$Type();
class PrimaryKeyDescriptor$Type extends MessageType {
  constructor() {
    super("cosmos.orm.v1.PrimaryKeyDescriptor", [
      {
        no: 1,
        name: "fields",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "auto_increment",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.fields = "";
    message.autoIncrement = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string fields */
        1:
          message.fields = reader.string();
          break;
        case /* bool auto_increment */
        2:
          message.autoIncrement = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fields !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.fields);
    if (message.autoIncrement !== false)
      writer.tag(2, WireType.Varint).bool(message.autoIncrement);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PrimaryKeyDescriptor = new PrimaryKeyDescriptor$Type();
class SecondaryIndexDescriptor$Type extends MessageType {
  constructor() {
    super("cosmos.orm.v1.SecondaryIndexDescriptor", [
      {
        no: 1,
        name: "fields",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "id",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "unique",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.fields = "";
    message.id = 0;
    message.unique = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string fields */
        1:
          message.fields = reader.string();
          break;
        case /* uint32 id */
        2:
          message.id = reader.uint32();
          break;
        case /* bool unique */
        3:
          message.unique = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fields !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.fields);
    if (message.id !== 0)
      writer.tag(2, WireType.Varint).uint32(message.id);
    if (message.unique !== false)
      writer.tag(3, WireType.Varint).bool(message.unique);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SecondaryIndexDescriptor = new SecondaryIndexDescriptor$Type();
class SingletonDescriptor$Type extends MessageType {
  constructor() {
    super("cosmos.orm.v1.SingletonDescriptor", [
      {
        no: 1,
        name: "id",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.id = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint32 id */
        1:
          message.id = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.id !== 0)
      writer.tag(1, WireType.Varint).uint32(message.id);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SingletonDescriptor = new SingletonDescriptor$Type();
export {
  PrimaryKeyDescriptor,
  SecondaryIndexDescriptor,
  SingletonDescriptor,
  TableDescriptor
};
