// @injectivelabs/core-proto-ts-v2/esm/generated/tendermint/state/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Timestamp } from "../../google/protobuf/timestamp_pb.js";
import { BlockID } from "../types/types_pb.js";
import { Consensus } from "../version/types_pb.js";
import { ResponseFinalizeBlock } from "../abci/types_pb.js";
import { ValidatorSet } from "../types/validator_pb.js";
import { ConsensusParams } from "../types/params_pb.js";
import { ValidatorUpdate } from "../abci/types_pb.js";
import { Event } from "../abci/types_pb.js";
import { ExecTxResult } from "../abci/types_pb.js";
class LegacyABCIResponses$Type extends MessageType {
  constructor() {
    super("tendermint.state.LegacyABCIResponses", [
      { no: 1, name: "deliver_txs", kind: "message", repeat: 2, T: () => ExecTxResult },
      { no: 2, name: "end_block", kind: "message", T: () => ResponseEndBlock },
      { no: 3, name: "begin_block", kind: "message", T: () => ResponseBeginBlock }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.deliverTxs = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated tendermint.abci.ExecTxResult deliver_txs */
        1:
          message.deliverTxs.push(ExecTxResult.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* tendermint.state.ResponseEndBlock end_block */
        2:
          message.endBlock = ResponseEndBlock.internalBinaryRead(reader, reader.uint32(), options, message.endBlock);
          break;
        case /* tendermint.state.ResponseBeginBlock begin_block */
        3:
          message.beginBlock = ResponseBeginBlock.internalBinaryRead(reader, reader.uint32(), options, message.beginBlock);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.deliverTxs.length; i++)
      ExecTxResult.internalBinaryWrite(message.deliverTxs[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.endBlock)
      ResponseEndBlock.internalBinaryWrite(message.endBlock, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.beginBlock)
      ResponseBeginBlock.internalBinaryWrite(message.beginBlock, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const LegacyABCIResponses = new LegacyABCIResponses$Type();
class ResponseBeginBlock$Type extends MessageType {
  constructor() {
    super("tendermint.state.ResponseBeginBlock", [
      { no: 1, name: "events", kind: "message", repeat: 2, T: () => Event, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "events,omitempty" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.events = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated tendermint.abci.Event events */
        1:
          message.events.push(Event.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.events.length; i++)
      Event.internalBinaryWrite(message.events[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ResponseBeginBlock = new ResponseBeginBlock$Type();
class ResponseEndBlock$Type extends MessageType {
  constructor() {
    super("tendermint.state.ResponseEndBlock", [
      { no: 1, name: "validator_updates", kind: "message", repeat: 2, T: () => ValidatorUpdate, options: { "gogoproto.nullable": false } },
      { no: 2, name: "consensus_param_updates", kind: "message", T: () => ConsensusParams },
      { no: 3, name: "events", kind: "message", repeat: 2, T: () => Event, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "events,omitempty" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.validatorUpdates = [];
    message.events = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated tendermint.abci.ValidatorUpdate validator_updates */
        1:
          message.validatorUpdates.push(ValidatorUpdate.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* tendermint.types.ConsensusParams consensus_param_updates */
        2:
          message.consensusParamUpdates = ConsensusParams.internalBinaryRead(reader, reader.uint32(), options, message.consensusParamUpdates);
          break;
        case /* repeated tendermint.abci.Event events */
        3:
          message.events.push(Event.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.validatorUpdates.length; i++)
      ValidatorUpdate.internalBinaryWrite(message.validatorUpdates[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.consensusParamUpdates)
      ConsensusParams.internalBinaryWrite(message.consensusParamUpdates, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.events.length; i++)
      Event.internalBinaryWrite(message.events[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ResponseEndBlock = new ResponseEndBlock$Type();
class ValidatorsInfo$Type extends MessageType {
  constructor() {
    super("tendermint.state.ValidatorsInfo", [
      { no: 1, name: "validator_set", kind: "message", T: () => ValidatorSet },
      {
        no: 2,
        name: "last_height_changed",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.lastHeightChanged = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.ValidatorSet validator_set */
        1:
          message.validatorSet = ValidatorSet.internalBinaryRead(reader, reader.uint32(), options, message.validatorSet);
          break;
        case /* int64 last_height_changed */
        2:
          message.lastHeightChanged = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.validatorSet)
      ValidatorSet.internalBinaryWrite(message.validatorSet, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.lastHeightChanged !== 0n)
      writer.tag(2, WireType.Varint).int64(message.lastHeightChanged);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ValidatorsInfo = new ValidatorsInfo$Type();
class ConsensusParamsInfo$Type extends MessageType {
  constructor() {
    super("tendermint.state.ConsensusParamsInfo", [
      { no: 1, name: "consensus_params", kind: "message", T: () => ConsensusParams, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "last_height_changed",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.lastHeightChanged = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.ConsensusParams consensus_params */
        1:
          message.consensusParams = ConsensusParams.internalBinaryRead(reader, reader.uint32(), options, message.consensusParams);
          break;
        case /* int64 last_height_changed */
        2:
          message.lastHeightChanged = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.consensusParams)
      ConsensusParams.internalBinaryWrite(message.consensusParams, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.lastHeightChanged !== 0n)
      writer.tag(2, WireType.Varint).int64(message.lastHeightChanged);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConsensusParamsInfo = new ConsensusParamsInfo$Type();
class ABCIResponsesInfo$Type extends MessageType {
  constructor() {
    super("tendermint.state.ABCIResponsesInfo", [
      { no: 1, name: "legacy_abci_responses", kind: "message", T: () => LegacyABCIResponses },
      {
        no: 2,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "response_finalize_block", kind: "message", T: () => ResponseFinalizeBlock }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.state.LegacyABCIResponses legacy_abci_responses */
        1:
          message.legacyAbciResponses = LegacyABCIResponses.internalBinaryRead(reader, reader.uint32(), options, message.legacyAbciResponses);
          break;
        case /* int64 height */
        2:
          message.height = reader.int64().toBigInt();
          break;
        case /* tendermint.abci.ResponseFinalizeBlock response_finalize_block */
        3:
          message.responseFinalizeBlock = ResponseFinalizeBlock.internalBinaryRead(reader, reader.uint32(), options, message.responseFinalizeBlock);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.legacyAbciResponses)
      LegacyABCIResponses.internalBinaryWrite(message.legacyAbciResponses, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.height !== 0n)
      writer.tag(2, WireType.Varint).int64(message.height);
    if (message.responseFinalizeBlock)
      ResponseFinalizeBlock.internalBinaryWrite(message.responseFinalizeBlock, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ABCIResponsesInfo = new ABCIResponsesInfo$Type();
class Version$Type extends MessageType {
  constructor() {
    super("tendermint.state.Version", [
      { no: 1, name: "consensus", kind: "message", T: () => Consensus, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "software",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.software = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.version.Consensus consensus */
        1:
          message.consensus = Consensus.internalBinaryRead(reader, reader.uint32(), options, message.consensus);
          break;
        case /* string software */
        2:
          message.software = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.consensus)
      Consensus.internalBinaryWrite(message.consensus, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.software !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.software);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Version = new Version$Type();
class State$Type extends MessageType {
  constructor() {
    super("tendermint.state.State", [
      { no: 1, name: "version", kind: "message", T: () => Version, options: { "gogoproto.nullable": false } },
      { no: 2, name: "chain_id", kind: "scalar", T: 9, options: { "gogoproto.customname": "ChainID" } },
      {
        no: 14,
        name: "initial_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "last_block_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "last_block_id", kind: "message", T: () => BlockID, options: { "gogoproto.nullable": false, "gogoproto.customname": "LastBlockID" } },
      { no: 5, name: "last_block_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      { no: 6, name: "next_validators", kind: "message", T: () => ValidatorSet },
      { no: 7, name: "validators", kind: "message", T: () => ValidatorSet },
      { no: 8, name: "last_validators", kind: "message", T: () => ValidatorSet },
      {
        no: 9,
        name: "last_height_validators_changed",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 10, name: "consensus_params", kind: "message", T: () => ConsensusParams, options: { "gogoproto.nullable": false } },
      {
        no: 11,
        name: "last_height_consensus_params_changed",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 12,
        name: "last_results_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 13,
        name: "app_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.chainId = "";
    message.initialHeight = 0n;
    message.lastBlockHeight = 0n;
    message.lastHeightValidatorsChanged = 0n;
    message.lastHeightConsensusParamsChanged = 0n;
    message.lastResultsHash = new Uint8Array(0);
    message.appHash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.state.Version version */
        1:
          message.version = Version.internalBinaryRead(reader, reader.uint32(), options, message.version);
          break;
        case /* string chain_id */
        2:
          message.chainId = reader.string();
          break;
        case /* int64 initial_height */
        14:
          message.initialHeight = reader.int64().toBigInt();
          break;
        case /* int64 last_block_height */
        3:
          message.lastBlockHeight = reader.int64().toBigInt();
          break;
        case /* tendermint.types.BlockID last_block_id */
        4:
          message.lastBlockId = BlockID.internalBinaryRead(reader, reader.uint32(), options, message.lastBlockId);
          break;
        case /* google.protobuf.Timestamp last_block_time */
        5:
          message.lastBlockTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.lastBlockTime);
          break;
        case /* tendermint.types.ValidatorSet next_validators */
        6:
          message.nextValidators = ValidatorSet.internalBinaryRead(reader, reader.uint32(), options, message.nextValidators);
          break;
        case /* tendermint.types.ValidatorSet validators */
        7:
          message.validators = ValidatorSet.internalBinaryRead(reader, reader.uint32(), options, message.validators);
          break;
        case /* tendermint.types.ValidatorSet last_validators */
        8:
          message.lastValidators = ValidatorSet.internalBinaryRead(reader, reader.uint32(), options, message.lastValidators);
          break;
        case /* int64 last_height_validators_changed */
        9:
          message.lastHeightValidatorsChanged = reader.int64().toBigInt();
          break;
        case /* tendermint.types.ConsensusParams consensus_params */
        10:
          message.consensusParams = ConsensusParams.internalBinaryRead(reader, reader.uint32(), options, message.consensusParams);
          break;
        case /* int64 last_height_consensus_params_changed */
        11:
          message.lastHeightConsensusParamsChanged = reader.int64().toBigInt();
          break;
        case /* bytes last_results_hash */
        12:
          message.lastResultsHash = reader.bytes();
          break;
        case /* bytes app_hash */
        13:
          message.appHash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.version)
      Version.internalBinaryWrite(message.version, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.chainId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.chainId);
    if (message.lastBlockHeight !== 0n)
      writer.tag(3, WireType.Varint).int64(message.lastBlockHeight);
    if (message.lastBlockId)
      BlockID.internalBinaryWrite(message.lastBlockId, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.lastBlockTime)
      Timestamp.internalBinaryWrite(message.lastBlockTime, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.nextValidators)
      ValidatorSet.internalBinaryWrite(message.nextValidators, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.validators)
      ValidatorSet.internalBinaryWrite(message.validators, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.lastValidators)
      ValidatorSet.internalBinaryWrite(message.lastValidators, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    if (message.lastHeightValidatorsChanged !== 0n)
      writer.tag(9, WireType.Varint).int64(message.lastHeightValidatorsChanged);
    if (message.consensusParams)
      ConsensusParams.internalBinaryWrite(message.consensusParams, writer.tag(10, WireType.LengthDelimited).fork(), options).join();
    if (message.lastHeightConsensusParamsChanged !== 0n)
      writer.tag(11, WireType.Varint).int64(message.lastHeightConsensusParamsChanged);
    if (message.lastResultsHash.length)
      writer.tag(12, WireType.LengthDelimited).bytes(message.lastResultsHash);
    if (message.appHash.length)
      writer.tag(13, WireType.LengthDelimited).bytes(message.appHash);
    if (message.initialHeight !== 0n)
      writer.tag(14, WireType.Varint).int64(message.initialHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const State = new State$Type();
export {
  ABCIResponsesInfo,
  ConsensusParamsInfo,
  LegacyABCIResponses,
  ResponseBeginBlock,
  ResponseEndBlock,
  State,
  ValidatorsInfo,
  Version
};
