// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/auth/module/v1/module_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Module$Type extends MessageType {
  constructor() {
    super("cosmos.auth.module.v1.Module", [
      {
        no: 1,
        name: "bech32_prefix",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "module_account_permissions", kind: "message", repeat: 2, T: () => ModuleAccountPermission },
      {
        no: 3,
        name: "authority",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "cosmos.app.v1alpha1.module": { goImport: "github.com/cosmos/cosmos-sdk/x/auth" } });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.bech32Prefix = "";
    message.moduleAccountPermissions = [];
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string bech32_prefix */
        1:
          message.bech32Prefix = reader.string();
          break;
        case /* repeated cosmos.auth.module.v1.ModuleAccountPermission module_account_permissions */
        2:
          message.moduleAccountPermissions.push(ModuleAccountPermission.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string authority */
        3:
          message.authority = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.bech32Prefix !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.bech32Prefix);
    for (let i = 0; i < message.moduleAccountPermissions.length; i++)
      ModuleAccountPermission.internalBinaryWrite(message.moduleAccountPermissions[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.authority !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.authority);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Module = new Module$Type();
class ModuleAccountPermission$Type extends MessageType {
  constructor() {
    super("cosmos.auth.module.v1.ModuleAccountPermission", [
      {
        no: 1,
        name: "account",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "permissions",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.account = "";
    message.permissions = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string account */
        1:
          message.account = reader.string();
          break;
        case /* repeated string permissions */
        2:
          message.permissions.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.account !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.account);
    for (let i = 0; i < message.permissions.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.permissions[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ModuleAccountPermission = new ModuleAccountPermission$Type();
export {
  Module,
  ModuleAccountPermission
};
