// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/bank/v1beta1/events_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class EventSetBalances$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.EventSetBalances", [
      { no: 1, name: "balance_updates", kind: "message", repeat: 2, T: () => BalanceUpdate }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.balanceUpdates = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.bank.v1beta1.BalanceUpdate balance_updates */
        1:
          message.balanceUpdates.push(BalanceUpdate.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.balanceUpdates.length; i++)
      BalanceUpdate.internalBinaryWrite(message.balanceUpdates[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventSetBalances = new EventSetBalances$Type();
class BalanceUpdate$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.BalanceUpdate", [
      {
        no: 1,
        name: "addr",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "denom",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "amt", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int", "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.addr = new Uint8Array(0);
    message.denom = new Uint8Array(0);
    message.amt = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes addr */
        1:
          message.addr = reader.bytes();
          break;
        case /* bytes denom */
        2:
          message.denom = reader.bytes();
          break;
        case /* string amt */
        3:
          message.amt = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.addr.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.addr);
    if (message.denom.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.denom);
    if (message.amt !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.amt);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BalanceUpdate = new BalanceUpdate$Type();
export {
  BalanceUpdate,
  EventSetBalances
};
