// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/bank/v1beta1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { SendEnabled } from "./bank_pb.js";
import { Params } from "./bank_pb.js";
import { Output } from "./bank_pb.js";
import { Input } from "./bank_pb.js";
import { Coin } from "../../base/v1beta1/coin_pb.js";
class MsgSend$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.MsgSend", [
      { no: 1, name: "from_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "to_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "amount", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgSend", "cosmos.msg.v1.signer": ["from_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.fromAddress = "";
    message.toAddress = "";
    message.amount = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string from_address */
        1:
          message.fromAddress = reader.string();
          break;
        case /* string to_address */
        2:
          message.toAddress = reader.string();
          break;
        case /* repeated cosmos.base.v1beta1.Coin amount */
        3:
          message.amount.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fromAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.fromAddress);
    if (message.toAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.toAddress);
    for (let i = 0; i < message.amount.length; i++)
      Coin.internalBinaryWrite(message.amount[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSend = new MsgSend$Type();
class MsgSendResponse$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.MsgSendResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSendResponse = new MsgSendResponse$Type();
class MsgMultiSend$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.MsgMultiSend", [
      { no: 1, name: "inputs", kind: "message", repeat: 2, T: () => Input, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "outputs", kind: "message", repeat: 2, T: () => Output, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgMultiSend", "cosmos.msg.v1.signer": ["inputs"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.inputs = [];
    message.outputs = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.bank.v1beta1.Input inputs */
        1:
          message.inputs.push(Input.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.bank.v1beta1.Output outputs */
        2:
          message.outputs.push(Output.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.inputs.length; i++)
      Input.internalBinaryWrite(message.inputs[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.outputs.length; i++)
      Output.internalBinaryWrite(message.outputs[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgMultiSend = new MsgMultiSend$Type();
class MsgMultiSendResponse$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.MsgMultiSendResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgMultiSendResponse = new MsgMultiSendResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.MsgUpdateParams", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "amino.name": "cosmos-sdk/x/bank/MsgUpdateParams", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* cosmos.bank.v1beta1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
class MsgSetSendEnabled$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.MsgSetSendEnabled", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "send_enabled", kind: "message", repeat: 2, T: () => SendEnabled },
      {
        no: 3,
        name: "use_default_for",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "amino.name": "cosmos-sdk/MsgSetSendEnabled", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    message.sendEnabled = [];
    message.useDefaultFor = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* repeated cosmos.bank.v1beta1.SendEnabled send_enabled */
        2:
          message.sendEnabled.push(SendEnabled.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated string use_default_for */
        3:
          message.useDefaultFor.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    for (let i = 0; i < message.sendEnabled.length; i++)
      SendEnabled.internalBinaryWrite(message.sendEnabled[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.useDefaultFor.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.useDefaultFor[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSetSendEnabled = new MsgSetSendEnabled$Type();
class MsgSetSendEnabledResponse$Type extends MessageType {
  constructor() {
    super("cosmos.bank.v1beta1.MsgSetSendEnabledResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSetSendEnabledResponse = new MsgSetSendEnabledResponse$Type();
const Msg = new ServiceType("cosmos.bank.v1beta1.Msg", [
  { name: "Send", options: {}, I: MsgSend, O: MsgSendResponse },
  { name: "MultiSend", options: {}, I: MsgMultiSend, O: MsgMultiSendResponse },
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse },
  { name: "SetSendEnabled", options: {}, I: MsgSetSendEnabled, O: MsgSetSendEnabledResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgMultiSend,
  MsgMultiSendResponse,
  MsgSend,
  MsgSendResponse,
  MsgSetSendEnabled,
  MsgSetSendEnabledResponse,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
