// @injectivelabs/core-proto-ts-v2/esm/index.js
import * as InjectiveAuctionV1Beta1AuctionPb from "./generated/injective/auction/v1beta1/auction_pb.js";
import * as InjectiveAuctionV1Beta1GenesisPb from "./generated/injective/auction/v1beta1/genesis_pb.js";
import * as InjectiveAuctionV1Beta1QueryPb from "./generated/injective/auction/v1beta1/query_pb.js";
import * as InjectiveAuctionV1Beta1TxPb from "./generated/injective/auction/v1beta1/tx_pb.js";
import { MsgClient } from "./generated/injective/auction/v1beta1/tx_pb.client.js";
import { QueryClient } from "./generated/injective/auction/v1beta1/query_pb.client.js";
import * as InjectiveCryptoV1Beta1Ethsecp256k1KeysPb from "./generated/injective/crypto/v1beta1/ethsecp256k1/keys_pb.js";
import * as InjectiveErc20V1Beta1Erc20Pb from "./generated/injective/erc20/v1beta1/erc20_pb.js";
import * as InjectiveErc20V1Beta1EventsPb from "./generated/injective/erc20/v1beta1/events_pb.js";
import * as InjectiveErc20V1Beta1GenesisPb from "./generated/injective/erc20/v1beta1/genesis_pb.js";
import * as InjectiveErc20V1Beta1ParamsPb from "./generated/injective/erc20/v1beta1/params_pb.js";
import * as InjectiveErc20V1Beta1QueryPb from "./generated/injective/erc20/v1beta1/query_pb.js";
import * as InjectiveErc20V1Beta1TxPb from "./generated/injective/erc20/v1beta1/tx_pb.js";
import { MsgClient as MsgClient2 } from "./generated/injective/erc20/v1beta1/tx_pb.client.js";
import { QueryClient as QueryClient2 } from "./generated/injective/erc20/v1beta1/query_pb.client.js";
import * as InjectiveEvmV1AccessTuplePb from "./generated/injective/evm/v1/access_tuple_pb.js";
import * as InjectiveEvmV1ChainConfigPb from "./generated/injective/evm/v1/chain_config_pb.js";
import * as InjectiveEvmV1ChainEventsPb from "./generated/injective/evm/v1/events_pb.js";
import * as InjectiveEvmV1GenesisPb from "./generated/injective/evm/v1/genesis_pb.js";
import * as InjectiveEvmV1LogPb from "./generated/injective/evm/v1/log_pb.js";
import * as InjectiveEvmV1ParamsPb from "./generated/injective/evm/v1/params_pb.js";
import * as InjectiveEvmV1QueryPb from "./generated/injective/evm/v1/query_pb.js";
import * as InjectiveEvmV1StatePb from "./generated/injective/evm/v1/state_pb.js";
import * as InjectiveEvmV1TraceConfigPb from "./generated/injective/evm/v1/trace_config_pb.js";
import * as InjectiveEvmV1TransactionLogsPb from "./generated/injective/evm/v1/transaction_logs_pb.js";
import * as InjectiveEvmV1TxResultPb from "./generated/injective/evm/v1/tx_result_pb.js";
import * as InjectiveEvmV1TxPb from "./generated/injective/evm/v1/tx_pb.js";
import { MsgClient as MsgClient3 } from "./generated/injective/evm/v1/tx_pb.client.js";
import { QueryClient as QueryClient3 } from "./generated/injective/evm/v1/query_pb.client.js";
import * as InjectiveExchangeV1Beta1AuthzPb from "./generated/injective/exchange/v1beta1/authz_pb.js";
import * as InjectiveExchangeV1Beta1EventsPb from "./generated/injective/exchange/v1beta1/events_pb.js";
import * as InjectiveExchangeV1Beta1ExchangePb from "./generated/injective/exchange/v1beta1/exchange_pb.js";
import * as InjectiveExchangeV1Beta1GenesisPb from "./generated/injective/exchange/v1beta1/genesis_pb.js";
import * as InjectiveExchangeV1Beta1ProposalPb from "./generated/injective/exchange/v1beta1/proposal_pb.js";
import * as InjectiveExchangeV1Beta1QueryPb from "./generated/injective/exchange/v1beta1/query_pb.js";
import * as InjectiveExchangeV1Beta1TxPb from "./generated/injective/exchange/v1beta1/tx_pb.js";
import { MsgClient as MsgClient4 } from "./generated/injective/exchange/v1beta1/tx_pb.client.js";
import { QueryClient as QueryClient4 } from "./generated/injective/exchange/v1beta1/query_pb.client.js";
import * as InjectiveExchangeV2AuthzPb from "./generated/injective/exchange/v2/authz_pb.js";
import * as InjectiveExchangeV2EventsPb from "./generated/injective/exchange/v2/events_pb.js";
import * as InjectiveExchangeV2ExchangePb from "./generated/injective/exchange/v2/exchange_pb.js";
import * as InjectiveExchangeV2GenesisPb from "./generated/injective/exchange/v2/genesis_pb.js";
import * as InjectiveExchangeV2MarketPb from "./generated/injective/exchange/v2/market_pb.js";
import * as InjectiveExchangeV2OrderPb from "./generated/injective/exchange/v2/order_pb.js";
import * as InjectiveExchangeV2OrderbookPb from "./generated/injective/exchange/v2/orderbook_pb.js";
import * as InjectiveExchangeV2ProposalPb from "./generated/injective/exchange/v2/proposal_pb.js";
import * as InjectiveExchangeV2QueryPb from "./generated/injective/exchange/v2/query_pb.js";
import * as InjectiveExchangeV2TxPb from "./generated/injective/exchange/v2/tx_pb.js";
import { MsgClient as MsgClient5 } from "./generated/injective/exchange/v2/tx_pb.client.js";
import { QueryClient as QueryClient5 } from "./generated/injective/exchange/v2/query_pb.client.js";
import * as InjectiveInsuranceV1Beta1EventsPb from "./generated/injective/insurance/v1beta1/events_pb.js";
import * as InjectiveInsuranceV1Beta1GenesisPb from "./generated/injective/insurance/v1beta1/genesis_pb.js";
import * as InjectiveInsuranceV1Beta1InsurancePb from "./generated/injective/insurance/v1beta1/insurance_pb.js";
import * as InjectiveInsuranceV1Beta1QueryPb from "./generated/injective/insurance/v1beta1/query_pb.js";
import * as InjectiveInsuranceV1Beta1TxPb from "./generated/injective/insurance/v1beta1/tx_pb.js";
import { MsgClient as MsgClient6 } from "./generated/injective/insurance/v1beta1/tx_pb.client.js";
import { QueryClient as QueryClient6 } from "./generated/injective/insurance/v1beta1/query_pb.client.js";
import * as InjectiveOcrV1Beta1GenesisPb from "./generated/injective/ocr/v1beta1/genesis_pb.js";
import * as InjectiveOcrV1Beta1OcrPb from "./generated/injective/ocr/v1beta1/ocr_pb.js";
import * as InjectiveOcrV1Beta1QueryPb from "./generated/injective/ocr/v1beta1/query_pb.js";
import * as InjectiveOcrV1Beta1TxPb from "./generated/injective/ocr/v1beta1/tx_pb.js";
import { MsgClient as MsgClient7 } from "./generated/injective/ocr/v1beta1/tx_pb.client.js";
import { QueryClient as QueryClient7 } from "./generated/injective/ocr/v1beta1/query_pb.client.js";
import * as InjectiveOracleV1Beta1EventsPb from "./generated/injective/oracle/v1beta1/events_pb.js";
import * as InjectiveOracleV1Beta1GenesisPb from "./generated/injective/oracle/v1beta1/genesis_pb.js";
import * as InjectiveOracleV1Beta1OraclePb from "./generated/injective/oracle/v1beta1/oracle_pb.js";
import * as InjectiveOracleV1Beta1ProposalPb from "./generated/injective/oracle/v1beta1/proposal_pb.js";
import * as InjectiveOracleV1Beta1QueryPb from "./generated/injective/oracle/v1beta1/query_pb.js";
import * as InjectiveOracleV1Beta1TxPb from "./generated/injective/oracle/v1beta1/tx_pb.js";
import { MsgClient as MsgClient8 } from "./generated/injective/oracle/v1beta1/tx_pb.client.js";
import { QueryClient as QueryClient8 } from "./generated/injective/oracle/v1beta1/query_pb.client.js";
import * as InjectivePeggyV1AttestationPb from "./generated/injective/peggy/v1/attestation_pb.js";
import * as InjectivePeggyV1BatchPb from "./generated/injective/peggy/v1/batch_pb.js";
import * as InjectivePeggyV1EthereumSignerPb from "./generated/injective/peggy/v1/ethereum_signer_pb.js";
import * as InjectivePeggyV1EventsPb from "./generated/injective/peggy/v1/events_pb.js";
import * as InjectivePeggyV1GenesisPb from "./generated/injective/peggy/v1/genesis_pb.js";
import * as InjectivePeggyV1MsgsPb from "./generated/injective/peggy/v1/msgs_pb.js";
import * as InjectivePeggyV1ParamsPb from "./generated/injective/peggy/v1/params_pb.js";
import * as InjectivePeggyV1PoolPb from "./generated/injective/peggy/v1/pool_pb.js";
import * as InjectivePeggyV1QueryPb from "./generated/injective/peggy/v1/query_pb.js";
import * as InjectivePeggyV1TypesPb from "./generated/injective/peggy/v1/types_pb.js";
import { MsgClient as MsgClient9 } from "./generated/injective/peggy/v1/msgs_pb.client.js";
import { QueryClient as QueryClient9 } from "./generated/injective/peggy/v1/query_pb.client.js";
import * as InjectivePermissionsV1Beta1EventsPb from "./generated/injective/permissions/v1beta1/events_pb.js";
import * as InjectivePermissionsV1Beta1GenesisPb from "./generated/injective/permissions/v1beta1/genesis_pb.js";
import * as InjectivePermissionsV1Beta1ParamsPb from "./generated/injective/permissions/v1beta1/params_pb.js";
import * as InjectivePermissionsV1Beta1PermissionsPb from "./generated/injective/permissions/v1beta1/permissions_pb.js";
import * as InjectivePermissionsV1Beta1QueryPb from "./generated/injective/permissions/v1beta1/query_pb.js";
import * as InjectivePermissionsV1Beta1TxPb from "./generated/injective/permissions/v1beta1/tx_pb.js";
import { MsgClient as MsgClient10 } from "./generated/injective/permissions/v1beta1/tx_pb.client.js";
import { QueryClient as QueryClient10 } from "./generated/injective/permissions/v1beta1/query_pb.client.js";
import * as InjectiveStreamV1Beta1QueryPb from "./generated/injective/stream/v1beta1/query_pb.js";
import { StreamClient } from "./generated/injective/stream/v1beta1/query_pb.client.js";
import * as InjectiveStreamV2QueryPb from "./generated/injective/stream/v2/query_pb.js";
import { StreamClient as StreamClient2 } from "./generated/injective/stream/v2/query_pb.client.js";
import * as InjectiveTokenFactoryV1Beta1AuthorityMetadataPb from "./generated/injective/tokenfactory/v1beta1/authorityMetadata_pb.js";
import * as InjectiveTokenFactoryV1Beta1EventsPb from "./generated/injective/tokenfactory/v1beta1/events_pb.js";
import * as InjectiveTokenFactoryV1Beta1GenesisPb from "./generated/injective/tokenfactory/v1beta1/genesis_pb.js";
import * as InjectiveTokenFactoryV1Beta1ParamsPb from "./generated/injective/tokenfactory/v1beta1/params_pb.js";
import * as InjectiveTokenFactoryV1Beta1QueryPb from "./generated/injective/tokenfactory/v1beta1/query_pb.js";
import * as InjectiveTokenFactoryV1Beta1TxPb from "./generated/injective/tokenfactory/v1beta1/tx_pb.js";
import { MsgClient as MsgClient11 } from "./generated/injective/tokenfactory/v1beta1/tx_pb.client.js";
import { QueryClient as QueryClient11 } from "./generated/injective/tokenfactory/v1beta1/query_pb.client.js";
import * as InjectiveTxFeesV1Beta1GenesisPb from "./generated/injective/txfees/v1beta1/genesis_pb.js";
import * as InjectiveTxFeesV1Beta1QueryPb from "./generated/injective/txfees/v1beta1/query_pb.js";
import * as InjectiveTxFeesV1Beta1TxPb from "./generated/injective/txfees/v1beta1/tx_pb.js";
import * as InjectiveTxFeesV1Beta1TxfeesPb from "./generated/injective/txfees/v1beta1/txfees_pb.js";
import { MsgClient as MsgClient12 } from "./generated/injective/txfees/v1beta1/tx_pb.client.js";
import { QueryClient as QueryClient12 } from "./generated/injective/txfees/v1beta1/query_pb.client.js";
import * as InjectiveTypesV1Beta1AccountPb from "./generated/injective/types/v1beta1/account_pb.js";
import * as InjectiveTypesV1Beta1IndexerPb from "./generated/injective/types/v1beta1/indexer_pb.js";
import * as InjectiveTypesV1Beta1TxExtPb from "./generated/injective/types/v1beta1/tx_ext_pb.js";
import * as InjectiveTypesV1TxResponsePb from "./generated/injective/types/v1beta1/tx_response_pb.js";
import * as InjectiveWasmxV1AuthzPb from "./generated/injective/wasmx/v1/authz_pb.js";
import * as InjectiveWasmxV1EventsPb from "./generated/injective/wasmx/v1/events_pb.js";
import * as InjectiveWasmxV1GenesisPb from "./generated/injective/wasmx/v1/genesis_pb.js";
import * as InjectiveWasmxV1ProposalPb from "./generated/injective/wasmx/v1/proposal_pb.js";
import * as InjectiveWasmxV1QueryPb from "./generated/injective/wasmx/v1/query_pb.js";
import * as InjectiveWasmxV1TxPb from "./generated/injective/wasmx/v1/tx_pb.js";
import * as InjectiveWasmxV1WasmxPb from "./generated/injective/wasmx/v1/wasmx_pb.js";
import { MsgClient as MsgClient13 } from "./generated/injective/wasmx/v1/tx_pb.client.js";
import { QueryClient as QueryClient13 } from "./generated/injective/wasmx/v1/query_pb.client.js";
import * as CosmosAuthV1Beta1AuthPb from "./generated/cosmos/auth/v1beta1/auth_pb.js";
import * as CosmosAuthV1Beta1GenesisPb from "./generated/cosmos/auth/v1beta1/genesis_pb.js";
import * as CosmosAuthV1Beta1QueryPb from "./generated/cosmos/auth/v1beta1/query_pb.js";
import { MsgClient as MsgClient14 } from "./generated/cosmos/auth/v1beta1/tx_pb.client.js";
import { QueryClient as QueryClient14 } from "./generated/cosmos/auth/v1beta1/query_pb.client.js";
import * as CosmosAuthzV1Beta1AuthzPb from "./generated/cosmos/authz/v1beta1/authz_pb.js";
import * as CosmosAuthzV1Beta1EventPb from "./generated/cosmos/authz/v1beta1/event_pb.js";
import * as CosmosAuthzV1Beta1GenesisPb from "./generated/cosmos/authz/v1beta1/genesis_pb.js";
import * as CosmosAuthzV1Beta1QueryPb from "./generated/cosmos/authz/v1beta1/query_pb.js";
import * as CosmosAuthzV1Beta1TxPb from "./generated/cosmos/authz/v1beta1/tx_pb.js";
import { QueryClient as QueryClient15 } from "./generated/cosmos/authz/v1beta1/query_pb.client.js";
import { MsgClient as MsgClient15 } from "./generated/cosmos/authz/v1beta1/tx_pb.client.js";
import * as CosmosBankV1Beta1BankPb from "./generated/cosmos/bank/v1beta1/bank_pb.js";
import * as CosmosBankV1Beta1GenesisPb from "./generated/cosmos/bank/v1beta1/genesis_pb.js";
import * as CosmosBankV1Beta1QueryPb from "./generated/cosmos/bank/v1beta1/query_pb.js";
import * as CosmosBankV1Beta1TxPb from "./generated/cosmos/bank/v1beta1/tx_pb.js";
import { QueryClient as QueryClient16 } from "./generated/cosmos/bank/v1beta1/query_pb.client.js";
import { MsgClient as MsgClient16 } from "./generated/cosmos/bank/v1beta1/tx_pb.client.js";
import * as CosmosBaseAbciV1Beta1AbciPb from "./generated/cosmos/base/abci/v1beta1/abci_pb.js";
import * as CosmosBaseKvV1Beta1KvPb from "./generated/cosmos/store/internal/kv/v1beta1/kv_pb.js";
import * as CosmosBaseQueryV1Beta1PaginationPb from "./generated/cosmos/base/query/v1beta1/pagination_pb.js";
import * as CosmosBaseReflectionV1Beta1ReflectionPb from "./generated/cosmos/base/reflection/v1beta1/reflection_pb.js";
import * as CosmosBaseReflectionV2Alpha1ReflectionPb from "./generated/cosmos/base/reflection/v2alpha1/reflection_pb.js";
import * as CosmosBaseSnapshotsV1Beta1SnapshotPb from "./generated/cosmos/store/snapshots/v1/snapshot_pb.js";
import * as CosmosBaseStoreV1Beta1CommitInfoPb from "./generated/cosmos/store/v1beta1/commit_info_pb.js";
import * as CosmosBaseStoreV1Beta1ListeningPb from "./generated/cosmos/store/v1beta1/listening_pb.js";
import * as CosmosBaseTendermintV1Beta1QueryPb from "./generated/cosmos/base/tendermint/v1beta1/query_pb.js";
import * as CosmosBaseV1Beta1CoinPb from "./generated/cosmos/base/v1beta1/coin_pb.js";
import * as CosmosCrisisV1Beta1GenesisPb from "./generated/cosmos/crisis/v1beta1/genesis_pb.js";
import * as CosmosCryptoEd255519KeysPb from "./generated/cosmos/crypto/ed25519/keys_pb.js";
import * as CosmosCryptoMultisigKeysPb from "./generated/cosmos/crypto/multisig/keys_pb.js";
import * as CosmosCryptoV1Beta1MultisigPb from "./generated/cosmos/crypto/multisig/v1beta1/multisig_pb.js";
import * as CosmosCryptoSecp256k1KeysPb from "./generated/cosmos/crypto/secp256k1/keys_pb.js";
import * as CosmosCryptoSecp256r1KeysPb from "./generated/cosmos/crypto/secp256r1/keys_pb.js";
import * as CosmosDistributionV1Beta1DistributionPb from "./generated/cosmos/distribution/v1beta1/distribution_pb.js";
import * as CosmosDistributionV1Beta1GenesisPb from "./generated/cosmos/distribution/v1beta1/genesis_pb.js";
import * as CosmosDistributionV1Beta1QueryPb from "./generated/cosmos/distribution/v1beta1/query_pb.js";
import * as CosmosDistributionV1Beta1TxPb from "./generated/cosmos/distribution/v1beta1/tx_pb.js";
import * as CosmosEvidenceV1Beta1EvidencePb from "./generated/cosmos/evidence/v1beta1/evidence_pb.js";
import * as CosmosEvidenceV1Beta1GenesisPb from "./generated/cosmos/evidence/v1beta1/genesis_pb.js";
import * as CosmosEvidenceV1Beta1QueryPb from "./generated/cosmos/evidence/v1beta1/query_pb.js";
import * as CosmosEvidenceV1Beta1TxPb from "./generated/cosmos/evidence/v1beta1/tx_pb.js";
import * as CosmosFeegrantV1Beta1FeegrantPb from "./generated/cosmos/feegrant/v1beta1/feegrant_pb.js";
import * as CosmosFeegrantV1Beta1GenesisPb from "./generated/cosmos/feegrant/v1beta1/genesis_pb.js";
import * as CosmosFeegrantV1Beta1QueryPb from "./generated/cosmos/feegrant/v1beta1/query_pb.js";
import * as CosmosFeegrantV1Beta1TxPb from "./generated/cosmos/feegrant/v1beta1/tx_pb.js";
import * as CosmosGenutilV1Beta1GenesisPb from "./generated/cosmos/genutil/v1beta1/genesis_pb.js";
import * as CosmosGovV1Beta1GovPb from "./generated/cosmos/gov/v1beta1/gov_pb.js";
import * as CosmosGovV1Beta1GenesisPb from "./generated/cosmos/gov/v1beta1/genesis_pb.js";
import * as CosmosGovV1Beta1QueryPb from "./generated/cosmos/gov/v1beta1/query_pb.js";
import * as CosmosGovV1Beta1TxPb from "./generated/cosmos/gov/v1beta1/tx_pb.js";
import * as CosmosGovV1GovPb from "./generated/cosmos/gov/v1/gov_pb.js";
import * as CosmosGovV1GenesisPb from "./generated/cosmos/gov/v1/genesis_pb.js";
import * as CosmosGovV1QueryPb from "./generated/cosmos/gov/v1/query_pb.js";
import * as CosmosGovV1TxPb from "./generated/cosmos/gov/v1/tx_pb.js";
import * as CosmosMintV1Beta1GenesisPb from "./generated/cosmos/mint/v1beta1/genesis_pb.js";
import * as CosmosMintV1Beta1MintPb from "./generated/cosmos/mint/v1beta1/mint_pb.js";
import * as CosmosMintV1Beta1QueryPb from "./generated/cosmos/mint/v1beta1/query_pb.js";
import * as CosmosParamsV1Beta1ParamsPb from "./generated/cosmos/params/v1beta1/params_pb.js";
import * as CosmosParamsV1Beta1QueryPb from "./generated/cosmos/params/v1beta1/query_pb.js";
import * as CosmosSlashingV1Beta1GenesisPb from "./generated/cosmos/slashing/v1beta1/genesis_pb.js";
import * as CosmosSlashingV1Beta1QueryPb from "./generated/cosmos/slashing/v1beta1/query_pb.js";
import * as CosmosSlashingV1Beta1SlashingPb from "./generated/cosmos/slashing/v1beta1/slashing_pb.js";
import * as CosmosSlashingV1Beta1TxPb from "./generated/cosmos/slashing/v1beta1/tx_pb.js";
import * as CosmosStakingV1Beta1AuthzPb from "./generated/cosmos/staking/v1beta1/authz_pb.js";
import * as CosmosStakingV1Beta1GenesisPb from "./generated/cosmos/staking/v1beta1/genesis_pb.js";
import * as CosmosStakingV1Beta1QueryPb from "./generated/cosmos/staking/v1beta1/query_pb.js";
import * as CosmosStakingV1Beta1StakingPb from "./generated/cosmos/staking/v1beta1/staking_pb.js";
import * as CosmosStakingV1Beta1TxPb from "./generated/cosmos/staking/v1beta1/tx_pb.js";
import * as CosmosTxSigningV1Beta1SigningPb from "./generated/cosmos/tx/signing/v1beta1/signing_pb.js";
import * as CosmosTxV1Beta1ServicePb from "./generated/cosmos/tx/v1beta1/service_pb.js";
import * as CosmosTxV1Beta1TxPb from "./generated/cosmos/tx/v1beta1/tx_pb.js";
import * as CosmosUpgradeV1Beta1QueryPb from "./generated/cosmos/upgrade/v1beta1/query_pb.js";
import * as CosmosUpgradeV1Beta1UpgradePb from "./generated/cosmos/upgrade/v1beta1/upgrade_pb.js";
import * as CosmosVestingV1Beta1TxPb from "./generated/cosmos/vesting/v1beta1/tx_pb.js";
import * as CosmosVestingV1Beta1VestingPb from "./generated/cosmos/vesting/v1beta1/vesting_pb.js";
import * as CosmwasmWasmV1AuthzPb from "./generated/cosmwasm/wasm/v1/authz_pb.js";
import * as CosmwasmWasmV1GenesisPb from "./generated/cosmwasm/wasm/v1/genesis_pb.js";
import * as CosmwasmWasmV1IbcPb from "./generated/cosmwasm/wasm/v1/ibc_pb.js";
import * as CosmwasmWasmV1ProposalLegacyPb from "./generated/cosmwasm/wasm/v1/proposal_legacy_pb.js";
import * as CosmwasmWasmV1QueryPb from "./generated/cosmwasm/wasm/v1/query_pb.js";
import * as CosmwasmWasmV1TxPb from "./generated/cosmwasm/wasm/v1/tx_pb.js";
import * as CosmwasmWasmV1TypesPb from "./generated/cosmwasm/wasm/v1/types_pb.js";
import * as GoogleApiHttpPb from "./generated/google/api/http_pb.js";
import * as GoogleProtobufAnyPb from "./generated/google/protobuf/any_pb.js";
import * as GoogleProtobufDescriptorPb from "./generated/google/protobuf/descriptor_pb.js";
import * as GoogleProtobufDurationPb from "./generated/google/protobuf/duration_pb.js";
import * as GoogleProtobufTimestampPb from "./generated/google/protobuf/timestamp_pb.js";
import * as IbcApplicationsTransferV1QueryPb from "./generated/ibc/applications/transfer/v1/query_pb.js";
import * as IbcApplicationsTransferV1TransferPb from "./generated/ibc/applications/transfer/v1/transfer_pb.js";
import * as IbcApplicationsTransferV1TxPb from "./generated/ibc/applications/transfer/v1/tx_pb.js";
import * as IbcCoreChannelV1ChannelPb from "./generated/ibc/core/channel/v1/channel_pb.js";
import * as IbcCoreChannelV1GenesisPb from "./generated/ibc/core/channel/v1/genesis_pb.js";
import * as IbcCoreChannelV1QueryPb from "./generated/ibc/core/channel/v1/query_pb.js";
import * as IbcCoreChannelV1TxPb from "./generated/ibc/core/channel/v1/tx_pb.js";
import * as IbcCoreClientV1ClientPb from "./generated/ibc/core/client/v1/client_pb.js";
import * as IbcCoreClientV1GenesisPb from "./generated/ibc/core/client/v1/genesis_pb.js";
import * as IbcCoreClientV1QueryPb from "./generated/ibc/core/client/v1/query_pb.js";
import * as IbcCoreClientV1TxPb from "./generated/ibc/core/client/v1/tx_pb.js";
import * as IbcCoreCommitmentV1CommitmentPb from "./generated/ibc/core/commitment/v1/commitment_pb.js";
import * as IbcCoreConnectionV1ConnectionPb from "./generated/ibc/core/connection/v1/connection_pb.js";
import * as IbcCoreConnectionV1GenesisPb from "./generated/ibc/core/connection/v1/genesis_pb.js";
import * as IbcCoreConnectionV1QueryPb from "./generated/ibc/core/connection/v1/query_pb.js";
import * as IbcCoreConnectionV1TxPb from "./generated/ibc/core/connection/v1/tx_pb.js";
import * as IbcCoreTypesV1GenesisPb from "./generated/ibc/core/types/v1/genesis_pb.js";
import * as IbcLightcientsSolomachineV2SolomachinePb from "./generated/ibc/lightclients/solomachine/v2/solomachine_pb.js";
import * as IbcLightcientsTendermintV1TendermintPb from "./generated/ibc/lightclients/tendermint/v1/tendermint_pb.js";
import * as CometAbciV1TypesPb from "./generated/cometbft/abci/v1/types_pb.js";
import * as CometCryptoV1KeysPb from "./generated/cometbft/crypto/v1/keys_pb.js";
import * as CometCryptoV1ProofPb from "./generated/cometbft/crypto/v1/proof_pb.js";
import * as CometP2PV1ConnPb from "./generated/cometbft/p2p/v1/types_pb.js";
import * as CometP2PV1TypesPb from "./generated/cometbft/p2p/v1/types_pb.js";
import * as CometTypesV1BlockPb from "./generated/cometbft/types/v1/block_pb.js";
import * as CometTypesV1EvidencePb from "./generated/cometbft/types/v1/evidence_pb.js";
import * as CometTypesV1ParamsPb from "./generated/cometbft/types/v1/params_pb.js";
import * as CometTypesV1TypesPb from "./generated/cometbft/types/v1/types_pb.js";
import * as CometTypesV1ValidatorPb from "./generated/cometbft/types/v1/validator_pb.js";
import * as CometVersionV1TypesPb from "./generated/cometbft/version/v1/types_pb.js";
import * as ConfioProofsPb from "./generated/cosmos/ics23/v1/proofs_pb.js";
export {
  CometAbciV1TypesPb,
  CometCryptoV1KeysPb,
  CometCryptoV1ProofPb,
  CometP2PV1ConnPb,
  CometP2PV1TypesPb,
  CometTypesV1BlockPb,
  CometTypesV1EvidencePb,
  CometTypesV1ParamsPb,
  CometTypesV1TypesPb,
  CometTypesV1ValidatorPb,
  CometVersionV1TypesPb,
  ConfioProofsPb,
  CosmosAuthV1Beta1AuthPb,
  CosmosAuthV1Beta1GenesisPb,
  CosmosAuthV1Beta1QueryPb,
  MsgClient14 as CosmosAuthV1BetaMsgClient,
  QueryClient14 as CosmosAuthV1BetaQueryClient,
  CosmosAuthzV1Beta1AuthzPb,
  CosmosAuthzV1Beta1EventPb,
  CosmosAuthzV1Beta1GenesisPb,
  CosmosAuthzV1Beta1QueryPb,
  CosmosAuthzV1Beta1TxPb,
  MsgClient15 as CosmosAuthzV1BetaMsgClient,
  QueryClient15 as CosmosAuthzV1BetaQueryClient,
  CosmosBankV1Beta1BankPb,
  CosmosBankV1Beta1GenesisPb,
  CosmosBankV1Beta1QueryPb,
  CosmosBankV1Beta1TxPb,
  MsgClient16 as CosmosBankV1BetaMsgClient,
  QueryClient16 as CosmosBankV1BetaQueryClient,
  CosmosBaseAbciV1Beta1AbciPb,
  CosmosBaseKvV1Beta1KvPb,
  CosmosBaseQueryV1Beta1PaginationPb,
  CosmosBaseReflectionV1Beta1ReflectionPb,
  CosmosBaseReflectionV2Alpha1ReflectionPb,
  CosmosBaseSnapshotsV1Beta1SnapshotPb,
  CosmosBaseStoreV1Beta1CommitInfoPb,
  CosmosBaseStoreV1Beta1ListeningPb,
  CosmosBaseTendermintV1Beta1QueryPb,
  CosmosBaseV1Beta1CoinPb,
  CosmosCrisisV1Beta1GenesisPb,
  CosmosCryptoEd255519KeysPb,
  CosmosCryptoMultisigKeysPb,
  CosmosCryptoSecp256k1KeysPb,
  CosmosCryptoSecp256r1KeysPb,
  CosmosCryptoV1Beta1MultisigPb,
  CosmosDistributionV1Beta1DistributionPb,
  CosmosDistributionV1Beta1GenesisPb,
  CosmosDistributionV1Beta1QueryPb,
  CosmosDistributionV1Beta1TxPb,
  CosmosEvidenceV1Beta1EvidencePb,
  CosmosEvidenceV1Beta1GenesisPb,
  CosmosEvidenceV1Beta1QueryPb,
  CosmosEvidenceV1Beta1TxPb,
  CosmosFeegrantV1Beta1FeegrantPb,
  CosmosFeegrantV1Beta1GenesisPb,
  CosmosFeegrantV1Beta1QueryPb,
  CosmosFeegrantV1Beta1TxPb,
  CosmosGenutilV1Beta1GenesisPb,
  CosmosGovV1Beta1GenesisPb,
  CosmosGovV1Beta1GovPb,
  CosmosGovV1Beta1QueryPb,
  CosmosGovV1Beta1TxPb,
  CosmosGovV1GenesisPb,
  CosmosGovV1GovPb,
  CosmosGovV1QueryPb,
  CosmosGovV1TxPb,
  CosmosMintV1Beta1GenesisPb,
  CosmosMintV1Beta1MintPb,
  CosmosMintV1Beta1QueryPb,
  CosmosParamsV1Beta1ParamsPb,
  CosmosParamsV1Beta1QueryPb,
  CosmosSlashingV1Beta1GenesisPb,
  CosmosSlashingV1Beta1QueryPb,
  CosmosSlashingV1Beta1SlashingPb,
  CosmosSlashingV1Beta1TxPb,
  CosmosStakingV1Beta1AuthzPb,
  CosmosStakingV1Beta1GenesisPb,
  CosmosStakingV1Beta1QueryPb,
  CosmosStakingV1Beta1StakingPb,
  CosmosStakingV1Beta1TxPb,
  CosmosTxSigningV1Beta1SigningPb,
  CosmosTxV1Beta1ServicePb,
  CosmosTxV1Beta1TxPb,
  CosmosUpgradeV1Beta1QueryPb,
  CosmosUpgradeV1Beta1UpgradePb,
  CosmosVestingV1Beta1TxPb,
  CosmosVestingV1Beta1VestingPb,
  CosmwasmWasmV1AuthzPb,
  CosmwasmWasmV1GenesisPb,
  CosmwasmWasmV1IbcPb,
  CosmwasmWasmV1ProposalLegacyPb,
  CosmwasmWasmV1QueryPb,
  CosmwasmWasmV1TxPb,
  CosmwasmWasmV1TypesPb,
  GoogleApiHttpPb,
  GoogleProtobufAnyPb,
  GoogleProtobufDescriptorPb,
  GoogleProtobufDurationPb,
  GoogleProtobufTimestampPb,
  IbcApplicationsTransferV1QueryPb,
  IbcApplicationsTransferV1TransferPb,
  IbcApplicationsTransferV1TxPb,
  IbcCoreChannelV1ChannelPb,
  IbcCoreChannelV1GenesisPb,
  IbcCoreChannelV1QueryPb,
  IbcCoreChannelV1TxPb,
  IbcCoreClientV1ClientPb,
  IbcCoreClientV1GenesisPb,
  IbcCoreClientV1QueryPb,
  IbcCoreClientV1TxPb,
  IbcCoreCommitmentV1CommitmentPb,
  IbcCoreConnectionV1ConnectionPb,
  IbcCoreConnectionV1GenesisPb,
  IbcCoreConnectionV1QueryPb,
  IbcCoreConnectionV1TxPb,
  IbcCoreTypesV1GenesisPb,
  IbcLightcientsSolomachineV2SolomachinePb,
  IbcLightcientsTendermintV1TendermintPb,
  InjectiveAuctionV1Beta1AuctionPb,
  InjectiveAuctionV1Beta1GenesisPb,
  MsgClient as InjectiveAuctionV1Beta1MsgClient,
  QueryClient as InjectiveAuctionV1Beta1QueryClient,
  InjectiveAuctionV1Beta1QueryPb,
  InjectiveAuctionV1Beta1TxPb,
  InjectiveCryptoV1Beta1Ethsecp256k1KeysPb,
  InjectiveErc20V1Beta1Erc20Pb,
  InjectiveErc20V1Beta1EventsPb,
  InjectiveErc20V1Beta1GenesisPb,
  MsgClient2 as InjectiveErc20V1Beta1MsgClient,
  InjectiveErc20V1Beta1ParamsPb,
  QueryClient2 as InjectiveErc20V1Beta1QueryClient,
  InjectiveErc20V1Beta1QueryPb,
  InjectiveErc20V1Beta1TxPb,
  InjectiveEvmV1AccessTuplePb,
  InjectiveEvmV1ChainConfigPb,
  InjectiveEvmV1ChainEventsPb,
  InjectiveEvmV1GenesisPb,
  InjectiveEvmV1LogPb,
  MsgClient3 as InjectiveEvmV1MsgClient,
  InjectiveEvmV1ParamsPb,
  QueryClient3 as InjectiveEvmV1QueryClient,
  InjectiveEvmV1QueryPb,
  InjectiveEvmV1StatePb,
  InjectiveEvmV1TraceConfigPb,
  InjectiveEvmV1TransactionLogsPb,
  InjectiveEvmV1TxPb,
  InjectiveEvmV1TxResultPb,
  InjectiveExchangeV1Beta1AuthzPb,
  InjectiveExchangeV1Beta1EventsPb,
  InjectiveExchangeV1Beta1ExchangePb,
  InjectiveExchangeV1Beta1GenesisPb,
  MsgClient4 as InjectiveExchangeV1Beta1MsgClient,
  InjectiveExchangeV1Beta1ProposalPb,
  QueryClient4 as InjectiveExchangeV1Beta1QueryClient,
  InjectiveExchangeV1Beta1QueryPb,
  InjectiveExchangeV1Beta1TxPb,
  InjectiveExchangeV2AuthzPb,
  InjectiveExchangeV2EventsPb,
  InjectiveExchangeV2ExchangePb,
  InjectiveExchangeV2GenesisPb,
  InjectiveExchangeV2MarketPb,
  MsgClient5 as InjectiveExchangeV2MsgClient,
  InjectiveExchangeV2OrderPb,
  InjectiveExchangeV2OrderbookPb,
  InjectiveExchangeV2ProposalPb,
  QueryClient5 as InjectiveExchangeV2QueryClient,
  InjectiveExchangeV2QueryPb,
  InjectiveExchangeV2TxPb,
  InjectiveInsuranceV1Beta1EventsPb,
  InjectiveInsuranceV1Beta1GenesisPb,
  InjectiveInsuranceV1Beta1InsurancePb,
  MsgClient6 as InjectiveInsuranceV1Beta1MsgClient,
  QueryClient6 as InjectiveInsuranceV1Beta1QueryClient,
  InjectiveInsuranceV1Beta1QueryPb,
  InjectiveInsuranceV1Beta1TxPb,
  InjectiveOcrV1Beta1GenesisPb,
  MsgClient7 as InjectiveOcrV1Beta1MsgClient,
  InjectiveOcrV1Beta1OcrPb,
  QueryClient7 as InjectiveOcrV1Beta1QueryClient,
  InjectiveOcrV1Beta1QueryPb,
  InjectiveOcrV1Beta1TxPb,
  InjectiveOracleV1Beta1EventsPb,
  InjectiveOracleV1Beta1GenesisPb,
  MsgClient8 as InjectiveOracleV1Beta1MsgClient,
  InjectiveOracleV1Beta1OraclePb,
  InjectiveOracleV1Beta1ProposalPb,
  QueryClient8 as InjectiveOracleV1Beta1QueryClient,
  InjectiveOracleV1Beta1QueryPb,
  InjectiveOracleV1Beta1TxPb,
  InjectivePeggyV1AttestationPb,
  InjectivePeggyV1BatchPb,
  InjectivePeggyV1EthereumSignerPb,
  InjectivePeggyV1EventsPb,
  InjectivePeggyV1GenesisPb,
  MsgClient9 as InjectivePeggyV1MsgClient,
  InjectivePeggyV1MsgsPb,
  InjectivePeggyV1ParamsPb,
  InjectivePeggyV1PoolPb,
  QueryClient9 as InjectivePeggyV1QueryClient,
  InjectivePeggyV1QueryPb,
  InjectivePeggyV1TypesPb,
  InjectivePermissionsV1Beta1EventsPb,
  InjectivePermissionsV1Beta1GenesisPb,
  MsgClient10 as InjectivePermissionsV1Beta1MsgClient,
  InjectivePermissionsV1Beta1ParamsPb,
  InjectivePermissionsV1Beta1PermissionsPb,
  QueryClient10 as InjectivePermissionsV1Beta1QueryClient,
  InjectivePermissionsV1Beta1QueryPb,
  InjectivePermissionsV1Beta1TxPb,
  StreamClient as InjectiveStreamV1Beta1QueryClient,
  InjectiveStreamV1Beta1QueryPb,
  StreamClient2 as InjectiveStreamV2QueryClient,
  InjectiveStreamV2QueryPb,
  InjectiveTokenFactoryV1Beta1AuthorityMetadataPb,
  InjectiveTokenFactoryV1Beta1EventsPb,
  InjectiveTokenFactoryV1Beta1GenesisPb,
  MsgClient11 as InjectiveTokenFactoryV1Beta1MsgClient,
  InjectiveTokenFactoryV1Beta1ParamsPb,
  QueryClient11 as InjectiveTokenFactoryV1Beta1QueryClient,
  InjectiveTokenFactoryV1Beta1QueryPb,
  InjectiveTokenFactoryV1Beta1TxPb,
  InjectiveTxFeesV1Beta1GenesisPb,
  MsgClient12 as InjectiveTxFeesV1Beta1MsgClient,
  QueryClient12 as InjectiveTxFeesV1Beta1QueryClient,
  InjectiveTxFeesV1Beta1QueryPb,
  InjectiveTxFeesV1Beta1TxPb,
  InjectiveTxFeesV1Beta1TxfeesPb,
  InjectiveTypesV1Beta1AccountPb,
  InjectiveTypesV1Beta1IndexerPb,
  InjectiveTypesV1Beta1TxExtPb,
  InjectiveTypesV1TxResponsePb,
  MsgClient13 as InjectiveWasmV1MsgClient,
  QueryClient13 as InjectiveWasmV1QueryClient,
  InjectiveWasmxV1AuthzPb,
  InjectiveWasmxV1EventsPb,
  InjectiveWasmxV1GenesisPb,
  InjectiveWasmxV1ProposalPb,
  InjectiveWasmxV1QueryPb,
  InjectiveWasmxV1TxPb,
  InjectiveWasmxV1WasmxPb
};
