// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/circuit/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Account returns account permissions.
   *
   * @generated from protobuf rpc: Account
   */
  account(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Account returns account permissions.
   *
   * @generated from protobuf rpc: Accounts
   */
  accounts(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * DisabledList returns a list of disabled message urls
   *
   * @generated from protobuf rpc: DisabledList
   */
  disabledList(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
