// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/circuit/v1/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
var Permissions_Level = /* @__PURE__ */ ((Permissions_Level2) => {
  Permissions_Level2[Permissions_Level2["NONE_UNSPECIFIED"] = 0] = "NONE_UNSPECIFIED";
  Permissions_Level2[Permissions_Level2["SOME_MSGS"] = 1] = "SOME_MSGS";
  Permissions_Level2[Permissions_Level2["ALL_MSGS"] = 2] = "ALL_MSGS";
  Permissions_Level2[Permissions_Level2["SUPER_ADMIN"] = 3] = "SUPER_ADMIN";
  return Permissions_Level2;
})(Permissions_Level || {});
class Permissions$Type extends MessageType {
  constructor() {
    super("cosmos.circuit.v1.Permissions", [
      { no: 1, name: "level", kind: "enum", T: () => ["cosmos.circuit.v1.Permissions.Level", Permissions_Level, "LEVEL_"] },
      {
        no: 2,
        name: "limit_type_urls",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.level = 0;
    message.limitTypeUrls = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.circuit.v1.Permissions.Level level */
        1:
          message.level = reader.int32();
          break;
        case /* repeated string limit_type_urls */
        2:
          message.limitTypeUrls.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.level !== 0)
      writer.tag(1, WireType.Varint).int32(message.level);
    for (let i = 0; i < message.limitTypeUrls.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.limitTypeUrls[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Permissions = new Permissions$Type();
class GenesisAccountPermissions$Type extends MessageType {
  constructor() {
    super("cosmos.circuit.v1.GenesisAccountPermissions", [
      {
        no: 1,
        name: "address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "permissions", kind: "message", T: () => Permissions }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* cosmos.circuit.v1.Permissions permissions */
        2:
          message.permissions = Permissions.internalBinaryRead(reader, reader.uint32(), options, message.permissions);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.permissions)
      Permissions.internalBinaryWrite(message.permissions, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisAccountPermissions = new GenesisAccountPermissions$Type();
class GenesisState$Type extends MessageType {
  constructor() {
    super("cosmos.circuit.v1.GenesisState", [
      { no: 1, name: "account_permissions", kind: "message", repeat: 2, T: () => GenesisAccountPermissions },
      {
        no: 2,
        name: "disabled_type_urls",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.accountPermissions = [];
    message.disabledTypeUrls = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.circuit.v1.GenesisAccountPermissions account_permissions */
        1:
          message.accountPermissions.push(GenesisAccountPermissions.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated string disabled_type_urls */
        2:
          message.disabledTypeUrls.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.accountPermissions.length; i++)
      GenesisAccountPermissions.internalBinaryWrite(message.accountPermissions[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.disabledTypeUrls.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.disabledTypeUrls[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
export {
  GenesisAccountPermissions,
  GenesisState,
  Permissions,
  Permissions_Level
};
