// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/distribution/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Params queries params of the distribution module.
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ValidatorDistributionInfo queries validator commission and self-delegation rewards for validator
   *
   * @generated from protobuf rpc: ValidatorDistributionInfo
   */
  validatorDistributionInfo(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ValidatorOutstandingRewards queries rewards of a validator address.
   *
   * @generated from protobuf rpc: ValidatorOutstandingRewards
   */
  validatorOutstandingRewards(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ValidatorCommission queries accumulated commission for a validator.
   *
   * @generated from protobuf rpc: ValidatorCommission
   */
  validatorCommission(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ValidatorSlashes queries slash events of a validator.
   *
   * @generated from protobuf rpc: ValidatorSlashes
   */
  validatorSlashes(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * DelegationRewards queries the total rewards accrued by a delegation.
   *
   * @generated from protobuf rpc: DelegationRewards
   */
  delegationRewards(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * DelegationTotalRewards queries the total rewards accrued by each
   * validator.
   *
   * @generated from protobuf rpc: DelegationTotalRewards
   */
  delegationTotalRewards(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * DelegatorValidators queries the validators of a delegator.
   *
   * @generated from protobuf rpc: DelegatorValidators
   */
  delegatorValidators(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * DelegatorWithdrawAddress queries withdraw address of a delegator.
   *
   * @generated from protobuf rpc: DelegatorWithdrawAddress
   */
  delegatorWithdrawAddress(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * CommunityPool queries the community pool coins.
   *
   * @generated from protobuf rpc: CommunityPool
   */
  communityPool(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
