// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/lightclients/solomachine/v3/solomachine_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../../google/protobuf/any_pb.js";
class ClientState$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v3.ClientState", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "is_frozen",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 3, name: "consensus_state", kind: "message", T: () => ConsensusState }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    message.isFrozen = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* bool is_frozen */
        2:
          message.isFrozen = reader.bool();
          break;
        case /* ibc.lightclients.solomachine.v3.ConsensusState consensus_state */
        3:
          message.consensusState = ConsensusState.internalBinaryRead(reader, reader.uint32(), options, message.consensusState);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    if (message.isFrozen !== false)
      writer.tag(2, WireType.Varint).bool(message.isFrozen);
    if (message.consensusState)
      ConsensusState.internalBinaryWrite(message.consensusState, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ClientState = new ClientState$Type();
class ConsensusState$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v3.ConsensusState", [
      { no: 1, name: "public_key", kind: "message", T: () => Any },
      {
        no: 2,
        name: "diversifier",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.diversifier = "";
    message.timestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any public_key */
        1:
          message.publicKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.publicKey);
          break;
        case /* string diversifier */
        2:
          message.diversifier = reader.string();
          break;
        case /* uint64 timestamp */
        3:
          message.timestamp = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.publicKey)
      Any.internalBinaryWrite(message.publicKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.diversifier !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.diversifier);
    if (message.timestamp !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.timestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ConsensusState = new ConsensusState$Type();
class Header$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v3.Header", [
      {
        no: 1,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "new_public_key", kind: "message", T: () => Any },
      {
        no: 4,
        name: "new_diversifier",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.timestamp = 0n;
    message.signature = new Uint8Array(0);
    message.newDiversifier = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 timestamp */
        1:
          message.timestamp = reader.uint64().toBigInt();
          break;
        case /* bytes signature */
        2:
          message.signature = reader.bytes();
          break;
        case /* google.protobuf.Any new_public_key */
        3:
          message.newPublicKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.newPublicKey);
          break;
        case /* string new_diversifier */
        4:
          message.newDiversifier = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.timestamp !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.timestamp);
    if (message.signature.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.signature);
    if (message.newPublicKey)
      Any.internalBinaryWrite(message.newPublicKey, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.newDiversifier !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.newDiversifier);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Header = new Header$Type();
class Misbehaviour$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v3.Misbehaviour", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "signature_one", kind: "message", T: () => SignatureAndData },
      { no: 3, name: "signature_two", kind: "message", T: () => SignatureAndData }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* ibc.lightclients.solomachine.v3.SignatureAndData signature_one */
        2:
          message.signatureOne = SignatureAndData.internalBinaryRead(reader, reader.uint32(), options, message.signatureOne);
          break;
        case /* ibc.lightclients.solomachine.v3.SignatureAndData signature_two */
        3:
          message.signatureTwo = SignatureAndData.internalBinaryRead(reader, reader.uint32(), options, message.signatureTwo);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    if (message.signatureOne)
      SignatureAndData.internalBinaryWrite(message.signatureOne, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.signatureTwo)
      SignatureAndData.internalBinaryWrite(message.signatureTwo, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Misbehaviour = new Misbehaviour$Type();
class SignatureAndData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v3.SignatureAndData", [
      {
        no: 1,
        name: "signature",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signature = new Uint8Array(0);
    message.path = new Uint8Array(0);
    message.data = new Uint8Array(0);
    message.timestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes signature */
        1:
          message.signature = reader.bytes();
          break;
        case /* bytes path */
        2:
          message.path = reader.bytes();
          break;
        case /* bytes data */
        3:
          message.data = reader.bytes();
          break;
        case /* uint64 timestamp */
        4:
          message.timestamp = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signature.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.signature);
    if (message.path.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.path);
    if (message.data.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.data);
    if (message.timestamp !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.timestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignatureAndData = new SignatureAndData$Type();
class TimestampedSignatureData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v3.TimestampedSignatureData", [
      {
        no: 1,
        name: "signature_data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signatureData = new Uint8Array(0);
    message.timestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes signature_data */
        1:
          message.signatureData = reader.bytes();
          break;
        case /* uint64 timestamp */
        2:
          message.timestamp = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signatureData.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.signatureData);
    if (message.timestamp !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.timestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TimestampedSignatureData = new TimestampedSignatureData$Type();
class SignBytes$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v3.SignBytes", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "diversifier",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "path",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    message.timestamp = 0n;
    message.diversifier = "";
    message.path = new Uint8Array(0);
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* uint64 timestamp */
        2:
          message.timestamp = reader.uint64().toBigInt();
          break;
        case /* string diversifier */
        3:
          message.diversifier = reader.string();
          break;
        case /* bytes path */
        4:
          message.path = reader.bytes();
          break;
        case /* bytes data */
        5:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    if (message.timestamp !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.timestamp);
    if (message.diversifier !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.diversifier);
    if (message.path.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.path);
    if (message.data.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SignBytes = new SignBytes$Type();
class HeaderData$Type extends MessageType {
  constructor() {
    super("ibc.lightclients.solomachine.v3.HeaderData", [
      { no: 1, name: "new_pub_key", kind: "message", T: () => Any },
      {
        no: 2,
        name: "new_diversifier",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.newDiversifier = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any new_pub_key */
        1:
          message.newPubKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.newPubKey);
          break;
        case /* string new_diversifier */
        2:
          message.newDiversifier = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.newPubKey)
      Any.internalBinaryWrite(message.newPubKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.newDiversifier !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.newDiversifier);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const HeaderData = new HeaderData$Type();
export {
  ClientState,
  ConsensusState,
  Header,
  HeaderData,
  Misbehaviour,
  SignBytes,
  SignatureAndData,
  TimestampedSignatureData
};
