// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/auth/v1beta1/auth_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../google/protobuf/any_pb.js";
class BaseAccount$Type extends MessageType {
  constructor() {
    super("cosmos.auth.v1beta1.BaseAccount", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "pub_key", kind: "message", T: () => Any, options: { "gogoproto.jsontag": "public_key,omitempty", "amino.field_name": "public_key" } },
      {
        no: 3,
        name: "account_number",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "cosmos_proto.implements_interface": ["cosmos.auth.v1beta1.AccountI"], "amino.name": "cosmos-sdk/BaseAccount" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.accountNumber = 0n;
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* google.protobuf.Any pub_key */
        2:
          message.pubKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.pubKey);
          break;
        case /* uint64 account_number */
        3:
          message.accountNumber = reader.uint64().toBigInt();
          break;
        case /* uint64 sequence */
        4:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.pubKey)
      Any.internalBinaryWrite(message.pubKey, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.accountNumber !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.accountNumber);
    if (message.sequence !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BaseAccount = new BaseAccount$Type();
class ModuleAccount$Type extends MessageType {
  constructor() {
    super("cosmos.auth.v1beta1.ModuleAccount", [
      { no: 1, name: "base_account", kind: "message", T: () => BaseAccount, options: { "gogoproto.embed": true } },
      {
        no: 2,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "permissions",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos_proto.implements_interface": ["cosmos.auth.v1beta1.ModuleAccountI"], "amino.name": "cosmos-sdk/ModuleAccount", "amino.message_encoding": "module_account" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.name = "";
    message.permissions = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.auth.v1beta1.BaseAccount base_account */
        1:
          message.baseAccount = BaseAccount.internalBinaryRead(reader, reader.uint32(), options, message.baseAccount);
          break;
        case /* string name */
        2:
          message.name = reader.string();
          break;
        case /* repeated string permissions */
        3:
          message.permissions.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.baseAccount)
      BaseAccount.internalBinaryWrite(message.baseAccount, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.name !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.name);
    for (let i = 0; i < message.permissions.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.permissions[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ModuleAccount = new ModuleAccount$Type();
class ModuleCredential$Type extends MessageType {
  constructor() {
    super("cosmos.auth.v1beta1.ModuleCredential", [
      {
        no: 1,
        name: "module_name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "derivation_keys",
        kind: "scalar",
        repeat: 2,
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "amino.name": "cosmos-sdk/GroupAccountCredential" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.moduleName = "";
    message.derivationKeys = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string module_name */
        1:
          message.moduleName = reader.string();
          break;
        case /* repeated bytes derivation_keys */
        2:
          message.derivationKeys.push(reader.bytes());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.moduleName !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.moduleName);
    for (let i = 0; i < message.derivationKeys.length; i++)
      writer.tag(2, WireType.LengthDelimited).bytes(message.derivationKeys[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ModuleCredential = new ModuleCredential$Type();
class Params$Type extends MessageType {
  constructor() {
    super("cosmos.auth.v1beta1.Params", [
      {
        no: 1,
        name: "max_memo_characters",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "tx_sig_limit",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "tx_size_cost_per_byte",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "sig_verify_cost_ed25519", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "SigVerifyCostED25519" } },
      { no: 5, name: "sig_verify_cost_secp256k1", kind: "scalar", jsonName: "sigVerifyCostSecp256k1", T: 4, L: 0, options: { "gogoproto.customname": "SigVerifyCostSecp256k1" } }
    ], { "gogoproto.equal": true, "amino.name": "cosmos-sdk/x/auth/Params" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.maxMemoCharacters = 0n;
    message.txSigLimit = 0n;
    message.txSizeCostPerByte = 0n;
    message.sigVerifyCostEd25519 = 0n;
    message.sigVerifyCostSecp256K1 = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 max_memo_characters */
        1:
          message.maxMemoCharacters = reader.uint64().toBigInt();
          break;
        case /* uint64 tx_sig_limit */
        2:
          message.txSigLimit = reader.uint64().toBigInt();
          break;
        case /* uint64 tx_size_cost_per_byte */
        3:
          message.txSizeCostPerByte = reader.uint64().toBigInt();
          break;
        case /* uint64 sig_verify_cost_ed25519 */
        4:
          message.sigVerifyCostEd25519 = reader.uint64().toBigInt();
          break;
        case /* uint64 sig_verify_cost_secp256k1 */
        5:
          message.sigVerifyCostSecp256K1 = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.maxMemoCharacters !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.maxMemoCharacters);
    if (message.txSigLimit !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.txSigLimit);
    if (message.txSizeCostPerByte !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.txSizeCostPerByte);
    if (message.sigVerifyCostEd25519 !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.sigVerifyCostEd25519);
    if (message.sigVerifyCostSecp256K1 !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.sigVerifyCostSecp256K1);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
export {
  BaseAccount,
  ModuleAccount,
  ModuleCredential,
  Params
};
