// @injectivelabs/core-proto-ts-v2/esm/generated/cosmwasm/wasm/v1/authz_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Coin } from "../../../cosmos/base/v1beta1/coin_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
import { AccessConfig } from "./types_pb.js";
class StoreCodeAuthorization$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.StoreCodeAuthorization", [
      { no: 1, name: "grants", kind: "message", repeat: 2, T: () => CodeGrant, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "cosmos_proto.implements_interface": ["cosmos.authz.v1beta1.Authorization"], "amino.name": "wasm/StoreCodeAuthorization" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.grants = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmwasm.wasm.v1.CodeGrant grants */
        1:
          message.grants.push(CodeGrant.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.grants.length; i++)
      CodeGrant.internalBinaryWrite(message.grants[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StoreCodeAuthorization = new StoreCodeAuthorization$Type();
class ContractExecutionAuthorization$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.ContractExecutionAuthorization", [
      { no: 1, name: "grants", kind: "message", repeat: 2, T: () => ContractGrant, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "cosmos_proto.implements_interface": ["cosmos.authz.v1beta1.Authorization"], "amino.name": "wasm/ContractExecutionAuthorization" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.grants = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmwasm.wasm.v1.ContractGrant grants */
        1:
          message.grants.push(ContractGrant.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.grants.length; i++)
      ContractGrant.internalBinaryWrite(message.grants[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ContractExecutionAuthorization = new ContractExecutionAuthorization$Type();
class ContractMigrationAuthorization$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.ContractMigrationAuthorization", [
      { no: 1, name: "grants", kind: "message", repeat: 2, T: () => ContractGrant, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "cosmos_proto.implements_interface": ["cosmos.authz.v1beta1.Authorization"], "amino.name": "wasm/ContractMigrationAuthorization" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.grants = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmwasm.wasm.v1.ContractGrant grants */
        1:
          message.grants.push(ContractGrant.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.grants.length; i++)
      ContractGrant.internalBinaryWrite(message.grants[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ContractMigrationAuthorization = new ContractMigrationAuthorization$Type();
class CodeGrant$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.CodeGrant", [
      {
        no: 1,
        name: "code_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "instantiate_permission", kind: "message", T: () => AccessConfig }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codeHash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes code_hash */
        1:
          message.codeHash = reader.bytes();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig instantiate_permission */
        2:
          message.instantiatePermission = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.instantiatePermission);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeHash.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.codeHash);
    if (message.instantiatePermission)
      AccessConfig.internalBinaryWrite(message.instantiatePermission, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CodeGrant = new CodeGrant$Type();
class ContractGrant$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.ContractGrant", [
      { no: 1, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "limit", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmwasm.wasm.v1.ContractAuthzLimitX" } },
      { no: 3, name: "filter", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmwasm.wasm.v1.ContractAuthzFilterX" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contract = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract */
        1:
          message.contract = reader.string();
          break;
        case /* google.protobuf.Any limit */
        2:
          message.limit = Any.internalBinaryRead(reader, reader.uint32(), options, message.limit);
          break;
        case /* google.protobuf.Any filter */
        3:
          message.filter = Any.internalBinaryRead(reader, reader.uint32(), options, message.filter);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contract !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contract);
    if (message.limit)
      Any.internalBinaryWrite(message.limit, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.filter)
      Any.internalBinaryWrite(message.filter, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ContractGrant = new ContractGrant$Type();
class MaxCallsLimit$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MaxCallsLimit", [
      {
        no: 1,
        name: "remaining",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "cosmos_proto.implements_interface": ["cosmwasm.wasm.v1.ContractAuthzLimitX"], "amino.name": "wasm/MaxCallsLimit" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.remaining = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 remaining */
        1:
          message.remaining = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.remaining !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.remaining);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MaxCallsLimit = new MaxCallsLimit$Type();
class MaxFundsLimit$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MaxFundsLimit", [
      { no: 1, name: "amounts", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } }
    ], { "cosmos_proto.implements_interface": ["cosmwasm.wasm.v1.ContractAuthzLimitX"], "amino.name": "wasm/MaxFundsLimit" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.amounts = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin amounts */
        1:
          message.amounts.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.amounts.length; i++)
      Coin.internalBinaryWrite(message.amounts[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MaxFundsLimit = new MaxFundsLimit$Type();
class CombinedLimit$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.CombinedLimit", [
      {
        no: 1,
        name: "calls_remaining",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "amounts", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } }
    ], { "cosmos_proto.implements_interface": ["cosmwasm.wasm.v1.ContractAuthzLimitX"], "amino.name": "wasm/CombinedLimit" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.callsRemaining = 0n;
    message.amounts = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 calls_remaining */
        1:
          message.callsRemaining = reader.uint64().toBigInt();
          break;
        case /* repeated cosmos.base.v1beta1.Coin amounts */
        2:
          message.amounts.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.callsRemaining !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.callsRemaining);
    for (let i = 0; i < message.amounts.length; i++)
      Coin.internalBinaryWrite(message.amounts[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CombinedLimit = new CombinedLimit$Type();
class AllowAllMessagesFilter$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.AllowAllMessagesFilter", [], { "cosmos_proto.implements_interface": ["cosmwasm.wasm.v1.ContractAuthzFilterX"], "amino.name": "wasm/AllowAllMessagesFilter" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AllowAllMessagesFilter = new AllowAllMessagesFilter$Type();
class AcceptedMessageKeysFilter$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.AcceptedMessageKeysFilter", [
      {
        no: 1,
        name: "keys",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "cosmos_proto.implements_interface": ["cosmwasm.wasm.v1.ContractAuthzFilterX"], "amino.name": "wasm/AcceptedMessageKeysFilter" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.keys = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string keys */
        1:
          message.keys.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.keys.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.keys[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AcceptedMessageKeysFilter = new AcceptedMessageKeysFilter$Type();
class AcceptedMessagesFilter$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.AcceptedMessagesFilter", [
      { no: 1, name: "messages", kind: "scalar", repeat: 2, T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } }
    ], { "cosmos_proto.implements_interface": ["cosmwasm.wasm.v1.ContractAuthzFilterX"], "amino.name": "wasm/AcceptedMessagesFilter" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.messages = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated bytes messages */
        1:
          message.messages.push(reader.bytes());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.messages.length; i++)
      writer.tag(1, WireType.LengthDelimited).bytes(message.messages[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AcceptedMessagesFilter = new AcceptedMessagesFilter$Type();
export {
  AcceptedMessageKeysFilter,
  AcceptedMessagesFilter,
  AllowAllMessagesFilter,
  CodeGrant,
  CombinedLimit,
  ContractExecutionAuthorization,
  ContractGrant,
  ContractMigrationAuthorization,
  MaxCallsLimit,
  MaxFundsLimit,
  StoreCodeAuthorization
};
