// @injectivelabs/core-proto-ts-v2/esm/generated/cosmwasm/wasm/v1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ContractCodeHistoryEntry } from "./types_pb.js";
import { Model } from "./types_pb.js";
import { ContractInfo } from "./types_pb.js";
import { CodeInfo } from "./types_pb.js";
import { Params } from "./types_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.GenesisState", [
      { no: 1, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 2, name: "codes", kind: "message", repeat: 2, T: () => Code, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "codes,omitempty", "amino.dont_omitempty": true } },
      { no: 3, name: "contracts", kind: "message", repeat: 2, T: () => Contract, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "contracts,omitempty", "amino.dont_omitempty": true } },
      { no: 4, name: "sequences", kind: "message", repeat: 2, T: () => Sequence, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "sequences,omitempty", "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codes = [];
    message.contracts = [];
    message.sequences = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmwasm.wasm.v1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        case /* repeated cosmwasm.wasm.v1.Code codes */
        2:
          message.codes.push(Code.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmwasm.wasm.v1.Contract contracts */
        3:
          message.contracts.push(Contract.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmwasm.wasm.v1.Sequence sequences */
        4:
          message.sequences.push(Sequence.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.codes.length; i++)
      Code.internalBinaryWrite(message.codes[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.contracts.length; i++)
      Contract.internalBinaryWrite(message.contracts[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.sequences.length; i++)
      Sequence.internalBinaryWrite(message.sequences[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class Code$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.Code", [
      { no: 1, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      { no: 2, name: "code_info", kind: "message", T: () => CodeInfo, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 3,
        name: "code_bytes",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "pinned",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codeId = 0n;
    message.codeBytes = new Uint8Array(0);
    message.pinned = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 code_id */
        1:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* cosmwasm.wasm.v1.CodeInfo code_info */
        2:
          message.codeInfo = CodeInfo.internalBinaryRead(reader, reader.uint32(), options, message.codeInfo);
          break;
        case /* bytes code_bytes */
        3:
          message.codeBytes = reader.bytes();
          break;
        case /* bool pinned */
        4:
          message.pinned = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.codeId);
    if (message.codeInfo)
      CodeInfo.internalBinaryWrite(message.codeInfo, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.codeBytes.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.codeBytes);
    if (message.pinned !== false)
      writer.tag(4, WireType.Varint).bool(message.pinned);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Code = new Code$Type();
class Contract$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.Contract", [
      { no: 1, name: "contract_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "contract_info", kind: "message", T: () => ContractInfo, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 3, name: "contract_state", kind: "message", repeat: 2, T: () => Model, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 4, name: "contract_code_history", kind: "message", repeat: 2, T: () => ContractCodeHistoryEntry, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contractAddress = "";
    message.contractState = [];
    message.contractCodeHistory = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract_address */
        1:
          message.contractAddress = reader.string();
          break;
        case /* cosmwasm.wasm.v1.ContractInfo contract_info */
        2:
          message.contractInfo = ContractInfo.internalBinaryRead(reader, reader.uint32(), options, message.contractInfo);
          break;
        case /* repeated cosmwasm.wasm.v1.Model contract_state */
        3:
          message.contractState.push(Model.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmwasm.wasm.v1.ContractCodeHistoryEntry contract_code_history */
        4:
          message.contractCodeHistory.push(ContractCodeHistoryEntry.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contractAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contractAddress);
    if (message.contractInfo)
      ContractInfo.internalBinaryWrite(message.contractInfo, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.contractState.length; i++)
      Model.internalBinaryWrite(message.contractState[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.contractCodeHistory.length; i++)
      ContractCodeHistoryEntry.internalBinaryWrite(message.contractCodeHistory[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Contract = new Contract$Type();
class Sequence$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.Sequence", [
      { no: 1, name: "id_key", kind: "scalar", T: 12, options: { "gogoproto.customname": "IDKey" } },
      {
        no: 2,
        name: "value",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.idKey = new Uint8Array(0);
    message.value = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes id_key */
        1:
          message.idKey = reader.bytes();
          break;
        case /* uint64 value */
        2:
          message.value = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.idKey.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.idKey);
    if (message.value !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Sequence = new Sequence$Type();
export {
  Code,
  Contract,
  GenesisState,
  Sequence
};
