// @injectivelabs/core-proto-ts-v2/esm/generated/cosmwasm/wasm/v1/ibc_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class MsgIBCSend$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgIBCSend", [
      { no: 2, name: "channel", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"source_channel"' } },
      { no: 4, name: "timeout_height", kind: "scalar", T: 4, L: 0, options: { "gogoproto.moretags": 'yaml:"timeout_height"' } },
      { no: 5, name: "timeout_timestamp", kind: "scalar", T: 4, L: 0, options: { "gogoproto.moretags": 'yaml:"timeout_timestamp"' } },
      {
        no: 6,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channel = "";
    message.timeoutHeight = 0n;
    message.timeoutTimestamp = 0n;
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string channel */
        2:
          message.channel = reader.string();
          break;
        case /* uint64 timeout_height */
        4:
          message.timeoutHeight = reader.uint64().toBigInt();
          break;
        case /* uint64 timeout_timestamp */
        5:
          message.timeoutTimestamp = reader.uint64().toBigInt();
          break;
        case /* bytes data */
        6:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channel !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channel);
    if (message.timeoutHeight !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.timeoutHeight);
    if (message.timeoutTimestamp !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.timeoutTimestamp);
    if (message.data.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgIBCSend = new MsgIBCSend$Type();
class MsgIBCSendResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgIBCSendResponse", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgIBCSendResponse = new MsgIBCSendResponse$Type();
class MsgIBCWriteAcknowledgementResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgIBCWriteAcknowledgementResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgIBCWriteAcknowledgementResponse = new MsgIBCWriteAcknowledgementResponse$Type();
class MsgIBCCloseChannel$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgIBCCloseChannel", [
      { no: 2, name: "channel", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"source_channel"' } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channel = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string channel */
        2:
          message.channel = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channel !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channel);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgIBCCloseChannel = new MsgIBCCloseChannel$Type();
export {
  MsgIBCCloseChannel,
  MsgIBCSend,
  MsgIBCSendResponse,
  MsgIBCWriteAcknowledgementResponse
};
