// @injectivelabs/core-proto-ts-v2/esm/generated/cosmwasm/wasm/v1/proposal_legacy_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Coin } from "../../../cosmos/base/v1beta1/coin_pb.js";
import { AccessConfig } from "./types_pb.js";
class StoreCodeProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.StoreCodeProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "run_as", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 4, name: "wasm_byte_code", kind: "scalar", T: 12, options: { "gogoproto.customname": "WASMByteCode" } },
      { no: 7, name: "instantiate_permission", kind: "message", T: () => AccessConfig },
      {
        no: 8,
        name: "unpin_code",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 9,
        name: "source",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "builder",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "code_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/StoreCodeProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.runAs = "";
    message.wasmByteCode = new Uint8Array(0);
    message.unpinCode = false;
    message.source = "";
    message.builder = "";
    message.codeHash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* string run_as */
        3:
          message.runAs = reader.string();
          break;
        case /* bytes wasm_byte_code */
        4:
          message.wasmByteCode = reader.bytes();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig instantiate_permission */
        7:
          message.instantiatePermission = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.instantiatePermission);
          break;
        case /* bool unpin_code */
        8:
          message.unpinCode = reader.bool();
          break;
        case /* string source */
        9:
          message.source = reader.string();
          break;
        case /* string builder */
        10:
          message.builder = reader.string();
          break;
        case /* bytes code_hash */
        11:
          message.codeHash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.runAs !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.runAs);
    if (message.wasmByteCode.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.wasmByteCode);
    if (message.instantiatePermission)
      AccessConfig.internalBinaryWrite(message.instantiatePermission, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.unpinCode !== false)
      writer.tag(8, WireType.Varint).bool(message.unpinCode);
    if (message.source !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.source);
    if (message.builder !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.builder);
    if (message.codeHash.length)
      writer.tag(11, WireType.LengthDelimited).bytes(message.codeHash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StoreCodeProposal = new StoreCodeProposal$Type();
class InstantiateContractProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.InstantiateContractProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "run_as", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 4, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 5, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      {
        no: 6,
        name: "label",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 7, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } },
      { no: 8, name: "funds", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/InstantiateContractProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.runAs = "";
    message.admin = "";
    message.codeId = 0n;
    message.label = "";
    message.msg = new Uint8Array(0);
    message.funds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* string run_as */
        3:
          message.runAs = reader.string();
          break;
        case /* string admin */
        4:
          message.admin = reader.string();
          break;
        case /* uint64 code_id */
        5:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string label */
        6:
          message.label = reader.string();
          break;
        case /* bytes msg */
        7:
          message.msg = reader.bytes();
          break;
        case /* repeated cosmos.base.v1beta1.Coin funds */
        8:
          message.funds.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.runAs !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.runAs);
    if (message.admin !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.admin);
    if (message.codeId !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.codeId);
    if (message.label !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.label);
    if (message.msg.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.msg);
    for (let i = 0; i < message.funds.length; i++)
      Coin.internalBinaryWrite(message.funds[i], writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const InstantiateContractProposal = new InstantiateContractProposal$Type();
class InstantiateContract2Proposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.InstantiateContract2Proposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "run_as", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 4, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 5, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      {
        no: 6,
        name: "label",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 7, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } },
      { no: 8, name: "funds", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } },
      {
        no: 9,
        name: "salt",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 10,
        name: "fix_msg",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/InstantiateContract2Proposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.runAs = "";
    message.admin = "";
    message.codeId = 0n;
    message.label = "";
    message.msg = new Uint8Array(0);
    message.funds = [];
    message.salt = new Uint8Array(0);
    message.fixMsg = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* string run_as */
        3:
          message.runAs = reader.string();
          break;
        case /* string admin */
        4:
          message.admin = reader.string();
          break;
        case /* uint64 code_id */
        5:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string label */
        6:
          message.label = reader.string();
          break;
        case /* bytes msg */
        7:
          message.msg = reader.bytes();
          break;
        case /* repeated cosmos.base.v1beta1.Coin funds */
        8:
          message.funds.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* bytes salt */
        9:
          message.salt = reader.bytes();
          break;
        case /* bool fix_msg */
        10:
          message.fixMsg = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.runAs !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.runAs);
    if (message.admin !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.admin);
    if (message.codeId !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.codeId);
    if (message.label !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.label);
    if (message.msg.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.msg);
    for (let i = 0; i < message.funds.length; i++)
      Coin.internalBinaryWrite(message.funds[i], writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    if (message.salt.length)
      writer.tag(9, WireType.LengthDelimited).bytes(message.salt);
    if (message.fixMsg !== false)
      writer.tag(10, WireType.Varint).bool(message.fixMsg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const InstantiateContract2Proposal = new InstantiateContract2Proposal$Type();
class MigrateContractProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MigrateContractProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 5, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      { no: 6, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/MigrateContractProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.contract = "";
    message.codeId = 0n;
    message.msg = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* string contract */
        4:
          message.contract = reader.string();
          break;
        case /* uint64 code_id */
        5:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* bytes msg */
        6:
          message.msg = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.contract !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.contract);
    if (message.codeId !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.codeId);
    if (message.msg.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.msg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MigrateContractProposal = new MigrateContractProposal$Type();
class SudoContractProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.SudoContractProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 4, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/SudoContractProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.contract = "";
    message.msg = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* string contract */
        3:
          message.contract = reader.string();
          break;
        case /* bytes msg */
        4:
          message.msg = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.contract !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.contract);
    if (message.msg.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.msg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SudoContractProposal = new SudoContractProposal$Type();
class ExecuteContractProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.ExecuteContractProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "run_as", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 4, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 5, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } },
      { no: 6, name: "funds", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/ExecuteContractProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.runAs = "";
    message.contract = "";
    message.msg = new Uint8Array(0);
    message.funds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* string run_as */
        3:
          message.runAs = reader.string();
          break;
        case /* string contract */
        4:
          message.contract = reader.string();
          break;
        case /* bytes msg */
        5:
          message.msg = reader.bytes();
          break;
        case /* repeated cosmos.base.v1beta1.Coin funds */
        6:
          message.funds.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.runAs !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.runAs);
    if (message.contract !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.contract);
    if (message.msg.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.msg);
    for (let i = 0; i < message.funds.length; i++)
      Coin.internalBinaryWrite(message.funds[i], writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ExecuteContractProposal = new ExecuteContractProposal$Type();
class UpdateAdminProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.UpdateAdminProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "new_admin", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"new_admin"', "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 4, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/UpdateAdminProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.newAdmin = "";
    message.contract = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* string new_admin */
        3:
          message.newAdmin = reader.string();
          break;
        case /* string contract */
        4:
          message.contract = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.newAdmin !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.newAdmin);
    if (message.contract !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.contract);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const UpdateAdminProposal = new UpdateAdminProposal$Type();
class ClearAdminProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.ClearAdminProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/ClearAdminProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.contract = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* string contract */
        3:
          message.contract = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.contract !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.contract);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ClearAdminProposal = new ClearAdminProposal$Type();
class PinCodesProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.PinCodesProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "code_ids", kind: "scalar", repeat: 1, T: 4, L: 0, options: { "gogoproto.customname": "CodeIDs", "gogoproto.moretags": 'yaml:"code_ids"' } }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/PinCodesProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.codeIds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* repeated uint64 code_ids */
        3:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.codeIds.push(reader.uint64().toBigInt());
          else
            message.codeIds.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.codeIds.length) {
      writer.tag(3, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.codeIds.length; i++)
        writer.uint64(message.codeIds[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PinCodesProposal = new PinCodesProposal$Type();
class UnpinCodesProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.UnpinCodesProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "code_ids", kind: "scalar", repeat: 1, T: 4, L: 0, options: { "gogoproto.customname": "CodeIDs", "gogoproto.moretags": 'yaml:"code_ids"' } }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/UnpinCodesProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.codeIds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* repeated uint64 code_ids */
        3:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.codeIds.push(reader.uint64().toBigInt());
          else
            message.codeIds.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.codeIds.length) {
      writer.tag(3, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.codeIds.length; i++)
        writer.uint64(message.codeIds[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const UnpinCodesProposal = new UnpinCodesProposal$Type();
class AccessConfigUpdate$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.AccessConfigUpdate", [
      { no: 1, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      { no: 2, name: "instantiate_permission", kind: "message", T: () => AccessConfig, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codeId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 code_id */
        1:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig instantiate_permission */
        2:
          message.instantiatePermission = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.instantiatePermission);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.codeId);
    if (message.instantiatePermission)
      AccessConfig.internalBinaryWrite(message.instantiatePermission, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AccessConfigUpdate = new AccessConfigUpdate$Type();
class UpdateInstantiateConfigProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.UpdateInstantiateConfigProposal", [
      { no: 1, name: "title", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"title"' } },
      { no: 2, name: "description", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"description"' } },
      { no: 3, name: "access_config_updates", kind: "message", repeat: 2, T: () => AccessConfigUpdate, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/UpdateInstantiateConfigProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.accessConfigUpdates = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* repeated cosmwasm.wasm.v1.AccessConfigUpdate access_config_updates */
        3:
          message.accessConfigUpdates.push(AccessConfigUpdate.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    for (let i = 0; i < message.accessConfigUpdates.length; i++)
      AccessConfigUpdate.internalBinaryWrite(message.accessConfigUpdates[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const UpdateInstantiateConfigProposal = new UpdateInstantiateConfigProposal$Type();
class StoreAndInstantiateContractProposal$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.StoreAndInstantiateContractProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "run_as", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 4, name: "wasm_byte_code", kind: "scalar", T: 12, options: { "gogoproto.customname": "WASMByteCode" } },
      { no: 5, name: "instantiate_permission", kind: "message", T: () => AccessConfig },
      {
        no: 6,
        name: "unpin_code",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 7,
        name: "admin",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "label",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 9, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } },
      { no: 10, name: "funds", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } },
      {
        no: 11,
        name: "source",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 12,
        name: "builder",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 13,
        name: "code_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasm/StoreAndInstantiateContractProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.runAs = "";
    message.wasmByteCode = new Uint8Array(0);
    message.unpinCode = false;
    message.admin = "";
    message.label = "";
    message.msg = new Uint8Array(0);
    message.funds = [];
    message.source = "";
    message.builder = "";
    message.codeHash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* string run_as */
        3:
          message.runAs = reader.string();
          break;
        case /* bytes wasm_byte_code */
        4:
          message.wasmByteCode = reader.bytes();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig instantiate_permission */
        5:
          message.instantiatePermission = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.instantiatePermission);
          break;
        case /* bool unpin_code */
        6:
          message.unpinCode = reader.bool();
          break;
        case /* string admin */
        7:
          message.admin = reader.string();
          break;
        case /* string label */
        8:
          message.label = reader.string();
          break;
        case /* bytes msg */
        9:
          message.msg = reader.bytes();
          break;
        case /* repeated cosmos.base.v1beta1.Coin funds */
        10:
          message.funds.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string source */
        11:
          message.source = reader.string();
          break;
        case /* string builder */
        12:
          message.builder = reader.string();
          break;
        case /* bytes code_hash */
        13:
          message.codeHash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.runAs !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.runAs);
    if (message.wasmByteCode.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.wasmByteCode);
    if (message.instantiatePermission)
      AccessConfig.internalBinaryWrite(message.instantiatePermission, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.unpinCode !== false)
      writer.tag(6, WireType.Varint).bool(message.unpinCode);
    if (message.admin !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.admin);
    if (message.label !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.label);
    if (message.msg.length)
      writer.tag(9, WireType.LengthDelimited).bytes(message.msg);
    for (let i = 0; i < message.funds.length; i++)
      Coin.internalBinaryWrite(message.funds[i], writer.tag(10, WireType.LengthDelimited).fork(), options).join();
    if (message.source !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.source);
    if (message.builder !== "")
      writer.tag(12, WireType.LengthDelimited).string(message.builder);
    if (message.codeHash.length)
      writer.tag(13, WireType.LengthDelimited).bytes(message.codeHash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const StoreAndInstantiateContractProposal = new StoreAndInstantiateContractProposal$Type();
export {
  AccessConfigUpdate,
  ClearAdminProposal,
  ExecuteContractProposal,
  InstantiateContract2Proposal,
  InstantiateContractProposal,
  MigrateContractProposal,
  PinCodesProposal,
  StoreAndInstantiateContractProposal,
  StoreCodeProposal,
  SudoContractProposal,
  UnpinCodesProposal,
  UpdateAdminProposal,
  UpdateInstantiateConfigProposal
};
