// @injectivelabs/core-proto-ts-v2/esm/generated/cosmwasm/wasm/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * ContractInfo gets the contract meta data
   *
   * @generated from protobuf rpc: ContractInfo
   */
  contractInfo(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ContractHistory gets the contract code history
   *
   * @generated from protobuf rpc: ContractHistory
   */
  contractHistory(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ContractsByCode lists all smart contracts for a code id
   *
   * @generated from protobuf rpc: ContractsByCode
   */
  contractsByCode(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * AllContractState gets all raw store data for a single contract
   *
   * @generated from protobuf rpc: AllContractState
   */
  allContractState(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * RawContractState gets single key from the raw store data of a contract
   *
   * @generated from protobuf rpc: RawContractState
   */
  rawContractState(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * SmartContractState get smart query result from the contract
   *
   * @generated from protobuf rpc: SmartContractState
   */
  smartContractState(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Code gets the binary code and metadata for a singe wasm code
   *
   * @generated from protobuf rpc: Code
   */
  code(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Codes gets the metadata for all stored wasm codes
   *
   * @generated from protobuf rpc: Codes
   */
  codes(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * PinnedCodes gets the pinned code ids
   *
   * @generated from protobuf rpc: PinnedCodes
   */
  pinnedCodes(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Params gets the module params
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ContractsByCreator gets the contracts by creator
   *
   * @generated from protobuf rpc: ContractsByCreator
   */
  contractsByCreator(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * BuildAddress builds a contract address
   *
   * @generated from protobuf rpc: BuildAddress
   */
  buildAddress(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
