// @injectivelabs/core-proto-ts-v2/esm/generated/cosmwasm/wasm/v1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * StoreCode to submit Wasm code to the system
   *
   * @generated from protobuf rpc: StoreCode
   */
  storeCode(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   *  InstantiateContract creates a new smart contract instance for the given
   *  code id.
   *
   * @generated from protobuf rpc: InstantiateContract
   */
  instantiateContract(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   *  InstantiateContract2 creates a new smart contract instance for the given
   *  code id with a predictable address
   *
   * @generated from protobuf rpc: InstantiateContract2
   */
  instantiateContract2(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Execute submits the given message data to a smart contract
   *
   * @generated from protobuf rpc: ExecuteContract
   */
  executeContract(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Migrate runs a code upgrade/ downgrade for a smart contract
   *
   * @generated from protobuf rpc: MigrateContract
   */
  migrateContract(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateAdmin sets a new admin for a smart contract
   *
   * @generated from protobuf rpc: UpdateAdmin
   */
  updateAdmin(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ClearAdmin removes any admin stored for a smart contract
   *
   * @generated from protobuf rpc: ClearAdmin
   */
  clearAdmin(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateInstantiateConfig updates instantiate config for a smart contract
   *
   * @generated from protobuf rpc: UpdateInstantiateConfig
   */
  updateInstantiateConfig(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateParams defines a governance operation for updating the x/wasm
   * module parameters. The authority is defined in the keeper.
   *
   * Since: 0.40
   *
   * @generated from protobuf rpc: UpdateParams
   */
  updateParams(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * SudoContract defines a governance operation for calling sudo
   * on a contract. The authority is defined in the keeper.
   *
   * Since: 0.40
   *
   * @generated from protobuf rpc: SudoContract
   */
  sudoContract(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * PinCodes defines a governance operation for pinning a set of
   * code ids in the wasmvm cache. The authority is defined in the keeper.
   *
   * Since: 0.40
   *
   * @generated from protobuf rpc: PinCodes
   */
  pinCodes(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UnpinCodes defines a governance operation for unpinning a set of
   * code ids in the wasmvm cache. The authority is defined in the keeper.
   *
   * Since: 0.40
   *
   * @generated from protobuf rpc: UnpinCodes
   */
  unpinCodes(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StoreAndInstantiateContract defines a governance operation for storing
   * and instantiating the contract. The authority is defined in the keeper.
   *
   * Since: 0.40
   *
   * @generated from protobuf rpc: StoreAndInstantiateContract
   */
  storeAndInstantiateContract(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * RemoveCodeUploadParamsAddresses defines a governance operation for
   * removing addresses from code upload params.
   * The authority is defined in the keeper.
   *
   * @generated from protobuf rpc: RemoveCodeUploadParamsAddresses
   */
  removeCodeUploadParamsAddresses(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * AddCodeUploadParamsAddresses defines a governance operation for
   * adding addresses to code upload params.
   * The authority is defined in the keeper.
   *
   * @generated from protobuf rpc: AddCodeUploadParamsAddresses
   */
  addCodeUploadParamsAddresses(input, options) {
    const method = this.methods[14], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * StoreAndMigrateContract defines a governance operation for storing
   * and migrating the contract. The authority is defined in the keeper.
   *
   * Since: 0.42
   *
   * @generated from protobuf rpc: StoreAndMigrateContract
   */
  storeAndMigrateContract(input, options) {
    const method = this.methods[15], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateContractLabel sets a new label for a smart contract
   *
   * Since: 0.43
   *
   * @generated from protobuf rpc: UpdateContractLabel
   */
  updateContractLabel(input, options) {
    const method = this.methods[16], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
