// @injectivelabs/core-proto-ts-v2/esm/generated/cosmwasm/wasm/v1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./types_pb.js";
import { Coin } from "../../../cosmos/base/v1beta1/coin_pb.js";
import { AccessConfig } from "./types_pb.js";
class MsgStoreCode$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgStoreCode", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "wasm_byte_code", kind: "scalar", T: 12, options: { "gogoproto.customname": "WASMByteCode" } },
      { no: 5, name: "instantiate_permission", kind: "message", T: () => AccessConfig }
    ], { "amino.name": "wasm/MsgStoreCode", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.wasmByteCode = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* bytes wasm_byte_code */
        2:
          message.wasmByteCode = reader.bytes();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig instantiate_permission */
        5:
          message.instantiatePermission = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.instantiatePermission);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.wasmByteCode.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.wasmByteCode);
    if (message.instantiatePermission)
      AccessConfig.internalBinaryWrite(message.instantiatePermission, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgStoreCode = new MsgStoreCode$Type();
class MsgStoreCodeResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgStoreCodeResponse", [
      { no: 1, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      {
        no: 2,
        name: "checksum",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codeId = 0n;
    message.checksum = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 code_id */
        1:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* bytes checksum */
        2:
          message.checksum = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.codeId);
    if (message.checksum.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.checksum);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgStoreCodeResponse = new MsgStoreCodeResponse$Type();
class MsgInstantiateContract$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgInstantiateContract", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      {
        no: 4,
        name: "label",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 5, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } },
      { no: 6, name: "funds", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } }
    ], { "amino.name": "wasm/MsgInstantiateContract", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.admin = "";
    message.codeId = 0n;
    message.label = "";
    message.msg = new Uint8Array(0);
    message.funds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string admin */
        2:
          message.admin = reader.string();
          break;
        case /* uint64 code_id */
        3:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string label */
        4:
          message.label = reader.string();
          break;
        case /* bytes msg */
        5:
          message.msg = reader.bytes();
          break;
        case /* repeated cosmos.base.v1beta1.Coin funds */
        6:
          message.funds.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.admin !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.admin);
    if (message.codeId !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.codeId);
    if (message.label !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.label);
    if (message.msg.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.msg);
    for (let i = 0; i < message.funds.length; i++)
      Coin.internalBinaryWrite(message.funds[i], writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgInstantiateContract = new MsgInstantiateContract$Type();
class MsgInstantiateContractResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgInstantiateContractResponse", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* bytes data */
        2:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.data.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgInstantiateContractResponse = new MsgInstantiateContractResponse$Type();
class MsgInstantiateContract2$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgInstantiateContract2", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      {
        no: 4,
        name: "label",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 5, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } },
      { no: 6, name: "funds", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } },
      {
        no: 7,
        name: "salt",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 8,
        name: "fix_msg",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ], { "amino.name": "wasm/MsgInstantiateContract2", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.admin = "";
    message.codeId = 0n;
    message.label = "";
    message.msg = new Uint8Array(0);
    message.funds = [];
    message.salt = new Uint8Array(0);
    message.fixMsg = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string admin */
        2:
          message.admin = reader.string();
          break;
        case /* uint64 code_id */
        3:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string label */
        4:
          message.label = reader.string();
          break;
        case /* bytes msg */
        5:
          message.msg = reader.bytes();
          break;
        case /* repeated cosmos.base.v1beta1.Coin funds */
        6:
          message.funds.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* bytes salt */
        7:
          message.salt = reader.bytes();
          break;
        case /* bool fix_msg */
        8:
          message.fixMsg = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.admin !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.admin);
    if (message.codeId !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.codeId);
    if (message.label !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.label);
    if (message.msg.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.msg);
    for (let i = 0; i < message.funds.length; i++)
      Coin.internalBinaryWrite(message.funds[i], writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.salt.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.salt);
    if (message.fixMsg !== false)
      writer.tag(8, WireType.Varint).bool(message.fixMsg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgInstantiateContract2 = new MsgInstantiateContract2$Type();
class MsgInstantiateContract2Response$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgInstantiateContract2Response", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* bytes data */
        2:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.data.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgInstantiateContract2Response = new MsgInstantiateContract2Response$Type();
class MsgExecuteContract$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgExecuteContract", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } },
      { no: 5, name: "funds", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } }
    ], { "amino.name": "wasm/MsgExecuteContract", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.contract = "";
    message.msg = new Uint8Array(0);
    message.funds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string contract */
        2:
          message.contract = reader.string();
          break;
        case /* bytes msg */
        3:
          message.msg = reader.bytes();
          break;
        case /* repeated cosmos.base.v1beta1.Coin funds */
        5:
          message.funds.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.contract !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.contract);
    if (message.msg.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.msg);
    for (let i = 0; i < message.funds.length; i++)
      Coin.internalBinaryWrite(message.funds[i], writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgExecuteContract = new MsgExecuteContract$Type();
class MsgExecuteContractResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgExecuteContractResponse", [
      {
        no: 1,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes data */
        1:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.data.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgExecuteContractResponse = new MsgExecuteContractResponse$Type();
class MsgMigrateContract$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgMigrateContract", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      { no: 4, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } }
    ], { "amino.name": "wasm/MsgMigrateContract", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.contract = "";
    message.codeId = 0n;
    message.msg = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string contract */
        2:
          message.contract = reader.string();
          break;
        case /* uint64 code_id */
        3:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* bytes msg */
        4:
          message.msg = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.contract !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.contract);
    if (message.codeId !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.codeId);
    if (message.msg.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.msg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgMigrateContract = new MsgMigrateContract$Type();
class MsgMigrateContractResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgMigrateContractResponse", [
      {
        no: 1,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes data */
        1:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.data.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgMigrateContractResponse = new MsgMigrateContractResponse$Type();
class MsgUpdateAdmin$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUpdateAdmin", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "new_admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "amino.name": "wasm/MsgUpdateAdmin", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.newAdmin = "";
    message.contract = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string new_admin */
        2:
          message.newAdmin = reader.string();
          break;
        case /* string contract */
        3:
          message.contract = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.newAdmin !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.newAdmin);
    if (message.contract !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.contract);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateAdmin = new MsgUpdateAdmin$Type();
class MsgUpdateAdminResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUpdateAdminResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateAdminResponse = new MsgUpdateAdminResponse$Type();
class MsgClearAdmin$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgClearAdmin", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "amino.name": "wasm/MsgClearAdmin", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.contract = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string contract */
        3:
          message.contract = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.contract !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.contract);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgClearAdmin = new MsgClearAdmin$Type();
class MsgClearAdminResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgClearAdminResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgClearAdminResponse = new MsgClearAdminResponse$Type();
class MsgUpdateInstantiateConfig$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUpdateInstantiateConfig", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      { no: 3, name: "new_instantiate_permission", kind: "message", T: () => AccessConfig }
    ], { "amino.name": "wasm/MsgUpdateInstantiateConfig", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.codeId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* uint64 code_id */
        2:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig new_instantiate_permission */
        3:
          message.newInstantiatePermission = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.newInstantiatePermission);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.codeId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.codeId);
    if (message.newInstantiatePermission)
      AccessConfig.internalBinaryWrite(message.newInstantiatePermission, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateInstantiateConfig = new MsgUpdateInstantiateConfig$Type();
class MsgUpdateInstantiateConfigResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUpdateInstantiateConfigResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateInstantiateConfigResponse = new MsgUpdateInstantiateConfigResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUpdateParams", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "amino.name": "wasm/MsgUpdateParams", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* cosmwasm.wasm.v1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
class MsgSudoContract$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgSudoContract", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } }
    ], { "amino.name": "wasm/MsgSudoContract", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    message.contract = "";
    message.msg = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* string contract */
        2:
          message.contract = reader.string();
          break;
        case /* bytes msg */
        3:
          message.msg = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.contract !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.contract);
    if (message.msg.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.msg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSudoContract = new MsgSudoContract$Type();
class MsgSudoContractResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgSudoContractResponse", [
      {
        no: 1,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes data */
        1:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.data.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSudoContractResponse = new MsgSudoContractResponse$Type();
class MsgPinCodes$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgPinCodes", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "code_ids", kind: "scalar", repeat: 1, T: 4, L: 0, options: { "gogoproto.customname": "CodeIDs", "gogoproto.moretags": 'yaml:"code_ids"' } }
    ], { "amino.name": "wasm/MsgPinCodes", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    message.codeIds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* repeated uint64 code_ids */
        2:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.codeIds.push(reader.uint64().toBigInt());
          else
            message.codeIds.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.codeIds.length) {
      writer.tag(2, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.codeIds.length; i++)
        writer.uint64(message.codeIds[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPinCodes = new MsgPinCodes$Type();
class MsgPinCodesResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgPinCodesResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPinCodesResponse = new MsgPinCodesResponse$Type();
class MsgUnpinCodes$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUnpinCodes", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "code_ids", kind: "scalar", repeat: 1, T: 4, L: 0, options: { "gogoproto.customname": "CodeIDs", "gogoproto.moretags": 'yaml:"code_ids"' } }
    ], { "amino.name": "wasm/MsgUnpinCodes", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    message.codeIds = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* repeated uint64 code_ids */
        2:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.codeIds.push(reader.uint64().toBigInt());
          else
            message.codeIds.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.codeIds.length) {
      writer.tag(2, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.codeIds.length; i++)
        writer.uint64(message.codeIds[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUnpinCodes = new MsgUnpinCodes$Type();
class MsgUnpinCodesResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUnpinCodesResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUnpinCodesResponse = new MsgUnpinCodesResponse$Type();
class MsgStoreAndInstantiateContract$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgStoreAndInstantiateContract", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "wasm_byte_code", kind: "scalar", T: 12, options: { "gogoproto.customname": "WASMByteCode" } },
      { no: 4, name: "instantiate_permission", kind: "message", T: () => AccessConfig },
      {
        no: 5,
        name: "unpin_code",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 6, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 7,
        name: "label",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 8, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } },
      { no: 9, name: "funds", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } },
      {
        no: 10,
        name: "source",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 11,
        name: "builder",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 12,
        name: "code_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "amino.name": "wasm/MsgStoreAndInstantiateContract", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    message.wasmByteCode = new Uint8Array(0);
    message.unpinCode = false;
    message.admin = "";
    message.label = "";
    message.msg = new Uint8Array(0);
    message.funds = [];
    message.source = "";
    message.builder = "";
    message.codeHash = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* bytes wasm_byte_code */
        3:
          message.wasmByteCode = reader.bytes();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig instantiate_permission */
        4:
          message.instantiatePermission = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.instantiatePermission);
          break;
        case /* bool unpin_code */
        5:
          message.unpinCode = reader.bool();
          break;
        case /* string admin */
        6:
          message.admin = reader.string();
          break;
        case /* string label */
        7:
          message.label = reader.string();
          break;
        case /* bytes msg */
        8:
          message.msg = reader.bytes();
          break;
        case /* repeated cosmos.base.v1beta1.Coin funds */
        9:
          message.funds.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string source */
        10:
          message.source = reader.string();
          break;
        case /* string builder */
        11:
          message.builder = reader.string();
          break;
        case /* bytes code_hash */
        12:
          message.codeHash = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.wasmByteCode.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.wasmByteCode);
    if (message.instantiatePermission)
      AccessConfig.internalBinaryWrite(message.instantiatePermission, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.unpinCode !== false)
      writer.tag(5, WireType.Varint).bool(message.unpinCode);
    if (message.admin !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.admin);
    if (message.label !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.label);
    if (message.msg.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.msg);
    for (let i = 0; i < message.funds.length; i++)
      Coin.internalBinaryWrite(message.funds[i], writer.tag(9, WireType.LengthDelimited).fork(), options).join();
    if (message.source !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.source);
    if (message.builder !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.builder);
    if (message.codeHash.length)
      writer.tag(12, WireType.LengthDelimited).bytes(message.codeHash);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgStoreAndInstantiateContract = new MsgStoreAndInstantiateContract$Type();
class MsgStoreAndInstantiateContractResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgStoreAndInstantiateContractResponse", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* bytes data */
        2:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.data.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgStoreAndInstantiateContractResponse = new MsgStoreAndInstantiateContractResponse$Type();
class MsgAddCodeUploadParamsAddresses$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgAddCodeUploadParamsAddresses", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "addresses", kind: "scalar", repeat: 2, T: 9, options: { "gogoproto.moretags": 'yaml:"addresses"' } }
    ], { "amino.name": "wasm/MsgAddCodeUploadParamsAddresses", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    message.addresses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* repeated string addresses */
        2:
          message.addresses.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    for (let i = 0; i < message.addresses.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.addresses[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgAddCodeUploadParamsAddresses = new MsgAddCodeUploadParamsAddresses$Type();
class MsgAddCodeUploadParamsAddressesResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgAddCodeUploadParamsAddressesResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgAddCodeUploadParamsAddressesResponse = new MsgAddCodeUploadParamsAddressesResponse$Type();
class MsgRemoveCodeUploadParamsAddresses$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgRemoveCodeUploadParamsAddresses", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "addresses", kind: "scalar", repeat: 2, T: 9, options: { "gogoproto.moretags": 'yaml:"addresses"' } }
    ], { "amino.name": "wasm/MsgRemoveCodeUploadParamsAddresses", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    message.addresses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* repeated string addresses */
        2:
          message.addresses.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    for (let i = 0; i < message.addresses.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.addresses[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRemoveCodeUploadParamsAddresses = new MsgRemoveCodeUploadParamsAddresses$Type();
class MsgRemoveCodeUploadParamsAddressesResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgRemoveCodeUploadParamsAddressesResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRemoveCodeUploadParamsAddressesResponse = new MsgRemoveCodeUploadParamsAddressesResponse$Type();
class MsgStoreAndMigrateContract$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgStoreAndMigrateContract", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "wasm_byte_code", kind: "scalar", T: 12, options: { "gogoproto.customname": "WASMByteCode" } },
      { no: 3, name: "instantiate_permission", kind: "message", T: () => AccessConfig },
      {
        no: 4,
        name: "contract",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 5, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } }
    ], { "amino.name": "wasm/MsgStoreAndMigrateContract", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    message.wasmByteCode = new Uint8Array(0);
    message.contract = "";
    message.msg = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* bytes wasm_byte_code */
        2:
          message.wasmByteCode = reader.bytes();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig instantiate_permission */
        3:
          message.instantiatePermission = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.instantiatePermission);
          break;
        case /* string contract */
        4:
          message.contract = reader.string();
          break;
        case /* bytes msg */
        5:
          message.msg = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.wasmByteCode.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.wasmByteCode);
    if (message.instantiatePermission)
      AccessConfig.internalBinaryWrite(message.instantiatePermission, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.contract !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.contract);
    if (message.msg.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.msg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgStoreAndMigrateContract = new MsgStoreAndMigrateContract$Type();
class MsgStoreAndMigrateContractResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgStoreAndMigrateContractResponse", [
      { no: 1, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      {
        no: 2,
        name: "checksum",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codeId = 0n;
    message.checksum = new Uint8Array(0);
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 code_id */
        1:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* bytes checksum */
        2:
          message.checksum = reader.bytes();
          break;
        case /* bytes data */
        3:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.codeId);
    if (message.checksum.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.checksum);
    if (message.data.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgStoreAndMigrateContractResponse = new MsgStoreAndMigrateContractResponse$Type();
class MsgUpdateContractLabel$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUpdateContractLabel", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "new_label",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "contract", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "amino.name": "wasm/MsgUpdateContractLabel", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.newLabel = "";
    message.contract = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string new_label */
        2:
          message.newLabel = reader.string();
          break;
        case /* string contract */
        3:
          message.contract = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.newLabel !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.newLabel);
    if (message.contract !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.contract);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateContractLabel = new MsgUpdateContractLabel$Type();
class MsgUpdateContractLabelResponse$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.MsgUpdateContractLabelResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateContractLabelResponse = new MsgUpdateContractLabelResponse$Type();
const Msg = new ServiceType("cosmwasm.wasm.v1.Msg", [
  { name: "StoreCode", options: {}, I: MsgStoreCode, O: MsgStoreCodeResponse },
  { name: "InstantiateContract", options: {}, I: MsgInstantiateContract, O: MsgInstantiateContractResponse },
  { name: "InstantiateContract2", options: {}, I: MsgInstantiateContract2, O: MsgInstantiateContract2Response },
  { name: "ExecuteContract", options: {}, I: MsgExecuteContract, O: MsgExecuteContractResponse },
  { name: "MigrateContract", options: {}, I: MsgMigrateContract, O: MsgMigrateContractResponse },
  { name: "UpdateAdmin", options: {}, I: MsgUpdateAdmin, O: MsgUpdateAdminResponse },
  { name: "ClearAdmin", options: {}, I: MsgClearAdmin, O: MsgClearAdminResponse },
  { name: "UpdateInstantiateConfig", options: {}, I: MsgUpdateInstantiateConfig, O: MsgUpdateInstantiateConfigResponse },
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse },
  { name: "SudoContract", options: {}, I: MsgSudoContract, O: MsgSudoContractResponse },
  { name: "PinCodes", options: {}, I: MsgPinCodes, O: MsgPinCodesResponse },
  { name: "UnpinCodes", options: {}, I: MsgUnpinCodes, O: MsgUnpinCodesResponse },
  { name: "StoreAndInstantiateContract", options: {}, I: MsgStoreAndInstantiateContract, O: MsgStoreAndInstantiateContractResponse },
  { name: "RemoveCodeUploadParamsAddresses", options: {}, I: MsgRemoveCodeUploadParamsAddresses, O: MsgRemoveCodeUploadParamsAddressesResponse },
  { name: "AddCodeUploadParamsAddresses", options: {}, I: MsgAddCodeUploadParamsAddresses, O: MsgAddCodeUploadParamsAddressesResponse },
  { name: "StoreAndMigrateContract", options: {}, I: MsgStoreAndMigrateContract, O: MsgStoreAndMigrateContractResponse },
  { name: "UpdateContractLabel", options: {}, I: MsgUpdateContractLabel, O: MsgUpdateContractLabelResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgAddCodeUploadParamsAddresses,
  MsgAddCodeUploadParamsAddressesResponse,
  MsgClearAdmin,
  MsgClearAdminResponse,
  MsgExecuteContract,
  MsgExecuteContractResponse,
  MsgInstantiateContract,
  MsgInstantiateContract2,
  MsgInstantiateContract2Response,
  MsgInstantiateContractResponse,
  MsgMigrateContract,
  MsgMigrateContractResponse,
  MsgPinCodes,
  MsgPinCodesResponse,
  MsgRemoveCodeUploadParamsAddresses,
  MsgRemoveCodeUploadParamsAddressesResponse,
  MsgStoreAndInstantiateContract,
  MsgStoreAndInstantiateContractResponse,
  MsgStoreAndMigrateContract,
  MsgStoreAndMigrateContractResponse,
  MsgStoreCode,
  MsgStoreCodeResponse,
  MsgSudoContract,
  MsgSudoContractResponse,
  MsgUnpinCodes,
  MsgUnpinCodesResponse,
  MsgUpdateAdmin,
  MsgUpdateAdminResponse,
  MsgUpdateContractLabel,
  MsgUpdateContractLabelResponse,
  MsgUpdateInstantiateConfig,
  MsgUpdateInstantiateConfigResponse,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
