// @injectivelabs/core-proto-ts-v2/esm/generated/cosmwasm/wasm/v1/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Coin } from "../../../cosmos/base/v1beta1/coin_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
var AccessType = /* @__PURE__ */ ((AccessType2) => {
  AccessType2[AccessType2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  AccessType2[AccessType2["NOBODY"] = 1] = "NOBODY";
  AccessType2[AccessType2["EVERYBODY"] = 3] = "EVERYBODY";
  AccessType2[AccessType2["ANY_OF_ADDRESSES"] = 4] = "ANY_OF_ADDRESSES";
  return AccessType2;
})(AccessType || {});
var ContractCodeHistoryOperationType = /* @__PURE__ */ ((ContractCodeHistoryOperationType2) => {
  ContractCodeHistoryOperationType2[ContractCodeHistoryOperationType2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  ContractCodeHistoryOperationType2[ContractCodeHistoryOperationType2["INIT"] = 1] = "INIT";
  ContractCodeHistoryOperationType2[ContractCodeHistoryOperationType2["MIGRATE"] = 2] = "MIGRATE";
  ContractCodeHistoryOperationType2[ContractCodeHistoryOperationType2["GENESIS"] = 3] = "GENESIS";
  return ContractCodeHistoryOperationType2;
})(ContractCodeHistoryOperationType || {});
class AccessTypeParam$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.AccessTypeParam", [
      { no: 1, name: "value", kind: "enum", T: () => ["cosmwasm.wasm.v1.AccessType", AccessType, "ACCESS_TYPE_"], options: { "gogoproto.moretags": 'yaml:"value"' } }
    ], { "gogoproto.goproto_stringer": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmwasm.wasm.v1.AccessType value */
        1:
          message.value = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value !== 0)
      writer.tag(1, WireType.Varint).int32(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AccessTypeParam = new AccessTypeParam$Type();
class AccessConfig$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.AccessConfig", [
      { no: 1, name: "permission", kind: "enum", T: () => ["cosmwasm.wasm.v1.AccessType", AccessType, "ACCESS_TYPE_"], options: { "gogoproto.moretags": 'yaml:"permission"' } },
      { no: 3, name: "addresses", kind: "scalar", repeat: 2, T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "gogoproto.goproto_stringer": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.permission = 0;
    message.addresses = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmwasm.wasm.v1.AccessType permission */
        1:
          message.permission = reader.int32();
          break;
        case /* repeated string addresses */
        3:
          message.addresses.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.permission !== 0)
      writer.tag(1, WireType.Varint).int32(message.permission);
    for (let i = 0; i < message.addresses.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.addresses[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AccessConfig = new AccessConfig$Type();
class Params$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.Params", [
      { no: 1, name: "code_upload_access", kind: "message", T: () => AccessConfig, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"code_upload_access"', "amino.dont_omitempty": true } },
      { no: 2, name: "instantiate_default_permission", kind: "enum", T: () => ["cosmwasm.wasm.v1.AccessType", AccessType, "ACCESS_TYPE_"], options: { "gogoproto.moretags": 'yaml:"instantiate_default_permission"' } }
    ], { "gogoproto.goproto_stringer": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.instantiateDefaultPermission = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmwasm.wasm.v1.AccessConfig code_upload_access */
        1:
          message.codeUploadAccess = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.codeUploadAccess);
          break;
        case /* cosmwasm.wasm.v1.AccessType instantiate_default_permission */
        2:
          message.instantiateDefaultPermission = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeUploadAccess)
      AccessConfig.internalBinaryWrite(message.codeUploadAccess, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.instantiateDefaultPermission !== 0)
      writer.tag(2, WireType.Varint).int32(message.instantiateDefaultPermission);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
class CodeInfo$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.CodeInfo", [
      {
        no: 1,
        name: "code_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "creator", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 5, name: "instantiate_config", kind: "message", T: () => AccessConfig, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codeHash = new Uint8Array(0);
    message.creator = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes code_hash */
        1:
          message.codeHash = reader.bytes();
          break;
        case /* string creator */
        2:
          message.creator = reader.string();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig instantiate_config */
        5:
          message.instantiateConfig = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.instantiateConfig);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeHash.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.codeHash);
    if (message.creator !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.creator);
    if (message.instantiateConfig)
      AccessConfig.internalBinaryWrite(message.instantiateConfig, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const CodeInfo = new CodeInfo$Type();
class ContractInfo$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.ContractInfo", [
      { no: 1, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      { no: 2, name: "creator", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 4,
        name: "label",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 5, name: "created", kind: "message", T: () => AbsoluteTxPosition },
      { no: 6, name: "ibc_port_id", kind: "scalar", T: 9, options: { "gogoproto.customname": "IBCPortID" } },
      { no: 7, name: "extension", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmwasm.wasm.v1.ContractInfoExtension" } }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codeId = 0n;
    message.creator = "";
    message.admin = "";
    message.label = "";
    message.ibcPortId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 code_id */
        1:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string creator */
        2:
          message.creator = reader.string();
          break;
        case /* string admin */
        3:
          message.admin = reader.string();
          break;
        case /* string label */
        4:
          message.label = reader.string();
          break;
        case /* cosmwasm.wasm.v1.AbsoluteTxPosition created */
        5:
          message.created = AbsoluteTxPosition.internalBinaryRead(reader, reader.uint32(), options, message.created);
          break;
        case /* string ibc_port_id */
        6:
          message.ibcPortId = reader.string();
          break;
        case /* google.protobuf.Any extension */
        7:
          message.extension = Any.internalBinaryRead(reader, reader.uint32(), options, message.extension);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.codeId);
    if (message.creator !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.creator);
    if (message.admin !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.admin);
    if (message.label !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.label);
    if (message.created)
      AbsoluteTxPosition.internalBinaryWrite(message.created, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.ibcPortId !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.ibcPortId);
    if (message.extension)
      Any.internalBinaryWrite(message.extension, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ContractInfo = new ContractInfo$Type();
class ContractCodeHistoryEntry$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.ContractCodeHistoryEntry", [
      { no: 1, name: "operation", kind: "enum", T: () => ["cosmwasm.wasm.v1.ContractCodeHistoryOperationType", ContractCodeHistoryOperationType, "CONTRACT_CODE_HISTORY_OPERATION_TYPE_"] },
      { no: 2, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      { no: 3, name: "updated", kind: "message", T: () => AbsoluteTxPosition },
      { no: 4, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage", "amino.encoding": "inline_json" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.operation = 0;
    message.codeId = 0n;
    message.msg = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmwasm.wasm.v1.ContractCodeHistoryOperationType operation */
        1:
          message.operation = reader.int32();
          break;
        case /* uint64 code_id */
        2:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* cosmwasm.wasm.v1.AbsoluteTxPosition updated */
        3:
          message.updated = AbsoluteTxPosition.internalBinaryRead(reader, reader.uint32(), options, message.updated);
          break;
        case /* bytes msg */
        4:
          message.msg = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.operation !== 0)
      writer.tag(1, WireType.Varint).int32(message.operation);
    if (message.codeId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.codeId);
    if (message.updated)
      AbsoluteTxPosition.internalBinaryWrite(message.updated, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.msg.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.msg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ContractCodeHistoryEntry = new ContractCodeHistoryEntry$Type();
class AbsoluteTxPosition$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.AbsoluteTxPosition", [
      {
        no: 1,
        name: "block_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "tx_index",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.blockHeight = 0n;
    message.txIndex = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 block_height */
        1:
          message.blockHeight = reader.uint64().toBigInt();
          break;
        case /* uint64 tx_index */
        2:
          message.txIndex = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockHeight !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.blockHeight);
    if (message.txIndex !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.txIndex);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AbsoluteTxPosition = new AbsoluteTxPosition$Type();
class Model$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.Model", [
      { no: 1, name: "key", kind: "scalar", T: 12, options: { "gogoproto.casttype": "github.com/cometbft/cometbft/libs/bytes.HexBytes" } },
      {
        no: 2,
        name: "value",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.key = new Uint8Array(0);
    message.value = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes key */
        1:
          message.key = reader.bytes();
          break;
        case /* bytes value */
        2:
          message.value = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.key.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.key);
    if (message.value.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Model = new Model$Type();
class EventCodeStored$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.EventCodeStored", [
      { no: 1, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      {
        no: 2,
        name: "creator",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "access_config", kind: "message", T: () => AccessConfig },
      {
        no: 4,
        name: "checksum",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codeId = 0n;
    message.creator = "";
    message.checksum = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 code_id */
        1:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string creator */
        2:
          message.creator = reader.string();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig access_config */
        3:
          message.accessConfig = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.accessConfig);
          break;
        case /* bytes checksum */
        4:
          message.checksum = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.codeId);
    if (message.creator !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.creator);
    if (message.accessConfig)
      AccessConfig.internalBinaryWrite(message.accessConfig, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.checksum.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.checksum);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventCodeStored = new EventCodeStored$Type();
class EventContractInstantiated$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.EventContractInstantiated", [
      {
        no: 1,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "admin",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      { no: 4, name: "funds", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins" } },
      { no: 5, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage" } },
      {
        no: 6,
        name: "label",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "creator",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contractAddress = "";
    message.admin = "";
    message.codeId = 0n;
    message.funds = [];
    message.msg = new Uint8Array(0);
    message.label = "";
    message.creator = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract_address */
        1:
          message.contractAddress = reader.string();
          break;
        case /* string admin */
        2:
          message.admin = reader.string();
          break;
        case /* uint64 code_id */
        3:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* repeated cosmos.base.v1beta1.Coin funds */
        4:
          message.funds.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* bytes msg */
        5:
          message.msg = reader.bytes();
          break;
        case /* string label */
        6:
          message.label = reader.string();
          break;
        case /* string creator */
        7:
          message.creator = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contractAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contractAddress);
    if (message.admin !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.admin);
    if (message.codeId !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.codeId);
    for (let i = 0; i < message.funds.length; i++)
      Coin.internalBinaryWrite(message.funds[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.msg.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.msg);
    if (message.label !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.label);
    if (message.creator !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.creator);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventContractInstantiated = new EventContractInstantiated$Type();
class EventContractMigrated$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.EventContractMigrated", [
      { no: 1, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.customname": "CodeID" } },
      {
        no: 2,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "msg", kind: "scalar", T: 12, options: { "gogoproto.casttype": "RawContractMessage" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.codeId = 0n;
    message.contractAddress = "";
    message.msg = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 code_id */
        1:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string contract_address */
        2:
          message.contractAddress = reader.string();
          break;
        case /* bytes msg */
        3:
          message.msg = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.codeId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.codeId);
    if (message.contractAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.contractAddress);
    if (message.msg.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.msg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventContractMigrated = new EventContractMigrated$Type();
class EventContractAdminSet$Type extends MessageType {
  constructor() {
    super("cosmwasm.wasm.v1.EventContractAdminSet", [
      {
        no: 1,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "new_admin",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contractAddress = "";
    message.newAdmin = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract_address */
        1:
          message.contractAddress = reader.string();
          break;
        case /* string new_admin */
        2:
          message.newAdmin = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contractAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contractAddress);
    if (message.newAdmin !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.newAdmin);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventContractAdminSet = new EventContractAdminSet$Type();
export {
  AbsoluteTxPosition,
  AccessConfig,
  AccessType,
  AccessTypeParam,
  CodeInfo,
  ContractCodeHistoryEntry,
  ContractCodeHistoryOperationType,
  ContractInfo,
  EventCodeStored,
  EventContractAdminSet,
  EventContractInstantiated,
  EventContractMigrated,
  Model,
  Params
};
