// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/tx/config/v1/config_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Config$Type extends MessageType {
  constructor() {
    super("cosmos.tx.config.v1.Config", [
      {
        no: 1,
        name: "skip_ante_handler",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 2,
        name: "skip_post_handler",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ], { "cosmos.app.v1alpha1.module": { goImport: "github.com/cosmos/cosmos-sdk/x/auth/tx" } });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.skipAnteHandler = false;
    message.skipPostHandler = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bool skip_ante_handler */
        1:
          message.skipAnteHandler = reader.bool();
          break;
        case /* bool skip_post_handler */
        2:
          message.skipPostHandler = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.skipAnteHandler !== false)
      writer.tag(1, WireType.Varint).bool(message.skipAnteHandler);
    if (message.skipPostHandler !== false)
      writer.tag(2, WireType.Varint).bool(message.skipPostHandler);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Config = new Config$Type();
export {
  Config
};
