// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/group/module/v1/module_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Duration } from "../../../../google/protobuf/duration_pb.js";
class Module$Type extends MessageType {
  constructor() {
    super("cosmos.group.module.v1.Module", [
      { no: 1, name: "max_execution_period", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true, "amino.dont_omitempty": true } },
      {
        no: 2,
        name: "max_metadata_len",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "cosmos.app.v1alpha1.module": { goImport: "github.com/cosmos/cosmos-sdk/x/group" } });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.maxMetadataLen = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Duration max_execution_period */
        1:
          message.maxExecutionPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.maxExecutionPeriod);
          break;
        case /* uint64 max_metadata_len */
        2:
          message.maxMetadataLen = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.maxExecutionPeriod)
      Duration.internalBinaryWrite(message.maxExecutionPeriod, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.maxMetadataLen !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.maxMetadataLen);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Module = new Module$Type();
export {
  Module
};
