// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/crypto/keyring/v1/record_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { BIP44Params } from "../../hd/v1/hd_pb.js";
import { Any } from "../../../../google/protobuf/any_pb.js";
class Record$Type extends MessageType {
  constructor() {
    super("cosmos.crypto.keyring.v1.Record", [
      {
        no: 1,
        name: "name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "pub_key", kind: "message", T: () => Any },
      { no: 3, name: "local", kind: "message", oneof: "item", T: () => Record_Local },
      { no: 4, name: "ledger", kind: "message", oneof: "item", T: () => Record_Ledger },
      { no: 5, name: "multi", kind: "message", oneof: "item", T: () => Record_Multi },
      { no: 6, name: "offline", kind: "message", oneof: "item", T: () => Record_Offline }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.name = "";
    message.item = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string name */
        1:
          message.name = reader.string();
          break;
        case /* google.protobuf.Any pub_key */
        2:
          message.pubKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.pubKey);
          break;
        case /* cosmos.crypto.keyring.v1.Record.Local local */
        3:
          message.item = {
            oneofKind: "local",
            local: Record_Local.internalBinaryRead(reader, reader.uint32(), options, message.item.local)
          };
          break;
        case /* cosmos.crypto.keyring.v1.Record.Ledger ledger */
        4:
          message.item = {
            oneofKind: "ledger",
            ledger: Record_Ledger.internalBinaryRead(reader, reader.uint32(), options, message.item.ledger)
          };
          break;
        case /* cosmos.crypto.keyring.v1.Record.Multi multi */
        5:
          message.item = {
            oneofKind: "multi",
            multi: Record_Multi.internalBinaryRead(reader, reader.uint32(), options, message.item.multi)
          };
          break;
        case /* cosmos.crypto.keyring.v1.Record.Offline offline */
        6:
          message.item = {
            oneofKind: "offline",
            offline: Record_Offline.internalBinaryRead(reader, reader.uint32(), options, message.item.offline)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.name !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.name);
    if (message.pubKey)
      Any.internalBinaryWrite(message.pubKey, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.item.oneofKind === "local")
      Record_Local.internalBinaryWrite(message.item.local, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.item.oneofKind === "ledger")
      Record_Ledger.internalBinaryWrite(message.item.ledger, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.item.oneofKind === "multi")
      Record_Multi.internalBinaryWrite(message.item.multi, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.item.oneofKind === "offline")
      Record_Offline.internalBinaryWrite(message.item.offline, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Record = new Record$Type();
class Record_Local$Type extends MessageType {
  constructor() {
    super("cosmos.crypto.keyring.v1.Record.Local", [
      { no: 1, name: "priv_key", kind: "message", T: () => Any }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any priv_key */
        1:
          message.privKey = Any.internalBinaryRead(reader, reader.uint32(), options, message.privKey);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.privKey)
      Any.internalBinaryWrite(message.privKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Record_Local = new Record_Local$Type();
class Record_Ledger$Type extends MessageType {
  constructor() {
    super("cosmos.crypto.keyring.v1.Record.Ledger", [
      { no: 1, name: "path", kind: "message", T: () => BIP44Params }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.crypto.hd.v1.BIP44Params path */
        1:
          message.path = BIP44Params.internalBinaryRead(reader, reader.uint32(), options, message.path);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.path)
      BIP44Params.internalBinaryWrite(message.path, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Record_Ledger = new Record_Ledger$Type();
class Record_Multi$Type extends MessageType {
  constructor() {
    super("cosmos.crypto.keyring.v1.Record.Multi", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Record_Multi = new Record_Multi$Type();
class Record_Offline$Type extends MessageType {
  constructor() {
    super("cosmos.crypto.keyring.v1.Record.Offline", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Record_Offline = new Record_Offline$Type();
export {
  Record,
  Record_Ledger,
  Record_Local,
  Record_Multi,
  Record_Offline
};
