// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/interchain_accounts/genesis/v1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params as Params$ } from "../../host/v1/host_pb.js";
import { Params } from "../../controller/v1/controller_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.genesis.v1.GenesisState", [
      { no: 1, name: "controller_genesis_state", kind: "message", T: () => ControllerGenesisState, options: { "gogoproto.nullable": false } },
      { no: 2, name: "host_genesis_state", kind: "message", T: () => HostGenesisState, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.applications.interchain_accounts.genesis.v1.ControllerGenesisState controller_genesis_state */
        1:
          message.controllerGenesisState = ControllerGenesisState.internalBinaryRead(reader, reader.uint32(), options, message.controllerGenesisState);
          break;
        case /* ibc.applications.interchain_accounts.genesis.v1.HostGenesisState host_genesis_state */
        2:
          message.hostGenesisState = HostGenesisState.internalBinaryRead(reader, reader.uint32(), options, message.hostGenesisState);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.controllerGenesisState)
      ControllerGenesisState.internalBinaryWrite(message.controllerGenesisState, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.hostGenesisState)
      HostGenesisState.internalBinaryWrite(message.hostGenesisState, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class ControllerGenesisState$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.genesis.v1.ControllerGenesisState", [
      { no: 1, name: "active_channels", kind: "message", repeat: 2, T: () => ActiveChannel, options: { "gogoproto.nullable": false } },
      { no: 2, name: "interchain_accounts", kind: "message", repeat: 2, T: () => RegisteredInterchainAccount, options: { "gogoproto.nullable": false } },
      {
        no: 3,
        name: "ports",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.activeChannels = [];
    message.interchainAccounts = [];
    message.ports = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.applications.interchain_accounts.genesis.v1.ActiveChannel active_channels */
        1:
          message.activeChannels.push(ActiveChannel.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.applications.interchain_accounts.genesis.v1.RegisteredInterchainAccount interchain_accounts */
        2:
          message.interchainAccounts.push(RegisteredInterchainAccount.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated string ports */
        3:
          message.ports.push(reader.string());
          break;
        case /* ibc.applications.interchain_accounts.controller.v1.Params params */
        4:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.activeChannels.length; i++)
      ActiveChannel.internalBinaryWrite(message.activeChannels[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.interchainAccounts.length; i++)
      RegisteredInterchainAccount.internalBinaryWrite(message.interchainAccounts[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.ports.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.ports[i]);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ControllerGenesisState = new ControllerGenesisState$Type();
class HostGenesisState$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.genesis.v1.HostGenesisState", [
      { no: 1, name: "active_channels", kind: "message", repeat: 2, T: () => ActiveChannel, options: { "gogoproto.nullable": false } },
      { no: 2, name: "interchain_accounts", kind: "message", repeat: 2, T: () => RegisteredInterchainAccount, options: { "gogoproto.nullable": false } },
      {
        no: 3,
        name: "port",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "params", kind: "message", T: () => Params$, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.activeChannels = [];
    message.interchainAccounts = [];
    message.port = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.applications.interchain_accounts.genesis.v1.ActiveChannel active_channels */
        1:
          message.activeChannels.push(ActiveChannel.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.applications.interchain_accounts.genesis.v1.RegisteredInterchainAccount interchain_accounts */
        2:
          message.interchainAccounts.push(RegisteredInterchainAccount.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string port */
        3:
          message.port = reader.string();
          break;
        case /* ibc.applications.interchain_accounts.host.v1.Params params */
        4:
          message.params = Params$.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.activeChannels.length; i++)
      ActiveChannel.internalBinaryWrite(message.activeChannels[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.interchainAccounts.length; i++)
      RegisteredInterchainAccount.internalBinaryWrite(message.interchainAccounts[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.port !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.port);
    if (message.params)
      Params$.internalBinaryWrite(message.params, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const HostGenesisState = new HostGenesisState$Type();
class ActiveChannel$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.genesis.v1.ActiveChannel", [
      {
        no: 1,
        name: "connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "is_middleware_enabled",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.connectionId = "";
    message.portId = "";
    message.channelId = "";
    message.isMiddlewareEnabled = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string connection_id */
        1:
          message.connectionId = reader.string();
          break;
        case /* string port_id */
        2:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        3:
          message.channelId = reader.string();
          break;
        case /* bool is_middleware_enabled */
        4:
          message.isMiddlewareEnabled = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.connectionId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.connectionId);
    if (message.portId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.channelId);
    if (message.isMiddlewareEnabled !== false)
      writer.tag(4, WireType.Varint).bool(message.isMiddlewareEnabled);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ActiveChannel = new ActiveChannel$Type();
class RegisteredInterchainAccount$Type extends MessageType {
  constructor() {
    super("ibc.applications.interchain_accounts.genesis.v1.RegisteredInterchainAccount", [
      {
        no: 1,
        name: "connection_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "account_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.connectionId = "";
    message.portId = "";
    message.accountAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string connection_id */
        1:
          message.connectionId = reader.string();
          break;
        case /* string port_id */
        2:
          message.portId = reader.string();
          break;
        case /* string account_address */
        3:
          message.accountAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.connectionId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.connectionId);
    if (message.portId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.portId);
    if (message.accountAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.accountAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RegisteredInterchainAccount = new RegisteredInterchainAccount$Type();
export {
  ActiveChannel,
  ControllerGenesisState,
  GenesisState,
  HostGenesisState,
  RegisteredInterchainAccount
};
