// @injectivelabs/utils/dist/esm/constants-BxcSsKjm.js
import { BigNumber } from "bignumber.js";

//#region src/classes/BigNumber.ts
var BigNumber_default = BigNumber;
/**
* Converts a value to BigNumber if it isn't already
*/
const toBigNumber = (value) => {
	if (value instanceof BigNumber) return value;
	return new BigNumber(value);
};
/**
* Converts a value to Wei (multiplies by 10^decimals) default to 18 decimals
*/
const toChainFormat = (value, decimals = 18) => {
	return toBigNumber(value).multipliedBy(new BigNumber(10).pow(decimals));
};
/**
* Converts a value from Wei to Base (divides by 10^decimals) default to 18 decimals
*/
const toHumanReadable = (value, decimals = 18) => {
	return toBigNumber(value).dividedBy(new BigNumber(10).pow(decimals));
};

//#endregion
//#region src/constants.ts
const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
const INJ_DENOM = "inj";
const INJECTIVE_DENOM = "inj";
const DEFAULT_FEE_DENOM = "inj";
const DEFAULT_GAS_LIMIT = 4e5;
const DEFAULT_IBC_GAS_LIMIT = 3e5;
const DEFAULT_GAS_PRICE = 16e7;
const DEFAULT_EXCHANGE_LIMIT = 2e5;
const DEFAULT_BRIDGE_FEE_DENOM = "inj";
const DEFAULT_BRIDGE_FEE_PRICE = "160000000";
const DEFAULT_BRIDGE_FEE_AMOUNT = "200000000000000";
const DEFAULT_BLOCK_TIMEOUT_HEIGHT = 120;
const DEFAULT_BLOCK_TIME_IN_SECONDS = .7;
/**
* Default timeout for transaction block inclusion polling.
*
* Calculation: 120 blocks × 0.7 seconds/block = 84 seconds = 84,000ms
*/
const DEFAULT_TX_BLOCK_INCLUSION_TIMEOUT_IN_MS = Math.floor(DEFAULT_BLOCK_TIMEOUT_HEIGHT * DEFAULT_BLOCK_TIME_IN_SECONDS * 1e3);
const DEFAULT_TIMESTAMP_TIMEOUT_MS = 60 * 1e3 * 3;
const getDefaultStdFee = () => ({
	amount: [{
		amount: new BigNumber_default(DEFAULT_GAS_LIMIT).times(DEFAULT_GAS_PRICE).toString(),
		denom: "inj"
	}],
	gas: DEFAULT_GAS_LIMIT.toString(),
	payer: "",
	granter: "",
	feePayer: ""
});
const DEFAULT_STD_FEE_BY_DENOM = (denom = "inj") => ({
	amount: [{
		denom,
		amount: new BigNumber_default(DEFAULT_GAS_LIMIT).times(DEFAULT_GAS_PRICE).toString()
	}],
	gas: DEFAULT_GAS_LIMIT.toString()
});

//#endregion
export { getDefaultStdFee as _, DEFAULT_BRIDGE_FEE_PRICE as a, toChainFormat as b, DEFAULT_GAS_LIMIT as c, DEFAULT_STD_FEE_BY_DENOM as d, DEFAULT_TIMESTAMP_TIMEOUT_MS as f, ZERO_ADDRESS as g, INJ_DENOM as h, DEFAULT_BRIDGE_FEE_DENOM as i, DEFAULT_GAS_PRICE as l, INJECTIVE_DENOM as m, DEFAULT_BLOCK_TIME_IN_SECONDS as n, DEFAULT_EXCHANGE_LIMIT as o, DEFAULT_TX_BLOCK_INCLUSION_TIMEOUT_IN_MS as p, DEFAULT_BRIDGE_FEE_AMOUNT as r, DEFAULT_FEE_DENOM as s, DEFAULT_BLOCK_TIMEOUT_HEIGHT as t, DEFAULT_IBC_GAS_LIMIT as u, BigNumber_default as v, toHumanReadable as x, toBigNumber as y };
