// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/transfer/v1/authz_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Coin } from "../../../../cosmos/base/v1beta1/coin_pb.js";
class Allocation$Type extends MessageType {
  constructor() {
    super("ibc.applications.transfer.v1.Allocation", [
      {
        no: 1,
        name: "source_port",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "source_channel",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "spend_limit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins" } },
      {
        no: 4,
        name: "allow_list",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "allowed_packet_data",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sourcePort = "";
    message.sourceChannel = "";
    message.spendLimit = [];
    message.allowList = [];
    message.allowedPacketData = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string source_port */
        1:
          message.sourcePort = reader.string();
          break;
        case /* string source_channel */
        2:
          message.sourceChannel = reader.string();
          break;
        case /* repeated cosmos.base.v1beta1.Coin spend_limit */
        3:
          message.spendLimit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated string allow_list */
        4:
          message.allowList.push(reader.string());
          break;
        case /* repeated string allowed_packet_data */
        5:
          message.allowedPacketData.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sourcePort !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sourcePort);
    if (message.sourceChannel !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.sourceChannel);
    for (let i = 0; i < message.spendLimit.length; i++)
      Coin.internalBinaryWrite(message.spendLimit[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.allowList.length; i++)
      writer.tag(4, WireType.LengthDelimited).string(message.allowList[i]);
    for (let i = 0; i < message.allowedPacketData.length; i++)
      writer.tag(5, WireType.LengthDelimited).string(message.allowedPacketData[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Allocation = new Allocation$Type();
class TransferAuthorization$Type extends MessageType {
  constructor() {
    super("ibc.applications.transfer.v1.TransferAuthorization", [
      { no: 1, name: "allocations", kind: "message", repeat: 2, T: () => Allocation, options: { "gogoproto.nullable": false } }
    ], { "cosmos_proto.implements_interface": ["cosmos.authz.v1beta1.Authorization"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.allocations = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.applications.transfer.v1.Allocation allocations */
        1:
          message.allocations.push(Allocation.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.allocations.length; i++)
      Allocation.internalBinaryWrite(message.allocations[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TransferAuthorization = new TransferAuthorization$Type();
export {
  Allocation,
  TransferAuthorization
};
