// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/transfer/v1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Coin } from "../../../../cosmos/base/v1beta1/coin_pb.js";
import { Params } from "./transfer_pb.js";
import { DenomTrace } from "./transfer_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("ibc.applications.transfer.v1.GenesisState", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "denom_traces", kind: "message", repeat: 2, T: () => DenomTrace, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "Traces" } },
      { no: 3, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } },
      { no: 4, name: "total_escrowed", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.denomTraces = [];
    message.totalEscrowed = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* repeated ibc.applications.transfer.v1.DenomTrace denom_traces */
        2:
          message.denomTraces.push(DenomTrace.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* ibc.applications.transfer.v1.Params params */
        3:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        case /* repeated cosmos.base.v1beta1.Coin total_escrowed */
        4:
          message.totalEscrowed.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    for (let i = 0; i < message.denomTraces.length; i++)
      DenomTrace.internalBinaryWrite(message.denomTraces[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.totalEscrowed.length; i++)
      Coin.internalBinaryWrite(message.totalEscrowed[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
export {
  GenesisState
};
