// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/transfer/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * DenomTraces queries all denomination traces.
   *
   * @generated from protobuf rpc: DenomTraces
   */
  denomTraces(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * DenomTrace queries a denomination trace information.
   *
   * @generated from protobuf rpc: DenomTrace
   */
  denomTrace(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Params queries all parameters of the ibc-transfer module.
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * DenomHash queries a denomination hash information.
   *
   * @generated from protobuf rpc: DenomHash
   */
  denomHash(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * EscrowAddress returns the escrow address for a particular port and channel id.
   *
   * @generated from protobuf rpc: EscrowAddress
   */
  escrowAddress(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TotalEscrowForDenom returns the total amount of tokens in escrow based on the denom.
   *
   * @generated from protobuf rpc: TotalEscrowForDenom
   */
  totalEscrowForDenom(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
