// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/applications/transfer/v1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./transfer_pb.js";
import { Height } from "../../../core/client/v1/client_pb.js";
import { Coin } from "../../../../cosmos/base/v1beta1/coin_pb.js";
class MsgTransfer$Type extends MessageType {
  constructor() {
    super("ibc.applications.transfer.v1.MsgTransfer", [
      {
        no: 1,
        name: "source_port",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "source_channel",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "token", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 4,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "receiver",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 6, name: "timeout_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 7,
        name: "timeout_timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 8,
        name: "memo",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "amino.name": "cosmos-sdk/MsgTransfer", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sourcePort = "";
    message.sourceChannel = "";
    message.sender = "";
    message.receiver = "";
    message.timeoutTimestamp = 0n;
    message.memo = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string source_port */
        1:
          message.sourcePort = reader.string();
          break;
        case /* string source_channel */
        2:
          message.sourceChannel = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin token */
        3:
          message.token = Coin.internalBinaryRead(reader, reader.uint32(), options, message.token);
          break;
        case /* string sender */
        4:
          message.sender = reader.string();
          break;
        case /* string receiver */
        5:
          message.receiver = reader.string();
          break;
        case /* ibc.core.client.v1.Height timeout_height */
        6:
          message.timeoutHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.timeoutHeight);
          break;
        case /* uint64 timeout_timestamp */
        7:
          message.timeoutTimestamp = reader.uint64().toBigInt();
          break;
        case /* string memo */
        8:
          message.memo = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sourcePort !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sourcePort);
    if (message.sourceChannel !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.sourceChannel);
    if (message.token)
      Coin.internalBinaryWrite(message.token, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.sender !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.sender);
    if (message.receiver !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.receiver);
    if (message.timeoutHeight)
      Height.internalBinaryWrite(message.timeoutHeight, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.timeoutTimestamp !== 0n)
      writer.tag(7, WireType.Varint).uint64(message.timeoutTimestamp);
    if (message.memo !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.memo);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTransfer = new MsgTransfer$Type();
class MsgTransferResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.transfer.v1.MsgTransferResponse", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTransferResponse = new MsgTransferResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("ibc.applications.transfer.v1.MsgUpdateParams", [
      {
        no: 1,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string signer */
        1:
          message.signer = reader.string();
          break;
        case /* ibc.applications.transfer.v1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.signer !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.signer);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("ibc.applications.transfer.v1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
const Msg = new ServiceType("ibc.applications.transfer.v1.Msg", [
  { name: "Transfer", options: {}, I: MsgTransfer, O: MsgTransferResponse },
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgTransfer,
  MsgTransferResponse,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
