// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/group/v1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Vote } from "./types_pb.js";
import { Proposal } from "./types_pb.js";
import { GroupPolicyInfo } from "./types_pb.js";
import { GroupMember } from "./types_pb.js";
import { GroupInfo } from "./types_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.GenesisState", [
      {
        no: 1,
        name: "group_seq",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "groups", kind: "message", repeat: 2, T: () => GroupInfo },
      { no: 3, name: "group_members", kind: "message", repeat: 2, T: () => GroupMember },
      {
        no: 4,
        name: "group_policy_seq",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 5, name: "group_policies", kind: "message", repeat: 2, T: () => GroupPolicyInfo },
      {
        no: 6,
        name: "proposal_seq",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 7, name: "proposals", kind: "message", repeat: 2, T: () => Proposal },
      { no: 8, name: "votes", kind: "message", repeat: 2, T: () => Vote }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.groupSeq = 0n;
    message.groups = [];
    message.groupMembers = [];
    message.groupPolicySeq = 0n;
    message.groupPolicies = [];
    message.proposalSeq = 0n;
    message.proposals = [];
    message.votes = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 group_seq */
        1:
          message.groupSeq = reader.uint64().toBigInt();
          break;
        case /* repeated cosmos.group.v1.GroupInfo groups */
        2:
          message.groups.push(GroupInfo.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.group.v1.GroupMember group_members */
        3:
          message.groupMembers.push(GroupMember.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* uint64 group_policy_seq */
        4:
          message.groupPolicySeq = reader.uint64().toBigInt();
          break;
        case /* repeated cosmos.group.v1.GroupPolicyInfo group_policies */
        5:
          message.groupPolicies.push(GroupPolicyInfo.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* uint64 proposal_seq */
        6:
          message.proposalSeq = reader.uint64().toBigInt();
          break;
        case /* repeated cosmos.group.v1.Proposal proposals */
        7:
          message.proposals.push(Proposal.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.group.v1.Vote votes */
        8:
          message.votes.push(Vote.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.groupSeq !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.groupSeq);
    for (let i = 0; i < message.groups.length; i++)
      GroupInfo.internalBinaryWrite(message.groups[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.groupMembers.length; i++)
      GroupMember.internalBinaryWrite(message.groupMembers[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.groupPolicySeq !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.groupPolicySeq);
    for (let i = 0; i < message.groupPolicies.length; i++)
      GroupPolicyInfo.internalBinaryWrite(message.groupPolicies[i], writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.proposalSeq !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.proposalSeq);
    for (let i = 0; i < message.proposals.length; i++)
      Proposal.internalBinaryWrite(message.proposals[i], writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.votes.length; i++)
      Vote.internalBinaryWrite(message.votes[i], writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
export {
  GenesisState
};
