// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/group/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * GroupInfo queries group info based on group id.
   *
   * @generated from protobuf rpc: GroupInfo
   */
  groupInfo(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GroupPolicyInfo queries group policy info based on account address of group policy.
   *
   * @generated from protobuf rpc: GroupPolicyInfo
   */
  groupPolicyInfo(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GroupMembers queries members of a group by group id.
   *
   * @generated from protobuf rpc: GroupMembers
   */
  groupMembers(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GroupsByAdmin queries groups by admin address.
   *
   * @generated from protobuf rpc: GroupsByAdmin
   */
  groupsByAdmin(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GroupPoliciesByGroup queries group policies by group id.
   *
   * @generated from protobuf rpc: GroupPoliciesByGroup
   */
  groupPoliciesByGroup(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GroupPoliciesByAdmin queries group policies by admin address.
   *
   * @generated from protobuf rpc: GroupPoliciesByAdmin
   */
  groupPoliciesByAdmin(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Proposal queries a proposal based on proposal id.
   *
   * @generated from protobuf rpc: Proposal
   */
  proposal(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ProposalsByGroupPolicy queries proposals based on account address of group policy.
   *
   * @generated from protobuf rpc: ProposalsByGroupPolicy
   */
  proposalsByGroupPolicy(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * VoteByProposalVoter queries a vote by proposal id and voter.
   *
   * @generated from protobuf rpc: VoteByProposalVoter
   */
  voteByProposalVoter(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * VotesByProposal queries a vote by proposal id.
   *
   * @generated from protobuf rpc: VotesByProposal
   */
  votesByProposal(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * VotesByVoter queries a vote by voter.
   *
   * @generated from protobuf rpc: VotesByVoter
   */
  votesByVoter(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * GroupsByMember queries groups by member address.
   *
   * @generated from protobuf rpc: GroupsByMember
   */
  groupsByMember(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TallyResult returns the tally result of a proposal. If the proposal is
   * still in voting period, then this query computes the current tally state,
   * which might not be final. On the other hand, if the proposal is final,
   * then it simply returns the `final_tally_result` state stored in the
   * proposal itself.
   *
   * @generated from protobuf rpc: TallyResult
   */
  tallyResult(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Groups queries all groups in state.
   *
   * Since: cosmos-sdk 0.47.1
   *
   * @generated from protobuf rpc: Groups
   */
  groups(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
