// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/group/v1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ProposalExecutorResult } from "./types_pb.js";
import { VoteOption } from "./types_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
import { MemberRequest } from "./types_pb.js";
var Exec = /* @__PURE__ */ ((Exec2) => {
  Exec2[Exec2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  Exec2[Exec2["TRY"] = 1] = "TRY";
  return Exec2;
})(Exec || {});
class MsgCreateGroup$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgCreateGroup", [
      { no: 1, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "members", kind: "message", repeat: 2, T: () => MemberRequest, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 3,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "amino.name": "cosmos-sdk/MsgCreateGroup", "cosmos.msg.v1.signer": ["admin"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.admin = "";
    message.members = [];
    message.metadata = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string admin */
        1:
          message.admin = reader.string();
          break;
        case /* repeated cosmos.group.v1.MemberRequest members */
        2:
          message.members.push(MemberRequest.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string metadata */
        3:
          message.metadata = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.admin !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.admin);
    for (let i = 0; i < message.members.length; i++)
      MemberRequest.internalBinaryWrite(message.members[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.metadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.metadata);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateGroup = new MsgCreateGroup$Type();
class MsgCreateGroupResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgCreateGroupResponse", [
      {
        no: 1,
        name: "group_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.groupId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 group_id */
        1:
          message.groupId = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.groupId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.groupId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateGroupResponse = new MsgCreateGroupResponse$Type();
class MsgUpdateGroupMembers$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupMembers", [
      { no: 1, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "group_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "member_updates", kind: "message", repeat: 2, T: () => MemberRequest, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "amino.name": "cosmos-sdk/MsgUpdateGroupMembers", "cosmos.msg.v1.signer": ["admin"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.admin = "";
    message.groupId = 0n;
    message.memberUpdates = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string admin */
        1:
          message.admin = reader.string();
          break;
        case /* uint64 group_id */
        2:
          message.groupId = reader.uint64().toBigInt();
          break;
        case /* repeated cosmos.group.v1.MemberRequest member_updates */
        3:
          message.memberUpdates.push(MemberRequest.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.admin !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.admin);
    if (message.groupId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.groupId);
    for (let i = 0; i < message.memberUpdates.length; i++)
      MemberRequest.internalBinaryWrite(message.memberUpdates[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupMembers = new MsgUpdateGroupMembers$Type();
class MsgUpdateGroupMembersResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupMembersResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupMembersResponse = new MsgUpdateGroupMembersResponse$Type();
class MsgUpdateGroupAdmin$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupAdmin", [
      { no: 1, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "group_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "new_admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "amino.name": "cosmos-sdk/MsgUpdateGroupAdmin", "cosmos.msg.v1.signer": ["admin"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.admin = "";
    message.groupId = 0n;
    message.newAdmin = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string admin */
        1:
          message.admin = reader.string();
          break;
        case /* uint64 group_id */
        2:
          message.groupId = reader.uint64().toBigInt();
          break;
        case /* string new_admin */
        3:
          message.newAdmin = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.admin !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.admin);
    if (message.groupId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.groupId);
    if (message.newAdmin !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.newAdmin);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupAdmin = new MsgUpdateGroupAdmin$Type();
class MsgUpdateGroupAdminResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupAdminResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupAdminResponse = new MsgUpdateGroupAdminResponse$Type();
class MsgUpdateGroupMetadata$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupMetadata", [
      { no: 1, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "group_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "amino.name": "cosmos-sdk/MsgUpdateGroupMetadata", "cosmos.msg.v1.signer": ["admin"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.admin = "";
    message.groupId = 0n;
    message.metadata = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string admin */
        1:
          message.admin = reader.string();
          break;
        case /* uint64 group_id */
        2:
          message.groupId = reader.uint64().toBigInt();
          break;
        case /* string metadata */
        3:
          message.metadata = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.admin !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.admin);
    if (message.groupId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.groupId);
    if (message.metadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.metadata);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupMetadata = new MsgUpdateGroupMetadata$Type();
class MsgUpdateGroupMetadataResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupMetadataResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupMetadataResponse = new MsgUpdateGroupMetadataResponse$Type();
class MsgCreateGroupPolicy$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgCreateGroupPolicy", [
      { no: 1, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "group_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "decision_policy", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.group.v1.DecisionPolicy" } }
    ], { "gogoproto.goproto_getters": false, "amino.name": "cosmos-sdk/MsgCreateGroupPolicy", "cosmos.msg.v1.signer": ["admin"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.admin = "";
    message.groupId = 0n;
    message.metadata = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string admin */
        1:
          message.admin = reader.string();
          break;
        case /* uint64 group_id */
        2:
          message.groupId = reader.uint64().toBigInt();
          break;
        case /* string metadata */
        3:
          message.metadata = reader.string();
          break;
        case /* google.protobuf.Any decision_policy */
        4:
          message.decisionPolicy = Any.internalBinaryRead(reader, reader.uint32(), options, message.decisionPolicy);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.admin !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.admin);
    if (message.groupId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.groupId);
    if (message.metadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.metadata);
    if (message.decisionPolicy)
      Any.internalBinaryWrite(message.decisionPolicy, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateGroupPolicy = new MsgCreateGroupPolicy$Type();
class MsgCreateGroupPolicyResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgCreateGroupPolicyResponse", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateGroupPolicyResponse = new MsgCreateGroupPolicyResponse$Type();
class MsgUpdateGroupPolicyAdmin$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupPolicyAdmin", [
      { no: 1, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "group_policy_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "new_admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "amino.name": "cosmos-sdk/MsgUpdateGroupPolicyAdmin", "cosmos.msg.v1.signer": ["admin"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.admin = "";
    message.groupPolicyAddress = "";
    message.newAdmin = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string admin */
        1:
          message.admin = reader.string();
          break;
        case /* string group_policy_address */
        2:
          message.groupPolicyAddress = reader.string();
          break;
        case /* string new_admin */
        3:
          message.newAdmin = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.admin !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.admin);
    if (message.groupPolicyAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.groupPolicyAddress);
    if (message.newAdmin !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.newAdmin);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupPolicyAdmin = new MsgUpdateGroupPolicyAdmin$Type();
class MsgUpdateGroupPolicyAdminResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupPolicyAdminResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupPolicyAdminResponse = new MsgUpdateGroupPolicyAdminResponse$Type();
class MsgCreateGroupWithPolicy$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgCreateGroupWithPolicy", [
      { no: 1, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "members", kind: "message", repeat: 2, T: () => MemberRequest, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      {
        no: 3,
        name: "group_metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "group_policy_metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "group_policy_as_admin",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 6, name: "decision_policy", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.group.v1.DecisionPolicy" } }
    ], { "gogoproto.goproto_getters": false, "amino.name": "cosmos-sdk/MsgCreateGroupWithPolicy", "cosmos.msg.v1.signer": ["admin"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.admin = "";
    message.members = [];
    message.groupMetadata = "";
    message.groupPolicyMetadata = "";
    message.groupPolicyAsAdmin = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string admin */
        1:
          message.admin = reader.string();
          break;
        case /* repeated cosmos.group.v1.MemberRequest members */
        2:
          message.members.push(MemberRequest.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string group_metadata */
        3:
          message.groupMetadata = reader.string();
          break;
        case /* string group_policy_metadata */
        4:
          message.groupPolicyMetadata = reader.string();
          break;
        case /* bool group_policy_as_admin */
        5:
          message.groupPolicyAsAdmin = reader.bool();
          break;
        case /* google.protobuf.Any decision_policy */
        6:
          message.decisionPolicy = Any.internalBinaryRead(reader, reader.uint32(), options, message.decisionPolicy);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.admin !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.admin);
    for (let i = 0; i < message.members.length; i++)
      MemberRequest.internalBinaryWrite(message.members[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.groupMetadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.groupMetadata);
    if (message.groupPolicyMetadata !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.groupPolicyMetadata);
    if (message.groupPolicyAsAdmin !== false)
      writer.tag(5, WireType.Varint).bool(message.groupPolicyAsAdmin);
    if (message.decisionPolicy)
      Any.internalBinaryWrite(message.decisionPolicy, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateGroupWithPolicy = new MsgCreateGroupWithPolicy$Type();
class MsgCreateGroupWithPolicyResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgCreateGroupWithPolicyResponse", [
      {
        no: 1,
        name: "group_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "group_policy_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.groupId = 0n;
    message.groupPolicyAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 group_id */
        1:
          message.groupId = reader.uint64().toBigInt();
          break;
        case /* string group_policy_address */
        2:
          message.groupPolicyAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.groupId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.groupId);
    if (message.groupPolicyAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.groupPolicyAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateGroupWithPolicyResponse = new MsgCreateGroupWithPolicyResponse$Type();
class MsgUpdateGroupPolicyDecisionPolicy$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupPolicyDecisionPolicy", [
      { no: 1, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "group_policy_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "decision_policy", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.group.v1.DecisionPolicy" } }
    ], { "gogoproto.goproto_getters": false, "amino.name": "cosmos-sdk/MsgUpdateGroupDecisionPolicy", "cosmos.msg.v1.signer": ["admin"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.admin = "";
    message.groupPolicyAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string admin */
        1:
          message.admin = reader.string();
          break;
        case /* string group_policy_address */
        2:
          message.groupPolicyAddress = reader.string();
          break;
        case /* google.protobuf.Any decision_policy */
        3:
          message.decisionPolicy = Any.internalBinaryRead(reader, reader.uint32(), options, message.decisionPolicy);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.admin !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.admin);
    if (message.groupPolicyAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.groupPolicyAddress);
    if (message.decisionPolicy)
      Any.internalBinaryWrite(message.decisionPolicy, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupPolicyDecisionPolicy = new MsgUpdateGroupPolicyDecisionPolicy$Type();
class MsgUpdateGroupPolicyDecisionPolicyResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupPolicyDecisionPolicyResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupPolicyDecisionPolicyResponse = new MsgUpdateGroupPolicyDecisionPolicyResponse$Type();
class MsgUpdateGroupPolicyMetadata$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupPolicyMetadata", [
      { no: 1, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "group_policy_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 3,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "amino.name": "cosmos-sdk/MsgUpdateGroupPolicyMetadata", "cosmos.msg.v1.signer": ["admin"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.admin = "";
    message.groupPolicyAddress = "";
    message.metadata = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string admin */
        1:
          message.admin = reader.string();
          break;
        case /* string group_policy_address */
        2:
          message.groupPolicyAddress = reader.string();
          break;
        case /* string metadata */
        3:
          message.metadata = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.admin !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.admin);
    if (message.groupPolicyAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.groupPolicyAddress);
    if (message.metadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.metadata);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupPolicyMetadata = new MsgUpdateGroupPolicyMetadata$Type();
class MsgUpdateGroupPolicyMetadataResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgUpdateGroupPolicyMetadataResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateGroupPolicyMetadataResponse = new MsgUpdateGroupPolicyMetadataResponse$Type();
class MsgSubmitProposal$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgSubmitProposal", [
      { no: 1, name: "group_policy_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "proposers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "messages", kind: "message", repeat: 2, T: () => Any },
      { no: 5, name: "exec", kind: "enum", T: () => ["cosmos.group.v1.Exec", Exec, "EXEC_"] },
      {
        no: 6,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "summary",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "amino.name": "cosmos-sdk/group/MsgSubmitProposal", "cosmos.msg.v1.signer": ["proposers"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.groupPolicyAddress = "";
    message.proposers = [];
    message.metadata = "";
    message.messages = [];
    message.exec = 0;
    message.title = "";
    message.summary = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string group_policy_address */
        1:
          message.groupPolicyAddress = reader.string();
          break;
        case /* repeated string proposers */
        2:
          message.proposers.push(reader.string());
          break;
        case /* string metadata */
        3:
          message.metadata = reader.string();
          break;
        case /* repeated google.protobuf.Any messages */
        4:
          message.messages.push(Any.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.group.v1.Exec exec */
        5:
          message.exec = reader.int32();
          break;
        case /* string title */
        6:
          message.title = reader.string();
          break;
        case /* string summary */
        7:
          message.summary = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.groupPolicyAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.groupPolicyAddress);
    for (let i = 0; i < message.proposers.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.proposers[i]);
    if (message.metadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.metadata);
    for (let i = 0; i < message.messages.length; i++)
      Any.internalBinaryWrite(message.messages[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.exec !== 0)
      writer.tag(5, WireType.Varint).int32(message.exec);
    if (message.title !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.title);
    if (message.summary !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.summary);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSubmitProposal = new MsgSubmitProposal$Type();
class MsgSubmitProposalResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgSubmitProposalResponse", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSubmitProposalResponse = new MsgSubmitProposalResponse$Type();
class MsgWithdrawProposal$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgWithdrawProposal", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "amino.name": "cosmos-sdk/group/MsgWithdrawProposal", "cosmos.msg.v1.signer": ["address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.address = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string address */
        2:
          message.address = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.address !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.address);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgWithdrawProposal = new MsgWithdrawProposal$Type();
class MsgWithdrawProposalResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgWithdrawProposalResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgWithdrawProposalResponse = new MsgWithdrawProposalResponse$Type();
class MsgVote$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgVote", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "voter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "option", kind: "enum", T: () => ["cosmos.group.v1.VoteOption", VoteOption, "VOTE_OPTION_"] },
      {
        no: 4,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 5, name: "exec", kind: "enum", T: () => ["cosmos.group.v1.Exec", Exec, "EXEC_"] }
    ], { "amino.name": "cosmos-sdk/group/MsgVote", "cosmos.msg.v1.signer": ["voter"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.voter = "";
    message.option = 0;
    message.metadata = "";
    message.exec = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string voter */
        2:
          message.voter = reader.string();
          break;
        case /* cosmos.group.v1.VoteOption option */
        3:
          message.option = reader.int32();
          break;
        case /* string metadata */
        4:
          message.metadata = reader.string();
          break;
        case /* cosmos.group.v1.Exec exec */
        5:
          message.exec = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.voter !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.voter);
    if (message.option !== 0)
      writer.tag(3, WireType.Varint).int32(message.option);
    if (message.metadata !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.metadata);
    if (message.exec !== 0)
      writer.tag(5, WireType.Varint).int32(message.exec);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgVote = new MsgVote$Type();
class MsgVoteResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgVoteResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgVoteResponse = new MsgVoteResponse$Type();
class MsgExec$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgExec", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "executor", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "amino.name": "cosmos-sdk/group/MsgExec", "cosmos.msg.v1.signer": ["executor"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.executor = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string executor */
        2:
          message.executor = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.executor !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.executor);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgExec = new MsgExec$Type();
class MsgExecResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgExecResponse", [
      { no: 2, name: "result", kind: "enum", T: () => ["cosmos.group.v1.ProposalExecutorResult", ProposalExecutorResult, "PROPOSAL_EXECUTOR_RESULT_"] }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.result = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.group.v1.ProposalExecutorResult result */
        2:
          message.result = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.result !== 0)
      writer.tag(2, WireType.Varint).int32(message.result);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgExecResponse = new MsgExecResponse$Type();
class MsgLeaveGroup$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgLeaveGroup", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "group_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "amino.name": "cosmos-sdk/group/MsgLeaveGroup", "cosmos.msg.v1.signer": ["address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.groupId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* uint64 group_id */
        2:
          message.groupId = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.groupId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.groupId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgLeaveGroup = new MsgLeaveGroup$Type();
class MsgLeaveGroupResponse$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MsgLeaveGroupResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgLeaveGroupResponse = new MsgLeaveGroupResponse$Type();
const Msg = new ServiceType("cosmos.group.v1.Msg", [
  { name: "CreateGroup", options: {}, I: MsgCreateGroup, O: MsgCreateGroupResponse },
  { name: "UpdateGroupMembers", options: {}, I: MsgUpdateGroupMembers, O: MsgUpdateGroupMembersResponse },
  { name: "UpdateGroupAdmin", options: {}, I: MsgUpdateGroupAdmin, O: MsgUpdateGroupAdminResponse },
  { name: "UpdateGroupMetadata", options: {}, I: MsgUpdateGroupMetadata, O: MsgUpdateGroupMetadataResponse },
  { name: "CreateGroupPolicy", options: {}, I: MsgCreateGroupPolicy, O: MsgCreateGroupPolicyResponse },
  { name: "CreateGroupWithPolicy", options: {}, I: MsgCreateGroupWithPolicy, O: MsgCreateGroupWithPolicyResponse },
  { name: "UpdateGroupPolicyAdmin", options: {}, I: MsgUpdateGroupPolicyAdmin, O: MsgUpdateGroupPolicyAdminResponse },
  { name: "UpdateGroupPolicyDecisionPolicy", options: {}, I: MsgUpdateGroupPolicyDecisionPolicy, O: MsgUpdateGroupPolicyDecisionPolicyResponse },
  { name: "UpdateGroupPolicyMetadata", options: {}, I: MsgUpdateGroupPolicyMetadata, O: MsgUpdateGroupPolicyMetadataResponse },
  { name: "SubmitProposal", options: {}, I: MsgSubmitProposal, O: MsgSubmitProposalResponse },
  { name: "WithdrawProposal", options: {}, I: MsgWithdrawProposal, O: MsgWithdrawProposalResponse },
  { name: "Vote", options: {}, I: MsgVote, O: MsgVoteResponse },
  { name: "Exec", options: {}, I: MsgExec, O: MsgExecResponse },
  { name: "LeaveGroup", options: {}, I: MsgLeaveGroup, O: MsgLeaveGroupResponse }
], { "cosmos.msg.v1.service": true });
export {
  Exec,
  Msg,
  MsgCreateGroup,
  MsgCreateGroupPolicy,
  MsgCreateGroupPolicyResponse,
  MsgCreateGroupResponse,
  MsgCreateGroupWithPolicy,
  MsgCreateGroupWithPolicyResponse,
  MsgExec,
  MsgExecResponse,
  MsgLeaveGroup,
  MsgLeaveGroupResponse,
  MsgSubmitProposal,
  MsgSubmitProposalResponse,
  MsgUpdateGroupAdmin,
  MsgUpdateGroupAdminResponse,
  MsgUpdateGroupMembers,
  MsgUpdateGroupMembersResponse,
  MsgUpdateGroupMetadata,
  MsgUpdateGroupMetadataResponse,
  MsgUpdateGroupPolicyAdmin,
  MsgUpdateGroupPolicyAdminResponse,
  MsgUpdateGroupPolicyDecisionPolicy,
  MsgUpdateGroupPolicyDecisionPolicyResponse,
  MsgUpdateGroupPolicyMetadata,
  MsgUpdateGroupPolicyMetadataResponse,
  MsgVote,
  MsgVoteResponse,
  MsgWithdrawProposal,
  MsgWithdrawProposalResponse
};
