// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/group/v1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * CreateGroup creates a new group with an admin account address, a list of members and some optional metadata.
   *
   * @generated from protobuf rpc: CreateGroup
   */
  createGroup(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateGroupMembers updates the group members with given group id and admin address.
   *
   * @generated from protobuf rpc: UpdateGroupMembers
   */
  updateGroupMembers(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateGroupAdmin updates the group admin with given group id and previous admin address.
   *
   * @generated from protobuf rpc: UpdateGroupAdmin
   */
  updateGroupAdmin(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateGroupMetadata updates the group metadata with given group id and admin address.
   *
   * @generated from protobuf rpc: UpdateGroupMetadata
   */
  updateGroupMetadata(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * CreateGroupPolicy creates a new group policy using given DecisionPolicy.
   *
   * @generated from protobuf rpc: CreateGroupPolicy
   */
  createGroupPolicy(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * CreateGroupWithPolicy creates a new group with policy.
   *
   * @generated from protobuf rpc: CreateGroupWithPolicy
   */
  createGroupWithPolicy(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateGroupPolicyAdmin updates a group policy admin.
   *
   * @generated from protobuf rpc: UpdateGroupPolicyAdmin
   */
  updateGroupPolicyAdmin(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateGroupPolicyDecisionPolicy allows a group policy's decision policy to be updated.
   *
   * @generated from protobuf rpc: UpdateGroupPolicyDecisionPolicy
   */
  updateGroupPolicyDecisionPolicy(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateGroupPolicyMetadata updates a group policy metadata.
   *
   * @generated from protobuf rpc: UpdateGroupPolicyMetadata
   */
  updateGroupPolicyMetadata(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * SubmitProposal submits a new proposal.
   *
   * @generated from protobuf rpc: SubmitProposal
   */
  submitProposal(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * WithdrawProposal withdraws a proposal.
   *
   * @generated from protobuf rpc: WithdrawProposal
   */
  withdrawProposal(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Vote allows a voter to vote on a proposal.
   *
   * @generated from protobuf rpc: Vote
   */
  vote(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Exec executes a proposal.
   *
   * @generated from protobuf rpc: Exec
   */
  exec(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * LeaveGroup allows a group member to leave the group.
   *
   * @generated from protobuf rpc: LeaveGroup
   */
  leaveGroup(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
