// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/group/v1/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Any } from "../../../google/protobuf/any_pb.js";
import { Duration } from "../../../google/protobuf/duration_pb.js";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
var VoteOption = /* @__PURE__ */ ((VoteOption2) => {
  VoteOption2[VoteOption2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  VoteOption2[VoteOption2["YES"] = 1] = "YES";
  VoteOption2[VoteOption2["ABSTAIN"] = 2] = "ABSTAIN";
  VoteOption2[VoteOption2["NO"] = 3] = "NO";
  VoteOption2[VoteOption2["NO_WITH_VETO"] = 4] = "NO_WITH_VETO";
  return VoteOption2;
})(VoteOption || {});
var ProposalStatus = /* @__PURE__ */ ((ProposalStatus2) => {
  ProposalStatus2[ProposalStatus2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  ProposalStatus2[ProposalStatus2["SUBMITTED"] = 1] = "SUBMITTED";
  ProposalStatus2[ProposalStatus2["ACCEPTED"] = 2] = "ACCEPTED";
  ProposalStatus2[ProposalStatus2["REJECTED"] = 3] = "REJECTED";
  ProposalStatus2[ProposalStatus2["ABORTED"] = 4] = "ABORTED";
  ProposalStatus2[ProposalStatus2["WITHDRAWN"] = 5] = "WITHDRAWN";
  return ProposalStatus2;
})(ProposalStatus || {});
var ProposalExecutorResult = /* @__PURE__ */ ((ProposalExecutorResult2) => {
  ProposalExecutorResult2[ProposalExecutorResult2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  ProposalExecutorResult2[ProposalExecutorResult2["NOT_RUN"] = 1] = "NOT_RUN";
  ProposalExecutorResult2[ProposalExecutorResult2["SUCCESS"] = 2] = "SUCCESS";
  ProposalExecutorResult2[ProposalExecutorResult2["FAILURE"] = 3] = "FAILURE";
  return ProposalExecutorResult2;
})(ProposalExecutorResult || {});
class Member$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.Member", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "weight",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "added_at", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.weight = "";
    message.metadata = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* string weight */
        2:
          message.weight = reader.string();
          break;
        case /* string metadata */
        3:
          message.metadata = reader.string();
          break;
        case /* google.protobuf.Timestamp added_at */
        4:
          message.addedAt = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.addedAt);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.weight !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.weight);
    if (message.metadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.metadata);
    if (message.addedAt)
      Timestamp.internalBinaryWrite(message.addedAt, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Member = new Member$Type();
class MemberRequest$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.MemberRequest", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "weight",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.weight = "";
    message.metadata = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* string weight */
        2:
          message.weight = reader.string();
          break;
        case /* string metadata */
        3:
          message.metadata = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.weight !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.weight);
    if (message.metadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.metadata);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MemberRequest = new MemberRequest$Type();
class ThresholdDecisionPolicy$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.ThresholdDecisionPolicy", [
      {
        no: 1,
        name: "threshold",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "windows", kind: "message", T: () => DecisionPolicyWindows }
    ], { "cosmos_proto.implements_interface": ["cosmos.group.v1.DecisionPolicy"], "amino.name": "cosmos-sdk/ThresholdDecisionPolicy" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.threshold = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string threshold */
        1:
          message.threshold = reader.string();
          break;
        case /* cosmos.group.v1.DecisionPolicyWindows windows */
        2:
          message.windows = DecisionPolicyWindows.internalBinaryRead(reader, reader.uint32(), options, message.windows);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.threshold !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.threshold);
    if (message.windows)
      DecisionPolicyWindows.internalBinaryWrite(message.windows, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ThresholdDecisionPolicy = new ThresholdDecisionPolicy$Type();
class PercentageDecisionPolicy$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.PercentageDecisionPolicy", [
      {
        no: 1,
        name: "percentage",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "windows", kind: "message", T: () => DecisionPolicyWindows }
    ], { "cosmos_proto.implements_interface": ["cosmos.group.v1.DecisionPolicy"], "amino.name": "cosmos-sdk/PercentageDecisionPolicy" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.percentage = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string percentage */
        1:
          message.percentage = reader.string();
          break;
        case /* cosmos.group.v1.DecisionPolicyWindows windows */
        2:
          message.windows = DecisionPolicyWindows.internalBinaryRead(reader, reader.uint32(), options, message.windows);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.percentage !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.percentage);
    if (message.windows)
      DecisionPolicyWindows.internalBinaryWrite(message.windows, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PercentageDecisionPolicy = new PercentageDecisionPolicy$Type();
class DecisionPolicyWindows$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.DecisionPolicyWindows", [
      { no: 1, name: "voting_period", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true, "amino.dont_omitempty": true } },
      { no: 2, name: "min_execution_period", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Duration voting_period */
        1:
          message.votingPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.votingPeriod);
          break;
        case /* google.protobuf.Duration min_execution_period */
        2:
          message.minExecutionPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.minExecutionPeriod);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.votingPeriod)
      Duration.internalBinaryWrite(message.votingPeriod, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.minExecutionPeriod)
      Duration.internalBinaryWrite(message.minExecutionPeriod, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DecisionPolicyWindows = new DecisionPolicyWindows$Type();
class GroupInfo$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.GroupInfo", [
      {
        no: 1,
        name: "id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 3,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "version",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "total_weight",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 6, name: "created_at", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.id = 0n;
    message.admin = "";
    message.metadata = "";
    message.version = 0n;
    message.totalWeight = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 id */
        1:
          message.id = reader.uint64().toBigInt();
          break;
        case /* string admin */
        2:
          message.admin = reader.string();
          break;
        case /* string metadata */
        3:
          message.metadata = reader.string();
          break;
        case /* uint64 version */
        4:
          message.version = reader.uint64().toBigInt();
          break;
        case /* string total_weight */
        5:
          message.totalWeight = reader.string();
          break;
        case /* google.protobuf.Timestamp created_at */
        6:
          message.createdAt = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.createdAt);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.id !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.id);
    if (message.admin !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.admin);
    if (message.metadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.metadata);
    if (message.version !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.version);
    if (message.totalWeight !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.totalWeight);
    if (message.createdAt)
      Timestamp.internalBinaryWrite(message.createdAt, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GroupInfo = new GroupInfo$Type();
class GroupMember$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.GroupMember", [
      {
        no: 1,
        name: "group_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "member", kind: "message", T: () => Member }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.groupId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 group_id */
        1:
          message.groupId = reader.uint64().toBigInt();
          break;
        case /* cosmos.group.v1.Member member */
        2:
          message.member = Member.internalBinaryRead(reader, reader.uint32(), options, message.member);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.groupId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.groupId);
    if (message.member)
      Member.internalBinaryWrite(message.member, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GroupMember = new GroupMember$Type();
class GroupPolicyInfo$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.GroupPolicyInfo", [
      { no: 1, name: "address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 2,
        name: "group_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "admin", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 4,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "version",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 6, name: "decision_policy", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.group.v1.DecisionPolicy" } },
      { no: 7, name: "created_at", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.address = "";
    message.groupId = 0n;
    message.admin = "";
    message.metadata = "";
    message.version = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string address */
        1:
          message.address = reader.string();
          break;
        case /* uint64 group_id */
        2:
          message.groupId = reader.uint64().toBigInt();
          break;
        case /* string admin */
        3:
          message.admin = reader.string();
          break;
        case /* string metadata */
        4:
          message.metadata = reader.string();
          break;
        case /* uint64 version */
        5:
          message.version = reader.uint64().toBigInt();
          break;
        case /* google.protobuf.Any decision_policy */
        6:
          message.decisionPolicy = Any.internalBinaryRead(reader, reader.uint32(), options, message.decisionPolicy);
          break;
        case /* google.protobuf.Timestamp created_at */
        7:
          message.createdAt = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.createdAt);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.address !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.address);
    if (message.groupId !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.groupId);
    if (message.admin !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.admin);
    if (message.metadata !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.metadata);
    if (message.version !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.version);
    if (message.decisionPolicy)
      Any.internalBinaryWrite(message.decisionPolicy, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.createdAt)
      Timestamp.internalBinaryWrite(message.createdAt, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GroupPolicyInfo = new GroupPolicyInfo$Type();
class Proposal$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.Proposal", [
      {
        no: 1,
        name: "id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "group_policy_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      {
        no: 3,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "proposers", kind: "scalar", repeat: 2, T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 5, name: "submit_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      {
        no: 6,
        name: "group_version",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "group_policy_version",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 8, name: "status", kind: "enum", T: () => ["cosmos.group.v1.ProposalStatus", ProposalStatus, "PROPOSAL_STATUS_"] },
      { no: 9, name: "final_tally_result", kind: "message", T: () => TallyResult, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 10, name: "voting_period_end", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      { no: 11, name: "executor_result", kind: "enum", T: () => ["cosmos.group.v1.ProposalExecutorResult", ProposalExecutorResult, "PROPOSAL_EXECUTOR_RESULT_"] },
      { no: 12, name: "messages", kind: "message", repeat: 2, T: () => Any },
      {
        no: 13,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 14,
        name: "summary",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.id = 0n;
    message.groupPolicyAddress = "";
    message.metadata = "";
    message.proposers = [];
    message.groupVersion = 0n;
    message.groupPolicyVersion = 0n;
    message.status = 0;
    message.executorResult = 0;
    message.messages = [];
    message.title = "";
    message.summary = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 id */
        1:
          message.id = reader.uint64().toBigInt();
          break;
        case /* string group_policy_address */
        2:
          message.groupPolicyAddress = reader.string();
          break;
        case /* string metadata */
        3:
          message.metadata = reader.string();
          break;
        case /* repeated string proposers */
        4:
          message.proposers.push(reader.string());
          break;
        case /* google.protobuf.Timestamp submit_time */
        5:
          message.submitTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.submitTime);
          break;
        case /* uint64 group_version */
        6:
          message.groupVersion = reader.uint64().toBigInt();
          break;
        case /* uint64 group_policy_version */
        7:
          message.groupPolicyVersion = reader.uint64().toBigInt();
          break;
        case /* cosmos.group.v1.ProposalStatus status */
        8:
          message.status = reader.int32();
          break;
        case /* cosmos.group.v1.TallyResult final_tally_result */
        9:
          message.finalTallyResult = TallyResult.internalBinaryRead(reader, reader.uint32(), options, message.finalTallyResult);
          break;
        case /* google.protobuf.Timestamp voting_period_end */
        10:
          message.votingPeriodEnd = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.votingPeriodEnd);
          break;
        case /* cosmos.group.v1.ProposalExecutorResult executor_result */
        11:
          message.executorResult = reader.int32();
          break;
        case /* repeated google.protobuf.Any messages */
        12:
          message.messages.push(Any.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string title */
        13:
          message.title = reader.string();
          break;
        case /* string summary */
        14:
          message.summary = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.id !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.id);
    if (message.groupPolicyAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.groupPolicyAddress);
    if (message.metadata !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.metadata);
    for (let i = 0; i < message.proposers.length; i++)
      writer.tag(4, WireType.LengthDelimited).string(message.proposers[i]);
    if (message.submitTime)
      Timestamp.internalBinaryWrite(message.submitTime, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.groupVersion !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.groupVersion);
    if (message.groupPolicyVersion !== 0n)
      writer.tag(7, WireType.Varint).uint64(message.groupPolicyVersion);
    if (message.status !== 0)
      writer.tag(8, WireType.Varint).int32(message.status);
    if (message.finalTallyResult)
      TallyResult.internalBinaryWrite(message.finalTallyResult, writer.tag(9, WireType.LengthDelimited).fork(), options).join();
    if (message.votingPeriodEnd)
      Timestamp.internalBinaryWrite(message.votingPeriodEnd, writer.tag(10, WireType.LengthDelimited).fork(), options).join();
    if (message.executorResult !== 0)
      writer.tag(11, WireType.Varint).int32(message.executorResult);
    for (let i = 0; i < message.messages.length; i++)
      Any.internalBinaryWrite(message.messages[i], writer.tag(12, WireType.LengthDelimited).fork(), options).join();
    if (message.title !== "")
      writer.tag(13, WireType.LengthDelimited).string(message.title);
    if (message.summary !== "")
      writer.tag(14, WireType.LengthDelimited).string(message.summary);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Proposal = new Proposal$Type();
class TallyResult$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.TallyResult", [
      {
        no: 1,
        name: "yes_count",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "abstain_count",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "no_count",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "no_with_veto_count",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.yesCount = "";
    message.abstainCount = "";
    message.noCount = "";
    message.noWithVetoCount = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string yes_count */
        1:
          message.yesCount = reader.string();
          break;
        case /* string abstain_count */
        2:
          message.abstainCount = reader.string();
          break;
        case /* string no_count */
        3:
          message.noCount = reader.string();
          break;
        case /* string no_with_veto_count */
        4:
          message.noWithVetoCount = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.yesCount !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.yesCount);
    if (message.abstainCount !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.abstainCount);
    if (message.noCount !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.noCount);
    if (message.noWithVetoCount !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.noWithVetoCount);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TallyResult = new TallyResult$Type();
class Vote$Type extends MessageType {
  constructor() {
    super("cosmos.group.v1.Vote", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "voter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "option", kind: "enum", T: () => ["cosmos.group.v1.VoteOption", VoteOption, "VOTE_OPTION_"] },
      {
        no: 4,
        name: "metadata",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 5, name: "submit_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.voter = "";
    message.option = 0;
    message.metadata = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string voter */
        2:
          message.voter = reader.string();
          break;
        case /* cosmos.group.v1.VoteOption option */
        3:
          message.option = reader.int32();
          break;
        case /* string metadata */
        4:
          message.metadata = reader.string();
          break;
        case /* google.protobuf.Timestamp submit_time */
        5:
          message.submitTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.submitTime);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.voter !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.voter);
    if (message.option !== 0)
      writer.tag(3, WireType.Varint).int32(message.option);
    if (message.metadata !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.metadata);
    if (message.submitTime)
      Timestamp.internalBinaryWrite(message.submitTime, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Vote = new Vote$Type();
export {
  DecisionPolicyWindows,
  GroupInfo,
  GroupMember,
  GroupPolicyInfo,
  Member,
  MemberRequest,
  PercentageDecisionPolicy,
  Proposal,
  ProposalExecutorResult,
  ProposalStatus,
  TallyResult,
  ThresholdDecisionPolicy,
  Vote,
  VoteOption
};
