// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/gov/v1beta1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { TallyParams } from "./gov_pb.js";
import { VotingParams } from "./gov_pb.js";
import { DepositParams } from "./gov_pb.js";
import { Proposal } from "./gov_pb.js";
import { Vote } from "./gov_pb.js";
import { Deposit } from "./gov_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.GenesisState", [
      {
        no: 1,
        name: "starting_proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "deposits", kind: "message", repeat: 2, T: () => Deposit, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "Deposits", "amino.dont_omitempty": true } },
      { no: 3, name: "votes", kind: "message", repeat: 2, T: () => Vote, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "Votes", "amino.dont_omitempty": true } },
      { no: 4, name: "proposals", kind: "message", repeat: 2, T: () => Proposal, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "Proposals", "amino.dont_omitempty": true } },
      { no: 5, name: "deposit_params", kind: "message", T: () => DepositParams, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 6, name: "voting_params", kind: "message", T: () => VotingParams, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 7, name: "tally_params", kind: "message", T: () => TallyParams, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.startingProposalId = 0n;
    message.deposits = [];
    message.votes = [];
    message.proposals = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 starting_proposal_id */
        1:
          message.startingProposalId = reader.uint64().toBigInt();
          break;
        case /* repeated cosmos.gov.v1beta1.Deposit deposits */
        2:
          message.deposits.push(Deposit.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.gov.v1beta1.Vote votes */
        3:
          message.votes.push(Vote.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.gov.v1beta1.Proposal proposals */
        4:
          message.proposals.push(Proposal.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.gov.v1beta1.DepositParams deposit_params */
        5:
          message.depositParams = DepositParams.internalBinaryRead(reader, reader.uint32(), options, message.depositParams);
          break;
        case /* cosmos.gov.v1beta1.VotingParams voting_params */
        6:
          message.votingParams = VotingParams.internalBinaryRead(reader, reader.uint32(), options, message.votingParams);
          break;
        case /* cosmos.gov.v1beta1.TallyParams tally_params */
        7:
          message.tallyParams = TallyParams.internalBinaryRead(reader, reader.uint32(), options, message.tallyParams);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.startingProposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.startingProposalId);
    for (let i = 0; i < message.deposits.length; i++)
      Deposit.internalBinaryWrite(message.deposits[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.votes.length; i++)
      Vote.internalBinaryWrite(message.votes[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.proposals.length; i++)
      Proposal.internalBinaryWrite(message.proposals[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.depositParams)
      DepositParams.internalBinaryWrite(message.depositParams, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.votingParams)
      VotingParams.internalBinaryWrite(message.votingParams, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.tallyParams)
      TallyParams.internalBinaryWrite(message.tallyParams, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
export {
  GenesisState
};
