// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/gov/v1beta1/gov_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Duration } from "../../../google/protobuf/duration_pb.js";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
import { Coin } from "../../base/v1beta1/coin_pb.js";
var VoteOption = /* @__PURE__ */ ((VoteOption2) => {
  VoteOption2[VoteOption2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  VoteOption2[VoteOption2["YES"] = 1] = "YES";
  VoteOption2[VoteOption2["ABSTAIN"] = 2] = "ABSTAIN";
  VoteOption2[VoteOption2["NO"] = 3] = "NO";
  VoteOption2[VoteOption2["NO_WITH_VETO"] = 4] = "NO_WITH_VETO";
  return VoteOption2;
})(VoteOption || {});
var ProposalStatus = /* @__PURE__ */ ((ProposalStatus2) => {
  ProposalStatus2[ProposalStatus2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  ProposalStatus2[ProposalStatus2["DEPOSIT_PERIOD"] = 1] = "DEPOSIT_PERIOD";
  ProposalStatus2[ProposalStatus2["VOTING_PERIOD"] = 2] = "VOTING_PERIOD";
  ProposalStatus2[ProposalStatus2["PASSED"] = 3] = "PASSED";
  ProposalStatus2[ProposalStatus2["REJECTED"] = 4] = "REJECTED";
  ProposalStatus2[ProposalStatus2["FAILED"] = 5] = "FAILED";
  return ProposalStatus2;
})(ProposalStatus || {});
class WeightedVoteOption$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.WeightedVoteOption", [
      { no: 1, name: "option", kind: "enum", T: () => ["cosmos.gov.v1beta1.VoteOption", VoteOption, "VOTE_OPTION_"] },
      { no: 2, name: "weight", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "cosmos_proto.scalar": "cosmos.Dec", "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.option = 0;
    message.weight = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.gov.v1beta1.VoteOption option */
        1:
          message.option = reader.int32();
          break;
        case /* string weight */
        2:
          message.weight = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.option !== 0)
      writer.tag(1, WireType.Varint).int32(message.option);
    if (message.weight !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.weight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const WeightedVoteOption = new WeightedVoteOption$Type();
class TextProposal$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.TextProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.equal": true, "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "cosmos-sdk/TextProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TextProposal = new TextProposal$Type();
class Deposit$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.Deposit", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "depositor", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "amount", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.depositor = "";
    message.amount = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string depositor */
        2:
          message.depositor = reader.string();
          break;
        case /* repeated cosmos.base.v1beta1.Coin amount */
        3:
          message.amount.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.depositor !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.depositor);
    for (let i = 0; i < message.amount.length; i++)
      Coin.internalBinaryWrite(message.amount[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Deposit = new Deposit$Type();
class Proposal$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.Proposal", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "content", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.gov.v1beta1.Content" } },
      { no: 3, name: "status", kind: "enum", T: () => ["cosmos.gov.v1beta1.ProposalStatus", ProposalStatus, "PROPOSAL_STATUS_"] },
      { no: 4, name: "final_tally_result", kind: "message", T: () => TallyResult, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } },
      { no: 5, name: "submit_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      { no: 6, name: "deposit_end_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      { no: 7, name: "total_deposit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.dont_omitempty": true } },
      { no: 8, name: "voting_start_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } },
      { no: 9, name: "voting_end_time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true, "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.status = 0;
    message.totalDeposit = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* google.protobuf.Any content */
        2:
          message.content = Any.internalBinaryRead(reader, reader.uint32(), options, message.content);
          break;
        case /* cosmos.gov.v1beta1.ProposalStatus status */
        3:
          message.status = reader.int32();
          break;
        case /* cosmos.gov.v1beta1.TallyResult final_tally_result */
        4:
          message.finalTallyResult = TallyResult.internalBinaryRead(reader, reader.uint32(), options, message.finalTallyResult);
          break;
        case /* google.protobuf.Timestamp submit_time */
        5:
          message.submitTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.submitTime);
          break;
        case /* google.protobuf.Timestamp deposit_end_time */
        6:
          message.depositEndTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.depositEndTime);
          break;
        case /* repeated cosmos.base.v1beta1.Coin total_deposit */
        7:
          message.totalDeposit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* google.protobuf.Timestamp voting_start_time */
        8:
          message.votingStartTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.votingStartTime);
          break;
        case /* google.protobuf.Timestamp voting_end_time */
        9:
          message.votingEndTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.votingEndTime);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.content)
      Any.internalBinaryWrite(message.content, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.status !== 0)
      writer.tag(3, WireType.Varint).int32(message.status);
    if (message.finalTallyResult)
      TallyResult.internalBinaryWrite(message.finalTallyResult, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.submitTime)
      Timestamp.internalBinaryWrite(message.submitTime, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.depositEndTime)
      Timestamp.internalBinaryWrite(message.depositEndTime, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.totalDeposit.length; i++)
      Coin.internalBinaryWrite(message.totalDeposit[i], writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.votingStartTime)
      Timestamp.internalBinaryWrite(message.votingStartTime, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    if (message.votingEndTime)
      Timestamp.internalBinaryWrite(message.votingEndTime, writer.tag(9, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Proposal = new Proposal$Type();
class TallyResult$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.TallyResult", [
      { no: 1, name: "yes", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } },
      { no: 2, name: "abstain", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } },
      { no: 3, name: "no", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } },
      { no: 4, name: "no_with_veto", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int", "cosmos_proto.scalar": "cosmos.Int" } }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.yes = "";
    message.abstain = "";
    message.no = "";
    message.noWithVeto = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string yes */
        1:
          message.yes = reader.string();
          break;
        case /* string abstain */
        2:
          message.abstain = reader.string();
          break;
        case /* string no */
        3:
          message.no = reader.string();
          break;
        case /* string no_with_veto */
        4:
          message.noWithVeto = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.yes !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.yes);
    if (message.abstain !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.abstain);
    if (message.no !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.no);
    if (message.noWithVeto !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.noWithVeto);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TallyResult = new TallyResult$Type();
class Vote$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.Vote", [
      { no: 1, name: "proposal_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.jsontag": "id", "amino.field_name": "id", "amino.dont_omitempty": true } },
      { no: 2, name: "voter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "option", kind: "enum", T: () => ["cosmos.gov.v1beta1.VoteOption", VoteOption, "VOTE_OPTION_"] },
      { no: 4, name: "options", kind: "message", repeat: 2, T: () => WeightedVoteOption, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.voter = "";
    message.option = 0;
    message.options = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string voter */
        2:
          message.voter = reader.string();
          break;
        case /* cosmos.gov.v1beta1.VoteOption option = 3 [deprecated = true] */
        3:
          message.option = reader.int32();
          break;
        case /* repeated cosmos.gov.v1beta1.WeightedVoteOption options */
        4:
          message.options.push(WeightedVoteOption.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.voter !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.voter);
    if (message.option !== 0)
      writer.tag(3, WireType.Varint).int32(message.option);
    for (let i = 0; i < message.options.length; i++)
      WeightedVoteOption.internalBinaryWrite(message.options[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Vote = new Vote$Type();
class DepositParams$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.DepositParams", [
      { no: 1, name: "min_deposit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "min_deposit,omitempty", "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins" } },
      { no: 2, name: "max_deposit_period", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "max_deposit_period,omitempty", "gogoproto.stdduration": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.minDeposit = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin min_deposit */
        1:
          message.minDeposit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* google.protobuf.Duration max_deposit_period */
        2:
          message.maxDepositPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.maxDepositPeriod);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.minDeposit.length; i++)
      Coin.internalBinaryWrite(message.minDeposit[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.maxDepositPeriod)
      Duration.internalBinaryWrite(message.maxDepositPeriod, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const DepositParams = new DepositParams$Type();
class VotingParams$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.VotingParams", [
      { no: 1, name: "voting_period", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.jsontag": "voting_period,omitempty", "gogoproto.stdduration": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Duration voting_period */
        1:
          message.votingPeriod = Duration.internalBinaryRead(reader, reader.uint32(), options, message.votingPeriod);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.votingPeriod)
      Duration.internalBinaryWrite(message.votingPeriod, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const VotingParams = new VotingParams$Type();
class TallyParams$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.TallyParams", [
      { no: 1, name: "quorum", kind: "scalar", T: 12, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "gogoproto.jsontag": "quorum,omitempty", "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 2, name: "threshold", kind: "scalar", T: 12, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "gogoproto.jsontag": "threshold,omitempty", "cosmos_proto.scalar": "cosmos.Dec" } },
      { no: 3, name: "veto_threshold", kind: "scalar", T: 12, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec", "gogoproto.jsontag": "veto_threshold,omitempty", "cosmos_proto.scalar": "cosmos.Dec" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.quorum = new Uint8Array(0);
    message.threshold = new Uint8Array(0);
    message.vetoThreshold = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes quorum */
        1:
          message.quorum = reader.bytes();
          break;
        case /* bytes threshold */
        2:
          message.threshold = reader.bytes();
          break;
        case /* bytes veto_threshold */
        3:
          message.vetoThreshold = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.quorum.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.quorum);
    if (message.threshold.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.threshold);
    if (message.vetoThreshold.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.vetoThreshold);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TallyParams = new TallyParams$Type();
export {
  Deposit,
  DepositParams,
  Proposal,
  ProposalStatus,
  TallyParams,
  TallyResult,
  TextProposal,
  Vote,
  VoteOption,
  VotingParams,
  WeightedVoteOption
};
