// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/gov/v1beta1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { WeightedVoteOption } from "./gov_pb.js";
import { VoteOption } from "./gov_pb.js";
import { Coin } from "../../base/v1beta1/coin_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
class MsgSubmitProposal$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.MsgSubmitProposal", [
      { no: 1, name: "content", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.gov.v1beta1.Content" } },
      { no: 2, name: "initial_deposit", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } },
      { no: 3, name: "proposer", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgSubmitProposal", "cosmos.msg.v1.signer": ["proposer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.initialDeposit = [];
    message.proposer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any content */
        1:
          message.content = Any.internalBinaryRead(reader, reader.uint32(), options, message.content);
          break;
        case /* repeated cosmos.base.v1beta1.Coin initial_deposit */
        2:
          message.initialDeposit.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* string proposer */
        3:
          message.proposer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.content)
      Any.internalBinaryWrite(message.content, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.initialDeposit.length; i++)
      Coin.internalBinaryWrite(message.initialDeposit[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.proposer !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.proposer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSubmitProposal = new MsgSubmitProposal$Type();
class MsgSubmitProposalResponse$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.MsgSubmitProposalResponse", [
      { no: 1, name: "proposal_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.jsontag": "proposal_id", "amino.dont_omitempty": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSubmitProposalResponse = new MsgSubmitProposalResponse$Type();
class MsgVote$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.MsgVote", [
      {
        no: 1,
        name: "proposal_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "voter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "option", kind: "enum", T: () => ["cosmos.gov.v1beta1.VoteOption", VoteOption, "VOTE_OPTION_"] }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgVote", "cosmos.msg.v1.signer": ["voter"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.voter = "";
    message.option = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string voter */
        2:
          message.voter = reader.string();
          break;
        case /* cosmos.gov.v1beta1.VoteOption option */
        3:
          message.option = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.voter !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.voter);
    if (message.option !== 0)
      writer.tag(3, WireType.Varint).int32(message.option);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgVote = new MsgVote$Type();
class MsgVoteResponse$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.MsgVoteResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgVoteResponse = new MsgVoteResponse$Type();
class MsgVoteWeighted$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.MsgVoteWeighted", [
      { no: 1, name: "proposal_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.jsontag": "proposal_id", "amino.dont_omitempty": true } },
      { no: 2, name: "voter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "options", kind: "message", repeat: 2, T: () => WeightedVoteOption, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgVoteWeighted", "cosmos.msg.v1.signer": ["voter"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.voter = "";
    message.options = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string voter */
        2:
          message.voter = reader.string();
          break;
        case /* repeated cosmos.gov.v1beta1.WeightedVoteOption options */
        3:
          message.options.push(WeightedVoteOption.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.voter !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.voter);
    for (let i = 0; i < message.options.length; i++)
      WeightedVoteOption.internalBinaryWrite(message.options[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgVoteWeighted = new MsgVoteWeighted$Type();
class MsgVoteWeightedResponse$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.MsgVoteWeightedResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgVoteWeightedResponse = new MsgVoteWeightedResponse$Type();
class MsgDeposit$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.MsgDeposit", [
      { no: 1, name: "proposal_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.jsontag": "proposal_id", "amino.dont_omitempty": true } },
      { no: 2, name: "depositor", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "amount", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.stringer": false, "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgDeposit", "cosmos.msg.v1.signer": ["depositor"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proposalId = 0n;
    message.depositor = "";
    message.amount = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 proposal_id */
        1:
          message.proposalId = reader.uint64().toBigInt();
          break;
        case /* string depositor */
        2:
          message.depositor = reader.string();
          break;
        case /* repeated cosmos.base.v1beta1.Coin amount */
        3:
          message.amount.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.proposalId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.proposalId);
    if (message.depositor !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.depositor);
    for (let i = 0; i < message.amount.length; i++)
      Coin.internalBinaryWrite(message.amount[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgDeposit = new MsgDeposit$Type();
class MsgDepositResponse$Type extends MessageType {
  constructor() {
    super("cosmos.gov.v1beta1.MsgDepositResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgDepositResponse = new MsgDepositResponse$Type();
const Msg = new ServiceType("cosmos.gov.v1beta1.Msg", [
  { name: "SubmitProposal", options: {}, I: MsgSubmitProposal, O: MsgSubmitProposalResponse },
  { name: "Vote", options: {}, I: MsgVote, O: MsgVoteResponse },
  { name: "VoteWeighted", options: {}, I: MsgVoteWeighted, O: MsgVoteWeightedResponse },
  { name: "Deposit", options: {}, I: MsgDeposit, O: MsgDepositResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgDeposit,
  MsgDepositResponse,
  MsgSubmitProposal,
  MsgSubmitProposalResponse,
  MsgVote,
  MsgVoteResponse,
  MsgVoteWeighted,
  MsgVoteWeightedResponse
};
