// @injectivelabs/sdk-ts/dist/cjs/AbacusGrpcApi-QbGB8ZMB.cjs
const require_chunk = require('./chunk-CbDLau6x.cjs');
const require_defineProperty = require('./defineProperty-dyr5BsvG.cjs');
const require_BaseGrpcConsumer = require('./BaseGrpcConsumer-Db3sCC6b.cjs');
let __injectivelabs_exceptions = require("@injectivelabs/exceptions");
let __injectivelabs_abacus_proto_ts_v2_generated_points_svc_pb = require("@injectivelabs/abacus-proto-ts-v2/generated/points_svc_pb");
__injectivelabs_abacus_proto_ts_v2_generated_points_svc_pb = require_chunk.__toESM(__injectivelabs_abacus_proto_ts_v2_generated_points_svc_pb);
let __injectivelabs_abacus_proto_ts_v2_generated_points_svc_pb_client = require("@injectivelabs/abacus-proto-ts-v2/generated/points_svc_pb.client");

//#region src/client/abacus/grpc/transformers/index.ts
var AbacusGrpcTransformer = class {
	static grpcPointsStatsDailyToPointsStatsDaily(response) {
		return response.rows.map((row) => ({
			day: row.day,
			week: row.week,
			volume: row.volume,
			points: row.points.toString(),
			pointsPrecise: row.pointsPrecise
		}));
	}
	static grpcPointsStatsWeeklyToPointsStatsWeekly(response) {
		return response.rows.map((row) => ({
			day: row.day,
			week: row.week,
			volume: row.volume,
			points: row.points.toString(),
			pointsPrecise: row.pointsPrecise
		}));
	}
	static grpcPointsLatestToPointsLatest(response) {
		return {
			league: response.league,
			updatedAt: response.updatedAt,
			rank: response.rank.toString(),
			pointsBonus: response.pointsBonus,
			pointsSeason1: response.pointsSeason1,
			totalPoints: response.totalPoints.toString(),
			totalPointsPrecise: response.totalPointsPrecise
		};
	}
};

//#endregion
//#region src/client/abacus/grpc/AbacusGrpcApi.ts
var AbacusGrpcApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", __injectivelabs_exceptions.IndexerErrorModule.Abacus);
	}
	get client() {
		return this.initClient(__injectivelabs_abacus_proto_ts_v2_generated_points_svc_pb_client.PointsSvcClient);
	}
	async fetchAccountLatestPoints(address) {
		const request = __injectivelabs_abacus_proto_ts_v2_generated_points_svc_pb.PointsLatestForAccountRequest.create({ accountAddress: address });
		const response = await this.executeGrpcCall(request, this.client.pointsLatestForAccount.bind(this.client));
		return AbacusGrpcTransformer.grpcPointsLatestToPointsLatest(response);
	}
	async fetchAccountDailyPoints(address, daysLimit) {
		const request = __injectivelabs_abacus_proto_ts_v2_generated_points_svc_pb.PointsStatsDailyForAccountRequest.create({
			accountAddress: address,
			daysLimit: daysLimit ? BigInt(daysLimit) : void 0
		});
		const response = await this.executeGrpcCall(request, this.client.pointsStatsDailyForAccount.bind(this.client));
		return AbacusGrpcTransformer.grpcPointsStatsDailyToPointsStatsDaily(response);
	}
	async fetchAccountWeeklyPoints(address, weeksLimit) {
		const request = __injectivelabs_abacus_proto_ts_v2_generated_points_svc_pb.PointsStatsWeeklyForAccountRequest.create({
			accountAddress: address,
			weeksLimit: weeksLimit ? BigInt(weeksLimit) : void 0
		});
		const response = await this.executeGrpcCall(request, this.client.pointsStatsWeeklyForAccount.bind(this.client));
		return AbacusGrpcTransformer.grpcPointsStatsWeeklyToPointsStatsWeekly(response);
	}
};

//#endregion
Object.defineProperty(exports, 'AbacusGrpcApi', {
  enumerable: true,
  get: function () {
    return AbacusGrpcApi;
  }
});
