// @injectivelabs/sdk-ts/dist/cjs/ChainGrpcExchangeApi-Xtflr7xp.cjs
const require_chunk = require('./chunk-CbDLau6x.cjs');
const require_utils = require('./utils-D3XFGv8y.cjs');
const require_defineProperty = require('./defineProperty-dyr5BsvG.cjs');
const require_ChainRestTendermintApi = require('./ChainRestTendermintApi-CyeoAmer.cjs');
const require_BaseGrpcConsumer = require('./BaseGrpcConsumer-Db3sCC6b.cjs');
let __injectivelabs_utils = require("@injectivelabs/utils");
let __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/exchange/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/exchange/v1beta1/query_pb.client");

//#region src/client/chain/transformers/ChainGrpcExchangeTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcExchangeTransformer = class ChainGrpcExchangeTransformer {
	static moduleParamsResponseToParams(response) {
		const params = response.params;
		const spotMarketInstantListingFee = params.spotMarketInstantListingFee;
		const derivativeMarketInstantListingFee = params.derivativeMarketInstantListingFee;
		return {
			spotMarketInstantListingFee: spotMarketInstantListingFee ? {
				amount: spotMarketInstantListingFee.amount,
				denom: spotMarketInstantListingFee.denom
			} : void 0,
			derivativeMarketInstantListingFee: derivativeMarketInstantListingFee ? {
				amount: derivativeMarketInstantListingFee.amount,
				denom: derivativeMarketInstantListingFee.denom
			} : void 0,
			defaultSpotMakerFeeRate: params.defaultSpotMakerFeeRate,
			defaultSpotTakerFeeRate: params.defaultSpotTakerFeeRate,
			defaultDerivativeMakerFeeRate: params.defaultDerivativeMakerFeeRate,
			defaultDerivativeTakerFeeRate: params.defaultDerivativeTakerFeeRate,
			defaultInitialMarginRatio: params.defaultInitialMarginRatio,
			defaultMaintenanceMarginRatio: params.defaultMaintenanceMarginRatio,
			defaultFundingInterval: parseInt(params.defaultFundingInterval.toString(), 10),
			fundingMultiple: parseInt(params.fundingMultiple.toString(), 10),
			relayerFeeShareRate: params.relayerFeeShareRate,
			defaultHourlyFundingRateCap: params.defaultHourlyFundingRateCap,
			defaultHourlyInterestRate: params.defaultHourlyInterestRate,
			maxDerivativeOrderSideCount: params.maxDerivativeOrderSideCount,
			injRewardStakedRequirementThreshold: params.injRewardStakedRequirementThreshold,
			tradingRewardsVestingDuration: parseInt(params.tradingRewardsVestingDuration.toString(), 10),
			liquidatorRewardShareRate: params.liquidatorRewardShareRate,
			binaryOptionsMarketInstantListingFee: params.binaryOptionsMarketInstantListingFee ? {
				amount: params.binaryOptionsMarketInstantListingFee.amount,
				denom: params.binaryOptionsMarketInstantListingFee.denom
			} : void 0,
			atomicMarketOrderAccessLevel: params.atomicMarketOrderAccessLevel.toString(),
			spotAtomicMarketOrderFeeMultiplier: params.spotAtomicMarketOrderFeeMultiplier,
			derivativeAtomicMarketOrderFeeMultiplier: params.derivativeAtomicMarketOrderFeeMultiplier,
			binaryOptionsAtomicMarketOrderFeeMultiplier: params.binaryOptionsAtomicMarketOrderFeeMultiplier,
			minimalProtocolFeeRate: params.minimalProtocolFeeRate,
			isInstantDerivativeMarketLaunchEnabled: params.isInstantDerivativeMarketLaunchEnabled,
			postOnlyModeHeightThreshold: params.postOnlyModeHeightThreshold.toString(),
			marginDecreasePriceTimestampThresholdSeconds: params.marginDecreasePriceTimestampThresholdSeconds.toString(),
			exchangeAdmins: params.exchangeAdmins
		};
	}
	static feeDiscountScheduleResponseToFeeDiscountSchedule(response) {
		const schedule = response.feeDiscountSchedule;
		return {
			bucketCount: parseInt(schedule.bucketCount.toString(), 10),
			bucketDuration: parseInt(schedule.bucketDuration.toString(), 10),
			quoteDenomsList: schedule.quoteDenoms,
			tierInfosList: schedule.tierInfos.map(ChainGrpcExchangeTransformer.grpcFeeDiscountTierInfoToFeeDiscountTierInfo).filter((info) => info !== void 0),
			disqualifiedMarketIdsList: schedule.disqualifiedMarketIds
		};
	}
	static tradingRewardsCampaignResponseToTradingRewardsCampaign(response) {
		return {
			tradingRewardCampaignInfo: ChainGrpcExchangeTransformer.grpcTradingRewardCampaignInfoToTradingRewardCampaignInfo(response.tradingRewardCampaignInfo),
			tradingRewardPoolCampaignScheduleList: response.tradingRewardPoolCampaignSchedule.map(ChainGrpcExchangeTransformer.grpcCampaignRewardPoolToCampaignRewardPool),
			pendingTradingRewardPoolCampaignScheduleList: response.pendingTradingRewardPoolCampaignSchedule.map(ChainGrpcExchangeTransformer.grpcCampaignRewardPoolToCampaignRewardPool),
			totalTradeRewardPoints: response.totalTradeRewardPoints,
			pendingTotalTradeRewardPointsList: response.pendingTotalTradeRewardPoints
		};
	}
	static feeDiscountAccountInfoResponseToFeeDiscountAccountInfo(response) {
		return {
			tierLevel: Number(response.tierLevel),
			accountInfo: ChainGrpcExchangeTransformer.grpcFeeDiscountTierInfoToFeeDiscountTierInfo(response.accountInfo),
			accountTtl: ChainGrpcExchangeTransformer.grpcFeeDiscountTierTTLToFeeDiscountTierTTL(response.accountTtl)
		};
	}
	static grpcFeeDiscountTierInfoToFeeDiscountTierInfo(info) {
		if (!info) return;
		return {
			makerDiscountRate: info.makerDiscountRate,
			takerDiscountRate: info.takerDiscountRate,
			stakedAmount: info.stakedAmount,
			volume: info.volume == void 0 ? "0" : info.volume
		};
	}
	static grpcFeeDiscountTierTTLToFeeDiscountTierTTL(info) {
		if (!info) return;
		return {
			tier: parseInt(info.tier.toString(), 10),
			ttlTimestamp: parseInt(info.ttlTimestamp.toString(), 10)
		};
	}
	static grpcPointsMultiplierToPointsMultiplier(point) {
		return {
			makerPointsMultiplier: point.makerPointsMultiplier,
			takerPointsMultiplier: point.takerPointsMultiplier
		};
	}
	static grpcTradingRewardCampaignBoostInfoToTradingRewardCampaignBoostInfo(info) {
		if (!info) return;
		return {
			boostedSpotMarketIdsList: info.boostedSpotMarketIds,
			boostedDerivativeMarketIdsList: info.boostedDerivativeMarketIds,
			spotMarketMultipliersList: info.spotMarketMultipliers.map(ChainGrpcExchangeTransformer.grpcPointsMultiplierToPointsMultiplier),
			derivativeMarketMultipliersList: info.derivativeMarketMultipliers.map(ChainGrpcExchangeTransformer.grpcPointsMultiplierToPointsMultiplier)
		};
	}
	static grpcTradingRewardCampaignInfoToTradingRewardCampaignInfo(info) {
		if (!info) return;
		return {
			campaignDurationSeconds: parseInt(info.campaignDurationSeconds.toString(), 10),
			quoteDenomsList: info.quoteDenoms,
			tradingRewardBoostInfo: ChainGrpcExchangeTransformer.grpcTradingRewardCampaignBoostInfoToTradingRewardCampaignBoostInfo(info.tradingRewardBoostInfo),
			disqualifiedMarketIdsList: info.disqualifiedMarketIds
		};
	}
	static grpcCampaignRewardPoolToCampaignRewardPool(pool) {
		return {
			startTimestamp: parseInt(pool.startTimestamp.toString(), 10),
			maxCampaignRewardsList: pool.maxCampaignRewards.map((coin) => ({
				amount: coin.amount,
				denom: coin.denom
			}))
		};
	}
	static grpcPositionToPosition(position) {
		return {
			islong: position.isLong,
			...position
		};
	}
	static positionsResponseToPositions(response) {
		return response.state.map((position) => {
			return {
				subaccountId: position.subaccountId,
				marketId: position.marketId,
				position: ChainGrpcExchangeTransformer.grpcPositionToPosition(position.position)
			};
		});
	}
	static isOptedOutOfRewardsResponseToIsOptedOutOfRewards(response) {
		return { isOptedOut: response.isOptedOut };
	}
	static activeStakeGrantResponseToActiveStakeGrant(response) {
		return {
			grant: response.grant,
			effectiveGrant: response.effectiveGrant
		};
	}
	static denomMinNotionalResponseToDenomMinNotional(response) {
		return response.amount;
	}
	static denomDecimalsResponseToDenomDecimals(response) {
		return response.denomDecimals.map((denomDecimals) => ({
			denom: denomDecimals.denom,
			decimals: denomDecimals.decimals.toString()
		}));
	}
	static denomMinNotionalsResponseToDenomMinNotionals(response) {
		return response.denomMinNotionals.map((denomDecimals) => ({
			denom: denomDecimals.denom,
			minNotional: (0, __injectivelabs_utils.toHumanReadable)(denomDecimals.minNotional).toFixed()
		}));
	}
	static spotMarketsResponseToSpotMarkets(response) {
		return response.markets.map((market) => {
			return ChainGrpcExchangeTransformer.grpcSpotMarketToSpotMarket(market);
		});
	}
	static grpcSpotMarketToSpotMarket(market) {
		const marketInfo = market;
		return {
			marketId: marketInfo.marketId,
			marketStatus: marketInfo.status.toString(),
			ticker: marketInfo.ticker,
			baseDenom: marketInfo.baseDenom,
			quoteDenom: marketInfo.quoteDenom,
			makerFeeRate: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.makerFeeRate).toFixed(),
			quoteToken: void 0,
			baseToken: void 0,
			takerFeeRate: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.takerFeeRate).toFixed(),
			serviceProviderFee: "",
			minPriceTickSize: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.minPriceTickSize).toNumber(),
			minQuantityTickSize: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.minQuantityTickSize).toNumber(),
			minNotional: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.minNotional).toNumber()
		};
	}
	static fullSpotMarketsResponseToSpotMarkets(response) {
		return response.markets.map((market) => {
			return ChainGrpcExchangeTransformer.grpcFullSpotMarketToSpotMarket(market);
		});
	}
	static grpcFullSpotMarketToSpotMarket(market) {
		const marketInfo = market.market;
		return {
			marketId: marketInfo.marketId,
			marketStatus: marketInfo.status.toString(),
			ticker: marketInfo.ticker,
			baseDenom: marketInfo.baseDenom,
			quoteDenom: marketInfo.quoteDenom,
			makerFeeRate: marketInfo.makerFeeRate,
			quoteToken: void 0,
			baseToken: void 0,
			takerFeeRate: marketInfo.takerFeeRate,
			serviceProviderFee: "",
			minPriceTickSize: Number(marketInfo.minPriceTickSize),
			minQuantityTickSize: Number(marketInfo.minQuantityTickSize),
			minNotional: Number(marketInfo.minNotional)
		};
	}
	static fullDerivativeMarketsResponseToDerivativeMarkets(response) {
		return response.markets.map((market) => {
			return ChainGrpcExchangeTransformer.grpcFullDerivativeMarketToDerivativeMarket(market);
		});
	}
	static grpcFullDerivativeMarketToDerivativeMarket(market) {
		var _market$info, _market$info$perpetua, _market$info$perpetua2, _market$info2, _market$info$perpetua3, _market$info$perpetua4, _market$info3, _market$info$perpetua5, _market$info$perpetua6, _market$info4, _market$info$perpetua7, _market$info$perpetua8;
		const marketInfo = market.market;
		return {
			oracleType: marketInfo.oracleType.toString(),
			marketId: marketInfo.marketId,
			marketStatus: marketInfo.status.toString(),
			ticker: marketInfo.ticker,
			quoteDenom: marketInfo.quoteDenom,
			makerFeeRate: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.makerFeeRate).toFixed(),
			takerFeeRate: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.takerFeeRate).toFixed(),
			serviceProviderFee: "",
			quoteToken: void 0,
			minPriceTickSize: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.minPriceTickSize).toNumber(),
			minQuantityTickSize: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.minQuantityTickSize).toNumber(),
			minNotional: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.minNotional).toNumber(),
			reduceMarginRatio: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.reduceMarginRatio).toFixed(),
			initialMarginRatio: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.initialMarginRatio).toFixed(),
			maintenanceMarginRatio: require_utils.denomAmountFromGrpcChainDenomAmount(marketInfo.maintenanceMarginRatio).toFixed(),
			isPerpetual: marketInfo.isPerpetual,
			oracleBase: marketInfo.oracleBase,
			oracleQuote: marketInfo.oracleQuote,
			oracleScaleFactor: marketInfo.oracleScaleFactor,
			perpetualMarketInfo: {
				hourlyFundingRateCap: require_utils.denomAmountFromGrpcChainDenomAmount(((_market$info = market.info) === null || _market$info === void 0 ? void 0 : _market$info.oneofKind) === "perpetualInfo" ? (_market$info$perpetua = (_market$info$perpetua2 = market.info.perpetualInfo.marketInfo) === null || _market$info$perpetua2 === void 0 || (_market$info$perpetua2 = _market$info$perpetua2.hourlyFundingRateCap) === null || _market$info$perpetua2 === void 0 ? void 0 : _market$info$perpetua2.toString()) !== null && _market$info$perpetua !== void 0 ? _market$info$perpetua : "0" : "0").toFixed(),
				hourlyInterestRate: require_utils.denomAmountFromGrpcChainDenomAmount(((_market$info2 = market.info) === null || _market$info2 === void 0 ? void 0 : _market$info2.oneofKind) === "perpetualInfo" ? (_market$info$perpetua3 = (_market$info$perpetua4 = market.info.perpetualInfo.marketInfo) === null || _market$info$perpetua4 === void 0 || (_market$info$perpetua4 = _market$info$perpetua4.hourlyInterestRate) === null || _market$info$perpetua4 === void 0 ? void 0 : _market$info$perpetua4.toString()) !== null && _market$info$perpetua3 !== void 0 ? _market$info$perpetua3 : "0" : "0").toFixed(),
				nextFundingTimestamp: parseInt(((_market$info3 = market.info) === null || _market$info3 === void 0 ? void 0 : _market$info3.oneofKind) === "perpetualInfo" ? (_market$info$perpetua5 = (_market$info$perpetua6 = market.info.perpetualInfo.marketInfo) === null || _market$info$perpetua6 === void 0 || (_market$info$perpetua6 = _market$info$perpetua6.nextFundingTimestamp) === null || _market$info$perpetua6 === void 0 ? void 0 : _market$info$perpetua6.toString()) !== null && _market$info$perpetua5 !== void 0 ? _market$info$perpetua5 : "" : "", 10),
				fundingInterval: parseInt(((_market$info4 = market.info) === null || _market$info4 === void 0 ? void 0 : _market$info4.oneofKind) === "perpetualInfo" ? (_market$info$perpetua7 = (_market$info$perpetua8 = market.info.perpetualInfo.marketInfo) === null || _market$info$perpetua8 === void 0 || (_market$info$perpetua8 = _market$info$perpetua8.fundingInterval) === null || _market$info$perpetua8 === void 0 ? void 0 : _market$info$perpetua8.toString()) !== null && _market$info$perpetua7 !== void 0 ? _market$info$perpetua7 : "" : "", 10)
			}
		};
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcExchangeApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcExchangeApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Exchange);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryExchangeParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.queryExchangeParams.bind(this.client));
		return ChainGrpcExchangeTransformer.moduleParamsResponseToParams(response);
	}
	async fetchModuleState() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryModuleStateRequest.create();
		return (await this.executeGrpcCall(request, this.client.exchangeModuleState.bind(this.client))).state;
	}
	async fetchFeeDiscountSchedule() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryFeeDiscountScheduleRequest.create();
		const response = await this.executeGrpcCall(request, this.client.feeDiscountSchedule.bind(this.client));
		return ChainGrpcExchangeTransformer.feeDiscountScheduleResponseToFeeDiscountSchedule(response);
	}
	async fetchFeeDiscountAccountInfo(injectiveAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryFeeDiscountAccountInfoRequest.create();
		request.account = injectiveAddress;
		const response = await this.executeGrpcCall(request, this.client.feeDiscountAccountInfo.bind(this.client));
		return ChainGrpcExchangeTransformer.feeDiscountAccountInfoResponseToFeeDiscountAccountInfo(response);
	}
	async fetchTradingRewardsCampaign() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryTradeRewardCampaignRequest.create();
		const response = await this.executeGrpcCall(request, this.client.tradeRewardCampaign.bind(this.client));
		return ChainGrpcExchangeTransformer.tradingRewardsCampaignResponseToTradingRewardsCampaign(response);
	}
	async fetchTradeRewardPoints(injectiveAddresses) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryTradeRewardPointsRequest.create();
		request.accounts = injectiveAddresses;
		return (await this.executeGrpcCall(request, this.client.tradeRewardPoints.bind(this.client))).accountTradeRewardPoints;
	}
	async fetchPendingTradeRewardPoints(injectiveAddresses, timestamp) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryTradeRewardPointsRequest.create();
		request.accounts = injectiveAddresses;
		if (timestamp) request.pendingPoolTimestamp = BigInt(timestamp);
		return (await this.executeGrpcCall(request, this.client.pendingTradeRewardPoints.bind(this.client))).accountTradeRewardPoints;
	}
	async fetchPositions() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryPositionsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.positions.bind(this.client));
		return ChainGrpcExchangeTransformer.positionsResponseToPositions(response);
	}
	async fetchSubaccountTradeNonce(subaccountId) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QuerySubaccountTradeNonceRequest.create();
		request.subaccountId = subaccountId;
		return await this.executeGrpcCall(request, this.client.subaccountTradeNonce.bind(this.client));
	}
	async fetchIsOptedOutOfRewards(account) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryIsOptedOutOfRewardsRequest.create();
		request.account = account;
		const response = await this.executeGrpcCall(request, this.client.isOptedOutOfRewards.bind(this.client));
		return ChainGrpcExchangeTransformer.isOptedOutOfRewardsResponseToIsOptedOutOfRewards(response);
	}
	async fetchActiveStakeGrant(account) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryActiveStakeGrantRequest.create();
		request.grantee = account;
		const response = await this.executeGrpcCall(request, this.client.activeStakeGrant.bind(this.client));
		return ChainGrpcExchangeTransformer.activeStakeGrantResponseToActiveStakeGrant(response);
	}
	async fetchDenomDecimal(denom) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryDenomDecimalRequest.create();
		request.denom = denom;
		return (await this.executeGrpcCall(request, this.client.denomDecimal.bind(this.client))).decimal;
	}
	async fetchDenomDecimals() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryDenomDecimalsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.denomDecimals.bind(this.client));
		return ChainGrpcExchangeTransformer.denomDecimalsResponseToDenomDecimals(response);
	}
	async fetchDenomMinNotional(denom) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryDenomMinNotionalRequest.create();
		request.denom = denom;
		return (await this.executeGrpcCall(request, this.client.denomMinNotional.bind(this.client))).amount;
	}
	async fetchDenomMinNotionals() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryDenomMinNotionalsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.denomMinNotionals.bind(this.client));
		return ChainGrpcExchangeTransformer.denomMinNotionalsResponseToDenomMinNotionals(response);
	}
	async fetchDerivativeMarkets(status, marketIds) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryDerivativeMarketsRequest.create();
		if (status) request.status = status;
		if (marketIds) request.marketIds = marketIds;
		const response = await this.executeGrpcCall(request, this.client.derivativeMarkets.bind(this.client));
		return ChainGrpcExchangeTransformer.fullDerivativeMarketsResponseToDerivativeMarkets(response);
	}
	async fetchSpotMarkets(status, marketIds) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QuerySpotMarketsRequest.create();
		if (status) request.status = status;
		if (marketIds) request.marketIds = marketIds;
		const response = await this.executeGrpcCall(request, this.client.spotMarkets.bind(this.client));
		return ChainGrpcExchangeTransformer.spotMarketsResponseToSpotMarkets(response);
	}
	async fetchFullSpotMarkets(status, marketIds) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_query_pb.QueryFullSpotMarketsRequest.create();
		if (status) request.status = status;
		if (marketIds) request.marketIds = marketIds;
		const response = await this.executeGrpcCall(request, this.client.fullSpotMarkets.bind(this.client));
		return ChainGrpcExchangeTransformer.fullSpotMarketsResponseToSpotMarkets(response);
	}
};

//#endregion
Object.defineProperty(exports, 'ChainGrpcExchangeApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcExchangeApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcExchangeTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcExchangeTransformer;
  }
});
