// @injectivelabs/sdk-ts/dist/cjs/ChainRestTendermintApi-CyeoAmer.cjs
const require_chunk = require('./chunk-CbDLau6x.cjs');
const require_utils = require('./utils-D3XFGv8y.cjs');
const require_defineProperty = require('./defineProperty-dyr5BsvG.cjs');
const require_BaseGrpcConsumer = require('./BaseGrpcConsumer-Db3sCC6b.cjs');
const require_BaseRestConsumer = require('./BaseRestConsumer-CRkyjiSo.cjs');
let __injectivelabs_exceptions = require("@injectivelabs/exceptions");
let __injectivelabs_core_proto_ts_v2_generated_injective_crypto_v1beta1_ethsecp256k1_keys_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/crypto/v1beta1/ethsecp256k1/keys_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_crypto_v1beta1_ethsecp256k1_keys_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_crypto_v1beta1_ethsecp256k1_keys_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/exchange/v1beta1/exchange_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_oracle_v1beta1_oracle_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/oracle/v1beta1/oracle_pb");
let __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_gov_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/gov/v1/gov_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_gov_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_gov_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_types_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmwasm/wasm/v1/types_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_types_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_types_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_permissions_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/permissions/v1beta1/permissions_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_permissions_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_permissions_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_types_v1beta1_account_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/types/v1beta1/account_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_types_v1beta1_account_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_types_v1beta1_account_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_auth_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/auth/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_auth_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_auth_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_auth_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/auth/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_cosmos_base_tendermint_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/base/tendermint/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_base_tendermint_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_base_tendermint_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_base_tendermint_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/base/tendermint/v1beta1/query_pb.client");

//#region src/client/chain/types/gov.ts
const VoteOptionMap = __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_gov_pb.VoteOption;
const ProposalStatusMap = __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_gov_pb.ProposalStatus;

//#endregion
//#region src/client/chain/types/wasm.ts
const ContractCodeHistoryOperationTypeMap = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_types_pb.ContractCodeHistoryOperationType;

//#endregion
//#region src/client/chain/types/staking.ts
const BondStatus = {
	UnBonded: "UnBonded",
	UnBonding: "UnBonding",
	Bonded: "Bonded"
};

//#endregion
//#region src/client/chain/types/exchange.ts
const GrpcOrderTypeMap = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb.OrderType;
const OrderTypeMap = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb.OrderType;
const GrpcMarketStatusMap = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb.MarketStatus;

//#endregion
//#region src/client/chain/types/insurance.ts
const OracleTypeMap = __injectivelabs_core_proto_ts_v2_generated_injective_oracle_v1beta1_oracle_pb.OracleType;

//#endregion
//#region src/client/chain/types/permissions.ts
const PermissionActionMap = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_permissions_pb.Action;

//#endregion
//#region src/client/chain/types/index.ts
const ChainModule = { ...__injectivelabs_exceptions.ChainErrorModule };

//#endregion
//#region src/client/chain/transformers/ChainGrpcAuthTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcAuthTransformer = class ChainGrpcAuthTransformer {
	static moduleParamsResponseToModuleParams(response) {
		const params = response.params;
		return {
			maxMemoCharacters: Number(params.maxMemoCharacters),
			txSigLimit: Number(params.txSigLimit),
			txSizeCostPerByte: Number(params.txSizeCostPerByte),
			sigVerifyCostEd25519: Number(params.sigVerifyCostEd25519),
			sigVerifyCostSecp256k1: Number(params.sigVerifyCostSecp256K1)
		};
	}
	static grpcAccountToAccount(ethAccount) {
		const account = __injectivelabs_core_proto_ts_v2_generated_injective_types_v1beta1_account_pb.EthAccount.fromBinary(ethAccount.value);
		const baseAccount = account.baseAccount;
		const pubKey = baseAccount.pubKey;
		return {
			codeHash: require_utils.uint8ArrayToString(account.codeHash),
			baseAccount: {
				address: baseAccount.address,
				pubKey: pubKey ? {
					key: require_utils.uint8ArrayToBase64(__injectivelabs_core_proto_ts_v2_generated_injective_crypto_v1beta1_ethsecp256k1_keys_pb.PubKey.fromBinary(pubKey.value).key),
					typeUrl: pubKey.typeUrl
				} : void 0,
				accountNumber: Number(baseAccount.accountNumber),
				sequence: Number(baseAccount.sequence)
			}
		};
	}
	static accountResponseToAccount(response) {
		return ChainGrpcAuthTransformer.grpcAccountToAccount(response.account);
	}
	static accountsResponseToAccounts(response) {
		return {
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination),
			accounts: response.accounts.map(ChainGrpcAuthTransformer.grpcAccountToAccount)
		};
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcAuthApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcAuthApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", ChainModule.Auth);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_cosmos_auth_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_auth_v1beta1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcAuthTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchAccount(address) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_auth_v1beta1_query_pb.QueryAccountRequest.create();
		request.address = address;
		const response = await this.executeGrpcCall(request, this.client.account.bind(this.client));
		return ChainGrpcAuthTransformer.accountResponseToAccount(response);
	}
	async fetchAccounts(pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_auth_v1beta1_query_pb.QueryAccountsRequest.create();
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.accounts.bind(this.client));
		return ChainGrpcAuthTransformer.accountsResponseToAccounts(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcTendermintApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcTendermintApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", ChainModule.Tendermint);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_cosmos_base_tendermint_v1beta1_query_pb_client.ServiceClient);
	}
	async fetchLatestBlock() {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_base_tendermint_v1beta1_query_pb.GetLatestBlockRequest.create();
		const response = await this.executeGrpcCall(request, this.client.getLatestBlock.bind(this.client));
		return response.sdkBlock || response.block;
	}
	async fetchBlock(height) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_base_tendermint_v1beta1_query_pb.GetBlockByHeightRequest.create();
		request.height = BigInt(height.toString());
		const response = await this.executeGrpcCall(request, this.client.getBlockByHeight.bind(this.client));
		return response.sdkBlock || response.block;
	}
};

//#endregion
//#region src/client/chain/rest/ChainRestAuthApi.ts
/**
* @category Chain Rest API
*/
var ChainRestAuthApi = class extends require_BaseRestConsumer.BaseRestConsumer {
	/**
	* Looks up the account information for the Injective address.
	*
	* @param address address of account to look up
	*/
	async fetchAccount(address, params = {}) {
		const endpoint = `cosmos/auth/v1beta1/accounts/${address}`;
		try {
			return (await this.retry(() => this.get(endpoint, params))).data;
		} catch (e) {
			if (e instanceof __injectivelabs_exceptions.HttpRequestException) throw e;
			throw new __injectivelabs_exceptions.HttpRequestException(new Error(e), {
				code: __injectivelabs_exceptions.UnspecifiedErrorCode,
				context: `${this.endpoint}/${endpoint}`,
				contextModule: ChainModule.Auth
			});
		}
	}
	/**
	* Looks up the account information for any cosmos chain address.
	*
	* @param address address of account to look up
	*/
	async fetchCosmosAccount(address, params = {}) {
		const endpoint = `cosmos/auth/v1beta1/accounts/${address}`;
		try {
			const isInjectiveAddress = address.startsWith("inj") || address.startsWith("evmos");
			const response = await this.retry(() => this.get(endpoint, params));
			return isInjectiveAddress ? response.data.account.base_account : response.data.account;
		} catch (e) {
			if (e instanceof __injectivelabs_exceptions.HttpRequestException) throw e;
			throw new __injectivelabs_exceptions.HttpRequestException(new Error(e), {
				code: __injectivelabs_exceptions.UnspecifiedErrorCode,
				context: `${this.endpoint}/${endpoint}`,
				contextModule: ChainModule.Auth
			});
		}
	}
};

//#endregion
//#region src/client/chain/rest/ChainRestTendermintApi.ts
/**
* @category Chain Rest API
*/
var ChainRestTendermintApi = class extends require_BaseRestConsumer.BaseRestConsumer {
	async fetchLatestBlock(params = {}) {
		const endpoint = `cosmos/base/tendermint/v1beta1/blocks/latest`;
		try {
			const response = await this.retry(() => this.get(endpoint, params));
			return response.data.sdk_block || response.data.block;
		} catch (e) {
			if (e instanceof __injectivelabs_exceptions.HttpRequestException) throw e;
			throw new __injectivelabs_exceptions.HttpRequestException(new Error(e), {
				code: __injectivelabs_exceptions.UnspecifiedErrorCode,
				context: `${this.endpoint}/${endpoint}`,
				contextModule: ChainModule.Tendermint
			});
		}
	}
	async fetchBlock(height, params = {}) {
		const endpoint = `cosmos/base/tendermint/v1beta1/blocks/${height}`;
		try {
			const response = await this.retry(() => this.get(endpoint, params));
			return response.data.sdk_block || response.data.block;
		} catch (e) {
			if (e instanceof __injectivelabs_exceptions.HttpRequestException) throw e;
			throw new __injectivelabs_exceptions.HttpRequestException(new Error(e), {
				code: __injectivelabs_exceptions.UnspecifiedErrorCode,
				context: `${this.endpoint}/${endpoint}`,
				contextModule: ChainModule.Tendermint
			});
		}
	}
	async fetchNodeInfo(params = {}) {
		const endpoint = `cosmos/base/tendermint/v1beta1/node_info`;
		try {
			const response = await this.retry(() => this.get(endpoint, params));
			return {
				nodeInfo: response.data.default_node_info,
				applicationVersion: response.data.application_version
			};
		} catch (e) {
			if (e instanceof __injectivelabs_exceptions.HttpRequestException) throw e;
			throw new __injectivelabs_exceptions.HttpRequestException(new Error(e), {
				code: __injectivelabs_exceptions.UnspecifiedErrorCode,
				context: `${this.endpoint}/${endpoint}`,
				contextModule: ChainModule.Tendermint
			});
		}
	}
};

//#endregion
Object.defineProperty(exports, 'BondStatus', {
  enumerable: true,
  get: function () {
    return BondStatus;
  }
});
Object.defineProperty(exports, 'ChainGrpcAuthApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcAuthApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcAuthTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcAuthTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcTendermintApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcTendermintApi;
  }
});
Object.defineProperty(exports, 'ChainModule', {
  enumerable: true,
  get: function () {
    return ChainModule;
  }
});
Object.defineProperty(exports, 'ChainRestAuthApi', {
  enumerable: true,
  get: function () {
    return ChainRestAuthApi;
  }
});
Object.defineProperty(exports, 'ChainRestTendermintApi', {
  enumerable: true,
  get: function () {
    return ChainRestTendermintApi;
  }
});
Object.defineProperty(exports, 'ContractCodeHistoryOperationTypeMap', {
  enumerable: true,
  get: function () {
    return ContractCodeHistoryOperationTypeMap;
  }
});
Object.defineProperty(exports, 'GrpcMarketStatusMap', {
  enumerable: true,
  get: function () {
    return GrpcMarketStatusMap;
  }
});
Object.defineProperty(exports, 'GrpcOrderTypeMap', {
  enumerable: true,
  get: function () {
    return GrpcOrderTypeMap;
  }
});
Object.defineProperty(exports, 'OracleTypeMap', {
  enumerable: true,
  get: function () {
    return OracleTypeMap;
  }
});
Object.defineProperty(exports, 'OrderTypeMap', {
  enumerable: true,
  get: function () {
    return OrderTypeMap;
  }
});
Object.defineProperty(exports, 'PermissionActionMap', {
  enumerable: true,
  get: function () {
    return PermissionActionMap;
  }
});
Object.defineProperty(exports, 'ProposalStatusMap', {
  enumerable: true,
  get: function () {
    return ProposalStatusMap;
  }
});
Object.defineProperty(exports, 'VoteOptionMap', {
  enumerable: true,
  get: function () {
    return VoteOptionMap;
  }
});
