// @injectivelabs/sdk-ts/dist/cjs/chain-CPs0kwL0.cjs
const require_chunk = require('./chunk-CbDLau6x.cjs');
const require_utils = require('./utils-D3XFGv8y.cjs');
const require_defineProperty = require('./defineProperty-dyr5BsvG.cjs');
const require_ChainRestTendermintApi = require('./ChainRestTendermintApi-CyeoAmer.cjs');
const require_BaseGrpcConsumer = require('./BaseGrpcConsumer-Db3sCC6b.cjs');
const require_BaseRestConsumer = require('./BaseRestConsumer-CRkyjiSo.cjs');
let __injectivelabs_utils = require("@injectivelabs/utils");
let __injectivelabs_exceptions = require("@injectivelabs/exceptions");
let __injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_authz_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/authz/v1beta1/authz_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_authz_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_authz_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/gov/v1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/gov/v1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_ibc_applications_transfer_v1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/ibc/applications/transfer/v1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_ibc_applications_transfer_v1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_ibc_applications_transfer_v1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_ibc_applications_transfer_v1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/ibc/applications/transfer/v1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/evm/v1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/evm/v1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/bank/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/bank/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_cosmos_mint_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/mint/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_mint_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_mint_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_mint_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/mint/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmwasm/wasm/v1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/cosmwasm/wasm/v1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/authz/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/authz/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_peggy_v1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/peggy/v1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_peggy_v1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_peggy_v1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_peggy_v1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/peggy/v1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/wasmx/v1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/wasmx/v1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_erc20_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/erc20/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_erc20_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_erc20_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_erc20_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/erc20/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_oracle_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/oracle/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_oracle_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_oracle_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_oracle_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/oracle/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_txfees_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/txfees/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_txfees_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_txfees_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_txfees_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/txfees/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_auction_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/auction/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_auction_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_auction_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_auction_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/auction/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/staking/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/staking/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/permissions/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/permissions/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/distribution/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/cosmos/distribution/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_tokenfactory_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/tokenfactory/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_tokenfactory_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_tokenfactory_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_tokenfactory_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/tokenfactory/v1beta1/query_pb.client");
let __injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/insurance/v1beta1/query_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb);
let __injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb_client = require("@injectivelabs/core-proto-ts-v2/generated/injective/insurance/v1beta1/query_pb.client");
let http_status_codes = require("http-status-codes");

//#region src/client/chain/transformers/ChainGrpcGovTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcGovTransformer = class ChainGrpcGovTransformer {
	static moduleParamsResponseToModuleParams(response) {
		var _params$minDeposit, _params$expeditedMinD, _params$maxDepositPer, _params$maxDepositPer2, _params$votingPeriod$, _params$votingPeriod, _params$expeditedVoti, _params$expeditedVoti2;
		const params = response.params;
		return {
			depositParams: {
				minDeposit: (_params$minDeposit = params === null || params === void 0 ? void 0 : params.minDeposit) !== null && _params$minDeposit !== void 0 ? _params$minDeposit : [],
				expeditedMinDeposit: (_params$expeditedMinD = params === null || params === void 0 ? void 0 : params.expeditedMinDeposit) !== null && _params$expeditedMinD !== void 0 ? _params$expeditedMinD : [],
				maxDepositPeriod: Number((_params$maxDepositPer = params === null || params === void 0 || (_params$maxDepositPer2 = params.maxDepositPeriod) === null || _params$maxDepositPer2 === void 0 ? void 0 : _params$maxDepositPer2.seconds) !== null && _params$maxDepositPer !== void 0 ? _params$maxDepositPer : 0n)
			},
			votingParams: {
				votingPeriod: Number((_params$votingPeriod$ = params === null || params === void 0 || (_params$votingPeriod = params.votingPeriod) === null || _params$votingPeriod === void 0 ? void 0 : _params$votingPeriod.seconds) !== null && _params$votingPeriod$ !== void 0 ? _params$votingPeriod$ : 0n),
				expeditedVotingPeriod: Number((_params$expeditedVoti = params === null || params === void 0 || (_params$expeditedVoti2 = params.expeditedVotingPeriod) === null || _params$expeditedVoti2 === void 0 ? void 0 : _params$expeditedVoti2.seconds) !== null && _params$expeditedVoti !== void 0 ? _params$expeditedVoti : 0n)
			},
			tallyParams: {
				quorum: !(params === null || params === void 0 ? void 0 : params.quorum) ? "0" : typeof params.quorum === "string" ? params.quorum : require_utils.uint8ArrayToString(params.quorum),
				threshold: !(params === null || params === void 0 ? void 0 : params.threshold) ? "0" : typeof params.threshold === "string" ? params.threshold : require_utils.uint8ArrayToString(params.threshold),
				vetoThreshold: !(params === null || params === void 0 ? void 0 : params.vetoThreshold) ? "0" : typeof params.vetoThreshold === "string" ? params.vetoThreshold : require_utils.uint8ArrayToString(params.vetoThreshold),
				expeditedThreshold: !(params === null || params === void 0 ? void 0 : params.expeditedThreshold) ? "0" : typeof params.expeditedThreshold === "string" ? params.expeditedThreshold : require_utils.uint8ArrayToString(params.expeditedThreshold)
			}
		};
	}
	static moduleParamsResponseToModuleParamsByType({ depositParams, votingParams, tallyParams }) {
		var _depositParams$maxDep, _depositParams$maxDep2, _votingParams$votingP, _votingParams$votingP2, _votingParams$expedit, _votingParams$expedit2;
		return {
			depositParams: {
				minDeposit: depositParams === null || depositParams === void 0 ? void 0 : depositParams.minDeposit,
				expeditedMinDeposit: depositParams === null || depositParams === void 0 ? void 0 : depositParams.expeditedMinDeposit,
				maxDepositPeriod: Number((_depositParams$maxDep = depositParams === null || depositParams === void 0 || (_depositParams$maxDep2 = depositParams.maxDepositPeriod) === null || _depositParams$maxDep2 === void 0 ? void 0 : _depositParams$maxDep2.seconds) !== null && _depositParams$maxDep !== void 0 ? _depositParams$maxDep : 0n)
			},
			votingParams: {
				votingPeriod: Number((_votingParams$votingP = (_votingParams$votingP2 = votingParams.votingPeriod) === null || _votingParams$votingP2 === void 0 ? void 0 : _votingParams$votingP2.seconds) !== null && _votingParams$votingP !== void 0 ? _votingParams$votingP : 0n),
				expeditedVotingPeriod: Number((_votingParams$expedit = (_votingParams$expedit2 = votingParams.expeditedVotingPeriod) === null || _votingParams$expedit2 === void 0 ? void 0 : _votingParams$expedit2.seconds) !== null && _votingParams$expedit !== void 0 ? _votingParams$expedit : 0n)
			},
			tallyParams: {
				quorum: typeof tallyParams.quorum === "string" ? tallyParams.quorum : require_utils.uint8ArrayToString(tallyParams.quorum),
				threshold: typeof tallyParams.threshold === "string" ? tallyParams.threshold : require_utils.uint8ArrayToString(tallyParams.threshold),
				vetoThreshold: typeof tallyParams.vetoThreshold === "string" ? tallyParams.vetoThreshold : require_utils.uint8ArrayToString(tallyParams.vetoThreshold),
				expeditedThreshold: typeof tallyParams.expeditedThreshold === "string" ? tallyParams.expeditedThreshold : require_utils.uint8ArrayToString(tallyParams.expeditedThreshold)
			}
		};
	}
	static proposalResponseToProposal(response) {
		if (!response.proposal) return;
		return ChainGrpcGovTransformer.grpcProposalToProposal(response.proposal);
	}
	static proposalsResponseToProposals(response) {
		const proposals = response.proposals.map((p) => ChainGrpcGovTransformer.grpcProposalToProposal(p)).filter((p) => p !== void 0);
		const pagination = response.pagination;
		return {
			proposals,
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(pagination)
		};
	}
	static depositsResponseToDeposits(response) {
		const pagination = response.pagination;
		return {
			deposits: response.deposits.map((deposit) => {
				return {
					depositor: deposit.depositor,
					amounts: deposit.amount.map((coin) => ({
						denom: coin.denom,
						amount: (0, __injectivelabs_utils.toHumanReadable)(coin.amount).toFixed()
					}))
				};
			}),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(pagination)
		};
	}
	static grpcVoteToVote(vote) {
		return {
			proposalId: Number(vote.proposalId),
			voter: vote.voter,
			metadata: vote.metadata,
			options: vote.options
		};
	}
	static votesResponseToVotes(response) {
		return {
			votes: response.votes.map(ChainGrpcGovTransformer.grpcVoteToVote),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
	static tallyResultResponseToTallyResult(response) {
		const result = response.tally;
		return ChainGrpcGovTransformer.grpcTallyResultToTallyResult(result);
	}
	static grpcTallyResultToTallyResult(result) {
		var _result$yesCount, _result$abstainCount, _result$noCount, _result$noWithVetoCou;
		return {
			yesCount: (_result$yesCount = result === null || result === void 0 ? void 0 : result.yesCount) !== null && _result$yesCount !== void 0 ? _result$yesCount : "0",
			abstainCount: (_result$abstainCount = result === null || result === void 0 ? void 0 : result.abstainCount) !== null && _result$abstainCount !== void 0 ? _result$abstainCount : "0",
			noCount: (_result$noCount = result === null || result === void 0 ? void 0 : result.noCount) !== null && _result$noCount !== void 0 ? _result$noCount : "0",
			noWithVetoCount: (_result$noWithVetoCou = result === null || result === void 0 ? void 0 : result.noWithVetoCount) !== null && _result$noWithVetoCou !== void 0 ? _result$noWithVetoCou : "0"
		};
	}
	static grpcProposalToProposal(proposal) {
		var _proposal$messages;
		const finalTallyResult = proposal.finalTallyResult;
		const [message] = (_proposal$messages = proposal.messages) !== null && _proposal$messages !== void 0 ? _proposal$messages : [];
		if (!message) return;
		return {
			proposalId: Number(proposal.id),
			title: proposal.title,
			summary: proposal.summary,
			proposer: proposal.proposer,
			content: {
				type: message.typeUrl,
				value: message.value
			},
			type: message.typeUrl,
			submitTime: require_utils.protobufTimestampToUnixSeconds(proposal.submitTime),
			status: proposal.status,
			expedited: proposal.expedited,
			failedReason: proposal.failedReason,
			finalTallyResult: ChainGrpcGovTransformer.grpcTallyResultToTallyResult(finalTallyResult),
			depositEndTime: require_utils.protobufTimestampToUnixSeconds(proposal.depositEndTime),
			totalDeposits: proposal.totalDeposit.map((coin) => ({
				denom: coin.denom,
				amount: (0, __injectivelabs_utils.toHumanReadable)(coin.amount).toFixed()
			})),
			votingStartTime: require_utils.protobufTimestampToUnixSeconds(proposal.votingStartTime),
			votingEndTime: require_utils.protobufTimestampToUnixSeconds(proposal.votingEndTime)
		};
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcEvmTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcEvmTransformer = class {
	static grpcBlobConfigToBlobConfig(config) {
		return {
			max: config.max.toString(),
			target: config.target.toString(),
			baseFeeUpdateFraction: config.baseFeeUpdateFraction.toString()
		};
	}
	static grpcBlockScheduleConfigToBlockScheduleConfig(config) {
		return {
			cancun: config.cancun ? this.grpcBlobConfigToBlobConfig(config.cancun) : void 0,
			prague: config.prague ? this.grpcBlobConfigToBlobConfig(config.prague) : void 0,
			osaka: config.osaka ? this.grpcBlobConfigToBlobConfig(config.osaka) : void 0,
			verkle: config.verkle ? this.grpcBlobConfigToBlobConfig(config.verkle) : void 0
		};
	}
	static grpcChainConfigToChainConfig(config) {
		return {
			homesteadBlock: config.homesteadBlock,
			daoForkBlock: config.daoForkBlock,
			daoForkSupport: config.daoForkSupport,
			eip150Block: config.eip150Block,
			eip150Hash: config.eip150Hash,
			eip155Block: config.eip155Block,
			eip158Block: config.eip158Block,
			byzantiumBlock: config.byzantiumBlock,
			constantinopleBlock: config.constantinopleBlock,
			petersburgBlock: config.petersburgBlock,
			istanbulBlock: config.istanbulBlock,
			muirGlacierBlock: config.muirGlacierBlock,
			berlinBlock: config.berlinBlock,
			londonBlock: config.londonBlock,
			arrowGlacierBlock: config.arrowGlacierBlock,
			grayGlacierBlock: config.grayGlacierBlock,
			mergeNetsplitBlock: config.mergeNetsplitBlock,
			shanghaiTime: config.shanghaiTime,
			cancunTime: config.cancunTime,
			pragueTime: config.pragueTime,
			eip155ChainId: config.eip155ChainId,
			blobScheduleConfig: config.blobScheduleConfig ? this.grpcBlockScheduleConfigToBlockScheduleConfig(config.blobScheduleConfig) : void 0
		};
	}
	/**
	* response transformers
	* */
	static accountResponseToAccount(response) {
		return {
			balance: response.balance,
			codeHash: response.codeHash,
			nonce: response.nonce.toString()
		};
	}
	static cosmosAccountResponseToCosmosAccount(response) {
		return {
			sequence: response.sequence.toString(),
			cosmosAddress: response.cosmosAddress,
			accountNumber: response.accountNumber.toString()
		};
	}
	static validatorAccountResponseToValidatorAccount(response) {
		return {
			sequence: response.sequence.toString(),
			accountNumber: response.accountNumber.toString(),
			accountAddress: response.accountAddress
		};
	}
	static paramsResponseToParams(response) {
		const { params } = response;
		if (!params) return;
		return {
			evmDenom: params.evmDenom,
			enableCreate: params.enableCreate,
			enableCall: params.enableCall,
			extraEips: params.extraEips.map((eip) => require_utils.bigIntToString(eip)),
			chainConfig: params.chainConfig ? this.grpcChainConfigToChainConfig(params.chainConfig) : void 0,
			allowUnprotectedTxs: params.allowUnprotectedTxs,
			authorizedDeployers: params.authorizedDeployers,
			permissioned: params.permissioned
		};
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcBankTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcBankTransformer = class ChainGrpcBankTransformer {
	static metadataToMetadata(metadata) {
		return metadata;
	}
	static moduleParamsResponseToModuleParams(response) {
		const params = response.params;
		return {
			sendEnabledList: params.sendEnabled,
			defaultSendEnabled: params.defaultSendEnabled
		};
	}
	static denomOwnersResponseToDenomOwners(response) {
		const denomOwners = response.denomOwners;
		const pagination = response.pagination;
		return {
			denomOwners: denomOwners.map((denomOwner) => ({
				address: denomOwner.address,
				balance: denomOwner.balance ? require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin(denomOwner.balance) : void 0
			})),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(pagination)
		};
	}
	static totalSupplyResponseToTotalSupply(response) {
		const balances = response.supply;
		const pagination = response.pagination;
		return {
			supply: balances.map(require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(pagination)
		};
	}
	static denomsMetadataResponseToDenomsMetadata(response) {
		const metadatas = response.metadatas;
		const pagination = response.pagination;
		return {
			metadatas: metadatas.map(ChainGrpcBankTransformer.metadataToMetadata),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(pagination)
		};
	}
	static balanceResponseToBalance(response) {
		return require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin(response.balance);
	}
	static balancesResponseToBalances(response) {
		const balances = response.balances;
		const pagination = response.pagination;
		return {
			balances: balances.map(require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(pagination)
		};
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcMintTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcMintTransformer = class {
	static moduleParamsResponseToModuleParams(response) {
		return response.params;
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcWasmTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcWasmTransformer = class ChainGrpcWasmTransformer {
	static allContractStateResponseToContractAccountsBalanceWithPagination(response) {
		const contractAccountsBalance = response.models.map((model) => {
			return {
				account: require_utils.uint8ArrayToString(model.key).split("balance").pop(),
				balance: require_utils.uint8ArrayToString(model.value).replace(/['"]+/g, "")
			};
		}).filter(({ account, balance }) => {
			return account && account.startsWith("inj") && balance;
		});
		const contractInfoModel = response.models.find((model) => {
			return require_utils.uint8ArrayToString(model.key) === "contract_info";
		});
		const contractInfoValue = require_utils.uint8ArrayToString((contractInfoModel === null || contractInfoModel === void 0 ? void 0 : contractInfoModel.value) || new Uint8Array());
		const tokenInfoModel = response.models.find((model) => {
			return require_utils.uint8ArrayToString(model.key) === "token_info";
		});
		const tokenInfoValue = require_utils.uint8ArrayToString((tokenInfoModel === null || tokenInfoModel === void 0 ? void 0 : tokenInfoModel.value) || new Uint8Array());
		const marketingInfoModel = response.models.find((model) => {
			return require_utils.uint8ArrayToString(model.key) === "marketing_info";
		});
		const marketingInfoValue = require_utils.uint8ArrayToString((marketingInfoModel === null || marketingInfoModel === void 0 ? void 0 : marketingInfoModel.value) || new Uint8Array());
		return {
			contractAccountsBalance,
			tokenInfo: JSON.parse(tokenInfoValue || "{}"),
			contractInfo: JSON.parse(contractInfoValue || "{}"),
			marketingInfo: JSON.parse(marketingInfoValue || "{}"),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
	static allContractStateResponseToContractState(response) {
		const contractAccountsBalance = response.models.map((model) => {
			return {
				account: require_utils.uint8ArrayToString(model.key).split("balance").pop(),
				balance: require_utils.uint8ArrayToString(model.value).replace(/['"]+/g, "")
			};
		}).filter(({ account, balance }) => {
			return account && account.startsWith("inj") && balance;
		});
		const contractInfoModel = response.models.find((model) => {
			return require_utils.uint8ArrayToString(model.key) === "contract_info";
		});
		const contractInfoValue = require_utils.uint8ArrayToString((contractInfoModel === null || contractInfoModel === void 0 ? void 0 : contractInfoModel.value) || new Uint8Array());
		const tokenInfoModel = response.models.find((model) => {
			return require_utils.uint8ArrayToString(model.key) === "token_info";
		});
		const tokenInfoValue = require_utils.uint8ArrayToString((tokenInfoModel === null || tokenInfoModel === void 0 ? void 0 : tokenInfoModel.value) || new Uint8Array());
		const marketingInfoModel = response.models.find((model) => {
			return require_utils.uint8ArrayToString(model.key) === "marketing_info";
		});
		const marketingInfoValue = require_utils.uint8ArrayToString((marketingInfoModel === null || marketingInfoModel === void 0 ? void 0 : marketingInfoModel.value) || new Uint8Array());
		return {
			contractAccountsBalance,
			tokenInfo: JSON.parse(tokenInfoValue || "{}"),
			contractInfo: JSON.parse(contractInfoValue || "{}"),
			marketingInfo: JSON.parse(marketingInfoValue || "{}"),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
	static contactInfoResponseToContractInfo(contractInfo) {
		const absoluteTxPosition = contractInfo.created;
		return {
			codeId: Number(contractInfo.codeId),
			creator: contractInfo.creator,
			admin: contractInfo.admin,
			label: contractInfo.label,
			created: absoluteTxPosition ? {
				blockHeight: Number(absoluteTxPosition.blockHeight),
				txIndex: Number(absoluteTxPosition.txIndex)
			} : void 0,
			ibcPortId: contractInfo.ibcPortId
		};
	}
	static grpcContractCodeHistoryEntryToContractCodeHistoryEntry(entry) {
		const updated = entry.updated;
		return {
			operation: entry.operation,
			codeId: Number(entry.codeId),
			updated: updated ? {
				blockHeight: Number(updated.blockHeight),
				txIndex: Number(updated.txIndex)
			} : void 0,
			msg: require_utils.fromUtf8(entry.msg)
		};
	}
	static grpcCodeInfoResponseToCodeInfoResponse(info) {
		return {
			codeId: Number(info.codeId),
			creator: info.creator,
			dataHash: info.dataHash
		};
	}
	static contactHistoryResponseToContractHistory(response) {
		return {
			entriesList: response.entries.map(ChainGrpcWasmTransformer.grpcContractCodeHistoryEntryToContractCodeHistoryEntry),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
	static contractCodesResponseToContractCodes(response) {
		return {
			codeInfosList: response.codeInfos.map(ChainGrpcWasmTransformer.grpcCodeInfoResponseToCodeInfoResponse),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
	static contractCodeResponseToContractCode(response) {
		return {
			codeInfo: ChainGrpcWasmTransformer.grpcCodeInfoResponseToCodeInfoResponse(response.codeInfo),
			data: response.data
		};
	}
	static contractByCodeResponseToContractByCode(response) {
		return {
			contractsList: response.contracts,
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcAuthZTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcAuthZTransformer = class ChainGrpcAuthZTransformer {
	static grpcGrantToGrant(grant) {
		const authorization = decodeAuthorizationNoThrow(grant.authorization);
		return {
			authorization: authorization === null || authorization === void 0 ? void 0 : authorization.authorization,
			authorizationType: (authorization === null || authorization === void 0 ? void 0 : authorization.authorizationType) || "",
			expiration: grant.expiration
		};
	}
	static grpcGrantAuthorizationToGrantAuthorization(grant) {
		const authorization = decodeAuthorizationNoThrow(grant.authorization);
		return {
			granter: grant.granter,
			grantee: grant.grantee,
			authorization: authorization === null || authorization === void 0 ? void 0 : authorization.authorization,
			authorizationType: (authorization === null || authorization === void 0 ? void 0 : authorization.authorizationType) || "",
			expiration: grant.expiration
		};
	}
	static grpcGrantsToGrants(response) {
		return {
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination),
			grants: response.grants.map(ChainGrpcAuthZTransformer.grpcGrantToGrant)
		};
	}
	static grpcGranteeGrantsToGranteeGrants(response) {
		return {
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination),
			grants: response.grants.map(ChainGrpcAuthZTransformer.grpcGrantAuthorizationToGrantAuthorization)
		};
	}
	static grpcGranterGrantsToGranterGrants(response) {
		return {
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination),
			grants: response.grants.map(ChainGrpcAuthZTransformer.grpcGrantAuthorizationToGrantAuthorization)
		};
	}
};
const decodeAuthorization = (authorization) => {
	switch (authorization.typeUrl) {
		case "/cosmos.authz.v1beta1.GenericAuthorization": return {
			authorization: __injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_authz_pb.GenericAuthorization.fromBinary(authorization.value),
			authorizationType: "/cosmos.authz.v1beta1.GenericAuthorization"
		};
		default: throw new __injectivelabs_exceptions.GeneralException(/* @__PURE__ */ new Error("Unsupported authorization type"));
	}
};
const decodeAuthorizationNoThrow = (authorization) => {
	if (!authorization) return;
	try {
		return decodeAuthorization(authorization);
	} catch (_unused) {
		return;
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcPeggyTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcPeggyTransformer = class {
	static moduleParamsResponseToModuleParams(response) {
		const params = response.params;
		const valsetReward = params.valsetReward;
		return {
			peggyId: params.peggyId,
			contractSourceHash: params.contractSourceHash,
			bridgeEthereumAddress: params.bridgeEthereumAddress,
			bridgeChainId: params.bridgeChainId.toString(),
			signedValsetsWindow: params.signedValsetsWindow.toString(),
			signedBatchesWindow: params.signedBatchesWindow.toString(),
			signedClaimsWindow: params.signedClaimsWindow.toString(),
			targetBatchTimeout: params.targetBatchTimeout.toString(),
			averageBlockTime: params.averageBlockTime.toString(),
			averageEthereumBlockTime: params.averageEthereumBlockTime.toString(),
			slashFractionValset: params.slashFractionValset,
			slashFractionBatch: params.slashFractionBatch,
			slashFractionClaim: params.slashFractionClaim,
			slashFractionConflictingClaim: params.slashFractionConflictingClaim,
			unbondSlashingValsetsWindow: params.unbondSlashingValsetsWindow.toString(),
			slashFractionBadEthSignature: params.slashFractionBadEthSignature,
			cosmosCoinDenom: params.cosmosCoinDenom,
			cosmosCoinErc20Contract: params.cosmosCoinErc20Contract,
			claimSlashingEnabled: params.claimSlashingEnabled,
			bridgeContractStartHeight: params.bridgeContractStartHeight.toString(),
			valsetReward: {
				denom: valsetReward.denom,
				amount: valsetReward.amount
			}
		};
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcErc20Transformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcErc20Transformer = class {
	static grpcTokenPairToTokenPair(grpcTokenPair) {
		return {
			bankDenom: grpcTokenPair.bankDenom,
			erc20Address: grpcTokenPair.erc20Address
		};
	}
	static paramsResponseToParams(response) {
		const params = response.params;
		return { denomCreationFee: params === null || params === void 0 ? void 0 : params.denomCreationFee };
	}
	static tokenPairsResponseToTokenPairs(response) {
		return {
			tokenPairs: response.tokenPairs.map((tokenPair) => ({
				bankDenom: tokenPair.bankDenom,
				erc20Address: tokenPair.erc20Address
			})),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcTxFeesTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcTxFeesTransformer = class {
	static moduleParamsResponseToModuleParams(response) {
		const params = response.params;
		return {
			maxGasWantedPerTx: params.maxGasWantedPerTx.toString(),
			highGasTxThreshold: params.highGasTxThreshold.toString(),
			minGasPriceForHighGasTx: params.minGasPriceForHighGasTx,
			mempool1559Enabled: params.mempool1559Enabled,
			minGasPrice: require_utils.denomAmountFromGrpcChainDenomAmount(params.minGasPrice).toFixed(),
			defaultBaseFeeMultiplier: require_utils.denomAmountFromGrpcChainDenomAmount(params.defaultBaseFeeMultiplier).toFixed(),
			maxBaseFeeMultiplier: require_utils.denomAmountFromGrpcChainDenomAmount(params.maxBaseFeeMultiplier).toFixed(),
			resetInterval: params.resetInterval.toString(),
			maxBlockChangeRate: require_utils.denomAmountFromGrpcChainDenomAmount(params.maxBlockChangeRate).toFixed(),
			targetBlockSpacePercentRate: require_utils.denomAmountFromGrpcChainDenomAmount(params.targetBlockSpacePercentRate).toFixed(),
			recheckFeeLowBaseFee: require_utils.denomAmountFromGrpcChainDenomAmount(params.recheckFeeLowBaseFee).toFixed(),
			recheckFeeHighBaseFee: require_utils.denomAmountFromGrpcChainDenomAmount(params.recheckFeeHighBaseFee).toFixed(),
			recheckFeeBaseFeeThresholdMultiplier: require_utils.denomAmountFromGrpcChainDenomAmount(params.recheckFeeBaseFeeThresholdMultiplier).toFixed()
		};
	}
	static eipBaseFeeResponseToEipBaseFee(response) {
		return { baseFee: response.baseFee ? require_utils.denomAmountFromGrpcChainDenomAmount(response.baseFee.baseFee).toFixed() : void 0 };
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcAuctionTransformer.ts
/**
* Transformer for Auction module gRPC responses
* @category Chain Grpc Transformer
*/
var ChainGrpcAuctionTransformer = class ChainGrpcAuctionTransformer {
	static grpcBidToBid(grpcBid) {
		return {
			bidder: grpcBid.bidder,
			amount: grpcBid.amount ? require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin(grpcBid.amount) : void 0
		};
	}
	static grpcLastAuctionResultToLastAuctionResult(grpcLastAuctionResult) {
		return {
			winner: grpcLastAuctionResult.winner,
			amount: grpcLastAuctionResult.amount ? require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin(grpcLastAuctionResult.amount) : void 0,
			round: grpcLastAuctionResult.round.toString()
		};
	}
	static moduleParamsResponseToModuleParams(response) {
		var _params$auctionPeriod, _params$minNextBidInc;
		if (!response.params) throw new __injectivelabs_exceptions.GeneralException(/* @__PURE__ */ new Error("Auction params not found in response"));
		const params = response.params;
		return {
			auctionPeriod: Number((_params$auctionPeriod = params.auctionPeriod) !== null && _params$auctionPeriod !== void 0 ? _params$auctionPeriod : 0n),
			minNextBidIncrementRate: (_params$minNextBidInc = params.minNextBidIncrementRate) !== null && _params$minNextBidInc !== void 0 ? _params$minNextBidInc : "0"
		};
	}
	static currentBasketResponseToCurrentBasket(response) {
		return {
			amountList: response.amount.map(require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin),
			auctionRound: Number(response.auctionRound),
			auctionClosingTime: Number(response.auctionClosingTime),
			highestBidder: response.highestBidAmount,
			highestBidAmount: response.highestBidAmount
		};
	}
	static auctionModuleStateResponseToAuctionModuleState(response) {
		if (!response.state) throw new __injectivelabs_exceptions.GeneralException(/* @__PURE__ */ new Error("Auction module state not found in response"));
		const state = response.state;
		if (!state.params) throw new __injectivelabs_exceptions.GeneralException(/* @__PURE__ */ new Error("Auction params not found in state"));
		const params = state.params;
		return {
			params: {
				auctionPeriod: Number(params.auctionPeriod),
				minNextBidIncrementRate: params.minNextBidIncrementRate
			},
			auctionRound: Number(state.auctionRound),
			highestBid: state.highestBid ? ChainGrpcAuctionTransformer.grpcBidToBid(state.highestBid) : void 0,
			auctionEndingTimestamp: Number(state.auctionEndingTimestamp),
			lastAuctionResult: state.lastAuctionResult ? ChainGrpcAuctionTransformer.grpcLastAuctionResultToLastAuctionResult(state.lastAuctionResult) : void 0
		};
	}
	static lastAuctionResultResponseToLastAuctionResult(response) {
		if (!response.lastAuctionResult) return;
		return ChainGrpcAuctionTransformer.grpcLastAuctionResultToLastAuctionResult(response.lastAuctionResult);
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcStakingTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcStakingTransformer = class ChainGrpcStakingTransformer {
	static moduleParamsResponseToModuleParams(response) {
		const params = response.params;
		return {
			unbondingTime: require_utils.protobufTimestampToUnixSeconds(params.unbondingTime),
			minCommissionRate: params.minCommissionRate,
			maxValidators: params.maxValidators,
			maxEntries: params.maxEntries,
			historicalEntries: params.historicalEntries,
			bondDenom: params.bondDenom
		};
	}
	static validatorResponseToValidator(response) {
		return ChainGrpcStakingTransformer.grpcValidatorToValidator(response.validator);
	}
	static validatorsResponseToValidators(response) {
		return {
			validators: response.validators.map((validator) => ChainGrpcStakingTransformer.grpcValidatorToValidator(validator)),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
	static delegationResponseToDelegation(response) {
		const grpcDelegation = response.delegationResponse;
		const delegation = grpcDelegation.delegation;
		const balance = grpcDelegation.balance;
		return {
			delegation: {
				delegatorAddress: delegation ? delegation.delegatorAddress : "",
				validatorAddress: delegation ? delegation.validatorAddress : "",
				shares: (0, __injectivelabs_utils.toHumanReadable)(delegation ? delegation.shares : 0).toFixed()
			},
			balance: {
				denom: balance ? balance.denom : "",
				amount: (0, __injectivelabs_utils.toBigNumber)(balance ? balance.amount : 0).toFixed()
			}
		};
	}
	static delegationsResponseToDelegations(response) {
		return {
			delegations: response.delegationResponses.map((grpcDelegator) => {
				const delegation = grpcDelegator.delegation;
				const balance = grpcDelegator.balance;
				return {
					delegation: {
						delegatorAddress: delegation ? delegation.delegatorAddress : "",
						validatorAddress: delegation ? delegation.validatorAddress : "",
						shares: (0, __injectivelabs_utils.toHumanReadable)(delegation ? delegation.shares : 0).toFixed()
					},
					balance: {
						denom: balance ? balance.denom : "",
						amount: (0, __injectivelabs_utils.toBigNumber)(balance ? balance.amount : 0).toFixed()
					}
				};
			}),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
	static unBondingDelegationsResponseToUnBondingDelegations(response) {
		return {
			unbondingDelegations: response.unbondingResponses.reduce((unbondingDelegations, grpcUnBondingDelegation) => {
				const mappedEntries = grpcUnBondingDelegation.entries.map((entry) => ({
					delegatorAddress: grpcUnBondingDelegation ? grpcUnBondingDelegation.delegatorAddress : "",
					validatorAddress: grpcUnBondingDelegation ? grpcUnBondingDelegation.validatorAddress : "",
					creationHeight: parseInt(entry.creationHeight.toString(), 10),
					completionTime: require_utils.protobufTimestampToUnixSeconds(entry.completionTime),
					initialBalance: (0, __injectivelabs_utils.toBigNumber)(entry.initialBalance).toFixed(),
					balance: (0, __injectivelabs_utils.toBigNumber)(entry.balance).toFixed()
				}));
				return [...unbondingDelegations, ...mappedEntries];
			}, []),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
	static reDelegationsResponseToReDelegations(response) {
		return {
			redelegations: response.redelegationResponses.reduce((uiReDelegator, grpcReDelegationCurrent) => {
				const grpcRedelegation = grpcReDelegationCurrent.redelegation;
				if (!grpcRedelegation) return uiReDelegator;
				const uiRedelegations = grpcReDelegationCurrent.entries.reduce((acc, entry) => {
					var _entry$redelegationEn;
					return [...acc, {
						delegation: {
							completionTime: require_utils.protobufTimestampToUnixSeconds((_entry$redelegationEn = entry.redelegationEntry) === null || _entry$redelegationEn === void 0 ? void 0 : _entry$redelegationEn.completionTime),
							delegatorAddress: grpcRedelegation.delegatorAddress || "",
							sourceValidatorAddress: grpcRedelegation.validatorSrcAddress || "",
							destinationValidatorAddress: (grpcRedelegation === null || grpcRedelegation === void 0 ? void 0 : grpcRedelegation.validatorDstAddress) || ""
						},
						balance: (0, __injectivelabs_utils.toBigNumber)(entry.balance).toFixed()
					}];
				}, []);
				return [...uiReDelegator, ...uiRedelegations];
			}, []),
			pagination: require_utils.ChainGrpcCommonTransformer.grpcPaginationToPaginationV2(response.pagination)
		};
	}
	static grpcValidatorToValidator(validator) {
		return {
			operatorAddress: validator.operatorAddress,
			jailed: validator.jailed,
			status: ChainGrpcStakingTransformer.grpcValidatorStatusToStatus(validator.status),
			tokens: (0, __injectivelabs_utils.toHumanReadable)(validator.tokens).toFixed(),
			delegatorShares: (0, __injectivelabs_utils.toHumanReadable)(validator.delegatorShares).toFixed(),
			description: ChainGrpcStakingTransformer.grpcValidatorDescriptionToDescription(validator.description),
			unbondingHeight: Number(validator.unbondingHeight),
			unbondingTime: validator.unbondingTime,
			commission: ChainGrpcStakingTransformer.grpcValidatorCommissionToCommission(validator.commission),
			minSelfDelegation: validator.minSelfDelegation
		};
	}
	static poolResponseToPool(response) {
		const pool = response.pool;
		if (!pool) return {
			notBondedTokens: "0",
			bondedTokens: "0"
		};
		return {
			notBondedTokens: (0, __injectivelabs_utils.toHumanReadable)(pool.notBondedTokens).toFixed(),
			bondedTokens: (0, __injectivelabs_utils.toHumanReadable)(pool.bondedTokens).toFixed()
		};
	}
	static grpcValidatorDescriptionToDescription(description) {
		return {
			moniker: description ? description.moniker : "",
			identity: description ? description.identity : "",
			website: description ? description.website : "",
			securityContact: description ? description.securityContact : "",
			details: description ? description.details : ""
		};
	}
	static grpcValidatorCommissionToCommission(commission) {
		var _protobufTimestampToD;
		const commissionRates = commission ? commission.commissionRates : null;
		return {
			commissionRates: {
				rate: (0, __injectivelabs_utils.toHumanReadable)(commissionRates ? commissionRates.rate : "0").toFixed(),
				maxRate: (0, __injectivelabs_utils.toHumanReadable)(commissionRates ? commissionRates.maxRate : "0").toFixed(),
				maxChangeRate: (0, __injectivelabs_utils.toHumanReadable)(commissionRates ? commissionRates.maxChangeRate : "0").toFixed()
			},
			updateTime: (_protobufTimestampToD = require_utils.protobufTimestampToDate(commission === null || commission === void 0 ? void 0 : commission.updateTime)) !== null && _protobufTimestampToD !== void 0 ? _protobufTimestampToD : /* @__PURE__ */ new Date()
		};
	}
	static grpcValidatorStatusToStatus(status) {
		switch (status) {
			case 1: return require_ChainRestTendermintApi.BondStatus.UnBonded;
			case 2: return require_ChainRestTendermintApi.BondStatus.UnBonding;
			case 3: return require_ChainRestTendermintApi.BondStatus.Bonded;
			default: return require_ChainRestTendermintApi.BondStatus.UnBonded;
		}
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcPermissionsTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcPermissionsTransformer = class ChainGrpcPermissionsTransformer {
	static grpcRoleToRole(grpcRole) {
		return {
			name: grpcRole.name,
			roleId: grpcRole.roleId,
			permissions: grpcRole.permissions
		};
	}
	static grpcActorRolesToActorRoles(grpcActorRoles) {
		return {
			actor: grpcActorRoles.actor,
			roles: grpcActorRoles.roles
		};
	}
	static grpcRoleManagerToRoleManager(grpcRoleManager) {
		return {
			roles: grpcRoleManager.roles,
			manager: grpcRoleManager.manager
		};
	}
	static grpcPolicyStatusToPolicyStatus(grpcPolicyStatus) {
		return {
			action: grpcPolicyStatus.action,
			isSealed: grpcPolicyStatus.isSealed,
			isDisabled: grpcPolicyStatus.isDisabled
		};
	}
	static grpcPolicyManagerCapabilityToPolicyManagerCapability(grpcPolicyManagerCapability) {
		return {
			action: grpcPolicyManagerCapability.action,
			canSeal: grpcPolicyManagerCapability.canSeal,
			manager: grpcPolicyManagerCapability.manager,
			canDisable: grpcPolicyManagerCapability.canDisable
		};
	}
	static grpcAddressVoucherToAddressVoucher(grpcAddressVoucher) {
		return {
			address: grpcAddressVoucher.address,
			voucher: grpcAddressVoucher.voucher ? require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin(grpcAddressVoucher.voucher) : void 0
		};
	}
	static grpcNamespaceToNamespace(grpcNamespace) {
		return {
			denom: grpcNamespace.denom,
			contractHook: grpcNamespace.contractHook,
			rolePermissions: grpcNamespace.rolePermissions.map(ChainGrpcPermissionsTransformer.grpcRoleToRole),
			actorRoles: grpcNamespace.actorRoles.map(ChainGrpcPermissionsTransformer.grpcActorRolesToActorRoles),
			roleManagers: grpcNamespace.roleManagers.map(ChainGrpcPermissionsTransformer.grpcRoleManagerToRoleManager),
			policyStatuses: grpcNamespace.policyStatuses.map(ChainGrpcPermissionsTransformer.grpcPolicyStatusToPolicyStatus),
			policyManagerCapabilities: grpcNamespace.policyManagerCapabilities.map(ChainGrpcPermissionsTransformer.grpcPolicyManagerCapabilityToPolicyManagerCapability)
		};
	}
	static moduleParamsResponseToModuleParams(response) {
		const params = response.params;
		if (!params) throw new Error("Params not found in response");
		return { wasmHookQueryMaxGas: params.wasmHookQueryMaxGas.toString() };
	}
	static nameSpaceDenomsResponseToNameSpaceDenoms(response) {
		return response.denoms.map((denom) => denom);
	}
	static namespaceResponseToNamespaces(response) {
		if (!response.namespace) return;
		return ChainGrpcPermissionsTransformer.grpcNamespaceToNamespace(response.namespace);
	}
	static namespacesResponseToNamespaces(response) {
		return response.namespaces.map((namespace) => {
			ChainGrpcPermissionsTransformer.grpcNamespaceToNamespace(namespace);
		});
	}
	static actorsByRoleResponseToActorsByRole(response) {
		return response.actors.map((role) => {
			return { roles: role };
		});
	}
	static rolesByActorResponseToRolesByActor(response) {
		return response.roles.map((role) => {
			return { roles: role };
		});
	}
	static roleManagerResponseToRoleManager(response) {
		if (!response.roleManager) return;
		return ChainGrpcPermissionsTransformer.grpcRoleManagerToRoleManager(response.roleManager);
	}
	static roleManagersResponseToRoleManagers(response) {
		return response.roleManagers.map(ChainGrpcPermissionsTransformer.grpcRoleManagerToRoleManager);
	}
	static policyStatusesResponseToPolicyStatuses(response) {
		return response.policyStatuses.map(ChainGrpcPermissionsTransformer.grpcPolicyStatusToPolicyStatus);
	}
	static policyManagerCapabilitiesResponseToPolicyManagerCapabilities(response) {
		return response.policyManagerCapabilities.map(ChainGrpcPermissionsTransformer.grpcPolicyManagerCapabilityToPolicyManagerCapability);
	}
	static voucherResponseToVoucher(response) {
		if (!response.voucher) return;
		return require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin(response.voucher);
	}
	static vouchersResponseToVouchers(response) {
		return response.vouchers.map(ChainGrpcPermissionsTransformer.grpcAddressVoucherToAddressVoucher);
	}
	static moduleStateResponseToModuleState(response) {
		if (!response.state) return;
		return {
			params: ChainGrpcPermissionsTransformer.moduleParamsResponseToModuleParams(response.state),
			namespaces: response.state.namespaces.map(ChainGrpcPermissionsTransformer.grpcNamespaceToNamespace),
			vouchers: response.state.vouchers.map(ChainGrpcPermissionsTransformer.grpcAddressVoucherToAddressVoucher)
		};
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcDistributionTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcDistributionTransformer = class {
	static moduleParamsResponseToModuleParams(response) {
		const params = response.params;
		return {
			communityTax: (0, __injectivelabs_utils.toHumanReadable)(params.communityTax).toFixed(),
			baseProposerReward: params.baseProposerReward,
			bonusProposerReward: params.bonusProposerReward,
			withdrawAddrEnabled: params.withdrawAddrEnabled
		};
	}
	static delegationRewardResponseToReward(response) {
		return response.rewards.map((grpcReward) => {
			return {
				amount: (0, __injectivelabs_utils.toHumanReadable)(grpcReward.amount).toFixed(),
				denom: grpcReward.denom
			};
		});
	}
	static totalDelegationRewardResponseToTotalReward(response) {
		return response.rewards.map((grpcReward) => {
			return {
				rewards: grpcReward.reward.map((reward) => ({
					amount: (0, __injectivelabs_utils.toHumanReadable)(reward.amount).toFixed(),
					denom: reward.denom
				})),
				validatorAddress: grpcReward.validatorAddress
			};
		});
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcTokenFactoryTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcTokenFactoryTransformer = class {
	static moduleParamsResponseToModuleParams(response) {
		return { denomCreationFee: response.params.denomCreationFee };
	}
	static moduleStateResponseToModuleState(response) {
		const state = response.state;
		return {
			denomCreationFee: state.params.denomCreationFee,
			factoryDenoms: state.factoryDenoms.map((item) => ({
				denom: item.denom,
				authorityMetadata: item.authorityMetadata
			}))
		};
	}
	static denomsCreatorResponseToDenomsString(response) {
		return response.denoms;
	}
	static authorityMetadataResponseToAuthorityMetadata(response) {
		return response.authorityMetadata;
	}
};

//#endregion
//#region src/client/chain/transformers/ChainGrpcInsuranceFundTransformer.ts
/**
* @category Chain Grpc Transformer
*/
var ChainGrpcInsuranceFundTransformer = class ChainGrpcInsuranceFundTransformer {
	static moduleParamsResponseToModuleParams(response) {
		const params = response.params;
		return { defaultRedemptionNoticePeriodDuration: Number(params.defaultRedemptionNoticePeriodDuration || 0n) };
	}
	static grpcInsuranceFundToInsuranceFund(grpcFund) {
		return {
			depositDenom: grpcFund.depositDenom,
			insurancePoolTokenDenom: grpcFund.insurancePoolTokenDenom,
			redemptionNoticePeriodDuration: Number(grpcFund.redemptionNoticePeriodDuration || 0n),
			balance: grpcFund.balance,
			totalShare: grpcFund.totalShare,
			marketId: grpcFund.marketId,
			marketTicker: grpcFund.marketTicker,
			oracleBase: grpcFund.oracleBase,
			oracleQuote: grpcFund.oracleQuote,
			oracleType: grpcFund.oracleType,
			expiry: Number(grpcFund.expiry)
		};
	}
	static insuranceFundResponseToInsuranceFund(response) {
		return ChainGrpcInsuranceFundTransformer.grpcInsuranceFundToInsuranceFund(response.fund);
	}
	static insuranceFundsResponseToInsuranceFunds(response) {
		return response.funds.map((fund) => ChainGrpcInsuranceFundTransformer.grpcInsuranceFundToInsuranceFund(fund));
	}
	static redemptionsResponseToRedemptions(response) {
		return response.amount.map((amount) => {
			return {
				amount: amount.amount,
				denom: amount.denom
			};
		});
	}
	static estimatedRedemptionsResponseToEstimatedRedemptions(response) {
		return response.amount.map((amount) => {
			return {
				amount: amount.amount,
				denom: amount.denom
			};
		});
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcGovApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcGovApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Gov);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const requests = [
			"voting",
			"deposit",
			"tallying"
		].map((type) => {
			const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb.QueryParamsRequest.create();
			request.paramsType = type;
			return request;
		});
		const [votingParams, depositParams, tallyParams] = await Promise.all(requests.map((request) => this.executeGrpcCall(request, this.client.params.bind(this.client))));
		return ChainGrpcGovTransformer.moduleParamsResponseToModuleParamsByType({
			votingParams: votingParams.params,
			tallyParams: tallyParams.params,
			depositParams: depositParams.params
		});
	}
	async fetchProposals({ status, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb.QueryProposalsRequest.create();
		request.proposalStatus = status;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.proposals.bind(this.client));
		return ChainGrpcGovTransformer.proposalsResponseToProposals(response);
	}
	async fetchProposal(proposalId) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb.QueryProposalRequest.create();
		request.proposalId = BigInt(proposalId);
		const response = await this.executeGrpcCall(request, this.client.proposal.bind(this.client));
		return ChainGrpcGovTransformer.proposalResponseToProposal(response);
	}
	async fetchProposalDeposits({ proposalId, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb.QueryDepositsRequest.create();
		request.proposalId = BigInt(proposalId);
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.deposits.bind(this.client));
		return ChainGrpcGovTransformer.depositsResponseToDeposits(response);
	}
	async fetchProposalVotes({ proposalId, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb.QueryVotesRequest.create();
		request.proposalId = BigInt(proposalId);
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.votes.bind(this.client));
		return ChainGrpcGovTransformer.votesResponseToVotes(response);
	}
	async fetchProposalTally(proposalId) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_gov_v1_query_pb.QueryTallyResultRequest.create();
		request.proposalId = BigInt(proposalId);
		const response = await this.executeGrpcCall(request, this.client.tallyResult.bind(this.client));
		return ChainGrpcGovTransformer.tallyResultResponseToTallyResult(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcIbcApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcIbcApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Ibc);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_ibc_applications_transfer_v1_query_pb_client.QueryClient);
	}
	async fetchDenomTrace(hash) {
		const request = __injectivelabs_core_proto_ts_v2_generated_ibc_applications_transfer_v1_query_pb.QueryDenomTraceRequest.create();
		request.hash = hash;
		return (await this.executeGrpcCall(request, this.client.denomTrace.bind(this.client))).denomTrace;
	}
	async fetchDenomsTrace(pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_ibc_applications_transfer_v1_query_pb.QueryDenomTracesRequest.create();
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		return (await this.executeGrpcCall(request, this.client.denomTraces.bind(this.client))).denomTraces;
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcEvmApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcEvmApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Evm);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb_client.QueryClient);
	}
	async fetchAccount(ethAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb.QueryAccountRequest.create();
		request.address = ethAddress;
		const response = await this.executeGrpcCall(request, this.client.account.bind(this.client));
		return ChainGrpcEvmTransformer.accountResponseToAccount(response);
	}
	async fetchCosmosAccount(ethAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb.QueryCosmosAccountRequest.create();
		request.address = ethAddress;
		const response = await this.executeGrpcCall(request, this.client.cosmosAccount.bind(this.client));
		return ChainGrpcEvmTransformer.cosmosAccountResponseToCosmosAccount(response);
	}
	async fetchValidatorAccount(consAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb.QueryValidatorAccountRequest.create();
		request.consAddress = consAddress;
		const response = await this.executeGrpcCall(request, this.client.validatorAccount.bind(this.client));
		return ChainGrpcEvmTransformer.validatorAccountResponseToValidatorAccount(response);
	}
	async fetchBalance(ethAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb.QueryBalanceRequest.create();
		request.address = ethAddress;
		return (await this.executeGrpcCall(request, this.client.balance.bind(this.client))).balance;
	}
	async fetchStorage(ethAddress, key) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb.QueryStorageRequest.create();
		request.address = ethAddress;
		request.key = key;
		return (await this.executeGrpcCall(request, this.client.storage.bind(this.client))).value;
	}
	async fetchCode(ethAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb.QueryCodeRequest.create();
		request.address = ethAddress;
		return (await this.executeGrpcCall(request, this.client.code.bind(this.client))).code;
	}
	async fetchParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcEvmTransformer.paramsResponseToParams(response);
	}
	async fetchBaseFee() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_evm_v1_query_pb.QueryBaseFeeRequest.create();
		return (await this.executeGrpcCall(request, this.client.baseFee.bind(this.client))).baseFee;
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcBankApi.ts
const MAX_LIMIT_FOR_SUPPLY$1 = 1e4;
/**
* @category Chain Grpc API
*/
var ChainGrpcBankApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Bank);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcBankTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchBalance({ accountAddress, denom }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb.QueryBalanceRequest.create();
		request.address = accountAddress;
		request.denom = denom;
		const response = await this.executeGrpcCall(request, this.client.balance.bind(this.client));
		return ChainGrpcBankTransformer.balanceResponseToBalance(response);
	}
	async fetchBalances(address, pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb.QueryAllBalancesRequest.create();
		request.address = address;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.allBalances.bind(this.client));
		return ChainGrpcBankTransformer.balancesResponseToBalances(response);
	}
	async fetchTotalSupply(pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb.QueryTotalSupplyRequest.create();
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.totalSupply.bind(this.client));
		return ChainGrpcBankTransformer.totalSupplyResponseToTotalSupply(response);
	}
	/** a way to ensure all total supply is fully fetched */
	async fetchAllTotalSupply(pagination = { limit: MAX_LIMIT_FOR_SUPPLY$1 }) {
		return require_utils.fetchAllWithPagination(pagination, this.fetchTotalSupply.bind(this));
	}
	async fetchSupplyOf(denom) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb.QuerySupplyOfRequest.create();
		request.denom = denom;
		const response = await this.executeGrpcCall(request, this.client.supplyOf.bind(this.client));
		return require_utils.ChainGrpcCommonTransformer.grpcCoinToCoin(response.amount);
	}
	async fetchDenomsMetadata(pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb.QueryDenomsMetadataRequest.create();
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.denomsMetadata.bind(this.client));
		return ChainGrpcBankTransformer.denomsMetadataResponseToDenomsMetadata(response);
	}
	async fetchDenomMetadata(denom) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb.QueryDenomMetadataRequest.create();
		request.denom = denom;
		const response = await this.executeGrpcCall(request, this.client.denomMetadata.bind(this.client));
		return ChainGrpcBankTransformer.metadataToMetadata(response.metadata);
	}
	async fetchDenomOwners(denom, pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_bank_v1beta1_query_pb.QueryDenomOwnersRequest.create();
		request.denom = denom;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.denomOwners.bind(this.client));
		return ChainGrpcBankTransformer.denomOwnersResponseToDenomOwners(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcMintApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcMintApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Mint);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_cosmos_mint_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_mint_v1beta1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcMintTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchInflation() {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_mint_v1beta1_query_pb.QueryInflationRequest.create();
		return { inflation: (0, __injectivelabs_utils.toHumanReadable)(require_utils.uint8ArrayToString((await this.executeGrpcCall(request, this.client.inflation.bind(this.client))).inflation)) };
	}
	async fetchAnnualProvisions() {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_mint_v1beta1_query_pb.QueryAnnualProvisionsRequest.create();
		return { annualProvisions: (0, __injectivelabs_utils.toHumanReadable)(require_utils.uint8ArrayToString((await this.executeGrpcCall(request, this.client.annualProvisions.bind(this.client))).annualProvisions)).toFixed() };
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcWasmApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcWasmApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Wasm);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb_client.QueryClient);
	}
	async fetchContractAccountsBalance({ contractAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb.QueryAllContractStateRequest.create();
		request.address = contractAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.allContractState.bind(this.client));
		return ChainGrpcWasmTransformer.allContractStateResponseToContractAccountsBalanceWithPagination(response);
	}
	async fetchContractState({ contractAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb.QueryAllContractStateRequest.create();
		request.address = contractAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.allContractState.bind(this.client));
		return ChainGrpcWasmTransformer.allContractStateResponseToContractState(response);
	}
	async fetchContractInfo(contractAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb.QueryContractInfoRequest.create();
		request.address = contractAddress;
		const contractInfo = (await this.executeGrpcCall(request, this.client.contractInfo.bind(this.client))).contractInfo;
		if (!contractInfo) return;
		return ChainGrpcWasmTransformer.contactInfoResponseToContractInfo(contractInfo);
	}
	async fetchContractHistory(contractAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb.QueryContractHistoryRequest.create();
		request.address = contractAddress;
		const response = await this.executeGrpcCall(request, this.client.contractHistory.bind(this.client));
		return ChainGrpcWasmTransformer.contactHistoryResponseToContractHistory(response);
	}
	async fetchSmartContractState(contractAddress, query) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb.QuerySmartContractStateRequest.create();
		request.address = contractAddress;
		if (query) request.queryData = require_utils.base64ToUint8Array(typeof query === "string" ? query : require_utils.toBase64(query));
		return await this.executeGrpcCall(request, this.client.smartContractState.bind(this.client));
	}
	async fetchRawContractState(contractAddress, query) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb.QueryRawContractStateRequest.create();
		request.address = contractAddress;
		if (query) request.queryData = require_utils.base64ToUint8Array(query);
		return await this.executeGrpcCall(request, this.client.rawContractState.bind(this.client));
	}
	async fetchContractCodes(pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb.QueryCodesRequest.create();
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.codes.bind(this.client));
		return ChainGrpcWasmTransformer.contractCodesResponseToContractCodes(response);
	}
	async fetchContractCode(codeId) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb.QueryCodeRequest.create();
		request.codeId = BigInt(codeId);
		const response = await this.executeGrpcCall(request, this.client.code.bind(this.client));
		return ChainGrpcWasmTransformer.contractCodeResponseToContractCode(response);
	}
	async fetchContractCodeContracts(codeId, pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmwasm_wasm_v1_query_pb.QueryContractsByCodeRequest.create();
		request.codeId = BigInt(codeId);
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.contractsByCode.bind(this.client));
		return ChainGrpcWasmTransformer.contractByCodeResponseToContractByCode(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcAuthZApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcAuthZApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Authz);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_query_pb_client.QueryClient);
	}
	async fetchGrants({ pagination, granter, grantee, msgTypeUrl }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_query_pb.QueryGrantsRequest.create();
		if (granter) request.granter = granter;
		if (grantee) request.grantee = grantee;
		if (msgTypeUrl) request.msgTypeUrl = msgTypeUrl;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.grants.bind(this.client));
		return ChainGrpcAuthZTransformer.grpcGrantsToGrants(response);
	}
	async fetchGranterGrants(granter, pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_query_pb.QueryGranterGrantsRequest.create();
		if (granter) request.granter = granter;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.granterGrants.bind(this.client));
		return ChainGrpcAuthZTransformer.grpcGranterGrantsToGranterGrants(response);
	}
	async fetchGranteeGrants(grantee, pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_authz_v1beta1_query_pb.QueryGranteeGrantsRequest.create();
		if (grantee) request.grantee = grantee;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.granteeGrants.bind(this.client));
		return ChainGrpcAuthZTransformer.grpcGranteeGrantsToGranteeGrants(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcPeggyApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcPeggyApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Peggy);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_peggy_v1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_peggy_v1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcPeggyTransformer.moduleParamsResponseToModuleParams(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcWasmXApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcWasmXApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.WasmX);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_query_pb.QueryWasmxParamsRequest.create();
		return await this.executeGrpcCall(request, this.client.wasmxParams.bind(this.client));
	}
	async fetchModuleState() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_query_pb.QueryModuleStateRequest.create();
		return (await this.executeGrpcCall(request, this.client.wasmxModuleState.bind(this.client))).state;
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcErc20Api.ts
const MAX_LIMIT_FOR_SUPPLY = 1e4;
/**
* @category Chain Grpc API
*/
var ChainGrpcErc20Api = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Erc20);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_erc20_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_erc20_v1beta1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcErc20Transformer.paramsResponseToParams(response);
	}
	async fetchTokenPairs(pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_erc20_v1beta1_query_pb.QueryAllTokenPairsRequest.create();
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.allTokenPairs.bind(this.client));
		return ChainGrpcErc20Transformer.tokenPairsResponseToTokenPairs(response);
	}
	async fetchAllTokenPairsWithPagination(pagination = { limit: MAX_LIMIT_FOR_SUPPLY }) {
		return require_utils.fetchAllWithPagination(pagination, this.fetchTokenPairs.bind(this));
	}
	async fetchTokenPairByDenom(denom) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_erc20_v1beta1_query_pb.QueryTokenPairByDenomRequest.create();
		request.bankDenom = denom;
		const response = await this.executeGrpcCall(request, this.client.tokenPairByDenom.bind(this.client));
		if (!response.tokenPair) return;
		return ChainGrpcErc20Transformer.grpcTokenPairToTokenPair(response.tokenPair);
	}
	async fetchTokenPairByErc20Address(erc20Address) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_erc20_v1beta1_query_pb.QueryTokenPairByERC20AddressRequest.create();
		request.erc20Address = erc20Address;
		const response = await this.executeGrpcCall(request, this.client.tokenPairByERC20Address.bind(this.client));
		if (!response.tokenPair) return;
		return ChainGrpcErc20Transformer.grpcTokenPairToTokenPair(response.tokenPair);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcOracleApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcOracleApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Oracle);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_oracle_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_oracle_v1beta1_query_pb.QueryParamsRequest.create();
		return (await this.executeGrpcCall(request, this.client.params.bind(this.client))).params;
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcTxFeesApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcTxFeesApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.TxFees);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_txfees_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_txfees_v1beta1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcTxFeesTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchEipBaseFee() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_txfees_v1beta1_query_pb.QueryEipBaseFeeRequest.create();
		const response = await this.executeGrpcCall(request, this.client.getEipBaseFee.bind(this.client));
		return ChainGrpcTxFeesTransformer.eipBaseFeeResponseToEipBaseFee(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcAuctionApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcAuctionApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Auction);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_auction_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_auction_v1beta1_query_pb.QueryAuctionParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.auctionParams.bind(this.client));
		return ChainGrpcAuctionTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchCurrentBasket() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_auction_v1beta1_query_pb.QueryCurrentAuctionBasketRequest.create();
		const response = await this.executeGrpcCall(request, this.client.currentAuctionBasket.bind(this.client));
		return ChainGrpcAuctionTransformer.currentBasketResponseToCurrentBasket(response);
	}
	async fetchModuleState() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_auction_v1beta1_query_pb.QueryModuleStateRequest.create();
		const response = await this.executeGrpcCall(request, this.client.auctionModuleState.bind(this.client));
		return ChainGrpcAuctionTransformer.auctionModuleStateResponseToAuctionModuleState(response);
	}
	async fetchLastAuctionResult() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_auction_v1beta1_query_pb.QueryLastAuctionResultRequest.create();
		const response = await this.executeGrpcCall(request, this.client.lastAuctionResult.bind(this.client));
		return ChainGrpcAuctionTransformer.lastAuctionResultResponseToLastAuctionResult(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcStakingApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcStakingApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Staking);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcStakingTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchPool() {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryPoolRequest.create();
		const response = await this.executeGrpcCall(request, this.client.pool.bind(this.client));
		return ChainGrpcStakingTransformer.poolResponseToPool(response);
	}
	async fetchValidators(pagination) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryValidatorsRequest.create();
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.validators.bind(this.client));
		return ChainGrpcStakingTransformer.validatorsResponseToValidators(response);
	}
	async fetchValidator(address) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryValidatorRequest.create();
		request.validatorAddr = address;
		const response = await this.executeGrpcCall(request, this.client.validator.bind(this.client));
		return ChainGrpcStakingTransformer.validatorResponseToValidator(response);
	}
	async fetchValidatorDelegations({ validatorAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryValidatorDelegationsRequest.create();
		request.validatorAddr = validatorAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.validatorDelegations.bind(this.client));
		return ChainGrpcStakingTransformer.delegationsResponseToDelegations(response);
	}
	async fetchValidatorDelegationsNoThrow({ validatorAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryValidatorDelegationsRequest.create();
		request.validatorAddr = validatorAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		try {
			const response = await this.executeGrpcCall(request, this.client.validatorDelegations.bind(this.client));
			return ChainGrpcStakingTransformer.delegationsResponseToDelegations(response);
		} catch (e) {
			if (e.message.includes("does not exist")) return {
				delegations: [],
				pagination: {
					total: 0,
					next: ""
				}
			};
			throw e;
		}
	}
	async fetchValidatorUnbondingDelegations({ validatorAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryValidatorUnbondingDelegationsRequest.create();
		request.validatorAddr = validatorAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.validatorUnbondingDelegations.bind(this.client));
		return ChainGrpcStakingTransformer.unBondingDelegationsResponseToUnBondingDelegations(response);
	}
	async fetchValidatorUnbondingDelegationsNoThrow({ validatorAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryValidatorUnbondingDelegationsRequest.create();
		request.validatorAddr = validatorAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		try {
			const response = await this.executeGrpcCall(request, this.client.validatorUnbondingDelegations.bind(this.client));
			return ChainGrpcStakingTransformer.unBondingDelegationsResponseToUnBondingDelegations(response);
		} catch (e) {
			if (e.message.includes("does not exist")) return {
				unbondingDelegations: [],
				pagination: {
					total: 0,
					next: ""
				}
			};
			throw e;
		}
	}
	async fetchDelegation({ injectiveAddress, validatorAddress }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryDelegationRequest.create();
		request.delegatorAddr = injectiveAddress;
		request.validatorAddr = validatorAddress;
		const response = await this.executeGrpcCall(request, this.client.delegation.bind(this.client));
		return ChainGrpcStakingTransformer.delegationResponseToDelegation(response);
	}
	async fetchDelegations({ injectiveAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryDelegatorDelegationsRequest.create();
		request.delegatorAddr = injectiveAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.delegatorDelegations.bind(this.client));
		return ChainGrpcStakingTransformer.delegationsResponseToDelegations(response);
	}
	async fetchDelegationsNoThrow({ injectiveAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryDelegatorDelegationsRequest.create();
		request.delegatorAddr = injectiveAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		try {
			const response = await this.executeGrpcCall(request, this.client.delegatorDelegations.bind(this.client));
			return ChainGrpcStakingTransformer.delegationsResponseToDelegations(response);
		} catch (e) {
			if (e.message.includes("does not exist")) return {
				delegations: [],
				pagination: {
					total: 0,
					next: ""
				}
			};
			throw e;
		}
	}
	async fetchDelegators({ validatorAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryValidatorDelegationsRequest.create();
		request.validatorAddr = validatorAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.validatorDelegations.bind(this.client));
		return ChainGrpcStakingTransformer.delegationsResponseToDelegations(response);
	}
	async fetchDelegatorsNoThrow({ validatorAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryValidatorDelegationsRequest.create();
		request.validatorAddr = validatorAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		try {
			const response = await this.executeGrpcCall(request, this.client.validatorDelegations.bind(this.client));
			return ChainGrpcStakingTransformer.delegationsResponseToDelegations(response);
		} catch (e) {
			if (e.message.includes("does not exist")) return {
				delegations: [],
				pagination: {
					total: 0,
					next: ""
				}
			};
			throw e;
		}
	}
	async fetchUnbondingDelegations({ injectiveAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryDelegatorUnbondingDelegationsRequest.create();
		request.delegatorAddr = injectiveAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.delegatorUnbondingDelegations.bind(this.client));
		return ChainGrpcStakingTransformer.unBondingDelegationsResponseToUnBondingDelegations(response);
	}
	async fetchUnbondingDelegationsNoThrow({ injectiveAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryDelegatorUnbondingDelegationsRequest.create();
		request.delegatorAddr = injectiveAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		try {
			const response = await this.executeGrpcCall(request, this.client.delegatorUnbondingDelegations.bind(this.client));
			return ChainGrpcStakingTransformer.unBondingDelegationsResponseToUnBondingDelegations(response);
		} catch (e) {
			if (e.message.includes("does not exist")) return {
				unbondingDelegations: [],
				pagination: {
					total: 0,
					next: ""
				}
			};
			throw e;
		}
	}
	async fetchReDelegations({ injectiveAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryRedelegationsRequest.create();
		request.delegatorAddr = injectiveAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		const response = await this.executeGrpcCall(request, this.client.redelegations.bind(this.client));
		return ChainGrpcStakingTransformer.reDelegationsResponseToReDelegations(response);
	}
	async fetchReDelegationsNoThrow({ injectiveAddress, pagination }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_staking_v1beta1_query_pb.QueryRedelegationsRequest.create();
		request.delegatorAddr = injectiveAddress;
		const paginationForRequest = require_utils.ChainGrpcCommonTransformer.pageRequestToGrpcPageRequestV2(pagination);
		if (paginationForRequest) request.pagination = paginationForRequest;
		try {
			const response = await this.executeGrpcCall(request, this.client.redelegations.bind(this.client));
			return ChainGrpcStakingTransformer.reDelegationsResponseToReDelegations(response);
		} catch (e) {
			if (e.message.includes("does not exist")) return {
				redelegations: [],
				pagination: {
					total: 0,
					next: ""
				}
			};
			throw e;
		}
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcPermissionsApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcPermissionsApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Permissions);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcPermissionsTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchNamespaceDenoms() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryNamespaceDenomsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.namespaceDenoms.bind(this.client));
		return ChainGrpcPermissionsTransformer.nameSpaceDenomsResponseToNameSpaceDenoms(response);
	}
	async fetchNamespaces() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryNamespacesRequest.create();
		const response = await this.executeGrpcCall(request, this.client.namespaces.bind(this.client));
		return ChainGrpcPermissionsTransformer.namespacesResponseToNamespaces(response);
	}
	async fetchNamespace(denom) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryNamespaceRequest.create();
		request.denom = denom;
		const response = await this.executeGrpcCall(request, this.client.namespace.bind(this.client));
		return ChainGrpcPermissionsTransformer.namespaceResponseToNamespaces(response);
	}
	async fetchActorsByRole({ denom, role }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryActorsByRoleRequest.create();
		request.denom = denom;
		request.role = role;
		const response = await this.executeGrpcCall(request, this.client.actorsByRole.bind(this.client));
		return ChainGrpcPermissionsTransformer.actorsByRoleResponseToActorsByRole(response);
	}
	async fetchRolesByActor({ actor, denom }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryRolesByActorRequest.create();
		request.actor = actor;
		request.denom = denom;
		const response = await this.executeGrpcCall(request, this.client.rolesByActor.bind(this.client));
		return ChainGrpcPermissionsTransformer.rolesByActorResponseToRolesByActor(response);
	}
	async fetchRoleManager({ denom, manager }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryRoleManagerRequest.create();
		request.denom = denom;
		request.manager = manager;
		const response = await this.executeGrpcCall(request, this.client.roleManager.bind(this.client));
		return ChainGrpcPermissionsTransformer.roleManagerResponseToRoleManager(response);
	}
	async fetchRoleManagers() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryRoleManagersRequest.create();
		const response = await this.executeGrpcCall(request, this.client.roleManagers.bind(this.client));
		return ChainGrpcPermissionsTransformer.roleManagersResponseToRoleManagers(response);
	}
	async fetchPolicyStatuses() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryPolicyStatusesRequest.create();
		const response = await this.executeGrpcCall(request, this.client.policyStatuses.bind(this.client));
		return ChainGrpcPermissionsTransformer.policyStatusesResponseToPolicyStatuses(response);
	}
	async fetchPolicyManagerCapabilities(denom) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryPolicyManagerCapabilitiesRequest.create();
		request.denom = denom;
		const response = await this.executeGrpcCall(request, this.client.policyManagerCapabilities.bind(this.client));
		return ChainGrpcPermissionsTransformer.policyManagerCapabilitiesResponseToPolicyManagerCapabilities(response);
	}
	async fetchVoucher({ denom, address }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryVoucherRequest.create();
		request.denom = denom;
		request.address = address;
		const response = await this.executeGrpcCall(request, this.client.voucher.bind(this.client));
		return ChainGrpcPermissionsTransformer.voucherResponseToVoucher(response);
	}
	async fetchVouchers(denom) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryVouchersRequest.create();
		request.denom = denom;
		const response = await this.executeGrpcCall(request, this.client.vouchers.bind(this.client));
		return ChainGrpcPermissionsTransformer.vouchersResponseToVouchers(response);
	}
	async fetchModuleState() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_permissions_v1beta1_query_pb.QueryModuleStateRequest.create();
		const response = await this.executeGrpcCall(request, this.client.permissionsModuleState.bind(this.client));
		return ChainGrpcPermissionsTransformer.moduleStateResponseToModuleState(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcDistributionApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcDistributionApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.Distribution);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcDistributionTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchDelegatorRewardsForValidator({ delegatorAddress, validatorAddress }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb.QueryDelegationRewardsRequest.create();
		request.validatorAddress = validatorAddress;
		request.delegatorAddress = delegatorAddress;
		const response = await this.executeGrpcCall(request, this.client.delegationRewards.bind(this.client));
		return ChainGrpcDistributionTransformer.delegationRewardResponseToReward(response);
	}
	async fetchDelegatorRewardsForValidatorNoThrow({ delegatorAddress, validatorAddress }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb.QueryDelegationRewardsRequest.create();
		request.validatorAddress = validatorAddress;
		request.delegatorAddress = delegatorAddress;
		try {
			const response = await this.executeGrpcCall(request, this.client.delegationRewards.bind(this.client));
			return ChainGrpcDistributionTransformer.delegationRewardResponseToReward(response);
		} catch (e) {
			if (e.message.includes("does not exist") || e.message.includes("no delegation for (address, validator) tuple")) return [];
			throw e;
		}
	}
	async fetchDelegatorRewards(injectiveAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb.QueryDelegationTotalRewardsRequest.create();
		request.delegatorAddress = injectiveAddress;
		const response = await this.executeGrpcCall(request, this.client.delegationTotalRewards.bind(this.client));
		return ChainGrpcDistributionTransformer.totalDelegationRewardResponseToTotalReward(response);
	}
	async fetchDelegatorRewardsNoThrow(injectiveAddress) {
		const request = __injectivelabs_core_proto_ts_v2_generated_cosmos_distribution_v1beta1_query_pb.QueryDelegationTotalRewardsRequest.create();
		request.delegatorAddress = injectiveAddress;
		try {
			const response = await this.executeGrpcCall(request, this.client.delegationTotalRewards.bind(this.client));
			return ChainGrpcDistributionTransformer.totalDelegationRewardResponseToTotalReward(response);
		} catch (e) {
			if (e.message.includes("does not exist")) return [];
			throw e;
		}
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcTokenFactoryApi.ts
/**
* @category TokenFactory Grpc API
*/
var ChainGrpcTokenFactoryApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.WasmX);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_tokenfactory_v1beta1_query_pb_client.QueryClient);
	}
	async fetchDenomsFromCreator(creator) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_tokenfactory_v1beta1_query_pb.QueryDenomsFromCreatorRequest.create();
		request.creator = creator;
		const response = await this.executeGrpcCall(request, this.client.denomsFromCreator.bind(this.client));
		return ChainGrpcTokenFactoryTransformer.denomsCreatorResponseToDenomsString(response);
	}
	async fetchDenomAuthorityMetadata(creator, subDenom) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_tokenfactory_v1beta1_query_pb.QueryDenomAuthorityMetadataRequest.create();
		request.creator = creator;
		request.subDenom = subDenom;
		const response = await this.executeGrpcCall(request, this.client.denomAuthorityMetadata.bind(this.client));
		return ChainGrpcTokenFactoryTransformer.authorityMetadataResponseToAuthorityMetadata(response);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_tokenfactory_v1beta1_query_pb.QueryParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.params.bind(this.client));
		return ChainGrpcTokenFactoryTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchModuleState() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_tokenfactory_v1beta1_query_pb.QueryModuleStateRequest.create();
		const response = await this.executeGrpcCall(request, this.client.tokenfactoryModuleState.bind(this.client));
		return ChainGrpcTokenFactoryTransformer.moduleStateResponseToModuleState(response);
	}
};

//#endregion
//#region src/client/chain/grpc/ChainGrpcInsuranceFundApi.ts
/**
* @category Chain Grpc API
*/
var ChainGrpcInsuranceFundApi = class extends require_BaseGrpcConsumer.BaseGrpcConsumer {
	constructor(..._args) {
		super(..._args);
		require_defineProperty._defineProperty(this, "module", require_ChainRestTendermintApi.ChainModule.InsuranceFund);
	}
	get client() {
		return this.initClient(__injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb_client.QueryClient);
	}
	async fetchModuleParams() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb.QueryInsuranceParamsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.insuranceParams.bind(this.client));
		return ChainGrpcInsuranceFundTransformer.moduleParamsResponseToModuleParams(response);
	}
	async fetchInsuranceFunds() {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb.QueryInsuranceFundsRequest.create();
		const response = await this.executeGrpcCall(request, this.client.insuranceFunds.bind(this.client));
		return ChainGrpcInsuranceFundTransformer.insuranceFundsResponseToInsuranceFunds(response);
	}
	async fetchInsuranceFund(marketId) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb.QueryInsuranceFundRequest.create();
		request.marketId = marketId;
		const response = await this.executeGrpcCall(request, this.client.insuranceFund.bind(this.client));
		return ChainGrpcInsuranceFundTransformer.insuranceFundResponseToInsuranceFund(response);
	}
	async fetchEstimatedRedemptions({ marketId, address }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb.QueryEstimatedRedemptionsRequest.create();
		request.marketId = marketId;
		request.address = address;
		const response = await this.executeGrpcCall(request, this.client.estimatedRedemptions.bind(this.client));
		return ChainGrpcInsuranceFundTransformer.estimatedRedemptionsResponseToEstimatedRedemptions(response);
	}
	async fetchPendingRedemptions({ marketId, address }) {
		const request = __injectivelabs_core_proto_ts_v2_generated_injective_insurance_v1beta1_query_pb.QueryPendingRedemptionsRequest.create();
		request.marketId = marketId;
		request.address = address;
		const response = await this.executeGrpcCall(request, this.client.pendingRedemptions.bind(this.client));
		return ChainGrpcInsuranceFundTransformer.redemptionsResponseToRedemptions(response);
	}
};

//#endregion
//#region src/client/chain/rest/ChainRestBankApi.ts
/**
* @category Chain Rest API
*/
var ChainRestBankApi = class extends require_BaseRestConsumer.BaseRestConsumer {
	/**
	* Get address's balance
	*
	* @param address address of account to look up
	*/
	async fetchBalances(address, params = {}) {
		const endpoint = `cosmos/bank/v1beta1/balances/${address}`;
		try {
			return (await this.retry(() => this.get(endpoint, params))).data;
		} catch (e) {
			if (e instanceof __injectivelabs_exceptions.HttpRequestException) throw e;
			throw new __injectivelabs_exceptions.HttpRequestException(new Error(e), {
				code: __injectivelabs_exceptions.UnspecifiedErrorCode,
				context: `${this.endpoint}/${endpoint}`,
				contextModule: require_ChainRestTendermintApi.ChainModule.Bank
			});
		}
	}
	/**
	* Get address's balances
	*
	* @param address address of account to look up
	*/
	async fetchBalance(address, denom, params = {}) {
		const endpoint = `cosmos/bank/v1beta1/balances/${address}`;
		try {
			const balance = (await this.retry(() => this.get(endpoint, params))).data.balances.find((balance$1) => balance$1.denom === denom);
			if (!balance) throw new __injectivelabs_exceptions.GeneralException(/* @__PURE__ */ new Error(`The ${denom} balance was not found`), {
				code: http_status_codes.StatusCodes.NOT_FOUND,
				type: __injectivelabs_exceptions.ErrorType.NotFoundError
			});
			return balance;
		} catch (e) {
			if (e instanceof __injectivelabs_exceptions.HttpRequestException) throw e;
			if (e instanceof __injectivelabs_exceptions.GeneralException) throw e;
			throw new __injectivelabs_exceptions.HttpRequestException(new Error(e), {
				code: __injectivelabs_exceptions.UnspecifiedErrorCode,
				context: `${this.endpoint}/${endpoint}`,
				contextModule: require_ChainRestTendermintApi.ChainModule.Bank
			});
		}
	}
	async fetchDenomOwners(denom, params = {}) {
		const endpoint = `cosmos/bank/v1beta1/denom_owners/${denom}`;
		try {
			return (await this.retry(() => this.get(endpoint, params))).data.denom_owners;
		} catch (e) {
			if (e instanceof __injectivelabs_exceptions.HttpRequestException) throw e;
			if (e instanceof __injectivelabs_exceptions.GeneralException) throw e;
			throw new __injectivelabs_exceptions.HttpRequestException(new Error(e), {
				code: __injectivelabs_exceptions.UnspecifiedErrorCode,
				context: `${this.endpoint}/${endpoint}`,
				contextModule: require_ChainRestTendermintApi.ChainModule.Bank
			});
		}
	}
};

//#endregion
//#region src/client/chain/rest/ChainRestWasmApi.ts
/**
* @category Chain Wasm API
*/
var ChainRestWasmApi = class extends require_BaseRestConsumer.BaseRestConsumer {
	async fetchSmartContractState(contractAddress, query, params = {}) {
		const endpoint = `cosmwasm/wasm/v1/contract/${contractAddress}/smart/${query}`;
		try {
			return (await this.retry(() => this.get(endpoint, params))).data;
		} catch (e) {
			if (e instanceof __injectivelabs_exceptions.HttpRequestException) throw e;
			throw new __injectivelabs_exceptions.HttpRequestException(new Error(e), {
				code: __injectivelabs_exceptions.UnspecifiedErrorCode,
				context: `${this.endpoint}/${endpoint}`,
				contextModule: require_ChainRestTendermintApi.ChainModule.Bank
			});
		}
	}
};

//#endregion
Object.defineProperty(exports, 'ChainGrpcAuctionApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcAuctionApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcAuctionTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcAuctionTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcAuthZApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcAuthZApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcAuthZTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcAuthZTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcBankApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcBankApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcBankTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcBankTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcDistributionApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcDistributionApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcDistributionTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcDistributionTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcErc20Api', {
  enumerable: true,
  get: function () {
    return ChainGrpcErc20Api;
  }
});
Object.defineProperty(exports, 'ChainGrpcErc20Transformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcErc20Transformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcEvmApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcEvmApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcEvmTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcEvmTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcGovApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcGovApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcGovTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcGovTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcIbcApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcIbcApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcInsuranceFundApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcInsuranceFundApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcInsuranceFundTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcInsuranceFundTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcMintApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcMintApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcMintTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcMintTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcOracleApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcOracleApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcPeggyApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcPeggyApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcPeggyTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcPeggyTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcPermissionsApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcPermissionsApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcPermissionsTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcPermissionsTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcStakingApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcStakingApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcStakingTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcStakingTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcTokenFactoryApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcTokenFactoryApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcTokenFactoryTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcTokenFactoryTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcTxFeesApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcTxFeesApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcTxFeesTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcTxFeesTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcWasmApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcWasmApi;
  }
});
Object.defineProperty(exports, 'ChainGrpcWasmTransformer', {
  enumerable: true,
  get: function () {
    return ChainGrpcWasmTransformer;
  }
});
Object.defineProperty(exports, 'ChainGrpcWasmXApi', {
  enumerable: true,
  get: function () {
    return ChainGrpcWasmXApi;
  }
});
Object.defineProperty(exports, 'ChainRestBankApi', {
  enumerable: true,
  get: function () {
    return ChainRestBankApi;
  }
});
Object.defineProperty(exports, 'ChainRestWasmApi', {
  enumerable: true,
  get: function () {
    return ChainRestWasmApi;
  }
});
