// @injectivelabs/sdk-ts/dist/cjs/ExecArgNeptuneWithdraw-4esQ6Su-.cjs
const require_chunk = require('./chunk-CbDLau6x.cjs');
const require_utils = require('./utils-D3XFGv8y.cjs');
const require_defineProperty = require('./defineProperty-dyr5BsvG.cjs');
const require_MsgBase = require('./MsgBase-D_1CCpXQ.cjs');
let __injectivelabs_exceptions = require("@injectivelabs/exceptions");
let __injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_tx_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/wasmx/v1/tx_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_tx_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_tx_pb);

//#region src/core/modules/wasm/ExecArgBase.ts
const dataToExecData = (action, data) => {
	return { [action]: data };
};
var ExecArgBase = class {
	constructor(params) {
		require_defineProperty._defineProperty(this, "params", void 0);
		this.params = params;
	}
	toJSON() {
		return require_utils.fromUtf8(require_utils.safeBigIntStringify(this.params));
	}
	toExecJSON() {
		return require_utils.fromUtf8(require_utils.safeBigIntStringify(this.toExecData()));
	}
};

//#endregion
//#region src/core/modules/wasm/exec-args/ExecArgNeptuneDeposit.ts
/**
* @category Contract Exec Arguments
*/
var ExecArgNeptuneDeposit = class ExecArgNeptuneDeposit extends ExecArgBase {
	static fromJSON(params) {
		return new ExecArgNeptuneDeposit(params);
	}
	toData() {
		return {};
	}
	toExecData() {
		return dataToExecData("lend", this.toData());
	}
};

//#endregion
//#region src/core/modules/wasm/msgs/MsgExecuteContractCompat.ts
/**
* @category Messages
*/
var MsgExecuteContractCompat = class MsgExecuteContractCompat extends require_MsgBase.MsgBase {
	static fromJSON(params) {
		return new MsgExecuteContractCompat(params);
	}
	toProto() {
		const { params } = this;
		const msg = this.getMsgObject();
		const funds = params.funds ? (Array.isArray(params.funds) ? params.funds : [params.funds]).map((coin) => `${coin.amount}${coin.denom}`).join(",") : "0";
		return __injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_tx_pb.MsgExecuteContractCompat.create({
			sender: params.sender,
			contract: params.contractAddress,
			msg: require_utils.safeBigIntStringify(msg),
			funds
		});
	}
	toData() {
		const proto = this.toProto();
		return {
			"@type": "/injective.wasmx.v1.MsgExecuteContractCompat",
			...proto
		};
	}
	toAmino() {
		const proto = this.toProto();
		return {
			type: "wasmx/MsgExecuteContractCompat",
			value: {
				sender: proto.sender,
				contract: proto.contract,
				msg: require_utils.safeBigIntStringify(this.getMsgObject()),
				funds: proto.funds
			}
		};
	}
	toWeb3Gw() {
		const { value } = this.toAmino();
		return {
			"@type": "/injective.wasmx.v1.MsgExecuteContractCompat",
			...value
		};
	}
	toDirectSign() {
		return {
			type: "/injective.wasmx.v1.MsgExecuteContractCompat",
			message: this.toProto()
		};
	}
	toBinary() {
		return __injectivelabs_core_proto_ts_v2_generated_injective_wasmx_v1_tx_pb.MsgExecuteContractCompat.toBinary(this.toProto());
	}
	getMsgObject() {
		const { params } = this;
		if ((params.exec || params.msg) && params.execArgs) throw new __injectivelabs_exceptions.GeneralException(/* @__PURE__ */ new Error("Please provide only one exec|msg argument"));
		if (params.execArgs) return params.execArgs.toExecData();
		if (params.exec) return { [params.exec.action]: params.exec.msg };
		if (params.msg) return params.msg;
		throw new __injectivelabs_exceptions.GeneralException(/* @__PURE__ */ new Error("Please provide at least one exec argument"));
	}
};

//#endregion
//#region src/core/modules/wasm/exec-args/ExecArgNeptuneWithdraw.ts
/**
* @category Contract Exec Arguments
*/
var ExecArgNeptuneWithdraw = class ExecArgNeptuneWithdraw extends ExecArgBase {
	static fromJSON(params) {
		return new ExecArgNeptuneWithdraw(params);
	}
	toData() {
		const { params } = this;
		const encodedMsg = require_utils.toBase64({ redeem: {} });
		return {
			amount: params.amount,
			contract: params.contract,
			msg: encodedMsg
		};
	}
	toExecData() {
		return dataToExecData("send", this.toData());
	}
};

//#endregion
Object.defineProperty(exports, 'ExecArgBase', {
  enumerable: true,
  get: function () {
    return ExecArgBase;
  }
});
Object.defineProperty(exports, 'ExecArgNeptuneDeposit', {
  enumerable: true,
  get: function () {
    return ExecArgNeptuneDeposit;
  }
});
Object.defineProperty(exports, 'ExecArgNeptuneWithdraw', {
  enumerable: true,
  get: function () {
    return ExecArgNeptuneWithdraw;
  }
});
Object.defineProperty(exports, 'MsgExecuteContractCompat', {
  enumerable: true,
  get: function () {
    return MsgExecuteContractCompat;
  }
});
Object.defineProperty(exports, 'dataToExecData', {
  enumerable: true,
  get: function () {
    return dataToExecData;
  }
});
