// @injectivelabs/sdk-ts/dist/cjs/types-BSrt7JzM.cjs
const require_chunk = require('./chunk-CbDLau6x.cjs');
let __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb = require("@injectivelabs/core-proto-ts-v2/generated/injective/exchange/v1beta1/exchange_pb");
__injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb = require_chunk.__toESM(__injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb);

//#region src/types/token.ts
const TokenType = {
	Ibc: "ibc",
	Cw20: "cw20",
	Spl: "spl",
	Erc20: "erc20",
	Lp: "lp",
	Evm: "evm",
	Native: "native",
	Symbol: "symbol",
	TokenFactory: "tokenFactory",
	InsuranceFund: "insuranceFund",
	Unknown: "unknown"
};
const TokenVerification = {
	Verified: "verified",
	Submitted: "submitted",
	Internal: "internal",
	External: "external",
	Unverified: "unverified"
};
const TokenSource = {
	Aptos: "aptos",
	Solana: "solana",
	Cosmos: "cosmos",
	Ethereum: "ethereum",
	EthereumWh: "ethereum-wormhole",
	Polygon: "polygon",
	Klaytn: "klaytn",
	Arbitrum: "arbitrum",
	Sui: "sui",
	Ibc: "ibc",
	BinanceSmartChain: "binance-smart-chain",
	Axelar: "axelar"
};

//#endregion
//#region src/types/stream.ts
/**
* Subscription interface for stream management with event-based lifecycle
*/
/**
* gRPC Status Codes
* @see https://grpc.io/docs/guides/status-codes/
*/
const GrpcStatusCode = {
	OK: 0,
	CANCELLED: 1,
	UNKNOWN: 2,
	INVALID_ARGUMENT: 3,
	DEADLINE_EXCEEDED: 4,
	NOT_FOUND: 5,
	ALREADY_EXISTS: 6,
	PERMISSION_DENIED: 7,
	RESOURCE_EXHAUSTED: 8,
	FAILED_PRECONDITION: 9,
	ABORTED: 10,
	OUT_OF_RANGE: 11,
	UNIMPLEMENTED: 12,
	INTERNAL: 13,
	UNAVAILABLE: 14,
	DATA_LOSS: 15,
	UNAUTHENTICATED: 16
};
/**
* Enhanced disconnect reasons with gRPC error code mapping
*/
const StreamDisconnectReason = {
	UserStopped: "user-stopped",
	StreamError: "stream-error",
	NetworkError: "network-error",
	Timeout: "timeout",
	AuthenticationError: "authentication-error",
	InvalidRequest: "invalid-request",
	RateLimited: "rate-limited",
	MaxRetries: "max-retries",
	StreamEnded: "stream-ended"
};
const StreamEvent = {
	Data: "data",
	Connect: "connect",
	Disconnect: "disconnect",
	Retry: "retry",
	StateChange: "state:change",
	Error: "error",
	Warn: "warn"
};
/**
* StreamManager
*/
const StreamState = {
	Idle: "idle",
	Connecting: "connecting",
	Connected: "connected",
	Reconnecting: "reconnecting",
	Stopped: "stopped"
};

//#endregion
//#region src/types/exchange.ts
/**
* Trade and Order types
*/
const TradeExecutionType = {
	Market: "market",
	LimitFill: "limitFill",
	LimitMatchRestingOrder: "limitMatchRestingOrder",
	LimitMatchNewOrder: "limitMatchNewOrder"
};
const TradeExecutionSide = {
	Maker: "maker",
	Taker: "taker"
};
const TradeDirection = {
	Buy: "buy",
	Sell: "sell",
	Long: "long",
	Short: "short"
};
const OrderState = {
	Unfilled: "unfilled",
	Booked: "booked",
	PartialFilled: "partial_filled",
	PartiallyFilled: "partially_filled",
	Filled: "filled",
	Canceled: "canceled",
	Triggered: "triggered"
};
const OrderSide = {
	Unspecified: "unspecified",
	Buy: "buy",
	Sell: "sell",
	StopBuy: "stop_buy",
	StopSell: "stop_sell",
	TakeBuy: "take_buy",
	TakeSell: "take_sell",
	BuyPO: "buy_po",
	SellPO: "sell_po",
	BuyAtomic: "buy_atomic",
	SellAtomic: "sell_atomic",
	Unrecognized: "unrecognized"
};
const OrderMaskMap = __injectivelabs_core_proto_ts_v2_generated_injective_exchange_v1beta1_exchange_pb.OrderMask;

//#endregion
//#region src/types/stream-legacy.ts
/**
* Legacy stream operation types
* Used by indexer stream transformers
*/
const StreamOperation = {
	Insert: "insert",
	Delete: "delete",
	Replace: "replace",
	Update: "update",
	Invalidate: "invalidate"
};

//#endregion
Object.defineProperty(exports, 'GrpcStatusCode', {
  enumerable: true,
  get: function () {
    return GrpcStatusCode;
  }
});
Object.defineProperty(exports, 'OrderMaskMap', {
  enumerable: true,
  get: function () {
    return OrderMaskMap;
  }
});
Object.defineProperty(exports, 'OrderSide', {
  enumerable: true,
  get: function () {
    return OrderSide;
  }
});
Object.defineProperty(exports, 'OrderState', {
  enumerable: true,
  get: function () {
    return OrderState;
  }
});
Object.defineProperty(exports, 'StreamDisconnectReason', {
  enumerable: true,
  get: function () {
    return StreamDisconnectReason;
  }
});
Object.defineProperty(exports, 'StreamEvent', {
  enumerable: true,
  get: function () {
    return StreamEvent;
  }
});
Object.defineProperty(exports, 'StreamOperation', {
  enumerable: true,
  get: function () {
    return StreamOperation;
  }
});
Object.defineProperty(exports, 'StreamState', {
  enumerable: true,
  get: function () {
    return StreamState;
  }
});
Object.defineProperty(exports, 'TokenSource', {
  enumerable: true,
  get: function () {
    return TokenSource;
  }
});
Object.defineProperty(exports, 'TokenType', {
  enumerable: true,
  get: function () {
    return TokenType;
  }
});
Object.defineProperty(exports, 'TokenVerification', {
  enumerable: true,
  get: function () {
    return TokenVerification;
  }
});
Object.defineProperty(exports, 'TradeDirection', {
  enumerable: true,
  get: function () {
    return TradeDirection;
  }
});
Object.defineProperty(exports, 'TradeExecutionSide', {
  enumerable: true,
  get: function () {
    return TradeExecutionSide;
  }
});
Object.defineProperty(exports, 'TradeExecutionType', {
  enumerable: true,
  get: function () {
    return TradeExecutionType;
  }
});
