// @injectivelabs/sdk-ts/dist/cjs/utils.cjs
const require_chunk = require('./chunk-CbDLau6x.cjs');
const require_utils = require('./utils-D3XFGv8y.cjs');
const require_grpc = require('./grpc-FG2z-ICE.cjs');
let __protobuf_ts_grpcweb_transport = require("@protobuf-ts/grpcweb-transport");
let __injectivelabs_utils = require("@injectivelabs/utils");

exports.BECH32_ADDR_ACC_PREFIX = require_utils.BECH32_ADDR_ACC_PREFIX;
exports.BECH32_ADDR_CONS_PREFIX = require_utils.BECH32_ADDR_CONS_PREFIX;
exports.BECH32_ADDR_VAL_PREFIX = require_utils.BECH32_ADDR_VAL_PREFIX;
exports.BECH32_PUBKEY_ACC_PREFIX = require_utils.BECH32_PUBKEY_ACC_PREFIX;
exports.BECH32_PUBKEY_CONS_PREFIX = require_utils.BECH32_PUBKEY_CONS_PREFIX;
exports.BECH32_PUBKEY_VAL_PREFIX = require_utils.BECH32_PUBKEY_VAL_PREFIX;
exports.DEFAULT_DERIVATION_PATH = require_utils.DEFAULT_DERIVATION_PATH;
exports.GrpcWebFetchTransport = __protobuf_ts_grpcweb_transport.GrpcWebFetchTransport;
exports.SignTypedDataVersionV4 = require_utils.SignTypedDataVersionV4;
exports.TypedDataUtilsHashStruct = require_utils.TypedDataUtilsHashStruct;
exports.TypedDataUtilsSanitizeData = require_utils.TypedDataUtilsSanitizeData;
exports.addHexPrefix = require_utils.addHexPrefix;
exports.amountToCosmosSdkDecAmount = require_utils.amountToCosmosSdkDecAmount;
exports.base64ToUint8Array = require_utils.base64ToUint8Array;
exports.base64ToUtf8 = require_utils.base64ToUtf8;
exports.bigIntReplacer = require_utils.bigIntReplacer;
exports.bigIntToNumber = require_utils.bigIntToNumber;
exports.bigIntToString = require_utils.bigIntToString;
exports.binaryToBase64 = require_utils.binaryToBase64;
exports.concatUint8Arrays = require_utils.concatUint8Arrays;
exports.cosmosSdkDecToBigNumber = require_utils.cosmosSdkDecToBigNumber;
exports.decompressPubKey = require_utils.decompressPubKey;
exports.denomAmountFromChainDenomAmount = require_utils.denomAmountFromChainDenomAmount;
exports.denomAmountFromChainDenomAmountToFixed = require_utils.denomAmountFromChainDenomAmountToFixed;
exports.denomAmountFromGrpcChainDenomAmount = require_utils.denomAmountFromGrpcChainDenomAmount;
exports.denomAmountToChainDenomAmount = require_utils.denomAmountToChainDenomAmount;
exports.denomAmountToChainDenomAmountToFixed = require_utils.denomAmountToChainDenomAmountToFixed;
exports.denomAmountToGrpcChainDenomAmount = require_utils.denomAmountToGrpcChainDenomAmount;
exports.derivativeMarginFromChainMargin = require_utils.derivativeMarginFromChainMargin;
exports.derivativeMarginFromChainMarginToFixed = require_utils.derivativeMarginFromChainMarginToFixed;
exports.derivativeMarginToChainMargin = require_utils.derivativeMarginToChainMargin;
exports.derivativeMarginToChainMarginToFixed = require_utils.derivativeMarginToChainMarginToFixed;
exports.derivativePriceFromChainPrice = require_utils.derivativePriceFromChainPrice;
exports.derivativePriceFromChainPriceToFixed = require_utils.derivativePriceFromChainPriceToFixed;
exports.derivativePriceToChainPrice = require_utils.derivativePriceToChainPrice;
exports.derivativePriceToChainPriceToFixed = require_utils.derivativePriceToChainPriceToFixed;
exports.derivativeQuantityFromChainQuantity = require_utils.derivativeQuantityFromChainQuantity;
exports.derivativeQuantityFromChainQuantityToFixed = require_utils.derivativeQuantityFromChainQuantityToFixed;
exports.derivativeQuantityToChainQuantity = require_utils.derivativeQuantityToChainQuantity;
exports.derivativeQuantityToChainQuantityToFixed = require_utils.derivativeQuantityToChainQuantityToFixed;
exports.domainHash = require_utils.domainHash;
exports.fetchAllWithPagination = require_utils.fetchAllWithPagination;
exports.formatAmountToAllowableAmount = require_utils.formatAmountToAllowableAmount;
exports.formatAmountToAllowableDecimals = require_utils.formatAmountToAllowableDecimals;
exports.formatNumberToAllowableDecimals = require_utils.formatNumberToAllowableDecimals;
exports.formatNumberToAllowableTensMultiplier = require_utils.formatNumberToAllowableTensMultiplier;
exports.formatPriceToAllowableDecimals = require_utils.formatPriceToAllowableDecimals;
exports.formatPriceToAllowablePrice = require_utils.formatPriceToAllowablePrice;
exports.fromBase64 = require_utils.fromBase64;
exports.fromUtf8 = require_utils.fromUtf8;
exports.getAddressFromInjectiveAddress = require_utils.getAddressFromInjectiveAddress;
exports.getChecksumAddress = require_utils.getChecksumAddress;
exports.getDefaultSubaccountId = require_utils.getDefaultSubaccountId;
exports.getDerivativeMarketDecimals = require_utils.getDerivativeMarketDecimals;
exports.getDerivativeMarketTensMultiplier = require_utils.getDerivativeMarketTensMultiplier;
exports.getErrorMessage = require_utils.getErrorMessage;
exports.getEthereumAddress = require_utils.getEthereumAddress;
exports.getExactDecimalsFromNumber = __injectivelabs_utils.getExactDecimalsFromNumber;
exports.getGasPriceBasedOnMessage = require_utils.getGasPriceBasedOnMessage;
exports.getGrpcWebTransport = require_grpc.getGrpcWebTransport;
exports.getInjectiveAddress = require_utils.getInjectiveAddress;
exports.getInjectiveAddressFromSubaccountId = require_utils.getInjectiveAddressFromSubaccountId;
exports.getSignificantDecimalsFromNumber = __injectivelabs_utils.getSignificantDecimalsFromNumber;
exports.getSpotMarketDecimals = require_utils.getSpotMarketDecimals;
exports.getSpotMarketTensMultiplier = require_utils.getSpotMarketTensMultiplier;
exports.getSubaccountId = require_utils.getSubaccountId;
exports.getTensMultiplier = require_utils.getTensMultiplier;
exports.grpcCoinToUiCoin = require_utils.grpcCoinToUiCoin;
exports.grpcPaginationToPagination = require_utils.grpcPaginationToPagination;
exports.grpcPagingToPaging = require_utils.grpcPagingToPaging;
exports.grpcPagingToPagingV2 = require_utils.grpcPagingToPagingV2;
exports.hashToHex = require_utils.hashToHex;
exports.hexToBase64 = require_utils.hexToBase64;
exports.hexToBuff = require_utils.hexToBuff;
exports.hexToNumber = require_utils.hexToNumber;
exports.hexToUint8Array = require_utils.hexToUint8Array;
exports.isBrowser = require_utils.isBrowser;
exports.isCw20ContractAddress = require_utils.isCw20ContractAddress;
exports.isJsonString = require_utils.isJsonString;
exports.isNode = require_utils.isNode;
exports.isNumber = require_utils.isNumber;
exports.isReactNative = require_utils.isReactNative;
exports.isServerSide = require_utils.isServerSide;
exports.makeTimeoutTimestamp = require_utils.makeTimeoutTimestamp;
exports.makeTimeoutTimestampInNs = require_utils.makeTimeoutTimestampInNs;
exports.messageHash = require_utils.messageHash;
exports.numberToCosmosSdkDecString = require_utils.numberToCosmosSdkDecString;
exports.objectToJson = require_utils.objectToJson;
exports.ofacList = require_utils.ofacList;
exports.pageRequestToGrpcPageRequestV2 = require_utils.pageRequestToGrpcPageRequestV2;
exports.paginationRequestFromPagination = require_utils.paginationRequestFromPagination;
exports.paginationUint8ArrayToString = require_utils.paginationUint8ArrayToString;
exports.parseCoins = require_utils.parseCoins;
exports.privateKeyHashToPublicKey = require_utils.privateKeyHashToPublicKey;
exports.privateKeyHashToPublicKeyBase64 = require_utils.privateKeyHashToPublicKeyBase64;
exports.privateKeyToPublicKey = require_utils.privateKeyToPublicKey;
exports.privateKeyToPublicKeyBase64 = require_utils.privateKeyToPublicKeyBase64;
exports.protoObjectToJson = require_utils.protoObjectToJson;
exports.protobufTimestampToDate = require_utils.protobufTimestampToDate;
exports.protobufTimestampToUnixMs = require_utils.protobufTimestampToUnixMs;
exports.protobufTimestampToUnixSeconds = require_utils.protobufTimestampToUnixSeconds;
exports.publicKeyToAddress = require_utils.publicKeyToAddress;
exports.recoverTypedSignaturePubKey = require_utils.recoverTypedSignaturePubKey;
exports.removeHexPrefix = require_utils.removeHexPrefix;
exports.ripemd160 = require_utils.ripemd160;
exports.safeBigIntStringify = require_utils.safeBigIntStringify;
exports.sanitizeTypedData = require_utils.sanitizeTypedData;
exports.sha256 = require_utils.sha256;
exports.sortObjectByKeys = require_utils.sortObjectByKeys;
exports.sortObjectByKeysWithReduce = require_utils.sortObjectByKeysWithReduce;
exports.spotPriceFromChainPrice = require_utils.spotPriceFromChainPrice;
exports.spotPriceFromChainPriceToFixed = require_utils.spotPriceFromChainPriceToFixed;
exports.spotPriceToChainPrice = require_utils.spotPriceToChainPrice;
exports.spotPriceToChainPriceToFixed = require_utils.spotPriceToChainPriceToFixed;
exports.spotQuantityFromChainQuantity = require_utils.spotQuantityFromChainQuantity;
exports.spotQuantityFromChainQuantityToFixed = require_utils.spotQuantityFromChainQuantityToFixed;
exports.spotQuantityToChainQuantity = require_utils.spotQuantityToChainQuantity;
exports.spotQuantityToChainQuantityToFixed = require_utils.spotQuantityToChainQuantityToFixed;
exports.stringToUint8Array = require_utils.stringToUint8Array;
exports.toBase64 = require_utils.toBase64;
exports.toUtf8 = require_utils.toUtf8;
exports.uint8ArrayToBase64 = require_utils.uint8ArrayToBase64;
exports.uint8ArrayToHex = require_utils.uint8ArrayToHex;
exports.uint8ArrayToString = require_utils.uint8ArrayToString;
