// @injectivelabs/core-proto-ts-v2/esm/generated/tendermint/rpc/grpc/types_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { ExecTxResult } from "../../abci/types_pb.js";
import { ResponseCheckTx } from "../../abci/types_pb.js";
class RequestPing$Type extends MessageType {
  constructor() {
    super("tendermint.rpc.grpc.RequestPing", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RequestPing = new RequestPing$Type();
class RequestBroadcastTx$Type extends MessageType {
  constructor() {
    super("tendermint.rpc.grpc.RequestBroadcastTx", [
      {
        no: 1,
        name: "tx",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.tx = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes tx */
        1:
          message.tx = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.tx.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.tx);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RequestBroadcastTx = new RequestBroadcastTx$Type();
class ResponsePing$Type extends MessageType {
  constructor() {
    super("tendermint.rpc.grpc.ResponsePing", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ResponsePing = new ResponsePing$Type();
class ResponseBroadcastTx$Type extends MessageType {
  constructor() {
    super("tendermint.rpc.grpc.ResponseBroadcastTx", [
      { no: 1, name: "check_tx", kind: "message", T: () => ResponseCheckTx },
      { no: 2, name: "tx_result", kind: "message", T: () => ExecTxResult }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.abci.ResponseCheckTx check_tx */
        1:
          message.checkTx = ResponseCheckTx.internalBinaryRead(reader, reader.uint32(), options, message.checkTx);
          break;
        case /* tendermint.abci.ExecTxResult tx_result */
        2:
          message.txResult = ExecTxResult.internalBinaryRead(reader, reader.uint32(), options, message.txResult);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.checkTx)
      ResponseCheckTx.internalBinaryWrite(message.checkTx, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.txResult)
      ExecTxResult.internalBinaryWrite(message.txResult, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ResponseBroadcastTx = new ResponseBroadcastTx$Type();
const BroadcastAPI = new ServiceType("tendermint.rpc.grpc.BroadcastAPI", [
  { name: "Ping", options: {}, I: RequestPing, O: ResponsePing },
  { name: "BroadcastTx", options: {}, I: RequestBroadcastTx, O: ResponseBroadcastTx }
]);
export {
  BroadcastAPI,
  RequestBroadcastTx,
  RequestPing,
  ResponseBroadcastTx,
  ResponsePing
};
