// @injectivelabs/core-proto-ts-v2/esm/generated/tendermint/p2p/conn_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { PublicKey } from "../crypto/keys_pb.js";
class PacketPing$Type extends MessageType {
  constructor() {
    super("tendermint.p2p.PacketPing", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketPing = new PacketPing$Type();
class PacketPong$Type extends MessageType {
  constructor() {
    super("tendermint.p2p.PacketPong", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketPong = new PacketPong$Type();
class PacketMsg$Type extends MessageType {
  constructor() {
    super("tendermint.p2p.PacketMsg", [
      { no: 1, name: "channel_id", kind: "scalar", T: 5, options: { "gogoproto.customname": "ChannelID" } },
      { no: 2, name: "eof", kind: "scalar", T: 8, options: { "gogoproto.customname": "EOF" } },
      {
        no: 3,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channelId = 0;
    message.eof = false;
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int32 channel_id */
        1:
          message.channelId = reader.int32();
          break;
        case /* bool eof */
        2:
          message.eof = reader.bool();
          break;
        case /* bytes data */
        3:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channelId !== 0)
      writer.tag(1, WireType.Varint).int32(message.channelId);
    if (message.eof !== false)
      writer.tag(2, WireType.Varint).bool(message.eof);
    if (message.data.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketMsg = new PacketMsg$Type();
class Packet$Type extends MessageType {
  constructor() {
    super("tendermint.p2p.Packet", [
      { no: 1, name: "packet_ping", kind: "message", oneof: "sum", T: () => PacketPing },
      { no: 2, name: "packet_pong", kind: "message", oneof: "sum", T: () => PacketPong },
      { no: 3, name: "packet_msg", kind: "message", oneof: "sum", T: () => PacketMsg }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sum = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.p2p.PacketPing packet_ping */
        1:
          message.sum = {
            oneofKind: "packetPing",
            packetPing: PacketPing.internalBinaryRead(reader, reader.uint32(), options, message.sum.packetPing)
          };
          break;
        case /* tendermint.p2p.PacketPong packet_pong */
        2:
          message.sum = {
            oneofKind: "packetPong",
            packetPong: PacketPong.internalBinaryRead(reader, reader.uint32(), options, message.sum.packetPong)
          };
          break;
        case /* tendermint.p2p.PacketMsg packet_msg */
        3:
          message.sum = {
            oneofKind: "packetMsg",
            packetMsg: PacketMsg.internalBinaryRead(reader, reader.uint32(), options, message.sum.packetMsg)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sum.oneofKind === "packetPing")
      PacketPing.internalBinaryWrite(message.sum.packetPing, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "packetPong")
      PacketPong.internalBinaryWrite(message.sum.packetPong, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "packetMsg")
      PacketMsg.internalBinaryWrite(message.sum.packetMsg, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Packet = new Packet$Type();
class AuthSigMessage$Type extends MessageType {
  constructor() {
    super("tendermint.p2p.AuthSigMessage", [
      { no: 1, name: "pub_key", kind: "message", T: () => PublicKey, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "sig",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sig = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.crypto.PublicKey pub_key */
        1:
          message.pubKey = PublicKey.internalBinaryRead(reader, reader.uint32(), options, message.pubKey);
          break;
        case /* bytes sig */
        2:
          message.sig = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pubKey)
      PublicKey.internalBinaryWrite(message.pubKey, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.sig.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.sig);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const AuthSigMessage = new AuthSigMessage$Type();
export {
  AuthSigMessage,
  Packet,
  PacketMsg,
  PacketPing,
  PacketPong
};
