// @injectivelabs/core-proto-ts-v2/esm/generated/tendermint/consensus/wal_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Timestamp } from "../../google/protobuf/timestamp_pb.js";
import { EventDataRoundState } from "../types/events_pb.js";
import { Duration } from "../../google/protobuf/duration_pb.js";
import { Message } from "./types_pb.js";
class MsgInfo$Type extends MessageType {
  constructor() {
    super("tendermint.consensus.MsgInfo", [
      { no: 1, name: "msg", kind: "message", T: () => Message, options: { "gogoproto.nullable": false } },
      { no: 2, name: "peer_id", kind: "scalar", T: 9, options: { "gogoproto.customname": "PeerID" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.peerId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.consensus.Message msg */
        1:
          message.msg = Message.internalBinaryRead(reader, reader.uint32(), options, message.msg);
          break;
        case /* string peer_id */
        2:
          message.peerId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.msg)
      Message.internalBinaryWrite(message.msg, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.peerId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.peerId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgInfo = new MsgInfo$Type();
class TimeoutInfo$Type extends MessageType {
  constructor() {
    super("tendermint.consensus.TimeoutInfo", [
      { no: 1, name: "duration", kind: "message", T: () => Duration, options: { "gogoproto.nullable": false, "gogoproto.stdduration": true } },
      {
        no: 2,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "round",
        kind: "scalar",
        T: 5
        /*ScalarType.INT32*/
      },
      {
        no: 4,
        name: "step",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    message.round = 0;
    message.step = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Duration duration */
        1:
          message.duration = Duration.internalBinaryRead(reader, reader.uint32(), options, message.duration);
          break;
        case /* int64 height */
        2:
          message.height = reader.int64().toBigInt();
          break;
        case /* int32 round */
        3:
          message.round = reader.int32();
          break;
        case /* uint32 step */
        4:
          message.step = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.duration)
      Duration.internalBinaryWrite(message.duration, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.height !== 0n)
      writer.tag(2, WireType.Varint).int64(message.height);
    if (message.round !== 0)
      writer.tag(3, WireType.Varint).int32(message.round);
    if (message.step !== 0)
      writer.tag(4, WireType.Varint).uint32(message.step);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TimeoutInfo = new TimeoutInfo$Type();
class EndHeight$Type extends MessageType {
  constructor() {
    super("tendermint.consensus.EndHeight", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 height */
        1:
          message.height = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).int64(message.height);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EndHeight = new EndHeight$Type();
class WALMessage$Type extends MessageType {
  constructor() {
    super("tendermint.consensus.WALMessage", [
      { no: 1, name: "event_data_round_state", kind: "message", oneof: "sum", T: () => EventDataRoundState },
      { no: 2, name: "msg_info", kind: "message", oneof: "sum", T: () => MsgInfo },
      { no: 3, name: "timeout_info", kind: "message", oneof: "sum", T: () => TimeoutInfo },
      { no: 4, name: "end_height", kind: "message", oneof: "sum", T: () => EndHeight }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sum = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.types.EventDataRoundState event_data_round_state */
        1:
          message.sum = {
            oneofKind: "eventDataRoundState",
            eventDataRoundState: EventDataRoundState.internalBinaryRead(reader, reader.uint32(), options, message.sum.eventDataRoundState)
          };
          break;
        case /* tendermint.consensus.MsgInfo msg_info */
        2:
          message.sum = {
            oneofKind: "msgInfo",
            msgInfo: MsgInfo.internalBinaryRead(reader, reader.uint32(), options, message.sum.msgInfo)
          };
          break;
        case /* tendermint.consensus.TimeoutInfo timeout_info */
        3:
          message.sum = {
            oneofKind: "timeoutInfo",
            timeoutInfo: TimeoutInfo.internalBinaryRead(reader, reader.uint32(), options, message.sum.timeoutInfo)
          };
          break;
        case /* tendermint.consensus.EndHeight end_height */
        4:
          message.sum = {
            oneofKind: "endHeight",
            endHeight: EndHeight.internalBinaryRead(reader, reader.uint32(), options, message.sum.endHeight)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sum.oneofKind === "eventDataRoundState")
      EventDataRoundState.internalBinaryWrite(message.sum.eventDataRoundState, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "msgInfo")
      MsgInfo.internalBinaryWrite(message.sum.msgInfo, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "timeoutInfo")
      TimeoutInfo.internalBinaryWrite(message.sum.timeoutInfo, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "endHeight")
      EndHeight.internalBinaryWrite(message.sum.endHeight, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const WALMessage = new WALMessage$Type();
class TimedWALMessage$Type extends MessageType {
  constructor() {
    super("tendermint.consensus.TimedWALMessage", [
      { no: 1, name: "time", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } },
      { no: 2, name: "msg", kind: "message", T: () => WALMessage }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Timestamp time */
        1:
          message.time = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.time);
          break;
        case /* tendermint.consensus.WALMessage msg */
        2:
          message.msg = WALMessage.internalBinaryRead(reader, reader.uint32(), options, message.msg);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.time)
      Timestamp.internalBinaryWrite(message.time, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.msg)
      WALMessage.internalBinaryWrite(message.msg, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const TimedWALMessage = new TimedWALMessage$Type();
export {
  EndHeight,
  MsgInfo,
  TimedWALMessage,
  TimeoutInfo,
  WALMessage
};
