// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/orm/query/v1alpha1/query_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Duration } from "../../../../google/protobuf/duration_pb.js";
import { Timestamp } from "../../../../google/protobuf/timestamp_pb.js";
import { PageResponse } from "../../../base/query/v1beta1/pagination_pb.js";
import { PageRequest } from "../../../base/query/v1beta1/pagination_pb.js";
import { Any } from "../../../../google/protobuf/any_pb.js";
class GetRequest$Type extends MessageType {
  constructor() {
    super("cosmos.orm.query.v1alpha1.GetRequest", [
      {
        no: 1,
        name: "message_name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "index",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "values", kind: "message", repeat: 2, T: () => IndexValue }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.messageName = "";
    message.index = "";
    message.values = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string message_name */
        1:
          message.messageName = reader.string();
          break;
        case /* string index */
        2:
          message.index = reader.string();
          break;
        case /* repeated cosmos.orm.query.v1alpha1.IndexValue values */
        3:
          message.values.push(IndexValue.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.messageName !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.messageName);
    if (message.index !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.index);
    for (let i = 0; i < message.values.length; i++)
      IndexValue.internalBinaryWrite(message.values[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetRequest = new GetRequest$Type();
class GetResponse$Type extends MessageType {
  constructor() {
    super("cosmos.orm.query.v1alpha1.GetResponse", [
      { no: 1, name: "result", kind: "message", T: () => Any }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any result */
        1:
          message.result = Any.internalBinaryRead(reader, reader.uint32(), options, message.result);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.result)
      Any.internalBinaryWrite(message.result, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GetResponse = new GetResponse$Type();
class ListRequest$Type extends MessageType {
  constructor() {
    super("cosmos.orm.query.v1alpha1.ListRequest", [
      {
        no: 1,
        name: "message_name",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "index",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "prefix", kind: "message", oneof: "query", T: () => ListRequest_Prefix },
      { no: 4, name: "range", kind: "message", oneof: "query", T: () => ListRequest_Range },
      { no: 5, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.messageName = "";
    message.index = "";
    message.query = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string message_name */
        1:
          message.messageName = reader.string();
          break;
        case /* string index */
        2:
          message.index = reader.string();
          break;
        case /* cosmos.orm.query.v1alpha1.ListRequest.Prefix prefix */
        3:
          message.query = {
            oneofKind: "prefix",
            prefix: ListRequest_Prefix.internalBinaryRead(reader, reader.uint32(), options, message.query.prefix)
          };
          break;
        case /* cosmos.orm.query.v1alpha1.ListRequest.Range range */
        4:
          message.query = {
            oneofKind: "range",
            range: ListRequest_Range.internalBinaryRead(reader, reader.uint32(), options, message.query.range)
          };
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        5:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.messageName !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.messageName);
    if (message.index !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.index);
    if (message.query.oneofKind === "prefix")
      ListRequest_Prefix.internalBinaryWrite(message.query.prefix, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.query.oneofKind === "range")
      ListRequest_Range.internalBinaryWrite(message.query.range, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListRequest = new ListRequest$Type();
class ListRequest_Prefix$Type extends MessageType {
  constructor() {
    super("cosmos.orm.query.v1alpha1.ListRequest.Prefix", [
      { no: 1, name: "values", kind: "message", repeat: 2, T: () => IndexValue }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.values = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.orm.query.v1alpha1.IndexValue values */
        1:
          message.values.push(IndexValue.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.values.length; i++)
      IndexValue.internalBinaryWrite(message.values[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListRequest_Prefix = new ListRequest_Prefix$Type();
class ListRequest_Range$Type extends MessageType {
  constructor() {
    super("cosmos.orm.query.v1alpha1.ListRequest.Range", [
      { no: 1, name: "start", kind: "message", repeat: 2, T: () => IndexValue },
      { no: 2, name: "end", kind: "message", repeat: 2, T: () => IndexValue }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.start = [];
    message.end = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.orm.query.v1alpha1.IndexValue start */
        1:
          message.start.push(IndexValue.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated cosmos.orm.query.v1alpha1.IndexValue end */
        2:
          message.end.push(IndexValue.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.start.length; i++)
      IndexValue.internalBinaryWrite(message.start[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.end.length; i++)
      IndexValue.internalBinaryWrite(message.end[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListRequest_Range = new ListRequest_Range$Type();
class ListResponse$Type extends MessageType {
  constructor() {
    super("cosmos.orm.query.v1alpha1.ListResponse", [
      { no: 1, name: "results", kind: "message", repeat: 2, T: () => Any },
      { no: 5, name: "pagination", kind: "message", T: () => PageResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.results = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated google.protobuf.Any results */
        1:
          message.results.push(Any.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        5:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.results.length; i++)
      Any.internalBinaryWrite(message.results[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListResponse = new ListResponse$Type();
class IndexValue$Type extends MessageType {
  constructor() {
    super("cosmos.orm.query.v1alpha1.IndexValue", [
      {
        no: 1,
        name: "uint",
        kind: "scalar",
        oneof: "value",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "int",
        kind: "scalar",
        oneof: "value",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "str",
        kind: "scalar",
        oneof: "value",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "bytes",
        kind: "scalar",
        oneof: "value",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "enum",
        kind: "scalar",
        oneof: "value",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "bool",
        kind: "scalar",
        oneof: "value",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 7, name: "timestamp", kind: "message", oneof: "value", T: () => Timestamp },
      { no: 8, name: "duration", kind: "message", oneof: "value", T: () => Duration }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.value = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 uint */
        1:
          message.value = {
            oneofKind: "uint",
            uint: reader.uint64().toBigInt()
          };
          break;
        case /* int64 int */
        2:
          message.value = {
            oneofKind: "int",
            int: reader.int64().toBigInt()
          };
          break;
        case /* string str */
        3:
          message.value = {
            oneofKind: "str",
            str: reader.string()
          };
          break;
        case /* bytes bytes */
        4:
          message.value = {
            oneofKind: "bytes",
            bytes: reader.bytes()
          };
          break;
        case /* string enum */
        5:
          message.value = {
            oneofKind: "enum",
            enum: reader.string()
          };
          break;
        case /* bool bool */
        6:
          message.value = {
            oneofKind: "bool",
            bool: reader.bool()
          };
          break;
        case /* google.protobuf.Timestamp timestamp */
        7:
          message.value = {
            oneofKind: "timestamp",
            timestamp: Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.value.timestamp)
          };
          break;
        case /* google.protobuf.Duration duration */
        8:
          message.value = {
            oneofKind: "duration",
            duration: Duration.internalBinaryRead(reader, reader.uint32(), options, message.value.duration)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.value.oneofKind === "uint")
      writer.tag(1, WireType.Varint).uint64(message.value.uint);
    if (message.value.oneofKind === "int")
      writer.tag(2, WireType.Varint).int64(message.value.int);
    if (message.value.oneofKind === "str")
      writer.tag(3, WireType.LengthDelimited).string(message.value.str);
    if (message.value.oneofKind === "bytes")
      writer.tag(4, WireType.LengthDelimited).bytes(message.value.bytes);
    if (message.value.oneofKind === "enum")
      writer.tag(5, WireType.LengthDelimited).string(message.value.enum);
    if (message.value.oneofKind === "bool")
      writer.tag(6, WireType.Varint).bool(message.value.bool);
    if (message.value.oneofKind === "timestamp")
      Timestamp.internalBinaryWrite(message.value.timestamp, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.value.oneofKind === "duration")
      Duration.internalBinaryWrite(message.value.duration, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const IndexValue = new IndexValue$Type();
const Query = new ServiceType("cosmos.orm.query.v1alpha1.Query", [
  { name: "Get", options: {}, I: GetRequest, O: GetResponse },
  { name: "List", options: {}, I: ListRequest, O: ListResponse }
]);
export {
  GetRequest,
  GetResponse,
  IndexValue,
  ListRequest,
  ListRequest_Prefix,
  ListRequest_Range,
  ListResponse,
  Query
};
