// @injectivelabs/utils/dist/cjs/constants-DEbTiKzu.cjs
const require_index = require('./index.cjs');
let bignumber_js = require("bignumber.js");

//#region src/classes/BigNumber.ts
var BigNumber_default = bignumber_js.BigNumber;
/**
* Converts a value to BigNumber if it isn't already
*/
const toBigNumber = (value) => {
	if (value instanceof bignumber_js.BigNumber) return value;
	return new bignumber_js.BigNumber(value);
};
/**
* Converts a value to Wei (multiplies by 10^decimals) default to 18 decimals
*/
const toChainFormat = (value, decimals = 18) => {
	return toBigNumber(value).multipliedBy(new bignumber_js.BigNumber(10).pow(decimals));
};
/**
* Converts a value from Wei to Base (divides by 10^decimals) default to 18 decimals
*/
const toHumanReadable = (value, decimals = 18) => {
	return toBigNumber(value).dividedBy(new bignumber_js.BigNumber(10).pow(decimals));
};

//#endregion
//#region src/constants.ts
const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
const INJ_DENOM = "inj";
const INJECTIVE_DENOM = "inj";
const DEFAULT_FEE_DENOM = "inj";
const DEFAULT_GAS_LIMIT = 4e5;
const DEFAULT_IBC_GAS_LIMIT = 3e5;
const DEFAULT_GAS_PRICE = 16e7;
const DEFAULT_EXCHANGE_LIMIT = 2e5;
const DEFAULT_BRIDGE_FEE_DENOM = "inj";
const DEFAULT_BRIDGE_FEE_PRICE = "160000000";
const DEFAULT_BRIDGE_FEE_AMOUNT = "200000000000000";
const DEFAULT_BLOCK_TIMEOUT_HEIGHT = 120;
const DEFAULT_BLOCK_TIME_IN_SECONDS = .7;
/**
* Default timeout for transaction block inclusion polling.
*
* Calculation: 120 blocks × 0.7 seconds/block = 84 seconds = 84,000ms
*/
const DEFAULT_TX_BLOCK_INCLUSION_TIMEOUT_IN_MS = Math.floor(DEFAULT_BLOCK_TIMEOUT_HEIGHT * DEFAULT_BLOCK_TIME_IN_SECONDS * 1e3);
const DEFAULT_TIMESTAMP_TIMEOUT_MS = 60 * 1e3 * 3;
const getDefaultStdFee = () => ({
	amount: [{
		amount: new BigNumber_default(DEFAULT_GAS_LIMIT).times(DEFAULT_GAS_PRICE).toString(),
		denom: "inj"
	}],
	gas: DEFAULT_GAS_LIMIT.toString(),
	payer: "",
	granter: "",
	feePayer: ""
});
const DEFAULT_STD_FEE_BY_DENOM = (denom = "inj") => ({
	amount: [{
		denom,
		amount: new BigNumber_default(DEFAULT_GAS_LIMIT).times(DEFAULT_GAS_PRICE).toString()
	}],
	gas: DEFAULT_GAS_LIMIT.toString()
});

//#endregion
Object.defineProperty(exports, 'BigNumber_default', {
  enumerable: true,
  get: function () {
    return BigNumber_default;
  }
});
Object.defineProperty(exports, 'DEFAULT_BLOCK_TIMEOUT_HEIGHT', {
  enumerable: true,
  get: function () {
    return DEFAULT_BLOCK_TIMEOUT_HEIGHT;
  }
});
Object.defineProperty(exports, 'DEFAULT_BLOCK_TIME_IN_SECONDS', {
  enumerable: true,
  get: function () {
    return DEFAULT_BLOCK_TIME_IN_SECONDS;
  }
});
Object.defineProperty(exports, 'DEFAULT_BRIDGE_FEE_AMOUNT', {
  enumerable: true,
  get: function () {
    return DEFAULT_BRIDGE_FEE_AMOUNT;
  }
});
Object.defineProperty(exports, 'DEFAULT_BRIDGE_FEE_DENOM', {
  enumerable: true,
  get: function () {
    return DEFAULT_BRIDGE_FEE_DENOM;
  }
});
Object.defineProperty(exports, 'DEFAULT_BRIDGE_FEE_PRICE', {
  enumerable: true,
  get: function () {
    return DEFAULT_BRIDGE_FEE_PRICE;
  }
});
Object.defineProperty(exports, 'DEFAULT_EXCHANGE_LIMIT', {
  enumerable: true,
  get: function () {
    return DEFAULT_EXCHANGE_LIMIT;
  }
});
Object.defineProperty(exports, 'DEFAULT_FEE_DENOM', {
  enumerable: true,
  get: function () {
    return DEFAULT_FEE_DENOM;
  }
});
Object.defineProperty(exports, 'DEFAULT_GAS_LIMIT', {
  enumerable: true,
  get: function () {
    return DEFAULT_GAS_LIMIT;
  }
});
Object.defineProperty(exports, 'DEFAULT_GAS_PRICE', {
  enumerable: true,
  get: function () {
    return DEFAULT_GAS_PRICE;
  }
});
Object.defineProperty(exports, 'DEFAULT_IBC_GAS_LIMIT', {
  enumerable: true,
  get: function () {
    return DEFAULT_IBC_GAS_LIMIT;
  }
});
Object.defineProperty(exports, 'DEFAULT_STD_FEE_BY_DENOM', {
  enumerable: true,
  get: function () {
    return DEFAULT_STD_FEE_BY_DENOM;
  }
});
Object.defineProperty(exports, 'DEFAULT_TIMESTAMP_TIMEOUT_MS', {
  enumerable: true,
  get: function () {
    return DEFAULT_TIMESTAMP_TIMEOUT_MS;
  }
});
Object.defineProperty(exports, 'DEFAULT_TX_BLOCK_INCLUSION_TIMEOUT_IN_MS', {
  enumerable: true,
  get: function () {
    return DEFAULT_TX_BLOCK_INCLUSION_TIMEOUT_IN_MS;
  }
});
Object.defineProperty(exports, 'INJECTIVE_DENOM', {
  enumerable: true,
  get: function () {
    return INJECTIVE_DENOM;
  }
});
Object.defineProperty(exports, 'INJ_DENOM', {
  enumerable: true,
  get: function () {
    return INJ_DENOM;
  }
});
Object.defineProperty(exports, 'ZERO_ADDRESS', {
  enumerable: true,
  get: function () {
    return ZERO_ADDRESS;
  }
});
Object.defineProperty(exports, 'getDefaultStdFee', {
  enumerable: true,
  get: function () {
    return getDefaultStdFee;
  }
});
Object.defineProperty(exports, 'toBigNumber', {
  enumerable: true,
  get: function () {
    return toBigNumber;
  }
});
Object.defineProperty(exports, 'toChainFormat', {
  enumerable: true,
  get: function () {
    return toChainFormat;
  }
});
Object.defineProperty(exports, 'toHumanReadable', {
  enumerable: true,
  get: function () {
    return toHumanReadable;
  }
});
