// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/vesting/v1beta1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Period } from "./vesting_pb.js";
import { Coin } from "../../base/v1beta1/coin_pb.js";
class MsgCreateVestingAccount$Type extends MessageType {
  constructor() {
    super("cosmos.vesting.v1beta1.MsgCreateVestingAccount", [
      { no: 1, name: "from_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "to_address", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "amount", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } },
      {
        no: 4,
        name: "end_time",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "delayed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ], { "gogoproto.equal": true, "amino.name": "cosmos-sdk/MsgCreateVestingAccount", "cosmos.msg.v1.signer": ["from_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.fromAddress = "";
    message.toAddress = "";
    message.amount = [];
    message.endTime = 0n;
    message.delayed = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string from_address */
        1:
          message.fromAddress = reader.string();
          break;
        case /* string to_address */
        2:
          message.toAddress = reader.string();
          break;
        case /* repeated cosmos.base.v1beta1.Coin amount */
        3:
          message.amount.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* int64 end_time */
        4:
          message.endTime = reader.int64().toBigInt();
          break;
        case /* bool delayed */
        5:
          message.delayed = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fromAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.fromAddress);
    if (message.toAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.toAddress);
    for (let i = 0; i < message.amount.length; i++)
      Coin.internalBinaryWrite(message.amount[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.endTime !== 0n)
      writer.tag(4, WireType.Varint).int64(message.endTime);
    if (message.delayed !== false)
      writer.tag(5, WireType.Varint).bool(message.delayed);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateVestingAccount = new MsgCreateVestingAccount$Type();
class MsgCreateVestingAccountResponse$Type extends MessageType {
  constructor() {
    super("cosmos.vesting.v1beta1.MsgCreateVestingAccountResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateVestingAccountResponse = new MsgCreateVestingAccountResponse$Type();
class MsgCreatePermanentLockedAccount$Type extends MessageType {
  constructor() {
    super("cosmos.vesting.v1beta1.MsgCreatePermanentLockedAccount", [
      { no: 1, name: "from_address", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"from_address"' } },
      { no: 2, name: "to_address", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"to_address"' } },
      { no: 3, name: "amount", kind: "message", repeat: 2, T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "github.com/cosmos/cosmos-sdk/types.Coins", "amino.encoding": "legacy_coins", "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": true, "amino.name": "cosmos-sdk/MsgCreatePermLockedAccount", "cosmos.msg.v1.signer": ["from_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.fromAddress = "";
    message.toAddress = "";
    message.amount = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string from_address */
        1:
          message.fromAddress = reader.string();
          break;
        case /* string to_address */
        2:
          message.toAddress = reader.string();
          break;
        case /* repeated cosmos.base.v1beta1.Coin amount */
        3:
          message.amount.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fromAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.fromAddress);
    if (message.toAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.toAddress);
    for (let i = 0; i < message.amount.length; i++)
      Coin.internalBinaryWrite(message.amount[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreatePermanentLockedAccount = new MsgCreatePermanentLockedAccount$Type();
class MsgCreatePermanentLockedAccountResponse$Type extends MessageType {
  constructor() {
    super("cosmos.vesting.v1beta1.MsgCreatePermanentLockedAccountResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreatePermanentLockedAccountResponse = new MsgCreatePermanentLockedAccountResponse$Type();
class MsgCreatePeriodicVestingAccount$Type extends MessageType {
  constructor() {
    super("cosmos.vesting.v1beta1.MsgCreatePeriodicVestingAccount", [
      {
        no: 1,
        name: "from_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "to_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "start_time",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 4, name: "vesting_periods", kind: "message", repeat: 2, T: () => Period, options: { "gogoproto.nullable": false, "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": false, "amino.name": "cosmos-sdk/MsgCreatePeriodVestAccount", "cosmos.msg.v1.signer": ["from_address"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.fromAddress = "";
    message.toAddress = "";
    message.startTime = 0n;
    message.vestingPeriods = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string from_address */
        1:
          message.fromAddress = reader.string();
          break;
        case /* string to_address */
        2:
          message.toAddress = reader.string();
          break;
        case /* int64 start_time */
        3:
          message.startTime = reader.int64().toBigInt();
          break;
        case /* repeated cosmos.vesting.v1beta1.Period vesting_periods */
        4:
          message.vestingPeriods.push(Period.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fromAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.fromAddress);
    if (message.toAddress !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.toAddress);
    if (message.startTime !== 0n)
      writer.tag(3, WireType.Varint).int64(message.startTime);
    for (let i = 0; i < message.vestingPeriods.length; i++)
      Period.internalBinaryWrite(message.vestingPeriods[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreatePeriodicVestingAccount = new MsgCreatePeriodicVestingAccount$Type();
class MsgCreatePeriodicVestingAccountResponse$Type extends MessageType {
  constructor() {
    super("cosmos.vesting.v1beta1.MsgCreatePeriodicVestingAccountResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreatePeriodicVestingAccountResponse = new MsgCreatePeriodicVestingAccountResponse$Type();
const Msg = new ServiceType("cosmos.vesting.v1beta1.Msg", [
  { name: "CreateVestingAccount", options: {}, I: MsgCreateVestingAccount, O: MsgCreateVestingAccountResponse },
  { name: "CreatePermanentLockedAccount", options: {}, I: MsgCreatePermanentLockedAccount, O: MsgCreatePermanentLockedAccountResponse },
  { name: "CreatePeriodicVestingAccount", options: {}, I: MsgCreatePeriodicVestingAccount, O: MsgCreatePeriodicVestingAccountResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgCreatePeriodicVestingAccount,
  MsgCreatePeriodicVestingAccountResponse,
  MsgCreatePermanentLockedAccount,
  MsgCreatePermanentLockedAccountResponse,
  MsgCreateVestingAccount,
  MsgCreateVestingAccountResponse
};
