// @injectivelabs/core-proto-ts-v2/esm/generated/injective/insurance/v1beta1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { RedemptionSchedule } from "./insurance_pb.js";
import { InsuranceFund } from "./insurance_pb.js";
import { Params } from "./insurance_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("injective.insurance.v1beta1.GenesisState", [
      { no: 1, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } },
      { no: 2, name: "insurance_funds", kind: "message", repeat: 2, T: () => InsuranceFund, options: { "gogoproto.nullable": false } },
      { no: 3, name: "redemption_schedule", kind: "message", repeat: 2, T: () => RedemptionSchedule, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "next_share_denom_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "next_redemption_schedule_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.insuranceFunds = [];
    message.redemptionSchedule = [];
    message.nextShareDenomId = 0n;
    message.nextRedemptionScheduleId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.insurance.v1beta1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        case /* repeated injective.insurance.v1beta1.InsuranceFund insurance_funds */
        2:
          message.insuranceFunds.push(InsuranceFund.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective.insurance.v1beta1.RedemptionSchedule redemption_schedule */
        3:
          message.redemptionSchedule.push(RedemptionSchedule.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* uint64 next_share_denom_id */
        4:
          message.nextShareDenomId = reader.uint64().toBigInt();
          break;
        case /* uint64 next_redemption_schedule_id */
        5:
          message.nextRedemptionScheduleId = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.insuranceFunds.length; i++)
      InsuranceFund.internalBinaryWrite(message.insuranceFunds[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.redemptionSchedule.length; i++)
      RedemptionSchedule.internalBinaryWrite(message.redemptionSchedule[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.nextShareDenomId !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.nextShareDenomId);
    if (message.nextRedemptionScheduleId !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.nextRedemptionScheduleId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
export {
  GenesisState
};
