// @injectivelabs/core-proto-ts-v2/esm/generated/injective/insurance/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Retrieves insurance params
   *
   * @generated from protobuf rpc: InsuranceParams
   */
  insuranceParams(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves individual insurance fund information from market id
   *
   * @generated from protobuf rpc: InsuranceFund
   */
  insuranceFund(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves all insurance funds
   *
   * @generated from protobuf rpc: InsuranceFunds
   */
  insuranceFunds(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrives the value of insurance fund share token at current price (not
   * pending redemption)
   *
   * @generated from protobuf rpc: EstimatedRedemptions
   */
  estimatedRedemptions(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves pending redemptions' share token at current price
   *
   * @generated from protobuf rpc: PendingRedemptions
   */
  pendingRedemptions(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the entire insurance module's state
   *
   * @generated from protobuf rpc: InsuranceModuleState
   */
  insuranceModuleState(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
