// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/consensus/v1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { FeatureParams } from "../../../cometbft/types/v1/params_pb.js";
import { SynchronyParams } from "../../../cometbft/types/v1/params_pb.js";
import { ABCIParams } from "../../../cometbft/types/v1/params_pb.js";
import { ValidatorParams } from "../../../cometbft/types/v1/params_pb.js";
import { EvidenceParams } from "../../../cometbft/types/v1/params_pb.js";
import { BlockParams } from "../../../cometbft/types/v1/params_pb.js";
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("cosmos.consensus.v1.MsgUpdateParams", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "block", kind: "message", T: () => BlockParams },
      { no: 3, name: "evidence", kind: "message", T: () => EvidenceParams },
      { no: 4, name: "validator", kind: "message", T: () => ValidatorParams },
      { no: 5, name: "abci", kind: "message", T: () => ABCIParams },
      { no: 6, name: "synchrony", kind: "message", T: () => SynchronyParams },
      { no: 7, name: "feature", kind: "message", T: () => FeatureParams }
    ], { "amino.name": "cosmos-sdk/x/consensus/MsgUpdateParams", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* cometbft.types.v1.BlockParams block */
        2:
          message.block = BlockParams.internalBinaryRead(reader, reader.uint32(), options, message.block);
          break;
        case /* cometbft.types.v1.EvidenceParams evidence */
        3:
          message.evidence = EvidenceParams.internalBinaryRead(reader, reader.uint32(), options, message.evidence);
          break;
        case /* cometbft.types.v1.ValidatorParams validator */
        4:
          message.validator = ValidatorParams.internalBinaryRead(reader, reader.uint32(), options, message.validator);
          break;
        case /* cometbft.types.v1.ABCIParams abci */
        5:
          message.abci = ABCIParams.internalBinaryRead(reader, reader.uint32(), options, message.abci);
          break;
        case /* cometbft.types.v1.SynchronyParams synchrony */
        6:
          message.synchrony = SynchronyParams.internalBinaryRead(reader, reader.uint32(), options, message.synchrony);
          break;
        case /* cometbft.types.v1.FeatureParams feature */
        7:
          message.feature = FeatureParams.internalBinaryRead(reader, reader.uint32(), options, message.feature);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.block)
      BlockParams.internalBinaryWrite(message.block, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.evidence)
      EvidenceParams.internalBinaryWrite(message.evidence, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.validator)
      ValidatorParams.internalBinaryWrite(message.validator, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.abci)
      ABCIParams.internalBinaryWrite(message.abci, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.synchrony)
      SynchronyParams.internalBinaryWrite(message.synchrony, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.feature)
      FeatureParams.internalBinaryWrite(message.feature, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("cosmos.consensus.v1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
const Msg = new ServiceType("cosmos.consensus.v1.Msg", [
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
