// @injectivelabs/core-proto-ts-v2/esm/generated/injective/tokenfactory/v1beta1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { DenomAuthorityMetadata } from "./authorityMetadata_pb.js";
import { Params } from "./params_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.GenesisState", [
      { no: 1, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } },
      { no: 2, name: "factory_denoms", kind: "message", repeat: 2, T: () => GenesisDenom, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"factory_denoms"' } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.factoryDenoms = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.tokenfactory.v1beta1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        case /* repeated injective.tokenfactory.v1beta1.GenesisDenom factory_denoms */
        2:
          message.factoryDenoms.push(GenesisDenom.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.factoryDenoms.length; i++)
      GenesisDenom.internalBinaryWrite(message.factoryDenoms[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class GenesisDenom$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.GenesisDenom", [
      { no: 1, name: "denom", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"denom"' } },
      { no: 2, name: "authority_metadata", kind: "message", T: () => DenomAuthorityMetadata, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"authority_metadata"' } },
      { no: 3, name: "name", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"name"' } },
      { no: 4, name: "symbol", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"symbol"' } },
      { no: 5, name: "decimals", kind: "scalar", T: 13, options: { "gogoproto.moretags": 'yaml:"decimals"' } }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.denom = "";
    message.name = "";
    message.symbol = "";
    message.decimals = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string denom */
        1:
          message.denom = reader.string();
          break;
        case /* injective.tokenfactory.v1beta1.DenomAuthorityMetadata authority_metadata */
        2:
          message.authorityMetadata = DenomAuthorityMetadata.internalBinaryRead(reader, reader.uint32(), options, message.authorityMetadata);
          break;
        case /* string name */
        3:
          message.name = reader.string();
          break;
        case /* string symbol */
        4:
          message.symbol = reader.string();
          break;
        case /* uint32 decimals */
        5:
          message.decimals = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.denom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.denom);
    if (message.authorityMetadata)
      DenomAuthorityMetadata.internalBinaryWrite(message.authorityMetadata, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.name !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.name);
    if (message.symbol !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.symbol);
    if (message.decimals !== 0)
      writer.tag(5, WireType.Varint).uint32(message.decimals);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisDenom = new GenesisDenom$Type();
export {
  GenesisDenom,
  GenesisState
};
