// @injectivelabs/core-proto-ts-v2/esm/generated/injective/tokenfactory/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Params defines a gRPC query method that returns the tokenfactory module's
   * parameters.
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * DenomAuthorityMetadata defines a gRPC query method for fetching
   * DenomAuthorityMetadata for a particular denom.
   *
   * @generated from protobuf rpc: DenomAuthorityMetadata
   */
  denomAuthorityMetadata(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * DenomsFromCreator defines a gRPC query method for fetching all
   * denominations created by a specific admin/creator.
   *
   * @generated from protobuf rpc: DenomsFromCreator
   */
  denomsFromCreator(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the entire auction module's state
   *
   * @generated from protobuf rpc: TokenfactoryModuleState
   */
  tokenfactoryModuleState(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
