// @injectivelabs/core-proto-ts-v2/esm/generated/injective/tokenfactory/v1beta1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./params_pb.js";
import { Metadata } from "../../../cosmos/bank/v1beta1/bank_pb.js";
import { Coin } from "../../../cosmos/base/v1beta1/coin_pb.js";
class MsgCreateDenom$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgCreateDenom", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"sender"' } },
      { no: 2, name: "subdenom", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"subdenom"' } },
      { no: 3, name: "name", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"name"' } },
      { no: 4, name: "symbol", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"symbol"' } },
      { no: 5, name: "decimals", kind: "scalar", T: 13, options: { "gogoproto.moretags": 'yaml:"decimals"' } },
      { no: 6, name: "allow_admin_burn", kind: "scalar", T: 8, options: { "gogoproto.moretags": 'yaml:"allow_admin_burn"' } }
    ], { "amino.name": "injective/tokenfactory/create-denom", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.subdenom = "";
    message.name = "";
    message.symbol = "";
    message.decimals = 0;
    message.allowAdminBurn = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string subdenom */
        2:
          message.subdenom = reader.string();
          break;
        case /* string name */
        3:
          message.name = reader.string();
          break;
        case /* string symbol */
        4:
          message.symbol = reader.string();
          break;
        case /* uint32 decimals */
        5:
          message.decimals = reader.uint32();
          break;
        case /* bool allow_admin_burn */
        6:
          message.allowAdminBurn = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.subdenom !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.subdenom);
    if (message.name !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.name);
    if (message.symbol !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.symbol);
    if (message.decimals !== 0)
      writer.tag(5, WireType.Varint).uint32(message.decimals);
    if (message.allowAdminBurn !== false)
      writer.tag(6, WireType.Varint).bool(message.allowAdminBurn);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateDenom = new MsgCreateDenom$Type();
class MsgCreateDenomResponse$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgCreateDenomResponse", [
      { no: 1, name: "new_token_denom", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"new_token_denom"' } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.newTokenDenom = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string new_token_denom */
        1:
          message.newTokenDenom = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.newTokenDenom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.newTokenDenom);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateDenomResponse = new MsgCreateDenomResponse$Type();
class MsgMint$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgMint", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"sender"' } },
      { no: 2, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"amount"' } },
      { no: 3, name: "receiver", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"receiver"' } }
    ], { "amino.name": "injective/tokenfactory/mint", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.receiver = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        2:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        case /* string receiver */
        3:
          message.receiver = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.receiver !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.receiver);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgMint = new MsgMint$Type();
class MsgMintResponse$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgMintResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgMintResponse = new MsgMintResponse$Type();
class MsgBurn$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgBurn", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"sender"' } },
      { no: 2, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"amount"' } },
      { no: 3, name: "burnFromAddress", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"burn_from_address"', "amino.dont_omitempty": true } }
    ], { "amino.name": "injective/tokenfactory/burn", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.burnFromAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        2:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        case /* string burnFromAddress */
        3:
          message.burnFromAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.burnFromAddress !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.burnFromAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgBurn = new MsgBurn$Type();
class MsgBurnResponse$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgBurnResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgBurnResponse = new MsgBurnResponse$Type();
class MsgChangeAdmin$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgChangeAdmin", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"sender"' } },
      { no: 2, name: "denom", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"denom"' } },
      { no: 3, name: "new_admin", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"new_admin"' } }
    ], { "amino.name": "injective/tokenfactory/change-admin", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.denom = "";
    message.newAdmin = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string denom */
        2:
          message.denom = reader.string();
          break;
        case /* string new_admin */
        3:
          message.newAdmin = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.denom !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.denom);
    if (message.newAdmin !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.newAdmin);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChangeAdmin = new MsgChangeAdmin$Type();
class MsgChangeAdminResponse$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgChangeAdminResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChangeAdminResponse = new MsgChangeAdminResponse$Type();
class MsgSetDenomMetadata$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgSetDenomMetadata", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"sender"' } },
      { no: 2, name: "metadata", kind: "message", T: () => Metadata, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"metadata"' } },
      { no: 3, name: "admin_burn_disabled", kind: "message", T: () => MsgSetDenomMetadata_AdminBurnDisabled, options: { "gogoproto.moretags": 'yaml:"admin_burn_disabled"' } }
    ], { "amino.name": "injective/tokenfactory/set-denom-metadata", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* cosmos.bank.v1beta1.Metadata metadata */
        2:
          message.metadata = Metadata.internalBinaryRead(reader, reader.uint32(), options, message.metadata);
          break;
        case /* injective.tokenfactory.v1beta1.MsgSetDenomMetadata.AdminBurnDisabled admin_burn_disabled */
        3:
          message.adminBurnDisabled = MsgSetDenomMetadata_AdminBurnDisabled.internalBinaryRead(reader, reader.uint32(), options, message.adminBurnDisabled);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.metadata)
      Metadata.internalBinaryWrite(message.metadata, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.adminBurnDisabled)
      MsgSetDenomMetadata_AdminBurnDisabled.internalBinaryWrite(message.adminBurnDisabled, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSetDenomMetadata = new MsgSetDenomMetadata$Type();
class MsgSetDenomMetadata_AdminBurnDisabled$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgSetDenomMetadata.AdminBurnDisabled", [
      { no: 1, name: "should_disable", kind: "scalar", T: 8, options: { "gogoproto.moretags": 'yaml:"should_disable"' } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.shouldDisable = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bool should_disable */
        1:
          message.shouldDisable = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.shouldDisable !== false)
      writer.tag(1, WireType.Varint).bool(message.shouldDisable);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSetDenomMetadata_AdminBurnDisabled = new MsgSetDenomMetadata_AdminBurnDisabled$Type();
class MsgSetDenomMetadataResponse$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgSetDenomMetadataResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSetDenomMetadataResponse = new MsgSetDenomMetadataResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgUpdateParams", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ], { "amino.name": "injective/tokenfactory/update-params", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* injective.tokenfactory.v1beta1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("injective.tokenfactory.v1beta1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
const Msg = new ServiceType("injective.tokenfactory.v1beta1.Msg", [
  { name: "CreateDenom", options: {}, I: MsgCreateDenom, O: MsgCreateDenomResponse },
  { name: "Mint", options: {}, I: MsgMint, O: MsgMintResponse },
  { name: "Burn", options: {}, I: MsgBurn, O: MsgBurnResponse },
  { name: "ChangeAdmin", options: {}, I: MsgChangeAdmin, O: MsgChangeAdminResponse },
  { name: "SetDenomMetadata", options: {}, I: MsgSetDenomMetadata, O: MsgSetDenomMetadataResponse },
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgBurn,
  MsgBurnResponse,
  MsgChangeAdmin,
  MsgChangeAdminResponse,
  MsgCreateDenom,
  MsgCreateDenomResponse,
  MsgMint,
  MsgMintResponse,
  MsgSetDenomMetadata,
  MsgSetDenomMetadataResponse,
  MsgSetDenomMetadata_AdminBurnDisabled,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
