// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/channel/v1/channel_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Height } from "../../client/v1/client_pb.js";
var State = /* @__PURE__ */ ((State2) => {
  State2[State2["UNINITIALIZED_UNSPECIFIED"] = 0] = "UNINITIALIZED_UNSPECIFIED";
  State2[State2["INIT"] = 1] = "INIT";
  State2[State2["TRYOPEN"] = 2] = "TRYOPEN";
  State2[State2["OPEN"] = 3] = "OPEN";
  State2[State2["CLOSED"] = 4] = "CLOSED";
  State2[State2["FLUSHING"] = 5] = "FLUSHING";
  State2[State2["FLUSHCOMPLETE"] = 6] = "FLUSHCOMPLETE";
  return State2;
})(State || {});
var Order = /* @__PURE__ */ ((Order2) => {
  Order2[Order2["NONE_UNSPECIFIED"] = 0] = "NONE_UNSPECIFIED";
  Order2[Order2["UNORDERED"] = 1] = "UNORDERED";
  Order2[Order2["ORDERED"] = 2] = "ORDERED";
  return Order2;
})(Order || {});
class Channel$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.Channel", [
      { no: 1, name: "state", kind: "enum", T: () => ["ibc.core.channel.v1.State", State, "STATE_"] },
      { no: 2, name: "ordering", kind: "enum", T: () => ["ibc.core.channel.v1.Order", Order, "ORDER_"] },
      { no: 3, name: "counterparty", kind: "message", T: () => Counterparty, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "connection_hops",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "upgrade_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.state = 0;
    message.ordering = 0;
    message.connectionHops = [];
    message.version = "";
    message.upgradeSequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.State state */
        1:
          message.state = reader.int32();
          break;
        case /* ibc.core.channel.v1.Order ordering */
        2:
          message.ordering = reader.int32();
          break;
        case /* ibc.core.channel.v1.Counterparty counterparty */
        3:
          message.counterparty = Counterparty.internalBinaryRead(reader, reader.uint32(), options, message.counterparty);
          break;
        case /* repeated string connection_hops */
        4:
          message.connectionHops.push(reader.string());
          break;
        case /* string version */
        5:
          message.version = reader.string();
          break;
        case /* uint64 upgrade_sequence */
        6:
          message.upgradeSequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.state !== 0)
      writer.tag(1, WireType.Varint).int32(message.state);
    if (message.ordering !== 0)
      writer.tag(2, WireType.Varint).int32(message.ordering);
    if (message.counterparty)
      Counterparty.internalBinaryWrite(message.counterparty, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.connectionHops.length; i++)
      writer.tag(4, WireType.LengthDelimited).string(message.connectionHops[i]);
    if (message.version !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.version);
    if (message.upgradeSequence !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.upgradeSequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Channel = new Channel$Type();
class IdentifiedChannel$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.IdentifiedChannel", [
      { no: 1, name: "state", kind: "enum", T: () => ["ibc.core.channel.v1.State", State, "STATE_"] },
      { no: 2, name: "ordering", kind: "enum", T: () => ["ibc.core.channel.v1.Order", Order, "ORDER_"] },
      { no: 3, name: "counterparty", kind: "message", T: () => Counterparty, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "connection_hops",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "upgrade_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.state = 0;
    message.ordering = 0;
    message.connectionHops = [];
    message.version = "";
    message.portId = "";
    message.channelId = "";
    message.upgradeSequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.State state */
        1:
          message.state = reader.int32();
          break;
        case /* ibc.core.channel.v1.Order ordering */
        2:
          message.ordering = reader.int32();
          break;
        case /* ibc.core.channel.v1.Counterparty counterparty */
        3:
          message.counterparty = Counterparty.internalBinaryRead(reader, reader.uint32(), options, message.counterparty);
          break;
        case /* repeated string connection_hops */
        4:
          message.connectionHops.push(reader.string());
          break;
        case /* string version */
        5:
          message.version = reader.string();
          break;
        case /* string port_id */
        6:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        7:
          message.channelId = reader.string();
          break;
        case /* uint64 upgrade_sequence */
        8:
          message.upgradeSequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.state !== 0)
      writer.tag(1, WireType.Varint).int32(message.state);
    if (message.ordering !== 0)
      writer.tag(2, WireType.Varint).int32(message.ordering);
    if (message.counterparty)
      Counterparty.internalBinaryWrite(message.counterparty, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.connectionHops.length; i++)
      writer.tag(4, WireType.LengthDelimited).string(message.connectionHops[i]);
    if (message.version !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.version);
    if (message.portId !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.channelId);
    if (message.upgradeSequence !== 0n)
      writer.tag(8, WireType.Varint).uint64(message.upgradeSequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const IdentifiedChannel = new IdentifiedChannel$Type();
class Counterparty$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.Counterparty", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Counterparty = new Counterparty$Type();
class Packet$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.Packet", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "source_port",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "source_channel",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "destination_port",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "destination_channel",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 7, name: "timeout_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 8,
        name: "timeout_timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    message.sourcePort = "";
    message.sourceChannel = "";
    message.destinationPort = "";
    message.destinationChannel = "";
    message.data = new Uint8Array(0);
    message.timeoutTimestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* string source_port */
        2:
          message.sourcePort = reader.string();
          break;
        case /* string source_channel */
        3:
          message.sourceChannel = reader.string();
          break;
        case /* string destination_port */
        4:
          message.destinationPort = reader.string();
          break;
        case /* string destination_channel */
        5:
          message.destinationChannel = reader.string();
          break;
        case /* bytes data */
        6:
          message.data = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height timeout_height */
        7:
          message.timeoutHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.timeoutHeight);
          break;
        case /* uint64 timeout_timestamp */
        8:
          message.timeoutTimestamp = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    if (message.sourcePort !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.sourcePort);
    if (message.sourceChannel !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.sourceChannel);
    if (message.destinationPort !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.destinationPort);
    if (message.destinationChannel !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.destinationChannel);
    if (message.data.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.data);
    if (message.timeoutHeight)
      Height.internalBinaryWrite(message.timeoutHeight, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.timeoutTimestamp !== 0n)
      writer.tag(8, WireType.Varint).uint64(message.timeoutTimestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Packet = new Packet$Type();
class PacketState$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.PacketState", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "data",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.sequence = 0n;
    message.data = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* uint64 sequence */
        3:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* bytes data */
        4:
          message.data = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.sequence !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.sequence);
    if (message.data.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.data);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketState = new PacketState$Type();
class PacketId$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.PacketId", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* uint64 sequence */
        3:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.sequence !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketId = new PacketId$Type();
class Acknowledgement$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.Acknowledgement", [
      {
        no: 21,
        name: "result",
        kind: "scalar",
        oneof: "response",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 22,
        name: "error",
        kind: "scalar",
        oneof: "response",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.response = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes result */
        21:
          message.response = {
            oneofKind: "result",
            result: reader.bytes()
          };
          break;
        case /* string error */
        22:
          message.response = {
            oneofKind: "error",
            error: reader.string()
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.response.oneofKind === "result")
      writer.tag(21, WireType.LengthDelimited).bytes(message.response.result);
    if (message.response.oneofKind === "error")
      writer.tag(22, WireType.LengthDelimited).string(message.response.error);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Acknowledgement = new Acknowledgement$Type();
class Timeout$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.Timeout", [
      { no: 1, name: "height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "timestamp",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.timestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.client.v1.Height height */
        1:
          message.height = Height.internalBinaryRead(reader, reader.uint32(), options, message.height);
          break;
        case /* uint64 timestamp */
        2:
          message.timestamp = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height)
      Height.internalBinaryWrite(message.height, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.timestamp !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.timestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Timeout = new Timeout$Type();
class Params$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.Params", [
      { no: 1, name: "upgrade_timeout", kind: "message", T: () => Timeout, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Timeout upgrade_timeout */
        1:
          message.upgradeTimeout = Timeout.internalBinaryRead(reader, reader.uint32(), options, message.upgradeTimeout);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.upgradeTimeout)
      Timeout.internalBinaryWrite(message.upgradeTimeout, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
export {
  Acknowledgement,
  Channel,
  Counterparty,
  IdentifiedChannel,
  Order,
  Packet,
  PacketId,
  PacketState,
  Params,
  State,
  Timeout
};
