// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/channel/v1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./channel_pb.js";
import { PacketState } from "./channel_pb.js";
import { IdentifiedChannel } from "./channel_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.GenesisState", [
      { no: 1, name: "channels", kind: "message", repeat: 2, T: () => IdentifiedChannel, options: { "gogoproto.nullable": false, "gogoproto.casttype": "IdentifiedChannel" } },
      { no: 2, name: "acknowledgements", kind: "message", repeat: 2, T: () => PacketState, options: { "gogoproto.nullable": false } },
      { no: 3, name: "commitments", kind: "message", repeat: 2, T: () => PacketState, options: { "gogoproto.nullable": false } },
      { no: 4, name: "receipts", kind: "message", repeat: 2, T: () => PacketState, options: { "gogoproto.nullable": false } },
      { no: 5, name: "send_sequences", kind: "message", repeat: 2, T: () => PacketSequence, options: { "gogoproto.nullable": false } },
      { no: 6, name: "recv_sequences", kind: "message", repeat: 2, T: () => PacketSequence, options: { "gogoproto.nullable": false } },
      { no: 7, name: "ack_sequences", kind: "message", repeat: 2, T: () => PacketSequence, options: { "gogoproto.nullable": false } },
      {
        no: 8,
        name: "next_channel_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 9, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channels = [];
    message.acknowledgements = [];
    message.commitments = [];
    message.receipts = [];
    message.sendSequences = [];
    message.recvSequences = [];
    message.ackSequences = [];
    message.nextChannelSequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.core.channel.v1.IdentifiedChannel channels */
        1:
          message.channels.push(IdentifiedChannel.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.core.channel.v1.PacketState acknowledgements */
        2:
          message.acknowledgements.push(PacketState.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.core.channel.v1.PacketState commitments */
        3:
          message.commitments.push(PacketState.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.core.channel.v1.PacketState receipts */
        4:
          message.receipts.push(PacketState.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.core.channel.v1.PacketSequence send_sequences */
        5:
          message.sendSequences.push(PacketSequence.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.core.channel.v1.PacketSequence recv_sequences */
        6:
          message.recvSequences.push(PacketSequence.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.core.channel.v1.PacketSequence ack_sequences */
        7:
          message.ackSequences.push(PacketSequence.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* uint64 next_channel_sequence */
        8:
          message.nextChannelSequence = reader.uint64().toBigInt();
          break;
        case /* ibc.core.channel.v1.Params params */
        9:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.channels.length; i++)
      IdentifiedChannel.internalBinaryWrite(message.channels[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.acknowledgements.length; i++)
      PacketState.internalBinaryWrite(message.acknowledgements[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.commitments.length; i++)
      PacketState.internalBinaryWrite(message.commitments[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.receipts.length; i++)
      PacketState.internalBinaryWrite(message.receipts[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.sendSequences.length; i++)
      PacketSequence.internalBinaryWrite(message.sendSequences[i], writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.recvSequences.length; i++)
      PacketSequence.internalBinaryWrite(message.recvSequences[i], writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.ackSequences.length; i++)
      PacketSequence.internalBinaryWrite(message.ackSequences[i], writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.nextChannelSequence !== 0n)
      writer.tag(8, WireType.Varint).uint64(message.nextChannelSequence);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(9, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class PacketSequence$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.PacketSequence", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* uint64 sequence */
        3:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.sequence !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const PacketSequence = new PacketSequence$Type();
export {
  GenesisState,
  PacketSequence
};
