// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/channel/v1/query_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./channel_pb.js";
import { Upgrade } from "./upgrade_pb.js";
import { ErrorReceipt } from "./upgrade_pb.js";
import { PacketState } from "./channel_pb.js";
import { Any } from "../../../../google/protobuf/any_pb.js";
import { IdentifiedClientState } from "../../client/v1/client_pb.js";
import { PageResponse } from "../../../../cosmos/base/query/v1beta1/pagination_pb.js";
import { IdentifiedChannel } from "./channel_pb.js";
import { PageRequest } from "../../../../cosmos/base/query/v1beta1/pagination_pb.js";
import { Height } from "../../client/v1/client_pb.js";
import { Channel } from "./channel_pb.js";
class QueryChannelRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelRequest = new QueryChannelRequest$Type();
class QueryChannelResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelResponse", [
      { no: 1, name: "channel", kind: "message", T: () => Channel },
      {
        no: 2,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Channel channel */
        1:
          message.channel = Channel.internalBinaryRead(reader, reader.uint32(), options, message.channel);
          break;
        case /* bytes proof */
        2:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channel)
      Channel.internalBinaryWrite(message.channel, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.proof.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelResponse = new QueryChannelResponse$Type();
class QueryChannelsRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelsRequest", [
      { no: 1, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        1:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelsRequest = new QueryChannelsRequest$Type();
class QueryChannelsResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelsResponse", [
      { no: 1, name: "channels", kind: "message", repeat: 2, T: () => IdentifiedChannel },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse },
      { no: 3, name: "height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channels = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.core.channel.v1.IdentifiedChannel channels */
        1:
          message.channels.push(IdentifiedChannel.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* ibc.core.client.v1.Height height */
        3:
          message.height = Height.internalBinaryRead(reader, reader.uint32(), options, message.height);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.channels.length; i++)
      IdentifiedChannel.internalBinaryWrite(message.channels[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.height)
      Height.internalBinaryWrite(message.height, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelsResponse = new QueryChannelsResponse$Type();
class QueryConnectionChannelsRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryConnectionChannelsRequest", [
      {
        no: 1,
        name: "connection",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.connection = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string connection */
        1:
          message.connection = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        2:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.connection !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.connection);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryConnectionChannelsRequest = new QueryConnectionChannelsRequest$Type();
class QueryConnectionChannelsResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryConnectionChannelsResponse", [
      { no: 1, name: "channels", kind: "message", repeat: 2, T: () => IdentifiedChannel },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse },
      { no: 3, name: "height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channels = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.core.channel.v1.IdentifiedChannel channels */
        1:
          message.channels.push(IdentifiedChannel.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* ibc.core.client.v1.Height height */
        3:
          message.height = Height.internalBinaryRead(reader, reader.uint32(), options, message.height);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.channels.length; i++)
      IdentifiedChannel.internalBinaryWrite(message.channels[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.height)
      Height.internalBinaryWrite(message.height, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryConnectionChannelsResponse = new QueryConnectionChannelsResponse$Type();
class QueryChannelClientStateRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelClientStateRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelClientStateRequest = new QueryChannelClientStateRequest$Type();
class QueryChannelClientStateResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelClientStateResponse", [
      { no: 1, name: "identified_client_state", kind: "message", T: () => IdentifiedClientState },
      {
        no: 2,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.client.v1.IdentifiedClientState identified_client_state */
        1:
          message.identifiedClientState = IdentifiedClientState.internalBinaryRead(reader, reader.uint32(), options, message.identifiedClientState);
          break;
        case /* bytes proof */
        2:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.identifiedClientState)
      IdentifiedClientState.internalBinaryWrite(message.identifiedClientState, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.proof.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelClientStateResponse = new QueryChannelClientStateResponse$Type();
class QueryChannelConsensusStateRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelConsensusStateRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "revision_number",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "revision_height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.revisionNumber = 0n;
    message.revisionHeight = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* uint64 revision_number */
        3:
          message.revisionNumber = reader.uint64().toBigInt();
          break;
        case /* uint64 revision_height */
        4:
          message.revisionHeight = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.revisionNumber !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.revisionNumber);
    if (message.revisionHeight !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.revisionHeight);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelConsensusStateRequest = new QueryChannelConsensusStateRequest$Type();
class QueryChannelConsensusStateResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelConsensusStateResponse", [
      { no: 1, name: "consensus_state", kind: "message", T: () => Any },
      {
        no: 2,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 4, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clientId = "";
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any consensus_state */
        1:
          message.consensusState = Any.internalBinaryRead(reader, reader.uint32(), options, message.consensusState);
          break;
        case /* string client_id */
        2:
          message.clientId = reader.string();
          break;
        case /* bytes proof */
        3:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        4:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.consensusState)
      Any.internalBinaryWrite(message.consensusState, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.clientId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.clientId);
    if (message.proof.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelConsensusStateResponse = new QueryChannelConsensusStateResponse$Type();
class QueryPacketCommitmentRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketCommitmentRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* uint64 sequence */
        3:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.sequence !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketCommitmentRequest = new QueryPacketCommitmentRequest$Type();
class QueryPacketCommitmentResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketCommitmentResponse", [
      {
        no: 1,
        name: "commitment",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.commitment = new Uint8Array(0);
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes commitment */
        1:
          message.commitment = reader.bytes();
          break;
        case /* bytes proof */
        2:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.commitment.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.commitment);
    if (message.proof.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketCommitmentResponse = new QueryPacketCommitmentResponse$Type();
class QueryPacketCommitmentsRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketCommitmentsRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "pagination", kind: "message", T: () => PageRequest }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        3:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketCommitmentsRequest = new QueryPacketCommitmentsRequest$Type();
class QueryPacketCommitmentsResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketCommitmentsResponse", [
      { no: 1, name: "commitments", kind: "message", repeat: 2, T: () => PacketState },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse },
      { no: 3, name: "height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.commitments = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.core.channel.v1.PacketState commitments */
        1:
          message.commitments.push(PacketState.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* ibc.core.client.v1.Height height */
        3:
          message.height = Height.internalBinaryRead(reader, reader.uint32(), options, message.height);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.commitments.length; i++)
      PacketState.internalBinaryWrite(message.commitments[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.height)
      Height.internalBinaryWrite(message.height, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketCommitmentsResponse = new QueryPacketCommitmentsResponse$Type();
class QueryPacketReceiptRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketReceiptRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* uint64 sequence */
        3:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.sequence !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketReceiptRequest = new QueryPacketReceiptRequest$Type();
class QueryPacketReceiptResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketReceiptResponse", [
      {
        no: 2,
        name: "received",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 3,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 4, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.received = false;
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bool received */
        2:
          message.received = reader.bool();
          break;
        case /* bytes proof */
        3:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        4:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.received !== false)
      writer.tag(2, WireType.Varint).bool(message.received);
    if (message.proof.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketReceiptResponse = new QueryPacketReceiptResponse$Type();
class QueryPacketAcknowledgementRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketAcknowledgementRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.sequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* uint64 sequence */
        3:
          message.sequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.sequence !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.sequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketAcknowledgementRequest = new QueryPacketAcknowledgementRequest$Type();
class QueryPacketAcknowledgementResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketAcknowledgementResponse", [
      {
        no: 1,
        name: "acknowledgement",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.acknowledgement = new Uint8Array(0);
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes acknowledgement */
        1:
          message.acknowledgement = reader.bytes();
          break;
        case /* bytes proof */
        2:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.acknowledgement.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.acknowledgement);
    if (message.proof.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketAcknowledgementResponse = new QueryPacketAcknowledgementResponse$Type();
class QueryPacketAcknowledgementsRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketAcknowledgementsRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "pagination", kind: "message", T: () => PageRequest },
      {
        no: 4,
        name: "packet_commitment_sequences",
        kind: "scalar",
        repeat: 1,
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.packetCommitmentSequences = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* cosmos.base.query.v1beta1.PageRequest pagination */
        3:
          message.pagination = PageRequest.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* repeated uint64 packet_commitment_sequences */
        4:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.packetCommitmentSequences.push(reader.uint64().toBigInt());
          else
            message.packetCommitmentSequences.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.pagination)
      PageRequest.internalBinaryWrite(message.pagination, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.packetCommitmentSequences.length) {
      writer.tag(4, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.packetCommitmentSequences.length; i++)
        writer.uint64(message.packetCommitmentSequences[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketAcknowledgementsRequest = new QueryPacketAcknowledgementsRequest$Type();
class QueryPacketAcknowledgementsResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryPacketAcknowledgementsResponse", [
      { no: 1, name: "acknowledgements", kind: "message", repeat: 2, T: () => PacketState },
      { no: 2, name: "pagination", kind: "message", T: () => PageResponse },
      { no: 3, name: "height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.acknowledgements = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.core.channel.v1.PacketState acknowledgements */
        1:
          message.acknowledgements.push(PacketState.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* cosmos.base.query.v1beta1.PageResponse pagination */
        2:
          message.pagination = PageResponse.internalBinaryRead(reader, reader.uint32(), options, message.pagination);
          break;
        case /* ibc.core.client.v1.Height height */
        3:
          message.height = Height.internalBinaryRead(reader, reader.uint32(), options, message.height);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.acknowledgements.length; i++)
      PacketState.internalBinaryWrite(message.acknowledgements[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.pagination)
      PageResponse.internalBinaryWrite(message.pagination, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.height)
      Height.internalBinaryWrite(message.height, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryPacketAcknowledgementsResponse = new QueryPacketAcknowledgementsResponse$Type();
class QueryUnreceivedPacketsRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryUnreceivedPacketsRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "packet_commitment_sequences",
        kind: "scalar",
        repeat: 1,
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.packetCommitmentSequences = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* repeated uint64 packet_commitment_sequences */
        3:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.packetCommitmentSequences.push(reader.uint64().toBigInt());
          else
            message.packetCommitmentSequences.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.packetCommitmentSequences.length) {
      writer.tag(3, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.packetCommitmentSequences.length; i++)
        writer.uint64(message.packetCommitmentSequences[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUnreceivedPacketsRequest = new QueryUnreceivedPacketsRequest$Type();
class QueryUnreceivedPacketsResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryUnreceivedPacketsResponse", [
      {
        no: 1,
        name: "sequences",
        kind: "scalar",
        repeat: 1,
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequences = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated uint64 sequences */
        1:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.sequences.push(reader.uint64().toBigInt());
          else
            message.sequences.push(reader.uint64().toBigInt());
          break;
        case /* ibc.core.client.v1.Height height */
        2:
          message.height = Height.internalBinaryRead(reader, reader.uint32(), options, message.height);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequences.length) {
      writer.tag(1, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.sequences.length; i++)
        writer.uint64(message.sequences[i]);
      writer.join();
    }
    if (message.height)
      Height.internalBinaryWrite(message.height, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUnreceivedPacketsResponse = new QueryUnreceivedPacketsResponse$Type();
class QueryUnreceivedAcksRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryUnreceivedAcksRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "packet_ack_sequences",
        kind: "scalar",
        repeat: 1,
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.packetAckSequences = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* repeated uint64 packet_ack_sequences */
        3:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.packetAckSequences.push(reader.uint64().toBigInt());
          else
            message.packetAckSequences.push(reader.uint64().toBigInt());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.packetAckSequences.length) {
      writer.tag(3, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.packetAckSequences.length; i++)
        writer.uint64(message.packetAckSequences[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUnreceivedAcksRequest = new QueryUnreceivedAcksRequest$Type();
class QueryUnreceivedAcksResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryUnreceivedAcksResponse", [
      {
        no: 1,
        name: "sequences",
        kind: "scalar",
        repeat: 1,
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequences = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated uint64 sequences */
        1:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.sequences.push(reader.uint64().toBigInt());
          else
            message.sequences.push(reader.uint64().toBigInt());
          break;
        case /* ibc.core.client.v1.Height height */
        2:
          message.height = Height.internalBinaryRead(reader, reader.uint32(), options, message.height);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequences.length) {
      writer.tag(1, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.sequences.length; i++)
        writer.uint64(message.sequences[i]);
      writer.join();
    }
    if (message.height)
      Height.internalBinaryWrite(message.height, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUnreceivedAcksResponse = new QueryUnreceivedAcksResponse$Type();
class QueryNextSequenceReceiveRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryNextSequenceReceiveRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryNextSequenceReceiveRequest = new QueryNextSequenceReceiveRequest$Type();
class QueryNextSequenceReceiveResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryNextSequenceReceiveResponse", [
      {
        no: 1,
        name: "next_sequence_receive",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nextSequenceReceive = 0n;
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 next_sequence_receive */
        1:
          message.nextSequenceReceive = reader.uint64().toBigInt();
          break;
        case /* bytes proof */
        2:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nextSequenceReceive !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.nextSequenceReceive);
    if (message.proof.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryNextSequenceReceiveResponse = new QueryNextSequenceReceiveResponse$Type();
class QueryNextSequenceSendRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryNextSequenceSendRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryNextSequenceSendRequest = new QueryNextSequenceSendRequest$Type();
class QueryNextSequenceSendResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryNextSequenceSendResponse", [
      {
        no: 1,
        name: "next_sequence_send",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nextSequenceSend = 0n;
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 next_sequence_send */
        1:
          message.nextSequenceSend = reader.uint64().toBigInt();
          break;
        case /* bytes proof */
        2:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.nextSequenceSend !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.nextSequenceSend);
    if (message.proof.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryNextSequenceSendResponse = new QueryNextSequenceSendResponse$Type();
class QueryUpgradeErrorRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryUpgradeErrorRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUpgradeErrorRequest = new QueryUpgradeErrorRequest$Type();
class QueryUpgradeErrorResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryUpgradeErrorResponse", [
      { no: 1, name: "error_receipt", kind: "message", T: () => ErrorReceipt, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.ErrorReceipt error_receipt */
        1:
          message.errorReceipt = ErrorReceipt.internalBinaryRead(reader, reader.uint32(), options, message.errorReceipt);
          break;
        case /* bytes proof */
        2:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.errorReceipt)
      ErrorReceipt.internalBinaryWrite(message.errorReceipt, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.proof.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUpgradeErrorResponse = new QueryUpgradeErrorResponse$Type();
class QueryUpgradeRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryUpgradeRequest", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUpgradeRequest = new QueryUpgradeRequest$Type();
class QueryUpgradeResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryUpgradeResponse", [
      { no: 1, name: "upgrade", kind: "message", T: () => Upgrade, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "proof",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proof = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Upgrade upgrade */
        1:
          message.upgrade = Upgrade.internalBinaryRead(reader, reader.uint32(), options, message.upgrade);
          break;
        case /* bytes proof */
        2:
          message.proof = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.upgrade)
      Upgrade.internalBinaryWrite(message.upgrade, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.proof.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proof);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryUpgradeResponse = new QueryUpgradeResponse$Type();
class QueryChannelParamsRequest$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelParamsRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelParamsRequest = new QueryChannelParamsRequest$Type();
class QueryChannelParamsResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.QueryChannelParamsResponse", [
      { no: 1, name: "params", kind: "message", T: () => Params }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryChannelParamsResponse = new QueryChannelParamsResponse$Type();
const Query = new ServiceType("ibc.core.channel.v1.Query", [
  { name: "Channel", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}" } }, I: QueryChannelRequest, O: QueryChannelResponse },
  { name: "Channels", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels" } }, I: QueryChannelsRequest, O: QueryChannelsResponse },
  { name: "ConnectionChannels", options: { "google.api.http": { get: "/ibc/core/channel/v1/connections/{connection}/channels" } }, I: QueryConnectionChannelsRequest, O: QueryConnectionChannelsResponse },
  { name: "ChannelClientState", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/client_state" } }, I: QueryChannelClientStateRequest, O: QueryChannelClientStateResponse },
  { name: "ChannelConsensusState", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/consensus_state/revision/{revision_number}/height/{revision_height}" } }, I: QueryChannelConsensusStateRequest, O: QueryChannelConsensusStateResponse },
  { name: "PacketCommitment", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/packet_commitments/{sequence}" } }, I: QueryPacketCommitmentRequest, O: QueryPacketCommitmentResponse },
  { name: "PacketCommitments", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/packet_commitments" } }, I: QueryPacketCommitmentsRequest, O: QueryPacketCommitmentsResponse },
  { name: "PacketReceipt", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/packet_receipts/{sequence}" } }, I: QueryPacketReceiptRequest, O: QueryPacketReceiptResponse },
  { name: "PacketAcknowledgement", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/packet_acks/{sequence}" } }, I: QueryPacketAcknowledgementRequest, O: QueryPacketAcknowledgementResponse },
  { name: "PacketAcknowledgements", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/packet_acknowledgements" } }, I: QueryPacketAcknowledgementsRequest, O: QueryPacketAcknowledgementsResponse },
  { name: "UnreceivedPackets", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/packet_commitments/{packet_commitment_sequences}/unreceived_packets" } }, I: QueryUnreceivedPacketsRequest, O: QueryUnreceivedPacketsResponse },
  { name: "UnreceivedAcks", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/packet_commitments/{packet_ack_sequences}/unreceived_acks" } }, I: QueryUnreceivedAcksRequest, O: QueryUnreceivedAcksResponse },
  { name: "NextSequenceReceive", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/next_sequence" } }, I: QueryNextSequenceReceiveRequest, O: QueryNextSequenceReceiveResponse },
  { name: "NextSequenceSend", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/next_sequence_send" } }, I: QueryNextSequenceSendRequest, O: QueryNextSequenceSendResponse },
  { name: "UpgradeError", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/upgrade_error" } }, I: QueryUpgradeErrorRequest, O: QueryUpgradeErrorResponse },
  { name: "Upgrade", options: { "google.api.http": { get: "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/upgrade" } }, I: QueryUpgradeRequest, O: QueryUpgradeResponse },
  { name: "ChannelParams", options: { "google.api.http": { get: "/ibc/core/channel/v1/params" } }, I: QueryChannelParamsRequest, O: QueryChannelParamsResponse }
]);
export {
  Query,
  QueryChannelClientStateRequest,
  QueryChannelClientStateResponse,
  QueryChannelConsensusStateRequest,
  QueryChannelConsensusStateResponse,
  QueryChannelParamsRequest,
  QueryChannelParamsResponse,
  QueryChannelRequest,
  QueryChannelResponse,
  QueryChannelsRequest,
  QueryChannelsResponse,
  QueryConnectionChannelsRequest,
  QueryConnectionChannelsResponse,
  QueryNextSequenceReceiveRequest,
  QueryNextSequenceReceiveResponse,
  QueryNextSequenceSendRequest,
  QueryNextSequenceSendResponse,
  QueryPacketAcknowledgementRequest,
  QueryPacketAcknowledgementResponse,
  QueryPacketAcknowledgementsRequest,
  QueryPacketAcknowledgementsResponse,
  QueryPacketCommitmentRequest,
  QueryPacketCommitmentResponse,
  QueryPacketCommitmentsRequest,
  QueryPacketCommitmentsResponse,
  QueryPacketReceiptRequest,
  QueryPacketReceiptResponse,
  QueryUnreceivedAcksRequest,
  QueryUnreceivedAcksResponse,
  QueryUnreceivedPacketsRequest,
  QueryUnreceivedPacketsResponse,
  QueryUpgradeErrorRequest,
  QueryUpgradeErrorResponse,
  QueryUpgradeRequest,
  QueryUpgradeResponse
};
