// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/channel/v1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./channel_pb.js";
import { ErrorReceipt } from "./upgrade_pb.js";
import { State } from "./channel_pb.js";
import { Upgrade } from "./upgrade_pb.js";
import { UpgradeFields } from "./upgrade_pb.js";
import { Packet } from "./channel_pb.js";
import { Height } from "../../client/v1/client_pb.js";
import { Channel } from "./channel_pb.js";
var ResponseResultType = /* @__PURE__ */ ((ResponseResultType2) => {
  ResponseResultType2[ResponseResultType2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  ResponseResultType2[ResponseResultType2["NOOP"] = 1] = "NOOP";
  ResponseResultType2[ResponseResultType2["SUCCESS"] = 2] = "SUCCESS";
  ResponseResultType2[ResponseResultType2["FAILURE"] = 3] = "FAILURE";
  return ResponseResultType2;
})(ResponseResultType || {});
class MsgChannelOpenInit$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelOpenInit", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "channel", kind: "message", T: () => Channel, options: { "gogoproto.nullable": false } },
      {
        no: 3,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* ibc.core.channel.v1.Channel channel */
        2:
          message.channel = Channel.internalBinaryRead(reader, reader.uint32(), options, message.channel);
          break;
        case /* string signer */
        3:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channel)
      Channel.internalBinaryWrite(message.channel, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelOpenInit = new MsgChannelOpenInit$Type();
class MsgChannelOpenInitResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelOpenInitResponse", [
      {
        no: 1,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.channelId = "";
    message.version = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string channel_id */
        1:
          message.channelId = reader.string();
          break;
        case /* string version */
        2:
          message.version = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.channelId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.channelId);
    if (message.version !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.version);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelOpenInitResponse = new MsgChannelOpenInitResponse$Type();
class MsgChannelOpenTry$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelOpenTry", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "previous_channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "channel", kind: "message", T: () => Channel, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "counterparty_version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "proof_init",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 6, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 7,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.previousChannelId = "";
    message.counterpartyVersion = "";
    message.proofInit = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string previous_channel_id = 2 [deprecated = true] */
        2:
          message.previousChannelId = reader.string();
          break;
        case /* ibc.core.channel.v1.Channel channel */
        3:
          message.channel = Channel.internalBinaryRead(reader, reader.uint32(), options, message.channel);
          break;
        case /* string counterparty_version */
        4:
          message.counterpartyVersion = reader.string();
          break;
        case /* bytes proof_init */
        5:
          message.proofInit = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        6:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        7:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.previousChannelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.previousChannelId);
    if (message.channel)
      Channel.internalBinaryWrite(message.channel, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.counterpartyVersion !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.counterpartyVersion);
    if (message.proofInit.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.proofInit);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelOpenTry = new MsgChannelOpenTry$Type();
class MsgChannelOpenTryResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelOpenTryResponse", [
      {
        no: 1,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.version = "";
    message.channelId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string version */
        1:
          message.version = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.version !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.version);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelOpenTryResponse = new MsgChannelOpenTryResponse$Type();
class MsgChannelOpenAck$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelOpenAck", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "counterparty_channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "counterparty_version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "proof_try",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 6, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 7,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.counterpartyChannelId = "";
    message.counterpartyVersion = "";
    message.proofTry = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* string counterparty_channel_id */
        3:
          message.counterpartyChannelId = reader.string();
          break;
        case /* string counterparty_version */
        4:
          message.counterpartyVersion = reader.string();
          break;
        case /* bytes proof_try */
        5:
          message.proofTry = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        6:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        7:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.counterpartyChannelId !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.counterpartyChannelId);
    if (message.counterpartyVersion !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.counterpartyVersion);
    if (message.proofTry.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.proofTry);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelOpenAck = new MsgChannelOpenAck$Type();
class MsgChannelOpenAckResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelOpenAckResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelOpenAckResponse = new MsgChannelOpenAckResponse$Type();
class MsgChannelOpenConfirm$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelOpenConfirm", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "proof_ack",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 4, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 5,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.proofAck = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* bytes proof_ack */
        3:
          message.proofAck = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        4:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        5:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.proofAck.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.proofAck);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelOpenConfirm = new MsgChannelOpenConfirm$Type();
class MsgChannelOpenConfirmResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelOpenConfirmResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelOpenConfirmResponse = new MsgChannelOpenConfirmResponse$Type();
class MsgChannelCloseInit$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelCloseInit", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* string signer */
        3:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.signer !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelCloseInit = new MsgChannelCloseInit$Type();
class MsgChannelCloseInitResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelCloseInitResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelCloseInitResponse = new MsgChannelCloseInitResponse$Type();
class MsgChannelCloseConfirm$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelCloseConfirm", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "proof_init",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 4, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 5,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 6,
        name: "counterparty_upgrade_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.proofInit = new Uint8Array(0);
    message.signer = "";
    message.counterpartyUpgradeSequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* bytes proof_init */
        3:
          message.proofInit = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        4:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        5:
          message.signer = reader.string();
          break;
        case /* uint64 counterparty_upgrade_sequence */
        6:
          message.counterpartyUpgradeSequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.proofInit.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.proofInit);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.signer);
    if (message.counterpartyUpgradeSequence !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.counterpartyUpgradeSequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelCloseConfirm = new MsgChannelCloseConfirm$Type();
class MsgChannelCloseConfirmResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelCloseConfirmResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelCloseConfirmResponse = new MsgChannelCloseConfirmResponse$Type();
class MsgRecvPacket$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgRecvPacket", [
      { no: 1, name: "packet", kind: "message", T: () => Packet, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "proof_commitment",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proofCommitment = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Packet packet */
        1:
          message.packet = Packet.internalBinaryRead(reader, reader.uint32(), options, message.packet);
          break;
        case /* bytes proof_commitment */
        2:
          message.proofCommitment = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        4:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packet)
      Packet.internalBinaryWrite(message.packet, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.proofCommitment.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proofCommitment);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRecvPacket = new MsgRecvPacket$Type();
class MsgRecvPacketResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgRecvPacketResponse", [
      { no: 1, name: "result", kind: "enum", T: () => ["ibc.core.channel.v1.ResponseResultType", ResponseResultType, "RESPONSE_RESULT_TYPE_"] }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.result = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.ResponseResultType result */
        1:
          message.result = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.result !== 0)
      writer.tag(1, WireType.Varint).int32(message.result);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgRecvPacketResponse = new MsgRecvPacketResponse$Type();
class MsgTimeout$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgTimeout", [
      { no: 1, name: "packet", kind: "message", T: () => Packet, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "proof_unreceived",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "next_sequence_recv",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proofUnreceived = new Uint8Array(0);
    message.nextSequenceRecv = 0n;
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Packet packet */
        1:
          message.packet = Packet.internalBinaryRead(reader, reader.uint32(), options, message.packet);
          break;
        case /* bytes proof_unreceived */
        2:
          message.proofUnreceived = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        3:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* uint64 next_sequence_recv */
        4:
          message.nextSequenceRecv = reader.uint64().toBigInt();
          break;
        case /* string signer */
        5:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packet)
      Packet.internalBinaryWrite(message.packet, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.proofUnreceived.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proofUnreceived);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.nextSequenceRecv !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.nextSequenceRecv);
    if (message.signer !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTimeout = new MsgTimeout$Type();
class MsgTimeoutResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgTimeoutResponse", [
      { no: 1, name: "result", kind: "enum", T: () => ["ibc.core.channel.v1.ResponseResultType", ResponseResultType, "RESPONSE_RESULT_TYPE_"] }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.result = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.ResponseResultType result */
        1:
          message.result = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.result !== 0)
      writer.tag(1, WireType.Varint).int32(message.result);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTimeoutResponse = new MsgTimeoutResponse$Type();
class MsgTimeoutOnClose$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgTimeoutOnClose", [
      { no: 1, name: "packet", kind: "message", T: () => Packet, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "proof_unreceived",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "proof_close",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 4, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 5,
        name: "next_sequence_recv",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "counterparty_upgrade_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.proofUnreceived = new Uint8Array(0);
    message.proofClose = new Uint8Array(0);
    message.nextSequenceRecv = 0n;
    message.signer = "";
    message.counterpartyUpgradeSequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Packet packet */
        1:
          message.packet = Packet.internalBinaryRead(reader, reader.uint32(), options, message.packet);
          break;
        case /* bytes proof_unreceived */
        2:
          message.proofUnreceived = reader.bytes();
          break;
        case /* bytes proof_close */
        3:
          message.proofClose = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        4:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* uint64 next_sequence_recv */
        5:
          message.nextSequenceRecv = reader.uint64().toBigInt();
          break;
        case /* string signer */
        6:
          message.signer = reader.string();
          break;
        case /* uint64 counterparty_upgrade_sequence */
        7:
          message.counterpartyUpgradeSequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packet)
      Packet.internalBinaryWrite(message.packet, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.proofUnreceived.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.proofUnreceived);
    if (message.proofClose.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.proofClose);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.nextSequenceRecv !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.nextSequenceRecv);
    if (message.signer !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.signer);
    if (message.counterpartyUpgradeSequence !== 0n)
      writer.tag(7, WireType.Varint).uint64(message.counterpartyUpgradeSequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTimeoutOnClose = new MsgTimeoutOnClose$Type();
class MsgTimeoutOnCloseResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgTimeoutOnCloseResponse", [
      { no: 1, name: "result", kind: "enum", T: () => ["ibc.core.channel.v1.ResponseResultType", ResponseResultType, "RESPONSE_RESULT_TYPE_"] }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.result = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.ResponseResultType result */
        1:
          message.result = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.result !== 0)
      writer.tag(1, WireType.Varint).int32(message.result);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTimeoutOnCloseResponse = new MsgTimeoutOnCloseResponse$Type();
class MsgAcknowledgement$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgAcknowledgement", [
      { no: 1, name: "packet", kind: "message", T: () => Packet, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "acknowledgement",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "proof_acked",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 4, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 5,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.acknowledgement = new Uint8Array(0);
    message.proofAcked = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Packet packet */
        1:
          message.packet = Packet.internalBinaryRead(reader, reader.uint32(), options, message.packet);
          break;
        case /* bytes acknowledgement */
        2:
          message.acknowledgement = reader.bytes();
          break;
        case /* bytes proof_acked */
        3:
          message.proofAcked = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        4:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        5:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.packet)
      Packet.internalBinaryWrite(message.packet, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.acknowledgement.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.acknowledgement);
    if (message.proofAcked.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.proofAcked);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgAcknowledgement = new MsgAcknowledgement$Type();
class MsgAcknowledgementResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgAcknowledgementResponse", [
      { no: 1, name: "result", kind: "enum", T: () => ["ibc.core.channel.v1.ResponseResultType", ResponseResultType, "RESPONSE_RESULT_TYPE_"] }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.result = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.ResponseResultType result */
        1:
          message.result = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.result !== 0)
      writer.tag(1, WireType.Varint).int32(message.result);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgAcknowledgementResponse = new MsgAcknowledgementResponse$Type();
class MsgChannelUpgradeInit$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeInit", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "fields", kind: "message", T: () => UpgradeFields, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* ibc.core.channel.v1.UpgradeFields fields */
        3:
          message.fields = UpgradeFields.internalBinaryRead(reader, reader.uint32(), options, message.fields);
          break;
        case /* string signer */
        4:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.fields)
      UpgradeFields.internalBinaryWrite(message.fields, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeInit = new MsgChannelUpgradeInit$Type();
class MsgChannelUpgradeInitResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeInitResponse", [
      { no: 1, name: "upgrade", kind: "message", T: () => Upgrade, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "upgrade_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.upgradeSequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Upgrade upgrade */
        1:
          message.upgrade = Upgrade.internalBinaryRead(reader, reader.uint32(), options, message.upgrade);
          break;
        case /* uint64 upgrade_sequence */
        2:
          message.upgradeSequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.upgrade)
      Upgrade.internalBinaryWrite(message.upgrade, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.upgradeSequence !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.upgradeSequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeInitResponse = new MsgChannelUpgradeInitResponse$Type();
class MsgChannelUpgradeTry$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeTry", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "proposed_upgrade_connection_hops",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 4, name: "counterparty_upgrade_fields", kind: "message", T: () => UpgradeFields, options: { "gogoproto.nullable": false } },
      {
        no: 5,
        name: "counterparty_upgrade_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "proof_channel",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 7,
        name: "proof_upgrade",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 8, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 9,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.proposedUpgradeConnectionHops = [];
    message.counterpartyUpgradeSequence = 0n;
    message.proofChannel = new Uint8Array(0);
    message.proofUpgrade = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* repeated string proposed_upgrade_connection_hops */
        3:
          message.proposedUpgradeConnectionHops.push(reader.string());
          break;
        case /* ibc.core.channel.v1.UpgradeFields counterparty_upgrade_fields */
        4:
          message.counterpartyUpgradeFields = UpgradeFields.internalBinaryRead(reader, reader.uint32(), options, message.counterpartyUpgradeFields);
          break;
        case /* uint64 counterparty_upgrade_sequence */
        5:
          message.counterpartyUpgradeSequence = reader.uint64().toBigInt();
          break;
        case /* bytes proof_channel */
        6:
          message.proofChannel = reader.bytes();
          break;
        case /* bytes proof_upgrade */
        7:
          message.proofUpgrade = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        8:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        9:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    for (let i = 0; i < message.proposedUpgradeConnectionHops.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.proposedUpgradeConnectionHops[i]);
    if (message.counterpartyUpgradeFields)
      UpgradeFields.internalBinaryWrite(message.counterpartyUpgradeFields, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.counterpartyUpgradeSequence !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.counterpartyUpgradeSequence);
    if (message.proofChannel.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.proofChannel);
    if (message.proofUpgrade.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.proofUpgrade);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(8, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeTry = new MsgChannelUpgradeTry$Type();
class MsgChannelUpgradeTryResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeTryResponse", [
      { no: 1, name: "upgrade", kind: "message", T: () => Upgrade, options: { "gogoproto.nullable": false } },
      {
        no: 2,
        name: "upgrade_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "result", kind: "enum", T: () => ["ibc.core.channel.v1.ResponseResultType", ResponseResultType, "RESPONSE_RESULT_TYPE_"] }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.upgradeSequence = 0n;
    message.result = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Upgrade upgrade */
        1:
          message.upgrade = Upgrade.internalBinaryRead(reader, reader.uint32(), options, message.upgrade);
          break;
        case /* uint64 upgrade_sequence */
        2:
          message.upgradeSequence = reader.uint64().toBigInt();
          break;
        case /* ibc.core.channel.v1.ResponseResultType result */
        3:
          message.result = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.upgrade)
      Upgrade.internalBinaryWrite(message.upgrade, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.upgradeSequence !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.upgradeSequence);
    if (message.result !== 0)
      writer.tag(3, WireType.Varint).int32(message.result);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeTryResponse = new MsgChannelUpgradeTryResponse$Type();
class MsgChannelUpgradeAck$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeAck", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "counterparty_upgrade", kind: "message", T: () => Upgrade, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "proof_channel",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "proof_upgrade",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 6, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 7,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.proofChannel = new Uint8Array(0);
    message.proofUpgrade = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* ibc.core.channel.v1.Upgrade counterparty_upgrade */
        3:
          message.counterpartyUpgrade = Upgrade.internalBinaryRead(reader, reader.uint32(), options, message.counterpartyUpgrade);
          break;
        case /* bytes proof_channel */
        4:
          message.proofChannel = reader.bytes();
          break;
        case /* bytes proof_upgrade */
        5:
          message.proofUpgrade = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        6:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        7:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.counterpartyUpgrade)
      Upgrade.internalBinaryWrite(message.counterpartyUpgrade, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.proofChannel.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.proofChannel);
    if (message.proofUpgrade.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.proofUpgrade);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeAck = new MsgChannelUpgradeAck$Type();
class MsgChannelUpgradeAckResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeAckResponse", [
      { no: 1, name: "result", kind: "enum", T: () => ["ibc.core.channel.v1.ResponseResultType", ResponseResultType, "RESPONSE_RESULT_TYPE_"] }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.result = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.ResponseResultType result */
        1:
          message.result = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.result !== 0)
      writer.tag(1, WireType.Varint).int32(message.result);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeAckResponse = new MsgChannelUpgradeAckResponse$Type();
class MsgChannelUpgradeConfirm$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeConfirm", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "counterparty_channel_state", kind: "enum", T: () => ["ibc.core.channel.v1.State", State, "STATE_"] },
      { no: 4, name: "counterparty_upgrade", kind: "message", T: () => Upgrade, options: { "gogoproto.nullable": false } },
      {
        no: 5,
        name: "proof_channel",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 6,
        name: "proof_upgrade",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 7, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 8,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.counterpartyChannelState = 0;
    message.proofChannel = new Uint8Array(0);
    message.proofUpgrade = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* ibc.core.channel.v1.State counterparty_channel_state */
        3:
          message.counterpartyChannelState = reader.int32();
          break;
        case /* ibc.core.channel.v1.Upgrade counterparty_upgrade */
        4:
          message.counterpartyUpgrade = Upgrade.internalBinaryRead(reader, reader.uint32(), options, message.counterpartyUpgrade);
          break;
        case /* bytes proof_channel */
        5:
          message.proofChannel = reader.bytes();
          break;
        case /* bytes proof_upgrade */
        6:
          message.proofUpgrade = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        7:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        8:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.counterpartyChannelState !== 0)
      writer.tag(3, WireType.Varint).int32(message.counterpartyChannelState);
    if (message.counterpartyUpgrade)
      Upgrade.internalBinaryWrite(message.counterpartyUpgrade, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.proofChannel.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.proofChannel);
    if (message.proofUpgrade.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.proofUpgrade);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeConfirm = new MsgChannelUpgradeConfirm$Type();
class MsgChannelUpgradeConfirmResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeConfirmResponse", [
      { no: 1, name: "result", kind: "enum", T: () => ["ibc.core.channel.v1.ResponseResultType", ResponseResultType, "RESPONSE_RESULT_TYPE_"] }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.result = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.ResponseResultType result */
        1:
          message.result = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.result !== 0)
      writer.tag(1, WireType.Varint).int32(message.result);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeConfirmResponse = new MsgChannelUpgradeConfirmResponse$Type();
class MsgChannelUpgradeOpen$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeOpen", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "counterparty_channel_state", kind: "enum", T: () => ["ibc.core.channel.v1.State", State, "STATE_"] },
      {
        no: 4,
        name: "counterparty_upgrade_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "proof_channel",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 6, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 7,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.counterpartyChannelState = 0;
    message.counterpartyUpgradeSequence = 0n;
    message.proofChannel = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* ibc.core.channel.v1.State counterparty_channel_state */
        3:
          message.counterpartyChannelState = reader.int32();
          break;
        case /* uint64 counterparty_upgrade_sequence */
        4:
          message.counterpartyUpgradeSequence = reader.uint64().toBigInt();
          break;
        case /* bytes proof_channel */
        5:
          message.proofChannel = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        6:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        7:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.counterpartyChannelState !== 0)
      writer.tag(3, WireType.Varint).int32(message.counterpartyChannelState);
    if (message.counterpartyUpgradeSequence !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.counterpartyUpgradeSequence);
    if (message.proofChannel.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.proofChannel);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeOpen = new MsgChannelUpgradeOpen$Type();
class MsgChannelUpgradeOpenResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeOpenResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeOpenResponse = new MsgChannelUpgradeOpenResponse$Type();
class MsgChannelUpgradeTimeout$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeTimeout", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "counterparty_channel", kind: "message", T: () => Channel, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "proof_channel",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 5, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 6,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.proofChannel = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* ibc.core.channel.v1.Channel counterparty_channel */
        3:
          message.counterpartyChannel = Channel.internalBinaryRead(reader, reader.uint32(), options, message.counterpartyChannel);
          break;
        case /* bytes proof_channel */
        4:
          message.proofChannel = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        5:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        6:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.counterpartyChannel)
      Channel.internalBinaryWrite(message.counterpartyChannel, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.proofChannel.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.proofChannel);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeTimeout = new MsgChannelUpgradeTimeout$Type();
class MsgChannelUpgradeTimeoutResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeTimeoutResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeTimeoutResponse = new MsgChannelUpgradeTimeoutResponse$Type();
class MsgChannelUpgradeCancel$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeCancel", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "error_receipt", kind: "message", T: () => ErrorReceipt, options: { "gogoproto.nullable": false } },
      {
        no: 4,
        name: "proof_error_receipt",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 5, name: "proof_height", kind: "message", T: () => Height, options: { "gogoproto.nullable": false } },
      {
        no: 6,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.proofErrorReceipt = new Uint8Array(0);
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* ibc.core.channel.v1.ErrorReceipt error_receipt */
        3:
          message.errorReceipt = ErrorReceipt.internalBinaryRead(reader, reader.uint32(), options, message.errorReceipt);
          break;
        case /* bytes proof_error_receipt */
        4:
          message.proofErrorReceipt = reader.bytes();
          break;
        case /* ibc.core.client.v1.Height proof_height */
        5:
          message.proofHeight = Height.internalBinaryRead(reader, reader.uint32(), options, message.proofHeight);
          break;
        case /* string signer */
        6:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.errorReceipt)
      ErrorReceipt.internalBinaryWrite(message.errorReceipt, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.proofErrorReceipt.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.proofErrorReceipt);
    if (message.proofHeight)
      Height.internalBinaryWrite(message.proofHeight, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    if (message.signer !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeCancel = new MsgChannelUpgradeCancel$Type();
class MsgChannelUpgradeCancelResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgChannelUpgradeCancelResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgChannelUpgradeCancelResponse = new MsgChannelUpgradeCancelResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgUpdateParams", [
      {
        no: 1,
        name: "authority",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* ibc.core.channel.v1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
class MsgPruneAcknowledgements$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgPruneAcknowledgements", [
      {
        no: 1,
        name: "port_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "channel_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "limit",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "signer",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "cosmos.msg.v1.signer": ["signer"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.portId = "";
    message.channelId = "";
    message.limit = 0n;
    message.signer = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string port_id */
        1:
          message.portId = reader.string();
          break;
        case /* string channel_id */
        2:
          message.channelId = reader.string();
          break;
        case /* uint64 limit */
        3:
          message.limit = reader.uint64().toBigInt();
          break;
        case /* string signer */
        4:
          message.signer = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.portId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.portId);
    if (message.channelId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.channelId);
    if (message.limit !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.limit);
    if (message.signer !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.signer);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPruneAcknowledgements = new MsgPruneAcknowledgements$Type();
class MsgPruneAcknowledgementsResponse$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.MsgPruneAcknowledgementsResponse", [
      {
        no: 1,
        name: "total_pruned_sequences",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "total_remaining_sequences",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.totalPrunedSequences = 0n;
    message.totalRemainingSequences = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 total_pruned_sequences */
        1:
          message.totalPrunedSequences = reader.uint64().toBigInt();
          break;
        case /* uint64 total_remaining_sequences */
        2:
          message.totalRemainingSequences = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.totalPrunedSequences !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.totalPrunedSequences);
    if (message.totalRemainingSequences !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.totalRemainingSequences);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgPruneAcknowledgementsResponse = new MsgPruneAcknowledgementsResponse$Type();
const Msg = new ServiceType("ibc.core.channel.v1.Msg", [
  { name: "ChannelOpenInit", options: {}, I: MsgChannelOpenInit, O: MsgChannelOpenInitResponse },
  { name: "ChannelOpenTry", options: {}, I: MsgChannelOpenTry, O: MsgChannelOpenTryResponse },
  { name: "ChannelOpenAck", options: {}, I: MsgChannelOpenAck, O: MsgChannelOpenAckResponse },
  { name: "ChannelOpenConfirm", options: {}, I: MsgChannelOpenConfirm, O: MsgChannelOpenConfirmResponse },
  { name: "ChannelCloseInit", options: {}, I: MsgChannelCloseInit, O: MsgChannelCloseInitResponse },
  { name: "ChannelCloseConfirm", options: {}, I: MsgChannelCloseConfirm, O: MsgChannelCloseConfirmResponse },
  { name: "RecvPacket", options: {}, I: MsgRecvPacket, O: MsgRecvPacketResponse },
  { name: "Timeout", options: {}, I: MsgTimeout, O: MsgTimeoutResponse },
  { name: "TimeoutOnClose", options: {}, I: MsgTimeoutOnClose, O: MsgTimeoutOnCloseResponse },
  { name: "Acknowledgement", options: {}, I: MsgAcknowledgement, O: MsgAcknowledgementResponse },
  { name: "ChannelUpgradeInit", options: {}, I: MsgChannelUpgradeInit, O: MsgChannelUpgradeInitResponse },
  { name: "ChannelUpgradeTry", options: {}, I: MsgChannelUpgradeTry, O: MsgChannelUpgradeTryResponse },
  { name: "ChannelUpgradeAck", options: {}, I: MsgChannelUpgradeAck, O: MsgChannelUpgradeAckResponse },
  { name: "ChannelUpgradeConfirm", options: {}, I: MsgChannelUpgradeConfirm, O: MsgChannelUpgradeConfirmResponse },
  { name: "ChannelUpgradeOpen", options: {}, I: MsgChannelUpgradeOpen, O: MsgChannelUpgradeOpenResponse },
  { name: "ChannelUpgradeTimeout", options: {}, I: MsgChannelUpgradeTimeout, O: MsgChannelUpgradeTimeoutResponse },
  { name: "ChannelUpgradeCancel", options: {}, I: MsgChannelUpgradeCancel, O: MsgChannelUpgradeCancelResponse },
  { name: "UpdateChannelParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse },
  { name: "PruneAcknowledgements", options: {}, I: MsgPruneAcknowledgements, O: MsgPruneAcknowledgementsResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgAcknowledgement,
  MsgAcknowledgementResponse,
  MsgChannelCloseConfirm,
  MsgChannelCloseConfirmResponse,
  MsgChannelCloseInit,
  MsgChannelCloseInitResponse,
  MsgChannelOpenAck,
  MsgChannelOpenAckResponse,
  MsgChannelOpenConfirm,
  MsgChannelOpenConfirmResponse,
  MsgChannelOpenInit,
  MsgChannelOpenInitResponse,
  MsgChannelOpenTry,
  MsgChannelOpenTryResponse,
  MsgChannelUpgradeAck,
  MsgChannelUpgradeAckResponse,
  MsgChannelUpgradeCancel,
  MsgChannelUpgradeCancelResponse,
  MsgChannelUpgradeConfirm,
  MsgChannelUpgradeConfirmResponse,
  MsgChannelUpgradeInit,
  MsgChannelUpgradeInitResponse,
  MsgChannelUpgradeOpen,
  MsgChannelUpgradeOpenResponse,
  MsgChannelUpgradeTimeout,
  MsgChannelUpgradeTimeoutResponse,
  MsgChannelUpgradeTry,
  MsgChannelUpgradeTryResponse,
  MsgPruneAcknowledgements,
  MsgPruneAcknowledgementsResponse,
  MsgRecvPacket,
  MsgRecvPacketResponse,
  MsgTimeout,
  MsgTimeoutOnClose,
  MsgTimeoutOnCloseResponse,
  MsgTimeoutResponse,
  MsgUpdateParams,
  MsgUpdateParamsResponse,
  ResponseResultType
};
