// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/channel/v1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * ChannelOpenInit defines a rpc handler method for MsgChannelOpenInit.
   *
   * @generated from protobuf rpc: ChannelOpenInit
   */
  channelOpenInit(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelOpenTry defines a rpc handler method for MsgChannelOpenTry.
   *
   * @generated from protobuf rpc: ChannelOpenTry
   */
  channelOpenTry(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelOpenAck defines a rpc handler method for MsgChannelOpenAck.
   *
   * @generated from protobuf rpc: ChannelOpenAck
   */
  channelOpenAck(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelOpenConfirm defines a rpc handler method for MsgChannelOpenConfirm.
   *
   * @generated from protobuf rpc: ChannelOpenConfirm
   */
  channelOpenConfirm(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelCloseInit defines a rpc handler method for MsgChannelCloseInit.
   *
   * @generated from protobuf rpc: ChannelCloseInit
   */
  channelCloseInit(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelCloseConfirm defines a rpc handler method for
   * MsgChannelCloseConfirm.
   *
   * @generated from protobuf rpc: ChannelCloseConfirm
   */
  channelCloseConfirm(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * RecvPacket defines a rpc handler method for MsgRecvPacket.
   *
   * @generated from protobuf rpc: RecvPacket
   */
  recvPacket(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Timeout defines a rpc handler method for MsgTimeout.
   *
   * @generated from protobuf rpc: Timeout
   */
  timeout(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TimeoutOnClose defines a rpc handler method for MsgTimeoutOnClose.
   *
   * @generated from protobuf rpc: TimeoutOnClose
   */
  timeoutOnClose(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Acknowledgement defines a rpc handler method for MsgAcknowledgement.
   *
   * @generated from protobuf rpc: Acknowledgement
   */
  acknowledgement(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelUpgradeInit defines a rpc handler method for MsgChannelUpgradeInit.
   *
   * @generated from protobuf rpc: ChannelUpgradeInit
   */
  channelUpgradeInit(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelUpgradeTry defines a rpc handler method for MsgChannelUpgradeTry.
   *
   * @generated from protobuf rpc: ChannelUpgradeTry
   */
  channelUpgradeTry(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelUpgradeAck defines a rpc handler method for MsgChannelUpgradeAck.
   *
   * @generated from protobuf rpc: ChannelUpgradeAck
   */
  channelUpgradeAck(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelUpgradeConfirm defines a rpc handler method for MsgChannelUpgradeConfirm.
   *
   * @generated from protobuf rpc: ChannelUpgradeConfirm
   */
  channelUpgradeConfirm(input, options) {
    const method = this.methods[13], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelUpgradeOpen defines a rpc handler method for MsgChannelUpgradeOpen.
   *
   * @generated from protobuf rpc: ChannelUpgradeOpen
   */
  channelUpgradeOpen(input, options) {
    const method = this.methods[14], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelUpgradeTimeout defines a rpc handler method for MsgChannelUpgradeTimeout.
   *
   * @generated from protobuf rpc: ChannelUpgradeTimeout
   */
  channelUpgradeTimeout(input, options) {
    const method = this.methods[15], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ChannelUpgradeCancel defines a rpc handler method for MsgChannelUpgradeCancel.
   *
   * @generated from protobuf rpc: ChannelUpgradeCancel
   */
  channelUpgradeCancel(input, options) {
    const method = this.methods[16], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateChannelParams defines a rpc handler method for MsgUpdateParams.
   *
   * @generated from protobuf rpc: UpdateChannelParams
   */
  updateChannelParams(input, options) {
    const method = this.methods[17], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * PruneAcknowledgements defines a rpc handler method for MsgPruneAcknowledgements.
   *
   * @generated from protobuf rpc: PruneAcknowledgements
   */
  pruneAcknowledgements(input, options) {
    const method = this.methods[18], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
