// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/channel/v1/upgrade_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Order } from "./channel_pb.js";
import { Timeout } from "./channel_pb.js";
class Upgrade$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.Upgrade", [
      { no: 1, name: "fields", kind: "message", T: () => UpgradeFields, options: { "gogoproto.nullable": false } },
      { no: 2, name: "timeout", kind: "message", T: () => Timeout, options: { "gogoproto.nullable": false } },
      {
        no: 3,
        name: "next_sequence_send",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.nextSequenceSend = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.UpgradeFields fields */
        1:
          message.fields = UpgradeFields.internalBinaryRead(reader, reader.uint32(), options, message.fields);
          break;
        case /* ibc.core.channel.v1.Timeout timeout */
        2:
          message.timeout = Timeout.internalBinaryRead(reader, reader.uint32(), options, message.timeout);
          break;
        case /* uint64 next_sequence_send */
        3:
          message.nextSequenceSend = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.fields)
      UpgradeFields.internalBinaryWrite(message.fields, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.timeout)
      Timeout.internalBinaryWrite(message.timeout, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.nextSequenceSend !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.nextSequenceSend);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Upgrade = new Upgrade$Type();
class UpgradeFields$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.UpgradeFields", [
      { no: 1, name: "ordering", kind: "enum", T: () => ["ibc.core.channel.v1.Order", Order, "ORDER_"] },
      {
        no: 2,
        name: "connection_hops",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "version",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.ordering = 0;
    message.connectionHops = [];
    message.version = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* ibc.core.channel.v1.Order ordering */
        1:
          message.ordering = reader.int32();
          break;
        case /* repeated string connection_hops */
        2:
          message.connectionHops.push(reader.string());
          break;
        case /* string version */
        3:
          message.version = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.ordering !== 0)
      writer.tag(1, WireType.Varint).int32(message.ordering);
    for (let i = 0; i < message.connectionHops.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.connectionHops[i]);
    if (message.version !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.version);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const UpgradeFields = new UpgradeFields$Type();
class ErrorReceipt$Type extends MessageType {
  constructor() {
    super("ibc.core.channel.v1.ErrorReceipt", [
      {
        no: 1,
        name: "sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "message",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sequence = 0n;
    message.message = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 sequence */
        1:
          message.sequence = reader.uint64().toBigInt();
          break;
        case /* string message */
        2:
          message.message = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sequence !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.sequence);
    if (message.message !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.message);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ErrorReceipt = new ErrorReceipt$Type();
export {
  ErrorReceipt,
  Upgrade,
  UpgradeFields
};
