// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/store/streaming/abci/grpc_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { StoreKVPair } from "../../v1beta1/listening_pb.js";
import { CommitResponse } from "../../../../cometbft/abci/v1/types_pb.js";
import { FinalizeBlockResponse } from "../../../../cometbft/abci/v1/types_pb.js";
import { FinalizeBlockRequest } from "../../../../cometbft/abci/v1/types_pb.js";
class ListenFinalizeBlockRequest$Type extends MessageType {
  constructor() {
    super("cosmos.store.streaming.abci.ListenFinalizeBlockRequest", [
      { no: 1, name: "req", kind: "message", T: () => FinalizeBlockRequest },
      { no: 2, name: "res", kind: "message", T: () => FinalizeBlockResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cometbft.abci.v1.FinalizeBlockRequest req */
        1:
          message.req = FinalizeBlockRequest.internalBinaryRead(reader, reader.uint32(), options, message.req);
          break;
        case /* cometbft.abci.v1.FinalizeBlockResponse res */
        2:
          message.res = FinalizeBlockResponse.internalBinaryRead(reader, reader.uint32(), options, message.res);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.req)
      FinalizeBlockRequest.internalBinaryWrite(message.req, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.res)
      FinalizeBlockResponse.internalBinaryWrite(message.res, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListenFinalizeBlockRequest = new ListenFinalizeBlockRequest$Type();
class ListenFinalizeBlockResponse$Type extends MessageType {
  constructor() {
    super("cosmos.store.streaming.abci.ListenFinalizeBlockResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListenFinalizeBlockResponse = new ListenFinalizeBlockResponse$Type();
class ListenCommitRequest$Type extends MessageType {
  constructor() {
    super("cosmos.store.streaming.abci.ListenCommitRequest", [
      {
        no: 1,
        name: "block_height",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "res", kind: "message", T: () => CommitResponse },
      { no: 3, name: "change_set", kind: "message", repeat: 2, T: () => StoreKVPair }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.blockHeight = 0n;
    message.changeSet = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 block_height */
        1:
          message.blockHeight = reader.int64().toBigInt();
          break;
        case /* cometbft.abci.v1.CommitResponse res */
        2:
          message.res = CommitResponse.internalBinaryRead(reader, reader.uint32(), options, message.res);
          break;
        case /* repeated cosmos.store.v1beta1.StoreKVPair change_set */
        3:
          message.changeSet.push(StoreKVPair.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.blockHeight !== 0n)
      writer.tag(1, WireType.Varint).int64(message.blockHeight);
    if (message.res)
      CommitResponse.internalBinaryWrite(message.res, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.changeSet.length; i++)
      StoreKVPair.internalBinaryWrite(message.changeSet[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListenCommitRequest = new ListenCommitRequest$Type();
class ListenCommitResponse$Type extends MessageType {
  constructor() {
    super("cosmos.store.streaming.abci.ListenCommitResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ListenCommitResponse = new ListenCommitResponse$Type();
const ABCIListenerService = new ServiceType("cosmos.store.streaming.abci.ABCIListenerService", [
  { name: "ListenFinalizeBlock", options: {}, I: ListenFinalizeBlockRequest, O: ListenFinalizeBlockResponse },
  { name: "ListenCommit", options: {}, I: ListenCommitRequest, O: ListenCommitResponse }
]);
export {
  ABCIListenerService,
  ListenCommitRequest,
  ListenCommitResponse,
  ListenFinalizeBlockRequest,
  ListenFinalizeBlockResponse
};
