// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/store/streaming/abci/grpc_pb.client.js
import { ABCIListenerService } from "./grpc_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class ABCIListenerServiceClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = ABCIListenerService.typeName;
    this.methods = ABCIListenerService.methods;
    this.options = ABCIListenerService.options;
  }
  /**
   * ListenFinalizeBlock is the corresponding endpoint for ABCIListener.ListenEndBlock
   *
   * @generated from protobuf rpc: ListenFinalizeBlock
   */
  listenFinalizeBlock(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ListenCommit is the corresponding endpoint for ABCIListener.ListenCommit
   *
   * @generated from protobuf rpc: ListenCommit
   */
  listenCommit(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  ABCIListenerServiceClient
};
