// @injectivelabs/core-proto-ts-v2/esm/generated/injective/ocr/v1beta1/ocr_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Coin } from "../../../cosmos/base/v1beta1/coin_pb.js";
class Params$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.Params", [
      {
        no: 1,
        name: "link_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "payout_block_interval",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "module_admin",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.equal": true, "amino.name": "ocr/Params" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.linkDenom = "";
    message.payoutBlockInterval = 0n;
    message.moduleAdmin = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string link_denom */
        1:
          message.linkDenom = reader.string();
          break;
        case /* uint64 payout_block_interval */
        2:
          message.payoutBlockInterval = reader.uint64().toBigInt();
          break;
        case /* string module_admin */
        3:
          message.moduleAdmin = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.linkDenom !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.linkDenom);
    if (message.payoutBlockInterval !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.payoutBlockInterval);
    if (message.moduleAdmin !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.moduleAdmin);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
class FeedConfig$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.FeedConfig", [
      {
        no: 1,
        name: "signers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "transmitters",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "f",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 4,
        name: "onchain_config",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "offchain_config_version",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "offchain_config",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 7, name: "module_params", kind: "message", T: () => ModuleParams }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.signers = [];
    message.transmitters = [];
    message.f = 0;
    message.onchainConfig = new Uint8Array(0);
    message.offchainConfigVersion = 0n;
    message.offchainConfig = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string signers */
        1:
          message.signers.push(reader.string());
          break;
        case /* repeated string transmitters */
        2:
          message.transmitters.push(reader.string());
          break;
        case /* uint32 f */
        3:
          message.f = reader.uint32();
          break;
        case /* bytes onchain_config */
        4:
          message.onchainConfig = reader.bytes();
          break;
        case /* uint64 offchain_config_version */
        5:
          message.offchainConfigVersion = reader.uint64().toBigInt();
          break;
        case /* bytes offchain_config */
        6:
          message.offchainConfig = reader.bytes();
          break;
        case /* injective.ocr.v1beta1.ModuleParams module_params */
        7:
          message.moduleParams = ModuleParams.internalBinaryRead(reader, reader.uint32(), options, message.moduleParams);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.signers.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.signers[i]);
    for (let i = 0; i < message.transmitters.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.transmitters[i]);
    if (message.f !== 0)
      writer.tag(3, WireType.Varint).uint32(message.f);
    if (message.onchainConfig.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.onchainConfig);
    if (message.offchainConfigVersion !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.offchainConfigVersion);
    if (message.offchainConfig.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.offchainConfig);
    if (message.moduleParams)
      ModuleParams.internalBinaryWrite(message.moduleParams, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FeedConfig = new FeedConfig$Type();
class FeedConfigInfo$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.FeedConfigInfo", [
      {
        no: 1,
        name: "latest_config_digest",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "f",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "n",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 4,
        name: "config_count",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "latest_config_block_number",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.latestConfigDigest = new Uint8Array(0);
    message.f = 0;
    message.n = 0;
    message.configCount = 0n;
    message.latestConfigBlockNumber = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes latest_config_digest */
        1:
          message.latestConfigDigest = reader.bytes();
          break;
        case /* uint32 f */
        2:
          message.f = reader.uint32();
          break;
        case /* uint32 n */
        3:
          message.n = reader.uint32();
          break;
        case /* uint64 config_count */
        4:
          message.configCount = reader.uint64().toBigInt();
          break;
        case /* int64 latest_config_block_number */
        5:
          message.latestConfigBlockNumber = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.latestConfigDigest.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.latestConfigDigest);
    if (message.f !== 0)
      writer.tag(2, WireType.Varint).uint32(message.f);
    if (message.n !== 0)
      writer.tag(3, WireType.Varint).uint32(message.n);
    if (message.configCount !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.configCount);
    if (message.latestConfigBlockNumber !== 0n)
      writer.tag(5, WireType.Varint).int64(message.latestConfigBlockNumber);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FeedConfigInfo = new FeedConfigInfo$Type();
class ModuleParams$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.ModuleParams", [
      {
        no: 1,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "min_answer", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } },
      { no: 3, name: "max_answer", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } },
      { no: 4, name: "link_per_observation", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      { no: 5, name: "link_per_transmission", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      {
        no: 6,
        name: "link_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 7,
        name: "unique_reports",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 8,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "feed_admin",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 10,
        name: "billing_admin",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.feedId = "";
    message.minAnswer = "";
    message.maxAnswer = "";
    message.linkPerObservation = "";
    message.linkPerTransmission = "";
    message.linkDenom = "";
    message.uniqueReports = false;
    message.description = "";
    message.feedAdmin = "";
    message.billingAdmin = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string feed_id */
        1:
          message.feedId = reader.string();
          break;
        case /* string min_answer */
        2:
          message.minAnswer = reader.string();
          break;
        case /* string max_answer */
        3:
          message.maxAnswer = reader.string();
          break;
        case /* string link_per_observation */
        4:
          message.linkPerObservation = reader.string();
          break;
        case /* string link_per_transmission */
        5:
          message.linkPerTransmission = reader.string();
          break;
        case /* string link_denom */
        6:
          message.linkDenom = reader.string();
          break;
        case /* bool unique_reports */
        7:
          message.uniqueReports = reader.bool();
          break;
        case /* string description */
        8:
          message.description = reader.string();
          break;
        case /* string feed_admin */
        9:
          message.feedAdmin = reader.string();
          break;
        case /* string billing_admin */
        10:
          message.billingAdmin = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feedId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.feedId);
    if (message.minAnswer !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.minAnswer);
    if (message.maxAnswer !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.maxAnswer);
    if (message.linkPerObservation !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.linkPerObservation);
    if (message.linkPerTransmission !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.linkPerTransmission);
    if (message.linkDenom !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.linkDenom);
    if (message.uniqueReports !== false)
      writer.tag(7, WireType.Varint).bool(message.uniqueReports);
    if (message.description !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.description);
    if (message.feedAdmin !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.feedAdmin);
    if (message.billingAdmin !== "")
      writer.tag(10, WireType.LengthDelimited).string(message.billingAdmin);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ModuleParams = new ModuleParams$Type();
class ContractConfig$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.ContractConfig", [
      {
        no: 1,
        name: "config_count",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "signers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "transmitters",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "f",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 5,
        name: "onchain_config",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 6,
        name: "offchain_config_version",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "offchain_config",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.configCount = 0n;
    message.signers = [];
    message.transmitters = [];
    message.f = 0;
    message.onchainConfig = new Uint8Array(0);
    message.offchainConfigVersion = 0n;
    message.offchainConfig = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 config_count */
        1:
          message.configCount = reader.uint64().toBigInt();
          break;
        case /* repeated string signers */
        2:
          message.signers.push(reader.string());
          break;
        case /* repeated string transmitters */
        3:
          message.transmitters.push(reader.string());
          break;
        case /* uint32 f */
        4:
          message.f = reader.uint32();
          break;
        case /* bytes onchain_config */
        5:
          message.onchainConfig = reader.bytes();
          break;
        case /* uint64 offchain_config_version */
        6:
          message.offchainConfigVersion = reader.uint64().toBigInt();
          break;
        case /* bytes offchain_config */
        7:
          message.offchainConfig = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.configCount !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.configCount);
    for (let i = 0; i < message.signers.length; i++)
      writer.tag(2, WireType.LengthDelimited).string(message.signers[i]);
    for (let i = 0; i < message.transmitters.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.transmitters[i]);
    if (message.f !== 0)
      writer.tag(4, WireType.Varint).uint32(message.f);
    if (message.onchainConfig.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.onchainConfig);
    if (message.offchainConfigVersion !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.offchainConfigVersion);
    if (message.offchainConfig.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.offchainConfig);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ContractConfig = new ContractConfig$Type();
class SetConfigProposal$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.SetConfigProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "config", kind: "message", T: () => FeedConfig }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "ocr/SetConfigProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* injective.ocr.v1beta1.FeedConfig config */
        3:
          message.config = FeedConfig.internalBinaryRead(reader, reader.uint32(), options, message.config);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.config)
      FeedConfig.internalBinaryWrite(message.config, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SetConfigProposal = new SetConfigProposal$Type();
class FeedProperties$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.FeedProperties", [
      {
        no: 1,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "f",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "onchain_config",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 4,
        name: "offchain_config_version",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "offchain_config",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 6, name: "min_answer", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } },
      { no: 7, name: "max_answer", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } },
      { no: 8, name: "link_per_observation", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      { no: 9, name: "link_per_transmission", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      {
        no: 10,
        name: "unique_reports",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 11,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.feedId = "";
    message.f = 0;
    message.onchainConfig = new Uint8Array(0);
    message.offchainConfigVersion = 0n;
    message.offchainConfig = new Uint8Array(0);
    message.minAnswer = "";
    message.maxAnswer = "";
    message.linkPerObservation = "";
    message.linkPerTransmission = "";
    message.uniqueReports = false;
    message.description = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string feed_id */
        1:
          message.feedId = reader.string();
          break;
        case /* uint32 f */
        2:
          message.f = reader.uint32();
          break;
        case /* bytes onchain_config */
        3:
          message.onchainConfig = reader.bytes();
          break;
        case /* uint64 offchain_config_version */
        4:
          message.offchainConfigVersion = reader.uint64().toBigInt();
          break;
        case /* bytes offchain_config */
        5:
          message.offchainConfig = reader.bytes();
          break;
        case /* string min_answer */
        6:
          message.minAnswer = reader.string();
          break;
        case /* string max_answer */
        7:
          message.maxAnswer = reader.string();
          break;
        case /* string link_per_observation */
        8:
          message.linkPerObservation = reader.string();
          break;
        case /* string link_per_transmission */
        9:
          message.linkPerTransmission = reader.string();
          break;
        case /* bool unique_reports */
        10:
          message.uniqueReports = reader.bool();
          break;
        case /* string description */
        11:
          message.description = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feedId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.feedId);
    if (message.f !== 0)
      writer.tag(2, WireType.Varint).uint32(message.f);
    if (message.onchainConfig.length)
      writer.tag(3, WireType.LengthDelimited).bytes(message.onchainConfig);
    if (message.offchainConfigVersion !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.offchainConfigVersion);
    if (message.offchainConfig.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.offchainConfig);
    if (message.minAnswer !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.minAnswer);
    if (message.maxAnswer !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.maxAnswer);
    if (message.linkPerObservation !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.linkPerObservation);
    if (message.linkPerTransmission !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.linkPerTransmission);
    if (message.uniqueReports !== false)
      writer.tag(10, WireType.Varint).bool(message.uniqueReports);
    if (message.description !== "")
      writer.tag(11, WireType.LengthDelimited).string(message.description);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const FeedProperties = new FeedProperties$Type();
class SetBatchConfigProposal$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.SetBatchConfigProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "signers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "transmitters",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "link_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 6, name: "feed_properties", kind: "message", repeat: 2, T: () => FeedProperties }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "ocr/SetBatchConfigProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.signers = [];
    message.transmitters = [];
    message.linkDenom = "";
    message.feedProperties = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* repeated string signers */
        3:
          message.signers.push(reader.string());
          break;
        case /* repeated string transmitters */
        4:
          message.transmitters.push(reader.string());
          break;
        case /* string link_denom */
        5:
          message.linkDenom = reader.string();
          break;
        case /* repeated injective.ocr.v1beta1.FeedProperties feed_properties */
        6:
          message.feedProperties.push(FeedProperties.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    for (let i = 0; i < message.signers.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.signers[i]);
    for (let i = 0; i < message.transmitters.length; i++)
      writer.tag(4, WireType.LengthDelimited).string(message.transmitters[i]);
    if (message.linkDenom !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.linkDenom);
    for (let i = 0; i < message.feedProperties.length; i++)
      FeedProperties.internalBinaryWrite(message.feedProperties[i], writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SetBatchConfigProposal = new SetBatchConfigProposal$Type();
class OracleObservationsCounts$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.OracleObservationsCounts", [
      {
        no: 1,
        name: "counts",
        kind: "scalar",
        repeat: 1,
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.counts = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated uint32 counts */
        1:
          if (wireType === WireType.LengthDelimited)
            for (let e = reader.int32() + reader.pos; reader.pos < e; )
              message.counts.push(reader.uint32());
          else
            message.counts.push(reader.uint32());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.counts.length) {
      writer.tag(1, WireType.LengthDelimited).fork();
      for (let i = 0; i < message.counts.length; i++)
        writer.uint32(message.counts[i]);
      writer.join();
    }
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const OracleObservationsCounts = new OracleObservationsCounts$Type();
class GasReimbursements$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.GasReimbursements", [
      { no: 1, name: "reimbursements", kind: "message", repeat: 2, T: () => Coin }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.reimbursements = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated cosmos.base.v1beta1.Coin reimbursements */
        1:
          message.reimbursements.push(Coin.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.reimbursements.length; i++)
      Coin.internalBinaryWrite(message.reimbursements[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GasReimbursements = new GasReimbursements$Type();
class Payee$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.Payee", [
      {
        no: 1,
        name: "transmitter_addr",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "payment_addr",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.transmitterAddr = "";
    message.paymentAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string transmitter_addr */
        1:
          message.transmitterAddr = reader.string();
          break;
        case /* string payment_addr */
        2:
          message.paymentAddr = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.transmitterAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.transmitterAddr);
    if (message.paymentAddr !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.paymentAddr);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Payee = new Payee$Type();
class Transmission$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.Transmission", [
      { no: 1, name: "answer", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } },
      {
        no: 2,
        name: "observations_timestamp",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "transmission_timestamp",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.answer = "";
    message.observationsTimestamp = 0n;
    message.transmissionTimestamp = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string answer */
        1:
          message.answer = reader.string();
          break;
        case /* int64 observations_timestamp */
        2:
          message.observationsTimestamp = reader.int64().toBigInt();
          break;
        case /* int64 transmission_timestamp */
        3:
          message.transmissionTimestamp = reader.int64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.answer !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.answer);
    if (message.observationsTimestamp !== 0n)
      writer.tag(2, WireType.Varint).int64(message.observationsTimestamp);
    if (message.transmissionTimestamp !== 0n)
      writer.tag(3, WireType.Varint).int64(message.transmissionTimestamp);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Transmission = new Transmission$Type();
class EpochAndRound$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.EpochAndRound", [
      {
        no: 1,
        name: "epoch",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "round",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.epoch = 0n;
    message.round = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 epoch */
        1:
          message.epoch = reader.uint64().toBigInt();
          break;
        case /* uint64 round */
        2:
          message.round = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.epoch !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.epoch);
    if (message.round !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.round);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EpochAndRound = new EpochAndRound$Type();
class Report$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.Report", [
      {
        no: 1,
        name: "observations_timestamp",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "observers",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 3, name: "observations", kind: "scalar", repeat: 2, T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.observationsTimestamp = 0n;
    message.observers = new Uint8Array(0);
    message.observations = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* int64 observations_timestamp */
        1:
          message.observationsTimestamp = reader.int64().toBigInt();
          break;
        case /* bytes observers */
        2:
          message.observers = reader.bytes();
          break;
        case /* repeated string observations */
        3:
          message.observations.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.observationsTimestamp !== 0n)
      writer.tag(1, WireType.Varint).int64(message.observationsTimestamp);
    if (message.observers.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.observers);
    for (let i = 0; i < message.observations.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.observations[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Report = new Report$Type();
class ReportToSign$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.ReportToSign", [
      {
        no: 1,
        name: "config_digest",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "epoch",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "round",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "extra_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "report",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.configDigest = new Uint8Array(0);
    message.epoch = 0n;
    message.round = 0n;
    message.extraHash = new Uint8Array(0);
    message.report = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes config_digest */
        1:
          message.configDigest = reader.bytes();
          break;
        case /* uint64 epoch */
        2:
          message.epoch = reader.uint64().toBigInt();
          break;
        case /* uint64 round */
        3:
          message.round = reader.uint64().toBigInt();
          break;
        case /* bytes extra_hash */
        4:
          message.extraHash = reader.bytes();
          break;
        case /* bytes report */
        5:
          message.report = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.configDigest.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.configDigest);
    if (message.epoch !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.epoch);
    if (message.round !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.round);
    if (message.extraHash.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.extraHash);
    if (message.report.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.report);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ReportToSign = new ReportToSign$Type();
class EventOraclePaid$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.EventOraclePaid", [
      {
        no: 1,
        name: "transmitter_addr",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "payee_addr",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.transmitterAddr = "";
    message.payeeAddr = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string transmitter_addr */
        1:
          message.transmitterAddr = reader.string();
          break;
        case /* string payee_addr */
        2:
          message.payeeAddr = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        3:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.transmitterAddr !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.transmitterAddr);
    if (message.payeeAddr !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.payeeAddr);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventOraclePaid = new EventOraclePaid$Type();
class EventAnswerUpdated$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.EventAnswerUpdated", [
      { no: 1, name: "current", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      { no: 2, name: "round_id", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      { no: 3, name: "updated_at", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.current = "";
    message.roundId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string current */
        1:
          message.current = reader.string();
          break;
        case /* string round_id */
        2:
          message.roundId = reader.string();
          break;
        case /* google.protobuf.Timestamp updated_at */
        3:
          message.updatedAt = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.updatedAt);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.current !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.current);
    if (message.roundId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.roundId);
    if (message.updatedAt)
      Timestamp.internalBinaryWrite(message.updatedAt, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventAnswerUpdated = new EventAnswerUpdated$Type();
class EventNewRound$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.EventNewRound", [
      { no: 1, name: "round_id", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      {
        no: 2,
        name: "started_by",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "started_at", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": false, "gogoproto.stdtime": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.roundId = "";
    message.startedBy = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string round_id */
        1:
          message.roundId = reader.string();
          break;
        case /* string started_by */
        2:
          message.startedBy = reader.string();
          break;
        case /* google.protobuf.Timestamp started_at */
        3:
          message.startedAt = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.startedAt);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.roundId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.roundId);
    if (message.startedBy !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.startedBy);
    if (message.startedAt)
      Timestamp.internalBinaryWrite(message.startedAt, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventNewRound = new EventNewRound$Type();
class EventTransmitted$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.EventTransmitted", [
      {
        no: 1,
        name: "config_digest",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "epoch",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.configDigest = new Uint8Array(0);
    message.epoch = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes config_digest */
        1:
          message.configDigest = reader.bytes();
          break;
        case /* uint64 epoch */
        2:
          message.epoch = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.configDigest.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.configDigest);
    if (message.epoch !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.epoch);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventTransmitted = new EventTransmitted$Type();
class EventNewTransmission$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.EventNewTransmission", [
      {
        no: 1,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "aggregator_round_id",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      { no: 3, name: "answer", kind: "scalar", T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } },
      {
        no: 4,
        name: "transmitter",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 5,
        name: "observations_timestamp",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 6, name: "observations", kind: "scalar", repeat: 2, T: 9, options: { "gogoproto.nullable": false, "gogoproto.customtype": "cosmossdk.io/math.LegacyDec" } },
      {
        no: 7,
        name: "observers",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 8,
        name: "config_digest",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 9, name: "epoch_and_round", kind: "message", T: () => EpochAndRound }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.feedId = "";
    message.aggregatorRoundId = 0;
    message.answer = "";
    message.transmitter = "";
    message.observationsTimestamp = 0n;
    message.observations = [];
    message.observers = new Uint8Array(0);
    message.configDigest = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string feed_id */
        1:
          message.feedId = reader.string();
          break;
        case /* uint32 aggregator_round_id */
        2:
          message.aggregatorRoundId = reader.uint32();
          break;
        case /* string answer */
        3:
          message.answer = reader.string();
          break;
        case /* string transmitter */
        4:
          message.transmitter = reader.string();
          break;
        case /* int64 observations_timestamp */
        5:
          message.observationsTimestamp = reader.int64().toBigInt();
          break;
        case /* repeated string observations */
        6:
          message.observations.push(reader.string());
          break;
        case /* bytes observers */
        7:
          message.observers = reader.bytes();
          break;
        case /* bytes config_digest */
        8:
          message.configDigest = reader.bytes();
          break;
        case /* injective.ocr.v1beta1.EpochAndRound epoch_and_round */
        9:
          message.epochAndRound = EpochAndRound.internalBinaryRead(reader, reader.uint32(), options, message.epochAndRound);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feedId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.feedId);
    if (message.aggregatorRoundId !== 0)
      writer.tag(2, WireType.Varint).uint32(message.aggregatorRoundId);
    if (message.answer !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.answer);
    if (message.transmitter !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.transmitter);
    if (message.observationsTimestamp !== 0n)
      writer.tag(5, WireType.Varint).int64(message.observationsTimestamp);
    for (let i = 0; i < message.observations.length; i++)
      writer.tag(6, WireType.LengthDelimited).string(message.observations[i]);
    if (message.observers.length)
      writer.tag(7, WireType.LengthDelimited).bytes(message.observers);
    if (message.configDigest.length)
      writer.tag(8, WireType.LengthDelimited).bytes(message.configDigest);
    if (message.epochAndRound)
      EpochAndRound.internalBinaryWrite(message.epochAndRound, writer.tag(9, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventNewTransmission = new EventNewTransmission$Type();
class EventConfigSet$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.EventConfigSet", [
      {
        no: 1,
        name: "config_digest",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "previous_config_block_number",
        kind: "scalar",
        T: 3,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 3, name: "config", kind: "message", T: () => FeedConfig },
      { no: 4, name: "config_info", kind: "message", T: () => FeedConfigInfo }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.configDigest = new Uint8Array(0);
    message.previousConfigBlockNumber = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes config_digest */
        1:
          message.configDigest = reader.bytes();
          break;
        case /* int64 previous_config_block_number */
        2:
          message.previousConfigBlockNumber = reader.int64().toBigInt();
          break;
        case /* injective.ocr.v1beta1.FeedConfig config */
        3:
          message.config = FeedConfig.internalBinaryRead(reader, reader.uint32(), options, message.config);
          break;
        case /* injective.ocr.v1beta1.FeedConfigInfo config_info */
        4:
          message.configInfo = FeedConfigInfo.internalBinaryRead(reader, reader.uint32(), options, message.configInfo);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.configDigest.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.configDigest);
    if (message.previousConfigBlockNumber !== 0n)
      writer.tag(2, WireType.Varint).int64(message.previousConfigBlockNumber);
    if (message.config)
      FeedConfig.internalBinaryWrite(message.config, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.configInfo)
      FeedConfigInfo.internalBinaryWrite(message.configInfo, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventConfigSet = new EventConfigSet$Type();
export {
  ContractConfig,
  EpochAndRound,
  EventAnswerUpdated,
  EventConfigSet,
  EventNewRound,
  EventNewTransmission,
  EventOraclePaid,
  EventTransmitted,
  FeedConfig,
  FeedConfigInfo,
  FeedProperties,
  GasReimbursements,
  ModuleParams,
  OracleObservationsCounts,
  Params,
  Payee,
  Report,
  ReportToSign,
  SetBatchConfigProposal,
  SetConfigProposal,
  Transmission
};
