// @injectivelabs/core-proto-ts-v2/esm/generated/injective/ocr/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the OCR FeedConfig for a given FeedId
   *
   * @generated from protobuf rpc: FeedConfig
   */
  feedConfig(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the OCR FeedConfigInfo for a given FeedId
   *
   * @generated from protobuf rpc: FeedConfigInfo
   */
  feedConfigInfo(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves latest round ID and data, including median answer for that round
   *
   * @generated from protobuf rpc: LatestRound
   */
  latestRound(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * LatestTransmissionDetails returns details about the latest trasmission
   * recorded on chain for the given feed ID.
   *
   * @generated from protobuf rpc: LatestTransmissionDetails
   */
  latestTransmissionDetails(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves transmitter's owed amount
   *
   * @generated from protobuf rpc: OwedAmount
   */
  owedAmount(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the entire OCR module's state
   *
   * @generated from protobuf rpc: OcrModuleState
   */
  ocrModuleState(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
