// @injectivelabs/core-proto-ts-v2/esm/generated/injective/ocr/v1beta1/query_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { GenesisState } from "./genesis_pb.js";
import { Coin } from "../../../cosmos/base/v1beta1/coin_pb.js";
import { Transmission } from "./ocr_pb.js";
import { EpochAndRound } from "./ocr_pb.js";
import { FeedConfig } from "./ocr_pb.js";
import { FeedConfigInfo } from "./ocr_pb.js";
import { Params } from "./ocr_pb.js";
class QueryParamsRequest$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryParamsRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryParamsRequest = new QueryParamsRequest$Type();
class QueryParamsResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryParamsResponse", [
      { no: 1, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.ocr.v1beta1.Params params */
        1:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryParamsResponse = new QueryParamsResponse$Type();
class QueryFeedConfigRequest$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryFeedConfigRequest", [
      {
        no: 1,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.feedId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string feed_id */
        1:
          message.feedId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feedId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.feedId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryFeedConfigRequest = new QueryFeedConfigRequest$Type();
class QueryFeedConfigResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryFeedConfigResponse", [
      { no: 1, name: "feed_config_info", kind: "message", T: () => FeedConfigInfo },
      { no: 2, name: "feed_config", kind: "message", T: () => FeedConfig }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.ocr.v1beta1.FeedConfigInfo feed_config_info */
        1:
          message.feedConfigInfo = FeedConfigInfo.internalBinaryRead(reader, reader.uint32(), options, message.feedConfigInfo);
          break;
        case /* injective.ocr.v1beta1.FeedConfig feed_config */
        2:
          message.feedConfig = FeedConfig.internalBinaryRead(reader, reader.uint32(), options, message.feedConfig);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feedConfigInfo)
      FeedConfigInfo.internalBinaryWrite(message.feedConfigInfo, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.feedConfig)
      FeedConfig.internalBinaryWrite(message.feedConfig, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryFeedConfigResponse = new QueryFeedConfigResponse$Type();
class QueryFeedConfigInfoRequest$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryFeedConfigInfoRequest", [
      {
        no: 1,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.feedId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string feed_id */
        1:
          message.feedId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feedId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.feedId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryFeedConfigInfoRequest = new QueryFeedConfigInfoRequest$Type();
class QueryFeedConfigInfoResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryFeedConfigInfoResponse", [
      { no: 1, name: "feed_config_info", kind: "message", T: () => FeedConfigInfo },
      { no: 2, name: "epoch_and_round", kind: "message", T: () => EpochAndRound }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.ocr.v1beta1.FeedConfigInfo feed_config_info */
        1:
          message.feedConfigInfo = FeedConfigInfo.internalBinaryRead(reader, reader.uint32(), options, message.feedConfigInfo);
          break;
        case /* injective.ocr.v1beta1.EpochAndRound epoch_and_round */
        2:
          message.epochAndRound = EpochAndRound.internalBinaryRead(reader, reader.uint32(), options, message.epochAndRound);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feedConfigInfo)
      FeedConfigInfo.internalBinaryWrite(message.feedConfigInfo, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.epochAndRound)
      EpochAndRound.internalBinaryWrite(message.epochAndRound, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryFeedConfigInfoResponse = new QueryFeedConfigInfoResponse$Type();
class QueryLatestRoundRequest$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryLatestRoundRequest", [
      {
        no: 1,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.feedId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string feed_id */
        1:
          message.feedId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feedId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.feedId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLatestRoundRequest = new QueryLatestRoundRequest$Type();
class QueryLatestRoundResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryLatestRoundResponse", [
      {
        no: 1,
        name: "latest_round_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 2, name: "data", kind: "message", T: () => Transmission }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.latestRoundId = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 latest_round_id */
        1:
          message.latestRoundId = reader.uint64().toBigInt();
          break;
        case /* injective.ocr.v1beta1.Transmission data */
        2:
          message.data = Transmission.internalBinaryRead(reader, reader.uint32(), options, message.data);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.latestRoundId !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.latestRoundId);
    if (message.data)
      Transmission.internalBinaryWrite(message.data, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLatestRoundResponse = new QueryLatestRoundResponse$Type();
class QueryLatestTransmissionDetailsRequest$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryLatestTransmissionDetailsRequest", [
      {
        no: 1,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.feedId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string feed_id */
        1:
          message.feedId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.feedId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.feedId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLatestTransmissionDetailsRequest = new QueryLatestTransmissionDetailsRequest$Type();
class QueryLatestTransmissionDetailsResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryLatestTransmissionDetailsResponse", [
      {
        no: 1,
        name: "config_digest",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 2, name: "epoch_and_round", kind: "message", T: () => EpochAndRound },
      { no: 3, name: "data", kind: "message", T: () => Transmission }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.configDigest = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes config_digest */
        1:
          message.configDigest = reader.bytes();
          break;
        case /* injective.ocr.v1beta1.EpochAndRound epoch_and_round */
        2:
          message.epochAndRound = EpochAndRound.internalBinaryRead(reader, reader.uint32(), options, message.epochAndRound);
          break;
        case /* injective.ocr.v1beta1.Transmission data */
        3:
          message.data = Transmission.internalBinaryRead(reader, reader.uint32(), options, message.data);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.configDigest.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.configDigest);
    if (message.epochAndRound)
      EpochAndRound.internalBinaryWrite(message.epochAndRound, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.data)
      Transmission.internalBinaryWrite(message.data, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryLatestTransmissionDetailsResponse = new QueryLatestTransmissionDetailsResponse$Type();
class QueryOwedAmountRequest$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryOwedAmountRequest", [
      {
        no: 1,
        name: "transmitter",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.transmitter = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string transmitter */
        1:
          message.transmitter = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.transmitter !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.transmitter);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryOwedAmountRequest = new QueryOwedAmountRequest$Type();
class QueryOwedAmountResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryOwedAmountResponse", [
      { no: 1, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* cosmos.base.v1beta1.Coin amount */
        1:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryOwedAmountResponse = new QueryOwedAmountResponse$Type();
class QueryModuleStateRequest$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryModuleStateRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryModuleStateRequest = new QueryModuleStateRequest$Type();
class QueryModuleStateResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.QueryModuleStateResponse", [
      { no: 1, name: "state", kind: "message", T: () => GenesisState }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* injective.ocr.v1beta1.GenesisState state */
        1:
          message.state = GenesisState.internalBinaryRead(reader, reader.uint32(), options, message.state);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.state)
      GenesisState.internalBinaryWrite(message.state, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const QueryModuleStateResponse = new QueryModuleStateResponse$Type();
const Query = new ServiceType("injective.ocr.v1beta1.Query", [
  { name: "Params", options: { "google.api.http": { get: "/chainlink/ocr/v1beta1/params" } }, I: QueryParamsRequest, O: QueryParamsResponse },
  { name: "FeedConfig", options: { "google.api.http": { get: "/chainlink/ocr/v1beta1/feed_config/{feed_id}" } }, I: QueryFeedConfigRequest, O: QueryFeedConfigResponse },
  { name: "FeedConfigInfo", options: { "google.api.http": { get: "/chainlink/ocr/v1beta1/feed_config_info/{feed_id}" } }, I: QueryFeedConfigInfoRequest, O: QueryFeedConfigInfoResponse },
  { name: "LatestRound", options: { "google.api.http": { get: "/chainlink/ocr/v1beta1/latest_round/{feed_id}" } }, I: QueryLatestRoundRequest, O: QueryLatestRoundResponse },
  { name: "LatestTransmissionDetails", options: { "google.api.http": { get: "/chainlink/ocr/v1beta1/latest_transmission_details/{feed_id}" } }, I: QueryLatestTransmissionDetailsRequest, O: QueryLatestTransmissionDetailsResponse },
  { name: "OwedAmount", options: { "google.api.http": { get: "/chainlink/ocr/v1beta1/owed_amount/{transmitter}" } }, I: QueryOwedAmountRequest, O: QueryOwedAmountResponse },
  { name: "OcrModuleState", options: { "google.api.http": { get: "/chainlink/ocr/v1beta1/module_state" } }, I: QueryModuleStateRequest, O: QueryModuleStateResponse }
]);
export {
  Query,
  QueryFeedConfigInfoRequest,
  QueryFeedConfigInfoResponse,
  QueryFeedConfigRequest,
  QueryFeedConfigResponse,
  QueryLatestRoundRequest,
  QueryLatestRoundResponse,
  QueryLatestTransmissionDetailsRequest,
  QueryLatestTransmissionDetailsResponse,
  QueryModuleStateRequest,
  QueryModuleStateResponse,
  QueryOwedAmountRequest,
  QueryOwedAmountResponse,
  QueryParamsRequest,
  QueryParamsResponse
};
