// @injectivelabs/core-proto-ts-v2/esm/generated/injective/ocr/v1beta1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * CreateFeed defines a method for creating feed by module admin
   *
   * @generated from protobuf rpc: CreateFeed
   */
  createFeed(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * CreateFeed defines a method for creating feed by feed admin or feed billing
   * admin
   *
   * @generated from protobuf rpc: UpdateFeed
   */
  updateFeed(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Transmit defines a method for transmitting the feed info by transmitter
   *
   * @generated from protobuf rpc: Transmit
   */
  transmit(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * FundFeedRewardPool defines a method to put funds into feed reward pool
   *
   * @generated from protobuf rpc: FundFeedRewardPool
   */
  fundFeedRewardPool(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * WithdrawFeedRewardPool defines a method to witdhraw feed reward by feed
   * admin or billing admin
   *
   * @generated from protobuf rpc: WithdrawFeedRewardPool
   */
  withdrawFeedRewardPool(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * SetPayees defines a method to set payees for transmitters (batch action)
   *
   * @generated from protobuf rpc: SetPayees
   */
  setPayees(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * TransferPayeeship defines a method for a payee to transfer reward receive
   * ownership
   *
   * @generated from protobuf rpc: TransferPayeeship
   */
  transferPayeeship(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * AcceptPayeeship defines a method for a new payee to accept reward receive
   * ownership
   *
   * @generated from protobuf rpc: AcceptPayeeship
   */
  acceptPayeeship(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * @generated from protobuf rpc: UpdateParams
   */
  updateParams(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
