// @injectivelabs/core-proto-ts-v2/esm/generated/injective/ocr/v1beta1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./ocr_pb.js";
import { Coin } from "../../../cosmos/base/v1beta1/coin_pb.js";
import { Report } from "./ocr_pb.js";
import { FeedConfig } from "./ocr_pb.js";
class MsgCreateFeed$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgCreateFeed", [
      {
        no: 1,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "config", kind: "message", T: () => FeedConfig }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "ocr/MsgCreateFeed", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* injective.ocr.v1beta1.FeedConfig config */
        2:
          message.config = FeedConfig.internalBinaryRead(reader, reader.uint32(), options, message.config);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.config)
      FeedConfig.internalBinaryWrite(message.config, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateFeed = new MsgCreateFeed$Type();
class MsgCreateFeedResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgCreateFeedResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateFeedResponse = new MsgCreateFeedResponse$Type();
class MsgUpdateFeed$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgUpdateFeed", [
      {
        no: 1,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "signers",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "transmitters",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 5, name: "link_per_observation", kind: "scalar", T: 9, options: { "gogoproto.nullable": true, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      { no: 6, name: "link_per_transmission", kind: "scalar", T: 9, options: { "gogoproto.nullable": true, "gogoproto.customtype": "cosmossdk.io/math.Int" } },
      {
        no: 7,
        name: "link_denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "feed_admin",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 9,
        name: "billing_admin",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "ocr/MsgUpdateFeed", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.feedId = "";
    message.signers = [];
    message.transmitters = [];
    message.linkPerObservation = "";
    message.linkPerTransmission = "";
    message.linkDenom = "";
    message.feedAdmin = "";
    message.billingAdmin = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string feed_id */
        2:
          message.feedId = reader.string();
          break;
        case /* repeated string signers */
        3:
          message.signers.push(reader.string());
          break;
        case /* repeated string transmitters */
        4:
          message.transmitters.push(reader.string());
          break;
        case /* string link_per_observation */
        5:
          message.linkPerObservation = reader.string();
          break;
        case /* string link_per_transmission */
        6:
          message.linkPerTransmission = reader.string();
          break;
        case /* string link_denom */
        7:
          message.linkDenom = reader.string();
          break;
        case /* string feed_admin */
        8:
          message.feedAdmin = reader.string();
          break;
        case /* string billing_admin */
        9:
          message.billingAdmin = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.feedId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.feedId);
    for (let i = 0; i < message.signers.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.signers[i]);
    for (let i = 0; i < message.transmitters.length; i++)
      writer.tag(4, WireType.LengthDelimited).string(message.transmitters[i]);
    if (message.linkPerObservation !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.linkPerObservation);
    if (message.linkPerTransmission !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.linkPerTransmission);
    if (message.linkDenom !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.linkDenom);
    if (message.feedAdmin !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.feedAdmin);
    if (message.billingAdmin !== "")
      writer.tag(9, WireType.LengthDelimited).string(message.billingAdmin);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateFeed = new MsgUpdateFeed$Type();
class MsgUpdateFeedResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgUpdateFeedResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateFeedResponse = new MsgUpdateFeedResponse$Type();
class MsgTransmit$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgTransmit", [
      {
        no: 1,
        name: "transmitter",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "config_digest",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "epoch",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 5,
        name: "round",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 6,
        name: "extra_hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      { no: 7, name: "report", kind: "message", T: () => Report },
      {
        no: 8,
        name: "signatures",
        kind: "scalar",
        repeat: 2,
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "ocr/MsgTransmit", "cosmos.msg.v1.signer": ["transmitter"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.transmitter = "";
    message.configDigest = new Uint8Array(0);
    message.feedId = "";
    message.epoch = 0n;
    message.round = 0n;
    message.extraHash = new Uint8Array(0);
    message.signatures = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string transmitter */
        1:
          message.transmitter = reader.string();
          break;
        case /* bytes config_digest */
        2:
          message.configDigest = reader.bytes();
          break;
        case /* string feed_id */
        3:
          message.feedId = reader.string();
          break;
        case /* uint64 epoch */
        4:
          message.epoch = reader.uint64().toBigInt();
          break;
        case /* uint64 round */
        5:
          message.round = reader.uint64().toBigInt();
          break;
        case /* bytes extra_hash */
        6:
          message.extraHash = reader.bytes();
          break;
        case /* injective.ocr.v1beta1.Report report */
        7:
          message.report = Report.internalBinaryRead(reader, reader.uint32(), options, message.report);
          break;
        case /* repeated bytes signatures */
        8:
          message.signatures.push(reader.bytes());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.transmitter !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.transmitter);
    if (message.configDigest.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.configDigest);
    if (message.feedId !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.feedId);
    if (message.epoch !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.epoch);
    if (message.round !== 0n)
      writer.tag(5, WireType.Varint).uint64(message.round);
    if (message.extraHash.length)
      writer.tag(6, WireType.LengthDelimited).bytes(message.extraHash);
    if (message.report)
      Report.internalBinaryWrite(message.report, writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.signatures.length; i++)
      writer.tag(8, WireType.LengthDelimited).bytes(message.signatures[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTransmit = new MsgTransmit$Type();
class MsgTransmitResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgTransmitResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTransmitResponse = new MsgTransmitResponse$Type();
class MsgFundFeedRewardPool$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgFundFeedRewardPool", [
      {
        no: 1,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "ocr/MsgFundFeedRewardPool", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.feedId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string feed_id */
        2:
          message.feedId = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        3:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.feedId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.feedId);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgFundFeedRewardPool = new MsgFundFeedRewardPool$Type();
class MsgFundFeedRewardPoolResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgFundFeedRewardPoolResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgFundFeedRewardPoolResponse = new MsgFundFeedRewardPoolResponse$Type();
class MsgWithdrawFeedRewardPool$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgWithdrawFeedRewardPool", [
      {
        no: 1,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "amount", kind: "message", T: () => Coin, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "ocr/MsgWithdrawFeedRewardPool", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.feedId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string feed_id */
        2:
          message.feedId = reader.string();
          break;
        case /* cosmos.base.v1beta1.Coin amount */
        3:
          message.amount = Coin.internalBinaryRead(reader, reader.uint32(), options, message.amount);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.feedId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.feedId);
    if (message.amount)
      Coin.internalBinaryWrite(message.amount, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgWithdrawFeedRewardPool = new MsgWithdrawFeedRewardPool$Type();
class MsgWithdrawFeedRewardPoolResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgWithdrawFeedRewardPoolResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgWithdrawFeedRewardPoolResponse = new MsgWithdrawFeedRewardPoolResponse$Type();
class MsgSetPayees$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgSetPayees", [
      {
        no: 1,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "transmitters",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "payees",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "ocr/MsgSetPayees", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.feedId = "";
    message.transmitters = [];
    message.payees = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string feed_id */
        2:
          message.feedId = reader.string();
          break;
        case /* repeated string transmitters */
        3:
          message.transmitters.push(reader.string());
          break;
        case /* repeated string payees */
        4:
          message.payees.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.feedId !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.feedId);
    for (let i = 0; i < message.transmitters.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.transmitters[i]);
    for (let i = 0; i < message.payees.length; i++)
      writer.tag(4, WireType.LengthDelimited).string(message.payees[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSetPayees = new MsgSetPayees$Type();
class MsgSetPayeesResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgSetPayeesResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgSetPayeesResponse = new MsgSetPayeesResponse$Type();
class MsgTransferPayeeship$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgTransferPayeeship", [
      {
        no: 1,
        name: "sender",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "transmitter",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "proposed",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "ocr/MsgTransferPayeeship", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.transmitter = "";
    message.feedId = "";
    message.proposed = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string transmitter */
        2:
          message.transmitter = reader.string();
          break;
        case /* string feed_id */
        3:
          message.feedId = reader.string();
          break;
        case /* string proposed */
        4:
          message.proposed = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.transmitter !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.transmitter);
    if (message.feedId !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.feedId);
    if (message.proposed !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.proposed);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTransferPayeeship = new MsgTransferPayeeship$Type();
class MsgTransferPayeeshipResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgTransferPayeeshipResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgTransferPayeeshipResponse = new MsgTransferPayeeshipResponse$Type();
class MsgAcceptPayeeship$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgAcceptPayeeship", [
      {
        no: 1,
        name: "payee",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "transmitter",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "feed_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "amino.name": "ocr/MsgAcceptPayeeship", "cosmos.msg.v1.signer": ["transmitter"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.payee = "";
    message.transmitter = "";
    message.feedId = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string payee */
        1:
          message.payee = reader.string();
          break;
        case /* string transmitter */
        2:
          message.transmitter = reader.string();
          break;
        case /* string feed_id */
        3:
          message.feedId = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.payee !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.payee);
    if (message.transmitter !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.transmitter);
    if (message.feedId !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.feedId);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgAcceptPayeeship = new MsgAcceptPayeeship$Type();
class MsgAcceptPayeeshipResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgAcceptPayeeshipResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgAcceptPayeeshipResponse = new MsgAcceptPayeeshipResponse$Type();
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgUpdateParams", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ], { "amino.name": "ocr/MsgUpdateParams", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* injective.ocr.v1beta1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("injective.ocr.v1beta1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
const Msg = new ServiceType("injective.ocr.v1beta1.Msg", [
  { name: "CreateFeed", options: {}, I: MsgCreateFeed, O: MsgCreateFeedResponse },
  { name: "UpdateFeed", options: {}, I: MsgUpdateFeed, O: MsgUpdateFeedResponse },
  { name: "Transmit", options: {}, I: MsgTransmit, O: MsgTransmitResponse },
  { name: "FundFeedRewardPool", options: {}, I: MsgFundFeedRewardPool, O: MsgFundFeedRewardPoolResponse },
  { name: "WithdrawFeedRewardPool", options: {}, I: MsgWithdrawFeedRewardPool, O: MsgWithdrawFeedRewardPoolResponse },
  { name: "SetPayees", options: {}, I: MsgSetPayees, O: MsgSetPayeesResponse },
  { name: "TransferPayeeship", options: {}, I: MsgTransferPayeeship, O: MsgTransferPayeeshipResponse },
  { name: "AcceptPayeeship", options: {}, I: MsgAcceptPayeeship, O: MsgAcceptPayeeshipResponse },
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgAcceptPayeeship,
  MsgAcceptPayeeshipResponse,
  MsgCreateFeed,
  MsgCreateFeedResponse,
  MsgFundFeedRewardPool,
  MsgFundFeedRewardPoolResponse,
  MsgSetPayees,
  MsgSetPayeesResponse,
  MsgTransferPayeeship,
  MsgTransferPayeeshipResponse,
  MsgTransmit,
  MsgTransmitResponse,
  MsgUpdateFeed,
  MsgUpdateFeedResponse,
  MsgUpdateParams,
  MsgUpdateParamsResponse,
  MsgWithdrawFeedRewardPool,
  MsgWithdrawFeedRewardPoolResponse
};
