// @injectivelabs/core-proto-ts-v2/esm/generated/injective/permissions/v1beta1/params_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Params$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.Params", [
      {
        no: 1,
        name: "wasm_hook_query_max_gas",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ], { "gogoproto.equal": true, "amino.name": "permissions/Params" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.wasmHookQueryMaxGas = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 wasm_hook_query_max_gas */
        1:
          message.wasmHookQueryMaxGas = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.wasmHookQueryMaxGas !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.wasmHookQueryMaxGas);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
export {
  Params
};
