// @injectivelabs/core-proto-ts-v2/esm/generated/injective/permissions/v1beta1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Params defines a gRPC query method that returns the permissions module's
   * parameters.
   *
   * @generated from protobuf rpc: Params
   */
  params(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * NamespaceDenoms defines a gRPC query method that returns the denoms for
   * which a namespace exists
   *
   * @generated from protobuf rpc: NamespaceDenoms
   */
  namespaceDenoms(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Namespaces defines a gRPC query method that returns the permissions
   * module's created namespaces.
   *
   * @generated from protobuf rpc: Namespaces
   */
  namespaces(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Namespace defines a gRPC query method that returns the permissions
   * module's namespace associated with the provided denom.
   *
   * @generated from protobuf rpc: Namespace
   */
  namespace(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * RolesByActor defines a gRPC query method that returns roles for the actor
   * in the namespace
   *
   * @generated from protobuf rpc: RolesByActor
   */
  rolesByActor(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ActorsByRole defines a gRPC query method that returns a namespace's roles
   * associated with the provided actor.
   *
   * @generated from protobuf rpc: ActorsByRole
   */
  actorsByRole(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * RoleManagers defines a gRPC query method that returns a namespace's role
   * managers
   *
   * @generated from protobuf rpc: RoleManagers
   */
  roleManagers(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * RoleManager defines a gRPC query method that returns the roles a given role
   * manager manages for a given namespace
   *
   * @generated from protobuf rpc: RoleManager
   */
  roleManager(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * PolicyStatuses defines a gRPC query method that returns a namespace's
   * policy statuses
   *
   * @generated from protobuf rpc: PolicyStatuses
   */
  policyStatuses(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * PolicyManagerCapabilities defines a gRPC query method that returns a
   * namespace's policy manager capabilities
   *
   * @generated from protobuf rpc: PolicyManagerCapabilities
   */
  policyManagerCapabilities(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Vouchers defines a gRPC query method for the vouchers for a given denom
   *
   * @generated from protobuf rpc: Vouchers
   */
  vouchers(input, options) {
    const method = this.methods[10], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Voucher defines a gRPC query method for the vouchers for a given denom and
   * address
   *
   * @generated from protobuf rpc: Voucher
   */
  voucher(input, options) {
    const method = this.methods[11], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the entire permissions module's state
   *
   * @generated from protobuf rpc: PermissionsModuleState
   */
  permissionsModuleState(input, options) {
    const method = this.methods[12], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
