// @injectivelabs/core-proto-ts-v2/esm/generated/injective/permissions/v1beta1/tx_pb.js
import { ServiceType } from "@protobuf-ts/runtime-rpc";
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { RoleActors } from "./permissions_pb.js";
import { PolicyManagerCapability } from "./permissions_pb.js";
import { PolicyStatus } from "./permissions_pb.js";
import { RoleManager } from "./permissions_pb.js";
import { Role } from "./permissions_pb.js";
import { Namespace } from "./permissions_pb.js";
import { Params } from "./params_pb.js";
class MsgUpdateParams$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgUpdateParams", [
      { no: 1, name: "authority", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } }
    ], { "amino.name": "permissions/MsgUpdateParams", "cosmos.msg.v1.signer": ["authority"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.authority = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string authority */
        1:
          message.authority = reader.string();
          break;
        case /* injective.permissions.v1beta1.Params params */
        2:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authority !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.authority);
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParams = new MsgUpdateParams$Type();
class MsgUpdateParamsResponse$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgUpdateParamsResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateParamsResponse = new MsgUpdateParamsResponse$Type();
class MsgCreateNamespace$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgCreateNamespace", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"sender"' } },
      { no: 2, name: "namespace", kind: "message", T: () => Namespace, options: { "gogoproto.nullable": false } }
    ], { "amino.name": "permissions/MsgCreateNamespace", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* injective.permissions.v1beta1.Namespace namespace */
        2:
          message.namespace = Namespace.internalBinaryRead(reader, reader.uint32(), options, message.namespace);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.namespace)
      Namespace.internalBinaryWrite(message.namespace, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateNamespace = new MsgCreateNamespace$Type();
class MsgCreateNamespaceResponse$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgCreateNamespaceResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgCreateNamespaceResponse = new MsgCreateNamespaceResponse$Type();
class MsgUpdateNamespace$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgUpdateNamespace", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"sender"' } },
      {
        no: 2,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "contract_hook", kind: "message", T: () => MsgUpdateNamespace_SetContractHook },
      { no: 4, name: "role_permissions", kind: "message", repeat: 2, T: () => Role },
      { no: 5, name: "role_managers", kind: "message", repeat: 2, T: () => RoleManager },
      { no: 6, name: "policy_statuses", kind: "message", repeat: 2, T: () => PolicyStatus },
      { no: 7, name: "policy_manager_capabilities", kind: "message", repeat: 2, T: () => PolicyManagerCapability }
    ], { "amino.name": "permissions/MsgUpdateNamespace", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.denom = "";
    message.rolePermissions = [];
    message.roleManagers = [];
    message.policyStatuses = [];
    message.policyManagerCapabilities = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string denom */
        2:
          message.denom = reader.string();
          break;
        case /* injective.permissions.v1beta1.MsgUpdateNamespace.SetContractHook contract_hook */
        3:
          message.contractHook = MsgUpdateNamespace_SetContractHook.internalBinaryRead(reader, reader.uint32(), options, message.contractHook);
          break;
        case /* repeated injective.permissions.v1beta1.Role role_permissions */
        4:
          message.rolePermissions.push(Role.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective.permissions.v1beta1.RoleManager role_managers */
        5:
          message.roleManagers.push(RoleManager.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective.permissions.v1beta1.PolicyStatus policy_statuses */
        6:
          message.policyStatuses.push(PolicyStatus.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective.permissions.v1beta1.PolicyManagerCapability policy_manager_capabilities */
        7:
          message.policyManagerCapabilities.push(PolicyManagerCapability.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.denom !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.denom);
    if (message.contractHook)
      MsgUpdateNamespace_SetContractHook.internalBinaryWrite(message.contractHook, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.rolePermissions.length; i++)
      Role.internalBinaryWrite(message.rolePermissions[i], writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.roleManagers.length; i++)
      RoleManager.internalBinaryWrite(message.roleManagers[i], writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.policyStatuses.length; i++)
      PolicyStatus.internalBinaryWrite(message.policyStatuses[i], writer.tag(6, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.policyManagerCapabilities.length; i++)
      PolicyManagerCapability.internalBinaryWrite(message.policyManagerCapabilities[i], writer.tag(7, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateNamespace = new MsgUpdateNamespace$Type();
class MsgUpdateNamespace_SetContractHook$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgUpdateNamespace.SetContractHook", [
      {
        no: 1,
        name: "new_value",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.newValue = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string new_value */
        1:
          message.newValue = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.newValue !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.newValue);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateNamespace_SetContractHook = new MsgUpdateNamespace_SetContractHook$Type();
class MsgUpdateNamespaceResponse$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgUpdateNamespaceResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateNamespaceResponse = new MsgUpdateNamespaceResponse$Type();
class MsgUpdateActorRoles$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgUpdateActorRoles", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"sender"' } },
      {
        no: 2,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "role_actors_to_add", kind: "message", repeat: 2, T: () => RoleActors },
      { no: 5, name: "role_actors_to_revoke", kind: "message", repeat: 2, T: () => RoleActors }
    ], { "amino.name": "permissions/MsgUpdateActorRoles", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.denom = "";
    message.roleActorsToAdd = [];
    message.roleActorsToRevoke = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string denom */
        2:
          message.denom = reader.string();
          break;
        case /* repeated injective.permissions.v1beta1.RoleActors role_actors_to_add */
        3:
          message.roleActorsToAdd.push(RoleActors.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated injective.permissions.v1beta1.RoleActors role_actors_to_revoke */
        5:
          message.roleActorsToRevoke.push(RoleActors.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.denom !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.denom);
    for (let i = 0; i < message.roleActorsToAdd.length; i++)
      RoleActors.internalBinaryWrite(message.roleActorsToAdd[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.roleActorsToRevoke.length; i++)
      RoleActors.internalBinaryWrite(message.roleActorsToRevoke[i], writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateActorRoles = new MsgUpdateActorRoles$Type();
class MsgUpdateActorRolesResponse$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgUpdateActorRolesResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgUpdateActorRolesResponse = new MsgUpdateActorRolesResponse$Type();
class MsgClaimVoucher$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgClaimVoucher", [
      { no: 1, name: "sender", kind: "scalar", T: 9, options: { "gogoproto.moretags": 'yaml:"sender"' } },
      {
        no: 2,
        name: "denom",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "amino.name": "permissions/MsgClaimVoucher", "cosmos.msg.v1.signer": ["sender"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sender = "";
    message.denom = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string sender */
        1:
          message.sender = reader.string();
          break;
        case /* string denom */
        2:
          message.denom = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sender !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.sender);
    if (message.denom !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.denom);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgClaimVoucher = new MsgClaimVoucher$Type();
class MsgClaimVoucherResponse$Type extends MessageType {
  constructor() {
    super("injective.permissions.v1beta1.MsgClaimVoucherResponse", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const MsgClaimVoucherResponse = new MsgClaimVoucherResponse$Type();
const Msg = new ServiceType("injective.permissions.v1beta1.Msg", [
  { name: "UpdateParams", options: {}, I: MsgUpdateParams, O: MsgUpdateParamsResponse },
  { name: "CreateNamespace", options: {}, I: MsgCreateNamespace, O: MsgCreateNamespaceResponse },
  { name: "UpdateNamespace", options: {}, I: MsgUpdateNamespace, O: MsgUpdateNamespaceResponse },
  { name: "UpdateActorRoles", options: {}, I: MsgUpdateActorRoles, O: MsgUpdateActorRolesResponse },
  { name: "ClaimVoucher", options: {}, I: MsgClaimVoucher, O: MsgClaimVoucherResponse }
], { "cosmos.msg.v1.service": true });
export {
  Msg,
  MsgClaimVoucher,
  MsgClaimVoucherResponse,
  MsgCreateNamespace,
  MsgCreateNamespaceResponse,
  MsgUpdateActorRoles,
  MsgUpdateActorRolesResponse,
  MsgUpdateNamespace,
  MsgUpdateNamespaceResponse,
  MsgUpdateNamespace_SetContractHook,
  MsgUpdateParams,
  MsgUpdateParamsResponse
};
