// @injectivelabs/core-proto-ts-v2/esm/generated/cosmos/authz/v1beta1/authz_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Timestamp } from "../../../google/protobuf/timestamp_pb.js";
import { Any } from "../../../google/protobuf/any_pb.js";
class GenericAuthorization$Type extends MessageType {
  constructor() {
    super("cosmos.authz.v1beta1.GenericAuthorization", [
      {
        no: 1,
        name: "msg",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "cosmos_proto.implements_interface": ["cosmos.authz.v1beta1.Authorization"], "amino.name": "cosmos-sdk/GenericAuthorization" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.msg = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string msg */
        1:
          message.msg = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.msg !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.msg);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenericAuthorization = new GenericAuthorization$Type();
class Grant$Type extends MessageType {
  constructor() {
    super("cosmos.authz.v1beta1.Grant", [
      { no: 1, name: "authorization", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.authz.v1beta1.Authorization" } },
      { no: 2, name: "expiration", kind: "message", T: () => Timestamp, options: { "gogoproto.nullable": true, "gogoproto.stdtime": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* google.protobuf.Any authorization */
        1:
          message.authorization = Any.internalBinaryRead(reader, reader.uint32(), options, message.authorization);
          break;
        case /* google.protobuf.Timestamp expiration */
        2:
          message.expiration = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.expiration);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.authorization)
      Any.internalBinaryWrite(message.authorization, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.expiration)
      Timestamp.internalBinaryWrite(message.expiration, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Grant = new Grant$Type();
class GrantAuthorization$Type extends MessageType {
  constructor() {
    super("cosmos.authz.v1beta1.GrantAuthorization", [
      { no: 1, name: "granter", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 2, name: "grantee", kind: "scalar", T: 9, options: { "cosmos_proto.scalar": "cosmos.AddressString" } },
      { no: 3, name: "authorization", kind: "message", T: () => Any, options: { "cosmos_proto.accepts_interface": "cosmos.authz.v1beta1.Authorization" } },
      { no: 4, name: "expiration", kind: "message", T: () => Timestamp, options: { "gogoproto.stdtime": true } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.granter = "";
    message.grantee = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string granter */
        1:
          message.granter = reader.string();
          break;
        case /* string grantee */
        2:
          message.grantee = reader.string();
          break;
        case /* google.protobuf.Any authorization */
        3:
          message.authorization = Any.internalBinaryRead(reader, reader.uint32(), options, message.authorization);
          break;
        case /* google.protobuf.Timestamp expiration */
        4:
          message.expiration = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.expiration);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.granter !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.granter);
    if (message.grantee !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.grantee);
    if (message.authorization)
      Any.internalBinaryWrite(message.authorization, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.expiration)
      Timestamp.internalBinaryWrite(message.expiration, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GrantAuthorization = new GrantAuthorization$Type();
class GrantQueueItem$Type extends MessageType {
  constructor() {
    super("cosmos.authz.v1beta1.GrantQueueItem", [
      {
        no: 1,
        name: "msg_type_urls",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.msgTypeUrls = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated string msg_type_urls */
        1:
          message.msgTypeUrls.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.msgTypeUrls.length; i++)
      writer.tag(1, WireType.LengthDelimited).string(message.msgTypeUrls[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GrantQueueItem = new GrantQueueItem$Type();
export {
  GenericAuthorization,
  Grant,
  GrantAuthorization,
  GrantQueueItem
};
