// @injectivelabs/core-proto-ts-v2/esm/generated/injective/wasmx/v1/events_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { FundingMode } from "./proposal_pb.js";
class EventContractExecution$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.EventContractExecution", [
      {
        no: 1,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "response",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 3,
        name: "other_error",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 4,
        name: "execution_error",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contractAddress = "";
    message.response = new Uint8Array(0);
    message.otherError = "";
    message.executionError = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract_address */
        1:
          message.contractAddress = reader.string();
          break;
        case /* bytes response */
        2:
          message.response = reader.bytes();
          break;
        case /* string other_error */
        3:
          message.otherError = reader.string();
          break;
        case /* string execution_error */
        4:
          message.executionError = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contractAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contractAddress);
    if (message.response.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.response);
    if (message.otherError !== "")
      writer.tag(3, WireType.LengthDelimited).string(message.otherError);
    if (message.executionError !== "")
      writer.tag(4, WireType.LengthDelimited).string(message.executionError);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventContractExecution = new EventContractExecution$Type();
class EventContractRegistered$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.EventContractRegistered", [
      {
        no: 1,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "gas_price",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "should_pin_contract",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 5,
        name: "is_migration_allowed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 6,
        name: "code_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "admin_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "granter_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 9, name: "funding_mode", kind: "enum", T: () => ["injective.wasmx.v1.FundingMode", FundingMode] }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contractAddress = "";
    message.gasPrice = 0n;
    message.shouldPinContract = false;
    message.isMigrationAllowed = false;
    message.codeId = 0n;
    message.adminAddress = "";
    message.granterAddress = "";
    message.fundingMode = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract_address */
        1:
          message.contractAddress = reader.string();
          break;
        case /* uint64 gas_price */
        3:
          message.gasPrice = reader.uint64().toBigInt();
          break;
        case /* bool should_pin_contract */
        4:
          message.shouldPinContract = reader.bool();
          break;
        case /* bool is_migration_allowed */
        5:
          message.isMigrationAllowed = reader.bool();
          break;
        case /* uint64 code_id */
        6:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string admin_address */
        7:
          message.adminAddress = reader.string();
          break;
        case /* string granter_address */
        8:
          message.granterAddress = reader.string();
          break;
        case /* injective.wasmx.v1.FundingMode funding_mode */
        9:
          message.fundingMode = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contractAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contractAddress);
    if (message.gasPrice !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.gasPrice);
    if (message.shouldPinContract !== false)
      writer.tag(4, WireType.Varint).bool(message.shouldPinContract);
    if (message.isMigrationAllowed !== false)
      writer.tag(5, WireType.Varint).bool(message.isMigrationAllowed);
    if (message.codeId !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.codeId);
    if (message.adminAddress !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.adminAddress);
    if (message.granterAddress !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.granterAddress);
    if (message.fundingMode !== 0)
      writer.tag(9, WireType.Varint).int32(message.fundingMode);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventContractRegistered = new EventContractRegistered$Type();
class EventContractDeregistered$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.EventContractDeregistered", [
      {
        no: 1,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contractAddress = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract_address */
        1:
          message.contractAddress = reader.string();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contractAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contractAddress);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const EventContractDeregistered = new EventContractDeregistered$Type();
export {
  EventContractDeregistered,
  EventContractExecution,
  EventContractRegistered
};
