// @injectivelabs/core-proto-ts-v2/esm/generated/injective/wasmx/v1/proposal_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { StoreCodeProposal } from "../../../cosmwasm/wasm/v1/proposal_legacy_pb.js";
var FundingMode = /* @__PURE__ */ ((FundingMode2) => {
  FundingMode2[FundingMode2["Unspecified"] = 0] = "Unspecified";
  FundingMode2[FundingMode2["SelfFunded"] = 1] = "SelfFunded";
  FundingMode2[FundingMode2["GrantOnly"] = 2] = "GrantOnly";
  FundingMode2[FundingMode2["Dual"] = 3] = "Dual";
  return FundingMode2;
})(FundingMode || {});
class ContractRegistrationRequestProposal$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.ContractRegistrationRequestProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "contract_registration_request", kind: "message", T: () => ContractRegistrationRequest, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasmx/ContractRegistrationRequestProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* injective.wasmx.v1.ContractRegistrationRequest contract_registration_request */
        3:
          message.contractRegistrationRequest = ContractRegistrationRequest.internalBinaryRead(reader, reader.uint32(), options, message.contractRegistrationRequest);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    if (message.contractRegistrationRequest)
      ContractRegistrationRequest.internalBinaryWrite(message.contractRegistrationRequest, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ContractRegistrationRequestProposal = new ContractRegistrationRequestProposal$Type();
class BatchContractRegistrationRequestProposal$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.BatchContractRegistrationRequestProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "contract_registration_requests", kind: "message", repeat: 2, T: () => ContractRegistrationRequest, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasmx/BatchContractRegistrationRequestProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.contractRegistrationRequests = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* repeated injective.wasmx.v1.ContractRegistrationRequest contract_registration_requests */
        3:
          message.contractRegistrationRequests.push(ContractRegistrationRequest.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    for (let i = 0; i < message.contractRegistrationRequests.length; i++)
      ContractRegistrationRequest.internalBinaryWrite(message.contractRegistrationRequests[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BatchContractRegistrationRequestProposal = new BatchContractRegistrationRequestProposal$Type();
class BatchContractDeregistrationProposal$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.BatchContractDeregistrationProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 3,
        name: "contracts",
        kind: "scalar",
        repeat: 2,
        T: 9
        /*ScalarType.STRING*/
      }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasmx/BatchContractDeregistrationProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.contracts = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* repeated string contracts */
        3:
          message.contracts.push(reader.string());
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    for (let i = 0; i < message.contracts.length; i++)
      writer.tag(3, WireType.LengthDelimited).string(message.contracts[i]);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BatchContractDeregistrationProposal = new BatchContractDeregistrationProposal$Type();
class ContractRegistrationRequest$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.ContractRegistrationRequest", [
      {
        no: 1,
        name: "contract_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "gas_limit",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "gas_price",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "should_pin_contract",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 5,
        name: "is_migration_allowed",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 6,
        name: "code_id",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 7,
        name: "admin_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 8,
        name: "granter_address",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 9, name: "funding_mode", kind: "enum", T: () => ["injective.wasmx.v1.FundingMode", FundingMode] }
    ], { "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"] });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.contractAddress = "";
    message.gasLimit = 0n;
    message.gasPrice = 0n;
    message.shouldPinContract = false;
    message.isMigrationAllowed = false;
    message.codeId = 0n;
    message.adminAddress = "";
    message.granterAddress = "";
    message.fundingMode = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string contract_address */
        1:
          message.contractAddress = reader.string();
          break;
        case /* uint64 gas_limit */
        2:
          message.gasLimit = reader.uint64().toBigInt();
          break;
        case /* uint64 gas_price */
        3:
          message.gasPrice = reader.uint64().toBigInt();
          break;
        case /* bool should_pin_contract */
        4:
          message.shouldPinContract = reader.bool();
          break;
        case /* bool is_migration_allowed */
        5:
          message.isMigrationAllowed = reader.bool();
          break;
        case /* uint64 code_id */
        6:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string admin_address */
        7:
          message.adminAddress = reader.string();
          break;
        case /* string granter_address */
        8:
          message.granterAddress = reader.string();
          break;
        case /* injective.wasmx.v1.FundingMode funding_mode */
        9:
          message.fundingMode = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.contractAddress !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.contractAddress);
    if (message.gasLimit !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.gasLimit);
    if (message.gasPrice !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.gasPrice);
    if (message.shouldPinContract !== false)
      writer.tag(4, WireType.Varint).bool(message.shouldPinContract);
    if (message.isMigrationAllowed !== false)
      writer.tag(5, WireType.Varint).bool(message.isMigrationAllowed);
    if (message.codeId !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.codeId);
    if (message.adminAddress !== "")
      writer.tag(7, WireType.LengthDelimited).string(message.adminAddress);
    if (message.granterAddress !== "")
      writer.tag(8, WireType.LengthDelimited).string(message.granterAddress);
    if (message.fundingMode !== 0)
      writer.tag(9, WireType.Varint).int32(message.fundingMode);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ContractRegistrationRequest = new ContractRegistrationRequest$Type();
class BatchStoreCodeProposal$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.BatchStoreCodeProposal", [
      {
        no: 1,
        name: "title",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      {
        no: 2,
        name: "description",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 3, name: "proposals", kind: "message", repeat: 2, T: () => StoreCodeProposal, options: { "gogoproto.nullable": false } }
    ], { "gogoproto.goproto_getters": false, "gogoproto.equal": false, "cosmos_proto.implements_interface": ["cosmos.gov.v1beta1.Content"], "amino.name": "wasmx/BatchStoreCodeProposal" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.title = "";
    message.description = "";
    message.proposals = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string title */
        1:
          message.title = reader.string();
          break;
        case /* string description */
        2:
          message.description = reader.string();
          break;
        case /* repeated cosmwasm.wasm.v1.StoreCodeProposal proposals */
        3:
          message.proposals.push(StoreCodeProposal.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.title !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.title);
    if (message.description !== "")
      writer.tag(2, WireType.LengthDelimited).string(message.description);
    for (let i = 0; i < message.proposals.length; i++)
      StoreCodeProposal.internalBinaryWrite(message.proposals[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const BatchStoreCodeProposal = new BatchStoreCodeProposal$Type();
export {
  BatchContractDeregistrationProposal,
  BatchContractRegistrationRequestProposal,
  BatchStoreCodeProposal,
  ContractRegistrationRequest,
  ContractRegistrationRequestProposal,
  FundingMode
};
