// @injectivelabs/core-proto-ts-v2/esm/generated/injective/wasmx/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * Retrieves wasmx params
   *
   * @generated from protobuf rpc: WasmxParams
   */
  wasmxParams(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves contract registration info
   *
   * @generated from protobuf rpc: ContractRegistrationInfo
   */
  contractRegistrationInfo(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Retrieves the entire wasmx module's state
   *
   * @generated from protobuf rpc: WasmxModuleState
   */
  wasmxModuleState(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
