// @injectivelabs/core-proto-ts-v2/esm/generated/injective/wasmx/v1/wasmx_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { FundingMode } from "./proposal_pb.js";
import { AccessConfig } from "../../../cosmwasm/wasm/v1/types_pb.js";
class Params$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.Params", [
      {
        no: 1,
        name: "is_execution_enabled",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 2,
        name: "max_begin_block_total_gas",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "max_contract_gas_limit",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 4,
        name: "min_gas_price",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      { no: 5, name: "register_contract_access", kind: "message", T: () => AccessConfig, options: { "gogoproto.nullable": false, "gogoproto.moretags": 'yaml:"register_contract_access"', "amino.dont_omitempty": true } }
    ], { "gogoproto.equal": true, "amino.name": "wasmx/Params" });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.isExecutionEnabled = false;
    message.maxBeginBlockTotalGas = 0n;
    message.maxContractGasLimit = 0n;
    message.minGasPrice = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bool is_execution_enabled */
        1:
          message.isExecutionEnabled = reader.bool();
          break;
        case /* uint64 max_begin_block_total_gas */
        2:
          message.maxBeginBlockTotalGas = reader.uint64().toBigInt();
          break;
        case /* uint64 max_contract_gas_limit */
        3:
          message.maxContractGasLimit = reader.uint64().toBigInt();
          break;
        case /* uint64 min_gas_price */
        4:
          message.minGasPrice = reader.uint64().toBigInt();
          break;
        case /* cosmwasm.wasm.v1.AccessConfig register_contract_access */
        5:
          message.registerContractAccess = AccessConfig.internalBinaryRead(reader, reader.uint32(), options, message.registerContractAccess);
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.isExecutionEnabled !== false)
      writer.tag(1, WireType.Varint).bool(message.isExecutionEnabled);
    if (message.maxBeginBlockTotalGas !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.maxBeginBlockTotalGas);
    if (message.maxContractGasLimit !== 0n)
      writer.tag(3, WireType.Varint).uint64(message.maxContractGasLimit);
    if (message.minGasPrice !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.minGasPrice);
    if (message.registerContractAccess)
      AccessConfig.internalBinaryWrite(message.registerContractAccess, writer.tag(5, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Params = new Params$Type();
class RegisteredContract$Type extends MessageType {
  constructor() {
    super("injective.wasmx.v1.RegisteredContract", [
      {
        no: 1,
        name: "gas_limit",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "gas_price",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 3,
        name: "is_executable",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      { no: 4, name: "code_id", kind: "scalar", T: 4, L: 0, options: { "gogoproto.nullable": true } },
      { no: 5, name: "admin_address", kind: "scalar", T: 9, options: { "gogoproto.nullable": true } },
      { no: 6, name: "granter_address", kind: "scalar", T: 9, options: { "gogoproto.nullable": true } },
      { no: 7, name: "fund_mode", kind: "enum", T: () => ["injective.wasmx.v1.FundingMode", FundingMode] }
    ], { "gogoproto.equal": true });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.gasLimit = 0n;
    message.gasPrice = 0n;
    message.isExecutable = false;
    message.codeId = 0n;
    message.adminAddress = "";
    message.granterAddress = "";
    message.fundMode = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 gas_limit */
        1:
          message.gasLimit = reader.uint64().toBigInt();
          break;
        case /* uint64 gas_price */
        2:
          message.gasPrice = reader.uint64().toBigInt();
          break;
        case /* bool is_executable */
        3:
          message.isExecutable = reader.bool();
          break;
        case /* uint64 code_id */
        4:
          message.codeId = reader.uint64().toBigInt();
          break;
        case /* string admin_address */
        5:
          message.adminAddress = reader.string();
          break;
        case /* string granter_address */
        6:
          message.granterAddress = reader.string();
          break;
        case /* injective.wasmx.v1.FundingMode fund_mode */
        7:
          message.fundMode = reader.int32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.gasLimit !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.gasLimit);
    if (message.gasPrice !== 0n)
      writer.tag(2, WireType.Varint).uint64(message.gasPrice);
    if (message.isExecutable !== false)
      writer.tag(3, WireType.Varint).bool(message.isExecutable);
    if (message.codeId !== 0n)
      writer.tag(4, WireType.Varint).uint64(message.codeId);
    if (message.adminAddress !== "")
      writer.tag(5, WireType.LengthDelimited).string(message.adminAddress);
    if (message.granterAddress !== "")
      writer.tag(6, WireType.LengthDelimited).string(message.granterAddress);
    if (message.fundMode !== 0)
      writer.tag(7, WireType.Varint).int32(message.fundMode);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const RegisteredContract = new RegisteredContract$Type();
export {
  Params,
  RegisteredContract
};
