// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/client/v1/genesis_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
import { Params } from "./client_pb.js";
import { ClientConsensusStates } from "./client_pb.js";
import { IdentifiedClientState } from "./client_pb.js";
class GenesisState$Type extends MessageType {
  constructor() {
    super("ibc.core.client.v1.GenesisState", [
      { no: 1, name: "clients", kind: "message", repeat: 2, T: () => IdentifiedClientState, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "IdentifiedClientStates" } },
      { no: 2, name: "clients_consensus", kind: "message", repeat: 2, T: () => ClientConsensusStates, options: { "gogoproto.nullable": false, "gogoproto.castrepeated": "ClientsConsensusStates" } },
      { no: 3, name: "clients_metadata", kind: "message", repeat: 2, T: () => IdentifiedGenesisMetadata, options: { "gogoproto.nullable": false } },
      { no: 4, name: "params", kind: "message", T: () => Params, options: { "gogoproto.nullable": false } },
      {
        no: 5,
        name: "create_localhost",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      },
      {
        no: 6,
        name: "next_client_sequence",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clients = [];
    message.clientsConsensus = [];
    message.clientsMetadata = [];
    message.createLocalhost = false;
    message.nextClientSequence = 0n;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* repeated ibc.core.client.v1.IdentifiedClientState clients */
        1:
          message.clients.push(IdentifiedClientState.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.core.client.v1.ClientConsensusStates clients_consensus */
        2:
          message.clientsConsensus.push(ClientConsensusStates.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* repeated ibc.core.client.v1.IdentifiedGenesisMetadata clients_metadata */
        3:
          message.clientsMetadata.push(IdentifiedGenesisMetadata.internalBinaryRead(reader, reader.uint32(), options));
          break;
        case /* ibc.core.client.v1.Params params */
        4:
          message.params = Params.internalBinaryRead(reader, reader.uint32(), options, message.params);
          break;
        case /* bool create_localhost = 5 [deprecated = true] */
        5:
          message.createLocalhost = reader.bool();
          break;
        case /* uint64 next_client_sequence */
        6:
          message.nextClientSequence = reader.uint64().toBigInt();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    for (let i = 0; i < message.clients.length; i++)
      IdentifiedClientState.internalBinaryWrite(message.clients[i], writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.clientsConsensus.length; i++)
      ClientConsensusStates.internalBinaryWrite(message.clientsConsensus[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    for (let i = 0; i < message.clientsMetadata.length; i++)
      IdentifiedGenesisMetadata.internalBinaryWrite(message.clientsMetadata[i], writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.params)
      Params.internalBinaryWrite(message.params, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    if (message.createLocalhost !== false)
      writer.tag(5, WireType.Varint).bool(message.createLocalhost);
    if (message.nextClientSequence !== 0n)
      writer.tag(6, WireType.Varint).uint64(message.nextClientSequence);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisState = new GenesisState$Type();
class GenesisMetadata$Type extends MessageType {
  constructor() {
    super("ibc.core.client.v1.GenesisMetadata", [
      {
        no: 1,
        name: "key",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 2,
        name: "value",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ], { "gogoproto.goproto_getters": false });
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.key = new Uint8Array(0);
    message.value = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* bytes key */
        1:
          message.key = reader.bytes();
          break;
        case /* bytes value */
        2:
          message.value = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.key.length)
      writer.tag(1, WireType.LengthDelimited).bytes(message.key);
    if (message.value.length)
      writer.tag(2, WireType.LengthDelimited).bytes(message.value);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const GenesisMetadata = new GenesisMetadata$Type();
class IdentifiedGenesisMetadata$Type extends MessageType {
  constructor() {
    super("ibc.core.client.v1.IdentifiedGenesisMetadata", [
      {
        no: 1,
        name: "client_id",
        kind: "scalar",
        T: 9
        /*ScalarType.STRING*/
      },
      { no: 2, name: "client_metadata", kind: "message", repeat: 2, T: () => GenesisMetadata, options: { "gogoproto.nullable": false } }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.clientId = "";
    message.clientMetadata = [];
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* string client_id */
        1:
          message.clientId = reader.string();
          break;
        case /* repeated ibc.core.client.v1.GenesisMetadata client_metadata */
        2:
          message.clientMetadata.push(GenesisMetadata.internalBinaryRead(reader, reader.uint32(), options));
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.clientId !== "")
      writer.tag(1, WireType.LengthDelimited).string(message.clientId);
    for (let i = 0; i < message.clientMetadata.length; i++)
      GenesisMetadata.internalBinaryWrite(message.clientMetadata[i], writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const IdentifiedGenesisMetadata = new IdentifiedGenesisMetadata$Type();
export {
  GenesisMetadata,
  GenesisState,
  IdentifiedGenesisMetadata
};
