// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/client/v1/query_pb.client.js
import { Query } from "./query_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class QueryClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Query.typeName;
    this.methods = Query.methods;
    this.options = Query.options;
  }
  /**
   * ClientState queries an IBC light client.
   *
   * @generated from protobuf rpc: ClientState
   */
  clientState(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ClientStates queries all the IBC light clients of a chain.
   *
   * @generated from protobuf rpc: ClientStates
   */
  clientStates(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ConsensusState queries a consensus state associated with a client state at
   * a given height.
   *
   * @generated from protobuf rpc: ConsensusState
   */
  consensusState(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ConsensusStates queries all the consensus state associated with a given
   * client.
   *
   * @generated from protobuf rpc: ConsensusStates
   */
  consensusStates(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ConsensusStateHeights queries the height of every consensus states associated with a given client.
   *
   * @generated from protobuf rpc: ConsensusStateHeights
   */
  consensusStateHeights(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * Status queries the status of an IBC client.
   *
   * @generated from protobuf rpc: ClientStatus
   */
  clientStatus(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * ClientParams queries all parameters of the ibc client submodule.
   *
   * @generated from protobuf rpc: ClientParams
   */
  clientParams(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpgradedClientState queries an Upgraded IBC light client.
   *
   * @generated from protobuf rpc: UpgradedClientState
   */
  upgradedClientState(input, options) {
    const method = this.methods[7], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpgradedConsensusState queries an Upgraded IBC consensus state.
   *
   * @generated from protobuf rpc: UpgradedConsensusState
   */
  upgradedConsensusState(input, options) {
    const method = this.methods[8], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * VerifyMembership queries an IBC light client for proof verification of a value at a given key path.
   *
   * @generated from protobuf rpc: VerifyMembership
   */
  verifyMembership(input, options) {
    const method = this.methods[9], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  QueryClient
};
