// @injectivelabs/core-proto-ts-v2/esm/generated/ibc/core/client/v1/tx_pb.client.js
import { Msg } from "./tx_pb.js";
import { stackIntercept } from "@protobuf-ts/runtime-rpc";
class MsgClient {
  constructor(_transport) {
    this._transport = _transport;
    this.typeName = Msg.typeName;
    this.methods = Msg.methods;
    this.options = Msg.options;
  }
  /**
   * CreateClient defines a rpc handler method for MsgCreateClient.
   *
   * @generated from protobuf rpc: CreateClient
   */
  createClient(input, options) {
    const method = this.methods[0], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateClient defines a rpc handler method for MsgUpdateClient.
   *
   * @generated from protobuf rpc: UpdateClient
   */
  updateClient(input, options) {
    const method = this.methods[1], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpgradeClient defines a rpc handler method for MsgUpgradeClient.
   *
   * @generated from protobuf rpc: UpgradeClient
   */
  upgradeClient(input, options) {
    const method = this.methods[2], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * SubmitMisbehaviour defines a rpc handler method for MsgSubmitMisbehaviour.
   *
   * @generated from protobuf rpc: SubmitMisbehaviour
   */
  submitMisbehaviour(input, options) {
    const method = this.methods[3], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * RecoverClient defines a rpc handler method for MsgRecoverClient.
   *
   * @generated from protobuf rpc: RecoverClient
   */
  recoverClient(input, options) {
    const method = this.methods[4], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * IBCSoftwareUpgrade defines a rpc handler method for MsgIBCSoftwareUpgrade.
   *
   * @generated from protobuf rpc: IBCSoftwareUpgrade
   */
  iBCSoftwareUpgrade(input, options) {
    const method = this.methods[5], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
  /**
   * UpdateClientParams defines a rpc handler method for MsgUpdateParams.
   *
   * @generated from protobuf rpc: UpdateClientParams
   */
  updateClientParams(input, options) {
    const method = this.methods[6], opt = this._transport.mergeOptions(options);
    return stackIntercept("unary", this._transport, method, opt, input);
  }
}
export {
  MsgClient
};
