// @injectivelabs/core-proto-ts-v2/esm/generated/tendermint/statesync/types_pb.js
import { WireType } from "@protobuf-ts/runtime";
import { UnknownFieldHandler } from "@protobuf-ts/runtime";
import { reflectionMergePartial } from "@protobuf-ts/runtime";
import { MessageType } from "@protobuf-ts/runtime";
class Message$Type extends MessageType {
  constructor() {
    super("tendermint.statesync.Message", [
      { no: 1, name: "snapshots_request", kind: "message", oneof: "sum", T: () => SnapshotsRequest },
      { no: 2, name: "snapshots_response", kind: "message", oneof: "sum", T: () => SnapshotsResponse },
      { no: 3, name: "chunk_request", kind: "message", oneof: "sum", T: () => ChunkRequest },
      { no: 4, name: "chunk_response", kind: "message", oneof: "sum", T: () => ChunkResponse }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.sum = { oneofKind: void 0 };
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* tendermint.statesync.SnapshotsRequest snapshots_request */
        1:
          message.sum = {
            oneofKind: "snapshotsRequest",
            snapshotsRequest: SnapshotsRequest.internalBinaryRead(reader, reader.uint32(), options, message.sum.snapshotsRequest)
          };
          break;
        case /* tendermint.statesync.SnapshotsResponse snapshots_response */
        2:
          message.sum = {
            oneofKind: "snapshotsResponse",
            snapshotsResponse: SnapshotsResponse.internalBinaryRead(reader, reader.uint32(), options, message.sum.snapshotsResponse)
          };
          break;
        case /* tendermint.statesync.ChunkRequest chunk_request */
        3:
          message.sum = {
            oneofKind: "chunkRequest",
            chunkRequest: ChunkRequest.internalBinaryRead(reader, reader.uint32(), options, message.sum.chunkRequest)
          };
          break;
        case /* tendermint.statesync.ChunkResponse chunk_response */
        4:
          message.sum = {
            oneofKind: "chunkResponse",
            chunkResponse: ChunkResponse.internalBinaryRead(reader, reader.uint32(), options, message.sum.chunkResponse)
          };
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.sum.oneofKind === "snapshotsRequest")
      SnapshotsRequest.internalBinaryWrite(message.sum.snapshotsRequest, writer.tag(1, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "snapshotsResponse")
      SnapshotsResponse.internalBinaryWrite(message.sum.snapshotsResponse, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "chunkRequest")
      ChunkRequest.internalBinaryWrite(message.sum.chunkRequest, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
    if (message.sum.oneofKind === "chunkResponse")
      ChunkResponse.internalBinaryWrite(message.sum.chunkResponse, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const Message = new Message$Type();
class SnapshotsRequest$Type extends MessageType {
  constructor() {
    super("tendermint.statesync.SnapshotsRequest", []);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SnapshotsRequest = new SnapshotsRequest$Type();
class SnapshotsResponse$Type extends MessageType {
  constructor() {
    super("tendermint.statesync.SnapshotsResponse", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "format",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "chunks",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 4,
        name: "hash",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "metadata",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    message.format = 0;
    message.chunks = 0;
    message.hash = new Uint8Array(0);
    message.metadata = new Uint8Array(0);
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 height */
        1:
          message.height = reader.uint64().toBigInt();
          break;
        case /* uint32 format */
        2:
          message.format = reader.uint32();
          break;
        case /* uint32 chunks */
        3:
          message.chunks = reader.uint32();
          break;
        case /* bytes hash */
        4:
          message.hash = reader.bytes();
          break;
        case /* bytes metadata */
        5:
          message.metadata = reader.bytes();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.height);
    if (message.format !== 0)
      writer.tag(2, WireType.Varint).uint32(message.format);
    if (message.chunks !== 0)
      writer.tag(3, WireType.Varint).uint32(message.chunks);
    if (message.hash.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.hash);
    if (message.metadata.length)
      writer.tag(5, WireType.LengthDelimited).bytes(message.metadata);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const SnapshotsResponse = new SnapshotsResponse$Type();
class ChunkRequest$Type extends MessageType {
  constructor() {
    super("tendermint.statesync.ChunkRequest", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "format",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "index",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    message.format = 0;
    message.index = 0;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 height */
        1:
          message.height = reader.uint64().toBigInt();
          break;
        case /* uint32 format */
        2:
          message.format = reader.uint32();
          break;
        case /* uint32 index */
        3:
          message.index = reader.uint32();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.height);
    if (message.format !== 0)
      writer.tag(2, WireType.Varint).uint32(message.format);
    if (message.index !== 0)
      writer.tag(3, WireType.Varint).uint32(message.index);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ChunkRequest = new ChunkRequest$Type();
class ChunkResponse$Type extends MessageType {
  constructor() {
    super("tendermint.statesync.ChunkResponse", [
      {
        no: 1,
        name: "height",
        kind: "scalar",
        T: 4,
        L: 0
        /*LongType.BIGINT*/
      },
      {
        no: 2,
        name: "format",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 3,
        name: "index",
        kind: "scalar",
        T: 13
        /*ScalarType.UINT32*/
      },
      {
        no: 4,
        name: "chunk",
        kind: "scalar",
        T: 12
        /*ScalarType.BYTES*/
      },
      {
        no: 5,
        name: "missing",
        kind: "scalar",
        T: 8
        /*ScalarType.BOOL*/
      }
    ]);
  }
  create(value) {
    const message = globalThis.Object.create(this.messagePrototype);
    message.height = 0n;
    message.format = 0;
    message.index = 0;
    message.chunk = new Uint8Array(0);
    message.missing = false;
    if (value !== void 0)
      reflectionMergePartial(this, message, value);
    return message;
  }
  internalBinaryRead(reader, length, options, target) {
    let message = target ?? this.create(), end = reader.pos + length;
    while (reader.pos < end) {
      let [fieldNo, wireType] = reader.tag();
      switch (fieldNo) {
        case /* uint64 height */
        1:
          message.height = reader.uint64().toBigInt();
          break;
        case /* uint32 format */
        2:
          message.format = reader.uint32();
          break;
        case /* uint32 index */
        3:
          message.index = reader.uint32();
          break;
        case /* bytes chunk */
        4:
          message.chunk = reader.bytes();
          break;
        case /* bool missing */
        5:
          message.missing = reader.bool();
          break;
        default:
          let u = options.readUnknownField;
          if (u === "throw")
            throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
          let d = reader.skip(wireType);
          if (u !== false)
            (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
      }
    }
    return message;
  }
  internalBinaryWrite(message, writer, options) {
    if (message.height !== 0n)
      writer.tag(1, WireType.Varint).uint64(message.height);
    if (message.format !== 0)
      writer.tag(2, WireType.Varint).uint32(message.format);
    if (message.index !== 0)
      writer.tag(3, WireType.Varint).uint32(message.index);
    if (message.chunk.length)
      writer.tag(4, WireType.LengthDelimited).bytes(message.chunk);
    if (message.missing !== false)
      writer.tag(5, WireType.Varint).bool(message.missing);
    let u = options.writeUnknownFields;
    if (u !== false)
      (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
    return writer;
  }
}
const ChunkResponse = new ChunkResponse$Type();
export {
  ChunkRequest,
  ChunkResponse,
  Message,
  SnapshotsRequest,
  SnapshotsResponse
};
