// @injectivelabs/sdk-ts/dist/cjs/core/modules.cjs
require('../utils-D3XFGv8y.cjs');
require('../defineProperty-dyr5BsvG.cjs');
const require_accounts = require('../accounts-HFzAaFRB.cjs');
require('../grpc-FG2z-ICE.cjs');
require('../MsgBase-D_1CCpXQ.cjs');
const require_MsgSetDenomMetadata = require('../MsgSetDenomMetadata-DzYvKtNj.cjs');
require('../ChainRestTendermintApi-CyeoAmer.cjs');
require('../BaseGrpcConsumer-Db3sCC6b.cjs');
require('../ChainGrpcExchangeApi-Xtflr7xp.cjs');
require('../BaseIndexerGrpcConsumer-BBXxg_BR.cjs');
require('../BaseRestConsumer-CRkyjiSo.cjs');
const require_ExecArgNeptuneWithdraw = require('../ExecArgNeptuneWithdraw-4esQ6Su-.cjs');
require('../IndexerGrpcWeb3GwApi-BxJoq1-k.cjs');
require('../tx-MgHAjEie.cjs');

exports.ContractExecutionAuthz = require_MsgSetDenomMetadata.ContractExecutionAuthorization;
exports.ContractExecutionCompatAuthz = require_MsgSetDenomMetadata.ContractExecutionCompatAuthorization;
exports.ExecArgCW20AdapterRedeemAndTransfer = require_MsgSetDenomMetadata.ExecArgCW20AdapterRedeemAndTransfer;
exports.ExecArgCW20Send = require_MsgSetDenomMetadata.ExecArgCW20Send;
exports.ExecArgCW20Transfer = require_MsgSetDenomMetadata.ExecArgCW20Transfer;
exports.ExecArgCreateCampaign = require_MsgSetDenomMetadata.ExecArgCreateCampaign;
exports.ExecArgCreatePerpGridStrategy = require_MsgSetDenomMetadata.ExecArgCreatePerpGridStrategy;
exports.ExecArgCreateRound = require_MsgSetDenomMetadata.ExecArgCreateRound;
exports.ExecArgCreateSpotGridStrategy = require_MsgSetDenomMetadata.ExecArgCreateSpotGridStrategy;
exports.ExecArgDepositTokens = require_MsgSetDenomMetadata.ExecArgDepositTokens;
exports.ExecArgFundCampaign = require_MsgSetDenomMetadata.ExecArgFundCampaign;
exports.ExecArgIncreaseAllowance = require_MsgSetDenomMetadata.ExecArgIncreaseAllowance;
exports.ExecArgInitiateTransfer = require_MsgSetDenomMetadata.ExecArgInitiateTransfer;
exports.ExecArgNeptuneDeposit = require_ExecArgNeptuneWithdraw.ExecArgNeptuneDeposit;
exports.ExecArgNeptuneWithdraw = require_ExecArgNeptuneWithdraw.ExecArgNeptuneWithdraw;
exports.ExecArgRemoveGridStrategy = require_MsgSetDenomMetadata.ExecArgRemoveGridStrategy;
exports.ExecArgSubmitVaa = require_MsgSetDenomMetadata.ExecArgSubmitVaa;
exports.ExecArgSwapExactOutput = require_MsgSetDenomMetadata.ExecArgSwapExactOutput;
exports.ExecArgSwapMinOutput = require_MsgSetDenomMetadata.ExecArgSwapMinOutput;
exports.ExecArgUpdateGridConfig = require_MsgSetDenomMetadata.ExecArgUpdateGridConfig;
exports.ExecPrivilegedArgOffChainVaultRedeem = require_MsgSetDenomMetadata.ExecPrivilegedArgOffChainVaultRedeem;
exports.ExecPrivilegedArgOffChainVaultSubscribe = require_MsgSetDenomMetadata.ExecPrivilegedArgOffChainVaultSubscribe;
exports.ExecPrivilegedArgVaultRedeem = require_MsgSetDenomMetadata.ExecPrivilegedArgVaultRedeem;
exports.ExecPrivilegedArgVaultSubscribe = require_MsgSetDenomMetadata.ExecPrivilegedArgVaultSubscribe;
exports.ExitType = require_MsgSetDenomMetadata.ExitType;
exports.GenericAuthz = require_MsgSetDenomMetadata.GenericAuthorization;
exports.GrantAuthorizationType = require_MsgSetDenomMetadata.GrantAuthorizationType;
exports.MsgAdminUpdateBinaryOptionsMarket = require_MsgSetDenomMetadata.MsgAdminUpdateBinaryOptionsMarket;
exports.MsgAuthorizeStakeGrants = require_MsgSetDenomMetadata.MsgAuthorizeStakeGrants;
exports.MsgAuthzExec = require_MsgSetDenomMetadata.MsgExec;
exports.MsgBatchCancelBinaryOptionsOrders = require_MsgSetDenomMetadata.MsgBatchCancelBinaryOptionsOrders;
exports.MsgBatchCancelDerivativeOrders = require_MsgSetDenomMetadata.MsgBatchCancelDerivativeOrders;
exports.MsgBatchCancelSpotOrders = require_MsgSetDenomMetadata.MsgBatchCancelSpotOrders;
exports.MsgBatchUpdateOrders = require_MsgSetDenomMetadata.MsgBatchUpdateOrders;
exports.MsgBeginRedelegate = require_MsgSetDenomMetadata.MsgBeginRedelegate;
exports.MsgBid = require_MsgSetDenomMetadata.MsgBid;
exports.MsgBurn = require_MsgSetDenomMetadata.MsgBurn;
exports.MsgCancelBinaryOptionsOrder = require_MsgSetDenomMetadata.MsgCancelBinaryOptionsOrder;
exports.MsgCancelDerivativeOrder = require_MsgSetDenomMetadata.MsgCancelDerivativeOrder;
exports.MsgCancelPostOnlyModeV2 = require_MsgSetDenomMetadata.MsgCancelPostOnlyModeV2;
exports.MsgCancelSpotOrder = require_MsgSetDenomMetadata.MsgCancelSpotOrder;
exports.MsgCancelUnbondingDelegation = require_MsgSetDenomMetadata.MsgCancelUnbondingDelegation;
exports.MsgChangeAdmin = require_MsgSetDenomMetadata.MsgChangeAdmin;
exports.MsgClaimVoucher = require_MsgSetDenomMetadata.MsgClaimVoucher;
exports.MsgCreateBinaryOptionsLimitOrder = require_MsgSetDenomMetadata.MsgCreateBinaryOptionsLimitOrder;
exports.MsgCreateBinaryOptionsMarketOrder = require_MsgSetDenomMetadata.MsgCreateBinaryOptionsMarketOrder;
exports.MsgCreateDenom = require_MsgSetDenomMetadata.MsgCreateDenom;
exports.MsgCreateDerivativeLimitOrder = require_MsgSetDenomMetadata.MsgCreateDerivativeLimitOrder;
exports.MsgCreateDerivativeMarketOrder = require_MsgSetDenomMetadata.MsgCreateDerivativeMarketOrder;
exports.MsgCreateInsuranceFund = require_MsgSetDenomMetadata.MsgCreateInsuranceFund;
exports.MsgCreateNamespace = require_MsgSetDenomMetadata.MsgCreateNamespace;
exports.MsgCreateSpotLimitOrder = require_MsgSetDenomMetadata.MsgCreateSpotLimitOrder;
exports.MsgCreateSpotMarketOrder = require_MsgSetDenomMetadata.MsgCreateSpotMarketOrder;
exports.MsgCreateValidator = require_MsgSetDenomMetadata.MsgCreateValidator;
exports.MsgDelegate = require_MsgSetDenomMetadata.MsgDelegate;
exports.MsgDeposit = require_MsgSetDenomMetadata.MsgDeposit$1;
exports.MsgEditValidator = require_MsgSetDenomMetadata.MsgEditValidator;
exports.MsgExecuteContract = require_MsgSetDenomMetadata.MsgExecuteContract;
exports.MsgExecuteContractCompat = require_ExecArgNeptuneWithdraw.MsgExecuteContractCompat;
exports.MsgExternalTransfer = require_MsgSetDenomMetadata.MsgExternalTransfer;
exports.MsgFundCommunityPool = require_MsgSetDenomMetadata.MsgFundCommunityPool;
exports.MsgGovDeposit = require_MsgSetDenomMetadata.MsgDeposit;
exports.MsgGrant = require_MsgSetDenomMetadata.MsgGrant;
exports.MsgGrantAllowance = require_MsgSetDenomMetadata.MsgGrantAllowance;
exports.MsgGrantWithAuthorization = require_MsgSetDenomMetadata.MsgGrantWithAuthorization;
exports.MsgIncreasePositionMargin = require_accounts.MsgIncreasePositionMargin;
exports.MsgInstantBinaryOptionsMarketLaunch = require_MsgSetDenomMetadata.MsgInstantBinaryOptionsMarketLaunch;
exports.MsgInstantSpotMarketLaunch = require_MsgSetDenomMetadata.MsgInstantSpotMarketLaunch;
exports.MsgInstantiateContract = require_MsgSetDenomMetadata.MsgInstantiateContract;
exports.MsgLiquidatePosition = require_MsgSetDenomMetadata.MsgLiquidatePosition;
exports.MsgMigrateContract = require_MsgSetDenomMetadata.MsgMigrateContract;
exports.MsgMint = require_MsgSetDenomMetadata.MsgMint;
exports.MsgMultiSend = require_MsgSetDenomMetadata.MsgMultiSend;
exports.MsgPrivilegedExecuteContract = require_MsgSetDenomMetadata.MsgPrivilegedExecuteContract;
exports.MsgReclaimLockedFunds = require_MsgSetDenomMetadata.MsgReclaimLockedFunds;
exports.MsgRequestRedemption = require_MsgSetDenomMetadata.MsgRequestRedemption;
exports.MsgRevoke = require_MsgSetDenomMetadata.MsgRevoke;
exports.MsgRevokeAllowance = require_MsgSetDenomMetadata.MsgRevokeAllowance;
exports.MsgRewardsOptOut = require_MsgSetDenomMetadata.MsgRewardsOptOut;
exports.MsgSend = require_MsgSetDenomMetadata.MsgSend;
exports.MsgSendToEth = require_MsgSetDenomMetadata.MsgSendToEth;
exports.MsgSetDelegationTransferReceivers = require_MsgSetDenomMetadata.MsgSetDelegationTransferReceivers;
exports.MsgSetDenomMetadata = require_MsgSetDenomMetadata.MsgSetDenomMetadata;
exports.MsgSignData = require_accounts.MsgSignData;
exports.MsgStoreCode = require_MsgSetDenomMetadata.MsgStoreCode;
exports.MsgSubmitGenericProposal = require_MsgSetDenomMetadata.MsgSubmitGenericProposal;
exports.MsgSubmitProposalExpiryFuturesMarketLaunch = require_MsgSetDenomMetadata.MsgSubmitProposalExpiryFuturesMarketLaunch;
exports.MsgSubmitProposalPerpetualMarketLaunch = require_MsgSetDenomMetadata.MsgSubmitProposalPerpetualMarketLaunch;
exports.MsgSubmitProposalPerpetualMarketLaunchV2 = require_MsgSetDenomMetadata.MsgSubmitProposalPerpetualMarketLaunchV2;
exports.MsgSubmitProposalSpotMarketLaunch = require_MsgSetDenomMetadata.MsgSubmitProposalSpotMarketLaunch;
exports.MsgSubmitProposalSpotMarketParamUpdate = require_MsgSetDenomMetadata.MsgSubmitProposalSpotMarketParamUpdate;
exports.MsgSubmitTextProposal = require_MsgSetDenomMetadata.MsgSubmitTextProposal;
exports.MsgTransfer = require_MsgSetDenomMetadata.MsgTransfer;
exports.MsgTransferDelegation = require_MsgSetDenomMetadata.MsgTransferDelegation;
exports.MsgUndelegate = require_MsgSetDenomMetadata.MsgUndelegate;
exports.MsgUnderwrite = require_MsgSetDenomMetadata.MsgUnderwrite;
exports.MsgUpdateActorRoles = require_MsgSetDenomMetadata.MsgUpdateActorRoles;
exports.MsgUpdateAdmin = require_MsgSetDenomMetadata.MsgUpdateAdmin;
exports.MsgUpdateDerivativeMarketV2 = require_MsgSetDenomMetadata.MsgUpdateDerivativeMarketV2;
exports.MsgUpdateNamespace = require_MsgSetDenomMetadata.MsgUpdateNamespace;
exports.MsgUpdateParams = require_MsgSetDenomMetadata.MsgUpdateParams;
exports.MsgUpdateSpotMarketV2 = require_MsgSetDenomMetadata.MsgUpdateSpotMarketV2;
exports.MsgVote = require_MsgSetDenomMetadata.MsgVote;
exports.MsgWithdraw = require_MsgSetDenomMetadata.MsgWithdraw;
exports.MsgWithdrawDelegatorReward = require_MsgSetDenomMetadata.MsgWithdrawDelegatorReward;
exports.MsgWithdrawValidatorCommission = require_MsgSetDenomMetadata.MsgWithdrawValidatorCommission;
exports.OrderHashManager = require_MsgSetDenomMetadata.OrderHashManager;
exports.ProposalDecomposer = require_MsgSetDenomMetadata.ProposalDecomposer;
exports.StrategyType = require_MsgSetDenomMetadata.StrategyType;
exports.getGenericAuthorizationFromMessageType = require_MsgSetDenomMetadata.getGenericAuthorizationFromMessageType;
exports.msgsOrMsgExecMsgs = require_MsgSetDenomMetadata.msgsOrMsgExecMsgs;
